diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.cpp
@@ -31,17 +31,17 @@
 #include "EventNames.h"
 #include "PaymentRequest.h"
 
 namespace WebCore {
 
-PaymentRequestUpdateEvent::PaymentRequestUpdateEvent(const AtomicString& type, const PaymentRequestUpdateEventInit& eventInit)
+PaymentRequestUpdateEvent::PaymentRequestUpdateEvent(const AtomString& type, const PaymentRequestUpdateEventInit& eventInit)
     : Event { type, eventInit, IsTrusted::No }
 {
     ASSERT(!isTrusted());
 }
 
-PaymentRequestUpdateEvent::PaymentRequestUpdateEvent(const AtomicString& type)
+PaymentRequestUpdateEvent::PaymentRequestUpdateEvent(const AtomString& type)
     : Event { type, CanBubble::No, IsCancelable::No }
 {
     ASSERT(isTrusted());
 }
 
