diff a/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/Makefile.am b/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/Makefile.am
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/Makefile.am
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/Makefile.am
@@ -36,10 +36,11 @@
 ## Use sed and then mv to avoid problems if the user interrupts.
 	sed -e 's?\@XSLT_LIBDIR\@?$(XSLT_LIBDIR)?g' \
 	    -e 's?\@XSLT_INCLUDEDIR\@?$(XSLT_INCLUDEDIR)?g' \
 	    -e 's?\@VERSION\@?$(VERSION)?g' \
 	    -e 's?\@XSLT_LIBS\@?$(XSLT_LIBS) $(EXTRA_LIBS)?g' \
+	    -e 's?\@XSLT_PRIVATE_LIBS\@?$(XSLT_PRIVATE_LIBS)?g' \
 	       < $(srcdir)/xsltConf.sh.in > xsltConf.tmp \
 	&& mv xsltConf.tmp xsltConf.sh
 
 CLEANFILES = xsltConf.sh
 
