diff a/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java b/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
--- a/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
+++ b/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
@@ -41,11 +41,11 @@
      * Public API                                                          *
      *                                                                     *
      **********************************************************************/
 
     /**
-     * Creates a default {@code TableFocusModel}.
+     * Constructor for subclasses to call.
      */
     public TableFocusModel() {
     }
 
     /**
