diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/filters/SourceAlpha.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/filters/SourceAlpha.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/filters/SourceAlpha.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/filters/SourceAlpha.cpp
@@ -33,13 +33,13 @@
 Ref<SourceAlpha> SourceAlpha::create(FilterEffect& sourceEffect)
 {
     return adoptRef(*new SourceAlpha(sourceEffect));
 }
 
-const AtomicString& SourceAlpha::effectName()
+const AtomString& SourceAlpha::effectName()
 {
-    static NeverDestroyed<const AtomicString> s_effectName("SourceAlpha", AtomicString::ConstructFromLiteral);
+    static NeverDestroyed<const AtomString> s_effectName("SourceAlpha", AtomString::ConstructFromLiteral);
     return s_effectName;
 }
 
 void SourceAlpha::determineAbsolutePaintRect()
 {
