diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/GridArea.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/GridArea.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/GridArea.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/GridArea.h
@@ -146,11 +146,11 @@
     enum GridSpanType {UntranslatedDefinite, TranslatedDefinite, Indefinite};
 
     GridSpan(int startLine, int endLine, GridSpanType type)
         : m_type(type)
     {
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
         ASSERT(startLine < endLine);
         if (type == TranslatedDefinite) {
             ASSERT(startLine >= 0);
             ASSERT(endLine > 0);
         }
