diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleBackgroundData.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleBackgroundData.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleBackgroundData.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleBackgroundData.cpp
@@ -58,6 +58,22 @@
     if (!outline.isVisible() && !other.outline.isVisible())
         return true;
     return outline == other.outline;
 }
 
+void StyleBackgroundData::dump(TextStream& ts, DumpStyleValues behavior) const
+{
+    if (behavior == DumpStyleValues::All || background != FillLayer(FillLayerType::Background))
+        ts.dumpProperty("background-image", background);
+    if (behavior == DumpStyleValues::All || color != RenderStyle::initialBackgroundColor())
+        ts.dumpProperty("background-color", color);
+    if (behavior == DumpStyleValues::All || outline != OutlineValue())
+        ts.dumpProperty("outline", outline);
+}
+
+TextStream& operator<<(TextStream& ts, const StyleBackgroundData& backgroundData)
+{
+    backgroundData.dump(ts);
+    return ts;
+}
+
 } // namespace WebCore
