diff a/src/java.base/share/classes/java/nio/channels/ReadableByteChannel.java b/src/java.base/share/classes/java/nio/channels/ReadableByteChannel.java
--- a/src/java.base/share/classes/java/nio/channels/ReadableByteChannel.java
+++ b/src/java.base/share/classes/java/nio/channels/ReadableByteChannel.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, 2001, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -82,10 +82,13 @@
      *         The buffer into which bytes are to be transferred
      *
      * @return  The number of bytes read, possibly zero, or {@code -1} if the
      *          channel has reached end-of-stream
      *
+     * @throws  IllegalArgumentException
+     *          If the buffer is read-only
+     *
      * @throws  NonReadableChannelException
      *          If this channel was not opened for reading
      *
      * @throws  ClosedChannelException
      *          If this channel is closed
