diff a/src/hotspot/share/utilities/globalDefinitions_solstudio.hpp b/src/hotspot/share/utilities/globalDefinitions_solstudio.hpp
--- a/src/hotspot/share/utilities/globalDefinitions_solstudio.hpp
+++ b/src/hotspot/share/utilities/globalDefinitions_solstudio.hpp
@@ -243,13 +243,11 @@
 #define FORMAT64_MODIFIER "ll"
 #endif // _LP64
 
 #define offset_of(klass,field) offsetof(klass,field)
 
-#ifndef USE_LIBRARY_BASED_TLS_ONLY
-#define THREAD_LOCAL_DECL __thread
-#endif
+#define THREAD_LOCAL __thread
 
 // Inlining support
 #define NOINLINE
 #define ALWAYSINLINE inline __attribute__((always_inline))
 
