diff a/src/hotspot/share/jvmci/jvmciCodeInstaller.cpp b/src/hotspot/share/jvmci/jvmciCodeInstaller.cpp
--- a/src/hotspot/share/jvmci/jvmciCodeInstaller.cpp
+++ b/src/hotspot/share/jvmci/jvmciCodeInstaller.cpp
@@ -1335,12 +1335,12 @@
       case UNVERIFIED_ENTRY:
         _offsets.set_value(CodeOffsets::Entry, pc_offset);
         break;
       case VERIFIED_ENTRY:
         _offsets.set_value(CodeOffsets::Verified_Entry, pc_offset);
-        _offsets.set_value(CodeOffsets::Verified_Value_Entry, pc_offset);
-        _offsets.set_value(CodeOffsets::Verified_Value_Entry_RO, pc_offset);
+        _offsets.set_value(CodeOffsets::Verified_Inline_Entry, pc_offset);
+        _offsets.set_value(CodeOffsets::Verified_Inline_Entry_RO, pc_offset);
         break;
       case OSR_ENTRY:
         _offsets.set_value(CodeOffsets::OSR_Entry, pc_offset);
         break;
       case EXCEPTION_HANDLER_ENTRY:
