diff a/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/PrettyPrinter.java b/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/PrettyPrinter.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/PrettyPrinter.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/PrettyPrinter.java
@@ -128,11 +128,11 @@
     }
 
     private static Type.Visitor<String, Void> typeVisitor = new Type.Visitor<>() {
         @Override
         public String visitPrimitive(Type.Primitive t, Void aVoid) {
-            return t.kind().toString() + t.layout().map(l -> "(layout = " + l + ")").orElse("");
+            return t.kind().toString() + t.kind().layout().map(l -> "(layout = " + l + ")").orElse("");
         }
 
         @Override
         public String visitDelegated(Type.Delegated t, Void aVoid) {
             switch (t.kind()) {
