diff a/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MarkdownToTextTests.java b/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MarkdownToTextTests.java
--- a/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MarkdownToTextTests.java
+++ b/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MarkdownToTextTests.java
@@ -53,6 +53,16 @@
 
     @Test
     void suggestion() {
         assertEquals("Suggestion:\n\nJust some text", MarkdownToText.removeFormatting("```suggestion\nJust some text\n```"));
     }
+
+    @Test
+    void escapes() {
+        assertEquals("Special chars: #$%&'()*+\\!", MarkdownToText.removeFormatting("Special chars: \\#\\$\\%\\&\\'\\(\\)\\*\\+\\\\!"));
+    }
+
+    @Test
+    void entities() {
+        assertEquals("space is here", MarkdownToText.removeFormatting("space&#32;is here"));
+    }
 }
