diff a/test/jdk/java/net/URLConnection/HttpContinueStackOverflow.java b/test/jdk/java/net/URLConnection/HttpContinueStackOverflow.java
--- a/test/jdk/java/net/URLConnection/HttpContinueStackOverflow.java
+++ b/test/jdk/java/net/URLConnection/HttpContinueStackOverflow.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, 2010, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -21,32 +21,35 @@
  * questions.
  */
 
 /* @test
  * @bug 4258697
+ * @library /test/lib
  * @summary Make sure that http CONTINUE status followed by invalid
  * response doesn't cause HttpClient to recursively loop and
  * eventually StackOverflow.
  *
  */
 import java.io.BufferedReader;
 import java.io.InputStreamReader;
 import java.io.IOException;
 import java.io.PrintStream;
+import java.net.InetAddress;
 import java.net.ServerSocket;
 import java.net.Socket;
 import java.net.URL;
 import java.net.HttpURLConnection;
+import jdk.test.lib.net.URIBuilder;
+import static java.net.Proxy.NO_PROXY;
 
 public class HttpContinueStackOverflow {
 
     static class Server implements Runnable {
-        int port;
         ServerSocket serverSock ;
 
         Server() throws IOException {
-            serverSock = new ServerSocket(0);
+            serverSock = new ServerSocket(0, 0, InetAddress.getLoopbackAddress());
         }
 
         int getLocalPort() {
             return serverSock.getLocalPort();
         }
@@ -80,13 +83,19 @@
     HttpContinueStackOverflow() throws Exception {
         /* create the server */
         Server s = new Server();
         (new Thread(s)).start();
 
-        /* connect to server, connect to server and get response code */
-        URL url = new URL("http", "localhost", s.getLocalPort(), "anything.html");
-        HttpURLConnection conn = (HttpURLConnection)url.openConnection();
+        /* connect to server and get response code */
+        URL url = URIBuilder.newBuilder()
+                .scheme("http")
+                .loopback()
+                .port(s.getLocalPort())
+                .path("/anything.html")
+                .toURL();
+
+        HttpURLConnection conn = (HttpURLConnection)url.openConnection(NO_PROXY);
         conn.getResponseCode();
         System.out.println("TEST PASSED");
     }
 
     public static void main(String args[]) throws Exception {
