diff a/webrev/src/main/java/org/openjdk/skara/webrev/HunkCoalescer.java b/webrev/src/main/java/org/openjdk/skara/webrev/HunkCoalescer.java
--- a/webrev/src/main/java/org/openjdk/skara/webrev/HunkCoalescer.java
+++ b/webrev/src/main/java/org/openjdk/skara/webrev/HunkCoalescer.java
@@ -254,11 +254,12 @@
         }
         var destAfterContextCount = destAfterContextEnd - destAfterContextStart;
 
         var afterContextCount = Math.min(sourceAfterContextCount, destAfterContextCount);
 
-        var sourceLineNumStart = hunk.source().lines().isEmpty() && hunk.source().range().start() == 0 ? sourceAfterContextStart + 1 : sourceAfterContextStart;
+        var sourceLineNumStart = hunk.source().lines().isEmpty() && hunk.source().range().start() == 0 ?
+            sourceAfterContextStart + 1 : sourceAfterContextStart;
         var sourceEndingLineNum = sourceLineNumStart + afterContextCount;
         var sourceContextAfter = new ArrayList<Line>();
         for (var lineNum = sourceLineNumStart; lineNum < sourceEndingLineNum; lineNum++) {
             var text = sourceContent.get(lineNum - 1);
             sourceContextAfter.add(new Line(lineNum, text));
@@ -262,13 +263,15 @@
         for (var lineNum = sourceLineNumStart; lineNum < sourceEndingLineNum; lineNum++) {
             var text = sourceContent.get(lineNum - 1);
             sourceContextAfter.add(new Line(lineNum, text));
         }
 
-        var destEndingLineNum = destAfterContextStart + afterContextCount;
+        var destLineNumStart = hunk.target().lines().isEmpty() && hunk.target().range().start() == 0 ?
+            destAfterContextStart + 1 : destAfterContextStart;
+        var destEndingLineNum = destLineNumStart + afterContextCount;
         var destContextAfter = new ArrayList<Line>();
-        for (var lineNum = destAfterContextStart; lineNum < destEndingLineNum; lineNum++) {
+        for (var lineNum = destLineNumStart; lineNum < destEndingLineNum; lineNum++) {
             var text = destContent.get(lineNum - 1);
             destContextAfter.add(new Line(lineNum, text));
         }
 
         return new Context(sourceContextAfter, destContextAfter);
