diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/PCToCodeOriginMap.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/PCToCodeOriginMap.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/PCToCodeOriginMap.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/PCToCodeOriginMap.h
@@ -28,11 +28,10 @@
 #if ENABLE(JIT)
 
 #include "CodeOrigin.h"
 #include "MacroAssembler.h"
 #include "VM.h"
-#include <wtf/Optional.h>
 #include <wtf/Vector.h>
 
 namespace JSC {
 
 #if ENABLE(FTL_JIT)
@@ -43,10 +42,11 @@
 
 class LinkBuffer;
 class PCToCodeOriginMapBuilder;
 
 class PCToCodeOriginMapBuilder {
+    WTF_MAKE_FAST_ALLOCATED;
     WTF_MAKE_NONCOPYABLE(PCToCodeOriginMapBuilder);
     friend class PCToCodeOriginMap;
 
 public:
     PCToCodeOriginMapBuilder(VM&);
@@ -74,10 +74,11 @@
     bool m_shouldBuildMapping;
 };
 
 // FIXME: <rdar://problem/39436658>
 class PCToCodeOriginMap {
+    WTF_MAKE_FAST_ALLOCATED;
     WTF_MAKE_NONCOPYABLE(PCToCodeOriginMap);
 public:
     PCToCodeOriginMap(PCToCodeOriginMapBuilder&&, LinkBuffer&);
     ~PCToCodeOriginMap();
 
