diff a/test/langtools/tools/javac/classfiles/ClassVersionChecker.java b/test/langtools/tools/javac/classfiles/ClassVersionChecker.java
--- a/test/langtools/tools/javac/classfiles/ClassVersionChecker.java
+++ b/test/langtools/tools/javac/classfiles/ClassVersionChecker.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2012, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2012, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -46,11 +46,13 @@
         EIGHT("8", 52),
         NINE("9", 53),
         TEN("10", 54),
         ELEVEN("11", 55),
         TWELVE("12", 56),
-        THIRTEEN("13", 57);
+        THIRTEEN("13", 57),
+        FOURTEEN("14", 58),
+        FIFTEEN("15", 59);
 
         private Version(String release, int classFileVer) {
             this.release = release;
             this.classFileVer = classFileVer;
         }
