diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLJITCode.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLJITCode.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLJITCode.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLJITCode.cpp
@@ -129,10 +129,18 @@
 DFG::CommonData* JITCode::dfgCommon()
 {
     return &common;
 }
 
+void JITCode::shrinkToFit(const ConcurrentJSLocker&)
+{
+    common.shrinkToFit();
+    osrExit.shrinkToFit();
+    osrExitDescriptors.shrinkToFit();
+    lazySlowPaths.shrinkToFit();
+}
+
 void JITCode::validateReferences(const TrackedReferences& trackedReferences)
 {
     common.validateReferences(trackedReferences);
 
     for (OSRExit& exit : osrExit)
