diff a/modules/javafx.graphics/src/main/native-glass/gtk_new/GlassCommonDialogs.cpp b/modules/javafx.graphics/src/main/native-glass/gtk_new/GlassCommonDialogs.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk_new/GlassCommonDialogs.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk_new/GlassCommonDialogs.cpp
@@ -112,13 +112,13 @@
         return create_empty_result();
     }
 
     GtkWidget* chooser = gtk_file_chooser_dialog_new(chooser_title, gdk_window_handle_to_gtk(parent),
             static_cast<GtkFileChooserAction>(chooser_type),
-            GTK_STOCK_CANCEL,
+            "_Cancel",
             GTK_RESPONSE_CANCEL,
-            (chooser_type == GTK_FILE_CHOOSER_ACTION_OPEN ? GTK_STOCK_OPEN : GTK_STOCK_SAVE),
+            (chooser_type == GTK_FILE_CHOOSER_ACTION_OPEN ? "_Open" : "_Save"),
             GTK_RESPONSE_ACCEPT,
             NULL);
 
     if (chooser_type == GTK_FILE_CHOOSER_ACTION_SAVE) {
         gtk_file_chooser_set_current_name(GTK_FILE_CHOOSER(chooser), chooser_filename);
@@ -199,13 +199,13 @@
 
     GtkWidget* chooser = gtk_file_chooser_dialog_new(
             chooser_title,
             gdk_window_handle_to_gtk(parent),
             GTK_FILE_CHOOSER_ACTION_SELECT_FOLDER,
-            GTK_STOCK_CANCEL,
+            "_Cancel",
             GTK_RESPONSE_CANCEL,
-            GTK_STOCK_OPEN,
+            "_Open",
             GTK_RESPONSE_ACCEPT,
             NULL);
 
     if (chooser_folder != NULL) {
         gtk_file_chooser_set_current_folder(GTK_FILE_CHOOSER(chooser),
