diff a/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java b/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java
--- a/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java
+++ b/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java
@@ -202,10 +202,16 @@
     /**
      * The title of the displayed file dialog.
      */
     private StringProperty title;
 
+    /**
+     * Creates a {@code FileChooser}.
+     */
+    public FileChooser() {
+    }
+
     public final void setTitle(final String value) {
         titleProperty().set(value);
     }
 
     public final String getTitle() {
