diff a/modules/javafx.web/src/main/native/Source/WebCore/crypto/gcrypt/GCryptUtilities.cpp b/modules/javafx.web/src/main/native/Source/WebCore/crypto/gcrypt/GCryptUtilities.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/crypto/gcrypt/GCryptUtilities.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/crypto/gcrypt/GCryptUtilities.cpp
@@ -26,10 +26,12 @@
  */
 
 #include "config.h"
 #include "GCryptUtilities.h"
 
+#include <wtf/Optional.h>
+
 namespace WebCore {
 
 Optional<const char*> hashAlgorithmName(CryptoAlgorithmIdentifier identifier)
 {
     switch (identifier) {
