diff a/src/hotspot/share/oops/inlineKlass.cpp b/src/hotspot/share/oops/inlineKlass.cpp
--- a/src/hotspot/share/oops/inlineKlass.cpp
+++ b/src/hotspot/share/oops/inlineKlass.cpp
@@ -181,15 +181,15 @@
   if (!UseFlatArray) {
     return false;
   }
   // Too big
   int elem_bytes = raw_value_byte_size();
-  if ((InlineArrayElemMaxFlatSize >= 0) && (elem_bytes > InlineArrayElemMaxFlatSize)) {
+  if ((FlatArrayElemMaxFlatSize >= 0) && (elem_bytes > FlatArrayElemMaxFlatSize)) {
     return false;
   }
   // Too many embedded oops
-  if ((InlineArrayElemMaxFlatOops >= 0) && (nonstatic_oop_count() > InlineArrayElemMaxFlatOops)) {
+  if ((FlatArrayElemMaxFlatOops >= 0) && (nonstatic_oop_count() > FlatArrayElemMaxFlatOops)) {
     return false;
   }
   // Declared atomic but not naturally atomic.
   if (is_declared_atomic() && !is_naturally_atomic()) {
     return false;
