diff a/src/java.base/share/classes/sun/security/jca/ProviderConfig.java b/src/java.base/share/classes/sun/security/jca/ProviderConfig.java
--- a/src/java.base/share/classes/sun/security/jca/ProviderConfig.java
+++ b/src/java.base/share/classes/sun/security/jca/ProviderConfig.java
@@ -319,11 +319,11 @@
         }
 
         /**
          * Loads the provider with the specified class name.
          *
-         * @param name the name of the provider
+         * @param pn the name of the provider
          * @return the Provider, or null if it cannot be found or loaded
          * @throws ProviderException all other exceptions are ignored
          */
         public Provider load(String pn) {
             if (debug != null) {
