diff a/make/autoconf/hotspot.m4 b/make/autoconf/hotspot.m4
--- a/make/autoconf/hotspot.m4
+++ b/make/autoconf/hotspot.m4
@@ -24,10 +24,13 @@
 #
 
 # All valid JVM variants
 VALID_JVM_VARIANTS="server client minimal core zero custom"
 
+# Valhalla temporarily disabled
+VALHALLA_TEMP=false
+
 ###############################################################################
 # Check if the specified JVM variant should be built. To be used in shell if
 # constructs, like this:
 # if HOTSPOT_CHECK_JVM_VARIANT(server); then
 #
