diff a/src/java.security.jgss/share/classes/org/ietf/jgss/package-info.java b/src/java.security.jgss/share/classes/org/ietf/jgss/package-info.java
--- a/src/java.security.jgss/share/classes/org/ietf/jgss/package-info.java
+++ b/src/java.security.jgss/share/classes/org/ietf/jgss/package-info.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -50,11 +50,11 @@
  *
  * The GSS-API does not perform any communication with the peer. It merely
  * produces tokens that the application must somehow transport to the
  * other end.
  *
- * <h3 id="useSubjectCredsOnly">Credential Acquisition</h3>
+ * <h2 id="useSubjectCredsOnly">Credential Acquisition</h2>
  * The GSS-API itself does not dictate how an underlying mechanism
  * obtains the credentials that are needed for authentication. It is
  * assumed that prior to calling the GSS-API, these credentials are
  * obtained and stored in a location that the mechanism provider is
  * aware of. However, the default model in the Java platform will be
@@ -91,31 +91,12 @@
  * the provider is free to use any credentials cache of its choice. Such
  * a credential cache might be a disk cache, an in-memory cache, or even
  * just the current Subject itself.
  *
  * <h2>Related Documentation</h2>
- * <p>
  * For an online tutorial on using Java GSS-API, please see
  * {@extLink security_guide_jgss_tutorial
  * Introduction to JAAS and Java GSS-API}.
- * </p>
- *
- * <!--
- * <h2>Package Specification</h2>
- *
- * ##### FILL IN ANY SPECS NEEDED BY JAVA COMPATIBILITY KIT #####
- * <ul>
- * <li><a href="">##### REFER TO ANY FRAMEMAKER SPECIFICATION HERE #####</a>
- * </ul>
- *
- * <h2>Related Documentation</h2>
- *
- * For overviews, tutorials, examples, guides, and tool documentation, please see:
- * <ul>
- * <li><a href="">##### REFER TO NON-SPEC DOCUMENTATION HERE #####</a>
- * </ul>
- *
- * -->
  *
  * @since 1.4
  * */
 package org.ietf.jgss;
