diff a/test/hotspot/jtreg/vmTestbase/vm/mlvm/indy/stress/java/loopsAndThreads/TestDescription.java b/test/hotspot/jtreg/vmTestbase/vm/mlvm/indy/stress/java/loopsAndThreads/TestDescription.java
--- a/test/hotspot/jtreg/vmTestbase/vm/mlvm/indy/stress/java/loopsAndThreads/TestDescription.java
+++ b/test/hotspot/jtreg/vmTestbase/vm/mlvm/indy/stress/java/loopsAndThreads/TestDescription.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -32,11 +32,10 @@
  *     A Stress test that starts 100 threads in parallel. Each thread executes 100 InvokeDynamic operation
  *     in 10,000 cycles (to provoke compiler).
  *
  * @library /vmTestbase
  *          /test/lib
- * @run driver jdk.test.lib.FileInstaller . .
  *
  * @comment build test class and indify classes
  * @build vm.mlvm.indy.stress.java.loopsAndThreads.INDIFY_Test
  * @run driver vm.mlvm.share.IndifiedClassesBuilder
  *
