diff a/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java b/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java
--- a/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java
+++ b/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java
@@ -236,12 +236,11 @@
             String s = classUrlString.substring(4,
                     classUrlString.lastIndexOf('!'));
             // Strip everything after the last "/" or "\" to get rid of the jar filename
             int lastIndexOfSlash = Math.max(
                     s.lastIndexOf('/'), s.lastIndexOf('\\'));
-            return new File(new URL(s.substring(0, lastIndexOfSlash + 1)).getPath())
-                    .getParentFile();
+            return new File(new URL(s.substring(0, lastIndexOfSlash + 1)).getPath());
         } catch (MalformedURLException e) {
             return null;
         }
     }
 
diff a/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MonocleApplication.java b/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MonocleApplication.java
--- a/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MonocleApplication.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MonocleApplication.java
@@ -223,11 +223,11 @@
         Screen screen = new Screen(1l, // dummy native pointer;
                                    ns.getDepth(),
                                    0, 0, ns.getWidth(), ns.getHeight(),
                                    0, 0, ns.getWidth(), ns.getHeight(),
                                    0, 0, ns.getWidth(), ns.getHeight(),
-                                   ns.getWidth(), ns.getHeight(),
+                                   ns.getDPI(), ns.getDPI(),
                                    1.f, 1.f, ns.getScale(), ns.getScale());
         // Move the cursor to the middle of the screen
         MouseState mouseState = new MouseState();
         mouseState.setX(ns.getWidth() / 2);
         mouseState.setY(ns.getHeight() / 2);
