diff a/test/jdk/tools/jextract/test8244959/Test8244959.java b/test/jdk/tools/jextract/test8244959/Test8244959.java
--- a/test/jdk/tools/jextract/test8244959/Test8244959.java
+++ b/test/jdk/tools/jextract/test8244959/Test8244959.java
@@ -38,13 +38,13 @@
  */
 public class Test8244959 {
     @Test
     public void testsPrintf() {
         try (MemorySegment s = MemorySegment.allocateNative(1024)) {
-            my_sprintf(s.address(),
-                toCString("%hhd %c %.2f %.2f %lld %lld %d %hd %d %d %lld %c").address(), 12,
+            my_sprintf(s,
+                toCString("%hhd %c %.2f %.2f %lld %lld %d %hd %d %d %lld %c"), 12,
                 (byte) 1, 'b', -1.25f, 5.5d, -200L, Long.MAX_VALUE, (byte) -2, (short) 2, 3, (short) -4, 5L, 'a');
-            String str = toJavaString(s.address());
+            String str = toJavaString(s);
             assertEquals(str, "1 b -1.25 5.50 -200 " + Long.MAX_VALUE + " -2 2 3 -4 5 a");
         }
     }
 }
