diff a/src/jdk.hotspot.agent/share/classes/module-info.java b/src/jdk.hotspot.agent/share/classes/module-info.java
--- a/src/jdk.hotspot.agent/share/classes/module-info.java
+++ b/src/jdk.hotspot.agent/share/classes/module-info.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2014, 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -28,14 +28,11 @@
  *
  * <p> This module includes the <em>{@index jhsdb jhsdb tool}</em> tool to
  * attach to a running Java Virtual Machine (JVM) or launch a postmortem
  * debugger to analyze the content of a core-dump from a crashed JVM.
  *
- * <dl style="font-family:'DejaVu Sans', Arial, Helvetica, sans serif">
- * <dt class="simpleTagLabel">Tool Guides:</dt>
- * <dd> {@extLink jhsdb_tool_reference jhsdb}</dd>
- * </dl>
+ * @toolGuide jhsdb
  *
  * @moduleGraph
  * @since 9
  */
 module jdk.hotspot.agent {
