diff a/src/java.base/share/classes/java/security/cert/CertStoreException.java b/src/java.base/share/classes/java/security/cert/CertStoreException.java
--- a/src/java.base/share/classes/java/security/cert/CertStoreException.java
+++ b/src/java.base/share/classes/java/security/cert/CertStoreException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -48,10 +48,11 @@
  * @since       1.4
  * @author      Sean Mullan
  */
 public class CertStoreException extends GeneralSecurityException {
 
+    @java.io.Serial
     private static final long serialVersionUID = 2395296107471573245L;
 
     /**
      * Creates a {@code CertStoreException} with {@code null} as
      * its detail message.
