diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/generator/Section.rb b/modules/javafx.web/src/main/native/Source/JavaScriptCore/generator/Section.rb
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/generator/Section.rb
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/generator/Section.rb
@@ -98,11 +98,14 @@
       if config[:emit_opcode_id_string_values_in_h_file]
           opcodes.each { |opcode|
               out.write("#define #{opcode.name}_value_string \"#{opcode.id}\"\n")
           }
           opcodes.each { |opcode|
-              out.write("#define #{opcode.name}_wide_value_string \"#{num_opcodes + opcode.id}\"\n")
+              out.write("#define #{opcode.name}_wide16_value_string \"#{num_opcodes + opcode.id}\"\n")
+          }
+          opcodes.each { |opcode|
+              out.write("#define #{opcode.name}_wide32_value_string \"#{num_opcodes * 2 + opcode.id}\"\n")
           }
       end
       out.string
   end
 end
