diff a/modules/javafx.web/src/main/native/Source/WebCore/loader/soup/ResourceLoaderSoup.cpp b/modules/javafx.web/src/main/native/Source/WebCore/loader/soup/ResourceLoaderSoup.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/loader/soup/ResourceLoaderSoup.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/loader/soup/ResourceLoaderSoup.cpp
@@ -64,11 +64,11 @@
         const auto* data = static_cast<const guchar*>(g_bytes_get_data(bytes.get(), &dataSize));
         GUniquePtr<char> fileName(g_path_get_basename(url.path().utf8().data()));
         GUniquePtr<char> contentType(g_content_type_guess(fileName.get(), data, dataSize, nullptr));
         ResourceResponse response { url, extractMIMETypeFromMediaType(contentType.get()), static_cast<long long>(dataSize), extractCharsetFromMediaType(contentType.get()) };
         response.setHTTPStatusCode(200);
-        response.setHTTPStatusText(ASCIILiteral("OK"));
+        response.setHTTPStatusText("OK"_s);
         response.setHTTPHeaderField(HTTPHeaderName::ContentType, contentType.get());
         response.setSource(ResourceResponse::Source::Network);
         loader->deliverResponseAndData(response, SharedBuffer::create(bytes.get()));
     }, protectedThis.leakRef()));
 
