diff a/modules/javafx.web/src/main/native/Source/WebCore/html/Autocapitalize.h b/modules/javafx.web/src/main/native/Source/WebCore/html/Autocapitalize.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/Autocapitalize.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/Autocapitalize.h
@@ -24,13 +24,13 @@
  */
 
 #pragma once
 
 #include "AutocapitalizeTypes.h"
-#include <wtf/text/AtomicString.h>
+#include <wtf/text/AtomString.h>
 
 namespace WebCore {
 
-AutocapitalizeType autocapitalizeTypeForAttributeValue(const AtomicString&);
-const AtomicString& stringForAutocapitalizeType(AutocapitalizeType);
+AutocapitalizeType autocapitalizeTypeForAttributeValue(const AtomString&);
+const AtomString& stringForAutocapitalizeType(AutocapitalizeType);
 
 } // namespace WebCore
