diff a/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.h b/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.h
@@ -30,12 +30,10 @@
 
 #pragma once
 
 #include "File.h"
 #include "TextEncoding.h"
-#include <wtf/Forward.h>
-#include <wtf/Optional.h>
 #include <wtf/RefCounted.h>
 #include <wtf/Variant.h>
 #include <wtf/text/WTFString.h>
 
 namespace WebCore {
