diff a/modules/javafx.web/src/main/native/Source/WebCore/html/GenericCachedHTMLCollection.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/GenericCachedHTMLCollection.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/GenericCachedHTMLCollection.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/GenericCachedHTMLCollection.cpp
@@ -29,15 +29,22 @@
 #include "HTMLAppletElement.h"
 #include "HTMLFieldSetElement.h"
 #include "HTMLNames.h"
 #include "HTMLObjectElement.h"
 #include "HTMLOptionElement.h"
+#include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
 using namespace HTMLNames;
 
+using GenericDescendantsCachedHTMLCollection = GenericCachedHTMLCollection<CollectionTraversalType::Descendants>;
+using GenericChildrenOnlyCachedHTMLCollection = GenericCachedHTMLCollection<CollectionTraversalType::ChildrenOnly>;
+
+WTF_MAKE_ISO_ALLOCATED_IMPL_TEMPLATE(GenericDescendantsCachedHTMLCollection);
+WTF_MAKE_ISO_ALLOCATED_IMPL_TEMPLATE(GenericChildrenOnlyCachedHTMLCollection);
+
 template <CollectionTraversalType traversalType>
 bool GenericCachedHTMLCollection<traversalType>::elementMatches(Element& element) const
 {
     switch (this->type()) {
     case NodeChildren:
