diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFrameElement.h b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFrameElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFrameElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFrameElement.h
@@ -43,11 +43,11 @@
     HTMLFrameElement(const QualifiedName&, Document&);
 
     void didAttachRenderers() final;
     bool rendererIsNeeded(const RenderStyle&) final;
     RenderPtr<RenderElement> createElementRenderer(RenderStyle&&, const RenderTreePosition&) final;
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
 
     bool m_frameBorder { true };
     bool m_frameBorderSet { false };
 };
 
