diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLParagraphElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLParagraphElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLParagraphElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLParagraphElement.cpp
@@ -49,11 +49,11 @@
 Ref<HTMLParagraphElement> HTMLParagraphElement::create(const QualifiedName& tagName, Document& document)
 {
     return adoptRef(*new HTMLParagraphElement(tagName, document));
 }
 
-void HTMLParagraphElement::collectStyleForPresentationAttribute(const QualifiedName& name, const AtomicString& value, MutableStyleProperties& style)
+void HTMLParagraphElement::collectStyleForPresentationAttribute(const QualifiedName& name, const AtomString& value, MutableStyleProperties& style)
 {
     if (name == alignAttr) {
         if (equalLettersIgnoringASCIICase(value, "middle") || equalLettersIgnoringASCIICase(value, "center"))
             addPropertyToPresentationAttributeStyle(style, CSSPropertyTextAlign, CSSValueWebkitCenter);
         else if (equalLettersIgnoringASCIICase(value, "left"))
