diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFieldSetElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFieldSetElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFieldSetElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLFieldSetElement.cpp
@@ -150,13 +150,13 @@
 bool HTMLFieldSetElement::supportsFocus() const
 {
     return HTMLElement::supportsFocus();
 }
 
-const AtomicString& HTMLFieldSetElement::formControlType() const
+const AtomString& HTMLFieldSetElement::formControlType() const
 {
-    static NeverDestroyed<const AtomicString> fieldset("fieldset", AtomicString::ConstructFromLiteral);
+    static NeverDestroyed<const AtomString> fieldset("fieldset", AtomString::ConstructFromLiteral);
     return fieldset;
 }
 
 RenderPtr<RenderElement> HTMLFieldSetElement::createElementRenderer(RenderStyle&& style, const RenderTreePosition&)
 {
