diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMenuElement.h b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMenuElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMenuElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMenuElement.h
@@ -36,11 +36,11 @@
 private:
     HTMLMenuElement(const QualifiedName&, Document&);
 
     InsertedIntoAncestorResult insertedIntoAncestor(InsertionType, ContainerNode&) final;
     void removedFromAncestor(RemovalType, ContainerNode&) final;
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
 
     bool m_isTouchBarMenu;
 };
 
 } // namespace WebCore
