diff a/modules/javafx.web/src/main/native/Source/WebCore/html/ImageBitmap.h b/modules/javafx.web/src/main/native/Source/WebCore/html/ImageBitmap.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/ImageBitmap.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/ImageBitmap.h
@@ -25,13 +25,11 @@
 
 #pragma once
 
 #include "JSDOMPromiseDeferred.h"
 #include "ScriptWrappable.h"
-#include <wtf/Optional.h>
 #include <wtf/RefCounted.h>
-#include <wtf/Variant.h>
 
 namespace WebCore {
 
 class Blob;
 class HTMLCanvasElement;
@@ -45,11 +43,12 @@
 class PendingImageBitmap;
 class ScriptExecutionContext;
 class TypedOMCSSImageValue;
 struct ImageBitmapOptions;
 
-class ImageBitmap : public ScriptWrappable, public RefCounted<ImageBitmap> {
+class ImageBitmap final : public ScriptWrappable, public RefCounted<ImageBitmap> {
+    WTF_MAKE_ISO_ALLOCATED(ImageBitmap);
 public:
     using Source = Variant<
         RefPtr<HTMLImageElement>,
 #if ENABLE(VIDEO)
         RefPtr<HTMLVideoElement>,
