diff a/modules/javafx.web/src/main/native/Source/WebCore/html/MediaController.h b/modules/javafx.web/src/main/native/Source/WebCore/html/MediaController.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/MediaController.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/MediaController.h
@@ -40,10 +40,11 @@
 namespace WebCore {
 
 class HTMLMediaElement;
 
 class MediaController final : public RefCounted<MediaController>, public MediaControllerInterface, public EventTargetWithInlineData {
+    WTF_MAKE_ISO_ALLOCATED(MediaController);
 public:
     static Ref<MediaController> create(ScriptExecutionContext&);
     virtual ~MediaController();
 
     Ref<TimeRanges> buffered() const final;
@@ -69,11 +70,11 @@
     ExceptionOr<void> setVolume(double) final;
 
     bool muted() const final { return m_muted; }
     void setMuted(bool) final;
 
-    const AtomicString& playbackState() const;
+    const AtomString& playbackState() const;
 
     using RefCounted::ref;
     using RefCounted::deref;
 
 private:
@@ -82,11 +83,11 @@
     void reportControllerState();
     void updateReadyState();
     void updatePlaybackState();
     void updateMediaElements();
     void bringElementUpToSpeed(HTMLMediaElement&);
-    void scheduleEvent(const AtomicString& eventName);
+    void scheduleEvent(const AtomString& eventName);
     void asyncEventTimerFired();
     void clearPositionTimerFired();
     bool hasEnded() const;
     void scheduleTimeupdateEvent();
     void startTimeupdateTimer();
