diff a/modules/javafx.web/src/main/native/Source/WebCore/html/URLInputType.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/URLInputType.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/URLInputType.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/URLInputType.cpp
@@ -38,11 +38,11 @@
 #include "LocalizedStrings.h"
 #include <wtf/URL.h>
 
 namespace WebCore {
 
-const AtomicString& URLInputType::formControlType() const
+const AtomString& URLInputType::formControlType() const
 {
     return InputTypeNames::url();
 }
 
 bool URLInputType::typeMismatchFor(const String& value) const
