diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp
@@ -31,10 +31,11 @@
 #include "B3CCallValue.h"
 #include "B3FenceValue.h"
 #include "B3MemoryValue.h"
 #include "B3PatchpointValue.h"
 #include "B3ValueInlines.h"
+#include "DateInstance.h"
 #include "DirectArguments.h"
 #include "FTLState.h"
 #include "GetterSetter.h"
 #include "JSPropertyNameEnumerator.h"
 #include "JSScope.h"
