diff a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HotSpotAgent.java b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HotSpotAgent.java
--- a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HotSpotAgent.java
+++ b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HotSpotAgent.java
@@ -481,11 +481,11 @@
 
     // Given a classname, load an alternate implementation of JVMDebugger.
     private void setupDebuggerAlternate(String alternateName) {
 
         try {
-            Class c = Class.forName(alternateName);
+            Class<?> c = Class.forName(alternateName);
             Constructor cons = c.getConstructor();
             debugger = (JVMDebugger) cons.newInstance();
             attachDebugger();
             setupDebuggerExisting();
 
@@ -624,11 +624,11 @@
             }
         } else {
           try {
             machDesc = (MachineDescription)
               Class.forName("sun.jvm.hotspot.debugger.MachineDescription" +
-                            cpu.toUpperCase()).newInstance();
+                            cpu.toUpperCase()).getDeclaredConstructor().newInstance();
           } catch (Exception e) {
             throw new DebuggerException("Linux not supported on machine type " + cpu);
           }
         }
 
