diff a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/ContributorTests.java b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/ContributorTests.java
--- a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/ContributorTests.java
+++ b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/ContributorTests.java
@@ -104,11 +104,11 @@
                                .findAny()
                                .orElseThrow();
             assertEquals("Co-authored-by: Test Person <test@test.test>", creditLine);
 
             var pushed = pr.comments().stream()
-                           .filter(comment -> comment.body().contains("change can now be integrated"))
+                           .filter(comment -> comment.body().contains("change now passes all automated"))
                            .count();
             assertEquals(1, pushed);
 
             // Add a second person
             pr.addComment("/contributor add Another Person <another@test.test>");
