diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/SincResampler.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/SincResampler.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/SincResampler.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/SincResampler.h
@@ -29,10 +29,11 @@
 #ifndef SincResampler_h
 #define SincResampler_h
 
 #include "AudioArray.h"
 #include "AudioSourceProvider.h"
+#include <wtf/RefPtr.h>
 
 namespace WebCore {
 
 // SincResampler is a high-quality sample-rate converter.
 
@@ -78,10 +79,12 @@
     // m_sourceProvider is used to provide the audio input stream to the resampler.
     AudioSourceProvider* m_sourceProvider;
 
     // The buffer is primed once at the very beginning of processing.
     bool m_isBufferPrimed;
+
+    RefPtr<AudioBus> m_internalBus;
 };
 
 } // namespace WebCore
 
 #endif // SincResampler_h
