diff a/modules/javafx.web/src/main/native/Source/WebCore/html/track/InbandGenericTextTrack.h b/modules/javafx.web/src/main/native/Source/WebCore/html/track/InbandGenericTextTrack.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/track/InbandGenericTextTrack.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/track/InbandGenericTextTrack.h
@@ -50,10 +50,11 @@
     CueToDataMap m_cueToDataMap;
     CueDataToCueMap m_dataToCueMap;
 };
 
 class InbandGenericTextTrack final : public InbandTextTrack, private WebVTTParserClient {
+    WTF_MAKE_ISO_ALLOCATED(InbandGenericTextTrack);
 public:
     static Ref<InbandGenericTextTrack> create(ScriptExecutionContext&, TextTrackClient&, InbandTextTrackPrivate&);
     virtual ~InbandGenericTextTrack();
 
 private:
