diff a/modules/javafx.web/src/main/native/Source/WebCore/html/track/TextTrackCueGeneric.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/track/TextTrackCueGeneric.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/track/TextTrackCueGeneric.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/track/TextTrackCueGeneric.cpp
@@ -42,10 +42,12 @@
 #include <wtf/IsoMallocInlines.h>
 #include <wtf/MathExtras.h>
 
 namespace WebCore {
 
+WTF_MAKE_ISO_ALLOCATED_IMPL(TextTrackCueGeneric);
+
 // This default value must be the same as the one specified in mediaControlsApple.css for -webkit-media-controls-closed-captions-container
 const static int DEFAULTCAPTIONFONTSIZE = 10;
 
 class TextTrackCueGenericBoxElement final : public VTTCueBox {
     WTF_MAKE_ISO_ALLOCATED_INLINE(TextTrackCueGenericBoxElement);
