diff a/tests/system/src/testscriptapp1/java/mymod/myapp1/FXMLScriptDeployment.java b/tests/system/src/testscriptapp1/java/mymod/myapp1/FXMLScriptDeployment.java
--- a/tests/system/src/testscriptapp1/java/mymod/myapp1/FXMLScriptDeployment.java
+++ b/tests/system/src/testscriptapp1/java/mymod/myapp1/FXMLScriptDeployment.java
@@ -99,16 +99,16 @@
              scene = new Scene(rootNode);
              btn = (Button) scene.lookup("#idButton");
         }
         catch (Exception ioe) {
             ioe.printStackTrace();
-            System.exit(-1);
+            System.exit(ERROR_UNEXPECTED_EXCEPTION);
         }
             // fire three events on the button
         btn.fire();
         btn.fireEvent(new ActionEvent());
-        btn.fireEvent(new MouseEvent( MouseEvent.MOUSE_CLICKED,
+        btn.fireEvent(new MouseEvent(MouseEvent.MOUSE_CLICKED,
                                        0,       // double x,
                                        0,       // double y,
                                        0,       // double screenX,
                                        0,       // double screenY,
                                        MouseButton.PRIMARY,         // MouseButton button,
@@ -180,11 +180,11 @@
             TreeMap<String,Object> globalBindings = scopes.get(200);
 
             Object obj = null;
             Button btn = null;
 
-                // global Bindings
+            // global Bindings
             Util.assertExists(IDROOT + " in global scope Bindings", globalBindings.containsKey(IDROOT));
             obj = globalBindings.get(IDROOT);
             Util.assertType(IDROOT, AnchorPane.class, obj);
 
             Util.assertExists(LOCATION + " in global scope Bindings", globalBindings.containsKey(LOCATION));
@@ -203,11 +203,11 @@
                 obj = globalBindings.get(IDBUTTON);
                 Util.assertType(IDBUTTON, Button.class, obj);
                 btn = (Button) obj;
             }
 
-                // engine Bindings
+            // engine Bindings
             Util.assertExists(FILENAME + " in engine scope Bindings", engineBindings.containsKey(FILENAME));
             if (invocation < 7) {  // no event objects, no arguments
                 Util.assertNotExists(ARGV + " in engine scope Bindings", engineBindings.containsKey(ARGV));
                 Util.assertNotExists(EVENT + " in engine scope Bindings", engineBindings.containsKey(EVENT));
             }
@@ -223,12 +223,11 @@
                 if (invocation == 9) {
                     Util.assertType(EVENT, MouseEvent.class, obj);
                     MouseEvent ev = (MouseEvent) obj;
                     Util.assertSame("MouseEvent.getSource() == btn", ev.getSource(), btn);
                     Util.assertSame("MouseEvent.MOUSE_CLICKED", MouseEvent.MOUSE_CLICKED, ev.getEventType());
-                }
-                else {
+                } else {
                     Util.assertType(EVENT, ActionEvent.class, obj);
                     ActionEvent ev = (ActionEvent) obj;
                     Util.assertSame("ActionEvent.getSource() == btn", ev.getSource(), btn);
                 }
             }
