diff a/test/hotspot/jtreg/vmTestbase/nsk/stress/thread/thread006.java b/test/hotspot/jtreg/vmTestbase/nsk/stress/thread/thread006.java
--- a/test/hotspot/jtreg/vmTestbase/nsk/stress/thread/thread006.java
+++ b/test/hotspot/jtreg/vmTestbase/nsk/stress/thread/thread006.java
@@ -174,11 +174,11 @@
      * free CPU time so that the thread's memory could not
      * be moved to swap file.
      */
     public void run() {
         while (!GO && !timeout())
-            yield();
+            Thread.yield();
         while (!STOP && !timeout())
             ;
     }
 
     private static long startTime = System.currentTimeMillis();
