diff a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
--- a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
@@ -646,11 +646,11 @@
     public void addLink(Link link) {
         throw new RuntimeException("not implemented yet");
     }
 
     @Override
-    public void removeLink(URI uri) {
+    public void removeLink(Link link) {
         throw new RuntimeException("not implemented yet");
     }
 
     @Override
     public List<String> fixVersions() {
@@ -664,6 +664,21 @@
 
     @Override
     public void removeFixVersion(String fixVersion) {
         throw new RuntimeException("not implemented yet");
     }
+
+    @Override
+    public Map<String, String> properties() {
+        throw new RuntimeException("not implemented yet");
+    }
+
+    @Override
+    public void setProperty(String name, String value) {
+        throw new RuntimeException("not implemented yet");
+    }
+
+    @Override
+    public void removePropery(String name) {
+        throw new RuntimeException("not implemented yet");
+    }
 }
