diff a/vcs/src/main/java/org/openjdk/skara/vcs/Commits.java b/vcs/src/main/java/org/openjdk/skara/vcs/Commits.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/Commits.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/Commits.java
@@ -22,10 +22,11 @@
  */
 package org.openjdk.skara.vcs;
 
 import java.io.*;
 import java.util.*;
+import java.util.function.Consumer;
 import java.util.stream.*;
 
 public interface Commits extends Closeable, Iterable<Commit> {
     default List<Commit> asList() throws IOException {
         var result = new ArrayList<Commit>();
@@ -37,8 +38,24 @@
 
         return result;
     }
 
     default Stream<Commit> stream() {
-        return StreamSupport.stream(spliterator(), false);
+        return StreamSupport.stream(spliterator(), false).onClose(() -> {
+            try {
+                close();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    @Override
+    default void forEach(Consumer<? super Commit> action) {
+        Iterable.super.forEach(action);
+        try {
+            close();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
     }
 }
