diff a/src/java.base/share/classes/sun/security/x509/X509CRLEntryImpl.java b/src/java.base/share/classes/sun/security/x509/X509CRLEntryImpl.java
--- a/src/java.base/share/classes/sun/security/x509/X509CRLEntryImpl.java
+++ b/src/java.base/share/classes/sun/security/x509/X509CRLEntryImpl.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1997, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -75,11 +75,10 @@
     private CRLExtensions extensions = null;
     private byte[] revokedCert = null;
     private X500Principal certIssuer;
 
     private static final boolean isExplicit = false;
-    private static final long YR_2050 = 2524636800000L;
 
     /**
      * Constructs a revoked certificate entry using the given
      * serial number and revocation date.
      *
@@ -160,11 +159,11 @@
             if (revokedCert == null) {
                 DerOutputStream tmp = new DerOutputStream();
                 // sequence { serialNumber, revocationDate, extensions }
                 serialNumber.encode(tmp);
 
-                if (revocationDate.getTime() < YR_2050) {
+                if (revocationDate.getTime() < CertificateValidity.YR_2050) {
                     tmp.putUTCTime(revocationDate);
                 } else {
                     tmp.putGeneralizedTime(revocationDate);
                 }
 
