diff a/make/TestImage.gmk b/make/TestImage.gmk
--- a/make/TestImage.gmk
+++ b/make/TestImage.gmk
@@ -1,7 +1,7 @@
 #
-# Copyright (c) 2017, 2019, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2017, 2020, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.  Oracle designates this
@@ -33,21 +33,26 @@
 ifeq ($(call isTargetOs, windows), true)
   FIXPATH_COPY := $(TEST_IMAGE_DIR)/bin/fixpath.exe
 
   $(FIXPATH_COPY): $(firstword $(FIXPATH))
 	$(call install-file)
+
+  FIXPATH_WORKSPACE_ROOT := $(call FixPath, $(WORKSPACE_ROOT))
+  FIXPATH_OUTPUTDIR := $(call FixPath, $(OUTPUTDIR))
+else
+  FIXPATH_WORKSPACE_ROOT := $(WORKSPACE_ROOT)
+  FIXPATH_OUTPUTDIR := $(OUTPUTDIR)
 endif
 
-BUILD_INFO_PROPERTIES := $(TEST_IMAGE_DIR)/build-info.properties
 
-FIXPATH_ECHO := $(FIXPATH) $(call FixPath, $(ECHO))
+BUILD_INFO_PROPERTIES := $(TEST_IMAGE_DIR)/build-info.properties
 
 $(BUILD_INFO_PROPERTIES):
 	$(call MakeTargetDir)
 	$(ECHO) "# Build info properties for JDK tests" > $@
-	$(FIXPATH_ECHO) "build.workspace.root=$(WORKSPACE_ROOT)" >> $@
-	$(FIXPATH_ECHO) "build.output.root=$(OUTPUTDIR)" >> $@
+	$(ECHO) "build.workspace.root=$(FIXPATH_WORKSPACE_ROOT)" >> $@
+	$(ECHO) "build.output.root=$(FIXPATH_OUTPUTDIR)" >> $@
 
 prepare-test-image: $(FIXPATH_COPY) $(BUILD_INFO_PROPERTIES)
 	$(call MakeDir, $(TEST_IMAGE_DIR))
 	$(ECHO) > $(TEST_IMAGE_DIR)/Readme.txt 'JDK test image'
 
