diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webvr/NavigatorWebVR.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webvr/NavigatorWebVR.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webvr/NavigatorWebVR.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webvr/NavigatorWebVR.cpp
@@ -92,11 +92,11 @@
 
 NavigatorWebVR* NavigatorWebVR::from(Navigator* navigator)
 {
     NavigatorWebVR* supplement = static_cast<NavigatorWebVR*>(Supplement<Navigator>::from(navigator, supplementName()));
     if (!supplement) {
-        auto newSupplement = std::make_unique<NavigatorWebVR>();
+        auto newSupplement = makeUnique<NavigatorWebVR>();
         supplement = newSupplement.get();
         provideTo(navigator, supplementName(), WTFMove(newSupplement));
     }
     return supplement;
 }
