diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/JITDisassembler.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/JITDisassembler.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/JITDisassembler.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/JITDisassembler.cpp
@@ -87,11 +87,11 @@
     compilation->addDescription(Profiler::CompiledBytecode(Profiler::OriginStack(), out.toCString()));
 }
 
 void JITDisassembler::dumpHeader(PrintStream& out, LinkBuffer& linkBuffer)
 {
-    out.print("Generated Baseline JIT code for ", CodeBlockWithJITType(m_codeBlock, JITCode::BaselineJIT), ", instruction count = ", m_codeBlock->instructionCount(), "\n");
+    out.print("Generated Baseline JIT code for ", CodeBlockWithJITType(m_codeBlock, JITType::BaselineJIT), ", instructions size = ", m_codeBlock->instructionsSize(), "\n");
     out.print("   Source: ", m_codeBlock->sourceCodeOnOneLine(), "\n");
     out.print("   Code at [", RawPointer(linkBuffer.debugAddress()), ", ", RawPointer(static_cast<char*>(linkBuffer.debugAddress()) + linkBuffer.size()), "):\n");
 }
 
 MacroAssembler::Label JITDisassembler::firstSlowLabel()
