diff a/test/langtools/tools/javac/tree/NoPrivateTypesExported.java b/test/langtools/tools/javac/tree/NoPrivateTypesExported.java
--- a/test/langtools/tools/javac/tree/NoPrivateTypesExported.java
+++ b/test/langtools/tools/javac/tree/NoPrivateTypesExported.java
@@ -175,10 +175,11 @@
         for (AnnotationMirror mirror : annotations) {
             Element annotationElement = mirror.getAnnotationType().asElement();
 
             if (annotationElement.getAnnotation(Documented.class) == null) {
                 note("Ignoring undocumented annotation: " + mirror.getAnnotationType());
+                continue;
             }
 
             verifyTypeAcceptable(mirror.getAnnotationType(), acceptable);
 
             for (AnnotationValue value : mirror.getElementValues().values()) {
