diff a/test/langtools/tools/javac/tree/TreeKindTest.java b/test/langtools/tools/javac/tree/TreeKindTest.java
--- a/test/langtools/tools/javac/tree/TreeKindTest.java
+++ b/test/langtools/tools/javac/tree/TreeKindTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2006, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2006, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -104,10 +104,11 @@
 
             case INTERFACE:
             case ANNOTATION_TYPE:
             case ENUM:
             case CLASS:
+            case RECORD:
                 ok = ok & verify(k, i, i == ClassTree.class);
                 break;
 
             case ANNOTATION:
             case TYPE_ANNOTATION:
