diff a/test/jdk/jdk/jfr/api/consumer/streaming/TestFilledChunks.java b/test/jdk/jdk/jfr/api/consumer/streaming/TestFilledChunks.java
--- a/test/jdk/jdk/jfr/api/consumer/streaming/TestFilledChunks.java
+++ b/test/jdk/jdk/jfr/api/consumer/streaming/TestFilledChunks.java
@@ -52,10 +52,11 @@
     // Will generate about 100 MB of data, or 8-9 chunks
     private static final int EVENT_COUNT = 5_000_000;
 
     public static void main(String... args) throws Exception {
         try (RecordingStream rs = new RecordingStream()) {
+            rs.setOrdered(false);
             rs.onEvent(FillEvent.class.getName(), e -> {
                 int id = e.getInt("id");
                 // Some events may get lost due to
                 // buffer being full.
                 if (id > EVENT_COUNT / 2) {
