diff a/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java b/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java
--- a/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java
+++ b/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java
@@ -50,12 +50,12 @@
         }
         return Collections.unmodifiableSet(deserializer.deserialize(current));
     }
 
     @Override
-    public void put(T item) {
-        var updated = serializer.serialize(item, current());
+    public void put(Collection<T> items) {
+        var updated = serializer.serialize(items, current());
         if (current.equals(updated)) {
             return;
         }
         try {
             Files.writeString(file, updated, StandardCharsets.UTF_8);
