diff a/src/java.desktop/share/native/libfreetype/include/freetype/internal/cffotypes.h b/src/java.desktop/share/native/libfreetype/include/freetype/internal/cffotypes.h
--- a/src/java.desktop/share/native/libfreetype/include/freetype/internal/cffotypes.h
+++ b/src/java.desktop/share/native/libfreetype/include/freetype/internal/cffotypes.h
@@ -1,21 +1,21 @@
-/***************************************************************************/
-/*                                                                         */
-/*  cffotypes.h                                                            */
-/*                                                                         */
-/*    Basic OpenType/CFF object type definitions (specification).          */
-/*                                                                         */
-/*  Copyright 2017-2018 by                                                 */
-/*  David Turner, Robert Wilhelm, and Werner Lemberg.                      */
-/*                                                                         */
-/*  This file is part of the FreeType project, and may only be used,       */
-/*  modified, and distributed under the terms of the FreeType project      */
-/*  license, LICENSE.TXT.  By continuing to use, modify, or distribute     */
-/*  this file you indicate that you have read the license and              */
-/*  understand and accept it fully.                                        */
-/*                                                                         */
-/***************************************************************************/
+/****************************************************************************
+ *
+ * cffotypes.h
+ *
+ *   Basic OpenType/CFF object type definitions (specification).
+ *
+ * Copyright (C) 2017-2019 by
+ * David Turner, Robert Wilhelm, and Werner Lemberg.
+ *
+ * This file is part of the FreeType project, and may only be used,
+ * modified, and distributed under the terms of the FreeType project
+ * license, LICENSE.TXT.  By continuing to use, modify, or distribute
+ * this file you indicate that you have read the license and
+ * understand and accept it fully.
+ *
+ */
 
 
 #ifndef CFFOTYPES_H_
 #define CFFOTYPES_H_
 
@@ -31,34 +31,34 @@
 
 
   typedef TT_Face  CFF_Face;
 
 
-  /*************************************************************************/
-  /*                                                                       */
-  /* <Type>                                                                */
-  /*    CFF_Size                                                           */
-  /*                                                                       */
-  /* <Description>                                                         */
-  /*    A handle to an OpenType size object.                               */
-  /*                                                                       */
+  /**************************************************************************
+   *
+   * @type:
+   *   CFF_Size
+   *
+   * @description:
+   *   A handle to an OpenType size object.
+   */
   typedef struct  CFF_SizeRec_
   {
     FT_SizeRec  root;
     FT_ULong    strike_index;    /* 0xFFFFFFFF to indicate invalid */
 
   } CFF_SizeRec, *CFF_Size;
 
 
-  /*************************************************************************/
-  /*                                                                       */
-  /* <Type>                                                                */
-  /*    CFF_GlyphSlot                                                      */
-  /*                                                                       */
-  /* <Description>                                                         */
-  /*    A handle to an OpenType glyph slot object.                         */
-  /*                                                                       */
+  /**************************************************************************
+   *
+   * @type:
+   *   CFF_GlyphSlot
+   *
+   * @description:
+   *   A handle to an OpenType glyph slot object.
+   */
   typedef struct  CFF_GlyphSlotRec_
   {
     FT_GlyphSlotRec  root;
 
     FT_Bool  hint;
@@ -68,30 +68,30 @@
     FT_Fixed  y_scale;
 
   } CFF_GlyphSlotRec, *CFF_GlyphSlot;
 
 
-  /*************************************************************************/
-  /*                                                                       */
-  /* <Type>                                                                */
-  /*    CFF_Internal                                                       */
-  /*                                                                       */
-  /* <Description>                                                         */
-  /*    The interface to the `internal' field of `FT_Size'.                */
-  /*                                                                       */
+  /**************************************************************************
+   *
+   * @type:
+   *   CFF_Internal
+   *
+   * @description:
+   *   The interface to the 'internal' field of `FT_Size`.
+   */
   typedef struct  CFF_InternalRec_
   {
     PSH_Globals  topfont;
     PSH_Globals  subfonts[CFF_MAX_CID_FONTS];
 
   } CFF_InternalRec, *CFF_Internal;
 
 
-  /*************************************************************************/
-  /*                                                                       */
-  /* Subglyph transformation record.                                       */
-  /*                                                                       */
+  /**************************************************************************
+   *
+   * Subglyph transformation record.
+   */
   typedef struct  CFF_Transform_
   {
     FT_Fixed    xx, xy;     /* transformation matrix coefficients */
     FT_Fixed    yx, yy;
     FT_F26Dot6  ox, oy;     /* offsets                            */
