diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MediaEngineConfigurationFactoryMock.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MediaEngineConfigurationFactoryMock.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MediaEngineConfigurationFactoryMock.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MediaEngineConfigurationFactoryMock.h
@@ -29,18 +29,19 @@
 
 #include <wtf/Function.h>
 
 namespace WebCore {
 
-struct MediaCapabilitiesInfo;
+struct MediaCapabilitiesDecodingInfo;
+struct MediaCapabilitiesEncodingInfo;
 struct MediaDecodingConfiguration;
 struct MediaEncodingConfiguration;
 
 struct MediaEngineConfigurationFactoryMock {
-    using DecodingConfigurationCallback = WTF::Function<void(MediaCapabilitiesInfo&&)>;
-    using EncodingConfigurationCallback = WTF::Function<void(MediaCapabilitiesInfo&&)>;
+    using DecodingConfigurationCallback = WTF::Function<void(MediaCapabilitiesDecodingInfo&&)>;
+    using EncodingConfigurationCallback = WTF::Function<void(MediaCapabilitiesEncodingInfo&&)>;
 
-    static void createDecodingConfiguration(MediaDecodingConfiguration&, DecodingConfigurationCallback&&);
-    static void createEncodingConfiguration(MediaEncodingConfiguration&, EncodingConfigurationCallback&&);
+    static void createDecodingConfiguration(MediaDecodingConfiguration&&, DecodingConfigurationCallback&&);
+    static void createEncodingConfiguration(MediaEncodingConfiguration&&, EncodingConfigurationCallback&&);
 };
 
 }
