diff a/test/jdk/lib/testlibrary/java/lang/UCDFiles.java b/test/jdk/lib/testlibrary/java/lang/UCDFiles.java
--- a/test/jdk/lib/testlibrary/java/lang/UCDFiles.java
+++ b/test/jdk/lib/testlibrary/java/lang/UCDFiles.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -53,7 +53,7 @@
     public static Path SPECIAL_CASING =
         UCD_DIR.resolve("SpecialCasing.txt");
     public static Path UNICODE_DATA =
         UCD_DIR.resolve("UnicodeData.txt");
     public static Path EMOJI_DATA =
-        UCD_DIR.resolve("emoji-data.txt");
+        UCD_DIR.resolve("emoji").resolve("emoji-data.txt");
 }
