diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/IDLTypes.h b/modules/javafx.web/src/main/native/Source/WebCore/bindings/IDLTypes.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/IDLTypes.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/IDLTypes.h
@@ -139,15 +139,15 @@
 
 template<typename T> struct IDLTreatNullAsEmptyAdaptor : IDLString<String> {
     using InnerType = T;
 };
 
-template<typename T> struct IDLAtomicStringAdaptor : IDLString<AtomicString> {
+template<typename T> struct IDLAtomStringAdaptor : IDLString<AtomString> {
     using InnerType = T;
 };
 
-template<typename T> struct IDLRequiresExistingAtomicStringAdaptor : IDLString<AtomicString> {
+template<typename T> struct IDLRequiresExistingAtomStringAdaptor : IDLString<AtomString> {
     using InnerType = T;
 };
 
 struct IDLObject : IDLType<JSC::Strong<JSC::JSObject>> {
     using NullableType = JSC::Strong<JSC::JSObject>;
