diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.h
@@ -55,11 +55,11 @@
     BMPImageDecoder(AlphaOption, GammaAndColorProfileOption);
     void tryDecodeSize(bool allDataReceived) final { decode(true, allDataReceived); }
 
     inline uint32_t readUint32(int offset) const
     {
-        return BMPImageReader::readUint32(m_data.get(), m_decodedOffset + offset);
+        return BMPImageReader::readUint32(*m_data, m_decodedOffset + offset);
     }
 
     // Decodes the image. If |onlySize| is true, stops decoding after
     // calculating the image size. If decoding fails but there is no more
     // data coming, sets the "decode failure" flag.
