diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/image-decoders/bmp/BMPImageDecoder.cpp
@@ -51,11 +51,11 @@
     if (failed())
         return;
 
     ScalableImageDecoder::setData(data, allDataReceived);
     if (m_reader)
-        m_reader->setData(&data);
+        m_reader->setData(*m_data);
 }
 
 ScalableImageDecoderFrame* BMPImageDecoder::frameBufferAtIndex(size_t index)
 {
     if (index)
@@ -96,12 +96,12 @@
     size_t imgDataOffset = 0;
     if ((m_decodedOffset < sizeOfFileHeader) && !processFileHeader(&imgDataOffset))
         return false;
 
     if (!m_reader) {
-        m_reader = std::make_unique<BMPImageReader>(this, m_decodedOffset, imgDataOffset, false);
-        m_reader->setData(m_data.get());
+        m_reader = makeUnique<BMPImageReader>(this, m_decodedOffset, imgDataOffset, false);
+        m_reader->setData(*m_data);
     }
 
     if (!m_frameBufferCache.isEmpty())
         m_reader->setBuffer(&m_frameBufferCache.first());
 
