diff a/test/jdk/java/nio/channels/DatagramChannel/UseDGWithIPv6.java b/test/jdk/java/nio/channels/DatagramChannel/UseDGWithIPv6.java
--- a/test/jdk/java/nio/channels/DatagramChannel/UseDGWithIPv6.java
+++ b/test/jdk/java/nio/channels/DatagramChannel/UseDGWithIPv6.java
@@ -20,10 +20,11 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
 
 /* @test
+ * @library /test/lib
  * @bug 6435300
  * @summary Check using IPv6 address does not crash the VM
  * @run main/othervm UseDGWithIPv6
  * @run main/othervm -Djava.net.preferIPv4Stack=true UseDGWithIPv6
  */
@@ -33,20 +34,24 @@
 import java.net.SocketAddress;
 import java.nio.ByteBuffer;
 import java.nio.channels.DatagramChannel;
 import java.nio.channels.UnsupportedAddressTypeException;
 
+import jdk.test.lib.net.IPSupport;
+
 public class UseDGWithIPv6 {
     static String[] targets = {
         "3ffe:e00:811:b::21:5",
         "15.70.186.80"
     };
     static int BUFFER_LEN = 10;
     static int port = 12345;
 
     public static void main(String[] args) throws IOException
     {
+        IPSupport.throwSkippedExceptionIfNonOperational();
+
         ByteBuffer data = ByteBuffer.wrap("TESTING DATA".getBytes());
         DatagramChannel dgChannel = DatagramChannel.open();
 
         for(int i = 0; i < targets.length; i++){
             data.rewind();
