diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DataView.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DataView.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DataView.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DataView.cpp
@@ -31,12 +31,11 @@
 #include "JSGlobalObject.h"
 
 namespace JSC {
 
 DataView::DataView(RefPtr<ArrayBuffer>&& buffer, unsigned byteOffset, unsigned byteLength)
-    : ArrayBufferView(WTFMove(buffer), byteOffset)
-    , m_byteLength(byteLength)
+    : ArrayBufferView(WTFMove(buffer), byteOffset, byteLength)
 {
 }
 
 Ref<DataView> DataView::create(
     RefPtr<ArrayBuffer>&& buffer, unsigned byteOffset, unsigned byteLength)
