diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DirectArguments.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DirectArguments.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DirectArguments.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/DirectArguments.h
@@ -84,11 +84,11 @@
         return m_length;
     }
 
     bool isMappedArgument(uint32_t i) const
     {
-        return i < m_length && (!m_mappedArguments || !m_mappedArguments[i]);
+        return i < m_length && (!m_mappedArguments || !m_mappedArguments.at(i, m_length));
     }
 
     bool isMappedArgumentInDFG(uint32_t i) const
     {
         return i < m_length && !overrodeThings();
@@ -180,9 +180,10 @@
     unsigned mappedArgumentsSize();
 
     WriteBarrier<JSFunction> m_callee;
     uint32_t m_length; // Always the actual length of captured arguments and never what was stored into the length property.
     uint32_t m_minCapacity; // The max of this and length determines the capacity of this object. It may be the actual capacity, or maybe something smaller. We arrange it this way to be kind to the JITs.
-    CagedBarrierPtr<Gigacage::Primitive, bool> m_mappedArguments; // If non-null, it means that length, callee, and caller are fully materialized properties.
+    using MappedArguments = CagedBarrierPtr<Gigacage::Primitive, bool>;
+    MappedArguments m_mappedArguments; // If non-null, it means that length, callee, and caller are fully materialized properties.
 };
 
 } // namespace JSC
