diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h
@@ -30,13 +30,12 @@
 
 namespace JSC {
 
 template<typename Adaptor>
 GenericTypedArrayView<Adaptor>::GenericTypedArrayView(
-    RefPtr<ArrayBuffer>&& buffer, unsigned byteOffset, unsigned length)
-    : ArrayBufferView(WTFMove(buffer), byteOffset)
-    , m_length(length)
+RefPtr<ArrayBuffer>&& buffer, unsigned byteOffset, unsigned length)
+    : ArrayBufferView(WTFMove(buffer), byteOffset, length * sizeof(typename Adaptor::Type))
 {
 }
 
 template<typename Adaptor>
 Ref<GenericTypedArrayView<Adaptor>> GenericTypedArrayView<Adaptor>::create(unsigned length)
