diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/HashMapImpl.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/HashMapImpl.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/HashMapImpl.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/HashMapImpl.h
@@ -24,10 +24,11 @@
  */
 
 #pragma once
 
 #include "ExceptionHelpers.h"
+#include "JSCJSValueInlines.h"
 #include "JSObject.h"
 
 namespace JSC {
 
 JS_EXPORT_PRIVATE const ClassInfo* getHashMapBucketKeyClassInfo();
@@ -243,11 +244,11 @@
     if (key.isInt32())
         return key;
 
     double d = key.asDouble();
     if (std::isnan(d))
-        return key;
+        return jsNaN();
 
     int i = static_cast<int>(d);
     if (i == d) {
         // When a key is -0, we convert it to positive zero.
         // When a key is the double representation for an integer, we convert it to an integer.
