diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StackFrame.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StackFrame.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StackFrame.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StackFrame.h
@@ -24,10 +24,11 @@
  */
 
 #pragma once
 
 #include "Heap.h"
+#include "VM.h"
 #include "WasmIndexOrName.h"
 #include "WriteBarrier.h"
 #include <limits.h>
 
 namespace JSC {
@@ -56,11 +57,11 @@
         ASSERT(hasBytecodeOffset());
         return m_bytecodeOffset;
     }
 
     void visitChildren(SlotVisitor&);
-    bool isMarked() const { return (!m_callee || Heap::isMarked(m_callee.get())) && (!m_codeBlock || Heap::isMarked(m_codeBlock.get())); }
+    bool isMarked(VM& vm) const { return (!m_callee || vm.heap.isMarked(m_callee.get())) && (!m_codeBlock || vm.heap.isMarked(m_codeBlock.get())); }
 
 private:
     WriteBarrier<JSCell> m_callee { };
     WriteBarrier<CodeBlock> m_codeBlock { };
     Wasm::IndexOrName m_wasmFunctionIndexOrName;
