diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StructureRareData.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StructureRareData.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StructureRareData.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/StructureRareData.h
@@ -33,12 +33,12 @@
 
 namespace JSC {
 
 class JSPropertyNameEnumerator;
 class Structure;
-class ObjectToStringAdaptiveStructureWatchpoint;
 class ObjectToStringAdaptiveInferredPropertyValueWatchpoint;
+class ObjectToStringAdaptiveStructureWatchpoint;
 
 class StructureRareData final : public JSCell {
 public:
     typedef JSCell Base;
     static const unsigned StructureFlags = Base::StructureFlags | StructureIsImmortal;
@@ -88,10 +88,12 @@
         return OBJECT_OFFSETOF(StructureRareData, m_cachedOwnKeys);
     }
 
     DECLARE_EXPORT_INFO;
 
+    void finalizeUnconditionally(VM&);
+
 private:
     friend class Structure;
     friend class ObjectToStringAdaptiveStructureWatchpoint;
     friend class ObjectToStringAdaptiveInferredPropertyValueWatchpoint;
 
