diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Symbol.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Symbol.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Symbol.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Symbol.h
@@ -47,11 +47,11 @@
 
     static Symbol* create(VM&);
     static Symbol* createWithDescription(VM&, const String&);
     JS_EXPORT_PRIVATE static Symbol* create(VM&, SymbolImpl& uid);
 
-    const PrivateName& privateName() const { return m_privateName; }
+    PrivateName privateName() const { return m_privateName; }
     String descriptiveString() const;
     String description() const;
 
     JSValue toPrimitive(ExecState*, PreferredPrimitiveType) const;
     bool getPrimitiveNumber(ExecState*, double& number, JSValue&) const;
