diff a/test/langtools/jdk/javadoc/doclet/testInterface/TestInterface.java b/test/langtools/jdk/javadoc/doclet/testInterface/TestInterface.java
--- a/test/langtools/jdk/javadoc/doclet/testInterface/TestInterface.java
+++ b/test/langtools/jdk/javadoc/doclet/testInterface/TestInterface.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -65,25 +65,25 @@
                 "-sourcepath", testSrc,
                 "pkg");
         checkExit(Exit.OK);
 
         checkOutput("pkg/Interface.html", true,
-                "<div class=\"memberSignature\"><span class=\"returnType\">int</span>&nbsp;"
-                + "<span class=\"memberName\">method</span>()</div>",
-                "<div class=\"memberSignature\"><span class=\"modifiers\">static final</span>&nbsp;"
-                + "<span class=\"returnType\">int</span>&nbsp;<span class=\"memberName\">field</span></div>",
+                "<div class=\"member-signature\"><span class=\"return-type\">int</span>&nbsp;"
+                + "<span class=\"member-name\">method</span>()</div>",
+                "<div class=\"member-signature\"><span class=\"modifiers\">static final</span>&nbsp;"
+                + "<span class=\"return-type\">int</span>&nbsp;<span class=\"member-name\">field</span></div>",
                 // Make sure known implementing class list is correct and omits type parameters.
-                "<dl>\n"
+                "<dl class=\"notes\">\n"
                 + "<dt>All Known Implementing Classes:</dt>\n"
                 + "<dd><code><a href=\"Child.html\" title=\"class in pkg\">Child"
                 + "</a></code>, <code><a href=\"Parent.html\" title=\"class in pkg\">Parent"
                 + "</a></code></dd>\n"
                 + "</dl>");
 
         checkOutput("pkg/Child.html", true,
                 // Make sure "All Implemented Interfaces": has substituted type parameters
-                "<dl>\n"
+                "<dl class=\"notes\">\n"
                 + "<dt>All Implemented Interfaces:</dt>\n"
                 + "<dd><code><a href=\"Interface.html\" title=\"interface in pkg\">"
                 + "Interface</a>&lt;CE&gt;</code></dd>\n"
                 + "</dl>",
                 //Make sure Class Tree has substituted type parameters.
@@ -91,26 +91,26 @@
                 + "<div class=\"inheritance\"><a href=\"Parent.html\""
                 + " title=\"class in pkg\">pkg.Parent</a>&lt;CE&gt;\n"
                 + "<div class=\"inheritance\">pkg.Child&lt;CE&gt;</div>\n"
                 + "</div>\n</div>",
                 //Make sure "Specified By" has substituted type parameters.
-                "<dt><span class=\"overrideSpecifyLabel\">Specified by:</span></dt>\n"
+                "<dt>Specified by:</dt>\n"
                 + "<dd><code><a href=\"Interface.html#method()\">method</a>"
                 + "</code>&nbsp;in interface&nbsp;<code>"
                 + "<a href=\"Interface.html\" title=\"interface in pkg\">"
                 + "Interface</a>&lt;<a href=\"Child.html\" title=\"type parameter in Child\">"
                 + "CE</a>&gt;</code></dd>",
                 //Make sure "Overrides" has substituted type parameters.
-                "<dt><span class=\"overrideSpecifyLabel\">Overrides:</span></dt>\n"
+                "<dt>Overrides:</dt>\n"
                 + "<dd><code><a href=\"Parent.html#method()\">method</a>"
                 + "</code>&nbsp;in class&nbsp;<code><a href=\"Parent.html\" "
                 + "title=\"class in pkg\">Parent</a>&lt;<a href=\"Child.html\" "
                 + "title=\"type parameter in Child\">CE</a>&gt;</code></dd>");
 
         checkOutput("pkg/Parent.html", true,
-                //Make sure "Direct Know Subclasses" omits type parameters
-                "<dl>\n"
+                //Make sure "Direct Known Subclasses" omits type parameters
+                "<dl class=\"notes\">\n"
                 + "<dt>Direct Known Subclasses:</dt>\n"
                 + "<dd><code><a href=\"Child.html\" title=\"class in pkg\">Child"
                 + "</a></code></dd>\n"
                 + "</dl>");
 
@@ -119,38 +119,38 @@
                 "public static final&nbsp;int field");
 
         checkOutput("pkg/ClassWithStaticMembers.html", false,
                 //Make sure "Specified By" does not appear on class documentation when
                 //the method is a static method in the interface.
-                "<dt><span class=\"overrideSpecifyLabel\">Specified by:</span></dt>\n");
+                "<dt>Specified by:</dt>\n");
 
         checkOutput("pkg/ClassWithStaticMembers.html", true,
                 "<section class=\"detail\" id=\"f\">\n"
                 + "<h3>f</h3>\n"
-                + "<div class=\"memberSignature\"><span class=\"modifiers\">public static</span>&nbsp;"
-                + "<span class=\"returnType\">int</span>&nbsp;<span class=\"memberName\">f</span></div>\n"
+                + "<div class=\"member-signature\"><span class=\"modifiers\">public static</span>&nbsp;"
+                + "<span class=\"return-type\">int</span>&nbsp;<span class=\"member-name\">f</span></div>\n"
                 + "<div class=\"block\">A hider field</div>",
 
-                "<td class=\"colFirst\"><code>static void</code></td>\n"
-                + "<th class=\"colSecond\" scope=\"row\"><code><span class=\"memberNameLink\">"
+                "<td class=\"col-first\"><code>static void</code></td>\n"
+                + "<th class=\"col-second\" scope=\"row\"><code><span class=\"member-name-link\">"
                 + "<a href=\"#m()\">m</a></span>()</code></th>\n"
-                + "<td class=\"colLast\">\n"
+                + "<td class=\"col-last\">\n"
                 + "<div class=\"block\">A hider method</div>\n"
                 + "</td>\n",
 
                 "<section class=\"detail\" id=\"staticMethod()\">\n"
                 + "<h3>staticMethod</h3>\n"
-                + "<div class=\"memberSignature\"><span class=\"modifiers\">public static</span>&nbsp;"
-                + "<span class=\"returnType\">void</span>&nbsp;<span class=\"memberName\">staticMethod</span>()</div>\n"
-                + "<div class=\"block\"><span class=\"descfrmTypeLabel\">"
+                + "<div class=\"member-signature\"><span class=\"modifiers\">public static</span>&nbsp;"
+                + "<span class=\"return-type\">void</span>&nbsp;<span class=\"member-name\">staticMethod</span>()</div>\n"
+                + "<div class=\"block\"><span class=\"descfrm-type-label\">"
                 + "Description copied from interface:&nbsp;<code>"
                 + "<a href=\"InterfaceWithStaticMembers.html#staticMethod()\">"
                 + "InterfaceWithStaticMembers</a></code></span></div>\n"
                 + "<div class=\"block\">A static method</div>\n");
 
         checkOutput("pkg/ClassWithStaticMembers.InnerClass.html", true,
-                "<pre>public static class <span class=\"typeNameLabel\">"
+                "<pre>public static class <span class=\"type-name-label\">"
                 + "ClassWithStaticMembers.InnerClass</span>\n"
                 + "extends java.lang.Object</pre>\n"
                 + "<div class=\"block\">A hider inner class</div>");
     }
 
@@ -161,11 +161,11 @@
                 "pkg1");
         checkExit(Exit.OK);
 
         checkOutput("pkg1/Child.html", true,
             // Ensure the correct Overrides in the inheritance hierarchy is reported
-            "<span class=\"overrideSpecifyLabel\">Overrides:</span></dt>\n" +
+            "<dt>Overrides:</dt>\n" +
             "<dd><code><a href=\"GrandParent.html#method1()\">method1</a></code>" +
             "&nbsp;in class&nbsp;" +
             "<code><a href=\"GrandParent.html\" title=\"class in pkg1\">GrandParent</a>" +
             "&lt;<a href=\"Child.html\" title=\"type parameter in Child\">CE</a>&gt;</code>");
     }
