diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/DeferGC.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/DeferGC.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/DeferGC.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/DeferGC.cpp
@@ -29,9 +29,9 @@
 #include "JSCInlines.h"
 
 namespace JSC {
 
 #ifndef NDEBUG
-WTF::ThreadSpecificKey DisallowGC::s_scopeReentryCount = 0;
+LazyNeverDestroyed<ThreadSpecific<unsigned, WTF::CanBeGCThread::True>> DisallowGC::s_scopeReentryCount;
 #endif
 
 } // namespace JSC
