diff a/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIHost.idl b/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIHost.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIHost.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIHost.idl
@@ -35,14 +35,14 @@
     ImplementationLacksVTable
 ] interface CommandLineAPIHost {
     void clearConsoleMessages();
     void copyText(DOMString text);
 
-    [CallWith=ExecState] void inspect(any objectToInspect, any hints);
-    [CallWith=ExecState] any inspectedObject();
+    [CallWith=GlobalObject] void inspect(any objectToInspect, any hints);
+    [CallWith=GlobalObject] any inspectedObject();
 
-    [CallWith=ExecState] record<DOMString, sequence<ListenerEntry>> getEventListeners(EventTarget target);
+    [CallWith=GlobalObject] record<DOMString, sequence<ListenerEntry>> getEventListeners(EventTarget target);
 
     DOMString databaseId(Database database);
     DOMString storageId(Storage storage);
 };
 
