diff a/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp b/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
--- a/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
@@ -1499,11 +1499,11 @@
   }
 }
 
 void MacroAssembler::test_klass_is_value(Register klass, Register temp_reg, Label& is_value) {
   ldrw(temp_reg, Address(klass, Klass::access_flags_offset()));
-  andr(temp_reg, temp_reg, JVM_ACC_VALUE);
+  andr(temp_reg, temp_reg, JVM_ACC_INLINE);
   cbnz(temp_reg, is_value);
 }
 
 void MacroAssembler::test_field_is_inline_type(Register flags, Register temp_reg, Label& is_inline) {
   (void) temp_reg; // keep signature uniform with x86
