diff a/src/com/sun/javatest/report/HTMLReport.java b/src/com/sun/javatest/report/HTMLReport.java
--- a/src/com/sun/javatest/report/HTMLReport.java
+++ b/src/com/sun/javatest/report/HTMLReport.java
@@ -251,20 +251,20 @@
                 repWriter.endTag(HTMLWriterEx.H3);
             }
 
             // info from sections for main report
             repWriter.startTag(HTMLWriterEx.UL);
-            for (HTMLSection mainSection1 : mainSections) {
+            for (HTMLSection section : mainSections) {
                 repWriter.startTag(HTMLWriterEx.LI);
-                mainSection1.writeContents(repWriter);
+                section.writeContents(repWriter);
                 repWriter.endTag(HTMLWriterEx.LI);
             }
             repWriter.endTag(HTMLWriterEx.UL);
 
-            for (HTMLSection mainSection : mainSections) {
+            for (HTMLSection section : mainSections) {
                 repWriter.startTag(HTMLWriterEx.HR);
-                mainSection.writeSummary(repWriter);
+                section.writeSummary(repWriter);
                 repWriter.newLine();
             }
 
             repWriter.close();
         }
