diff a/test/jdk/jdk/jfr/event/runtime/TestShutdownEvent.java b/test/jdk/jdk/jfr/event/runtime/TestShutdownEvent.java
--- a/test/jdk/jdk/jfr/event/runtime/TestShutdownEvent.java
+++ b/test/jdk/jdk/jfr/event/runtime/TestShutdownEvent.java
@@ -228,13 +228,10 @@
         @Override
         public int  attempts() {
             if (Platform.isWindows()) {
                 return 0;
             }
-            if (signalName.equals("HUP") && Platform.isSolaris()) {
-                return 0;
-            }
             return 1;
         }
 
         public TestSig(String signalName) {
             this.signalName = signalName;
