diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStream.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStream.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStream.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStream.idl
@@ -26,11 +26,11 @@
     ActiveDOMObject,
     Conditional=MEDIA_STREAM,
     Constructor,
     Constructor(MediaStream stream),
     Constructor(sequence<MediaStreamTrack> tracks),
-    ConstructorCallWith=ScriptExecutionContext,
+    ConstructorCallWith=Document,
     ExportToWrappedFunction,
     PrivateIdentifier,
     PublicIdentifier
 ] interface MediaStream : EventTarget {
 
