diff a/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/DOMFormData.cpp
@@ -46,10 +46,11 @@
     : m_encoding(UTF8Encoding())
 {
     if (!form)
         return;
 
+    ASSERT(isMainThread());
     for (auto& element : form->copyAssociatedElementsVector()) {
         if (!element->asHTMLElement().isDisabledFormControl())
             element->appendFormData(*this, true);
     }
 }
