diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLEmbedElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLEmbedElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLEmbedElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLEmbedElement.cpp
@@ -92,12 +92,12 @@
 
 void HTMLEmbedElement::collectStyleForPresentationAttribute(const QualifiedName& name, const AtomString& value, MutableStyleProperties& style)
 {
     if (name == hiddenAttr) {
         if (equalLettersIgnoringASCIICase(value, "yes") || equalLettersIgnoringASCIICase(value, "true")) {
-            addPropertyToPresentationAttributeStyle(style, CSSPropertyWidth, 0, CSSPrimitiveValue::CSS_PX);
-            addPropertyToPresentationAttributeStyle(style, CSSPropertyHeight, 0, CSSPrimitiveValue::CSS_PX);
+            addPropertyToPresentationAttributeStyle(style, CSSPropertyWidth, 0, CSSUnitType::CSS_PX);
+            addPropertyToPresentationAttributeStyle(style, CSSPropertyHeight, 0, CSSUnitType::CSS_PX);
         }
     } else
         HTMLPlugInImageElement::collectStyleForPresentationAttribute(name, value, style);
 }
 
