diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLVideoElement.idl b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLVideoElement.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLVideoElement.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLVideoElement.idl
@@ -23,10 +23,11 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
  */
 
 [
     Conditional=VIDEO,
+    ExportMacro=WEBCORE_EXPORT,
     JSGenerateToNativeObject,
 ] interface HTMLVideoElement : HTMLMediaElement {
     [CEReactions=NotNeeded, Reflect] attribute unsigned long width;
     [CEReactions=NotNeeded, Reflect] attribute unsigned long height;
     readonly attribute unsigned long videoWidth;
@@ -56,6 +57,6 @@
     [Conditional=VIDEO_PRESENTATION_MODE] boolean webkitSupportsPresentationMode(VideoPresentationMode mode);
     [Conditional=VIDEO_PRESENTATION_MODE] readonly attribute VideoPresentationMode webkitPresentationMode;
     [Conditional=VIDEO_PRESENTATION_MODE] void webkitSetPresentationMode(VideoPresentationMode mode);
 };
 
-[Conditional=VIDEO_PRESENTATION_MODE] enum VideoPresentationMode { "fullscreen", "picture-in-picture", "inline" };
+[Conditional=VIDEO_PRESENTATION_MODE] enum VideoPresentationMode { "inline", "fullscreen", "picture-in-picture" };
