diff a/modules/javafx.web/src/main/native/Source/WebCore/html/MediaDocument.h b/modules/javafx.web/src/main/native/Source/WebCore/html/MediaDocument.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/MediaDocument.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/MediaDocument.h
@@ -32,30 +32,28 @@
 namespace WebCore {
 
 class MediaDocument final : public HTMLDocument {
     WTF_MAKE_ISO_ALLOCATED(MediaDocument);
 public:
-    static Ref<MediaDocument> create(PAL::SessionID sessionID, Frame* frame, const URL& url)
+    static Ref<MediaDocument> create(Frame* frame, const URL& url)
     {
-        return adoptRef(*new MediaDocument(sessionID, frame, url));
+        return adoptRef(*new MediaDocument(frame, url));
     }
     virtual ~MediaDocument();
 
-    void mediaElementSawUnsupportedTracks();
     void mediaElementNaturalSizeChanged(const IntSize&);
     String outgoingReferrer() const { return m_outgoingReferrer; }
 
 private:
-    MediaDocument(PAL::SessionID, Frame*, const URL&);
+    MediaDocument(Frame*, const URL&);
 
     Ref<DocumentParser> createParser() override;
 
     void defaultEventHandler(Event&) override;
 
     void replaceMediaElementTimerFired();
 
-    Timer m_replaceMediaElementTimer;
     String m_outgoingReferrer;
 };
 
 } // namespace WebCore
 
