diff a/make/modules/jdk.jfr/Copy.gmk b/make/modules/jdk.jfr/Copy.gmk
--- a/make/modules/jdk.jfr/Copy.gmk
+++ b/make/modules/jdk.jfr/Copy.gmk
@@ -25,18 +25,10 @@
 
 include CopyCommon.gmk
 
 ################################################################################
 
-$(eval $(call SetupCopyFiles, COPY_JFR_METADATA, \
-    SRC := $(TOPDIR)/src/hotspot/share/jfr/metadata, \
-    DEST := $(JDK_OUTPUTDIR)/modules/jdk.jfr/jdk/jfr/internal/types, \
-    FILES := metadata.xml \
-))
-
-TARGETS += $(COPY_JFR_METADATA)
-
 JFR_CONF_DIR := $(TOPDIR)/src/jdk.jfr/share/conf/jfr
 $(eval $(call SetupCopyFiles, COPY_JFR_CONF, \
     DEST := $(LIB_DST_DIR)/jfr, \
     FILES := $(wildcard $(JFR_CONF_DIR)/*.jfc), \
     FLATTEN := true, \
