diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/NicosiaPlatformLayer.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/NicosiaPlatformLayer.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/NicosiaPlatformLayer.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/NicosiaPlatformLayer.h
@@ -110,10 +110,11 @@
             union {
                 struct {
                     bool positionChanged : 1;
                     bool anchorPointChanged : 1;
                     bool sizeChanged : 1;
+                    bool boundsOriginChanged : 1;
                     bool transformChanged : 1;
                     bool childrenTransformChanged : 1;
                     bool contentsRectChanged : 1;
                     bool contentsTilingChanged : 1;
                     bool opacityChanged : 1;
@@ -155,10 +156,11 @@
         } flags;
 
         WebCore::FloatPoint position;
         WebCore::FloatPoint3D anchorPoint;
         WebCore::FloatSize size;
+        WebCore::FloatPoint boundsOrigin;
 
         WebCore::TransformationMatrix transform;
         WebCore::TransformationMatrix childrenTransform;
 
         WebCore::FloatRect contentsRect;
@@ -213,10 +215,12 @@
             staging.position = pending.position;
         if (pending.delta.anchorPointChanged)
             staging.anchorPoint = pending.anchorPoint;
         if (pending.delta.sizeChanged)
             staging.size = pending.size;
+        if (pending.delta.boundsOriginChanged)
+            staging.boundsOrigin = pending.boundsOrigin;
 
         if (pending.delta.transformChanged)
             staging.transform = pending.transform;
         if (pending.delta.childrenTransformChanged)
             staging.childrenTransform = pending.childrenTransform;
