diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java
@@ -200,11 +200,11 @@
     private String subjectPrefix() {
         var ret = new StringBuilder();
         var branchName = pr.targetRef();
         var repoName = Path.of(pr.repository().name()).getFileName().toString();
         var useBranchInSubject = bot.branchInSubject().matcher(branchName).matches();
-        var useRepoInSubject = bot.repoInSubject().matcher(repoName).matches();
+        var useRepoInSubject = bot.repoInSubject();
 
         if (useBranchInSubject || useRepoInSubject) {
             ret.append("[");
             if (useRepoInSubject) {
                 ret.append(repoName);
