diff a/src/java.desktop/unix/native/libawt_xawt/xawt/XWindow.c b/src/java.desktop/unix/native/libawt_xawt/xawt/XWindow.c
--- a/src/java.desktop/unix/native/libawt_xawt/xawt/XWindow.c
+++ b/src/java.desktop/unix/native/libawt_xawt/xawt/XWindow.c
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2002, 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2002, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -44,11 +44,10 @@
 
 #include "sun_awt_X11_XWindow.h"
 
 #include "awt_p.h"
 #include "awt_GraphicsEnv.h"
-#include "awt_AWTEvent.h"
 
 #define XK_KATAKANA
 #include <X11/keysym.h>     /* standard X keysyms */
 #include <X11/DECkeysym.h>  /* DEC vendor-specific */
 #include <X11/Sunkeysym.h>  /* Sun vendor-specific */
