diff a/cli/src/main/java/org/openjdk/skara/cli/GitPr.java b/cli/src/main/java/org/openjdk/skara/cli/GitPr.java
--- a/cli/src/main/java/org/openjdk/skara/cli/GitPr.java
+++ b/cli/src/main/java/org/openjdk/skara/cli/GitPr.java
@@ -504,10 +504,14 @@
                   .optional(),
             Switch.shortcut("")
                   .fullname("no-draft")
                   .helptext("Hide all pull requests in draft state")
                   .optional(),
+            Switch.shortcut("")
+                  .fullname("draft")
+                  .helptext("Create a pull request in draft state")
+                  .optional(),
             Switch.shortcut("")
                   .fullname("ignore-workspace")
                   .helptext("Ignore local changes in worktree and staging area when creating pull request")
                   .optional(),
             Switch.shortcut("")
@@ -1058,11 +1062,12 @@
                             .collect(Collectors.toList());
             } else {
                 body = Collections.emptyList();
             }
 
-            var pr = remoteRepo.createPullRequest(parentRepo, targetBranch, currentBranch.name(), title, body);
+            var isDraft = getSwitch("draft", "create", arguments);
+            var pr = remoteRepo.createPullRequest(parentRepo, targetBranch, currentBranch.name(), title, body, isDraft);
             var assigneesOption = getOption("assignees", "create", arguments);
             if (assigneesOption != null) {
                 var usernames = Arrays.asList(assigneesOption.split(","));
                 var assignees = usernames.stream()
                                          .map(u -> host.user(u))
@@ -1389,20 +1394,31 @@
                 exit("error: missing pull request identifier");
             }
 
             var remoteRepo = getHostedRepositoryFor(uri, repo, host);
             var pr = remoteRepo.pullRequest(prId.asString());
+
+            var setDraft = getSwitch("draft", "set", arguments);
+            if (!pr.isDraft() && setDraft) {
+                exit("error: cannot transition non-draft pull request to draft");
+            }
+
             var assigneesOption = getOption("assignees", "set", arguments);
             if (assigneesOption != null) {
                 var usernames = Arrays.asList(assigneesOption.split(","));
                 var assignees = usernames.stream()
                     .map(u -> host.user(u))
                     .filter(Optional::isPresent)
                     .map(Optional::get)
                     .collect(Collectors.toList());
                 pr.setAssignees(assignees);
             }
+
+            var setNoDraft = getSwitch("no-draft", "set", arguments);
+            if (setNoDraft) {
+                pr.makeNotDraft();
+            }
         } else if (action.equals("status")) {
             String id = pullRequestIdArgument(arguments, repo);
             var pr = getPullRequest(uri, repo, host, id);
             var noDecoration = getSwitch("no-decoration", "status", arguments);
             var decoration = noDecoration ? "" : "Status: ";
