diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/speech/SpeechSynthesisEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/speech/SpeechSynthesisEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/speech/SpeechSynthesisEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/speech/SpeechSynthesisEvent.h
@@ -31,20 +31,20 @@
 
 namespace WebCore {
 
 class SpeechSynthesisEvent : public Event {
 public:
-    static Ref<SpeechSynthesisEvent> create(const AtomicString& type, unsigned charIndex, float elapsedTime, const String& name);
+    static Ref<SpeechSynthesisEvent> create(const AtomString& type, unsigned charIndex, float elapsedTime, const String& name);
 
     unsigned long charIndex() const { return m_charIndex; }
     float elapsedTime() const { return m_elapsedTime; }
     const String& name() const { return m_name; }
 
     virtual EventInterface eventInterface() const { return SpeechSynthesisEventInterfaceType; }
 
 private:
-    SpeechSynthesisEvent(const AtomicString& type, unsigned charIndex, float elapsedTime, const String& name);
+    SpeechSynthesisEvent(const AtomString& type, unsigned charIndex, float elapsedTime, const String& name);
 
     unsigned long m_charIndex;
     float m_elapsedTime;
     String m_name;
 };
