diff a/email/src/main/java/org/openjdk/skara/email/WordWrap.java b/email/src/main/java/org/openjdk/skara/email/WordWrap.java
--- a/email/src/main/java/org/openjdk/skara/email/WordWrap.java
+++ b/email/src/main/java/org/openjdk/skara/email/WordWrap.java
@@ -90,11 +90,11 @@
             var split = split(line.substring(indentation.length()), lineLength);
             if (!split.getValue().isBlank()) {
                 var nextLine = lines.peekFirst();
                 if (nextLine != null) {
                     var nextIndent = indentation(nextLine);
-                    if (!indentation.equals(filterIndent(nextIndent))) {
+                    if (nextLine.isBlank() || !indentation.equals(filterIndent(nextIndent))) {
                         lines.addFirst(filterIndent(indentation) + split.getValue());
                     } else {
                         lines.removeFirst();
                         lines.addFirst(filterIndent(indentation) + split.getValue() + " " + nextLine.substring(indentation.length()));
                     }
