diff a/application/org.openjdk.jmc.greychart/src/main/java/org/openjdk/jmc/greychart/providers/AveragingSampleBuffer.java b/application/org.openjdk.jmc.greychart/src/main/java/org/openjdk/jmc/greychart/providers/AveragingSampleBuffer.java
--- a/application/org.openjdk.jmc.greychart/src/main/java/org/openjdk/jmc/greychart/providers/AveragingSampleBuffer.java
+++ b/application/org.openjdk.jmc.greychart/src/main/java/org/openjdk/jmc/greychart/providers/AveragingSampleBuffer.java
@@ -83,11 +83,11 @@
 	}
 
 	private void addSegmentToBucket(
 		int bucketIndex, double segmentHeigth, double segmentGradient, double segmentWidth) {
 		double segmentArea = (segmentHeigth + segmentGradient * segmentWidth / 2) * segmentWidth;
-		SamplePoint sp = getSamplePoint(bucketIndex);	
+		SamplePoint sp = getSamplePoint(bucketIndex);
 		sp.y += segmentArea;
 		sp.count++;
 	}
 
 	public void fixSamples() {
