diff a/modules/javafx.web/src/main/native/Source/WebCore/loader/FrameLoadRequest.h b/modules/javafx.web/src/main/native/Source/WebCore/loader/FrameLoadRequest.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/loader/FrameLoadRequest.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/loader/FrameLoadRequest.h
@@ -89,12 +89,12 @@
 
     const AtomString& downloadAttribute() const { return m_downloadAttribute; }
 
     InitiatedByMainFrame initiatedByMainFrame() const { return m_initiatedByMainFrame; }
 
-    bool isSystemPreview() const { return m_systemPreviewInfo.isSystemPreview; }
-    const IntRect& systemPreviewRect() const { return m_systemPreviewInfo.systemPreviewRect; }
+    bool isSystemPreview() const { return m_systemPreviewInfo.isPreview; }
+    const SystemPreviewInfo& systemPreviewInfo() const { return m_systemPreviewInfo; }
 
     void setIsRequestFromClientOrUserInput() { m_isRequestFromClientOrUserInput = true; }
     bool isRequestFromClientOrUserInput() const { return m_isRequestFromClientOrUserInput; }
 
 private:
