diff a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/RuntimeHelper.java.template b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/RuntimeHelper.java.template
--- a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/RuntimeHelper.java.template
+++ b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/RuntimeHelper.java.template
@@ -84,19 +84,19 @@
         } catch (Throwable ex) {
             throw new AssertionError(ex);
         }
     }
 
-    public static final class Scope extends NativeAllocationScope {
+    public static final class CScope extends NativeAllocationScope {
         private final NativeAllocationScope impl;
         private final List<MemorySegment> segments = new ArrayList<>();
 
-        public Scope() {
+        public CScope() {
             impl = NativeAllocationScope.unboundedScope();
         }
 
-        public Scope(long size) {
+        public CScope(long size) {
             impl = NativeAllocationScope.boundedScope(size);
         }
 
         @Override
         public OptionalLong byteSize() {
