diff a/src/hotspot/share/ci/ciInstanceKlass.cpp b/src/hotspot/share/ci/ciInstanceKlass.cpp
--- a/src/hotspot/share/ci/ciInstanceKlass.cpp
+++ b/src/hotspot/share/ci/ciInstanceKlass.cpp
@@ -799,11 +799,11 @@
           _out->print("%s", klass_name);
         }
       } else if (value->is_array()) {
         typeArrayOop ta = (typeArrayOop)value;
         _out->print("%d", ta->length());
-        if (value->is_objArray() || value->is_valueArray()) {
+        if (value->is_objArray() || value->is_inlineArray()) {
           objArrayOop oa = (objArrayOop)value;
           const char* klass_name  = value->klass()->name()->as_quoted_ascii();
           _out->print(" %s", klass_name);
         }
       } else {
