diff a/webrev/src/main/java/org/openjdk/skara/webrev/AddedFileView.java b/webrev/src/main/java/org/openjdk/skara/webrev/AddedFileView.java
--- a/webrev/src/main/java/org/openjdk/skara/webrev/AddedFileView.java
+++ b/webrev/src/main/java/org/openjdk/skara/webrev/AddedFileView.java
@@ -25,21 +25,26 @@
 import org.openjdk.skara.vcs.*;
 
 import java.io.*;
 import java.nio.file.*;
 import java.util.List;
+import java.util.stream.Collectors;
 
 class AddedFileView implements FileView {
     private final Patch patch;
     private final Path out;
+    private final List<CommitMetadata> commits;
+    private final MetadataFormatter formatter;
     private final List<String> newContent;
     private final byte[] binaryContent;
     private final WebrevStats stats;
 
-    public AddedFileView(ReadOnlyRepository repo, Hash head, Patch patch, Path out) throws IOException {
+    public AddedFileView(ReadOnlyRepository repo, Hash base, Hash head, List<CommitMetadata> commits, MetadataFormatter formatter, Patch patch, Path out) throws IOException {
         this.patch = patch;
         this.out = out;
+        this.commits = commits;
+        this.formatter = formatter;
         if (patch.isTextual()) {
             binaryContent = null;
             if (head == null) {
                 newContent = Files.readAllLines(repo.root().resolve(patch.target().path().get()));
             } else {
@@ -102,10 +107,15 @@
 
         w.write("<p>\n");
 
         if (patch.isTextual()) {
             w.write("<blockquote>\n");
+            w.write("  <pre>\n");
+            w.write(commits.stream()
+                           .map(formatter::format)
+                           .collect(Collectors.joining("\n")));
+            w.write("  </pre>\n");
             w.write("  <span class=\"stat\">\n");
             w.write(stats.toString());
             w.write("  </span>");
             w.write("</blockquote>\n");
         }
