diff a/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java b/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
--- a/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
+++ b/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
@@ -170,20 +170,26 @@
                     next = patches.get(modified.get(i + 1)).target().path().get();
                 }
                 navigations.addLast(new Navigation(prev, next));
             }
 
+            var headHash = head == null ? repository.head() : head;
             var fileViews = new ArrayList<FileView>();
+            var formatter = new MetadataFormatter();
             for (var patch : patches) {
                 var status = patch.status();
+                var path = status.isDeleted() ?
+                    patch.source().path().get() :
+                    patch.target().path().get();
+                var commits = repository.commitMetadata(tailEnd, headHash, List.of(path));
                 if (status.isModified() || status.isRenamed() || status.isCopied()) {
                     var nav = navigations.removeFirst();
-                    fileViews.add(new ModifiedFileView(repository, tailEnd, head, patch, output, nav));
+                    fileViews.add(new ModifiedFileView(repository, tailEnd, head, commits, formatter, patch, output, nav));
                 } else if (status.isAdded()) {
-                    fileViews.add(new AddedFileView(repository, head, patch, output));
+                    fileViews.add(new AddedFileView(repository, tailEnd, head, commits, formatter, patch, output));
                 } else if (status.isDeleted()) {
-                    fileViews.add(new RemovedFileView(repository, tailEnd, patch, output));
+                    fileViews.add(new RemovedFileView(repository, tailEnd, head, commits, formatter, patch, output));
                 }
             }
 
             var total = fileViews.stream().map(FileView::stats).mapToInt(WebrevStats::total).sum();
             var stats = new WebrevStats(diff.added(), diff.removed(), diff.modified(), total);
