diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/graphics/filters/SVGFEImage.cpp b/modules/javafx.web/src/main/native/Source/WebCore/svg/graphics/filters/SVGFEImage.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/graphics/filters/SVGFEImage.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/graphics/filters/SVGFEImage.cpp
@@ -91,12 +91,12 @@
 {
     RenderElement* renderer = referencedRenderer();
     if (!m_image && !renderer)
         return;
 
-    // FEImage results are always in ColorSpaceSRGB
-    setResultColorSpace(ColorSpaceSRGB);
+    // FEImage results are always in ColorSpace::SRGB
+    setResultColorSpace(ColorSpace::SRGB);
 
     ImageBuffer* resultImage = createImageBufferResult();
     if (!resultImage)
         return;
 
