diff a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/OutputFactory.java b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/OutputFactory.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/OutputFactory.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/OutputFactory.java
@@ -26,12 +26,10 @@
 
 import jdk.incubator.foreign.*;
 import jdk.incubator.jextract.Declaration;
 import jdk.incubator.jextract.Type;
 import jdk.incubator.jextract.Type.Primitive;
-import jdk.internal.foreign.abi.SharedUtils;
-
 import javax.tools.JavaFileObject;
 import javax.tools.SimpleJavaFileObject;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.lang.constant.ClassDesc;
@@ -119,11 +117,11 @@
         this.constantHelper = constantHelper;
     }
 
     private static String getCLangConstantsHolder() {
         String prefix = "jdk.incubator.foreign.CSupport.";
-        String abi = SharedUtils.getSystemLinker().name();
+        String abi = CSupport.getSystemLinker().name();
         switch (abi) {
             case CSupport.SysV.NAME:
                 return prefix + "SysV";
             case CSupport.Win64.NAME:
                 return prefix + "Win64";
