diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/spi/darwin/dyldSPI.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/spi/darwin/dyldSPI.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/spi/darwin/dyldSPI.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/spi/darwin/dyldSPI.h
@@ -39,18 +39,26 @@
 
 #ifndef DYLD_IOS_VERSION_12_0
 #define DYLD_IOS_VERSION_12_0 0x000C0000
 #endif
 
+#ifndef DYLD_IOS_VERSION_13_0
+#define DYLD_IOS_VERSION_13_0 0x000D0000
+#endif
+
 #ifndef DYLD_MACOSX_VERSION_10_13
 #define DYLD_MACOSX_VERSION_10_13 0x000A0D00
 #endif
 
 #ifndef DYLD_MACOSX_VERSION_10_14
 #define DYLD_MACOSX_VERSION_10_14 0x000A0E00
 #endif
 
+#ifndef DYLD_MACOSX_VERSION_10_15
+#define DYLD_MACOSX_VERSION_10_15 0x000A0F00
+#endif
+
 #else
 
 #define DYLD_IOS_VERSION_3_0 0x00030000
 #define DYLD_IOS_VERSION_4_2 0x00040200
 #define DYLD_IOS_VERSION_5_0 0x00050000
@@ -59,15 +67,17 @@
 #define DYLD_IOS_VERSION_9_0 0x00090000
 #define DYLD_IOS_VERSION_10_0 0x000A0000
 #define DYLD_IOS_VERSION_11_0 0x000B0000
 #define DYLD_IOS_VERSION_11_3 0x000B0300
 #define DYLD_IOS_VERSION_12_0 0x000C0000
+#define DYLD_IOS_VERSION_13_0 0x000D0000
 
 #define DYLD_MACOSX_VERSION_10_11 0x000A0B00
 #define DYLD_MACOSX_VERSION_10_12 0x000A0C00
 #define DYLD_MACOSX_VERSION_10_13 0x000A0D00
 #define DYLD_MACOSX_VERSION_10_14 0x000A0E00
+#define DYLD_MACOSX_VERSION_10_15 0x000A0F00
 
 #endif
 
 WTF_EXTERN_C_BEGIN
 
