diff a/src/hotspot/share/jfr/jni/jfrJavaCall.hpp b/src/hotspot/share/jfr/jni/jfrJavaCall.hpp
--- a/src/hotspot/share/jfr/jni/jfrJavaCall.hpp
+++ b/src/hotspot/share/jfr/jni/jfrJavaCall.hpp
@@ -26,10 +26,11 @@
 #define SHARE_JFR_JNI_JFRJAVACALL_HPP
 
 #include "jni.h"
 #include "jfr/utilities/jfrAllocation.hpp"
 #include "utilities/exceptions.hpp"
+#include "utilities/globalDefinitions.hpp"
 
 class JavaCallArguments;
 class JavaThread;
 class JavaValue;
 class Klass;
@@ -45,15 +46,15 @@
   Klass* klass() const;
   void set_klass(const char* klass_name, TRAPS);
   void set_klass(const Klass* klass);
 
   Symbol* name() const;
-  void set_name(const char* name, TRAPS);
+  void set_name(const char* name);
   void set_name(const Symbol* name);
 
   Symbol* signature() const;
-  void set_signature(const char* signature, TRAPS);
+  void set_signature(const char* signature);
   void set_signature(const Symbol* signature);
 
   int array_length() const;
   void set_array_length(int length);
 
@@ -84,12 +85,11 @@
     JavaValue _storage[SIZE];
     int _storage_index;
     int _java_stack_slots;
 
     Parameters();
-    Parameters(const Parameters&); // no impl
-    Parameters& operator=(const Parameters&); // no impl
+    NONCOPYABLE(Parameters);
 
     void push(const JavaValue& value);
     void push_large(const JavaValue& value);
 
     void push_oop(const oop obj);
