diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/geolocation/PositionErrorCallback.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/geolocation/PositionErrorCallback.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/geolocation/PositionErrorCallback.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/geolocation/PositionErrorCallback.idl
@@ -22,6 +22,6 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 [
     Conditional=GEOLOCATION,
-] callback PositionErrorCallback = void (PositionError error);
+] callback PositionErrorCallback = void (GeolocationPositionError error);
