diff a/src/jdk.hotspot.agent/linux/native/libsaproc/ps_proc.c b/src/jdk.hotspot.agent/linux/native/libsaproc/ps_proc.c
--- a/src/jdk.hotspot.agent/linux/native/libsaproc/ps_proc.c
+++ b/src/jdk.hotspot.agent/linux/native/libsaproc/ps_proc.c
@@ -277,15 +277,16 @@
       if (process_doesnt_exist(pid)) {
         print_debug("Thread with pid %d does not exist\n", pid);
         return ATTACH_THREAD_DEAD;
       }
     }
-    char buf[200];
-    if (strerror_r(errno, buf, sizeof(buf) == 0)) {
-      snprintf(err_buf, err_buf_len, "ptrace(PTRACE_ATTACH, ..) failed for %d: %s", pid, buf);
-      print_error("%s\n", err_buf);
-    }
+    // To improve portability across platforms and avoid conflicts
+    // between GNU and XSI versions of strerror_r, plain strerror is used.
+    // It's safe because this code is not used in any multithreaded environment.
+    char* msg = strerror(errno);
+    snprintf(err_buf, err_buf_len, "ptrace(PTRACE_ATTACH, ..) failed for %d: %s", pid, msg);
+    print_error("%s\n", err_buf);
     return ATTACH_FAIL;
   } else {
     attach_state_t wait_ret = ptrace_waitpid(pid);
     if (wait_ret == ATTACH_THREAD_DEAD) {
       print_debug("Thread with pid %d does not exist\n", pid);
