diff a/core/org.openjdk.jmc.agent/src/test/java/org/openjdk/jmc/agent/test/util/TestToolkit.java b/core/org.openjdk.jmc.agent/src/test/java/org/openjdk/jmc/agent/test/util/TestToolkit.java
--- a/core/org.openjdk.jmc.agent/src/test/java/org/openjdk/jmc/agent/test/util/TestToolkit.java
+++ b/core/org.openjdk.jmc.agent/src/test/java/org/openjdk/jmc/agent/test/util/TestToolkit.java
@@ -50,12 +50,13 @@
 	private TestToolkit() {
 		throw new UnsupportedOperationException("Not to be instantiated."); //$NON-NLS-1$
 	}
 
 	public static byte[] getByteCode(Class<?> c) throws IOException {
-		InputStream is = c.getClassLoader().getResourceAsStream(c.getName().replace('.', '/') + ".class"); //$NON-NLS-1$
-		return readFully(is, -1, true);
+		try (InputStream is = c.getClassLoader().getResourceAsStream(c.getName().replace('.', '/') + ".class")) { //$NON-NLS-1$
+			return readFully(is, -1, true);
+		}
 	}
 
 	public static byte[] readFully(InputStream is, int length, boolean readAll) throws IOException {
 		byte[] output = {};
 		if (length == -1) {
@@ -104,14 +105,13 @@
 		template = template.replaceAll("%TEST_NAME%", testName); //$NON-NLS-1$
 		return new ByteArrayInputStream(template.getBytes());
 	}
 
 	public static String readTemplate(Class<?> resouceClass, String templateName) throws IOException {
-		InputStream inputStream = resouceClass.getResourceAsStream(templateName); // $NON-NLS-1$
-		String s = readString(inputStream);
-		closeSilently(inputStream);
-		return s;
+		try (InputStream inputStream = resouceClass.getResourceAsStream(templateName)) {
+			return readString(inputStream);
+		}
 	}
 
 	public static String readString(InputStream in) throws IOException {
 		return readString(new BufferedReader(new InputStreamReader(in), 8192));
 	}
