diff a/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java b/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java
--- a/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java
+++ b/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2009, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2009, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -105,12 +105,11 @@
                 "-sourcepath", testSrc,
                 "pkg3");
         checkExit(Exit.OK);
 
         checkOutput("pkg3/package-summary.html", true,
-                "<div class=\"contentContainer\">\n"
-                + "<section class=\"packageDescription\" id=\"package.description\">\n"
+                "<section class=\"package-description\" id=\"package.description\">\n"
                 + "<div class=\"block\"><p>This is the first line. Note the newlines before the &lt;p&gt; is relevant.</div>\n"
                 + "</section>");
 
         checkOutput("pkg3/A.DatatypeFactory.html", true,
                 "<div class=\"block\"><p>\n"
@@ -153,11 +152,11 @@
                 + " <code>IllegalStateException</code>.\n"
                 + " </li>\n"
                 + " </ol></div>");
 
         checkOutput("pkg3/A.ActivationDesc.html", true,
-                "<pre>public class <span class=\"typeNameLabel\">A.ActivationDesc</span>\n"
+                "<pre>public class <span class=\"type-name-label\">A.ActivationDesc</span>\n"
                 + "extends java.lang.Object\n"
                 + "implements java.io.Serializable</pre>\n"
                 + "<div class=\"block\">An activation descriptor contains the information necessary to activate\n"
                 + " an object: <ul>\n"
                 + " <li> the object's group identifier,\n"
@@ -174,11 +173,11 @@
                 + " <code>MarshalledObject</code> in the object's descriptor is passed as the\n"
                 + " second argument to the remote object's constructor for object to use\n"
                 + " during reinitialization/activation.</div>");
 
         checkOutput("pkg3/A.ActivationGroupID.html", true,
-                "<pre>public class <span class=\"typeNameLabel\">A.ActivationGroupID</span>\n"
+                "<pre>public class <span class=\"type-name-label\">A.ActivationGroupID</span>\n"
                 + "extends java.lang.Object\n"
                 + "implements java.io.Serializable</pre>\n"
                 + "<div class=\"block\">The identifier for a registered activation group serves several purposes:\n"
                 + " <ul>\n"
                 + " <li>identifies the group uniquely within the activation system, and\n"
@@ -188,8 +187,8 @@
                 + " The <code>ActivationGroupID</code> is returned from the call to\n"
                 + " <code>ActivationSystem.registerGroup</code> and is used to identify the\n"
                 + " group within the activation system. This group id is passed as one of the\n"
                 + " arguments to the activation group's special constructor when an\n"
                 + " activation group is created/recreated.</div>\n"
-                + "<dl>");
+                + "<dl class=\"notes\">");
     }
 }
