diff a/bot/src/main/java/module-info.java b/bot/src/main/java/module-info.java
--- a/bot/src/main/java/module-info.java
+++ b/bot/src/main/java/module-info.java
@@ -22,10 +22,11 @@
  */
 module org.openjdk.skara.bot {
     requires transitive org.openjdk.skara.host;
     requires transitive org.openjdk.skara.json;
     requires transitive org.openjdk.skara.census;
+    requires org.openjdk.skara.network;
     requires org.openjdk.skara.vcs;
     requires java.logging;
     requires jdk.httpserver;
 
     exports org.openjdk.skara.bot;
