diff a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLDebugShaders.idl b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLDebugShaders.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLDebugShaders.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLDebugShaders.idl
@@ -26,7 +26,7 @@
 [
     Conditional=WEBGL,
     GenerateIsReachable=ImplWebGLRenderingContext,
     NoInterfaceObject,
 ] interface WebGLDebugShaders {
-    DOMString? getTranslatedShaderSource(WebGLShader? shader); // FIXME: The return value and the parameter should not be nullable.
+    DOMString getTranslatedShaderSource(WebGLShader shader);
 };
