diff a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLTransformFeedback.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLTransformFeedback.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLTransformFeedback.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLTransformFeedback.cpp
@@ -44,18 +44,16 @@
 }
 
 WebGLTransformFeedback::WebGLTransformFeedback(WebGLRenderingContextBase& ctx)
     : WebGLSharedObject(ctx)
 {
-    // FIXME: Call createTransformFeedback from GraphicsContext3D.
+    setObject(ctx.graphicsContextGL()->createTransformFeedback());
 }
 
-void WebGLTransformFeedback::deleteObjectImpl(GraphicsContext3D* context3d, Platform3DObject object)
+void WebGLTransformFeedback::deleteObjectImpl(GraphicsContextGLOpenGL* context3d, PlatformGLObject object)
 {
-    UNUSED_PARAM(context3d);
-    UNUSED_PARAM(object);
-    // FIXME: Call deleteTransformFeedback from GraphicsContext3D.
+    context3d->deleteTransformFeedback(object);
 }
 
 }
 
 #endif // ENABLE(WEBGL)
