diff a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Head.java b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Head.java
--- a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Head.java
+++ b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Head.java
@@ -309,11 +309,13 @@
         for (DocPath path : additionalStylesheets) {
             addStylesheet(tree, path);
         }
 
         if (index) {
-            addStylesheet(tree, DocPaths.JQUERY_FILES.resolve(DocPaths.JQUERY_STYLESHEET_FILE));
+            // The order of the addStylesheet(...) calls is important
+            addStylesheet(tree, DocPaths.JQUERY_FILES.resolve(DocPaths.JQUERY_UI_CSS));
+            addStylesheet(tree, DocPaths.JQUERY_OVERRIDES_CSS);
         }
     }
 
     private void addStylesheet(HtmlTree tree, DocPath stylesheet) {
         tree.add(HtmlTree.LINK("stylesheet", "text/css",
@@ -330,12 +332,12 @@
                 mainBodyScript.append("var pathtoroot = ")
                         .appendStringLiteral(ptrPath + "/")
                         .append(";\n")
                         .append("loadScripts(document, 'script');");
             }
-            addJQueryFile(tree, DocPaths.JQUERY_JS_3_4);
             addJQueryFile(tree, DocPaths.JQUERY_JS);
+            addJQueryFile(tree, DocPaths.JQUERY_UI_JS);
         }
         for (Script script : scripts) {
             tree.add(script.asContent());
         }
     }
