diff a/modules/javafx.web/src/main/native/Source/WebCore/loader/HistoryController.cpp b/modules/javafx.web/src/main/native/Source/WebCore/loader/HistoryController.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/loader/HistoryController.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/loader/HistoryController.cpp
@@ -749,11 +749,11 @@
 // We now traverse the frame tree and item tree a second time, loading frames that
 // do have the content the item requests.
 void HistoryController::recursiveGoToItem(HistoryItem& item, HistoryItem* fromItem, FrameLoadType type, ShouldTreatAsContinuingLoad shouldTreatAsContinuingLoad)
 {
     if (!itemsAreClones(item, fromItem)) {
-        m_frame.loader().loadItem(item, type, shouldTreatAsContinuingLoad);
+        m_frame.loader().loadItem(item, fromItem, type, shouldTreatAsContinuingLoad);
         return;
     }
 
     // Just iterate over the rest, looking for frames to navigate.
     for (auto& childItem : item.children()) {
