diff a/modules/javafx.web/src/main/native/Source/WebCore/loader/ResourceTimingInformation.h b/modules/javafx.web/src/main/native/Source/WebCore/loader/ResourceTimingInformation.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/loader/ResourceTimingInformation.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/loader/ResourceTimingInformation.h
@@ -39,16 +39,16 @@
 class ResourceTimingInformation {
 public:
     static bool shouldAddResourceTiming(CachedResource&);
 
     void addResourceTiming(CachedResource&, Document&, ResourceTiming&&);
-    void storeResourceTimingInitiatorInformation(const CachedResourceHandle<CachedResource>&, const AtomicString&, Frame*);
+    void storeResourceTimingInitiatorInformation(const CachedResourceHandle<CachedResource>&, const AtomString&, Frame*);
 
 private:
     enum AlreadyAdded { NotYetAdded, Added };
     struct InitiatorInfo {
-        AtomicString name;
+        AtomString name;
         AlreadyAdded added;
     };
     HashMap<CachedResource*, InitiatorInfo> m_initiatorMap;
 };
 
