diff a/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/AssociationChange.java b/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/AssociationChange.java
--- a/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/AssociationChange.java
+++ b/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/AssociationChange.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2009, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2009, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -91,11 +91,11 @@
         this.association = association;
     }
 
     @Override
     public Association association() {
-        assert association != null;
+        assert event == AssocChangeEvent.CANT_START ? true : association != null;
         return association;
     }
 
     @Override
     public AssocChangeEvent event() {
