diff a/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpNet.java b/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpNet.java
--- a/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpNet.java
+++ b/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpNet.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2009, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2009, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -32,10 +32,11 @@
 import java.nio.channels.AlreadyBoundException;
 import java.util.Set;
 import java.util.HashSet;
 import java.security.AccessController;
 import java.security.PrivilegedAction;
+import sun.net.util.IPAddressUtil;
 import sun.nio.ch.IOUtil;
 import sun.nio.ch.Net;
 import com.sun.nio.sctp.SctpSocketOption;
 import static com.sun.nio.sctp.SctpStandardSocketOptions.*;
 
@@ -167,13 +168,17 @@
 
             Net.checkAddress(addr);
             InetSocketAddress netAddr = (InetSocketAddress)addr;
 
             if (name.equals(SCTP_PRIMARY_ADDR)) {
+                InetAddress inetAddress = netAddr.getAddress();
+                if (inetAddress.isLinkLocalAddress()) {
+                    inetAddress = IPAddressUtil.toScopedAddress(inetAddress);
+                }
                 setPrimAddrOption0(fd,
                                    assocId,
-                                   netAddr.getAddress(),
+                                   inetAddress,
                                    netAddr.getPort());
             } else {
                 setPeerPrimAddrOption0(fd,
                                        assocId,
                                        netAddr.getAddress(),
@@ -333,9 +338,17 @@
     static native void shutdown0(int fd, int assocId);
 
     static native void init();
 
     static {
+        IOUtil.load();   // loads nio & net native libraries
+        java.security.AccessController.doPrivileged(
+                new java.security.PrivilegedAction<Void>() {
+                    public Void run() {
+                        System.loadLibrary("sctp");
+                        return null;
+                    }
+                });
         init();
     }
 }
 
