diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/PathJava.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/PathJava.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/PathJava.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/PathJava.cpp
@@ -26,10 +26,11 @@
 #include "config.h"
 
 #include "Path.h"
 #include "FloatRect.h"
 #include "StrokeStyleApplier.h"
+#include "PlatformContextJava.h"
 #include "PlatformJavaClasses.h"
 #include "NotImplemented.h"
 #include "GraphicsContextJava.h"
 #include "RQRef.h"
 #include "GraphicsContext.h"
