diff a/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java b/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java
--- a/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java
+++ b/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java
@@ -107,11 +107,11 @@
 
         for (var entry : config.get("issuetrackers").fields()) {
             if (entry.value().contains("jira")) {
                 var jira = entry.value().get("jira");
                 var uri = URIBuilder.base(jira.get("url").asString()).build();
-                ret.put(entry.name(), IssueTrackerFactory.createJiraHost(uri, null));
+                ret.put(entry.name(), IssueTracker.from("jira", uri, null, jira.asObject()));
             } else {
                 throw new ConfigurationError("Host " + entry.name());
             }
         }
 
