diff a/src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp b/src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp
--- a/src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp
+++ b/src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp
@@ -150,11 +150,11 @@
 
   // Is the previous value null?
   if (preloaded && not_null) {
 #ifdef ASSERT
     __ cmpdi(CCR0, pre_val, 0);
-    __ asm_assert_ne("null oop not allowed (G1 pre)", 0x321); // Checked by caller.
+    __ asm_assert_ne("null oop not allowed (G1 pre)"); // Checked by caller.
 #endif
   } else {
     __ cmpdi(CCR0, pre_val, 0);
     __ beq(CCR0, filtered);
   }
@@ -221,11 +221,11 @@
 
   // Crosses regions, storing NULL?
   if (not_null) {
 #ifdef ASSERT
     __ cmpdi(CCR0, new_val, 0);
-    __ asm_assert_ne("null oop not allowed (G1 post)", 0x322); // Checked by caller.
+    __ asm_assert_ne("null oop not allowed (G1 post)"); // Checked by caller.
 #endif
   } else {
     __ cmpdi(CCR0, new_val, 0);
     __ beq(CCR0, filtered);
   }
