diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveWorkItem.java
@@ -42,15 +42,17 @@
 import java.util.stream.Collectors;
 
 class ArchiveWorkItem implements WorkItem {
     private final PullRequest pr;
     private final MailingListBridgeBot bot;
+    private final Consumer<RuntimeException> exceptionConsumer;
     private final Logger log = Logger.getLogger("org.openjdk.skara.bots.mlbridge");
 
-    ArchiveWorkItem(PullRequest pr, MailingListBridgeBot bot) {
+    ArchiveWorkItem(PullRequest pr, MailingListBridgeBot bot, Consumer<RuntimeException> exceptionConsumer) {
         this.pr = pr;
         this.bot = bot;
+        this.exceptionConsumer = exceptionConsumer;
     }
 
     @Override
     public String toString() {
         return "ArchiveWorkItem@" + bot.codeRepo().getName() + "#" + pr.getId();
@@ -716,6 +718,11 @@
         var listMails = combineMails(newMails);
         for (var mail : listMails) {
             list.post(mail);
         }
     }
+
+    @Override
+    public void handleRuntimeException(RuntimeException e) {
+        exceptionConsumer.accept(e);
+    }
 }
