diff a/src/hotspot/share/opto/subnode.hpp b/src/hotspot/share/opto/subnode.hpp
--- a/src/hotspot/share/opto/subnode.hpp
+++ b/src/hotspot/share/opto/subnode.hpp
@@ -193,10 +193,11 @@
 // Compare 2 long values, returning condition codes (-1, 0 or 1).
 class CmpLNode : public CmpNode {
 public:
   CmpLNode( Node *in1, Node *in2 ) : CmpNode(in1,in2) {}
   virtual int    Opcode() const;
+  virtual Node *Ideal(PhaseGVN *phase, bool can_reshape);
   virtual const Type *sub( const Type *, const Type * ) const;
 };
 
 //------------------------------CmpULNode---------------------------------------
 // Compare 2 unsigned long values, returning condition codes (-1, 0 or 1).
