diff a/test/jdk/valhalla/valuetypes/ObjectMethods.java b/test/jdk/valhalla/valuetypes/ObjectMethods.java
--- a/test/jdk/valhalla/valuetypes/ObjectMethods.java
+++ b/test/jdk/valhalla/valuetypes/ObjectMethods.java
@@ -206,13 +206,13 @@
         return hc;
     }
 
     static inline class MyValue1 {
         private Point p;
-        private Point? np;
+        private Point.ref np;
 
-        MyValue1(int x, int y, Point? np) {
+        MyValue1(int x, int y, Point.ref np) {
             this.p = Point.makePoint(x, y);
             this.np = np;
         }
     }
 
