diff a/test/jdk/java/net/InetAddress/GetLocalHostWithSM.java b/test/jdk/java/net/InetAddress/GetLocalHostWithSM.java
--- a/test/jdk/java/net/InetAddress/GetLocalHostWithSM.java
+++ b/test/jdk/java/net/InetAddress/GetLocalHostWithSM.java
@@ -22,10 +22,11 @@
  */
 
 /**
  * @test
  * @bug 4531817 8026245
+ * @library /test/lib
  * @summary Inet[46]Address.localHost need doPrivileged
  * @run main/othervm GetLocalHostWithSM
  * @run main/othervm -Djava.net.preferIPv4Stack=true GetLocalHostWithSM
  * @run main/othervm -Djava.net.preferIPv6Addresses=true GetLocalHostWithSM
  * files needed: GetLocalHostWithSM.java, MyPrincipal.java, and policy.file
@@ -36,13 +37,16 @@
 import javax.security.auth.Subject;
 import java.security.Principal;
 import java.security.PrivilegedExceptionAction;
 import java.util.*;
 
+import jdk.test.lib.net.IPSupport;
+
 public class GetLocalHostWithSM {
 
         public static void main(String[] args) throws Exception {
+            IPSupport.throwSkippedExceptionIfNonOperational();
 
             // try setting the local hostname
             InetAddress localHost = InetAddress.getLocalHost();
             if (localHost.isLoopbackAddress()) {
                 System.err.println("Local host name is resolved into a loopback address. Quit now!");
