diff a/modules/javafx.web/src/main/native/Source/WebCore/loader/SinkDocument.cpp b/modules/javafx.web/src/main/native/Source/WebCore/loader/SinkDocument.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/loader/SinkDocument.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/loader/SinkDocument.cpp
@@ -52,11 +52,11 @@
     {
     }
 };
 
 SinkDocument::SinkDocument(Frame& frame, const URL& url)
-    : HTMLDocument(frame.sessionID(), &frame, url)
+    : HTMLDocument(&frame, url)
 {
     setCompatibilityMode(DocumentCompatibilityMode::QuirksMode);
     lockCompatibilityMode();
 }
 
