diff a/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java b/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java
@@ -46,10 +46,11 @@
     void fetchAll() throws IOException;
     void pushAll(URI uri) throws IOException;
     void push(Hash hash, URI uri, String ref, boolean force) throws IOException;
     void push(Branch branch, String remote, boolean setUpstream) throws IOException;
     void clean() throws IOException;
+    void reset(Hash target, boolean hard) throws IOException;
     void revert(Hash parent) throws IOException;
     Repository reinitialize() throws IOException;
     void squash(Hash h) throws IOException;
     void add(List<Path> files) throws IOException;
     default void add(Path... files) throws IOException {
@@ -170,11 +171,11 @@
 
         return localRepo;
     }
 
     static Repository clone(URI from) throws IOException {
-        var to = Path.of(from.getPath()).getFileName();
+        var to = Path.of(from).getFileName();
         if (to.toString().endsWith(".git")) {
             to = Path.of(to.toString().replace(".git", ""));
         }
         return clone(from, to);
     }
