diff a/test/hotspot/jtreg/gc/logging/TestMetaSpaceLog.java b/test/hotspot/jtreg/gc/logging/TestMetaSpaceLog.java
--- a/test/hotspot/jtreg/gc/logging/TestMetaSpaceLog.java
+++ b/test/hotspot/jtreg/gc/logging/TestMetaSpaceLog.java
@@ -37,19 +37,21 @@
 
 /*
  * @test TestMetaSpaceLog
  * @bug 8211123
  * @summary Ensure that the Metaspace is updated in the log
- * @requires vm.gc=="null"
  * @key gc
  * @library /test/lib
  * @modules java.base/jdk.internal.misc
  *          java.management
+ * @requires vm.gc != "Epsilon"
+ * @requires vm.gc != "Z"
+ * @requires vm.gc != "Shenandoah"
  *
  * @compile TestMetaSpaceLog.java
  * @run driver ClassFileInstaller sun.hotspot.WhiteBox
- * @run main gc.logging.TestMetaSpaceLog
+ * @run driver gc.logging.TestMetaSpaceLog
  */
 
 public class TestMetaSpaceLog {
   private static Pattern metaSpaceRegexp;
 
@@ -57,14 +59,11 @@
     // Do this once here.
     metaSpaceRegexp = Pattern.compile(".*Metaspace: ([0-9]+).*->([0-9]+).*");
   }
 
   public static void main(String[] args) throws Exception {
-    testMetaSpaceUpdate("UseParallelGC");
-    testMetaSpaceUpdate("UseG1GC");
-    testMetaSpaceUpdate("UseConcMarkSweepGC");
-    testMetaSpaceUpdate("UseSerialGC");
+    testMetaSpaceUpdate();
   }
 
   private static void verifyContainsMetaSpaceUpdate(OutputAnalyzer output) {
     // At least one metaspace line from GC should show GC being collected.
     boolean foundCollectedMetaSpace = output.asLines().stream()
@@ -79,18 +78,17 @@
     long before = Long.parseLong(m.group(1));
     long after  = Long.parseLong(m.group(2));
     return before > after;
   }
 
-  private static void testMetaSpaceUpdate(String gcFlag) throws Exception {
+  private static void testMetaSpaceUpdate() throws Exception {
     // Propagate test.src for the jar file.
     String testSrc= "-Dtest.src=" + System.getProperty("test.src", ".");
 
-    System.err.println("Testing with GC Flag: " + gcFlag);
     ProcessBuilder pb =
       ProcessTools.createJavaProcessBuilder(
-          "-XX:+" + gcFlag,
+          true,
           "-Xlog:gc*",
           "-Xbootclasspath/a:.",
           "-XX:+UnlockDiagnosticVMOptions",
           "-XX:+WhiteBoxAPI",
           "-Xmx1000M",
