diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/ContentType.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/ContentType.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/ContentType.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/ContentType.h
@@ -31,12 +31,12 @@
 
 namespace WebCore {
 
 class ContentType {
 public:
-    explicit ContentType(String&& type);
-    explicit ContentType(const String& type);
+    WEBCORE_EXPORT explicit ContentType(String&& type);
+    WEBCORE_EXPORT explicit ContentType(const String& type);
     ContentType() = default;
 
     static const String& codecsParameter();
     static const String& profilesParameter();
 
@@ -47,10 +47,27 @@
     const String& raw() const { return m_type; }
     bool isEmpty() const { return m_type.isEmpty(); }
 
     String toJSONString() const;
 
+    template<class Encoder>
+    void encode(Encoder& encoder) const
+    {
+        encoder << m_type;
+    }
+
+    template <class Decoder>
+    static Optional<ContentType> decode(Decoder& decoder)
+    {
+        Optional<String> type;
+        decoder >> type;
+        if (!type)
+            return WTF::nullopt;
+
+        return { ContentType(WTFMove(*type)) };
+    }
+
 private:
     String m_type;
 };
 
 } // namespace WebCore
