diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/LocalizedStrings.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/LocalizedStrings.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/LocalizedStrings.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/LocalizedStrings.h
@@ -66,11 +66,11 @@
     String contextMenuItemTagGoBack();
     String contextMenuItemTagGoForward();
     String contextMenuItemTagStop();
     String contextMenuItemTagReload();
     String contextMenuItemTagCut();
-    String contextMenuItemTagPaste();
+    WEBCORE_EXPORT String contextMenuItemTagPaste();
 #if PLATFORM(GTK)
     String contextMenuItemTagDelete();
     String contextMenuItemTagInputMethods();
     String contextMenuItemTagUnicode();
     String contextMenuItemTagUnicodeInsertLRMMark();
@@ -184,10 +184,13 @@
     String AXTimeFieldText();
     String AXDateTimeFieldText();
     String AXMonthFieldText();
     String AXNumberFieldText();
     String AXWeekFieldText();
+    String AXARIAContentGroupText(const String& ariaType);
+    String AXHorizontalRuleDescriptionText();
+    String AXMarkText();
 
     String AXButtonActionVerb();
     String AXRadioButtonActionVerb();
     String AXTextFieldActionVerb();
     String AXCheckedCheckBoxActionVerb();
@@ -201,13 +204,10 @@
     // weekFormatInLDML() returns week and year format in LDML, Unicode
     // technical standard 35, Locale Data Markup Language, e.g. "'Week' ww, yyyy"
     String weekFormatInLDML();
 #endif
 #if PLATFORM(COCOA)
-    String AXARIAContentGroupText(const String& ariaType);
-    String AXHorizontalRuleDescriptionText();
-    String AXMarkText();
 #if ENABLE(METER_ELEMENT)
     String AXMeterGaugeRegionOptimumText();
     String AXMeterGaugeRegionSuboptimalText();
     String AXMeterGaugeRegionLessGoodText();
 #endif
