diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/PopupMenu.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/PopupMenu.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/PopupMenu.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/PopupMenu.h
@@ -29,11 +29,11 @@
 class IntRect;
 
 class PopupMenu : public RefCounted<PopupMenu> {
 public:
     virtual ~PopupMenu() = default;
-    virtual void show(const IntRect&, FrameView*, int index) = 0;
+    virtual void show(const IntRect&, FrameView*, int selectedIndex) = 0;
     virtual void hide() = 0;
     virtual void updateFromElement() = 0;
     virtual void disconnectClient() = 0;
 };
 
