diff a/test/jdk/java/net/InetAddress/CheckJNI.java b/test/jdk/java/net/InetAddress/CheckJNI.java
--- a/test/jdk/java/net/InetAddress/CheckJNI.java
+++ b/test/jdk/java/net/InetAddress/CheckJNI.java
@@ -81,10 +81,14 @@
     }
 
     static void testDatagram(InetAddress ia) throws Exception {
         DatagramSocket s1 = new DatagramSocket(0, ia);
         DatagramSocket s2 = new DatagramSocket(0, ia);
+        System.out.println("s1: local address=" + s1.getLocalAddress()
+                            + ", local port=" + s1.getLocalPort());
+        System.out.println("s2: local address=" + s2.getLocalAddress()
+                            + ", local port=" + s2.getLocalPort());
 
         DatagramPacket p1 = new DatagramPacket (
                 "hello world".getBytes(),
                 0, "hello world".length(), s2.getLocalAddress(),
                 s2.getLocalPort()
