diff a/test/jdk/java/net/HttpURLConnection/NoProxyTest.java b/test/jdk/java/net/HttpURLConnection/NoProxyTest.java
--- a/test/jdk/java/net/HttpURLConnection/NoProxyTest.java
+++ b/test/jdk/java/net/HttpURLConnection/NoProxyTest.java
@@ -22,23 +22,26 @@
  */
 
  /*
  * @test
  * @bug 8144008
+ * @library /test/lib
  * @summary Setting NO_PROXY on HTTP URL connections does not stop proxying
  * @run main/othervm NoProxyTest
  */
 
 import java.io.IOException;
+import java.net.InetAddress;
 import java.net.MalformedURLException;
 import java.net.Proxy;
 import java.net.ProxySelector;
 import java.net.SocketAddress;
 import java.net.URI;
 import java.net.URL;
 import java.net.URLConnection;
 import java.util.List;
+import jdk.test.lib.net.URIBuilder;
 
 public class NoProxyTest {
 
     static class NoProxyTestSelector extends ProxySelector {
         @Override
@@ -50,11 +53,16 @@
     }
 
     public static void main(String args[]) throws MalformedURLException {
         ProxySelector.setDefault(new NoProxyTestSelector());
 
-        URL url = URI.create("http://127.0.0.1/").toURL();
+        URL url = URIBuilder.newBuilder()
+            .scheme("http")
+            .loopback()
+            .path("/")
+            .toURLUnchecked();
+        System.out.println("URL: " + url);
         URLConnection connection;
         try {
             connection = url.openConnection(Proxy.NO_PROXY);
             connection.connect();
         } catch (IOException ignore) {
