diff a/src/utils/hsdis/hsdis.c b/src/utils/hsdis/hsdis.c
--- a/src/utils/hsdis/hsdis.c
+++ b/src/utils/hsdis/hsdis.c
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2008, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2008, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * The Universal Permissive License (UPL), Version 1.0
  *
  * Subject to the condition set forth below, permission is hereby granted to
@@ -566,11 +566,16 @@
 
   dinfo->flavour = bfd_get_flavour(abfd);
   dinfo->arch = bfd_get_arch(abfd);
   dinfo->mach = bfd_get_mach(abfd);
   dinfo->disassembler_options = disassembler_options;
+#if BFD_VERSION >= 234000000
+  /* bfd_octets_per_byte() has 2 args since binutils 2.34 */
+  dinfo->octets_per_byte = bfd_octets_per_byte (abfd, NULL);
+#else
   dinfo->octets_per_byte = bfd_octets_per_byte (abfd);
+#endif
   dinfo->skip_zeroes = sizeof(void*) * 2;
   dinfo->skip_zeroes_at_end = sizeof(void*)-1;
   dinfo->disassembler_needs_relocs = FALSE;
 
   if (bfd_big_endian(abfd))
