diff a/src/java.base/share/classes/jdk/internal/misc/UnsafeConstants.java b/src/java.base/share/classes/jdk/internal/misc/UnsafeConstants.java
--- a/src/java.base/share/classes/jdk/internal/misc/UnsafeConstants.java
+++ b/src/java.base/share/classes/jdk/internal/misc/UnsafeConstants.java
@@ -43,11 +43,11 @@
  * is required to prevent the fields from being considered constant
  * variables, so the field values will be not be compiled directly into
  * any class that uses them.
  */
 
-final class UnsafeConstants {
+public final class UnsafeConstants {
 
     /**
      * This constructor is private because the class is not meant to
      * be instantiated.
      */
@@ -110,13 +110,23 @@
      * The actual value for this field is injected by the JVM.
      */
 
     static final int DATA_CACHE_LINE_FLUSH_SIZE;
 
+    /**
+     * The exact log (base 2) of the number of lines in the Scoped cache.
+     *
+     * @implNote
+     * The actual value for this field is injected by the JVM.
+     */
+
+    public static final int SCOPED_CACHE_SHIFT;
+
     static {
         ADDRESS_SIZE0 = 0;
         PAGE_SIZE = 0;
         BIG_ENDIAN = false;
         UNALIGNED_ACCESS = false;
         DATA_CACHE_LINE_FLUSH_SIZE = 0;
+        SCOPED_CACHE_SHIFT = -1;
     }
 }
