diff a/modules/javafx.web/src/main/native/Source/WebCore/css/parser/CSSTokenizer.h b/modules/javafx.web/src/main/native/Source/WebCore/css/parser/CSSTokenizer.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/parser/CSSTokenizer.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/parser/CSSTokenizer.h
@@ -43,11 +43,11 @@
 
 class CSSTokenizer {
     WTF_MAKE_NONCOPYABLE(CSSTokenizer);
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    CSSTokenizer(const String&);
+    explicit CSSTokenizer(const String&);
     CSSTokenizer(const String&, CSSParserObserverWrapper&); // For the inspector
 
     CSSParserTokenRange tokenRange() const;
     unsigned tokenCount();
 
