diff a/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraIssueTrackerFactory.java b/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraIssueTrackerFactory.java
--- a/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraIssueTrackerFactory.java
+++ b/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraIssueTrackerFactory.java
@@ -23,10 +23,11 @@
 package org.openjdk.skara.issuetracker.jira;
 
 import org.openjdk.skara.host.Credential;
 import org.openjdk.skara.issuetracker.*;
 import org.openjdk.skara.json.JSONObject;
+import org.openjdk.skara.network.URIBuilder;
 
 import java.net.URI;
 
 public class JiraIssueTrackerFactory implements IssueTrackerFactory {
     @Override
@@ -37,9 +38,15 @@
     @Override
     public IssueTracker create(URI uri, Credential credential, JSONObject configuration) {
         if (credential == null) {
             return new JiraHost(uri);
         } else {
-            throw new RuntimeException("authentication not implemented yet");
+            if (credential.username().startsWith("https://")) {
+                var vaultUrl = URIBuilder.base(credential.username()).build();
+                var jiraVault = new JiraVault(vaultUrl, credential.password());
+                return new JiraHost(uri, jiraVault);
+            } else {
+                throw new RuntimeException("basic authentication not implemented yet");
+            }
         }
     }
 }
