diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGAnimatedPropertyDescription.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGAnimatedPropertyDescription.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGAnimatedPropertyDescription.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGAnimatedPropertyDescription.h
@@ -20,11 +20,11 @@
  */
 
 #pragma once
 
 #include <wtf/HashMap.h>
-#include <wtf/text/AtomicString.h>
+#include <wtf/text/AtomString.h>
 
 namespace WebCore {
 
 class SVGElement;
 
@@ -45,11 +45,11 @@
     bool isHashTableDeletedValue() const
     {
         return m_element == reinterpret_cast<SVGElement*>(-1);
     }
 
-    SVGAnimatedPropertyDescription(SVGElement* element, const AtomicString& attributeName)
+    SVGAnimatedPropertyDescription(SVGElement* element, const AtomString& attributeName)
         : m_element(element)
         , m_attributeName(attributeName.impl())
     {
         ASSERT(m_element);
         ASSERT(m_attributeName);
@@ -59,11 +59,11 @@
     {
         return m_element == other.m_element && m_attributeName == other.m_attributeName;
     }
 
     SVGElement* m_element;
-    AtomicStringImpl* m_attributeName;
+    AtomStringImpl* m_attributeName;
 };
 
 struct SVGAnimatedPropertyDescriptionHash {
     static unsigned hash(const SVGAnimatedPropertyDescription& key)
     {
