diff a/test/hotspot/jtreg/runtime/cds/appcds/test-classes/JimageClassPackage.java b/test/hotspot/jtreg/runtime/cds/appcds/test-classes/JimageClassPackage.java
--- a/test/hotspot/jtreg/runtime/cds/appcds/test-classes/JimageClassPackage.java
+++ b/test/hotspot/jtreg/runtime/cds/appcds/test-classes/JimageClassPackage.java
@@ -28,12 +28,11 @@
         // The following classes are archived. See runtime/AppCDS/Package.java.
         //     java.util.Dictionary (testcase 0),
         //     sun.tools.javac.Main (testcase 1),
         //     jdk.nio.zipfs.ZipInfo (testcase 2),
         //     java.net.URL (testcase 3),
-        //     sun.rmi.rmic.Main (testcase 4),
-        //     com.sun.jndi.dns.DnsName (testcase 5)
+        //     com.sun.jndi.dns.DnsName (testcase 4)
         String testcases[][] =
             {{"Loading shared boot module class first", "java.util",
               "java.util.Dictionary", "java.util.ServiceConfigurationError"},
 
              {"Loading shared app module class first", "sun.tools.javac",
@@ -43,13 +42,10 @@
               "jdk.nio.zipfs.ZipInfo", "jdk.nio.zipfs.ZipPath"},
 
              {"Loading non-shared boot module class first", "java.net",
               "java.net.HttpCookie", "java.net.URL"},
 
-             {"Loading non-shared app module class first", "sun.rmi.rmic",
-              "sun.rmi.rmic.RMIGenerator", "sun.rmi.rmic.Main"},
-
              {"Loading non-shared ext module class first", "com.sun.jndi.dns",
               "com.sun.jndi.dns.Resolver", "com.sun.jndi.dns.DnsName"}};
 
         JimageClassPackage test = new JimageClassPackage();
         for (int i = 0; i < testcases.length; i++) {
