diff a/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java b/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
--- a/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
+++ b/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
@@ -41,11 +41,11 @@
      * Public API                                                          *
      *                                                                     *
      **********************************************************************/
 
     /**
-     * Creates a default {@code TableFocusModel}.
+     * Constructor for subclasses to call.
      */
     public TableFocusModel() {
     }
 
     /**
diff a/modules/javafx.controls/src/main/java/javafx/scene/control/TableSelectionModel.java b/modules/javafx.controls/src/main/java/javafx/scene/control/TableSelectionModel.java
--- a/modules/javafx.controls/src/main/java/javafx/scene/control/TableSelectionModel.java
+++ b/modules/javafx.controls/src/main/java/javafx/scene/control/TableSelectionModel.java
@@ -36,11 +36,11 @@
  * @since JavaFX 8.0
  */
 public abstract class TableSelectionModel<T> extends MultipleSelectionModelBase<T> {
 
     /**
-     * Creates a default {@code TableSelectionModel}.
+     * Constructor for subclasses to call.
      */
     public TableSelectionModel() {
     }
 
     /**
diff a/modules/javafx.graphics/src/main/java/javafx/application/Preloader.java b/modules/javafx.graphics/src/main/java/javafx/application/Preloader.java
--- a/modules/javafx.graphics/src/main/java/javafx/application/Preloader.java
+++ b/modules/javafx.graphics/src/main/java/javafx/application/Preloader.java
@@ -113,11 +113,11 @@
         String prop = AccessController.doPrivileged((PrivilegedAction<String>) () -> System.getProperty("line.separator"));
         lineSeparator = prop != null ? prop : "\n";
     }
 
     /**
-     * Creates a default {@code Preloader}.
+     * Constructor for subclasses to call.
      */
     public Preloader() {
     }
 
     /**
diff a/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java b/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java
--- a/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java
+++ b/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java
@@ -144,11 +144,11 @@
  * @since JavaFX 8.0
  */
 public abstract class ScheduledService<V> extends Service<V> {
 
     /**
-     * Creates a default {@code ScheduledService}.
+     * Constructor for subclasses to call.
      */
     public ScheduledService() {
     }
 
     /**
diff a/modules/javafx.graphics/src/main/java/javafx/css/PseudoClass.java b/modules/javafx.graphics/src/main/java/javafx/css/PseudoClass.java
--- a/modules/javafx.graphics/src/main/java/javafx/css/PseudoClass.java
+++ b/modules/javafx.graphics/src/main/java/javafx/css/PseudoClass.java
@@ -73,11 +73,11 @@
  * @since JavaFX 8.0
  */
 public abstract class PseudoClass {
 
     /**
-     * Creates a default {@code PseudoClass}.
+     * Constructor for subclasses to call.
      */
     public PseudoClass() {
     }
 
     /**
diff a/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java b/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java
--- a/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java
+++ b/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java
@@ -85,11 +85,11 @@
  * @since JavaFX 8.0
  */
 public class StyleConverter<F, T> {
 
     /**
-     * Creates a default {@code StyleConverter}.
+     * Creates a {@code StyleConverter}.
      */
     public StyleConverter() {
     }
 
     /**
diff a/modules/javafx.graphics/src/main/java/javafx/scene/input/ClipboardContent.java b/modules/javafx.graphics/src/main/java/javafx/scene/input/ClipboardContent.java
--- a/modules/javafx.graphics/src/main/java/javafx/scene/input/ClipboardContent.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/input/ClipboardContent.java
@@ -38,11 +38,11 @@
  * @since JavaFX 2.0
  */
 public class ClipboardContent extends HashMap<DataFormat, Object> {
 
     /**
-     * Creates a default {@code ClipboardContent}.
+     * Creates a {@code ClipboardContent}.
      */
     public ClipboardContent() {
     }
 
     /**
diff a/modules/javafx.graphics/src/main/java/javafx/scene/shape/PathElement.java b/modules/javafx.graphics/src/main/java/javafx/scene/shape/PathElement.java
--- a/modules/javafx.graphics/src/main/java/javafx/scene/shape/PathElement.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/shape/PathElement.java
@@ -70,11 +70,11 @@
      * is attached to.
      */
     WeakReferenceQueue nodes = new WeakReferenceQueue();
 
     /**
-     * Creates a default {@code PathElement}.
+     * Constructor for subclasses to call.
      */
     public PathElement() {
     }
 
     void addNode(final Node n) {
diff a/modules/javafx.graphics/src/main/java/javafx/scene/transform/Transform.java b/modules/javafx.graphics/src/main/java/javafx/scene/transform/Transform.java
--- a/modules/javafx.graphics/src/main/java/javafx/scene/transform/Transform.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/transform/Transform.java
@@ -126,11 +126,11 @@
             }
         });
     }
 
     /**
-     * Creates a default {@code Transform}.
+     * Constructor for subclasses to call.
      */
     public Transform() {
     }
 
     /* *************************************************************************
diff a/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java b/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java
--- a/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java
+++ b/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java
@@ -48,11 +48,11 @@
      * The title of the displayed dialog.
      */
     private StringProperty title;
 
     /**
-     * Creates a default {@code DirectoryChooser}.
+     * Creates a {@code DirectoryChooser}.
      */
     public DirectoryChooser() {
     }
 
     public final void setTitle(final String value) {
diff a/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java b/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java
--- a/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java
+++ b/modules/javafx.graphics/src/main/java/javafx/stage/FileChooser.java
@@ -203,11 +203,11 @@
      * The title of the displayed file dialog.
      */
     private StringProperty title;
 
     /**
-     * Creates a default {@code FileChooser}.
+     * Creates a {@code FileChooser}.
      */
     public FileChooser() {
     }
 
     public final void setTitle(final String value) {
diff a/modules/javafx.graphics/src/main/java/javafx/stage/Popup.java b/modules/javafx.graphics/src/main/java/javafx/stage/Popup.java
--- a/modules/javafx.graphics/src/main/java/javafx/stage/Popup.java
+++ b/modules/javafx.graphics/src/main/java/javafx/stage/Popup.java
@@ -37,11 +37,11 @@
  * @since JavaFX 2.0
  */
 public class Popup extends PopupWindow {
 
     /**
-     * Creates a default {@code Popup}.
+     * Creates a {@code Popup}.
      */
     public Popup() {
     }
 
     /**
