diff a/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/FrameLoaderClientJava.h b/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/FrameLoaderClientJava.h
--- a/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/FrameLoaderClientJava.h
+++ b/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/FrameLoaderClientJava.h
@@ -49,15 +49,14 @@
     bool hasWebView() const override;
 
     void makeRepresentation(DocumentLoader*) override;
     void forceLayoutForNonHTML() override;
 
-    Optional<uint64_t> pageID() const final;
-    Optional<uint64_t> frameID() const final;
+    Optional<PageIdentifier> pageID() const final;
+    Optional<FrameIdentifier> frameID() const final;
     PAL::SessionID sessionID() const final;
 
-
     void setCopiesOnScroll() override;
 
     void detachedFromParent2() override;
     void detachedFromParent3() override;
 
@@ -82,20 +81,20 @@
     void dispatchWillClose() override;
     void dispatchDidReceiveIcon() override;
     void dispatchDidStartProvisionalLoad() override;
     void dispatchDidReceiveTitle(const StringWithDirection&) override;
     void dispatchDidCommitLoad(Optional<HasInsecureContent>) override;
-    void dispatchDidFailProvisionalLoad(const ResourceError&) override;
+    void dispatchDidFailProvisionalLoad(const ResourceError&, WillContinueLoading) override;
     void dispatchDidFailLoad(const ResourceError&) override;
     void dispatchDidFinishDocumentLoad() override;
     void dispatchDidFinishLoad() override;
     void dispatchDidClearWindowObjectInWorld(WebCore::DOMWrapperWorld&) override;
 
     Frame* dispatchCreatePage(const NavigationAction&) override;
     void dispatchShow() override;
 
-    void dispatchDecidePolicyForResponse(const ResourceResponse&, const ResourceRequest&, PolicyCheckIdentifier, FramePolicyFunction&&) override;
+    void dispatchDecidePolicyForResponse(const ResourceResponse&, const ResourceRequest&, PolicyCheckIdentifier, const String& downloadAttribute, FramePolicyFunction&&) override;
     void dispatchDecidePolicyForNewWindowAction(const NavigationAction&, const ResourceRequest&, FormState*, const String& frameName, PolicyCheckIdentifier, FramePolicyFunction&&) override;
     void dispatchDecidePolicyForNavigationAction(const NavigationAction&, const ResourceRequest&, const ResourceResponse& redirectResponse, FormState*, PolicyDecisionMode, PolicyCheckIdentifier, FramePolicyFunction&&) override;
     void cancelPolicyCheck() override;
 
     void dispatchUnableToImplementPolicy(const ResourceError&) override;
@@ -110,11 +109,10 @@
 
     RefPtr<Frame> createFrame(const URL& url, const String& name, HTMLFrameOwnerElement& ownerElement,
                                const String& referrer) override;
     ObjectContentType objectContentType(const URL& url, const String& mimeTypeIn) override;
     RefPtr<Widget> createPlugin(const IntSize&, HTMLPlugInElement&, const URL&, const Vector<String>&, const Vector<String>&, const String&, bool loadManually) override;
-    void recreatePlugin(Widget*) override {}
     void redirectDataToPlugin(Widget&) override;
     RefPtr<Widget> createJavaAppletWidget(const IntSize&, HTMLAppletElement&, const URL& baseURL, const Vector<String>& paramNames, const Vector<String>& paramValues) override;
     String overrideMediaType() const override;
 
     void setMainFrameDocumentReady(bool) override;
@@ -180,17 +178,16 @@
     String userAgent(const URL&) override;
 
     void savePlatformDataToCachedFrame(CachedFrame*) override;
     void transitionToCommittedFromCachedFrame(CachedFrame*) override;
     void transitionToCommittedForNewPage() override;
+    void didSaveToPageCache() override;
+    void didRestoreFromPageCache() override;
 
     bool canCachePage() const override;
     void convertMainResourceLoadToDownload(DocumentLoader*, PAL::SessionID, const ResourceRequest&, const ResourceResponse&) override;
 
-    void didSaveToPageCache() override;
-    void didRestoreFromPageCache() override;
-
     void dispatchDidBecomeFrameset(bool) override; // Can change due to navigation or DOM modification override.
 
     Ref<FrameNetworkingContext> createNetworkingContext() override;
 
     void registerForIconNotification() override;
