diff a/src/java.base/share/classes/java/security/KeyFactorySpi.java b/src/java.base/share/classes/java/security/KeyFactorySpi.java
--- a/src/java.base/share/classes/java/security/KeyFactorySpi.java
+++ b/src/java.base/share/classes/java/security/KeyFactorySpi.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1997, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -113,11 +113,11 @@
      * @param keySpec the specification class in which
      * the key material should be returned.
      *
      * @return the underlying key specification (key material) in an instance
      * of the requested specification class.
-
+     *
      * @throws    InvalidKeySpecException if the requested key specification is
      * inappropriate for the given key, or the given key cannot be dealt with
      * (e.g., the given key has an unrecognized format).
      */
     protected abstract <T extends KeySpec>
