diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGRoot.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGRoot.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGRoot.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGRoot.h
@@ -44,10 +44,11 @@
     bool isEmbeddedThroughFrameContainingSVGDocument() const;
 
     void computeIntrinsicRatioInformation(FloatSize& intrinsicSize, double& intrinsicRatio) const override;
 
     bool isLayoutSizeChanged() const { return m_isLayoutSizeChanged; }
+    bool isInLayout() const { return m_inLayout; }
     void setNeedsBoundariesUpdate() override { m_needsBoundariesOrTransformUpdate = true; }
     bool needsBoundariesUpdate() override { return m_needsBoundariesOrTransformUpdate; }
     void setNeedsTransformUpdate() override { m_needsBoundariesOrTransformUpdate = true; }
 
     IntSize containerSize() const { return m_containerSize; }
@@ -104,11 +105,12 @@
     void updateCachedBoundaries();
     void buildLocalToBorderBoxTransform();
 
     IntSize m_containerSize;
     FloatRect m_objectBoundingBox;
-    bool m_objectBoundingBoxValid;
+    bool m_objectBoundingBoxValid { false };
+    bool m_inLayout { false };
     FloatRect m_strokeBoundingBox;
     FloatRect m_repaintBoundingBox;
     FloatRect m_repaintBoundingBoxExcludingShadow;
     mutable AffineTransform m_localToParentTransform;
     AffineTransform m_localToBorderBoxTransform;
