diff a/modules/javafx.graphics/src/main/native-glass/gtk/GlassApplication.cpp b/modules/javafx.graphics/src/main/native-glass/gtk/GlassApplication.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk/GlassApplication.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/GlassApplication.cpp
@@ -435,14 +435,17 @@
 
     glass_evloop_call_hooks(event);
 
     if (ctx != NULL) {
         EventsCounterHelper helper(ctx);
-        gtk_main_do_event(event);
 
-        if (event->type == GDK_DRAG_LEAVE) {
+        if (event->type == GDK_EXPOSE) {
+            ctx->process_expose(&event->expose);
+        } else if (event->type == GDK_DRAG_LEAVE) {
             dnd_drag_leave_callback(ctx);
+        } else {
+            gtk_main_do_event(event);
         }
     } else {
         if (window == gdk_screen_get_root_window(gdk_screen_get_default())) {
             if (event->any.type == GDK_PROPERTY_NOTIFY) {
                 if (event->property.atom == gdk_atom_intern_static_string("_NET_WORKAREA")
