diff a/test/jdk/sun/security/x509/X500Name/DerValueConstructor.java b/test/jdk/sun/security/x509/X500Name/DerValueConstructor.java
--- a/test/jdk/sun/security/x509/X500Name/DerValueConstructor.java
+++ b/test/jdk/sun/security/x509/X500Name/DerValueConstructor.java
@@ -45,11 +45,11 @@
 
         // encode
         X500Name dn = new X500Name(name);
         System.err.println("DEBUG: dn: " + dn.toString());
         debugDER = new DerOutputStream();
-        dn.encode(debugDER);
+        dn.derEncode(debugDER);
         ba = debugDER.toByteArray();
         System.err.print("DEBUG: encoded X500Name bytes: ");
         System.out.println(toHexString(ba));
         System.err.println();
 
