diff a/modules/javafx.web/src/main/native/Tools/DumpRenderTree/CMakeLists.txt b/modules/javafx.web/src/main/native/Tools/DumpRenderTree/CMakeLists.txt
--- a/modules/javafx.web/src/main/native/Tools/DumpRenderTree/CMakeLists.txt
+++ b/modules/javafx.web/src/main/native/Tools/DumpRenderTree/CMakeLists.txt
@@ -21,14 +21,16 @@
     ${WEBKIT_TESTRUNNER_SHARED_DIR}/Bindings/JSWrapper.cpp
 )
 
 set(DumpRenderTree_LIBRARIES
     WebCoreTestSupport
-    JavaScriptCore
+    WebKit::JavaScriptCore
 )
 
 set(DumpRenderTree_INCLUDE_DIRECTORIES
+    ${WebCore_PRIVATE_FRAMEWORK_HEADERS_DIR}
+    ${WebKitLegacy_FRAMEWORK_HEADERS_DIR}
     ${DERIVED_SOURCES_DIR}/DumpRenderTree
     ${WEBKIT_TESTRUNNER_UISCRIPTCONTEXT_DIR}
     ${WEBKIT_TESTRUNNER_SHARED_DIR}/Bindings
     ${TOOLS_DIR}/DumpRenderTree
     ${CMAKE_SOURCE_DIR}/Source
@@ -63,12 +65,11 @@
     TestNetscapePlugIn/Tests/ToStringAndValueOfObject.cpp
     TestNetscapePlugIn/Tests/URLRedirect.cpp
 )
 
 set(TestNetscapePlugIn_LIBRARIES
-    JavaScriptCore
-    WTF
+    WebKit::JavaScriptCore
     WebCoreTestSupport
 )
 
 set(DumpRenderTree_IDL_FILES
     "${WEBKIT_TESTRUNNER_UISCRIPTCONTEXT_DIR}/Bindings/UIScriptController.idl"
