diff a/test/jdk/tools/jextract/ConstantsTest.java b/test/jdk/tools/jextract/ConstantsTest.java
--- a/test/jdk/tools/jextract/ConstantsTest.java
+++ b/test/jdk/tools/jextract/ConstantsTest.java
@@ -36,11 +36,11 @@
 
 /*
  * @test
  * @modules jdk.incubator.jextract
  * @build ConstantsTest
- * @run testng ConstantsTest
+ * @run testng/othervm -Djdk.incubator.foreign.Foreign=permit ConstantsTest
  */
 public class ConstantsTest extends JextractToolRunner {
     private Class<?> constants;
     private Path dirPath;
     private Loader loader;
