diff a/src/java.base/share/classes/sun/security/internal/spec/TlsKeyMaterialSpec.java b/src/java.base/share/classes/sun/security/internal/spec/TlsKeyMaterialSpec.java
--- a/src/java.base/share/classes/sun/security/internal/spec/TlsKeyMaterialSpec.java
+++ b/src/java.base/share/classes/sun/security/internal/spec/TlsKeyMaterialSpec.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2005, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -43,15 +43,20 @@
  * release.
  */
 @Deprecated
 public class TlsKeyMaterialSpec implements KeySpec, SecretKey {
 
+    @java.io.Serial
     static final long serialVersionUID = 812912859129525028L;
 
     private final SecretKey clientMacKey, serverMacKey;
     private final SecretKey clientCipherKey, serverCipherKey;
-    private final IvParameterSpec clientIv, serverIv;
+
+    @SuppressWarnings("serial") // Not statically typed as Serializable
+    private final IvParameterSpec clientIv;
+    @SuppressWarnings("serial") // Not statically typed as Serializable
+    private final IvParameterSpec serverIv;
 
     /**
      * Constructs a new TlsKeymaterialSpec from the client and server MAC
      * keys.
      * This call is equivalent to
