diff a/test/jdk/sun/rmi/log/ReliableLog/Recovery.java b/test/jdk/sun/rmi/log/ReliableLog/Recovery.java
--- a/test/jdk/sun/rmi/log/ReliableLog/Recovery.java
+++ b/test/jdk/sun/rmi/log/ReliableLog/Recovery.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, 2002, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -86,17 +86,11 @@
                     } catch (Exception e) {
                     }
 
                     Recovery handler = new Recovery();
                     ReliableLog log;
-                    if (size == 4 && deathpoint == 6) {
-                        Runtime.getRuntime().traceMethodCalls(true);
-                    }
                     log = new ReliableLog(dir, handler);
-                    if (size == 4 && deathpoint == 6) {
-                        Runtime.getRuntime().traceMethodCalls(false);
-                    }
 
                     // Generate a number of updates (size - 1) until failing
                     int i;
                     StringBuffer writ = new StringBuffer();
                     char[] u = new char[1];
