diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCDataChannelEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCDataChannelEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCDataChannelEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCDataChannelEvent.h
@@ -32,10 +32,11 @@
 #include <wtf/text/AtomString.h>
 
 namespace WebCore {
 
 class RTCDataChannelEvent final : public Event {
+    WTF_MAKE_ISO_ALLOCATED(RTCDataChannelEvent);
 public:
     struct Init : EventInit {
         RefPtr<RTCDataChannel> channel;
     };
 
