diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/locutil.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/locutil.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/locutil.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/locutil.cpp
@@ -9,10 +9,11 @@
 #include "unicode/utypes.h"
 
 #if !UCONFIG_NO_SERVICE || !UCONFIG_NO_TRANSLITERATION
 
 #include "unicode/resbund.h"
+#include "unicode/uenum.h"
 #include "cmemory.h"
 #include "ustrfmt.h"
 #include "locutil.h"
 #include "charstr.h"
 #include "ucln_cmn.h"
@@ -227,19 +228,18 @@
         htp = new Hashtable(status);
         if (htp && U_SUCCESS(status)) {
             CharString cbundleID;
             cbundleID.appendInvariantChars(bundleID, status);
             const char* path = cbundleID.isEmpty() ? NULL : cbundleID.data();
-            UEnumeration *uenum = ures_openAvailableLocales(path, &status);
+            icu::LocalUEnumerationPointer uenum(ures_openAvailableLocales(path, &status));
             for (;;) {
-                const UChar* id = uenum_unext(uenum, NULL, &status);
+                const UChar* id = uenum_unext(uenum.getAlias(), NULL, &status);
                 if (id == NULL) {
                     break;
                 }
                 htp->put(UnicodeString(id), (void*)htp, status);
             }
-            uenum_close(uenum);
             if (U_FAILURE(status)) {
                 delete htp;
                 return NULL;
             }
             umtx_lock(NULL);
