diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/sharedobject.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/sharedobject.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/sharedobject.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/sharedobject.h
@@ -102,11 +102,11 @@
     inline UBool hasHardReferences() const { return getRefCount() != 0; }
 
     /**
      * Deletes this object if it has no references.
      * Available for non-cached SharedObjects only. Ownership of cached objects
-     * is with the UnifiedCache, which is soley responsible for eviction and deletion.
+     * is with the UnifiedCache, which is solely responsible for eviction and deletion.
      */
     void deleteIfZeroRefCount() const;
 
 
     /**
