diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucln_cmn.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucln_cmn.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucln_cmn.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucln_cmn.cpp
@@ -38,12 +38,12 @@
  ************************************************/
 U_CAPI void U_EXPORT2
 u_cleanup(void)
 {
     UTRACE_ENTRY_OC(UTRACE_U_CLEANUP);
-    umtx_lock(NULL);     /* Force a memory barrier, so that we are sure to see   */
-    umtx_unlock(NULL);   /*   all state left around by any other threads.        */
+    icu::umtx_lock(NULL);     /* Force a memory barrier, so that we are sure to see   */
+    icu::umtx_unlock(NULL);   /*   all state left around by any other threads.        */
 
     ucln_lib_cleanup();
 
     cmemory_cleanup();       /* undo any heap functions set by u_setMemoryFunctions(). */
     UTRACE_EXIT();           /* Must be before utrace_cleanup(), which turns off tracing. */
