diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucnv_ct.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucnv_ct.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucnv_ct.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/ucnv_ct.cpp
@@ -178,11 +178,11 @@
 static const char* U_CALLCONV
 _CompoundTextgetName(const UConverter* cnv);
 
 
 static int32_t findNextEsc(const char *source, const char *sourceLimit) {
-    int32_t length = sourceLimit - source;
+    int32_t length = static_cast<int32_t>(sourceLimit - source);
     int32_t i;
     for (i = 1; i < length; i++) {
         if (*(source + i) == 0x1B) {
             return i;
         }
