diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/umapfile.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/umapfile.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/umapfile.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/umapfile.h
@@ -27,11 +27,11 @@
 
 #include "unicode/putil.h"
 #include "unicode/udata.h"
 #include "putilimp.h"
 
-U_CFUNC UBool uprv_mapFile(UDataMemory *pdm, const char *path);
+U_CFUNC UBool uprv_mapFile(UDataMemory *pdm, const char *path, UErrorCode *status);
 U_CFUNC void  uprv_unmapFile(UDataMemory *pData);
 
 /* MAP_NONE: no memory mapping, no file access at all */
 #define MAP_NONE        0
 #define MAP_WIN32       1
