diff a/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java b/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java
--- a/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java
@@ -47,15 +47,23 @@
         var apiBase = URIBuilder
                 .base(gitHubHost.getURI())
                 .appendSubDomain("api")
                 .setPath("/repos/" + repository + "/")
                 .build();
-        request = new RestRequest(apiBase, () -> Arrays.asList(
-                "Authorization", "token " + gitHubHost.getInstallationToken(),
+        request = new RestRequest(apiBase, () -> {
+            var headers = new ArrayList<>(List.of(
                 "Accept", "application/vnd.github.machine-man-preview+json",
                 "Accept", "application/vnd.github.antiope-preview+json",
-                "Accept", "application/vnd.github.shadow-cat-preview+json"));
+                "Accept", "application/vnd.github.shadow-cat-preview+json"
+            ));
+            var token = gitHubHost.getInstallationToken();
+            if (token.isPresent()) {
+                headers.add("Authorization");
+                headers.add("token " + token.get());
+            }
+            return headers;
+        });
         json = gitHubHost.getProjectInfo(repository);
         var urlPattern = gitHubHost.getWebURI("/" + repository + "/pull/").toString();
         pullRequestPattern = Pattern.compile(urlPattern + "(\\d+)");
     }
 
@@ -140,15 +148,17 @@
         return repository;
     }
 
     @Override
     public URI url() {
-        return URIBuilder
-                .base(gitHubHost.getURI())
-                .setPath("/" + repository + ".git")
-                .setAuthentication("x-access-token:" + gitHubHost.getInstallationToken())
-                .build();
+        var builder = URIBuilder.base(gitHubHost.getURI())
+                                .setPath("/" + repository + ".git");
+        var token = gitHubHost.getInstallationToken();
+        if (token.isPresent()) {
+            builder.setAuthentication("x-access-token:" + token.get());
+        }
+        return builder.build();
     }
 
     @Override
     public URI webUrl() {
         var endpoint = "/" + repository;
