diff a/src/jdk.jfr/share/classes/jdk/jfr/events/FileWriteEvent.java b/src/jdk.jfr/share/classes/jdk/jfr/events/FileWriteEvent.java
--- a/src/jdk.jfr/share/classes/jdk/jfr/events/FileWriteEvent.java
+++ b/src/jdk.jfr/share/classes/jdk/jfr/events/FileWriteEvent.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2012, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2012, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -36,26 +36,17 @@
 @Label("File Write")
 @Category("Java Application")
 @Description("Writing data to a file")
 public final class FileWriteEvent extends AbstractJDKEvent {
 
-    public static final ThreadLocal<FileWriteEvent> EVENT =
-        new ThreadLocal<>() {
-            @Override protected FileWriteEvent initialValue() {
-                return new FileWriteEvent();
-            }
-        };
+    // The order of these fields must be the same as the parameters in
+    // EventHandler::write(..., String, long)
 
     @Label("Path")
     @Description("Full path of the file")
     public String path;
 
     @Label("Bytes Written")
     @Description("Number of bytes written to the file")
     @DataAmount
     public long bytesWritten;
-
-    public void reset() {
-        path = null;
-        bytesWritten = 0;
-    }
 }
