diff a/src/org/openjdk/asmtools/jasm/Environment.java b/src/org/openjdk/asmtools/jasm/Environment.java
--- a/src/org/openjdk/asmtools/jasm/Environment.java
+++ b/src/org/openjdk/asmtools/jasm/Environment.java
@@ -25,10 +25,11 @@
 import static org.openjdk.asmtools.jasm.Constants.EOF;
 import static org.openjdk.asmtools.jasm.Constants.OFFSETBITS;
 import org.openjdk.asmtools.util.I18NResourceBundle;
 
 import java.io.*;
+import java.nio.file.Paths;
 import java.text.MessageFormat;
 
 /**
  * An input stream for java programs. The stream treats either "\n", "\r" or "\r\n" as the
  * end of a line, it always returns \n. It also parses UNICODE characters expressed as
@@ -92,13 +93,11 @@
         return inputFileName;
     }
 
     public String getSimpleInputFileName() {
         if( simpleInputFileName == null ) {
-            char separatorChar = (inputFileName.matches("^[A-Za-z]+:.*")) ? '/' : File.separatorChar;
-            int index = inputFileName.lastIndexOf(separatorChar);
-            simpleInputFileName = (index < 0) ? inputFileName.substring(0) : inputFileName.substring(index + 1);
+            simpleInputFileName = Paths.get(inputFileName).getFileName().toString();
         }
         return simpleInputFileName;
     }
 
     int lookForward() {
