diff a/src/java.security.jgss/share/classes/sun/security/krb5/internal/crypto/KeyUsage.java b/src/java.security.jgss/share/classes/sun/security/krb5/internal/crypto/KeyUsage.java
--- a/src/java.security.jgss/share/classes/sun/security/krb5/internal/crypto/KeyUsage.java
+++ b/src/java.security.jgss/share/classes/sun/security/krb5/internal/crypto/KeyUsage.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2004, 2012, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2004, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -54,10 +54,11 @@
     public static final int KU_ENC_KRB_PRIV_PART = 13;          // KrbPriv
     public static final int KU_ENC_KRB_CRED_PART = 14;          // KrbCred
     public static final int KU_KRB_SAFE_CKSUM = 15;             // KrbSafe
     public static final int KU_PA_FOR_USER_ENC_CKSUM = 17;      // S4U2user
     public static final int KU_AD_KDC_ISSUED_CKSUM = 19;
+    public static final int KU_AS_REQ = 56;
 
     public static final boolean isValid(int usage) {
         return usage >= 0;
     }
 }
