diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstconfig.h b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstconfig.h
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstconfig.h
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstconfig.h
@@ -160,11 +160,15 @@
 #endif
 
 #ifdef GSTREAMER_LITE
   // We using def file to limit export, so not need to export all APIs
   #ifndef GST_API
-    #define GST_API
+    #if defined(__GNUC__)
+      #define GST_API GST_EXPORT
+    #else
+      #define GST_API
+    #endif
   #endif
 #else // GSTREAMER_LITE
   #ifndef GST_API
     #define GST_API GST_EXPORT
   #endif
