diff a/test/hotspot/jtreg/runtime/cds/appcds/SharedArchiveConsistency.java b/test/hotspot/jtreg/runtime/cds/appcds/SharedArchiveConsistency.java
--- a/test/hotspot/jtreg/runtime/cds/appcds/SharedArchiveConsistency.java
+++ b/test/hotspot/jtreg/runtime/cds/appcds/SharedArchiveConsistency.java
@@ -22,10 +22,11 @@
  *
  */
 
 /*
  * @test
+ * @key randomness
  * @summary SharedArchiveConsistency
  * @requires vm.cds
  * @library /test/lib
  * @build sun.hotspot.WhiteBox
  * @compile test-classes/Hello.java
