diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/animation/SVGSMILElement.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/animation/SVGSMILElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/animation/SVGSMILElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/animation/SVGSMILElement.h
@@ -44,17 +44,17 @@
     WTF_MAKE_ISO_ALLOCATED(SVGSMILElement);
 public:
     SVGSMILElement(const QualifiedName&, Document&);
     virtual ~SVGSMILElement();
 
-    void parseAttribute(const QualifiedName&, const AtomicString&) override;
+    void parseAttribute(const QualifiedName&, const AtomString&) override;
     void svgAttributeChanged(const QualifiedName&) override;
     InsertedIntoAncestorResult insertedIntoAncestor(InsertionType, ContainerNode&) override;
     void removedFromAncestor(RemovalType, ContainerNode&) override;
 
-    virtual bool hasValidAttributeType() = 0;
-    virtual bool hasValidAttributeName();
+    virtual bool hasValidAttributeType() const = 0;
+    virtual bool hasValidAttributeName() const;
     virtual void animationAttributeChanged() = 0;
 
     SMILTimeContainer* timeContainer() { return m_timeContainer.get(); }
 
     SVGElement* targetElement() const { return m_targetElement; }
