diff a/build.gradle b/build.gradle
--- a/build.gradle
+++ b/build.gradle
@@ -1831,11 +1831,14 @@
 
     if (JFX_DEPS_URL != "") {
         repositories {
             ivy {
                 url JFX_DEPS_URL
-                layout "pattern", {
+                metadataSources {
+                    artifact()
+                }
+                patternLayout {
                     artifact "[artifact]-[revision](-[classifier]).[ext]"
                     artifact "[artifact].[ext]"
                 }
             }
         }
@@ -1844,28 +1847,37 @@
     if (JFX_DEPS_URL == "") {
         repositories {
             mavenCentral()
             ivy {
                 url "https://download.eclipse.org/eclipse/updates/4.6/R-4.6.3-201703010400/plugins/"
-                layout "pattern", {
+                metadataSources {
+                    artifact()
+                }
+                patternLayout {
                     artifact "[artifact].[ext]"
                 }
             }
         }
     }
 
     if (JFX_DEPS_URL == "" && IS_BUILD_LIBAV_STUBS) {
         repositories {
             ivy {
                 url libAVRepositoryURL
-                layout "pattern", {
+                metadataSources {
+                    artifact()
+                }
+                patternLayout {
                     artifact "[artifact].[ext]"
                 }
             }
             ivy {
                 url FFmpegRepositoryURL
-                layout "pattern", {
+                metadataSources {
+                    artifact()
+                }
+                patternLayout {
                     artifact "[artifact].[ext]"
                 }
             }
         }
     }
@@ -3808,10 +3820,11 @@
         }
         // no jars needed for modules
         project.jar.enabled = false
 
         // and redirect the resources into the module
+        project.sourceSets.main.output.resourcesDir = project.moduleDir
         project.processResources.destinationDir = project.moduleDir
     }
 
     if (project.hasProperty('moduleSourcePathShim') &&
             project.sourceSets.hasProperty('shims')) {
@@ -3841,10 +3854,11 @@
 
         // shims resources should have the main resouces as a base
         project.sourceSets.shims.resources.srcDirs += project.sourceSets.main.resources.srcDirs
 
         // and redirect the resources into the module
+        project.sourceSets.shims.output.resourcesDir = project.moduleShimsDir
         project.processShimsResources.destinationDir = project.moduleShimsDir
 
        compileTestJava.dependsOn(copyGeneratedShimsTask)
     }
 
diff a/build.properties b/build.properties
--- a/build.properties
+++ b/build.properties
@@ -83,11 +83,11 @@
 # gradle/wrapper/gradle-wrapper.properties and should also be recorded in
 # gradle/legal/gradle.md.
 # The jfx.gradle.version.min property defines the minimum version of gradle
 # that is supported. It must be <= jfx.gradle.version.
 jfx.gradle.version=5.3
-jfx.gradle.version.min=4.8
+jfx.gradle.version.min=5.3
 
 # Toolchains
 jfx.build.linux.gcc.version=gcc8.2.0-OL6.4+1.0
 jfx.build.windows.msvc.version=VS2017-15.9.6+1.0
 jfx.build.macosx.xcode.version=Xcode10.1-MacOSX10.14+1.0
diff a/buildSrc/build.gradle b/buildSrc/build.gradle
--- a/buildSrc/build.gradle
+++ b/buildSrc/build.gradle
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2013, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -58,11 +58,14 @@
 
 repositories {
     if (buildClosed) {
         ivy {
             url jfxRepositoryURL
-            layout "pattern", {
+            metadataSources {
+                artifact()
+            }
+            patternLayout {
                 artifact "[artifact]-[revision].[ext]"
                 artifact "[artifact].[ext]"
             }
         }
     } else {
diff a/settings.gradle b/settings.gradle
--- a/settings.gradle
+++ b/settings.gradle
@@ -33,14 +33,10 @@
 project(":fxml").projectDir = file("modules/javafx.fxml")
 project(":web").projectDir = file("modules/javafx.web")
 project(":media").projectDir = file("modules/javafx.media")
 project(":systemTests").projectDir = file("tests/system")
 
-// Stable publishing behavior is the default in gradle 5.x.
-// This setting enables it in gradle 4.8 to help with the transition.
-enableFeaturePreview('STABLE_PUBLISHING')
-
 def closedDir = file("../rt-closed")
 def buildClosed = closedDir.isDirectory()
 
 if (buildClosed) {
     File supplementalSettingsFile = new File("../rt-closed/closed-settings.gradle");
