diff a/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutContainer.cpp b/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutContainer.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutContainer.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutContainer.cpp
@@ -35,11 +35,11 @@
 namespace Layout {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(Container);
 
 Container::Container(Optional<ElementAttributes> attributes, RenderStyle&& style, BaseTypeFlags baseTypeFlags)
-    : Box(attributes, WTFMove(style), baseTypeFlags | ContainerFlag)
+    : Box(attributes, { }, WTFMove(style), baseTypeFlags | ContainerFlag)
 {
 }
 
 const Box* Container::firstInFlowChild() const
 {
@@ -89,8 +89,21 @@
 void Container::setLastChild(Box& childBox)
 {
     m_lastChild = &childBox;
 }
 
+void Container::appendChild(Box& childBox)
+{
+    childBox.setParent(*this);
+
+    if (m_lastChild) {
+        m_lastChild->setNextSibling(childBox);
+        childBox.setPreviousSibling(*m_lastChild);
+    } else
+        m_firstChild = &childBox;
+
+    m_lastChild = &childBox;
+}
+
 }
 }
 #endif
