diff a/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java b/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java
--- a/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java
+++ b/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java
@@ -46,29 +46,46 @@
 import static java.lang.constant.ConstantDescs.BSM_INVOKE;
 import static java.lang.constant.ConstantDescs.CD_String;
 import static java.lang.constant.ConstantDescs.CD_long;
 
 abstract class AbstractLayout implements MemoryLayout {
+<<<<<<< HEAD
 
     private final OptionalLong size;
     final long alignment;
     final Map<String, Constable> attributes;
+=======
+    // memory layout attribute key for layout name
+    static final String NAME = "name";
+
+    private final OptionalLong size;
+    final long alignment;
+    protected final Map<String, Constable> attributes;
+>>>>>>> cd397502f19f72fa8a926f4508d3913c8ace9059
 
     public AbstractLayout(OptionalLong size, long alignment, Map<String, Constable> attributes) {
         this.size = size;
         this.alignment = alignment;
         this.attributes = Collections.unmodifiableMap(attributes);
     }
 
     @Override
     public AbstractLayout withName(String name) {
+<<<<<<< HEAD
         return withAttribute(LAYOUT_NAME, name);
+=======
+        return withAttribute(NAME, name);
+>>>>>>> cd397502f19f72fa8a926f4508d3913c8ace9059
     }
 
     @Override
     public final Optional<String> name() {
+<<<<<<< HEAD
         return attribute(LAYOUT_NAME).map(String.class::cast);
+=======
+        return attribute(NAME).map(String.class::cast);
+>>>>>>> cd397502f19f72fa8a926f4508d3913c8ace9059
     }
 
     @Override
     public Optional<Constable> attribute(String name) {
         return Optional.ofNullable(attributes.get(name));
