diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/text/CString.cpp b/modules/javafx.web/src/main/native/Source/WTF/wtf/text/CString.cpp
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/text/CString.cpp
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/text/CString.cpp
@@ -26,21 +26,24 @@
 
 #include "config.h"
 #include <wtf/text/CString.h>
 
 #include <string.h>
+#include <wtf/NeverDestroyed.h>
 #include <wtf/text/StringHasher.h>
 
 namespace WTF {
 
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(CStringBuffer);
+
 Ref<CStringBuffer> CStringBuffer::createUninitialized(size_t length)
 {
     RELEASE_ASSERT(length < (std::numeric_limits<unsigned>::max() - sizeof(CStringBuffer)));
 
     // The +1 is for the terminating null character.
     size_t size = sizeof(CStringBuffer) + length + 1;
-    CStringBuffer* stringBuffer = static_cast<CStringBuffer*>(fastMalloc(size));
+    CStringBuffer* stringBuffer = static_cast<CStringBuffer*>(CStringBufferMalloc::malloc(size));
     return adoptRef(*new (NotNull, stringBuffer) CStringBuffer(length));
 }
 
 CString::CString(const char* str)
 {
