diff a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestInstance.java b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestInstance.java
--- a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestInstance.java
+++ b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestInstance.java
@@ -41,19 +41,18 @@
     private final Hash targetHash;
     private final Hash headHash;
     private final Hash baseHash;
     private final boolean ignoreStaleReviews;
 
-    PullRequestInstance(Path localRepoPath, Path seedPath, PullRequest pr, boolean ignoreStaleReviews) throws IOException  {
+    PullRequestInstance(Path localRepoPath, HostedRepositoryPool hostedRepositoryPool, PullRequest pr, boolean ignoreStaleReviews) throws IOException  {
         this.pr = pr;
         this.ignoreStaleReviews = ignoreStaleReviews;
 
-        // Materialize the PR's target ref
+        // Materialize the PR's source and target ref
         var repository = pr.repository();
-        var hostedRepositoryInstance = new HostedRepositoryInstance(repository, seedPath, pr.targetRef());
-        localRepo = hostedRepositoryInstance.checkout(localRepoPath);
-        localRepo.fetch(repository.url(), "+" + pr.sourceRef() + ":pr_prinstance_" + pr.id());
+        localRepo = hostedRepositoryPool.checkout(pr, localRepoPath);
+        localRepo.fetch(repository.url(), "+" + pr.targetRef() + ":pr_prinstance");
 
         targetHash = pr.targetHash();
         headHash = pr.headHash();
         baseHash = localRepo.mergeBase(targetHash, headHash);
     }
