diff a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
--- a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
@@ -131,15 +131,16 @@
         return url();
     }
 
     @Override
     public URI webUrl(Hash hash) {
-        try {
-            return new URI(url().toString() + "/" + hash.hex());
-        } catch (URISyntaxException e) {
-            throw new RuntimeException(e);
-        }
+        return URI.create(url().toString() + "/" + hash.hex());
+    }
+
+    @Override
+    public URI webUrl(String baseRef, String headRef) {
+        return URI.create(url().toString() + "/" + baseRef + "..." + headRef);
     }
 
     @Override
     public VCS repositoryType() {
         return VCS.GIT;
