diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/texmap/NicosiaBackingStoreTextureMapperImpl.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/texmap/NicosiaBackingStoreTextureMapperImpl.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/texmap/NicosiaBackingStoreTextureMapperImpl.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/nicosia/texmap/NicosiaBackingStoreTextureMapperImpl.cpp
@@ -35,11 +35,11 @@
 
 auto BackingStoreTextureMapperImpl::createFactory() -> Factory
 {
     return Factory(
         [](BackingStore&) {
-            return std::make_unique<BackingStoreTextureMapperImpl>();
+            return makeUnique<BackingStoreTextureMapperImpl>();
         });
 }
 
 BackingStoreTextureMapperImpl::BackingStoreTextureMapperImpl() = default;
 BackingStoreTextureMapperImpl::~BackingStoreTextureMapperImpl() = default;
