diff a/src/jdk.jfr/share/classes/jdk/jfr/internal/jfc/JFC.java b/src/jdk.jfr/share/classes/jdk/jfr/internal/jfc/JFC.java
--- a/src/jdk.jfr/share/classes/jdk/jfr/internal/jfc/JFC.java
+++ b/src/jdk.jfr/share/classes/jdk/jfr/internal/jfc/JFC.java
@@ -125,11 +125,15 @@
         return filename.toString();
     }
 
     public static String nameFromPath(Path file) throws IOException {
         String f = nullSafeFileName(file);
-        return f.substring(0, f.length() - JFCParser.FILE_EXTENSION.length());
+        if (f.endsWith(JFCParser.FILE_EXTENSION)) {
+            return f.substring(0, f.length() - JFCParser.FILE_EXTENSION.length());
+        } else  {
+            return f;
+        }
     }
 
     // Invoked by DCmdStart
     public static Configuration createKnown(String name) throws IOException, ParseException {
         // Known name, no need for permission
