diff a/test/hotspot/jtreg/vmTestbase/jit/tiered/Test.java b/test/hotspot/jtreg/vmTestbase/jit/tiered/Test.java
--- a/test/hotspot/jtreg/vmTestbase/jit/tiered/Test.java
+++ b/test/hotspot/jtreg/vmTestbase/jit/tiered/Test.java
@@ -53,11 +53,11 @@
     private static String UNSUPPORTED_OPTION_MESSAGE = "-XX:+TieredCompilation not supported in this VM";
     private static String REGEXP = "^[0-9.]+: \\[compile level=\\d";
     public static void main(String[] args) throws Exception {
         {
             System.out.println("TieredCompilation is enabled");
-            var pb = ProcessTools.createJavaProcessBuilder(true,
+            var pb = ProcessTools.createTestJvm(
                     "-XX:+TieredCompilation",
                     "-XX:+PrintTieredEvents",
                     "-version");
             var output = new OutputAnalyzer(pb.start());
             if (output.getStderr().contains(UNSUPPORTED_OPTION_MESSAGE)) {
@@ -66,11 +66,11 @@
             output.shouldHaveExitValue(0)
                   .stdoutShouldMatch(REGEXP);
         }
         {
             System.out.println("TieredCompilation is disabled");
-            var pb = ProcessTools.createJavaProcessBuilder(true,
+            var pb = ProcessTools.createTestJvm(
                     "-XX:-TieredCompilation",
                     "-XX:+PrintTieredEvents",
                     "-version");
             var output = new OutputAnalyzer(pb.start())
                     .shouldHaveExitValue(0)
