diff a/test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestLWorld.java b/test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestLWorld.java
--- a/test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestLWorld.java
+++ b/test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestLWorld.java
@@ -3287,6 +3287,22 @@
     @DontCompile
     public void test118_verifier(boolean warmup) {
         boolean res = test118(MyValueEmpty.default, MyValueEmpty.default, new MyValueEmpty());
         Asserts.assertTrue(res);
     }
+
+    // Test re-allocation of empty inline type array during deoptimization
+    @Test
+    public void test119(boolean deopt) {
+        MyValueEmpty[] arr = new MyValueEmpty[]{MyValueEmpty.default};
+        if (deopt) {
+            // uncommon trap
+            WHITE_BOX.deoptimizeMethod(tests.get(getClass().getSimpleName() + "::test119"));
+        }
+        Asserts.assertEquals(arr[0], MyValueEmpty.default);
+    }
+
+    @DontCompile
+    public void test119_verifier(boolean warmup) {
+        test119(!warmup);
+    }
 }
