diff a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/ModuleSummaryWriter.java b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/ModuleSummaryWriter.java
--- a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/ModuleSummaryWriter.java
+++ b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/ModuleSummaryWriter.java
@@ -74,19 +74,10 @@
      * @param moduleContentTree the content tree to which the module description
      *                           will be added
      */
     void addModuleDescription(Content moduleContentTree);
 
-    /**
-     * Adds the tag information from the "module-info.java" file to the documentation
-     * tree.
-     *
-     * @param moduleContentTree the content tree to which the module tags will
-     *                           be added
-     */
-    void addModuleTags(Content moduleContentTree);
-
     /**
      * Adds the summary of modules to the list of summaries.
      *
      * @param summariesList the list of summaries
      */
