diff a/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingCoordinator.cpp b/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingCoordinator.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingCoordinator.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingCoordinator.cpp
@@ -36,11 +36,10 @@
 #include "PlatformWheelEvent.h"
 #include "PluginViewBase.h"
 #include "Region.h"
 #include "RenderLayerCompositor.h"
 #include "RenderView.h"
-#include "RuntimeEnabledFeatures.h"
 #include "ScrollAnimator.h"
 #include "Settings.h"
 #include <wtf/MainThread.h>
 #include <wtf/text/StringBuilder.h>
 #include <wtf/text/TextStream.h>
@@ -74,11 +73,11 @@
 {
     ASSERT(isMainThread());
     ASSERT(m_page);
 
     if (!frameView.frame().isMainFrame() && !m_page->settings().scrollingTreeIncludesFrames()
-#if PLATFORM(MAC)
+#if PLATFORM(MAC) || USE(NICOSIA)
         && !m_page->settings().asyncFrameScrollingEnabled()
 #endif
     )
         return false;
 
@@ -497,15 +496,6 @@
         break;
     }
     return ts;
 }
 
-TextStream& operator<<(TextStream& ts, ScrollType scrollType)
-{
-    switch (scrollType) {
-    case ScrollType::User: ts << "user"; break;
-    case ScrollType::Programmatic: ts << "programmatic"; break;
-    }
-    return ts;
-}
-
 } // namespace WebCore
