diff a/test/jdk/java/awt/ComponentOrientation/BorderTest.java b/test/jdk/java/awt/ComponentOrientation/BorderTest.java
--- a/test/jdk/java/awt/ComponentOrientation/BorderTest.java
+++ b/test/jdk/java/awt/ComponentOrientation/BorderTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2018, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -37,13 +37,12 @@
  * may not be removed.
  */
 
 import java.awt.*;
 import java.awt.event.*;
-import java.applet.Applet;
 
-public class BorderTest extends Applet {
+public class BorderTest extends Panel {
     Panel       panel1;
     Panel       panel2;
 
     public BorderTest() {
         setLayout(new GridLayout(0,2));
@@ -106,12 +105,10 @@
                 System.exit(0);
             };
         } );
 
         BorderTest BorderTest = new BorderTest();
-        BorderTest.init();
-        BorderTest.start();
 
         f.add("Center", BorderTest);
         f.setSize(450, 300);
         f.show();
     }
