diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/agents/InspectorAgent.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/agents/InspectorAgent.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/agents/InspectorAgent.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/agents/InspectorAgent.cpp
@@ -118,11 +118,11 @@
 {
     if (extraDomains.isEmpty())
         return;
 
     auto domainNames = JSON::ArrayOf<String>::create();
-    for (auto domainName : extraDomains)
+    for (const auto& domainName : extraDomains)
         domainNames->addItem(domainName);
 
     m_frontendDispatcher->activateExtraDomains(WTFMove(domainNames));
 }
 #endif
