diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/audio/audio-format.h b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/audio/audio-format.h
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/audio/audio-format.h
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/audio/audio-format.h
@@ -186,10 +186,11 @@
 } GstAudioPackFlags;
 
 /**
  * GstAudioFormatUnpack:
  * @info: a #GstAudioFormatInfo
+ * @flags: #GstAudioPackFlags
  * @dest: (array) (element-type guint8): a destination array
  * @data: (array) (element-type guint8): pointer to the audio data
  * @length: the amount of samples to unpack.
  *
  * Unpacks @length samples from the given data of format @info.
@@ -197,25 +198,26 @@
  * interleaved. @dest should at least be big enough to hold @length *
  * channels * size(unpack_format) bytes.
  */
 typedef void (*GstAudioFormatUnpack)         (const GstAudioFormatInfo *info,
                                               GstAudioPackFlags flags, gpointer dest,
-                                              const gpointer data, gint length);
+                                              gconstpointer data, gint length);
 /**
  * GstAudioFormatPack:
  * @info: a #GstAudioFormatInfo
+ * @flags: #GstAudioPackFlags
  * @src: (array) (element-type guint8): a source array
  * @data: (array) (element-type guint8): pointer to the destination
  *   data
  * @length: the amount of samples to pack.
  *
  * Packs @length samples from @src to the data array in format @info.
  * The samples from source have each channel interleaved
  * and will be packed into @data.
  */
 typedef void (*GstAudioFormatPack)           (const GstAudioFormatInfo *info,
-                                              GstAudioPackFlags flags, const gpointer src,
+                                              GstAudioPackFlags flags, gconstpointer src,
                                               gpointer data, gint length);
 
 /**
  * GstAudioFormatInfo:
  * @format: #GstAudioFormat
@@ -231,10 +233,11 @@
  * @pack_func: function to pack samples
  *
  * Information for an audio format.
  */
 struct _GstAudioFormatInfo {
+  /*< public >*/
   GstAudioFormat format;
   const gchar *name;
   const gchar *description;
   GstAudioFormatFlags flags;
   gint endianness;
