diff a/test/jdk/java/io/Serializable/class/SerialA_2/A.java b/test/jdk/java/io/Serializable/class/SerialA_2/A.java
--- a/test/jdk/java/io/Serializable/class/SerialA_2/A.java
+++ b/test/jdk/java/io/Serializable/class/SerialA_2/A.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -22,10 +22,10 @@
  */
 
 import java.io.Serializable;
 
 public class A implements Serializable {
-    static final long serialVersionUID = 746945609796141988L;
+    private static final long serialVersionUID = 746945609796141988L;
 
     int field1;
     int field2;
 };
