diff a/modules/javafx.web/src/main/native/Source/WebCore/page/PageDebuggable.h b/modules/javafx.web/src/main/native/Source/WebCore/page/PageDebuggable.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/PageDebuggable.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/PageDebuggable.h
@@ -39,11 +39,11 @@
     WTF_MAKE_NONCOPYABLE(PageDebuggable);
 public:
     PageDebuggable(Page&);
     ~PageDebuggable() = default;
 
-    Inspector::RemoteControllableTarget::Type type() const final { return Inspector::RemoteControllableTarget::Type::Web; }
+    Inspector::RemoteControllableTarget::Type type() const final { return Inspector::RemoteControllableTarget::Type::Page; }
 
     String name() const final;
     String url() const final;
     bool hasLocalDebugger() const final;
 
@@ -60,8 +60,8 @@
     String m_nameOverride;
 };
 
 } // namespace WebCore
 
-SPECIALIZE_TYPE_TRAITS_CONTROLLABLE_TARGET(WebCore::PageDebuggable, Web);
+SPECIALIZE_TYPE_TRAITS_CONTROLLABLE_TARGET(WebCore::PageDebuggable, Page);
 
 #endif // ENABLE(REMOTE_INSPECTOR)
