diff a/modules/javafx.web/src/main/native/Source/WebCore/page/PointerLockController.cpp b/modules/javafx.web/src/main/native/Source/WebCore/page/PointerLockController.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/PointerLockController.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/PointerLockController.cpp
@@ -32,11 +32,10 @@
 #include "Element.h"
 #include "Event.h"
 #include "EventNames.h"
 #include "Page.h"
 #include "PlatformMouseEvent.h"
-#include "RuntimeEnabledFeatures.h"
 #include "UserGestureIndicator.h"
 #include "VoidCallback.h"
 
 #if ENABLE(POINTER_EVENTS)
 #include "PointerCaptureController.h"
@@ -214,12 +213,13 @@
         enqueueEvent(type, &element->document());
 }
 
 void PointerLockController::enqueueEvent(const AtomString& type, Document* document)
 {
-    if (document)
-        document->enqueueDocumentEvent(Event::create(type, Event::CanBubble::Yes, Event::IsCancelable::No));
+    // FIXME: Spec doesn't specify which task source use.
+    if (auto protectedDocument = makeRefPtr(document))
+        protectedDocument->queueTaskToDispatchEvent(TaskSource::UserInteraction, Event::create(type, Event::CanBubble::Yes, Event::IsCancelable::No));
 }
 
 } // namespace WebCore
 
 #endif // ENABLE(POINTER_LOCK)
