diff a/modules/javafx.base/src/test/java/test/com/sun/javafx/runtime/VersionInfoTest.java b/modules/javafx.base/src/test/java/test/com/sun/javafx/runtime/VersionInfoTest.java
--- a/modules/javafx.base/src/test/java/test/com/sun/javafx/runtime/VersionInfoTest.java
+++ b/modules/javafx.base/src/test/java/test/com/sun/javafx/runtime/VersionInfoTest.java
@@ -87,11 +87,11 @@
     @Test
     public void testMajorVersion() {
         String version = VersionInfo.getVersion();
         // Need to update major version number when we develop the next
         // major release.
-        assertTrue(version.startsWith("14"));
+        assertTrue(version.startsWith("15"));
         String runtimeVersion = VersionInfo.getRuntimeVersion();
         assertTrue(runtimeVersion.startsWith(version));
     }
 
     @Test
