diff a/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java b/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java
--- a/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java
+++ b/modules/javafx.graphics/src/main/java/javafx/stage/DirectoryChooser.java
@@ -47,10 +47,16 @@
     /**
      * The title of the displayed dialog.
      */
     private StringProperty title;
 
+    /**
+     * Creates a default {@code DirectoryChooser}.
+     */
+    public DirectoryChooser() {
+    }
+
     public final void setTitle(final String value) {
         titleProperty().set(value);
     }
 
     public final String getTitle() {
