diff a/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/internal/MCMBeanServerConnection.java b/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/internal/MCMBeanServerConnection.java
--- a/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/internal/MCMBeanServerConnection.java
+++ b/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/internal/MCMBeanServerConnection.java
@@ -249,11 +249,10 @@
 			}
 		}
 		return filteredAttributes;
 	}
 
-	
 	@Override
 	public Object invoke(ObjectName name, String operationName, Object[] params, String[] signature)
 			throws InstanceNotFoundException, MBeanException, ReflectionException, IOException {
 		logInvokeMessage(name, operationName, params);
 		return delegate.invoke(name, operationName, params, signature);
