diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java
@@ -80,10 +80,15 @@
         var body = marker + "\n" +
                 "*Mailing list message from [" + email.author().fullName().orElse(email.author().localPart()) +
                 "](mailto:" + email.author().address() + ") on [" + email.sender().localPart() +
                 "](mailto:" + email.sender().address() + "):*\n\n" +
                 TextToMarkdown.escapeFormatting(email.body());
+        if (body.length() > 64000) {
+            body = body.substring(0, 64000) + "...\n\n" + "" +
+                    "This message was too large to bridge in full, and has been truncated. " +
+                    "Please check the mailing list archive to see the full text.";
+        }
         pr.addComment(body);
     }
 
     @Override
     public void run(Path scratchPath) {
