diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/ucol.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/ucol.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/ucol.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/ucol.h
@@ -1147,11 +1147,11 @@
  * See http://www.unicode.org/reports/tr10/#Merging_Sort_Keys
  *
  * The recommended way to achieve "merged" sorting is by
  * concatenating strings with U+FFFE between them.
  * The concatenation has the same sort order as the merged sort keys,
- * but merge(getSortKey(str1), getSortKey(str2)) may differ from getSortKey(str1 + '\uFFFE' + str2).
+ * but merge(getSortKey(str1), getSortKey(str2)) may differ from getSortKey(str1 + '\\uFFFE' + str2).
  * Using strings with U+FFFE may yield shorter sort keys.
  *
  * For details about Sort Key Features see
  * http://userguide.icu-project.org/collation/api#TOC-Sort-Key-Features
  *
@@ -1292,10 +1292,11 @@
  *
  * Beginning with ICU 53, the variable top is pinned to
  * the top of one of the supported reordering groups,
  * and it must not be beyond the last of those groups.
  * See ucol_setMaxVariable().
+ * @param coll collator to be set
  * @param varTop primary weight, as returned by ucol_setVariableTop or ucol_getVariableTop
  * @param status error code
  * @see ucol_getVariableTop
  * @see ucol_setVariableTop
  * @deprecated ICU 53 Call ucol_setMaxVariable() instead.
