diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/udatpg.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/udatpg.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/udatpg.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/udatpg.h
@@ -93,24 +93,22 @@
      * @deprecated ICU 58 The numeric value may change over time, see ICU ticket #12420.
      */
     UDATPG_FIELD_COUNT
 } UDateTimePatternField;
 
-#ifndef U_HIDE_DRAFT_API
 /**
  * Field display name width constants for udatpg_getFieldDisplayName().
- * @draft ICU 61
+ * @stable ICU 61
  */
 typedef enum UDateTimePGDisplayWidth {
-    /** @draft ICU 61 */
+    /** @stable ICU 61 */
     UDATPG_WIDE,
-    /** @draft ICU 61 */
+    /** @stable ICU 61 */
     UDATPG_ABBREVIATED,
-    /** @draft ICU 61 */
+    /** @stable ICU 61 */
     UDATPG_NARROW
 } UDateTimePGDisplayWidth;
-#endif  // U_HIDE_DRAFT_API
 
 /**
  * Masks to control forcing the length of specified fields in the returned
  * pattern to match those in the skeleton (when this would not happen
  * otherwise). These may be combined to force the length of multiple fields.
@@ -438,11 +436,10 @@
 U_STABLE const UChar * U_EXPORT2
 udatpg_getAppendItemName(const UDateTimePatternGenerator *dtpg,
                          UDateTimePatternField field,
                          int32_t *pLength);
 
-#ifndef U_HIDE_DRAFT_API
 /**
  * The general interface to get a display name for a particular date/time field,
  * in one of several possible display widths.
  *
  * @param dtpg
@@ -462,19 +459,18 @@
  * @param pErrorCode
  *          A pointer to a UErrorCode to receive any errors
  * @return
  *         The full length of the name; if greater than capacity, fieldName contains a
  *         truncated result.
- * @draft ICU 61
+ * @stable ICU 61
  */
-U_DRAFT int32_t U_EXPORT2
+U_STABLE int32_t U_EXPORT2
 udatpg_getFieldDisplayName(const UDateTimePatternGenerator *dtpg,
                            UDateTimePatternField field,
                            UDateTimePGDisplayWidth width,
                            UChar *fieldName, int32_t capacity,
                            UErrorCode *pErrorCode);
-#endif  // U_HIDE_DRAFT_API
 
 /**
  * The DateTimeFormat is a message format pattern used to compose date and
  * time patterns. The default pattern in the root locale is "{1} {0}", where
  * {1} will be replaced by the date pattern and {0} will be replaced by the
