diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/usearch.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/usearch.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/usearch.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/unicode/usearch.h
@@ -255,14 +255,13 @@
      * in the pattern will match a plain e or an e with any diacritic in the
      * searched text, but an e with diacritic in the pattern will only
      * match an e with the same diacritic or a plain e in the searched text.
      *
      * This option is similar to "asymmetric search" as described in
-     * <a href="http://www.unicode.org/reports/tr10/#Asymmetric_Search">
-     * UTS #10 Unicode Collation Algorithm</a, but also allows unmarked
-     * characters in the searched text to match marked or unmarked versions of
-     * that character in the pattern.
+     * [UTS #10 Unicode Collation Algorithm](http://www.unicode.org/reports/tr10/#Asymmetric_Search),
+     * but also allows unmarked characters in the searched text to match
+     * marked or unmarked versions of that character in the pattern.
      *
      * @stable ICU 4.4
      */
     USEARCH_ANY_BASE_WEIGHT_IS_WILDCARD,
 
