diff a/test/jdk/javax/net/ssl/compatibility/README b/test/jdk/javax/net/ssl/compatibility/README
--- a/test/jdk/javax/net/ssl/compatibility/README
+++ b/test/jdk/javax/net/ssl/compatibility/README
@@ -28,11 +28,11 @@
 
 ##### Usage #####
 jtreg [-options] \
     [-Dtest.debug=<true|false>] \
     [-Dtest.jdk.list.file=</path/to/jdkListFile>] \
-    $JDK_WORKSPACE/test/jdk/javax/net/ssl/compatibility/Compatibility.java
+    $JDK_WORKSPACE/test/jdk/javax/net/ssl/compatibility/<XXX.java>
 
 Besides the common jtreg options, like -jdk, this test introduces some more
 properties:
 test.debug
     It indicates if the test enable -Djavax.net.debug=all. This is a boolean
@@ -52,10 +52,10 @@
 /path/to/jdk10
 
 $ jtreg -jdk:/path/to/latest/jdk \
     -Ddebug=true \
     -Dtest.jdk.list.file=/path/to/jdkList \
-    $JDK_WS/jdk/test/javax/net/ssl/compatibility/Compatibility.java
+    $JDK_WS/jdk/test/javax/net/ssl/compatibility/<XXX.java>
 The above example uses a file "/path/to/jdkList" to contain the paths of local
 different JDK builds through 8 to 10. The execution uses each of JDK builds as
 server and client respectively. And it enables SSL debug flag, and tests the
 full parameter value set.
