diff a/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java b/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java
@@ -184,11 +184,11 @@
     public String spelling() {
         return LibClang.CXStrToString(Index_h.clang_getTypeSpelling(type));
     }
 
     public int kind0() {
-        return Index_h.CXType$kind$get(type);
+        return Index_h.CXType.kind$get(type);
     }
 
     private long size0() {
         return Index_h.clang_Type_getSizeOf(type);
     }
