diff a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
@@ -982,18 +982,18 @@
     // if window has no decoration, adjustments will be ZERO
     // this will always be content size
     int newW, newH;
     if (w > 0) {
         newW = w - geometry.adjust_w;
-        geometry.current_w = w;
+        geometry.current_w = newW;
     } else {
         newW = cw;
     }
 
     if (h > 0) {
         newH = h - geometry.adjust_h;
-        geometry.current_h = h;
+        geometry.current_h = newH;
     } else {
         newH = ch;
     }
 
     gboolean size_changed = FALSE;
