diff a/test/jdk/java/util/concurrent/locks/Lock/CheckedLockLoops.java b/test/jdk/java/util/concurrent/locks/Lock/CheckedLockLoops.java
--- a/test/jdk/java/util/concurrent/locks/Lock/CheckedLockLoops.java
+++ b/test/jdk/java/util/concurrent/locks/Lock/CheckedLockLoops.java
@@ -134,11 +134,11 @@
             barrier.await();
             barrier.await();
             long time = timer.getTime();
             long tpi = time / (iters * nthreads);
             System.out.print("\t" + LoopHelpers.rightJustify(tpi) + " ns per update");
-            //                double secs = (double)(time) / 1000000000.0;
+            //                double secs = (double)time / 1000000000.0;
             //                System.out.print("\t " + secs + "s run time");
             System.out.println();
 
             if (result == 0) // avoid overoptimization
                 System.out.println("useless result: " + result);
