diff a/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/InspectorCPUProfilerAgent.h b/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/InspectorCPUProfilerAgent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/InspectorCPUProfilerAgent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/InspectorCPUProfilerAgent.h
@@ -39,18 +39,19 @@
 class InspectorCPUProfilerAgent final : public InspectorAgentBase, public Inspector::CPUProfilerBackendDispatcherHandler {
     WTF_MAKE_NONCOPYABLE(InspectorCPUProfilerAgent);
     WTF_MAKE_FAST_ALLOCATED;
 public:
     InspectorCPUProfilerAgent(PageAgentContext&);
-    virtual ~InspectorCPUProfilerAgent() = default;
+    virtual ~InspectorCPUProfilerAgent();
 
-    void didCreateFrontendAndBackend(Inspector::FrontendRouter*, Inspector::BackendDispatcher*) override;
-    void willDestroyFrontendAndBackend(Inspector::DisconnectReason) override;
+    // InspectorAgentBase
+    void didCreateFrontendAndBackend(Inspector::FrontendRouter*, Inspector::BackendDispatcher*);
+    void willDestroyFrontendAndBackend(Inspector::DisconnectReason);
 
     // CPUProfilerBackendDispatcherHandler
-    void startTracking(ErrorString&) override;
-    void stopTracking(ErrorString&) override;
+    void startTracking(ErrorString&);
+    void stopTracking(ErrorString&);
 
 private:
     void collectSample(const ResourceUsageData&);
 
     std::unique_ptr<Inspector::CPUProfilerFrontendDispatcher> m_frontendDispatcher;
