diff a/test/jdk/javax/net/ssl/TLSCommon/interop/Utilities.java b/test/jdk/javax/net/ssl/TLSCommon/interop/Utilities.java
--- a/test/jdk/javax/net/ssl/TLSCommon/interop/Utilities.java
+++ b/test/jdk/javax/net/ssl/TLSCommon/interop/Utilities.java
@@ -78,12 +78,11 @@
     public static final String LOG_PATH = System.getProperty("test.log.path");
 
     public static final String PARAM_DELIMITER = ";";
     public static final String VALUE_DELIMITER = ",";
 
-    public static final CipherSuite[] ALL_CIPHER_SUITES
-            = Utilities.getAllCipherSuites();
+    public static final CipherSuite[] ALL_CIPHER_SUITES = getAllCipherSuites();
 
     /*
      * Gets all supported cipher suites.
      */
     private static CipherSuite[] getAllCipherSuites() {
@@ -99,12 +98,11 @@
 
         CipherSuite[] cipherSuites = Arrays.stream(supportedCipherSuites)
                 .map(cipherSuite -> {
                     return CipherSuite.cipherSuite(cipherSuite);})
                 .filter(cipherSuite -> {
-                    return cipherSuite
-                            != CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV;})
+                    return cipherSuite != CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV; })
                 .toArray(CipherSuite[]::new);
 
         return cipherSuites;
     }
 
