diff a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ExtraRowTableViewer.java b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ExtraRowTableViewer.java
--- a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ExtraRowTableViewer.java
+++ b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ExtraRowTableViewer.java
@@ -58,10 +58,14 @@
 
 	public void setExtraMessage(String message) {
 		this.message = message;
 	}
 
+	public long getNumRowsDisplayed() {
+		return getFilteredChildren(getInput()).length;
+	}
+
 	@Override
 	public void refresh(Object element) {
 		if (message == null) {
 			super.refresh(element);
 		} else {
