diff a/src/hotspot/cpu/x86/interp_masm_x86.cpp b/src/hotspot/cpu/x86/interp_masm_x86.cpp
--- a/src/hotspot/cpu/x86/interp_masm_x86.cpp
+++ b/src/hotspot/cpu/x86/interp_masm_x86.cpp
@@ -1161,13 +1161,13 @@
     movptr(rdi, Address(rdi, Method::const_offset()));
     load_unsigned_byte(rdi, Address(rdi, ConstMethod::result_type_offset()));
     cmpl(rdi, T_INLINE_TYPE);
     jcc(Assembler::notEqual, skip);
 
-    // We are returning a value type, load its fields into registers
+    // We are returning an inline type, load its fields into registers
 #ifndef _LP64
-    super_call_VM_leaf(StubRoutines::load_value_type_fields_in_regs());
+    super_call_VM_leaf(StubRoutines::load_inline_type_fields_in_regs());
 #else
     // Load fields from a buffered value with a value class specific handler
     Register tmp_load_klass = LP64_ONLY(rscratch1) NOT_LP64(noreg);
     load_klass(rdi, rax, tmp_load_klass);
     movptr(rdi, Address(rdi, InstanceKlass::adr_inlineklass_fixed_block_offset()));
