diff a/modules/javafx.web/src/main/native/Source/WebCore/style/StyleChange.cpp b/modules/javafx.web/src/main/native/Source/WebCore/style/StyleChange.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/style/StyleChange.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/style/StyleChange.cpp
@@ -47,19 +47,18 @@
     // When text-combine is on, we use RenderCombineText, otherwise RenderText.
     // https://bugs.webkit.org/show_bug.cgi?id=55069
     if (s1.hasTextCombine() != s2.hasTextCombine())
         return Detach;
 
-    if (s1 != s2) {
-        if (s1.inheritedNotEqual(&s2))
-            return Inherit;
+    if (!s1.inheritedEqual(s2))
+        return Inherit;
 
-        if (s1.alignItems() != s2.alignItems() || s1.justifyItems() != s2.justifyItems())
-            return Inherit;
+    if (!s1.descendantAffectingNonInheritedPropertiesEqual(s2))
+        return Inherit;
 
+    if (s1 != s2)
         return NoInherit;
-    }
 
     // If the pseudoStyles have changed, we want any StyleChange that is not NoChange
     // because setStyle will do the right thing with anything else.
     if (s1.hasAnyPublicPseudoStyles()) {
         for (PseudoId pseudoId = PseudoId::FirstPublicPseudoId; pseudoId < PseudoId::FirstInternalPseudoId; pseudoId = static_cast<PseudoId>(static_cast<unsigned>(pseudoId) + 1)) {
