diff a/modules/javafx.web/src/main/native/Source/WebCore/style/StyleInvalidationFunctions.h b/modules/javafx.web/src/main/native/Source/WebCore/style/StyleInvalidationFunctions.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/style/StyleInvalidationFunctions.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/style/StyleInvalidationFunctions.h
@@ -23,15 +23,15 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #pragma once
 
-#include "DocumentRuleSets.h"
 #include "HTMLSlotElement.h"
 #include "ShadowRoot.h"
 #include "StyleResolver.h"
 #include "StyleScope.h"
+#include "StyleScopeRuleSets.h"
 
 namespace WebCore {
 namespace Style {
 
 template <typename TraverseFunction>
