diff a/issuetracker/src/main/java/org/openjdk/skara/issuetracker/IssueTrackerFactory.java b/issuetracker/src/main/java/org/openjdk/skara/issuetracker/IssueTrackerFactory.java
--- a/issuetracker/src/main/java/org/openjdk/skara/issuetracker/IssueTrackerFactory.java
+++ b/issuetracker/src/main/java/org/openjdk/skara/issuetracker/IssueTrackerFactory.java
@@ -25,12 +25,12 @@
 import org.openjdk.skara.host.*;
 
 import java.net.URI;
 
 public class IssueTrackerFactory {
-    public static IssueTracker createJiraHost(URI uri, PersonalAccessToken pat) {
-        if (pat != null) {
+    public static IssueTracker createJiraHost(URI uri, Credential credential) {
+        if (credential != null) {
             throw new RuntimeException("authentication not implemented yet");
         }
         return new JiraHost(uri);
     }
 }
