diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/number_microprops.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/number_microprops.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/number_microprops.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/number_microprops.h
@@ -30,10 +30,11 @@
     Padder padding;
     IntegerWidth integerWidth;
     UNumberSignDisplay sign;
     UNumberDecimalSeparatorDisplay decimal;
     bool useCurrency;
+    char nsName[9];
 
     // Note: This struct has no direct ownership of the following pointers.
     const DecimalFormatSymbols* symbols;
     const Modifier* modOuter;
     const Modifier* modMiddle;
