diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/numparse_scientific.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/numparse_scientific.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/numparse_scientific.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/numparse_scientific.cpp
@@ -53,10 +53,15 @@
 bool ScientificMatcher::match(StringSegment& segment, ParsedNumber& result, UErrorCode& status) const {
     // Only accept scientific notation after the mantissa.
     if (!result.seenNumber()) {
         return false;
     }
+
+    // Only accept one exponent per string.
+    if (0 != (result.flags & FLAG_HAS_EXPONENT)) {
+        return false;
+    }
 
     // First match the scientific separator, and then match another number after it.
     // NOTE: This is guarded by the smoke test; no need to check fExponentSeparatorString length again.
     int overlap1 = segment.getCommonPrefixLength(fExponentSeparatorString);
     if (overlap1 == fExponentSeparatorString.length()) {
