diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/rbtz.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/rbtz.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/rbtz.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/rbtz.cpp
@@ -144,14 +144,14 @@
     }
     // Mark dirty, so transitions are recalculated at next complete() call
     fUpToDate = FALSE;
 }
 
-static UMutex gLock = U_MUTEX_INITIALIZER;
 
 void
 RuleBasedTimeZone::completeConst(UErrorCode& status) const {
+    static UMutex gLock = U_MUTEX_INITIALIZER;
     if (U_FAILURE(status)) {
         return;
     }
     umtx_lock(&gLock);
     if (!fUpToDate) {
