diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tmutfmt.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tmutfmt.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tmutfmt.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tmutfmt.cpp
@@ -222,11 +222,11 @@
                 if (pattern->getArgTypeCount() != 0) {
                     Formattable& temp = parsed[0];
                     if (temp.getType() == Formattable::kString) {
                         UnicodeString tmpString;
                         UErrorCode pStatus = U_ZERO_ERROR;
-                        getNumberFormat().parse(temp.getString(tmpString), tmpNumber, pStatus);
+                        getNumberFormatInternal().parse(temp.getString(tmpString), tmpNumber, pStatus);
                         if (U_FAILURE(pStatus)) {
                             continue;
                         }
                     } else if (temp.isNumeric()) {
                         tmpNumber = temp;
