diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tzgnames.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tzgnames.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tzgnames.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/i18n/tzgnames.h
@@ -44,11 +44,11 @@
     virtual ~TimeZoneGenericNames();
 
     static TimeZoneGenericNames* createInstance(const Locale& locale, UErrorCode& status);
 
     virtual UBool operator==(const TimeZoneGenericNames& other) const;
-    virtual UBool operator!=(const TimeZoneGenericNames& other) const {return !operator==(other);};
+    virtual UBool operator!=(const TimeZoneGenericNames& other) const {return !operator==(other);}
     virtual TimeZoneGenericNames* clone() const;
 
     UnicodeString& getDisplayName(const TimeZone& tz, UTimeZoneGenericNameType type,
                         UDate date, UnicodeString& name) const;
 
