diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/BytecodeGeneratorification.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/BytecodeGeneratorification.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/BytecodeGeneratorification.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/BytecodeGeneratorification.h
@@ -29,11 +29,11 @@
 namespace JSC {
 
 class BytecodeGenerator;
 class InstructionStreamWriter;
 class SymbolTable;
-class UnlinkedCodeBlock;
+class UnlinkedCodeBlockGenerator;
 class SymbolTable;
 
-void performGeneratorification(BytecodeGenerator&, UnlinkedCodeBlock*, InstructionStreamWriter&, SymbolTable* generatorFrameSymbolTable, int generatorFrameSymbolTableIndex);
+void performGeneratorification(BytecodeGenerator&, UnlinkedCodeBlockGenerator*, InstructionStreamWriter&, SymbolTable* generatorFrameSymbolTable, int generatorFrameSymbolTableIndex);
 
 } // namespace JSC
