diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUBufferDescriptor.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUBufferDescriptor.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUBufferDescriptor.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUBufferDescriptor.h
@@ -25,34 +25,22 @@
 
 #pragma once
 
 #if ENABLE(WEBGPU)
 
-#include <wtf/RefCounted.h>
+#include "GPUBufferUsage.h"
 
 namespace WebCore {
 
-using GPUBufferUsageFlags = unsigned long;
-
 struct GPUBufferDescriptor {
-    unsigned long size;
+    uint64_t size;
     GPUBufferUsageFlags usage;
 };
 
-class GPUBufferUsage : public RefCounted<GPUBufferUsage> {
-public:
-    enum Flags : GPUBufferUsageFlags {
-        None = 0,
-        MapRead = 1,
-        MapWrite = 2,
-        TransferSrc = 4,
-        TransferDst = 8,
-        Index = 16,
-        Vertex = 32,
-        Uniform = 64,
-        Storage = 128
-    };
+enum class GPUBufferMappedOption {
+    IsMapped,
+    NotMapped
 };
 
 } // namespace WebCore
 
 #endif // ENABLE(WEBGPU)
