diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUPipelineStageDescriptor.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUPipelineStageDescriptor.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUPipelineStageDescriptor.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUPipelineStageDescriptor.h
@@ -26,18 +26,27 @@
 #pragma once
 
 #if ENABLE(WEBGPU)
 
 #include "GPUShaderModule.h"
-
+#include <wtf/Ref.h>
 #include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
-struct GPUPipelineStageDescriptor {
-    const GPUShaderModule* module = nullptr;
+struct GPUPipelineStageDescriptorBase {
     String entryPoint;
 };
 
+struct GPUPipelineStageDescriptor : GPUPipelineStageDescriptorBase {
+    GPUPipelineStageDescriptor(Ref<const GPUShaderModule>&& module, const GPUPipelineStageDescriptorBase& base)
+        : GPUPipelineStageDescriptorBase(base)
+        , module { WTFMove(module) }
+    {
+    }
+
+    Ref<const GPUShaderModule> module;
+};
+
 } // namespace WebCore
 
 #endif // ENABLE(WEBGPU)
