diff a/cli/src/main/java/org/openjdk/skara/cli/pr/GitPrSet.java b/cli/src/main/java/org/openjdk/skara/cli/pr/GitPrSet.java
--- a/cli/src/main/java/org/openjdk/skara/cli/pr/GitPrSet.java
+++ b/cli/src/main/java/org/openjdk/skara/cli/pr/GitPrSet.java
@@ -113,11 +113,13 @@
         var host = getForge(uri, repo, arguments);
         var id = pullRequestIdArgument(repo, arguments);
         var pr = getPullRequest(uri, repo, host, id);
 
         var assigneesOption = getOption("assignees", "set", arguments);
-        if (assigneesOption != null) {
+        if (assigneesOption == null) {
+            pr.setAssignees(List.of());
+        } else {
             var usernames = Arrays.asList(assigneesOption.split(","));
             var assignees = usernames.stream()
                 .map(u -> host.user(u))
                 .filter(Optional::isPresent)
                 .map(Optional::get)
