diff a/application/tests/org.openjdk.jmc.ui.common.test/src/test/java/org/openjdk/jmc/ui/common/labelingrules/NameConverterTest.java b/application/tests/org.openjdk.jmc.ui.common.test/src/test/java/org/openjdk/jmc/ui/common/labelingrules/NameConverterTest.java
--- a/application/tests/org.openjdk.jmc.ui.common.test/src/test/java/org/openjdk/jmc/ui/common/labelingrules/NameConverterTest.java
+++ b/application/tests/org.openjdk.jmc.ui.common.test/src/test/java/org/openjdk/jmc/ui/common/labelingrules/NameConverterTest.java
@@ -37,13 +37,10 @@
 import static org.junit.Assert.assertTrue;
 
 import java.text.MessageFormat;
 
 import org.junit.Test;
-
-import org.openjdk.jmc.ui.common.labelingrules.NameConverter;
-import org.openjdk.jmc.ui.common.labelingrules.NamingRule;
 import org.openjdk.jmc.ui.common.util.Environment;
 
 @SuppressWarnings("nls")
 public class NameConverterTest {
 	private final static Object[] EXAMPLE_VALUES1 = new Object[] {"1.5", "[Unknown]",
