diff a/modules/javafx.graphics/src/main/native-iio/jpegloader.c b/modules/javafx.graphics/src/main/native-iio/jpegloader.c
--- a/modules/javafx.graphics/src/main/native-iio/jpegloader.c
+++ b/modules/javafx.graphics/src/main/native-iio/jpegloader.c
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2009, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2009, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -43,26 +43,18 @@
 #else
 #define jlong_to_ptr(a) ((void*)(int)(a))
 #define ptr_to_jlong(a) ((jlong)(int)(a))
 #endif
 
-/* On non iOS platforms we use JNI_OnLoad() shared library entrypoint. */
-#define USING_BUILTIN_LIBRARY_ENTRYPOINT 0
-
-/* On iOS we use builtin libraries, thus JNI_OnLoad_javafx_iio() is the entrypoint */
 #ifdef __APPLE__
 
 #include <TargetConditionals.h>
 
 /* RT-37125: use setjmp/longjmp versions that do not save/restore the signal mask */
 #define longjmp _longjmp
 #define setjmp _setjmp
 
-#if TARGET_OS_IPHONE /* iOS */
-#undef  USING_BUILTIN_LIBRARY_ENTRYPOINT
-#define USING_BUILTIN_LIBRARY_ENTRYPOINT 1
-#endif
 #endif
 
 static jboolean checkAndClearException(JNIEnv *env) {
     if (!(*env)->ExceptionCheck(env)) {
         return JNI_FALSE;
@@ -106,11 +98,11 @@
 
 /* Initialize the Java VM instance variable when the library is
    first loaded */
 static JavaVM *jvm;
 
-#if USING_BUILTIN_LIBRARY_ENTRYPOINT
+#ifdef STATIC_BUILD
 
 JNIEXPORT jint JNICALL
 JNI_OnLoad_javafx_iio(JavaVM *vm, void *reserved) {
     jvm = vm;
 #ifdef JNI_VERSION_1_8
@@ -131,11 +123,11 @@
 JNI_OnLoad(JavaVM *vm, void *reserved) {
     jvm = vm;
     return JNI_VERSION_1_2;
 }
 
-#endif
+#endif // STATIC_BUILD
 
 
 /*
  * The following sets of defines must match the warning messages in the
  * Java code.
@@ -1670,12 +1662,11 @@
             }
         }
 
         num_scanlines = jpeg_read_scanlines(cinfo, &scanline_ptr, 1);
         if (num_scanlines == 1) {
-            jboolean iscopy = FALSE;
-            jbyte *body = (*env)->GetPrimitiveArrayCritical(env, barray, &iscopy);
+            jbyte *body = (*env)->GetPrimitiveArrayCritical(env, barray, NULL);
             if (body == NULL) {
                 fprintf(stderr, "decompressIndirect: GetPrimitiveArrayCritical returns NULL: out of memory\n");
                 return JNI_FALSE;
             }
             memcpy(body+offset,scanline_ptr, bytes_per_row);
