diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.cpp
@@ -36,11 +36,11 @@
     : m_codeBlock(codeBlock->vm(), codeBlock)
     , m_rootJITType(JITType::None)
 {
 }
 
-DeferredSourceDump::DeferredSourceDump(CodeBlock* codeBlock, CodeBlock* rootCodeBlock, JITType rootJITType, unsigned callerBytecodeIndex)
+DeferredSourceDump::DeferredSourceDump(CodeBlock* codeBlock, CodeBlock* rootCodeBlock, JITType rootJITType, BytecodeIndex callerBytecodeIndex)
     : m_codeBlock(codeBlock->vm(), codeBlock)
     , m_rootCodeBlock(codeBlock->vm(), rootCodeBlock)
     , m_rootJITType(rootJITType)
     , m_callerBytecodeIndex(callerBytecodeIndex)
 {
@@ -54,11 +54,11 @@
     else
         dataLog("Compiled ");
     dataLog(*m_codeBlock);
 
     if (isInlinedFrame)
-        dataLog(" at ", CodeBlockWithJITType(*m_rootCodeBlock, m_rootJITType), " ", "bc#", m_callerBytecodeIndex);
+        dataLog(" at ", CodeBlockWithJITType(*m_rootCodeBlock, m_rootJITType), " ", m_callerBytecodeIndex);
 
     dataLog("\n'''");
     m_codeBlock->dumpSource();
     dataLog("'''\n");
 }
