diff a/src/java.base/share/classes/java/util/GregorianCalendar.java b/src/java.base/share/classes/java/util/GregorianCalendar.java
--- a/src/java.base/share/classes/java/util/GregorianCalendar.java
+++ b/src/java.base/share/classes/java/util/GregorianCalendar.java
@@ -729,11 +729,11 @@
 
     /**
      * Constructs an empty GregorianCalendar.
      *
      * @param zone    the given time zone
-     * @param aLocale the given locale
+     * @param locale  the given locale
      * @param flag    the flag requesting an empty instance
      */
     GregorianCalendar(TimeZone zone, Locale locale, boolean flag) {
         super(zone, locale);
         gdate = (BaseCalendar.Date) gcal.newCalendarDate(getZone());
