diff a/src/demo/share/jfc/SwingSet2/DemoModule.java b/src/demo/share/jfc/SwingSet2/DemoModule.java
--- a/src/demo/share/jfc/SwingSet2/DemoModule.java
+++ b/src/demo/share/jfc/SwingSet2/DemoModule.java
@@ -49,11 +49,11 @@
 /**
  * A generic SwingSet2 demo module
  *
  * @author Jeff Dinkins
  */
-public class DemoModule extends JApplet {
+public class DemoModule extends JFrame {
 
     // The preferred size of the demo
     private int PREFERRED_WIDTH = 680;
     private int PREFERRED_HEIGHT = 600;
 
@@ -188,11 +188,11 @@
         JFrame frame = new JFrame(getName());
         frame.getContentPane().setLayout(new BorderLayout());
         frame.getContentPane().add(getDemoPanel(), BorderLayout.CENTER);
         getDemoPanel().setPreferredSize(new Dimension(PREFERRED_WIDTH, PREFERRED_HEIGHT));
         frame.pack();
-        frame.show();
+        frame.setVisible(true);
     }
 
     public JPanel createHorizontalPanel(boolean threeD) {
         JPanel p = new JPanel();
         p.setLayout(new BoxLayout(p, BoxLayout.X_AXIS));
