diff a/test/hotspot/jtreg/vmTestbase/nsk/jvmti/Allocate/alloc001/alloc001.bash b/test/hotspot/jtreg/vmTestbase/nsk/jvmti/Allocate/alloc001/alloc001.bash
--- a/test/hotspot/jtreg/vmTestbase/nsk/jvmti/Allocate/alloc001/alloc001.bash
+++ b/test/hotspot/jtreg/vmTestbase/nsk/jvmti/Allocate/alloc001/alloc001.bash
@@ -1,7 +1,7 @@
 #!/bin/bash
-# Copyright (c) 2008, 2019, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2008, 2020, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.
@@ -26,11 +26,11 @@
 . ${TESTSRC}/../../../../../test_env.sh
 
 # Set virtual memory usage limit to be not 'unlimited' on unix platforms
 # This is workaround for 6683371.
 case $VM_OS in
-aix | bsd | linux | solaris)
+aix | bsd | linux)
     echo "Check virtual memory usage limits"
     soft_limit=`ulimit -S -v` || ""
     hard_limit=`ulimit -H -v` || ""
     echo "Virtual memory usage limit (hard): $hard_limit"
     echo "Virtual memory usage limit (soft): $soft_limit"
