diff a/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java b/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java
--- a/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java
+++ b/modules/javafx.graphics/src/main/java/javafx/concurrent/ScheduledService.java
@@ -144,11 +144,11 @@
  * @since JavaFX 8.0
  */
 public abstract class ScheduledService<V> extends Service<V> {
 
     /**
-     * Creates a default {@code ScheduledService}.
+     * Constructor for subclasses to call.
      */
     public ScheduledService() {
     }
 
     /**
