diff a/bots/submit/build.gradle b/bots/submit/build.gradle
--- a/bots/submit/build.gradle
+++ b/bots/submit/build.gradle
@@ -29,10 +29,11 @@
         opens 'org.openjdk.skara.bots.submit' to 'org.junit.platform.commons'
     }
 }
 
 dependencies {
+    implementation project(':ci')
     implementation project(':bot')
     implementation project(':host')
     implementation project(':forge')
     implementation project(':issuetracker')
     implementation project(':census')
