diff a/test/langtools/tools/javac/diags/examples/VarNotAllowedHere.java b/test/langtools/tools/javac/diags/examples/VarNotAllowedHere.java
--- a/test/langtools/tools/javac/diags/examples/VarNotAllowedHere.java
+++ b/test/langtools/tools/javac/diags/examples/VarNotAllowedHere.java
@@ -19,10 +19,10 @@
  * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
 
-// key: compiler.err.var.not.allowed.here
+// key: compiler.err.restricted.type.not.allowed.here
 
 class VarNotAllowedField {
     var s = "";
 }
