diff a/test/langtools/tools/javac/EarlyAssertWrapper.java b/test/langtools/tools/javac/EarlyAssertWrapper.java
--- a/test/langtools/tools/javac/EarlyAssertWrapper.java
+++ b/test/langtools/tools/javac/EarlyAssertWrapper.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2010, 2014, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2010, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -60,12 +60,15 @@
         PrintWriter pw = new PrintWriter(sw);
 
         String line;
         DataInputStream in = new DataInputStream(p.getInputStream());
         try {
-        while ((line = in.readLine()) != null)
-            pw.println(line);
+            while ((line = in.readLine()) != null) {
+                if (!line.matches("^Picked up .*JAVA.*OPTIONS:.*")) {
+                    pw.println(line);
+                }
+            }
         } finally {
             in.close();
         }
         pw.close();
 
