diff a/test/langtools/tools/javac/TextBlockLang.java b/test/langtools/tools/javac/TextBlockLang.java
--- a/test/langtools/tools/javac/TextBlockLang.java
+++ b/test/langtools/tools/javac/TextBlockLang.java
@@ -23,12 +23,12 @@
 
 /*
  * @test
  * @bug 8223967
  * @summary Unit tests for Text Block language changes
- * @compile --enable-preview -source ${jdk.version} -encoding utf8 TextBlockLang.java
- * @run main/othervm --enable-preview TextBlockLang
+ * @compile -encoding utf8 TextBlockLang.java
+ * @run main TextBlockLang
  */
 
 public class TextBlockLang {
     public static void main(String... args) {
         test1();
