diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/MediaSourceRegistry.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/MediaSourceRegistry.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/MediaSourceRegistry.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/MediaSourceRegistry.cpp
@@ -56,11 +56,11 @@
     MediaSource& source = static_cast<MediaSource&>(registrable);
     source.addedToRegistry();
     m_mediaSources.set(url.string(), &source);
 }
 
-void MediaSourceRegistry::unregisterURL(ScriptExecutionContext&, const URL& url)
+void MediaSourceRegistry::unregisterURL(const URL& url)
 {
     ASSERT(isMainThread());
     if (auto source = m_mediaSources.take(url.string()))
         source->removedFromRegistry();
 }
