diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/SourceBufferList.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/SourceBufferList.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/SourceBufferList.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasource/SourceBufferList.h
@@ -76,17 +76,13 @@
     void scheduleEvent(const AtomString&);
 
     void refEventTarget() override { ref(); }
     void derefEventTarget() override { deref(); }
 
-    bool canSuspendForDocumentSuspension() const final;
-    void suspend(ReasonForSuspension) final;
-    void resume() final;
-    void stop() final;
     const char* activeDOMObjectName() const final;
 
-    GenericEventQueue m_asyncEventQueue;
+    UniqueRef<MainThreadGenericEventQueue> m_asyncEventQueue;
 
     Vector<RefPtr<SourceBuffer>> m_list;
 };
 
 } // namespace WebCore
