diff a/vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java b/vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java
@@ -669,10 +669,15 @@
             }
             return new Branch(res.stdout().get(0));
         }
     }
 
+    @Override
+    public Optional<Bookmark> currentBookmark() throws IOException {
+        throw new RuntimeException("git does not have bookmarks");
+    }
+
     @Override
     public Branch defaultBranch() throws IOException {
         try (var p = capture("git", "symbolic-ref", "--short", "refs/remotes/origin/HEAD")) {
             var res = p.await();
             if (res.status() == 0 && res.stdout().size() == 1) {
