diff a/modules/javafx.graphics/src/main/java/com/sun/javafx/tk/Toolkit.java b/modules/javafx.graphics/src/main/java/com/sun/javafx/tk/Toolkit.java
--- a/modules/javafx.graphics/src/main/java/com/sun/javafx/tk/Toolkit.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/javafx/tk/Toolkit.java
@@ -151,12 +151,12 @@
         "api-ms-win-crt-utility-l1-1-0",
         "ucrtbase",
 
         // Finally load VS 2017 DLLs in the following order
         "vcruntime140",
-        "msvcp140",
-        "concrt140"
+        "vcruntime140_1",
+        "msvcp140"
 };
 
     private static String lookupToolkitClass(String name) {
         if ("prism".equalsIgnoreCase(name)) {
             return QUANTUM_TOOLKIT;
@@ -205,11 +205,11 @@
             // java class, VersionInfo, created at build time.
             VersionInfo.setupSystemProperties();
             return null;
         });
 
-        // This loading of msvcp140.dll and vcruntime140.dll (VS2017) is required on Windows platforms
+        // Load required Microsoft runtime DLLs on Windows platforms
         if (PlatformUtil.isWindows()) {
             loadMSWindowsLibraries();
         }
 
         boolean userSpecifiedToolkit = true;
