diff a/src/demo/share/jfc/J2Ddemo/java2d/demos/Paint/GradAnim.java b/src/demo/share/jfc/J2Ddemo/java2d/demos/Paint/GradAnim.java
--- a/src/demo/share/jfc/J2Ddemo/java2d/demos/Paint/GradAnim.java
+++ b/src/demo/share/jfc/J2Ddemo/java2d/demos/Paint/GradAnim.java
@@ -264,17 +264,17 @@
 
 
     class DemoControls extends CustomControls implements ActionListener {
 
         GradAnim demo;
-        JComboBox combo;
+        JComboBox<String> combo;
 
         @SuppressWarnings("LeakingThisInConstructor")
         public DemoControls(GradAnim demo) {
             super(demo.name);
             this.demo = demo;
-            combo = new JComboBox();
+            combo = new JComboBox<>();
             combo.addActionListener(this);
             combo.addItem("2-color GradientPaint");
             combo.addItem("3-color LinearGradientPaint");
             combo.addItem("4-color RadialGradientPaint");
             combo.addItem("4-color RadialGradientPaint with focus");
