diff a/src/java.base/share/classes/sun/security/x509/DistributionPoint.java b/src/java.base/share/classes/sun/security/x509/DistributionPoint.java
--- a/src/java.base/share/classes/sun/security/x509/DistributionPoint.java
+++ b/src/java.base/share/classes/sun/security/x509/DistributionPoint.java
@@ -275,11 +275,11 @@
      * Write the DistributionPoint value to the DerOutputStream.
      *
      * @param out the DerOutputStream to write the extension to.
      * @exception IOException on error.
      */
-    public void encode(DerOutputStream out) throws IOException {
+    public void encode(DerOutputStream out) {
         DerOutputStream tagged = new DerOutputStream();
 
         // NOTE: only one of pointNames and pointRDN can be set
         if ((fullName != null) || (relativeName != null)) {
             DerOutputStream distributionPoint = new DerOutputStream();
@@ -289,11 +289,11 @@
                 distributionPoint.writeImplicit(
                     DerValue.createTag(DerValue.TAG_CONTEXT, true, TAG_FULL_NAME),
                     derOut);
             } else if (relativeName != null) {
                 DerOutputStream derOut = new DerOutputStream();
-                relativeName.encode(derOut);
+                relativeName.derEncode(derOut);
                 distributionPoint.writeImplicit(
                     DerValue.createTag(DerValue.TAG_CONTEXT, true, TAG_REL_NAME),
                     derOut);
             }
             tagged.write(
