diff a/src/java.base/share/classes/sun/security/x509/OtherName.java b/src/java.base/share/classes/sun/security/x509/OtherName.java
--- a/src/java.base/share/classes/sun/security/x509/OtherName.java
+++ b/src/java.base/share/classes/sun/security/x509/OtherName.java
@@ -147,14 +147,15 @@
      * Encode the Other name into the DerOutputStream.
      *
      * @param out the DER stream to encode the Other-Name to.
      * @exception IOException on encoding errors.
      */
-    public void encode(DerOutputStream out) throws IOException {
+    @Override
+    public void derEncode(DerOutputStream out) {
         if (gni != null) {
             // This OtherName has a supported class
-            gni.encode(out);
+            gni.derEncode(out);
             return;
         } else {
             // This OtherName has no supporting class
             DerOutputStream tmp = new DerOutputStream();
             tmp.putOID(oid);
