diff a/src/java.base/share/classes/sun/security/x509/ReasonFlags.java b/src/java.base/share/classes/sun/security/x509/ReasonFlags.java
--- a/src/java.base/share/classes/sun/security/x509/ReasonFlags.java
+++ b/src/java.base/share/classes/sun/security/x509/ReasonFlags.java
@@ -238,11 +238,11 @@
      * Write the extension to the DerOutputStream.
      *
      * @param out the DerOutputStream to write the extension to.
      * @exception IOException on encoding errors.
      */
-    public void encode(DerOutputStream out) throws IOException {
+    public void encode(DerOutputStream out) {
         out.putTruncatedUnalignedBitString(new BitArray(this.bitString));
     }
 
     /**
      * Return an enumeration of names of attributes existing within this
