diff a/make/CompileJavaModules.gmk b/make/CompileJavaModules.gmk
--- a/make/CompileJavaModules.gmk
+++ b/make/CompileJavaModules.gmk
@@ -301,11 +301,11 @@
     $(patsubst %, $(TOPDIR)/src/jdk.compiler/share/classes/%/*.properties, \
         sun/tools/serialver/resources))
 
 ################################################################################
 
-jdk.hotspot.agent_DISABLED_WARNINGS += deprecation rawtypes serial unchecked \
+jdk.hotspot.agent_DISABLED_WARNINGS += rawtypes serial unchecked \
     cast static overrides fallthrough
 jdk.hotspot.agent_COPY += .gif .png sa.js .properties
 
 ################################################################################
 
@@ -670,16 +670,10 @@
   $($(MODULE)_COMPILE_TARGET): $(JDK_OUTPUTDIR)/modules/$(MODULE)/_imported.marker
 endif
 
 ################################################################################
 
-ifeq ($(MODULE), jdk.scripting.nashorn)
-  include CompileJavaModulesNashorn.gmk
-endif
-
-################################################################################
-
 $(eval $(call IncludeCustomExtension, CompileJavaModules-post.gmk))
 
 ################################################################################
 
 all: $(TARGETS)
