diff a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java
@@ -1022,11 +1022,11 @@
 
         /** All the interfaces of this class, including missing ones.
          */
         public List<Type> all_interfaces_field;
 
-        /* The `other' projection: If `this' is type of an inline class, then 'projection' is the
+        /* The 'other' projection: If 'this' is type of an inline class, then 'projection' is the
            its doppleganger in its referene projection world and vice versa.
         */
         public ClassType projection;
 
         public ClassType(Type outer, List<Type> typarams, TypeSymbol tsym) {
