diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
@@ -286,11 +286,13 @@
                     issue.setState(Issue.State.RESOLVED);
                     if (issue.assignees().isEmpty()) {
                         var username = findIssueUsername(commit);
                         if (username.isPresent()) {
                             var assignee = issueProject.issueTracker().user(username.get());
-                            issue.setAssignees(List.of(assignee));
+                            if (assignee.isPresent()) {
+                                issue.setAssignees(List.of(assignee.get()));
+                            }
                         }
                     }
                 }
 
                 if (commitLink) {
