diff a/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp b/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp
--- a/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp
+++ b/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp
@@ -24,14 +24,10 @@
  */
 
 #ifndef CPU_PPC_GLOBALDEFINITIONS_PPC_HPP
 #define CPU_PPC_GLOBALDEFINITIONS_PPC_HPP
 
-#ifdef CC_INTERP
-#error "CC_INTERP is no longer supported. Removed in change 8145117."
-#endif
-
 #ifndef FILE_AND_LINE
 #define FILE_AND_LINE __FILE__ ":" XSTR(__LINE__)
 #endif
 
 // Size of PPC Instructions
