diff a/src/java.desktop/macosx/native/libawt_lwawt/awt/shaders.metal b/src/java.desktop/macosx/native/libawt_lwawt/awt/shaders.metal
--- a/src/java.desktop/macosx/native/libawt_lwawt/awt/shaders.metal
+++ b/src/java.desktop/macosx/native/libawt_lwawt/awt/shaders.metal
@@ -185,14 +185,21 @@
 }
 
 fragment half4 aa_frag_txt(
         TxtShaderInOut vert [[stage_in]],
         texture2d<float, access::sample> renderTexture [[texture(0)]],
+        texture2d<float, access::sample> stencilTexture [[texture(1)]],
         constant TxtFrameUniforms& uniforms [[buffer(1)]],
         sampler textureSampler [[sampler(0)]]
 ) {
     float4 pixelColor = renderTexture.sample(textureSampler, vert.texCoords);
+    if (!is_null_texture(stencilTexture)) {
+        float4 stencil = stencilTexture.sample(textureSampler, vert.texCoords);
+        if (stencil.r ==  0.0) {
+            discard_fragment();
+        }
+    }
     return half4(pixelColor.r, pixelColor.g, pixelColor.b, pixelColor.a);
 }
 
 fragment half4 frag_grad(GradShaderInOut in [[stage_in]],
                          constant GradFrameUniforms& uniforms [[buffer(0)]]) {
