diff a/test/hotspot/jtreg/vmTestbase/gc/gctests/gctest02/gctest02.java b/test/hotspot/jtreg/vmTestbase/gc/gctests/gctest02/gctest02.java
--- a/test/hotspot/jtreg/vmTestbase/gc/gctests/gctest02/gctest02.java
+++ b/test/hotspot/jtreg/vmTestbase/gc/gctests/gctest02/gctest02.java
@@ -178,11 +178,11 @@
 class Escaper extends Thread {
         public void run() {
                 while ( ThreadCount.get() > 0 ) {
                         int buf[] = new int[32];
                         {
-                                                yield();
+                                                Thread.yield();
                         }
                 }
         }
 }
 
