diff a/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLTransform.h b/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLTransform.h
--- a/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLTransform.h
+++ b/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLTransform.h
@@ -13,9 +13,11 @@
 - (void)setTransformM00:(jdouble) m00 M10:(jdouble) m10
                     M01:(jdouble) m01 M11:(jdouble) m11
                     M02:(jdouble) m02 M12:(jdouble) m12;
 - (void)resetTransform;
 
-- (void)setVertexMatrix:(id<MTLRenderCommandEncoder>)encoder destWidth:(jint)dw destHeight:(jint)dh;
+- (void)setVertexMatrix:(id<MTLRenderCommandEncoder>)encoder
+              destWidth:(NSUInteger)dw
+             destHeight:(NSUInteger)dh;
 @end
 
 #endif // MTLTransform_h_Included
