diff a/make/hotspot/lib/JvmFeatures.gmk b/make/hotspot/lib/JvmFeatures.gmk
--- a/make/hotspot/lib/JvmFeatures.gmk
+++ b/make/hotspot/lib/JvmFeatures.gmk
@@ -171,10 +171,11 @@
 endif
 
 ifneq ($(call check-jvm-feature, jfr), true)
   JVM_CFLAGS_FEATURES += -DINCLUDE_JFR=0
   JVM_EXCLUDE_PATTERNS += jfr
+  JVM_EXCLUDE_FILES += compilerEvent.cpp
 endif
 
 ################################################################################
 
 ifeq ($(call check-jvm-feature, link-time-opt), true)
