diff a/modules/javafx.controls/src/main/java/javafx/scene/control/skin/ComboBoxListViewSkin.java b/modules/javafx.controls/src/main/java/javafx/scene/control/skin/ComboBoxListViewSkin.java
--- a/modules/javafx.controls/src/main/java/javafx/scene/control/skin/ComboBoxListViewSkin.java
+++ b/modules/javafx.controls/src/main/java/javafx/scene/control/skin/ComboBoxListViewSkin.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2010, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2010, 2018, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -415,25 +415,22 @@
                 int index = comboBox.getSelectionModel().getSelectedIndex();
                 if (index >= 0 && index < comboBoxItems.size()) {
                     T itemsObj = comboBoxItems.get(index);
                     if ((itemsObj != null && itemsObj.equals(newValue)) || (itemsObj == null && newValue == null)) {
                         listViewSM.select(index);
-                        listView.scrollTo(index);
                     } else {
                         listViewSM.select(newValue);
-                        listView.scrollTo(newValue);
                     }
                 } else {
                     // just select the first instance of newValue in the list
                     int listViewIndex = comboBoxItems.indexOf(newValue);
                     if (listViewIndex == -1) {
                         // RT-21336 Show the ComboBox value even though it doesn't
                         // exist in the ComboBox items list (part one of fix)
                         updateDisplayNode();
                     } else {
                         listViewSM.select(listViewIndex);
-                        listView.scrollTo(listViewIndex);
                     }
                 }
             }
         }
     }
