diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.idl
@@ -30,12 +30,12 @@
     readonly attribute [OverrideIDLType=IDLIDBKey] any lower;
     readonly attribute [OverrideIDLType=IDLIDBKey] any upper;
     readonly attribute boolean lowerOpen;
     readonly attribute boolean upperOpen;
 
-    [CallWith=ExecState, MayThrowException] static IDBKeyRange only(any value);
-    [CallWith=ExecState, MayThrowException] static IDBKeyRange lowerBound(any lower, optional boolean open = false);
-    [CallWith=ExecState, MayThrowException] static IDBKeyRange upperBound(any upper, optional boolean open = false);
-    [CallWith=ExecState, MayThrowException] static IDBKeyRange bound(any lower, any upper, optional boolean lowerOpen = false, optional boolean upperOpen = false);
+    [CallWith=GlobalObject, MayThrowException] static IDBKeyRange only(any value);
+    [CallWith=GlobalObject, MayThrowException] static IDBKeyRange lowerBound(any lower, optional boolean open = false);
+    [CallWith=GlobalObject, MayThrowException] static IDBKeyRange upperBound(any upper, optional boolean open = false);
+    [CallWith=GlobalObject, MayThrowException] static IDBKeyRange bound(any lower, any upper, optional boolean lowerOpen = false, optional boolean upperOpen = false);
 
-    [CallWith=ExecState, MayThrowException] boolean includes(any key);
+    [CallWith=GlobalObject, MayThrowException] boolean includes(any key);
 };
