diff a/src/java.desktop/share/classes/java/applet/Applet.java b/src/java.desktop/share/classes/java/applet/Applet.java
--- a/src/java.desktop/share/classes/java/applet/Applet.java
+++ b/src/java.desktop/share/classes/java/applet/Applet.java
@@ -542,10 +542,11 @@
     //
 
     /**
      * The accessible context associated with this {@code Applet}.
      */
+    @SuppressWarnings("serial") // Not statically typed as Serializable
     AccessibleContext accessibleContext = null;
 
     /**
      * Gets the {@code AccessibleContext} associated with this {@code Applet}.
      * For applets, the {@code AccessibleContext} takes the form of an
