diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/classfile/AccessFlags.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/classfile/AccessFlags.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/classfile/AccessFlags.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/classfile/AccessFlags.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2017, 2020, Oracle and/or its affiliates. All rights reserved.
  */
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
  * this work for additional information regarding copyright ownership.
@@ -23,21 +23,22 @@
 
 /**
  * Super class for all objects that have modifiers like private, final, ... I.e.
  * classes, fields, and methods.
  *
- * @version $Id: AccessFlags.java 1748636 2016-06-15 20:45:17Z dbrosius $
+ * @LastModified: Jan 2020
  */
 public abstract class AccessFlags {
 
     private int access_flags;
 
     public AccessFlags() {
     }
 
     /**
-     * @param a inital access flags
+     * @param a
+     *            inital access flags
      */
     public AccessFlags(final int a) {
         access_flags = a;
     }
 
@@ -56,20 +57,22 @@
     }
 
     /**
      * Set access flags aka "modifiers".
      *
-     * @param access_flags Access flags of the object.
+     * @param access_flags
+     *            Access flags of the object.
      */
     public final void setAccessFlags(final int access_flags) {
         this.access_flags = access_flags;
     }
 
     /**
      * Set access flags aka "modifiers".
      *
-     * @param access_flags Access flags of the object.
+     * @param access_flags
+     *            Access flags of the object.
      */
     public final void setModifiers(final int access_flags) {
         setAccessFlags(access_flags);
     }
 
