diff a/modules/javafx.graphics/src/main/java/com/sun/prism/impl/QueuedPixelSource.java b/modules/javafx.graphics/src/main/java/com/sun/prism/impl/QueuedPixelSource.java
--- a/modules/javafx.graphics/src/main/java/com/sun/prism/impl/QueuedPixelSource.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/prism/impl/QueuedPixelSource.java
@@ -98,11 +98,11 @@
     }
 
     private boolean usesSameBuffer(Pixels p1, Pixels p2) {
         if (p1 == p2) return true;
         if (p1 == null || p2 == null) return false;
-        return (p1.getPixels() == p2.getPixels());
+        return (p1.getBuffer() == p2.getBuffer());
     }
 
     /**
      * Return an unused Pixels with the indicated dimensions and scale.
      * The returned object may either be saved from a previous use, but
