diff a/modules/javafx.web/src/main/java/javafx/scene/web/HTMLEditorSkin.java b/modules/javafx.web/src/main/java/javafx/scene/web/HTMLEditorSkin.java
--- a/modules/javafx.web/src/main/java/javafx/scene/web/HTMLEditorSkin.java
+++ b/modules/javafx.web/src/main/java/javafx/scene/web/HTMLEditorSkin.java
@@ -961,17 +961,17 @@
         fontFamilyComboBox.setDisable(!isCommandEnabled(FONT_FAMILY.getCommand()));
         final String fontFamilyValue = getCommandValue(FONT_FAMILY.getCommand());
         if (fontFamilyValue != null) {
             String fontFamilyStr = fontFamilyValue;
 
-            // stripping out apostrophe characters, which are appended to either
+            // stripping out apostrophe and quote characters, which are appended to either
             // end of the font face name when the font face has one or more spaces.
-            if (fontFamilyStr.startsWith("'")) {
+            if (fontFamilyStr.startsWith("'") || fontFamilyStr.startsWith("\"")) {
                 fontFamilyStr = fontFamilyStr.substring(1);
             }
-            if (fontFamilyStr.endsWith("'")) {
-                fontFamilyStr = fontFamilyStr.substring(0,fontFamilyStr.length() - 1);
+            if (fontFamilyStr.endsWith("'") || fontFamilyStr.endsWith("\"")) {
+                fontFamilyStr = fontFamilyStr.substring(0, fontFamilyStr.length() - 1);
             }
 
             Object selectedFont = fontFamilyComboBox.getValue();
             if (selectedFont instanceof String) {
                 if (!selectedFont.equals(fontFamilyStr)) {
