diff a/application/org.openjdk.jmc.joverflow/src/main/java/org/openjdk/jmc/joverflow/heap/model/JavaClass.java b/application/org.openjdk.jmc.joverflow/src/main/java/org/openjdk/jmc/joverflow/heap/model/JavaClass.java
--- a/application/org.openjdk.jmc.joverflow/src/main/java/org/openjdk/jmc/joverflow/heap/model/JavaClass.java
+++ b/application/org.openjdk.jmc.joverflow/src/main/java/org/openjdk/jmc/joverflow/heap/model/JavaClass.java
@@ -611,11 +611,12 @@
 	public String getName() {
 		return name;
 	}
 
 	/**
-	 * Same as {@link org.openjdk.jmc.joverflow.util.ClassUtils#getShortNameForPopularClass(String)}, but
+	 * Same as
+	 * {@link org.openjdk.jmc.joverflow.util.ClassUtils#getShortNameForPopularClass(String)}, but
 	 * also: - makes names for anonymous classes, like MyFooClass$6, more informative, by adding
 	 * "(SuperClassName)" to them; - For array classes, returns a human-friendly name, such as
 	 * "boolean[]" instead of "[B" or "Object[]" instead of "[Ljava.lang.Object;"
 	 */
 	public String getHumanFriendlyName() {
