diff a/modules/javafx.web/src/main/native/Source/WebCore/plugins/DOMPlugin.h b/modules/javafx.web/src/main/native/Source/WebCore/plugins/DOMPlugin.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/plugins/DOMPlugin.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/plugins/DOMPlugin.h
@@ -28,11 +28,12 @@
 namespace WebCore {
 
 class Plugin;
 class PluginData;
 
-class DOMPlugin : public ScriptWrappable, public RefCounted<DOMPlugin>, public FrameDestructionObserver {
+class DOMPlugin final : public ScriptWrappable, public RefCounted<DOMPlugin>, public FrameDestructionObserver {
+    WTF_MAKE_ISO_ALLOCATED(DOMPlugin);
 public:
     static Ref<DOMPlugin> create(PluginData* pluginData, Frame* frame, PluginInfo pluginInfo) { return adoptRef(*new DOMPlugin(pluginData, frame, WTFMove(pluginInfo))); }
     ~DOMPlugin();
 
     String name() const;
@@ -40,12 +41,12 @@
     String description() const;
 
     unsigned length() const;
 
     RefPtr<DOMMimeType> item(unsigned index);
-    RefPtr<DOMMimeType> namedItem(const AtomicString& propertyName);
-    Vector<AtomicString> supportedPropertyNames();
+    RefPtr<DOMMimeType> namedItem(const AtomString& propertyName);
+    Vector<AtomString> supportedPropertyNames();
 
 private:
     DOMPlugin(PluginData*, Frame*, PluginInfo);
     RefPtr<PluginData> m_pluginData;
     PluginInfo m_pluginInfo;
