diff a/src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp b/src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp
--- a/src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp
+++ b/src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp
@@ -396,10 +396,23 @@
   ptrIDebugSymbols->SetImagePath(imagePath);
   ptrIDebugSymbols->SetSymbolPath(symbolPath);
   return true;
 }
 
+static HRESULT WaitForEvent(IDebugControl *ptrIDebugControl) {
+  HRESULT hr = ptrIDebugControl->WaitForEvent(DEBUG_WAIT_DEFAULT, INFINITE);
+  // see JDK-8204994: sometimes WaitForEvent fails with E_ACCESSDENIED,
+  // but succeeds on 2nd call.
+  // To minimize possible noise retry 3 times.
+  for (int i = 0; hr == E_ACCESSDENIED && i < 3; i++) {
+    // yield current thread use of a processor (short delay).
+    SwitchToThread();
+    hr = ptrIDebugControl->WaitForEvent(DEBUG_WAIT_DEFAULT, INFINITE);
+  }
+  return hr;
+}
+
 static bool openDumpFile(JNIEnv* env, jobject obj, jstring coreFileName) {
   // open the dump file
   AutoJavaString coreFile(env, coreFileName);
   CHECK_EXCEPTION_(false);
   if (!setImageAndSymbolPath(env, obj)) {
@@ -411,11 +424,11 @@
   COM_VERIFY_OK_(ptrIDebugClient->OpenDumpFile(coreFile),
                  "Windbg Error: OpenDumpFile failed!", false);
 
   IDebugControl* ptrIDebugControl = (IDebugControl*)env->GetLongField(obj, ptrIDebugControl_ID);
   CHECK_EXCEPTION_(false);
-  COM_VERIFY_OK_(ptrIDebugControl->WaitForEvent(DEBUG_WAIT_DEFAULT, INFINITE),
+  COM_VERIFY_OK_(WaitForEvent(ptrIDebugControl),
                  "Windbg Error: WaitForEvent failed!", false);
 
   return true;
 }
 
@@ -448,11 +461,11 @@
                  "Windbg Error: AttachProcess failed!", false);
 
   IDebugControl* ptrIDebugControl = (IDebugControl*) env->GetLongField(obj,
                                                      ptrIDebugControl_ID);
   CHECK_EXCEPTION_(false);
-  COM_VERIFY_OK_(ptrIDebugControl->WaitForEvent(DEBUG_WAIT_DEFAULT, INFINITE),
+  COM_VERIFY_OK_(WaitForEvent(ptrIDebugControl),
                  "Windbg Error: WaitForEvent failed!", false);
 
   return true;
 }
 
