diff a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/UnboundedAllocationScope.java b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/UnboundedAllocationScope.java
--- a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/UnboundedAllocationScope.java
+++ b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/UnboundedAllocationScope.java
@@ -23,24 +23,24 @@
  * questions.
  */
 
 package jdk.internal.foreign;
 
-import jdk.incubator.foreign.AllocationScope;
+import jdk.incubator.foreign.NativeAllocationScope;
 import jdk.incubator.foreign.MemoryAddress;
 import jdk.incubator.foreign.MemorySegment;
 
 import java.util.ArrayList;
 import java.util.List;
 import java.util.OptionalLong;
 import java.util.function.LongFunction;
 
-public class UnboundedAllocationScope extends AllocationScope {
+public class UnboundedAllocationScope extends NativeAllocationScope {
 
     private static final long BLOCK_SIZE = 4 * 1024;
+    private static final long MAX_ALLOC_SIZE = BLOCK_SIZE / 2;
 
-    private final LongFunction<MemorySegment> segmentFactory;
     private final List<MemorySegment> usedSegments = new ArrayList<>();
     private MemorySegment segment;
     private long sp = 0L;
     private long size = 0L;
 
@@ -52,17 +52,22 @@
     @Override
     public long allocatedBytes() {
         return size;
     }
 
-    public UnboundedAllocationScope(LongFunction<MemorySegment> segmentFactory) {
-        this.segmentFactory = segmentFactory;
-        this.segment = segmentFactory.apply(BLOCK_SIZE);
+    public UnboundedAllocationScope() {
+        this.segment = MemorySegment.allocateNative(BLOCK_SIZE);
     }
 
     @Override
     public MemoryAddress allocate(long bytesSize, long bytesAlignment) {
+        if (bytesSize > MAX_ALLOC_SIZE) {
+            MemorySegment segment = MemorySegment.allocateNative(bytesSize, bytesAlignment);
+            usedSegments.add(segment);
+            return segment.withAccessModes(MemorySegment.READ | MemorySegment.WRITE | MemorySegment.ACQUIRE)
+                    .baseAddress();
+        }
         for (int i = 0; i < 2; i++) {
             long min = ((MemoryAddressImpl) segment.baseAddress()).unsafeGetOffset();
             long start = Utils.alignUp(min + sp, bytesAlignment) - min;
             try {
                 MemorySegment slice = segment.asSlice(start, bytesSize)
@@ -71,14 +76,14 @@
                 size += Utils.alignUp(bytesSize, bytesAlignment);
                 return slice.baseAddress();
             } catch (IndexOutOfBoundsException ex) {
                 sp = 0L;
                 usedSegments.add(segment);
-                segment = segmentFactory.apply(BLOCK_SIZE);
+                segment = MemorySegment.allocateNative(BLOCK_SIZE);
             }
         }
-        throw new OutOfMemoryError("Allocation request exceeds scope block size");
+        throw new AssertionError("Cannot get here!");
     }
 
     @Override
     public void close() {
         segment.close();
