diff a/test/langtools/jdk/javadoc/tool/sourceOption/SourceOption.java b/test/langtools/jdk/javadoc/tool/sourceOption/SourceOption.java
--- a/test/langtools/jdk/javadoc/tool/sourceOption/SourceOption.java
+++ b/test/langtools/jdk/javadoc/tool/sourceOption/SourceOption.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2006, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2006, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -23,11 +23,10 @@
 
 /*
  * @test
  * @bug     6507179
  * @summary Ensure that "-source" option isn't ignored.
- * @author  Scott Seligman
  * @modules jdk.javadoc/jdk.javadoc.internal.tool
  * @run main/fail SourceOption 7
  * @run main      SourceOption 9
  * @run main      SourceOption
  */
