diff a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/model/MemoryStatisticsItem.java b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/model/MemoryStatisticsItem.java
--- a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/model/MemoryStatisticsItem.java
+++ b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/model/MemoryStatisticsItem.java
@@ -31,12 +31,12 @@
  * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 package org.openjdk.jmc.joverflow.ui.model;
 
 /**
- * Class holding an aggregate of mOverhead/mMemory/mSize for a number of {@code ObjectCluster}. The group is identified by
- * {@code id}
+ * Class holding an aggregate of mOverhead/mMemory/mSize for a number of {@code ObjectCluster}. The
+ * group is identified by {@code id}
  */
 public class MemoryStatisticsItem {
 	private final Object id;
 	private Integer index;
 
