diff a/src/hotspot/share/oops/access.hpp b/src/hotspot/share/oops/access.hpp
--- a/src/hotspot/share/oops/access.hpp
+++ b/src/hotspot/share/oops/access.hpp
@@ -218,11 +218,11 @@
   static inline void clone(oop src, oop dst, size_t size) {
     verify_decorators<IN_HEAP>();
     AccessInternal::clone<decorators>(src, dst, size);
   }
 
-  // Value type inline heap access (flattened)...
+  // inline type heap access (when allocated inline)...
 
   // Copy value type data from src to dst
   static inline void value_copy(void* src, void* dst, ValueKlass* md) {
     verify_heap_value_decorators<IN_HEAP>();
     AccessInternal::value_copy<decorators>(src, dst, md);
