diff a/src/java.desktop/share/classes/javax/swing/plaf/metal/DefaultMetalTheme.java b/src/java.desktop/share/classes/javax/swing/plaf/metal/DefaultMetalTheme.java
--- a/src/java.desktop/share/classes/javax/swing/plaf/metal/DefaultMetalTheme.java
+++ b/src/java.desktop/share/classes/javax/swing/plaf/metal/DefaultMetalTheme.java
@@ -40,11 +40,11 @@
  * the default theme.
  * <p>
  * All colors returned by {@code DefaultMetalTheme} are completely
  * opaque.
  *
- * <h3><a id="fontStyle"></a>Font Style</h3>
+ * <h2><a id="fontStyle"></a>Font Style</h2>
  *
  * {@code DefaultMetalTheme} uses bold fonts for many controls.  To make all
  * controls (with the exception of the internal frame title bars and
  * client decorated frame title bars) use plain fonts you can do either of
  * the following:
