diff a/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java b/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java
--- a/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java
+++ b/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java
@@ -1651,10 +1651,13 @@
                 var elapsed = Duration.between(start, Instant.now());
                 if (elapsed.compareTo(cooldown) < 0) {
                     break;
                 } else {
                     log.info("Didn't do the test in time - retrying (elapsed: " + elapsed + " required: " + cooldown + ")");
+                    // Make sure to run the bot once more after the cooldown has expired to reset the state
+                    Thread.sleep(cooldown.toMillis());
+                    TestBotRunner.runPeriodicItems(mlBot);
                     listServer.processIncoming();
                     cooldown = cooldown.multipliedBy(2);
                     mlBot = mlBotBuilder.cooldown(cooldown).build();
                 }
             }
@@ -1731,10 +1734,13 @@
                 var elapsed = Duration.between(start, Instant.now());
                 if (elapsed.compareTo(cooldown) < 0) {
                     break;
                 } else {
                     log.info("Didn't do the test in time - retrying (elapsed: " + elapsed + " required: " + cooldown + ")");
+                    // Make sure to run the bot once more after the cooldown has expired to reset the state
+                    Thread.sleep(cooldown.toMillis());
+                    TestBotRunner.runPeriodicItems(mlBot);
                     listServer.processIncoming();
                     cooldown = cooldown.multipliedBy(2);
                     mlBot = mlBotBuilder.cooldown(cooldown).build();
                 }
             }
