diff a/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java b/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/Repository.java
@@ -27,11 +27,11 @@
 
 import java.io.IOException;
 import java.net.URI;
 import java.nio.file.Path;
 import java.nio.file.Files;
-import java.time.Instant;
+import java.time.ZonedDateTime;
 import java.util.*;
 
 public interface Repository extends ReadOnlyRepository {
     Repository init() throws IOException;
     void checkout(Hash h, boolean force) throws IOException;
@@ -60,11 +60,11 @@
                 String authorName,
                 String authorEmail) throws IOException;
     Hash commit(String message,
                 String authorName,
                 String authorEmail,
-                Instant timestamp) throws IOException;
+                ZonedDateTime date) throws IOException;
     Hash commit(String message,
                 String authorName,
                 String authorEmail,
                 String committerName,
                 String committerEmail) throws IOException;
@@ -69,14 +69,14 @@
                 String committerName,
                 String committerEmail) throws IOException;
     Hash commit(String message,
                 String authorName,
                 String authorEmail,
-                Instant authorDate,
+                ZonedDateTime authorDate,
                 String committerName,
                 String committerEmail,
-                Instant committerDate) throws IOException;
+                ZonedDateTime committerDate) throws IOException;
     Hash amend(String message,
                String authorName,
                String authorEmail) throws IOException;
     Hash amend(String message,
                String authorName,
