diff a/src/java.base/share/classes/java/time/DateTimeException.java b/src/java.base/share/classes/java/time/DateTimeException.java
--- a/src/java.base/share/classes/java/time/DateTimeException.java
+++ b/src/java.base/share/classes/java/time/DateTimeException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2012, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2012, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -75,10 +75,11 @@
 public class DateTimeException extends RuntimeException {
 
     /**
      * Serialization version.
      */
+    @java.io.Serial
     private static final long serialVersionUID = -1632418723876261839L;
 
     /**
      * Constructs a new date-time exception with the specified message.
      *
