diff a/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java b/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
--- a/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
+++ b/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
@@ -202,11 +202,11 @@
             }
 
             var headHash = head == null ? repository.head() : head;
             var filesDesc = files.isEmpty() ? "" :
                             " for files " +
-                            String.join(", ", files.stream().map(Path::toString).collect(Collectors.toList()));
+                            files.stream().map(Path::toString).collect(Collectors.joining(", "));
             log.fine("Generating webrev from " + tailEnd + " to " + headHash + filesDesc);
 
             var fileViews = new ArrayList<FileView>();
             var formatter = new MetadataFormatter(issueLinker);
             for (var patch : patches) {
