diff a/modules/javafx.graphics/src/main/native-glass/mac/GlassWindow+Java.m b/modules/javafx.graphics/src/main/native-glass/mac/GlassWindow+Java.m
--- a/modules/javafx.graphics/src/main/native-glass/mac/GlassWindow+Java.m
+++ b/modules/javafx.graphics/src/main/native-glass/mac/GlassWindow+Java.m
@@ -93,13 +93,13 @@
 {
     if (self->suppressWindowResizeEvent == NO)
     {
         GET_MAIN_JENV;
 
-        if([self->nsWindow isMiniaturized]) {
+        if ([self->nsWindow isMiniaturized]) {
             type = com_sun_glass_events_WindowEvent_MINIMIZE;
-        } else if([self->nsWindow isZoomed]) {
+        } else if ([self->nsWindow isZoomed]) {
             type = com_sun_glass_events_WindowEvent_MAXIMIZE;
         }
 
         (*env)->CallVoidMethod(env, jWindow, jWindowNotifyResize, type,
              (int)frame.size.width, (int)frame.size.height);
