diff a/src/jdk.compiler/share/classes/com/sun/tools/sjavac/client/SjavacClient.java b/src/jdk.compiler/share/classes/com/sun/tools/sjavac/client/SjavacClient.java
--- a/src/jdk.compiler/share/classes/com/sun/tools/sjavac/client/SjavacClient.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/sjavac/client/SjavacClient.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2014, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -171,11 +171,11 @@
 
         return result;
     }
 
     /*
-     * Makes MAX_CONNECT_ATTEMPTS attepmts to connect to server.
+     * Makes MAX_CONNECT_ATTEMPTS attempts to connect to server.
      */
     private Socket tryConnect() throws IOException, InterruptedException {
         makeSureServerIsRunning(portFile);
         int attempt = 0;
         while (true) {
