diff a/modules/javafx.web/src/main/native/Source/WebCore/xml/NativeXPathNSResolver.cpp b/modules/javafx.web/src/main/native/Source/WebCore/xml/NativeXPathNSResolver.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/xml/NativeXPathNSResolver.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/xml/NativeXPathNSResolver.cpp
@@ -30,11 +30,11 @@
 #include "XMLNames.h"
 #include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
-NativeXPathNSResolver::NativeXPathNSResolver(RefPtr<Node>&& node)
+NativeXPathNSResolver::NativeXPathNSResolver(Ref<Node>&& node)
     : m_node(WTFMove(node))
 {
 }
 
 NativeXPathNSResolver::~NativeXPathNSResolver() = default;
@@ -44,9 +44,9 @@
     // This is not done by Node::lookupNamespaceURI as per the DOM3 Core spec,
     // but the XPath spec says that we should do it for XPathNSResolver.
     if (prefix == "xml")
         return XMLNames::xmlNamespaceURI.get();
 
-    return m_node ? m_node->lookupNamespaceURI(prefix).string() : String();
+    return m_node->lookupNamespaceURI(prefix).string();
 }
 
 } // namespace WebCore
