diff a/modules/javafx.web/src/main/java/javafx/scene/web/WebEngine.java b/modules/javafx.web/src/main/java/javafx/scene/web/WebEngine.java
--- a/modules/javafx.web/src/main/java/javafx/scene/web/WebEngine.java
+++ b/modules/javafx.web/src/main/java/javafx/scene/web/WebEngine.java
@@ -1610,11 +1610,11 @@
         float width = (float) pl.getPrintableWidth();
         float height = (float) pl.getPrintableHeight();
         int pageCount = page.beginPrinting(width, height);
 
         JobSettings jobSettings = job.getJobSettings();
-        if(jobSettings.getPageRanges() != null) {
+        if (jobSettings.getPageRanges() != null) {
             PageRange[] pageRanges = jobSettings.getPageRanges();
             for (PageRange p : pageRanges) {
                 for (int i = p.getStartPage(); i <= p.getEndPage() && i <= pageCount; ++i) {
                     if (printStatusOK(job)) {
                         Node printable = new Printable(page, i - 1, width);
