diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapValue.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapValue.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapValue.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapValue.h
@@ -28,15 +28,18 @@
 #if ENABLE(B3_JIT)
 
 #include "B3ConstrainedValue.h"
 #include "B3Value.h"
 #include "B3ValueRep.h"
-#include "CCallHelpers.h"
 #include "RegisterSet.h"
 #include <wtf/SharedTask.h>
 
-namespace JSC { namespace B3 {
+namespace JSC {
+
+class CCallHelpers;
+
+namespace B3 {
 
 class StackmapGenerationParams;
 
 typedef void StackmapGeneratorFunction(CCallHelpers&, const StackmapGenerationParams&);
 typedef SharedTask<StackmapGeneratorFunction> StackmapGenerator;
