diff a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabHost.java b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabHost.java
--- a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabHost.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabHost.java
@@ -112,11 +112,12 @@
 
     private HostUser parseUserDetails(JSONObject details) {
         var id = details.get("id").asInt();
         var username = details.get("username").asString();
         var name = details.get("name").asString();
-        return new HostUser(id, username, name);
+        var email = details.get("email").asString();
+        return new HostUser(id, username, name, email);
     }
 
     @Override
     public HostUser user(String username) {
         var details = request.get("users").param("username", username).execute().asArray();
