diff a/modules/javafx.web/src/main/native/Tools/DumpRenderTree/java/TestRunnerJava.cpp b/modules/javafx.web/src/main/native/Tools/DumpRenderTree/java/TestRunnerJava.cpp
--- a/modules/javafx.web/src/main/native/Tools/DumpRenderTree/java/TestRunnerJava.cpp
+++ b/modules/javafx.web/src/main/native/Tools/DumpRenderTree/java/TestRunnerJava.cpp
@@ -245,15 +245,10 @@
 void TestRunner::setTabKeyCyclesThroughElements(bool cycles)
 {
     // FIXME: implement
 }
 
-void TestRunner::setUseDashboardCompatibilityMode(bool flag)
-{
-    // FIXME: implement
-}
-
 void TestRunner::setUserStyleSheetEnabled(bool flag)
 {
     // FIXME: implement
 }
 
@@ -604,14 +599,9 @@
 void TestRunner::setSpellCheckerResults(JSContextRef, JSObjectRef)
 {
     fprintf(testResult, "ERROR: TestRunner::setSpellCheckerResults() not implemented\n");
 }
 
-void TestRunner::setIDBPerOriginQuota(uint64_t)
-{
-    fprintf(testResult, "ERROR: TestRunner::setIDBPerOriginQuota() not implemented\n");
-}
-
 void TestRunner::setOnlyAcceptFirstPartyCookies(bool)
 {
     fprintf(testResult, "ERROR: TestRunner::setOnlyAcceptFirstPartyCookies() not implemented\n");
 }
