diff a/host/src/main/java/org/openjdk/skara/host/github/GitHubHost.java b/host/src/main/java/org/openjdk/skara/host/github/GitHubHost.java
--- a/host/src/main/java/org/openjdk/skara/host/github/GitHubHost.java
+++ b/host/src/main/java/org/openjdk/skara/host/github/GitHubHost.java
@@ -122,13 +122,17 @@
         var details = getUserDetails(userName);
         return details.fullName();
     }
 
     // Most GitHub API's return user information in this format
-    HostUserDetails parseUserDetails(JSONValue json) {
-        return new HostUserDetails(json.get("user").get("id").asInt(), json.get("user").get("login").asString(),
-                                   () -> getFullName(json.get("user").get("login").asString()));
+    HostUserDetails parseUserField(JSONValue json) {
+        return parseUserObject(json.get("user"));
+    }
+
+    HostUserDetails parseUserObject(JSONValue json) {
+        return new HostUserDetails(json.get("id").asInt(), json.get("login").asString(),
+                                   () -> getFullName(json.get("login").asString()));
     }
 
     @Override
     public boolean isValid() {
         var endpoints = request.get("")
