diff a/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/stacktrace/StacktraceModelUtils.java b/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/stacktrace/StacktraceModelUtils.java
--- a/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/stacktrace/StacktraceModelUtils.java
+++ b/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/stacktrace/StacktraceModelUtils.java
@@ -2,13 +2,11 @@
 
 import org.openjdk.jmc.flightrecorder.stacktrace.StacktraceModel.Branch;
 import org.openjdk.jmc.flightrecorder.stacktrace.StacktraceModel.Fork;
 
 /**
- * 
- * StacktraceUtils useful shared functionality on {@link StacktraceModel} 
- *
+ * StacktraceUtils useful shared functionality on {@link StacktraceModel}
  */
 public final class StacktraceModelUtils {
 
 	// See JMC-6787
 	@SuppressWarnings("deprecation")
