diff a/src/jdk.incubator.jpackage/share/native/applauncher/JvmLauncher.cpp b/src/jdk.incubator.jpackage/share/native/applauncher/JvmLauncher.cpp
--- a/src/jdk.incubator.jpackage/share/native/applauncher/JvmLauncher.cpp
+++ b/src/jdk.incubator.jpackage/share/native/applauncher/JvmLauncher.cpp
@@ -167,11 +167,23 @@
         jboolean cpwildcard,
         jboolean javaw,
         jint ergo);
 
     std::vector<char*> argv;
+#ifdef TSTRINGS_WITH_WCHAR
+    std::vector<std::string> mbcs_args;
+    do {
+        tstring_array::const_iterator it = args.begin();
+        const tstring_array::const_iterator end = args.end();
+        for (; it != end; ++it) {
+            mbcs_args.push_back(tstrings::toACP(*it));
+        }
+    } while (0);
+    convertArgs(mbcs_args, argv);
+#else
     convertArgs(args, argv);
+#endif
 
     // Don't count terminal '0'.
     const int argc = (int)argv.size() - 1;
 
     LOG_TRACE(tstrings::any() << "JVM library: \"" << jvmPath << "\"");
