diff a/modules/javafx.web/src/main/native/Source/WebCore/fileapi/File.idl b/modules/javafx.web/src/main/native/Source/WebCore/fileapi/File.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/fileapi/File.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/fileapi/File.idl
@@ -26,10 +26,11 @@
 typedef (BufferSource or Blob or USVString) BlobPart;
 
 // FIXME: This should be exposed on Workers as well.
 [
     Constructor(sequence<BlobPart> fileBits, USVString fileName, optional FilePropertyBag options),
+    ConstructorCallWith=ScriptExecutionContext,
     ExportMacro=WEBCORE_EXPORT,
     Exposed=(Window),
     JSGenerateToJSObject,
     JSGenerateToNativeObject,
 ] interface File : Blob {
