diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveItem.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveItem.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveItem.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveItem.java
@@ -149,10 +149,32 @@
             }
         }
         return Optional.empty();
     }
 
+    static boolean containsQuote(String quote, String body) {
+        var compactQuote = quote.lines()
+                                .takeWhile(line -> line.startsWith(">"))
+                                .map(line -> line.replaceAll("\\W", ""))
+                                .collect(Collectors.joining());
+        if (!compactQuote.isBlank()) {
+            var compactBody = body.replaceAll("\\W", "");
+            return compactBody.contains(compactQuote);
+        } else {
+            return false;
+        }
+    }
+
+    private static Optional<ArchiveItem> findLastQuoted(String commentText, List<ArchiveItem> eligibleParents) {
+        for (int i = eligibleParents.size() - 1; i != 0; --i) {
+            if (containsQuote(commentText, eligibleParents.get(i).body())) {
+                return Optional.of(eligibleParents.get(i));
+            }
+        }
+        return Optional.empty();
+    }
+
     static ArchiveItem findParent(List<ArchiveItem> generated, Comment comment) {
         ArchiveItem lastCommentOrReview = generated.get(0);
         var eligible = new ArrayList<ArchiveItem>();
         eligible.add(lastCommentOrReview);
         for (var item : generated) {
@@ -166,10 +188,14 @@
 
         var lastMention = findLastMention(comment.body(), eligible);
         if (lastMention.isPresent()) {
             return lastMention.get();
         }
+        var lastQuoted = findLastQuoted(comment.body(), eligible);
+        if (lastQuoted.isPresent()) {
+            return lastQuoted.get();
+        }
 
         return lastCommentOrReview;
     }
 
     static ArchiveItem findRevisionItem(List<ArchiveItem> generated, Hash hash) {
