diff a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java
--- a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java
+++ b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java
@@ -358,12 +358,14 @@
         // write interfaces count
         dos.writeShort((short) len);
         for (int i = 0; i < len; i++) {
            Klass k = interfaces.getAt(i);
            Short index = (Short) classToIndex.get(k.getName().asString());
-           dos.writeShort(index.shortValue());
-           if (DEBUG) debugMessage("\t" + index);
+           if (index != null) { // Injected interfaces have no constant pool entry
+               dos.writeShort(index.shortValue());
+               if (DEBUG) debugMessage("\t" + index);
+           }
         }
     }
 
     protected void writeFields() throws IOException {
         final int javaFieldsCount = klass.getJavaFieldsCount();
