diff a/src/java.base/share/classes/java/security/DomainLoadStoreParameter.java b/src/java.base/share/classes/java/security/DomainLoadStoreParameter.java
--- a/src/java.base/share/classes/java/security/DomainLoadStoreParameter.java
+++ b/src/java.base/share/classes/java/security/DomainLoadStoreParameter.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -123,11 +123,11 @@
      *     A {@code java.util.Collections.EMPTY_MAP} should be used
      *     when protection parameters are not required or when they have
      *     been specified by properties in the domain configuration data.
      *     It is cloned to prevent subsequent modification.
      *
-     * @exception NullPointerException if {@code configuration} or
+     * @throws    NullPointerException if {@code configuration} or
      *     {@code protectionParams} is {@code null}
      */
     public DomainLoadStoreParameter(URI configuration,
         Map<String,ProtectionParameter> protectionParams) {
         if (configuration == null || protectionParams == null) {
