diff a/src/java.base/share/classes/java/security/URIParameter.java b/src/java.base/share/classes/java/security/URIParameter.java
--- a/src/java.base/share/classes/java/security/URIParameter.java
+++ b/src/java.base/share/classes/java/security/URIParameter.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2005, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -41,11 +41,11 @@
      * Constructs a URIParameter with the URI pointing to
      * data intended for an SPI implementation.
      *
      * @param uri the URI pointing to the data.
      *
-     * @exception NullPointerException if the specified URI is null.
+     * @throws    NullPointerException if the specified URI is null.
      */
     public URIParameter(java.net.URI uri) {
         if (uri == null) {
             throw new NullPointerException("invalid null URI");
         }
