diff a/src/hotspot/share/classfile/classLoaderStats.hpp b/src/hotspot/share/classfile/classLoaderStats.hpp
--- a/src/hotspot/share/classfile/classLoaderStats.hpp
+++ b/src/hotspot/share/classfile/classLoaderStats.hpp
@@ -75,14 +75,10 @@
 
   size_t            _chunk_sz;
   size_t            _block_sz;
   uintx             _classes_count;
 
-  size_t            _anon_chunk_sz;
-  size_t            _anon_block_sz;
-  uintx             _anon_classes_count;
-
   size_t            _hidden_chunk_sz;
   size_t            _hidden_block_sz;
   uintx             _hidden_classes_count;
 
   ClassLoaderStats() :
@@ -90,13 +86,10 @@
     _class_loader(0),
     _parent(0),
     _chunk_sz(0),
     _block_sz(0),
     _classes_count(0),
-    _anon_chunk_sz(0),
-    _anon_block_sz(0),
-    _anon_classes_count(0),
     _hidden_chunk_sz(0),
     _hidden_block_sz(0),
     _hidden_classes_count(0) {
   }
 };
