diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/GeneratedImage.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/GeneratedImage.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/GeneratedImage.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/GeneratedImage.h
@@ -38,11 +38,11 @@
     bool usesContainerSize() const override { return true; }
     bool hasRelativeWidth() const override { return true; }
     bool hasRelativeHeight() const override { return true; }
     void computeIntrinsicDimensions(Length& intrinsicWidth, Length& intrinsicHeight, FloatSize& intrinsicRatio) override;
 
-    FloatSize size() const override { return m_size; }
+    FloatSize size(ImageOrientation = ImageOrientation::FromImage) const override { return m_size; }
 
     // Assume that generated content has no decoded data we need to worry about
     void destroyDecodedData(bool /*destroyAll*/ = true) override { }
 
 protected:
