diff a/test/hotspot/gtest/oops/test_oop.cpp b/test/hotspot/gtest/oops/test_oop.cpp
--- a/test/hotspot/gtest/oops/test_oop.cpp
+++ b/test/hotspot/gtest/oops/test_oop.cpp
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -24,11 +24,11 @@
 #include "precompiled.hpp"
 #include "oops/oop.inline.hpp"
 #include "unittest.hpp"
 #include "utilities/globalDefinitions.hpp"
 
-static char memory[32];
+static unsigned char memory[32];
 
 oop fake_object() {
   return cast_to_oop(memory);
 }
 
