diff a/src/com/sun/javatest/BasicParameters.java b/src/com/sun/javatest/BasicParameters.java
--- a/src/com/sun/javatest/BasicParameters.java
+++ b/src/com/sun/javatest/BasicParameters.java
@@ -837,14 +837,11 @@
         TestFilter statusFilter = getPriorStatusFilter();
         if (statusFilter != null) {
             v.add(statusFilter);
         }
 
-        TestFilter testSuiteFilter = getRelevantTestFilter();
-        if (testSuiteFilter != null) {
-            v.add(testSuiteFilter);
-        }
+        v.addAll(getAllRelevantFiltersInTheSuite());
 
         if (v.isEmpty()) {
             return null;
         } else if (equal(v, cachedTestFilters)) {
             return cachedTestFilters;
