diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h
@@ -27,13 +27,10 @@
 
 namespace JSC {
 
 namespace Wasm {
 
-void registerCode(void* start, void* end);
-void unregisterCode(void* start, void* end);
-
 bool fastMemoryEnabled();
 JS_EXPORT_PRIVATE void enableFastMemory();
 
 } } // namespace JSC::Wasm
 
