diff a/host/build.gradle b/host/build.gradle
--- a/host/build.gradle
+++ b/host/build.gradle
@@ -25,10 +25,11 @@
     name = 'org.openjdk.skara.host'
     test {
         requires 'org.openjdk.skara.test'
         requires 'org.junit.jupiter.api'
         requires 'jdk.httpserver'
+        opens 'org.openjdk.skara.host' to 'org.junit.platform.commons'
         opens 'org.openjdk.skara.host.network' to 'org.junit.platform.commons'
         opens 'org.openjdk.skara.host.github' to 'org.junit.platform.commons'
     }
 }
 
