diff a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
--- a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
@@ -21,10 +21,11 @@
  * questions.
  */
 package org.openjdk.skara.test;
 
 import org.openjdk.skara.forge.*;
+import org.openjdk.skara.issuetracker.Issue;
 import org.openjdk.skara.json.JSONValue;
 import org.openjdk.skara.vcs.*;
 
 import java.io.*;
 import java.net.*;
@@ -68,11 +69,13 @@
         return host.getPullRequest(this, id);
     }
 
     @Override
     public List<PullRequest> pullRequests() {
-        return new ArrayList<>(host.getPullRequests(this));
+        return host.getPullRequests(this).stream()
+                   .filter(pr -> pr.state().equals(Issue.State.OPEN))
+                   .collect(Collectors.toList());
     }
 
     @Override
     public List<PullRequest> pullRequests(ZonedDateTime updatedAfter) {
         return host.getPullRequests(this).stream()
