diff a/src/hotspot/cpu/zero/vm_version_ext_zero.hpp b/src/hotspot/cpu/zero/vm_version_ext_zero.hpp
--- a/src/hotspot/cpu/zero/vm_version_ext_zero.hpp
+++ b/src/hotspot/cpu/zero/vm_version_ext_zero.hpp
@@ -23,12 +23,12 @@
  */
 
 #ifndef CPU_ZERO_VM_VERSION_EXT_ZERO_HPP
 #define CPU_ZERO_VM_VERSION_EXT_ZERO_HPP
 
+#include "runtime/vm_version.hpp"
 #include "utilities/macros.hpp"
-#include "vm_version_zero.hpp"
 
 class VM_Version_Ext : public VM_Version {
  private:
   static const size_t      CPU_TYPE_DESC_BUF_SIZE = 256;
   static const size_t      CPU_DETAILED_DESC_BUF_SIZE = 4096;
