diff a/test/jdk/java/foreign/callarranger/CallArrangerTestBase.java b/test/jdk/java/foreign/callarranger/CallArrangerTestBase.java
--- a/test/jdk/java/foreign/callarranger/CallArrangerTestBase.java
+++ b/test/jdk/java/foreign/callarranger/CallArrangerTestBase.java
@@ -45,16 +45,6 @@
     }
 
     public static void checkReturnBindings(CallingSequence callingSequence, Binding[] returnBindings) {
         assertEquals(callingSequence.returnBindings(), Arrays.asList(returnBindings));
     }
-
-    public static FunctionDescriptor descAddArgument(FunctionDescriptor desc, MemoryLayout... layouts) {
-        var args = new ArrayList<>(desc.argumentLayouts());
-        args.addAll(Arrays.asList(layouts));
-        var argsArray = args.toArray(MemoryLayout[]::new);
-        return desc.returnLayout().isEmpty()
-            ? FunctionDescriptor.ofVoid(false, argsArray)
-            : FunctionDescriptor.of(desc.returnLayout().get(), false, argsArray);
-    }
-
 }
