diff a/test/langtools/tools/javac/valhalla/lworld-values/CheckExtends.java b/test/langtools/tools/javac/valhalla/lworld-values/CheckExtends.java
--- a/test/langtools/tools/javac/valhalla/lworld-values/CheckExtends.java
+++ b/test/langtools/tools/javac/valhalla/lworld-values/CheckExtends.java
@@ -1,9 +1,11 @@
 /*
  * @test /nodynamiccopyright/
- * @summary Values may not extend a concrete type other than jlO
+ * @summary Values may not extend an identity class
  *
  * @compile/fail/ref=CheckExtends.out -XDrawDiagnostics CheckExtends.java
  */
 
 final inline class CheckExtends extends Object {
+    static class Nested {}
+    static inline class NestedValue extends Nested {}
 }
