diff a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java
--- a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java
+++ b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java
@@ -224,12 +224,12 @@
             var masterHash = localRepo.resolve("master").orElseThrow();
             assertFalse(CheckableRepository.hasBeenEdited(localRepo));
             localRepo.push(masterHash, author.getUrl(), "master", true);
 
             // Make a change in another branch that will not pass jcheck
-            var otherHash = CheckableRepository.appendAndCommit(localRepo, "Change in other",
-                                                                "Unreviewed change in other");
+            var otherHash = CheckableRepository.appendAndCommit(localRepo, "Change in other with trails..   ",
+                                                                "Bad change in other");
             localRepo.push(otherHash, author.getUrl(), "other", true);
 
             // Go back to the original master
             localRepo.checkout(masterHash, true);
 
