diff a/modules/javafx.graphics/src/main/native-glass/gtk_new/glass_window.cpp b/modules/javafx.graphics/src/main/native-glass/gtk_new/glass_window.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk_new/glass_window.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk_new/glass_window.cpp
@@ -591,20 +591,14 @@
         CHECK_JNI_EXCEPTION(mainEnv)
     }
 }
 
 void WindowContext::process_mouse_scroll(GdkEventScroll *event) {
-    jdouble dx = 0;
-    jdouble dy = 0;
+    jdouble dx = 0, dy = 0;
 
     // converting direction to change in pixels
     switch (event->direction) {
-#if GTK_CHECK_VERSION(3, 4, 0)
-        case GDK_SCROLL_SMOOTH:
-            //FIXME 3.4 ???
-            break;
-#endif
         case GDK_SCROLL_UP:
             dy = 1;
             break;
         case GDK_SCROLL_DOWN:
             dy = -1;
@@ -613,10 +607,12 @@
             dx = 1;
             break;
         case GDK_SCROLL_RIGHT:
             dx = -1;
             break;
+        default:
+            break;
     }
     if (event->state & GDK_SHIFT_MASK) {
         jdouble t = dy;
         dy = dx;
         dx = t;
