diff a/test/hotspot/jtreg/runtime/valhalla/valuetypes/identityObject/AbstractTypeImplementingIdentityObject.java b/test/hotspot/jtreg/runtime/valhalla/valuetypes/identityObject/AbstractTypeImplementingIdentityObject.java
--- /dev/null
+++ b/test/hotspot/jtreg/runtime/valhalla/valuetypes/identityObject/AbstractTypeImplementingIdentityObject.java
@@ -0,0 +1,3 @@
+public abstract class AbstractTypeImplementingIdentityObject implements IdentityObject {
+
+}
