diff a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/viewers/MemoryStatisticsTableViewer.java b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/viewers/MemoryStatisticsTableViewer.java
--- a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/viewers/MemoryStatisticsTableViewer.java
+++ b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/viewers/MemoryStatisticsTableViewer.java
@@ -216,14 +216,14 @@
 		if (mPieChartViewer != null) {
 			mPieChartViewer.setHighlightedItem(item);
 		}
 	}
 
-	private TableViewerColumn createTableColumnViewer(String label,
-			Function<MemoryStatisticsItem, String> labelProvider,
-			Function<MemoryStatisticsItem, String> toolTipProvider, Function<MemoryStatisticsItem, Color> colorProvider,
-			Comparator<MemoryStatisticsItem> comparator) {
+	private TableViewerColumn createTableColumnViewer(
+		String label, Function<MemoryStatisticsItem, String> labelProvider,
+		Function<MemoryStatisticsItem, String> toolTipProvider, Function<MemoryStatisticsItem, Color> colorProvider,
+		Comparator<MemoryStatisticsItem> comparator) {
 		TableViewerColumn column = new TableViewerColumn(this, SWT.NONE);
 		column.getColumn().setWidth(200);
 		column.getColumn().setText(label);
 		column.getColumn().setMoveable(true);
 
