diff a/src/jdk.incubator.jextract/share/classes/module-info.java b/src/jdk.incubator.jextract/share/classes/module-info.java
--- a/src/jdk.incubator.jextract/share/classes/module-info.java
+++ b/src/jdk.incubator.jextract/share/classes/module-info.java
@@ -1,9 +1,11 @@
+import jdk.incubator.jextract.JextractTool;
+
 module jdk.incubator.jextract {
     requires transitive java.compiler;
     requires transitive jdk.incubator.foreign;
     requires jdk.internal.opt;
     exports jdk.incubator.jextract;
 
     provides java.util.spi.ToolProvider with
-        jdk.incubator.jextract.tool.Main.JextractToolProvider;
+        JextractTool.JextractToolProvider;
 }
