diff a/modules/javafx.web/src/main/native/Source/WebCore/html/DateTimeInputType.h b/modules/javafx.web/src/main/native/Source/WebCore/html/DateTimeInputType.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/DateTimeInputType.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/DateTimeInputType.h
@@ -37,14 +37,14 @@
 
 namespace WebCore {
 
 class DateTimeInputType final : public BaseChooserOnlyDateAndTimeInputType {
 public:
-    explicit DateTimeInputType(HTMLInputElement& element) : BaseChooserOnlyDateAndTimeInputType(element) { }
+    explicit DateTimeInputType(HTMLInputElement& element) : BaseDateTimeInputType(element) { }
 
 private:
-    const AtomicString& formControlType() const override;
+    const AtomString& formControlType() const override;
     DateComponents::Type dateType() const override;
     StepRange createStepRange(AnyStepHandling) const override;
     Decimal defaultValueForStepUp() const override;
     bool parseToDateComponentsInternal(const UChar*, unsigned length, DateComponents*) const override;
     bool setMillisecondToDateComponents(double, DateComponents*) const override;
