diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLObjectElement.idl b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLObjectElement.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLObjectElement.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLObjectElement.idl
@@ -44,8 +44,9 @@
     boolean checkValidity();
     [EnabledAtRuntime=InteractiveFormValidation] boolean reportValidity();
     void setCustomValidity(DOMString error);
 
     [CheckSecurityForNode] readonly attribute Document contentDocument;
+    readonly attribute WindowProxy contentWindow;
 
     [CheckSecurityForNode, MayThrowException] Document getSVGDocument();
 };
