diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLOListElement.h b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLOListElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLOListElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLOListElement.h
@@ -46,13 +46,13 @@
 private:
     HTMLOListElement(const QualifiedName&, Document&);
 
     WEBCORE_EXPORT unsigned itemCount() const;
 
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
     bool isPresentationAttribute(const QualifiedName&) const final;
-    void collectStyleForPresentationAttribute(const QualifiedName&, const AtomicString&, MutableStyleProperties&) final;
+    void collectStyleForPresentationAttribute(const QualifiedName&, const AtomString&, MutableStyleProperties&) final;
 
     Optional<int> m_start;
     mutable Optional<unsigned> m_itemCount;
     bool m_isReversed { false };
 };
