diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLStyleElement.h b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLStyleElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLStyleElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLStyleElement.h
@@ -52,11 +52,11 @@
     void finishParsingChildren() final;
 
 private:
     HTMLStyleElement(const QualifiedName&, Document&, bool createdByParser);
 
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
     InsertedIntoAncestorResult insertedIntoAncestor(InsertionType, ContainerNode&) final;
     void removedFromAncestor(RemovalType, ContainerNode&) final;
     void childrenChanged(const ChildChange&) final;
 
     bool isLoading() const { return m_styleSheetOwner.isLoading(); }
