diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLTableRowElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLTableRowElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLTableRowElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLTableRowElement.cpp
@@ -125,11 +125,11 @@
         result = appendChild(cell);
     else
         result = insertBefore(cell, index < 1 ? firstChild() : children->item(index));
     if (result.hasException())
         return result.releaseException();
-    return WTFMove(cell);
+    return cell;
 }
 
 ExceptionOr<void> HTMLTableRowElement::deleteCell(int index)
 {
     auto children = cells();
