diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMapElement.h b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMapElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMapElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMapElement.h
@@ -34,24 +34,24 @@
 public:
     static Ref<HTMLMapElement> create(Document&);
     static Ref<HTMLMapElement> create(const QualifiedName&, Document&);
     virtual ~HTMLMapElement();
 
-    const AtomicString& getName() const { return m_name; }
+    const AtomString& getName() const { return m_name; }
 
     bool mapMouseEvent(LayoutPoint location, const LayoutSize&, HitTestResult&);
 
     HTMLImageElement* imageElement();
     WEBCORE_EXPORT Ref<HTMLCollection> areas();
 
 private:
     HTMLMapElement(const QualifiedName&, Document&);
 
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
 
     InsertedIntoAncestorResult insertedIntoAncestor(InsertionType, ContainerNode&) final;
     void removedFromAncestor(RemovalType, ContainerNode&) final;
 
-    AtomicString m_name;
+    AtomString m_name;
 };
 
 } // namespaces
