diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDetailsElement.h b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDetailsElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDetailsElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDetailsElement.h
@@ -45,14 +45,15 @@
 
 private:
     HTMLDetailsElement(const QualifiedName&, Document&);
 
     RenderPtr<RenderElement> createElementRenderer(RenderStyle&&, const RenderTreePosition&) final;
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
 
     void didAddUserAgentShadowRoot(ShadowRoot&) final;
     bool hasCustomFocusLogic() const final { return true; }
+    bool isInteractiveContent() const final { return true; }
 
     bool m_isOpen { false };
     HTMLSlotElement* m_summarySlot { nullptr };
     HTMLSummaryElement* m_defaultSummary { nullptr };
     RefPtr<HTMLSlotElement> m_defaultSlot;
