diff a/modules/javafx.web/src/main/native/Source/WebCore/html/TextFieldInputType.h b/modules/javafx.web/src/main/native/Source/WebCore/html/TextFieldInputType.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/TextFieldInputType.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/TextFieldInputType.h
@@ -51,11 +51,11 @@
 #endif
 {
 protected:
     explicit TextFieldInputType(HTMLInputElement&);
     virtual ~TextFieldInputType();
-    void handleKeydownEvent(KeyboardEvent&) override;
+    ShouldCallBaseEventHandler handleKeydownEvent(KeyboardEvent&) override;
     void handleKeydownEventForSpinButton(KeyboardEvent&);
 #if ENABLE(DATALIST_ELEMENT)
     void handleClickEvent(MouseEvent&) final;
 #endif
 
