diff a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSGridAutoRepeatValue.cpp b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSGridAutoRepeatValue.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSGridAutoRepeatValue.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSGridAutoRepeatValue.cpp
@@ -44,6 +44,11 @@
     result.append(CSSValueList::customCSSText());
     result.append(')');
     return result.toString();
 }
 
+bool CSSGridAutoRepeatValue::equals(const CSSGridAutoRepeatValue& other) const
+{
+    return m_autoRepeatID == other.m_autoRepeatID && CSSValueList::equals(other);
+}
+
 } // namespace WebCore
