diff a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValueKeywords.in b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValueKeywords.in
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValueKeywords.in
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValueKeywords.in
@@ -220,44 +220,59 @@
 threedlightshadow
 threedshadow
 window
 windowframe
 windowtext
-#if defined(WTF_PLATFORM_MAC) && WTF_PLATFORM_MAC
+#if (defined(WTF_PLATFORM_MAC) && WTF_PLATFORM_MAC) || (defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY)
 -apple-system-header-text
 -apple-system-text-background
 -apple-system-control-background
--apple-system-alternate-selected-text
--apple-system-control-accent
--apple-system-even-alternating-content-background
--apple-system-odd-alternating-content-background
--apple-system-selected-content-background
--apple-system-unemphasized-selected-content-background
--apple-system-selected-text
--apple-system-unemphasized-selected-text
--apple-system-selected-text-background
--apple-system-unemphasized-selected-text-background
 -apple-system-placeholder-text
--apple-system-find-highlight-background
 -apple-system-label
 -apple-system-secondary-label
 -apple-system-tertiary-label
 -apple-system-quaternary-label
 -apple-system-grid
 -apple-system-separator
 -apple-system-container-border
-#endif
+-apple-system-background
+-apple-system-secondary-background
+-apple-system-tertiary-background
+-apple-system-grouped-background
+-apple-system-secondary-grouped-background
+-apple-system-tertiary-grouped-background
+-apple-system-selected-content-background
+-apple-system-unemphasized-selected-content-background
 -apple-wireless-playback-target-active
 -apple-system-blue
 -apple-system-brown
 -apple-system-gray
 -apple-system-green
 -apple-system-orange
 -apple-system-pink
 -apple-system-purple
 -apple-system-red
 -apple-system-yellow
+#endif
+#if defined(WTF_PLATFORM_MAC) && WTF_PLATFORM_MAC
+-apple-system-alternate-selected-text
+-apple-system-control-accent
+-apple-system-even-alternating-content-background
+-apple-system-odd-alternating-content-background
+-apple-system-selected-text
+-apple-system-unemphasized-selected-text
+-apple-system-selected-text-background
+-apple-system-unemphasized-selected-text-background
+-apple-system-find-highlight-background
+#endif
+#if defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY
+-apple-system-indigo
+-apple-system-teal
+#endif
+#if defined(HAVE_OS_DARK_MODE_SUPPORT) && HAVE_OS_DARK_MODE_SUPPORT
+-webkit-control-background
+#endif
 -webkit-focus-ring-color
 currentcolor
 //
 // colors in non strict mode
 grey
@@ -437,10 +452,11 @@
 inline-flex
 -webkit-inline-flex
 contents
 grid
 inline-grid
+flow-root
 //none
 //
 // CSS_PROP_CURSOR:
 // The order here must match the order of the CursorType enum in RenderStyleConstants.h.
 //
@@ -754,10 +770,15 @@
 //
 // CSS_PROP_WORD_WRAP
 //
 break-word
 
+//
+// CSS_PROP_WHITE_SPACE
+//
+break-spaces
+
 //
 // CSS_PROP__KHTML_NBSP_MODE
 //
 space
 
@@ -767,10 +788,11 @@
 // auto
 loose
 // normal
 strict
 after-white-space
+anywhere
 
 // -webkit-appearance
 // The order here must match the order in the ControlPart enum in ThemeTypes.h.
 // All appearance values that should be accepted by the parser should be listed between 'checkbox' and 'textarea':
 checkbox
@@ -1030,11 +1052,11 @@
 
 #if defined(ENABLE_LETTERPRESS) && ENABLE_LETTERPRESS
 -webkit-letterpress
 #endif
 
-#if defined(ENABLE_ACCELERATED_OVERFLOW_SCROLLING) && ENABLE_ACCELERATED_OVERFLOW_SCROLLING
+#if defined(ENABLE_OVERFLOW_SCROLLING_TOUCH) && ENABLE_OVERFLOW_SCROLLING_TOUCH
 // -webkit-overflow-scrolling
 // auto
 touch
 #endif
 
@@ -1134,10 +1156,11 @@
 edges
 
 alphabetic
 
 // position
+sticky
 -webkit-sticky
 
 // (pointer:) media feature
 // none
 coarse
@@ -1198,11 +1221,10 @@
 -webkit-repeating-radial-gradient
 -webkit-image-set
 -webkit-named-image
 filter
 -webkit-filter
-dashboard-region
 #if defined(ENABLE_CSS_PAINTING_API) && ENABLE_CSS_PAINTING_API
 paint
 #endif
 
 // deprecated gradients
@@ -1388,11 +1410,11 @@
 // prefers-color-scheme
 light
 dark
 // no-preference
 
-// supported-color-schemes
+// color-scheme
 only
 // light
 // dark
 #endif
 
