diff a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValuePool.cpp b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValuePool.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValuePool.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSValuePool.cpp
@@ -129,11 +129,11 @@
     return *m_fontFamilyValueCache.ensure({ familyName, isFromSystemID }, [&familyName, isFromSystemID] {
         return CSSPrimitiveValue::create(CSSFontFamily { familyName, isFromSystemID });
     }).iterator->value;
 }
 
-RefPtr<CSSValueList> CSSValuePool::createFontFaceValue(const AtomicString& string)
+RefPtr<CSSValueList> CSSValuePool::createFontFaceValue(const AtomString& string)
 {
     // Remove one entry at random if the cache grows too large.
     // FIXME: Use TinyLRUCache instead?
     const int maximumFontFaceCacheSize = 128;
     if (m_fontFaceValueCache.size() >= maximumFontFaceCacheSize)
