diff a/modules/javafx.web/src/main/native/Source/WebCore/css/FontFace.h b/modules/javafx.web/src/main/native/Source/WebCore/css/FontFace.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/FontFace.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/FontFace.h
@@ -53,11 +53,11 @@
     using Source = Variant<String, RefPtr<JSC::ArrayBuffer>, RefPtr<JSC::ArrayBufferView>>;
     static ExceptionOr<Ref<FontFace>> create(Document&, const String& family, Source&&, const Descriptors&);
     static Ref<FontFace> create(CSSFontFace&);
     virtual ~FontFace();
 
-    ExceptionOr<void> setFamily(const String&);
+    ExceptionOr<void> setFamily(Document&, const String&);
     ExceptionOr<void> setStyle(const String&);
     ExceptionOr<void> setWeight(const String&);
     ExceptionOr<void> setStretch(const String&);
     ExceptionOr<void> setUnicodeRange(const String&);
     ExceptionOr<void> setVariant(const String&);
