diff a/modules/javafx.web/src/main/native/Source/WebCore/css/PageRuleCollector.cpp b/modules/javafx.web/src/main/native/Source/WebCore/css/PageRuleCollector.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/PageRuleCollector.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/PageRuleCollector.cpp
@@ -92,11 +92,11 @@
 
 static bool checkPageSelectorComponents(const CSSSelector* selector, bool isLeftPage, bool isFirstPage, const String& pageName)
 {
     for (const CSSSelector* component = selector; component; component = component->tagHistory()) {
         if (component->match() == CSSSelector::Tag) {
-            const AtomicString& localName = component->tagQName().localName();
+            const AtomString& localName = component->tagQName().localName();
             if (localName != starAtom() && localName != pageName)
                 return false;
         } else if (component->match() == CSSSelector::PagePseudoClass) {
             CSSSelector::PagePseudoClassType pseudoType = component->pagePseudoClassType();
             if ((pseudoType == CSSSelector::PagePseudoClassLeft && !isLeftPage)
