diff a/src/hotspot/share/prims/whitebox.cpp b/src/hotspot/share/prims/whitebox.cpp
--- a/src/hotspot/share/prims/whitebox.cpp
+++ b/src/hotspot/share/prims/whitebox.cpp
@@ -1911,11 +1911,11 @@
     return (jobjectArray)JNIHandles::make_local(env, create_results(THREAD));
   }
 
   void add_oop(oop o) {
     // Value might be oop, but JLS can't see as Object, just iterate through it...
-    if (o != NULL && o->is_value()) {
+    if (o != NULL && o->is_inline_type()) {
       o->oop_iterate(this);
     } else {
       array->append(Handle(Thread::current(), o));
     }
   }
