diff a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontFaceSource.h b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontFaceSource.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontFaceSource.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontFaceSource.h
@@ -68,12 +68,13 @@
     const AtomString& familyNameOrURI() const { return m_familyNameOrURI; }
 
     void opportunisticallyStartFontDataURLLoading(CSSFontSelector&);
 
     void load(CSSFontSelector*);
-    RefPtr<Font> font(const FontDescription&, bool syntheticBold, bool syntheticItalic, const FontFeatureSettings&, const FontVariantSettings&, FontSelectionSpecifiedCapabilities);
+    RefPtr<Font> font(const FontDescription&, bool syntheticBold, bool syntheticItalic, const FontFeatureSettings&, FontSelectionSpecifiedCapabilities);
 
+    CachedFont* cachedFont() const { return m_font.get(); }
     bool requiresExternalResource() const { return m_font; }
 
 #if ENABLE(SVG_FONTS)
     bool isSVGFontFaceSource() const;
 #endif
