diff a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontStyleValue.h b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontStyleValue.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontStyleValue.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontStyleValue.h
@@ -50,11 +50,11 @@
     {
         if (!obliqueValue) {
             auto valueID = fontStyleValue->valueID();
             return valueID == CSSValueItalic || valueID == CSSValueOblique;
         }
-        return obliqueValue->value<float>(CSSPrimitiveValue::CSS_DEG) >= static_cast<float>(italicValue());
+        return obliqueValue->value<float>(CSSUnitType::CSS_DEG) >= static_cast<float>(italicValue());
     }
 
     Ref<CSSPrimitiveValue> fontStyleValue;
     RefPtr<CSSPrimitiveValue> obliqueValue;
 
