diff a/modules/javafx.web/src/main/native/Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp b/modules/javafx.web/src/main/native/Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp
@@ -250,11 +250,11 @@
     if (UNLIKELY(propertyID == CSSPropertyCustom)) {
         Document* document = nullptr;
 
         if (parentElement())
             document = &parentElement()->document();
-        else
+        else if (parentStyleSheet())
             document = parentStyleSheet()->ownerDocument();
 
         changed = m_propertySet->setCustomProperty(document, propertyName, value, important, cssParserContext());
     } else
         changed = m_propertySet->setProperty(propertyID, value, important, cssParserContext());
@@ -410,11 +410,15 @@
 {
     auto* styleSheet = contextStyleSheet();
     if (!styleSheet)
         return PropertySetCSSStyleDeclaration::cssParserContext();
 
-    return styleSheet->parserContext();
+    auto context = styleSheet->parserContext();
+    if (m_parentRule)
+        context.enclosingRuleType = static_cast<StyleRuleType>(m_parentRule->type());
+
+    return context;
 }
 
 void StyleRuleCSSStyleDeclaration::reattach(MutableStyleProperties& propertySet)
 {
     m_propertySet->deref();
