diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevStorage.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevStorage.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevStorage.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevStorage.java
@@ -148,25 +148,32 @@
     private void awaitPublication(URI uri, Duration timeout) throws IOException {
         var end = Instant.now().plus(timeout);
         var uriBuilder = URIBuilder.base(uri);
         var client = HttpClient.newBuilder()
                                .connectTimeout(Duration.ofSeconds(30))
-                               .followRedirects(HttpClient.Redirect.NORMAL)
                                .build();
         while (Instant.now().isBefore(end)) {
             var uncachedUri = uriBuilder.setQuery(Map.of("nocache", UUID.randomUUID().toString())).build();
+            log.fine("Validating webrev URL: " + uncachedUri);
             var request = HttpRequest.newBuilder(uncachedUri)
                                      .timeout(Duration.ofSeconds(30))
                                      .GET()
                                      .build();
             try {
                 var response = client.send(request, HttpResponse.BodyHandlers.ofString());
-                if (response.statusCode() < 400) {
+                if (response.statusCode() < 300) {
                     log.info(response.statusCode() + " when checking " + uncachedUri + " - success!");
-                    // Success!
                     return;
                 }
+                if (response.statusCode() < 400) {
+                    var newLocation = response.headers().firstValue("location");
+                    if (newLocation.isPresent()) {
+                        log.info("Webrev url redirection: " + newLocation.get());
+                        uriBuilder = URIBuilder.base(newLocation.get());
+                        continue;
+                    }
+                }
                 log.info(response.statusCode() + " when checking " + uncachedUri + " - waiting...");
                 Thread.sleep(Duration.ofSeconds(10).toMillis());
             } catch (InterruptedException ignored) {
             }
         }
diff a/test/src/main/java/org/openjdk/skara/test/TestWebrevServer.java b/test/src/main/java/org/openjdk/skara/test/TestWebrevServer.java
--- a/test/src/main/java/org/openjdk/skara/test/TestWebrevServer.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestWebrevServer.java
@@ -39,11 +39,11 @@
         public void handle(HttpExchange exchange) throws IOException {
             checked = true;
             var response = "ok!";
             var responseBytes = response.getBytes(StandardCharsets.UTF_8);
             if (!exchange.getRequestURI().toString().contains("final=true")) {
-                exchange.getResponseHeaders().add("Location", exchange.getRequestURI().toString() + "&final=true");
+                exchange.getResponseHeaders().add("Location", uri() + "&final=true");
                 exchange.sendResponseHeaders(302, responseBytes.length);
             } else {
                 redirectFollowed = true;
                 exchange.sendResponseHeaders(200, responseBytes.length);
             }
