diff a/src/hotspot/share/memory/heapShared.inline.hpp b/src/hotspot/share/memory/heapShared.inline.hpp
--- a/src/hotspot/share/memory/heapShared.inline.hpp
+++ b/src/hotspot/share/memory/heapShared.inline.hpp
@@ -25,10 +25,11 @@
 #ifndef SHARE_MEMORY_HEAPSHARED_INLINE_HPP
 #define SHARE_MEMORY_HEAPSHARED_INLINE_HPP
 
 #include "oops/compressedOops.inline.hpp"
 #include "memory/heapShared.hpp"
+#include "utilities/align.hpp"
 #if INCLUDE_G1GC
 #include "gc/g1/g1Allocator.inline.hpp"
 #endif
 
 #if INCLUDE_CDS_JAVA_HEAP
@@ -38,11 +39,11 @@
 }
 
 inline oop HeapShared::decode_from_archive(narrowOop v) {
   assert(!CompressedOops::is_null(v), "narrow oop value can never be zero");
   oop result = (oop)(void*)((uintptr_t)_narrow_oop_base + ((uintptr_t)v << _narrow_oop_shift));
-  assert(check_obj_alignment(result), "address not aligned: " INTPTR_FORMAT, p2i((void*) result));
+  assert(is_object_aligned(result), "address not aligned: " INTPTR_FORMAT, p2i((void*) result));
   return result;
 }
 
 #endif
 
