diff a/src/java.desktop/share/native/libjsound/SoundDefs.h b/src/java.desktop/share/native/libjsound/SoundDefs.h
--- a/src/java.desktop/share/native/libjsound/SoundDefs.h
+++ b/src/java.desktop/share/native/libjsound/SoundDefs.h
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2007, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2007, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -27,14 +27,13 @@
 #define __SOUNDDEFS_INCLUDED__
 
 
 // types for X_PLATFORM
 #define X_WINDOWS       1
-#define X_SOLARIS       2
-#define X_LINUX         3
-#define X_BSD           4
-#define X_MACOSX        5
+#define X_LINUX         2
+#define X_BSD           3
+#define X_MACOSX        4
 
 // **********************************
 // Make sure you set X_PLATFORM defines correctly.
 // Everything depends upon this flag being setup correctly.
 // **********************************
@@ -43,11 +42,11 @@
 #error "You need to define X_PLATFORM outside of the source. Use the types above."
 #endif
 
 
 // following is needed for _LP64
-#if ((X_PLATFORM == X_SOLARIS) || (X_PLATFORM == X_LINUX) || (X_PLATFORM == X_MACOSX))
+#if ((X_PLATFORM == X_LINUX) || (X_PLATFORM == X_MACOSX))
 #include <sys/types.h>
 #endif
 
 #if X_PLATFORM == X_WINDOWS
 #ifndef WIN32_LEAN_AND_MEAN
@@ -113,15 +112,10 @@
 #include <stdlib.h>
 #define INLINE          _inline
 #endif
 
 
-#if X_PLATFORM == X_SOLARIS
-#define INLINE
-#endif
-
-
 #if X_PLATFORM == X_LINUX
 #define INLINE          inline
 #endif
 
 
