diff a/test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/hotswap/HS204/hs204t004/hs204t004.cpp b/test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/hotswap/HS204/hs204t004/hs204t004.cpp
--- a/test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/hotswap/HS204/hs204t004/hs204t004.cpp
+++ b/test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/hotswap/HS204/hs204t004/hs204t004.cpp
@@ -95,11 +95,11 @@
     clas = jni->FindClass(CLASS_LOADER_CLASS_NAME);
     nsk_printf("Agent::  Break Pont Reached..\n");
     /* Redefine the class loader and then pop the
        frame and resume the thread..*/
     nsk_jvmti_getFileName(redefineNumber, FILE_NAME, fileName, sizeof(fileName)/sizeof(char));
-    if (nsk_jvmti_redefineClass(jvmti_env, clas, fileName) == NSK_TRUE) {
+    if (nsk_jvmti_redefineClass(jvmti_env, clas, fileName)) {
         nsk_printf("\nMyClass :: Successfully redefined..\n");
     } else {
         nsk_printf("\nMyClass :: Failed to redefine ..\n");
     }
     nsk_printf(" End of REDEFINE CLASS LOADER \n");
