diff a/src/java.desktop/share/classes/java/awt/FileDialog.java b/src/java.desktop/share/classes/java/awt/FileDialog.java
--- a/src/java.desktop/share/classes/java/awt/FileDialog.java
+++ b/src/java.desktop/share/classes/java/awt/FileDialog.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1995, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1995, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -123,10 +123,11 @@
      * @serial
      * @see #getFilenameFilter()
      * @see #setFilenameFilter()
      * @see FileNameFilter
      */
+    @SuppressWarnings("serial") // Not statically typed as Serializable
     FilenameFilter filter;
 
     private static final String base = "filedlg";
     private static int nameCounter = 0;
 
@@ -442,11 +443,11 @@
      *
      * @param     dir   the specified directory
      * @see       java.awt.FileDialog#getDirectory
      */
     public void setDirectory(String dir) {
-        this.dir = (dir != null && dir.equals("")) ? null : dir;
+        this.dir = (dir != null && dir.isEmpty()) ? null : dir;
         FileDialogPeer peer = (FileDialogPeer)this.peer;
         if (peer != null) {
             peer.setDirectory(this.dir);
         }
     }
@@ -522,11 +523,11 @@
      * @param    file   the file being set
      * @see      #getFile
      * @see      #getFiles
      */
     public void setFile(String file) {
-        this.file = (file != null && file.equals("")) ? null : file;
+        this.file = (file != null && file.isEmpty()) ? null : file;
         FileDialogPeer peer = (FileDialogPeer)this.peer;
         if (peer != null) {
             peer.setFile(this.file);
         }
     }
@@ -603,14 +604,14 @@
         throws ClassNotFoundException, IOException
     {
         s.defaultReadObject();
 
         // 1.1 Compatibility: "" is not converted to null in 1.1
-        if (dir != null && dir.equals("")) {
+        if (dir != null && dir.isEmpty()) {
             dir = null;
         }
-        if (file != null && file.equals("")) {
+        if (file != null && file.isEmpty()) {
             file = null;
         }
     }
 
     /**
