diff a/src/java.desktop/share/classes/java/awt/MenuItem.java b/src/java.desktop/share/classes/java/awt/MenuItem.java
--- a/src/java.desktop/share/classes/java/awt/MenuItem.java
+++ b/src/java.desktop/share/classes/java/awt/MenuItem.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1995, 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1995, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -49,12 +49,12 @@
  * <p>
  * The default {@code MenuItem} object embodies
  * a simple labeled menu item.
  * <p>
  * This picture of a menu bar shows five menu items:
- * <IMG SRC="doc-files/MenuBar-1.gif" alt="The following text describes this graphic."
- * style="float:center; margin: 7px 10px;">
+ * <IMG SRC="doc-files/MenuBar-1.gif" alt="The following text describes this
+ * graphic." style="margin: 7px 10px;">
  * <br style="clear:left;">
  * The first two items are simple menu items, labeled
  * {@code "Basic"} and {@code "Simple"}.
  * Following these two items is a separator, which is itself
  * a menu item, created with the label {@code "-"}.
