diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/AsyncIteratorPrototype.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/AsyncIteratorPrototype.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/AsyncIteratorPrototype.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/AsyncIteratorPrototype.cpp
@@ -37,11 +37,10 @@
 
 void AsyncIteratorPrototype::finishCreation(VM& vm, JSGlobalObject* globalObject)
 {
     Base::finishCreation(vm);
     ASSERT(inherits(vm, info()));
-    didBecomePrototype();
 
     JSFunction* asyncIteratorPrototypeFunction = JSFunction::create(vm, asyncIteratorPrototypeSymbolAsyncIteratorGetterCodeGenerator(vm), globalObject);
     putDirectWithoutTransition(vm, vm.propertyNames->asyncIteratorSymbol, asyncIteratorPrototypeFunction, static_cast<unsigned>(PropertyAttribute::DontEnum));
 }
 
