diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CommonIdentifiers.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CommonIdentifiers.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CommonIdentifiers.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CommonIdentifiers.h
@@ -1,7 +1,7 @@
 /*
- *  Copyright (C) 2003-2017 Apple Inc. All rights reserved.
+ *  Copyright (C) 2003-2019 Apple Inc. All rights reserved.
  *
  *  This library is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU Library General Public
  *  License as published by the Free Software Foundation; either
  *  version 2 of the License, or (at your option) any later version.
@@ -51,10 +51,11 @@
     macro(RegExp) \
     macro(Set) \
     macro(SharedArrayBuffer) \
     macro(String) \
     macro(Symbol) \
+    macro(WeakRef) \
     macro(__defineGetter__) \
     macro(__defineSetter__) \
     macro(__lookupGetter__) \
     macro(__lookupSetter__) \
     macro(add) \
@@ -86,10 +87,11 @@
     macro(constructor) \
     macro(count) \
     macro(counters) \
     macro(day) \
     macro(defineProperty) \
+    macro(deref) \
     macro(description) \
     macro(descriptions) \
     macro(detail) \
     macro(displayName) \
     macro(done) \
@@ -252,10 +254,11 @@
     macro(hasInstance) \
     macro(isConcatSpreadable) \
     macro(asyncIterator) \
     macro(iterator) \
     macro(match) \
+    macro(matchAll) \
     macro(replace) \
     macro(search) \
     macro(species) \
     macro(split) \
     macro(toPrimitive) \
@@ -277,11 +280,11 @@
     class BuiltinNames;
 
     class CommonIdentifiers {
         WTF_MAKE_NONCOPYABLE(CommonIdentifiers); WTF_MAKE_FAST_ALLOCATED;
     private:
-        CommonIdentifiers(VM*);
+        CommonIdentifiers(VM&);
         ~CommonIdentifiers();
         friend class VM;
 
     public:
         const BuiltinNames& builtinNames() const { return *m_builtinNames; }
