diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConstructData.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConstructData.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConstructData.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConstructData.cpp
@@ -1,7 +1,7 @@
 /*
- * Copyright (C) 2008, 2016 Apple Inc. All Rights Reserved.
+ * Copyright (C) 2008-2019 Apple Inc. All Rights Reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
  * are met:
  * 1. Redistributions of source code must retain the above copyright
@@ -39,12 +39,14 @@
     VM& vm = exec->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
 
     ConstructData constructData;
     ConstructType constructType = getConstructData(vm, constructorObject, constructData);
-    if (constructType == ConstructType::None)
-        return throwTypeError(exec, scope, errorMessage);
+    if (UNLIKELY(constructType == ConstructType::None)) {
+        throwTypeError(exec, scope, errorMessage);
+        return nullptr;
+    }
 
     RELEASE_AND_RETURN(scope, construct(exec, constructorObject, constructType, constructData, args, constructorObject));
 }
 
 
