diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/IndexingType.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/IndexingType.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/IndexingType.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/IndexingType.cpp
@@ -24,10 +24,12 @@
  */
 
 #include "config.h"
 #include "IndexingType.h"
 
+#include "JSCInlines.h"
+
 namespace JSC {
 
 IndexingType leastUpperBoundOfIndexingTypes(IndexingType a, IndexingType b)
 {
     // It doesn't make sense to LUB something that is an array with something that isn't.
