diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp
@@ -32,10 +32,11 @@
 
 const ClassInfo JSImmutableButterfly::s_info = { "Immutable Butterfly", nullptr, nullptr, nullptr, CREATE_METHOD_TABLE(JSImmutableButterfly) };
 
 void JSImmutableButterfly::visitChildren(JSCell* cell, SlotVisitor& visitor)
 {
+    ASSERT_GC_OBJECT_INHERITS(cell, info());
     Base::visitChildren(cell, visitor);
     if (!hasContiguous(cell->indexingType())) {
         ASSERT(hasDouble(cell->indexingType()) || hasInt32(cell->indexingType()));
         return;
     }
