diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h
@@ -33,12 +33,13 @@
 public:
     typedef InternalFunction Base;
 
     static NullGetterFunction* create(VM& vm, Structure* structure)
     {
+        // Since NullGetterFunction is per JSGlobalObject, we use put-without-transition in InternalFunction::finishCreation.
         NullGetterFunction* function = new (NotNull, allocateCell< NullGetterFunction>(vm.heap))  NullGetterFunction(vm, structure);
-        function->finishCreation(vm, String());
+        function->finishCreation(vm, String(), NameVisibility::Visible, NameAdditionMode::WithoutStructureTransition);
         return function;
     }
 
     DECLARE_EXPORT_INFO;
 
