diff a/test/langtools/jdk/javadoc/doclet/testModules/TestModules.java b/test/langtools/jdk/javadoc/doclet/testModules/TestModules.java
--- a/test/langtools/jdk/javadoc/doclet/testModules/TestModules.java
+++ b/test/langtools/jdk/javadoc/doclet/testModules/TestModules.java
@@ -494,19 +494,19 @@
                 + "<p>@Deprecated(forRemoval=true)\n"
                 + "</p>\n"
                 + "<h1 title=\"Module\" class=\"title\">Module&nbsp;moduleA</h1>\n"
                 + "</div>"
                 + "<ul class=\"block-list\">\n"
-                + "<li class=\"block-list\">\n"
+                + "<li>\n"
                 + "<ul class=\"block-list\">\n"
-                + "<li class=\"block-list\">\n"
+                + "<li>\n"
                 + "<!-- ============ PACKAGES SUMMARY =========== -->");
         checkOutput("moduleB/module-summary.html", found,
                 "<ul class=\"block-list\">\n"
-                + "<li class=\"block-list\">\n"
+                + "<li>\n"
                 + "<ul class=\"block-list\">\n"
-                + "<li class=\"block-list\">\n"
+                + "<li>\n"
                 + "<!-- ============ PACKAGES SUMMARY =========== -->");
     }
 
     void checkHtml5Description(boolean found) {
         checkOutput("moduleA/module-summary.html", found,
@@ -549,11 +549,11 @@
                 + "</p>\n"
                 + "<h1 title=\"Module\" class=\"title\">Module&nbsp;moduleA</h1>\n"
                 + "</div>\n"
                 + "<section class=\"summary\">\n"
                 + "<ul class=\"block-list\">\n"
-                + "<li class=\"block-list\">\n"
+                + "<li>\n"
                 + "<section class=\"packages-summary\" id=\"packages.summary\">\n"
                 + "<!-- ============ PACKAGES SUMMARY =========== -->");
         checkOutput("moduleB/module-summary.html", found,
                 "<p><a href=\"testpkgmdlB/AnnotationType.html\" title=\"annotation in testpkgmdlB\">@AnnotationType</a>"
                 + "(<a href=\"testpkgmdlB/AnnotationType.html#optional()\">optional</a>=\"Module Annotation\",\n"
@@ -561,11 +561,11 @@
                 + "</p>\n"
                 + "<h1 title=\"Module\" class=\"title\">Module&nbsp;moduleB</h1>\n"
                 + "</div>\n"
                 + "<section class=\"summary\">\n"
                 + "<ul class=\"block-list\">\n"
-                + "<li class=\"block-list\">\n"
+                + "<li>\n"
                 + "<section class=\"packages-summary\" id=\"packages.summary\">\n"
                 + "<!-- ============ PACKAGES SUMMARY =========== -->");
     }
 
     void checkModuleLink() {
