diff a/modules/javafx.web/src/main/native/Source/WebCore/style/InlineTextBoxStyle.cpp b/modules/javafx.web/src/main/native/Source/WebCore/style/InlineTextBoxStyle.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/style/InlineTextBoxStyle.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/style/InlineTextBoxStyle.cpp
@@ -27,10 +27,11 @@
 #include "InlineTextBoxStyle.h"
 
 #include "FontCascade.h"
 #include "InlineTextBox.h"
 #include "RootInlineBox.h"
+#include "TextUnderlineOffset.h"
 
 namespace WebCore {
 
 float computeUnderlineOffset(TextUnderlinePosition underlinePosition, TextUnderlineOffset underlineOffset, const FontMetrics& fontMetrics, const InlineTextBox* inlineTextBox, float defaultGap)
 {
