diff a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
--- a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
@@ -254,14 +254,25 @@
     public Hash headHash() {
         return new Hash(json.get("sha").asString());
     }
 
     @Override
-    public String sourceRef() {
+    public String fetchRef() {
         return "merge-requests/" + id() + "/head";
     }
 
+    @Override
+    public String sourceRef() {
+        return json.get("source_branch").asString();
+    }
+
+    @Override
+    public HostedRepository sourceRepository() {
+        return new GitLabRepository((GitLabHost) repository.forge(),
+                                    json.get("head").get("source_project_id").asString());
+    }
+
     @Override
     public String targetRef() {
         return json.get("target_branch").asString();
     }
 
