diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/unicode/UTF8Conversion.cpp b/modules/javafx.web/src/main/native/Source/WTF/wtf/unicode/UTF8Conversion.cpp
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/unicode/UTF8Conversion.cpp
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/unicode/UTF8Conversion.cpp
@@ -90,10 +90,11 @@
 bool convertUTF8ToUTF16(const char* source, const char* sourceEnd, UChar** targetStart, UChar* targetEnd, bool* sourceAllASCII)
 {
     RELEASE_ASSERT(sourceEnd - source <= std::numeric_limits<int>::max());
     UBool error = false;
     UChar* target = *targetStart;
+    RELEASE_ASSERT(targetEnd - target <= std::numeric_limits<int>::max());
     UChar32 orAllData = 0;
     int targetOffset = 0;
     for (int sourceOffset = 0; sourceOffset < sourceEnd - source; ) {
         UChar32 character;
         U8_NEXT(reinterpret_cast<const uint8_t*>(source), sourceOffset, sourceEnd - source, character);
@@ -102,10 +103,11 @@
         U16_APPEND(target, targetOffset, targetEnd - target, character, error);
         if (error)
             return false;
         orAllData |= character;
     }
+    RELEASE_ASSERT(target + targetOffset <= targetEnd);
     *targetStart = target + targetOffset;
     if (sourceAllASCII)
         *sourceAllASCII = isASCII(orAllData);
     return true;
 }
