diff a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLExtension.h b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLExtension.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLExtension.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLExtension.h
@@ -23,10 +23,12 @@
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #pragma once
 
+#if ENABLE(WEBGL)
+
 #include "WebGLRenderingContextBase.h"
 
 namespace WebCore {
 
 class WebGLExtension {
@@ -69,5 +71,7 @@
     WebGLExtension(WebGLRenderingContextBase&);
     WebGLRenderingContextBase& m_context;
 };
 
 } // namespace WebCore
+
+#endif
