diff a/modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkDnDClipboard.java b/modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkDnDClipboard.java
--- a/modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkDnDClipboard.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkDnDClipboard.java
@@ -36,11 +36,10 @@
 
     protected void pushToSystem(HashMap<String, Object> cacheData,
                                 int supportedActions) {
         final int performedAction = pushToSystemImpl(cacheData,
                                                      supportedActions);
-
         actionPerformed(performedAction);
     }
 
     @Override
     protected native boolean isOwner();
