diff a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Opcodes.java b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Opcodes.java
--- a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Opcodes.java
+++ b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Opcodes.java
@@ -309,10 +309,11 @@
     int V11 = 0 << 16 | 55;
     int V12 = 0 << 16 | 56;
     int V13 = 0 << 16 | 57;
     int V14 = 0 << 16 | 58;
     int V15 = 0 << 16 | 59;
+    int V16 = 0 << 16 | 60;
 
     /**
       * Version flag indicating that the class is using 'preview' features.
       *
       * <p>{@code version & V_PREVIEW == V_PREVIEW} tests if a version is flagged with {@code
