diff a/src/java.desktop/share/classes/javax/swing/text/html/FormView.java b/src/java.desktop/share/classes/javax/swing/text/html/FormView.java
--- a/src/java.desktop/share/classes/javax/swing/text/html/FormView.java
+++ b/src/java.desktop/share/classes/javax/swing/text/html/FormView.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -600,11 +600,11 @@
         String x = mouseCoords.substring(0, sep);
         String y = mouseCoords.substring(++sep);
         String name = (String) getElement().getAttributes().getAttribute(HTML.Attribute.NAME);
 
         String data;
-        if (name == null || name.equals("")) {
+        if (name == null || name.isEmpty()) {
             data = "x="+ x +"&y="+ y;
         } else {
             name = URLEncoder.encode(name);
             data = name + ".x" +"="+ x +"&"+ name +".y"+"="+ y;
         }
