diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java
@@ -109,10 +109,11 @@
                                           .readyComments(readyComments)
                                           .issueTracker(issueTracker)
                                           .headers(headers)
                                           .sendInterval(interval)
                                           .cooldown(cooldown)
+                                          .seedStorage(configuration.storageFolder().resolve("seeds"))
                                           .build();
             ret.add(bot);
 
             allListNames.add(list);
             allRepositories.add(configuration.repository(repo));
