diff a/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ot-post-macroman.hh b/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ot-post-macroman.hh
--- a/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ot-post-macroman.hh
+++ b/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ot-post-macroman.hh
@@ -25,11 +25,11 @@
  */
 
 #ifndef HB_OT_POST_MACROMAN_HH
 #if 0 /* Make checks happy. */
 #define HB_OT_POST_MACROMAN_HH
-#include "hb-private.hh"
+#include "hb.hh"
 #endif
 
 
 _S(".notdef")
 _S(".null")
