diff a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/CompilationsPage.java b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/CompilationsPage.java
--- a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/CompilationsPage.java
+++ b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/CompilationsPage.java
@@ -123,11 +123,11 @@
 			return FlightRecorderUI.getDefault().getMCImageDescriptor(ImageConstants.PAGE_COMPILATIONS);
 		}
 
 		@Override
 		public String[] getTopics(IState state) {
-			return new String[] {JfrRuleTopics.COMPILATIONS_TOPIC};
+			return new String[] {JfrRuleTopics.COMPILATIONS};
 		}
 
 		@Override
 		public IDisplayablePage createPage(IPageDefinition dpd, StreamModel items, IPageContainer editor) {
 			return new CompilationsPage(dpd, items, editor);
