diff a/mailinglist/src/main/java/org/openjdk/skara/mailinglist/Mbox.java b/mailinglist/src/main/java/org/openjdk/skara/mailinglist/Mbox.java
--- a/mailinglist/src/main/java/org/openjdk/skara/mailinglist/Mbox.java
+++ b/mailinglist/src/main/java/org/openjdk/skara/mailinglist/Mbox.java
@@ -127,10 +127,10 @@
         mboxMail.println("Date: " + mail.date().format(DateTimeFormatter.RFC_1123_DATE_TIME));
         mboxMail.println("Subject: " + MimeText.encode(mail.subject()));
         mboxMail.println("Message-Id: " + mail.id());
         mail.headers().forEach(header -> mboxMail.println(header + ": " + MimeText.encode(mail.headerValue(header))));
         mboxMail.println();
-        mboxMail.println(encodeFromStrings(MimeText.encode(mail.body())));
+        mboxMail.println(encodeFromStrings(mail.body()));
 
         return mboxString.toString();
     }
 }
diff a/mailinglist/src/main/java/org/openjdk/skara/mailinglist/mboxfile/MboxFileList.java b/mailinglist/src/main/java/org/openjdk/skara/mailinglist/mboxfile/MboxFileList.java
--- a/mailinglist/src/main/java/org/openjdk/skara/mailinglist/mboxfile/MboxFileList.java
+++ b/mailinglist/src/main/java/org/openjdk/skara/mailinglist/mboxfile/MboxFileList.java
@@ -53,24 +53,24 @@
                     throw new UncheckedIOException(e);
                 }
             }
         }
         try {
-            Files.writeString(file, mboxMail, StandardCharsets.US_ASCII, StandardOpenOption.APPEND);
+            Files.writeString(file, mboxMail, StandardCharsets.UTF_8, StandardOpenOption.APPEND);
         } catch (IOException e) {
             try {
-                Files.writeString(file, mboxMail, StandardCharsets.US_ASCII, StandardOpenOption.CREATE_NEW);
+                Files.writeString(file, mboxMail, StandardCharsets.UTF_8, StandardOpenOption.CREATE_NEW);
             } catch (IOException e1) {
                 throw new UncheckedIOException(e);
             }
         }
     }
 
     private void postReply(Email mail) {
         var mboxMail = Mbox.fromMail(mail);
         try {
-            Files.writeString(file, mboxMail, StandardCharsets.US_ASCII, StandardOpenOption.APPEND);
+            Files.writeString(file, mboxMail, StandardCharsets.UTF_8, StandardOpenOption.APPEND);
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
@@ -85,11 +85,11 @@
 
     @Override
     public List<Conversation> conversations(Duration maxAge) {
         String mbox;
         try {
-            mbox = Files.readString(file, StandardCharsets.US_ASCII);
+            mbox = Files.readString(file, StandardCharsets.UTF_8);
         } catch (IOException e) {
             log.info("Failed to open mbox file");
             log.throwing("MboxFileList", "conversations", e);
             return new LinkedList<>();
         }
