diff a/src/java.base/share/classes/java/security/AlgorithmParameterGeneratorSpi.java b/src/java.base/share/classes/java/security/AlgorithmParameterGeneratorSpi.java
--- a/src/java.base/share/classes/java/security/AlgorithmParameterGeneratorSpi.java
+++ b/src/java.base/share/classes/java/security/AlgorithmParameterGeneratorSpi.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1997, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -58,10 +58,15 @@
  * @since 1.2
  */
 
 public abstract class AlgorithmParameterGeneratorSpi {
 
+    /**
+     * Constructor for subclasses to call.
+     */
+    public AlgorithmParameterGeneratorSpi() {}
+
     /**
      * Initializes this parameter generator for a certain size
      * and source of randomness.
      *
      * @param size the size (number of bits).
