diff a/modules/javafx.base/src/main/java/javafx/beans/binding/DoubleBinding.java b/modules/javafx.base/src/main/java/javafx/beans/binding/DoubleBinding.java
--- a/modules/javafx.base/src/main/java/javafx/beans/binding/DoubleBinding.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/binding/DoubleBinding.java
@@ -115,11 +115,11 @@
     private boolean valid;
     private BindingHelperObserver observer;
     private ExpressionHelper<Number> helper = null;
 
     /**
-     * Creates a default DoubleBinding.
+     * Creates a default {@code DoubleBinding}.
      */
     public DoubleBinding() {
     }
 
     @Override
