diff a/modules/javafx.base/src/main/java/javafx/beans/binding/IntegerBinding.java b/modules/javafx.base/src/main/java/javafx/beans/binding/IntegerBinding.java
--- a/modules/javafx.base/src/main/java/javafx/beans/binding/IntegerBinding.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/binding/IntegerBinding.java
@@ -65,11 +65,11 @@
     private boolean valid = false;
     private BindingHelperObserver observer;
     private ExpressionHelper<Number> helper = null;
 
     /**
-     * Creates a default IntegerBinding.
+     * Creates a default {@code IntegerBinding}.
      */
     public IntegerBinding() {
     }
 
     @Override
