diff a/vcs/src/main/java/org/openjdk/skara/vcs/tools/GitRange.java b/vcs/src/main/java/org/openjdk/skara/vcs/tools/GitRange.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/tools/GitRange.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/tools/GitRange.java
@@ -33,11 +33,11 @@
             return new Range(start, 1);
         }
 
         var start = Integer.parseInt(s.substring(0, separatorIndex));
 
-        // Need to work arond a bug in git where git sometimes print -1
+        // Need to work around a bug in git where git sometimes print -1
         // as an unsigned int for the count part of the range
         var countString = s.substring(separatorIndex + 1, s.length());
         var count =
             countString.equals("18446744073709551615") ?  0 : Integer.parseInt(countString);
 
