diff a/modules/javafx.web/src/main/native/Source/WebCore/editing/SpellingCorrectionCommand.cpp b/modules/javafx.web/src/main/native/Source/WebCore/editing/SpellingCorrectionCommand.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/editing/SpellingCorrectionCommand.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/editing/SpellingCorrectionCommand.cpp
@@ -128,11 +128,11 @@
 }
 
 RefPtr<DataTransfer> SpellingCorrectionCommand::inputEventDataTransfer() const
 {
     if (!isEditingTextAreaOrTextInput())
-        return DataTransfer::createForInputEvent(m_correction, serializeFragment(*m_correctionFragment, SerializedNodes::SubtreeIncludingNode));
+        return DataTransfer::createForInputEvent(document(), m_correction, serializeFragment(*m_correctionFragment, SerializedNodes::SubtreeIncludingNode));
 
     return CompositeEditCommand::inputEventDataTransfer();
 }
 
 bool SpellingCorrectionCommand::shouldRetainAutocorrectionIndicator() const
