diff a/src/hotspot/share/classfile/modules.hpp b/src/hotspot/share/classfile/modules.hpp
--- a/src/hotspot/share/classfile/modules.hpp
+++ b/src/hotspot/share/classfile/modules.hpp
@@ -105,15 +105,10 @@
   // NullPointerException is thrown if package is null.
   // IllegalArgumentException is thrown if loader is neither null nor a subtype of
   // java/lang/ClassLoader.
   static jobject get_named_module(Handle h_loader, const char* package, TRAPS);
 
-  // If package is defined by loader, return the
-  // java.lang.Module object for the module in which the package is defined.
-  // Returns NULL if package is invalid or not defined by loader.
-  static jobject get_module(Symbol* package_name, Handle h_loader, TRAPS);
-
   // Marks the specified package as exported to all unnamed modules.
   // If either module or package is null then NullPointerException is thrown.
   // If module or package is bad, or module is unnamed, or package is not in
   // module then IllegalArgumentException is thrown.
   static void add_module_exports_to_all_unnamed(jobject module, const char* package, TRAPS);
