diff a/test/langtools/tools/javap/8006334/JavapTaskCtorFailWithNPE.java b/test/langtools/tools/javap/8006334/JavapTaskCtorFailWithNPE.java
--- a/test/langtools/tools/javap/8006334/JavapTaskCtorFailWithNPE.java
+++ b/test/langtools/tools/javap/8006334/JavapTaskCtorFailWithNPE.java
@@ -45,11 +45,11 @@
 public class JavapTaskCtorFailWithNPE {
 
     //we will also check the output just to confirm that we get the expected one
     private static final String expOutput =
         "Compiled from \"JavapTaskCtorFailWithNPE.java\"\n" +
-        "public class JavapTaskCtorFailWithNPE implements java.lang.IdentityObject {\n" +
+        "public class JavapTaskCtorFailWithNPE {\n" +
         "  public JavapTaskCtorFailWithNPE();\n" +
         "  public static void main(java.lang.String[]);\n" +
         "}\n";
 
     public static void main(String[] args) {
