diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/mediasource/MockBox.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/mediasource/MockBox.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/mediasource/MockBox.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/mediasource/MockBox.cpp
@@ -49,11 +49,11 @@
 String MockBox::peekType(ArrayBuffer* data)
 {
     StringBuilder builder;
     auto array = JSC::Int8Array::create(data, 0, 4);
     for (int i = 0; i < 4; ++i)
-        builder.append(array->item(i));
+        builder.append(static_cast<char>(array->item(i)));
     return builder.toString();
 }
 
 size_t MockBox::peekLength(ArrayBuffer* data)
 {
@@ -70,11 +70,11 @@
     m_trackID = view->get<int32_t>(8, true);
 
     StringBuilder builder;
     auto array = JSC::Int8Array::create(data, 12, 4);
     for (int i = 0; i < 4; ++i)
-        builder.append(array->item(i));
+        builder.append(static_cast<char>(array->item(i)));
     m_codec = builder.toString();
 
     m_kind = static_cast<TrackKind>(view->get<uint8_t>(16, true));
 }
 
