diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingMethodSelectedEvent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingMethodSelectedEvent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingMethodSelectedEvent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingMethodSelectedEvent.cpp
@@ -26,14 +26,17 @@
 #include "config.h"
 #include "ApplePayShippingMethodSelectedEvent.h"
 
 #if ENABLE(APPLE_PAY)
 
+#include <wtf/IsoMallocInlines.h>
 #include <wtf/text/StringBuilder.h>
 
 namespace WebCore {
 
+WTF_MAKE_ISO_ALLOCATED_IMPL(ApplePayShippingMethodSelectedEvent);
+
 static inline ApplePayShippingMethod convert(const ApplePaySessionPaymentRequest::ShippingMethod& shippingMethod)
 {
     ApplePayShippingMethod convertedMethod;
     convertedMethod.label = shippingMethod.label;
     convertedMethod.detail = shippingMethod.detail;
