diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h
@@ -35,10 +35,12 @@
 
 class Document;
 class Payment;
 class PaymentContact;
 class PaymentMethod;
+class PaymentSessionError;
+class ScriptExecutionContext;
 
 class PaymentSession : public virtual PaymentSessionBase {
 public:
     static ExceptionOr<void> canCreateSession(Document&);
     static bool enabledForContext(ScriptExecutionContext&);
@@ -47,11 +49,11 @@
     virtual void validateMerchant(URL&&) = 0;
     virtual void didAuthorizePayment(const Payment&) = 0;
     virtual void didSelectShippingMethod(const ApplePaySessionPaymentRequest::ShippingMethod&) = 0;
     virtual void didSelectShippingContact(const PaymentContact&) = 0;
     virtual void didSelectPaymentMethod(const PaymentMethod&) = 0;
-    virtual void didCancelPaymentSession() = 0;
+    virtual void didCancelPaymentSession(PaymentSessionError&&) = 0;
 };
 
 } // namespace WebCore
 
 #endif // ENABLE(APPLE_PAY)
