diff a/test/hotspot/jtreg/compiler/arraycopy/TestCloneAccessStressGCM.java b/test/hotspot/jtreg/compiler/arraycopy/TestCloneAccessStressGCM.java
--- a/test/hotspot/jtreg/compiler/arraycopy/TestCloneAccessStressGCM.java
+++ b/test/hotspot/jtreg/compiler/arraycopy/TestCloneAccessStressGCM.java
@@ -21,18 +21,23 @@
  * questions.
  */
 
 /*
  * @test
- * @bug 8235332
+ * @bug 8235332 8248226
  * @summary Test cloning with more than 8 (=ArrayCopyLoadStoreMaxElem) fields with StressGCM
  * @library /
  *
  * @run main/othervm -Xbatch
  *                   -XX:CompileCommand=dontinline,compiler.arraycopy.TestCloneAccessStressGCM::test
  *                   -XX:+UnlockDiagnosticVMOptions -XX:+StressGCM -XX:-ReduceInitialCardMarks
  *                   compiler.arraycopy.TestCloneAccessStressGCM
+ * @run main/othervm -Xbatch
+ *                   -XX:CompileCommand=dontinline,compiler.arraycopy.TestCloneAccessStressGCM::test
+ *                   -XX:+UnlockDiagnosticVMOptions -XX:+StressGCM -XX:-ReduceInitialCardMarks
+ *                   -XX:-ReduceBulkZeroing
+ *                   compiler.arraycopy.TestCloneAccessStressGCM
  */
 
 package compiler.arraycopy;
 
 public class TestCloneAccessStressGCM {
