diff a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java
--- a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java
@@ -267,10 +267,10 @@
 
     @Override
     public List<HostedBranch> branches() {
         var branches = request.get("branches").execute();
         return branches.stream()
-                       .map(b -> new GitLabBranch(b.get("name").asString(),
+                       .map(b -> new HostedBranch(b.get("name").asString(),
                                                   new Hash(b.get("commit").get("id").asString())))
                        .collect(Collectors.toList());
     }
 }
