diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Const32Value.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Const32Value.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Const32Value.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Const32Value.cpp
@@ -301,13 +301,8 @@
 void Const32Value::dumpMeta(CommaPrinter& comma, PrintStream& out) const
 {
     out.print(comma, m_value);
 }
 
-Value* Const32Value::cloneImpl() const
-{
-    return new Const32Value(*this);
-}
-
 } } // namespace JSC::B3
 
 #endif // ENABLE(B3_JIT)
