diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3FenceValue.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3FenceValue.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3FenceValue.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3FenceValue.cpp
@@ -23,26 +23,22 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #include "config.h"
 #include "B3FenceValue.h"
+#include "B3ValueInlines.h"
 
 #if ENABLE(B3_JIT)
 
 namespace JSC { namespace B3 {
 
 FenceValue::~FenceValue()
 {
 }
 
-Value* FenceValue::cloneImpl() const
-{
-    return new FenceValue(*this);
-}
-
 FenceValue::FenceValue(Origin origin, HeapRange read, HeapRange write)
-    : Value(CheckedOpcode, Fence, Void, origin)
+    : Value(CheckedOpcode, Fence, Void, Zero, origin)
     , read(read)
     , write(write)
 {
 }
 
