diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3WasmAddressValue.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3WasmAddressValue.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3WasmAddressValue.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3WasmAddressValue.h
@@ -38,18 +38,21 @@
 
     ~WasmAddressValue();
 
     GPRReg pinnedGPR() const { return m_pinnedGPR; }
 
+    B3_SPECIALIZE_VALUE_FOR_FIXED_CHILDREN(1)
+    B3_SPECIALIZE_VALUE_FOR_FINAL_SIZE_FIXED_CHILDREN
+
 protected:
     void dumpMeta(CommaPrinter&, PrintStream&) const override;
 
-    Value* cloneImpl() const override;
-
 private:
     friend class Procedure;
+    friend class Value;
 
+    static Opcode opcodeFromConstructor(Origin, Value*, GPRReg) { return WasmAddress; }
     WasmAddressValue(Origin, Value*, GPRReg);
 
     GPRReg m_pinnedGPR;
 };
 
