diff a/modules/javafx.graphics/src/main/java/com/sun/prism/j2d/print/J2DPrinter.java b/modules/javafx.graphics/src/main/java/com/sun/prism/j2d/print/J2DPrinter.java
--- a/modules/javafx.graphics/src/main/java/com/sun/prism/j2d/print/J2DPrinter.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/prism/j2d/print/J2DPrinter.java
@@ -628,13 +628,13 @@
         }
         return paperSet;
     }
 
     private static Map<MediaSizeName, Paper> predefinedPaperMap;
-    private static Map<MediaTray, PaperSource> preDefinedTrayMap;
+    private static Map<MediaTray, PaperSource> predefinedTrayMap;
 
-    private static void initPrefinedMediaMaps() {
+    private static void initPredefinedMediaMaps() {
         if (predefinedPaperMap == null) {
             // North American papers
             predefinedPaperMap = Map.ofEntries(
                 Map.entry(MediaSizeName.NA_LETTER,             Paper.NA_LETTER),
                 Map.entry(MediaSizeName.TABLOID,               Paper.TABLOID),
@@ -660,12 +660,12 @@
                 Map.entry(MediaSizeName.JIS_B5,                Paper.JIS_B5),
                 Map.entry(MediaSizeName.JIS_B6,                Paper.JIS_B6),
                 Map.entry(MediaSizeName.JAPANESE_POSTCARD,     Paper.JAPANESE_POSTCARD));
         }
 
-        if (preDefinedTrayMap == null) {
-            preDefinedTrayMap = Map.of(
+        if (predefinedTrayMap == null) {
+            predefinedTrayMap = Map.of(
                 MediaTray.MAIN,           PaperSource.MAIN,
                 MediaTray.MANUAL,         PaperSource.MANUAL,
                 MediaTray.BOTTOM,         PaperSource.BOTTOM,
                 MediaTray.MIDDLE,         PaperSource.MIDDLE,
                 MediaTray.TOP,            PaperSource.TOP,
@@ -674,11 +674,11 @@
                 MediaTray.LARGE_CAPACITY, PaperSource.LARGE_CAPACITY);
         }
     }
 
     private void populateMedia() {
-        initPrefinedMediaMaps();
+        initPredefinedMediaMaps();
 
         if (paperSet != null) {
             return; // already inited
         }
         Media[] media =
@@ -752,11 +752,11 @@
         return sourceToTrayMap.get(source);
     }
 
     private synchronized final PaperSource addPaperSource(MediaTray tray) {
 
-        PaperSource source = preDefinedTrayMap.get(tray);
+        PaperSource source = predefinedTrayMap.get(tray);
 
         if (source == null) {
             source = PrintHelper.createPaperSource(tray.toString());
         }
 
