diff a/test/langtools/jdk/javadoc/doclet/testIndexInPackageFiles/TestIndexInPackageFiles.java b/test/langtools/jdk/javadoc/doclet/testIndexInPackageFiles/TestIndexInPackageFiles.java
--- a/test/langtools/jdk/javadoc/doclet/testIndexInPackageFiles/TestIndexInPackageFiles.java
+++ b/test/langtools/jdk/javadoc/doclet/testIndexInPackageFiles/TestIndexInPackageFiles.java
@@ -86,19 +86,19 @@
         // Here, we specifically check for both ends of each link.
         // However, we assume the search index files are generated appropriately,
         // to match the A-Z index files checked here.
 
         checkOutput("p/q/package-summary.html", true,
-            "<a id=\"test.name.1\" class=\"searchTagResult\">test.name.1</a>",
-            "<a id=\"test.property.1\" class=\"searchTagResult\">test.property.1</a>");
+            "<span id=\"test.name.1\" class=\"searchTagResult\">test.name.1</span>",
+            "<span id=\"test.property.1\" class=\"searchTagResult\">test.property.1</span>");
 
         checkOutput("p/q/doc-files/extra.html", true,
-            "<a id=\"test.name.2\" class=\"searchTagResult\">test.name.2</a>",
-            "<a id=\"test.property.2\" class=\"searchTagResult\">test.property.2</a>");
+            "<span id=\"test.name.2\" class=\"searchTagResult\">test.name.2</span>",
+            "<span id=\"test.property.2\" class=\"searchTagResult\">test.property.2</span>");
 
         checkOutput("index.html", true,
-            "<a id=\"test.name.3\" class=\"searchTagResult\">test.name.3</a>");
+            "<span id=\"test.name.3\" class=\"searchTagResult\">test.name.3</span>");
 
         checkOutput("index-all.html", true,
             "<span class=\"searchTagLink\"><a href=\"p/q/package-summary.html#test.name.1\">test.name.1</a></span>",
             "<span class=\"searchTagLink\"><a href=\"p/q/doc-files/extra.html#test.name.2\">test.name.2</a></span>",
             "<span class=\"searchTagLink\"><a href=\"index.html#test.name.3\">test.name.3</a></span> - Search tag in Overview</dt>",
