diff a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java
--- a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java
+++ b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java
@@ -241,11 +241,11 @@
     }
 
     static class StorageCalculator {
         private final boolean forArguments;
 
-        private int nRegs[] = new int[] { 0, 0 };
+        private final int[] nRegs = new int[] { 0, 0 };
         private long stackOffset = 0;
 
         public StorageCalculator(boolean forArguments) {
             this.forArguments = forArguments;
         }
@@ -459,11 +459,10 @@
                 .move(INDIRECT_RESULT, long.class)
                 .convertAddress()
                 .build();
         }
 
-        @SuppressWarnings("fallthrough")
         @Override
         List<Binding> getBindings(Class<?> carrier, MemoryLayout layout) {
             TypeClass argumentClass = classifyType(layout);
             Binding.Builder bindings = Binding.builder();
             switch (argumentClass) {
