diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/mediastream/libwebrtc/GStreamerVideoEncoderFactory.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/mediastream/libwebrtc/GStreamerVideoEncoderFactory.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/mediastream/libwebrtc/GStreamerVideoEncoderFactory.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/mediastream/libwebrtc/GStreamerVideoEncoderFactory.h
@@ -26,10 +26,11 @@
 #include <gst/gst.h>
 
 namespace WebCore {
 
 class GStreamerVideoEncoderFactory final : public webrtc::VideoEncoderFactory {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     GStreamerVideoEncoderFactory();
 
 private:
     std::vector<webrtc::SdpVideoFormat> GetSupportedFormats() const override;
