diff a/make/RunTestsPrebuilt.gmk b/make/RunTestsPrebuilt.gmk
--- a/make/RunTestsPrebuilt.gmk
+++ b/make/RunTestsPrebuilt.gmk
@@ -1,7 +1,7 @@
 #
-# Copyright (c) 2017, 2019, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2017, 2020, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.  Oracle designates this
@@ -334,11 +334,11 @@
 
 SPEC := $(NEW_SPEC)
 
 default: all
 
-run-test-prebuilt:
+test-prebuilt:
         # Need to make sure the failure logs output dir exists since
         # ExecuteWithLog is called in RunTests.gmk. The PrepareFailureLogs macro
         # is unfortunately not available at this point.
 	$(call MakeDir, $(MAKESUPPORT_OUTPUTDIR)/failure-logs)
 	@$(RM) -f $(MAKESUPPORT_OUTPUTDIR)/exit-with-error
@@ -349,8 +349,8 @@
 	@cd $(TOPDIR) && $(MAKE) $(MAKE_ARGS) -f make/common/FindTests.gmk \
 	    SPEC=$(SPEC) FINDTESTS_STAND_ALONE=true
 	@cd $(TOPDIR) && $(MAKE) $(MAKE_ARGS) -f make/RunTests.gmk run-test \
 	    TEST="$(TEST)"
 
-all: run-test-prebuilt
+all: test-prebuilt
 
-.PHONY: default all
+.PHONY: default all test-prebuilt
