diff a/test/lib/jdk/test/lib/jfr/EventNames.java b/test/lib/jdk/test/lib/jfr/EventNames.java
--- a/test/lib/jdk/test/lib/jfr/EventNames.java
+++ b/test/lib/jdk/test/lib/jfr/EventNames.java
@@ -139,10 +139,17 @@
     public final static String G1EvacuationOldStatistics = PREFIX + "G1EvacuationOldStatistics";
     public final static String G1BasicIHOP = PREFIX + "G1BasicIHOP";
     public final static String AllocationRequiringGC = PREFIX + "AllocationRequiringGC";
     public final static String GCPhaseParallel = PREFIX + "GCPhaseParallel";
     public final static String GCPhaseConcurrent = PREFIX + "GCPhaseConcurrent";
+    public final static String GCPhaseConcurrentLevel1 = PREFIX + "GCPhaseConcurrentLevel1";
+    public final static String ZAllocationStall = PREFIX + "ZAllocationStall";
+    public final static String ZPageAllocation = PREFIX + "ZPageAllocation";
+    public final static String ZPageCacheFlush = PREFIX + "ZPageCacheFlush";
+    public final static String ZRelocationSet = PREFIX + "ZRelocationSet";
+    public final static String ZRelocationSetGroup = PREFIX + "ZRelocationSetGroup";
+    public final static String ZUncommit = PREFIX + "ZUncommit";
 
     // Compiler
     public final static String Compilation = PREFIX + "Compilation";
     public final static String CompilerPhase = PREFIX + "CompilerPhase";
     public final static String CompilationFailure = PREFIX + "CompilationFailure";
@@ -169,10 +176,11 @@
     public final static String ThreadContextSwitchRate = PREFIX + "ThreadContextSwitchRate";
     public final static String InitialEnvironmentVariable = PREFIX + "InitialEnvironmentVariable";
     public final static String NativeLibrary = PREFIX + "NativeLibrary";
     public final static String PhysicalMemory = PREFIX + "PhysicalMemory";
     public final static String NetworkUtilization = PREFIX + "NetworkUtilization";
+    public static final String ProcessStart = PREFIX + "ProcessStart";
 
     // JDK
     public static final String FileForce  = PREFIX + "FileForce";
     public static final String FileRead = PREFIX + "FileRead";
     public static final String FileWrite = PREFIX + "FileWrite";
