diff a/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/Makefile.am b/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/Makefile.am
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/Makefile.am
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/Makefile.am
@@ -62,11 +62,10 @@
 LIBXSLT_VERSION_SCRIPT =
 endif
 
 libxslt_la_LIBADD = $(LIBXML_LIBS) $(EXTRA_LIBS) $(M_LIBS)
 libxslt_la_LDFLAGS =					\
-		$(WIN32_EXTRA_LDFLAGS)			\
 		$(LIBXSLT_VERSION_SCRIPT)		\
 		-version-info $(LIBXSLT_VERSION_INFO)
 
 man_MANS = libxslt.3
 
