diff a/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/security.c b/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/security.c
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/security.c
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/libxslt/src/libxslt/security.c
@@ -383,20 +383,22 @@
     }
     if ((uri->scheme == NULL) ||
     (xmlStrEqual(BAD_CAST uri->scheme, BAD_CAST "file"))) {
 
 #if defined(_WIN32) && !defined(__CYGWIN__)
-    if ((uri->path)&&(uri->path[0]=='/')&&
-        (uri->path[1]!='\0')&&(uri->path[2]==':'))
-    ret = xsltCheckWritePath(sec, ctxt, uri->path+1);
-    else
+        if ((uri->path)&&(uri->path[0]=='/')&&
+            (uri->path[1]!='\0')&&(uri->path[2]==':'))
+            ret = xsltCheckWritePath(sec, ctxt, uri->path+1);
+        else
 #endif
+        {
+            /*
+             * Check if we are allowed to write this file
+             */
+        ret = xsltCheckWritePath(sec, ctxt, uri->path);
+        }
 
-    /*
-     * Check if we are allowed to write this file
-     */
-    ret = xsltCheckWritePath(sec, ctxt, uri->path);
     if (ret <= 0) {
         xmlFreeURI(uri);
         return(ret);
     }
     } else {
