diff a/modules/javafx.web/src/main/native/Source/WebCore/html/track/AudioTrack.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/track/AudioTrack.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/track/AudioTrack.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/track/AudioTrack.cpp
@@ -158,11 +158,11 @@
     setLanguage(language);
 }
 
 void AudioTrack::willRemove()
 {
-    auto element = makeRefPtr(mediaElement());
+    auto element = makeRefPtr(mediaElement().get());
     if (!element)
         return;
     element->removeAudioTrack(*this);
 }
 
@@ -194,11 +194,11 @@
         ASSERT_NOT_REACHED();
         break;
     }
 }
 
-void AudioTrack::setMediaElement(HTMLMediaElement* element)
+void AudioTrack::setMediaElement(WeakPtr<HTMLMediaElement> element)
 {
     TrackBase::setMediaElement(element);
 #if !RELEASE_LOG_DISABLED
     m_private->setLogger(logger(), logIdentifier());
 #endif
