diff a/test/jdk/java/util/Formatter/Basic.sh b/test/jdk/java/util/Formatter/Basic.sh
--- a/test/jdk/java/util/Formatter/Basic.sh
+++ b/test/jdk/java/util/Formatter/Basic.sh
@@ -1,7 +1,7 @@
 #
-# Copyright (c) 2004, 2013, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2004, 2019, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.
@@ -22,10 +22,11 @@
 #
 
 #
 
 EXTRAOPTS="--add-exports java.base/jdk.internal.math=ALL-UNNAMED"
+LOCALEOPTS="-Djava.locale.providers=CLDR"
 ${COMPILEJAVA}/bin/javac ${TESTJAVACOPTS} ${TESTTOOLVMOPTS} ${EXTRAOPTS} -cp ${TESTSRC} -d . \
     ${TESTSRC}/Basic.java
 
 expectPass() {
   if [ $1 -eq 0 ]
@@ -38,11 +39,11 @@
 
 runTest() {
   echo "Testing:" ${1}
   TZ="${1}"; export TZ
   echo "  " $TZ
-  ${TESTJAVA}/bin/java ${TESTVMOPTS} ${EXTRAOPTS} Basic
+  ${TESTJAVA}/bin/java ${TESTVMOPTS} ${EXTRAOPTS} ${LOCALEOPTS} Basic
   expectPass $?
 }
 
 runTest US/Pacific
 runTest Asia/Novosibirsk
