diff a/test/jdk/tools/jjs/strict.js b/test/jdk/tools/jjs/strict.js
--- a/test/jdk/tools/jjs/strict.js
+++ /dev/null
@@ -1,12 +0,0 @@
-/*
- * This is the test JavaScript program used in jjs-strictTest.sh
- */
-
-try {
-    v = "hello";
-    throw new Error("should have thrown ReferenceError");
-} catch (e) {
-    if (! (e instanceof ReferenceError)) {
-        throw new Error("ReferenceError expected, got " + e);
-    }
-}
