diff a/bots/csr/src/main/java/org/openjdk/skara/bots/csr/CSRBot.java b/bots/csr/src/main/java/org/openjdk/skara/bots/csr/CSRBot.java
--- a/bots/csr/src/main/java/org/openjdk/skara/bots/csr/CSRBot.java
+++ b/bots/csr/src/main/java/org/openjdk/skara/bots/csr/CSRBot.java
@@ -28,13 +28,11 @@
 import org.openjdk.skara.forge.PullRequestUpdateCache;
 import org.openjdk.skara.issuetracker.IssueProject;
 import org.openjdk.skara.issuetracker.Issue;
 
 import java.nio.file.Path;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
 import java.util.logging.Logger;
 
 class CSRBot implements Bot, WorkItem {
     private final static String CSR_LABEL = "csr";
     private final Logger log = Logger.getLogger("org.openjdk.skara.bots");;
@@ -61,11 +59,11 @@
     private String describe(PullRequest pr) {
         return repo.name() + "#" + pr.id();
     }
 
     @Override
-    public void run(Path scratchPath) {
+    public Collection<WorkItem> run(Path scratchPath) {
         var prs = repo.pullRequests();
         for (var pr : prs) {
             if (!cache.needsUpdate(pr)) {
                 continue;
             }
@@ -130,10 +128,11 @@
                     pr.removeLabel(CSR_LABEL);
                     hasCSRLabel.remove(pr.id());
                 }
             }
         }
+        return List.of();
     }
 
     @Override
     public String toString() {
         return "CSRBot@" + repo.name();
