diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGGradientElement.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGGradientElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGGradientElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGGradientElement.h
@@ -21,11 +21,10 @@
 
 #pragma once
 
 #include "Gradient.h"
 #include "SVGElement.h"
-#include "SVGExternalResourcesRequired.h"
 #include "SVGNames.h"
 #include "SVGURIReference.h"
 #include "SVGUnitTypes.h"
 
 namespace WebCore {
@@ -68,11 +67,11 @@
             return SVGSpreadMethodRepeat;
         return SVGSpreadMethodUnknown;
     }
 };
 
-class SVGGradientElement : public SVGElement, public SVGExternalResourcesRequired, public SVGURIReference {
+class SVGGradientElement : public SVGElement, public SVGURIReference {
     WTF_MAKE_ISO_ALLOCATED(SVGGradientElement);
 public:
     enum {
         SVG_SPREADMETHOD_UNKNOWN = SVGSpreadMethodUnknown,
         SVG_SPREADMETHOD_PAD = SVGSpreadMethodReflect,
@@ -80,11 +79,11 @@
         SVG_SPREADMETHOD_REPEAT = SVGSpreadMethodUnknown
     };
 
     Vector<Gradient::ColorStop> buildStops();
 
-    using PropertyRegistry = SVGPropertyOwnerRegistry<SVGGradientElement, SVGElement, SVGExternalResourcesRequired, SVGURIReference>;
+    using PropertyRegistry = SVGPropertyOwnerRegistry<SVGGradientElement, SVGElement, SVGURIReference>;
 
     SVGSpreadMethodType spreadMethod() const { return m_spreadMethod->currentValue<SVGSpreadMethodType>(); }
     SVGUnitTypes::SVGUnitType gradientUnits() const { return m_gradientUnits->currentValue<SVGUnitTypes::SVGUnitType>(); }
     const SVGTransformList& gradientTransform() const { return m_gradientTransform->currentValue(); }
 
