diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGImageElement.idl b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGImageElement.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGImageElement.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGImageElement.idl
@@ -29,7 +29,6 @@
     readonly attribute SVGAnimatedLength width;
     readonly attribute SVGAnimatedLength height;
     readonly attribute SVGAnimatedPreserveAspectRatio preserveAspectRatio;
 };
 
-SVGImageElement implements SVGExternalResourcesRequired;
 SVGImageElement implements SVGURIReference;
