diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGPathByteStreamBuilder.cpp b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGPathByteStreamBuilder.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGPathByteStreamBuilder.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGPathByteStreamBuilder.cpp
@@ -21,10 +21,11 @@
 #include "config.h"
 #include "SVGPathByteStreamBuilder.h"
 
 #include "SVGPathSeg.h"
 #include "SVGPathStringSource.h"
+#include <wtf/text/StringConcatenateNumbers.h>
 
 namespace WebCore {
 
 SVGPathByteStreamBuilder::SVGPathByteStreamBuilder(SVGPathByteStream& byteStream)
     : m_byteStream(byteStream)
