diff a/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ucdn/ucdn.h b/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ucdn/ucdn.h
--- a/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ucdn/ucdn.h
+++ b/src/java.desktop/share/native/libfontmanager/harfbuzz/hb-ucdn/ucdn.h
@@ -454,10 +454,8 @@
  * @param b second codepoint
  * @return success
  */
 int ucdn_compose(uint32_t *code, uint32_t a, uint32_t b);
 
-#ifdef __cplusplus
-}
-#endif
+HB_END_HEADER
 
 #endif
