diff a/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java b/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java
--- a/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java
+++ b/bot/src/main/java/org/openjdk/skara/bot/BotRunnerConfiguration.java
@@ -76,14 +76,19 @@
                 if (github.contains("weburl")) {
                     webUriPattern = Pattern.compile(github.get("weburl").get("pattern").asString());
                     webUriReplacement = github.get("weburl").get("replacement").asString();
                 }
 
-                var keyFile = cwd.resolve(github.get("app").get("key").asString());
-                ret.put(entry.name(), HostFactory.createGitHubHost(uri, webUriPattern, webUriReplacement, keyFile.toString(),
-                                                                   github.get("app").get("id").asString(),
-                                                                   github.get("app").get("installation").asString()));
+                if (github.contains("app")) {
+                    var keyFile = cwd.resolve(github.get("app").get("key").asString());
+                    ret.put(entry.name(), HostFactory.createGitHubHost(uri, webUriPattern, webUriReplacement, keyFile.toString(),
+                                                                       github.get("app").get("id").asString(),
+                                                                       github.get("app").get("installation").asString()));
+                } else {
+                    var pat = new PersonalAccessToken(github.get("username").asString(), github.get("pat").asString());
+                    ret.put(entry.name(), HostFactory.createGitHubHost(uri, pat));
+                }
             } else {
                 throw new ConfigurationError("Host " + entry.name());
             }
         }
 
