diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGPropertyTraits.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGPropertyTraits.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGPropertyTraits.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGPropertyTraits.h
@@ -25,12 +25,10 @@
 #include "Color.h"
 #include "FloatPoint.h"
 #include "FloatRect.h"
 #include "QualifiedName.h"
 #include "SVGParserUtilities.h"
-#include <wtf/text/StringBuilder.h>
-#include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
 template<typename PropertyType>
 struct SVGPropertyTraits { };
@@ -101,11 +99,11 @@
         float number;
         if (!parseNumberFromString(string, number))
             return WTF::nullopt;
         return number;
     }
-    static String toString(float type) { return String::numberToStringFixedPrecision(type); }
+    static String toString(float type) { return String::number(type); }
 };
 
 template<>
 struct SVGPropertyTraits<std::pair<float, float>> {
     static std::pair<float, float> initialValue() { return { }; }
@@ -137,15 +135,11 @@
             return WTF::nullopt;
         return point;
     }
     static String toString(const FloatPoint& type)
     {
-        StringBuilder builder;
-        builder.appendFixedPrecisionNumber(type.x());
-        builder.append(' ');
-        builder.appendFixedPrecisionNumber(type.y());
-        return builder.toString();
+        return makeString(type.x(), ' ', type.y());
     }
 };
 
 template<>
 struct SVGPropertyTraits<FloatRect> {
@@ -164,19 +158,11 @@
             return WTF::nullopt;
         return rect;
     }
     static String toString(const FloatRect& type)
     {
-        StringBuilder builder;
-        builder.appendFixedPrecisionNumber(type.x());
-        builder.append(' ');
-        builder.appendFixedPrecisionNumber(type.y());
-        builder.append(' ');
-        builder.appendFixedPrecisionNumber(type.width());
-        builder.append(' ');
-        builder.appendFixedPrecisionNumber(type.height());
-        return builder.toString();
+        return makeString(type.x(), ' ', type.y(), ' ', type.width(), ' ', type.height());
     }
 };
 
 template<>
 struct SVGPropertyTraits<String> {
