diff a/modules/javafx.web/src/main/native/Source/WebCore/animation/AnimationTimeline.h b/modules/javafx.web/src/main/native/Source/WebCore/animation/AnimationTimeline.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/animation/AnimationTimeline.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/animation/AnimationTimeline.h
@@ -75,11 +75,11 @@
     virtual ~AnimationTimeline();
 
 protected:
     explicit AnimationTimeline();
 
-    ListHashSet<WebAnimation*> m_allAnimations;
+    Vector<WeakPtr<WebAnimation>> m_allAnimations;
     ListHashSet<RefPtr<WebAnimation>> m_animations;
     HashMap<Element*, PropertyToTransitionMap> m_elementToCompletedCSSTransitionByCSSPropertyID;
 
 private:
     RefPtr<WebAnimation> cssAnimationForElementAndProperty(Element&, CSSPropertyID);
