diff a/modules/javafx.web/src/main/native/Source/WebCore/animation/CSSAnimation.h b/modules/javafx.web/src/main/native/Source/WebCore/animation/CSSAnimation.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/animation/CSSAnimation.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/animation/CSSAnimation.h
@@ -33,10 +33,11 @@
 class Animation;
 class Element;
 class RenderStyle;
 
 class CSSAnimation final : public DeclarativeAnimation {
+    WTF_MAKE_ISO_ALLOCATED(CSSAnimation);
 public:
     static Ref<CSSAnimation> create(Element&, const Animation&, const RenderStyle* oldStyle, const RenderStyle& newStyle);
     ~CSSAnimation() = default;
 
     bool isCSSAnimation() const override { return true; }
