diff a/tests/system/src/test/java/test/robot/javafx/embed/swing/JFXPanelScaledTest.java b/tests/system/src/test/java/test/robot/javafx/embed/swing/JFXPanelScaledTest.java
--- a/tests/system/src/test/java/test/robot/javafx/embed/swing/JFXPanelScaledTest.java
+++ b/tests/system/src/test/java/test/robot/javafx/embed/swing/JFXPanelScaledTest.java
@@ -26,10 +26,11 @@
 package test.robot.javafx.embed.swing;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
+import static org.junit.Assume.assumeTrue;
 import static test.util.Util.TIMEOUT;
 
 import java.awt.Color;
 import java.awt.Dimension;
 import java.awt.image.BufferedImage;
@@ -38,10 +39,11 @@
 
 import javax.swing.JFrame;
 import javax.swing.SwingUtilities;
 import javax.swing.Timer;
 
+import com.sun.javafx.PlatformUtil;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
 import javafx.application.Platform;
@@ -58,16 +60,18 @@
 
     private static Timer t;
     static int cnt;
 
     /* Base size, so that with a scaling of 125% there are different results for Math.round and Math.ceil */
-    final static int TAGET_BASE_SIZE = 101;
+    final static int TARGET_BASE_SIZE = 101;
 
-    final static int TARGET_SCALED_SIZE = (int) Math.ceil(TAGET_BASE_SIZE *1.25);
+    final static int TARGET_SCALED_SIZE = (int) Math.ceil(TARGET_BASE_SIZE *1.25);
 
     @BeforeClass
     public static void setupOnce() throws Exception {
+        assumeTrue(PlatformUtil.isWindows());
+
         System.setProperty("sun.java2d.uiScale.enabled", "true");
         System.setProperty("sun.java2d.uiScale", "125%");
         System.setProperty("glass.win.uiScale", "1.25");
         System.setProperty("glass.gtk.uiScale", "1.25");
         launchLatch = new CountDownLatch(1);
@@ -131,11 +135,11 @@
                 switch (cnt) {
                 case 0:
                     jfxPanel.setSize(new Dimension(201, 201));
                     break;
                 case 1:
-                    jfxPanel.setSize(new Dimension(TAGET_BASE_SIZE, TAGET_BASE_SIZE));
+                    jfxPanel.setSize(new Dimension(TARGET_BASE_SIZE, TARGET_BASE_SIZE));
                     break;
                 case 2:
                     t.stop();
                     launchLatch.countDown();
                     break;
