diff a/src/hotspot/os/bsd/globals_bsd.hpp b/src/hotspot/os/bsd/globals_bsd.hpp
--- a/src/hotspot/os/bsd/globals_bsd.hpp
+++ b/src/hotspot/os/bsd/globals_bsd.hpp
@@ -34,19 +34,11 @@
                          product_pd, \
                          diagnostic, \
                          diagnostic_pd, \
                          notproduct, \
                          range, \
-                         constraint) \
-                                                                                \
-  product(bool, UseOprofile, false,                                             \
-        "enable support for Oprofile profiler")                                 \
-                                                                                \
-  /*  NB: The default value of UseBsdPosixThreadCPUClocks may be  */            \
-  /*  overridden in Arguments::parse_each_vm_init_arg.            */            \
-  product(bool, UseBsdPosixThreadCPUClocks, true,                               \
-          "enable fast Bsd Posix clocks where available")                       \
+                         constraint)
 
 //
 // Defines Bsd-specific default values. The flags are available on all
 // platforms, but they may have different default values on other platforms.
 //
