diff a/src/hotspot/cpu/s390/c1_MacroAssembler_s390.hpp b/src/hotspot/cpu/s390/c1_MacroAssembler_s390.hpp
--- a/src/hotspot/cpu/s390/c1_MacroAssembler_s390.hpp
+++ b/src/hotspot/cpu/s390/c1_MacroAssembler_s390.hpp
@@ -39,18 +39,18 @@
 
   void initialize_header(Register obj, Register klass, Register len, Register Rzero, Register t1);
   void initialize_body(Register objectFields, Register len_in_bytes, Register Rzero);
 
   // locking
-  // hdr     : Used to hold locked markOop to be CASed into obj, contents destroyed.
+  // hdr     : Used to hold locked markWord to be CASed into obj, contents destroyed.
   // obj     : Must point to the object to lock, contents preserved.
   // disp_hdr: Must point to the displaced header location, contents preserved.
   // Returns code offset at which to add null check debug information.
   void lock_object(Register hdr, Register obj, Register disp_hdr, Label& slow_case);
 
   // unlocking
-  // hdr     : Used to hold original markOop to be CASed back into obj, contents destroyed.
+  // hdr     : Used to hold original markWord to be CASed back into obj, contents destroyed.
   // obj     : Must point to the object to lock, contents preserved.
   // disp_hdr: Must point to the displaced header location, contents destroyed.
   void unlock_object(Register hdr, Register obj, Register lock, Label& slow_case);
 
   void initialize_object(
