diff a/test/jdk/java/rmi/activation/rmidViaInheritedChannel/InheritedChannelNotServerSocket.java b/test/jdk/java/rmi/activation/rmidViaInheritedChannel/InheritedChannelNotServerSocket.java
--- a/test/jdk/java/rmi/activation/rmidViaInheritedChannel/InheritedChannelNotServerSocket.java
+++ b/test/jdk/java/rmi/activation/rmidViaInheritedChannel/InheritedChannelNotServerSocket.java
@@ -136,11 +136,11 @@
     public static class SP extends SelectorProvider {
         private final SelectorProvider provider;
         private volatile SocketChannel channel = null;
 
         public SP() {
-            provider = sun.nio.ch.DefaultSelectorProvider.create();
+            provider = sun.nio.ch.DefaultSelectorProvider.get();
         }
 
         public DatagramChannel openDatagramChannel() throws IOException {
             return provider.openDatagramChannel();
         }
