diff a/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java b/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java
--- a/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java
+++ b/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java
@@ -21,11 +21,12 @@
  * questions.
  */
 package org.openjdk.skara.bots.notify.issue;
 
 import org.junit.jupiter.api.*;
-import org.openjdk.skara.bots.notify.*;
+import org.openjdk.skara.bots.notify.NotifyBot;
+import org.openjdk.skara.issuetracker.Issue;
 import org.openjdk.skara.json.JSON;
 import org.openjdk.skara.test.*;
 
 import java.io.IOException;
 import java.net.URI;
@@ -321,10 +322,11 @@
             assertEquals(reviewIcon, links.get(0).iconUrl().orElseThrow());
 
             // Simulate integration
             pr.addComment("Pushed as commit " + editHash.hex() + ".");
             pr.addLabel("integrated");
+            pr.setState(Issue.State.CLOSED);
             localRepo.push(editHash, repo.url(), "other");
             TestBotRunner.runPeriodicItems(notifyBot);
 
             // The changeset should be reflected in another link
             var updatedIssue = issueProject.issue(issue.id()).orElseThrow();
