diff a/test/langtools/jdk/javadoc/doclet/AccessSkipNav/AccessSkipNav.java b/test/langtools/jdk/javadoc/doclet/AccessSkipNav/AccessSkipNav.java
--- a/test/langtools/jdk/javadoc/doclet/AccessSkipNav/AccessSkipNav.java
+++ b/test/langtools/jdk/javadoc/doclet/AccessSkipNav/AccessSkipNav.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2002, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2002, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -50,16 +50,16 @@
         // Testing only for the presence of the <a href> and <a id>
         checkOutput("p1/C1.html", true,
                 // Top navbar <a href>
                 "<a href=\"#skip.navbar.top\" title=\"Skip navigation links\">Skip navigation links</a>",
                 // Top navbar <span id>
-                "<span class=\"skipNav\" id=\"skip.navbar.top\">\n"
+                "<span class=\"skip-nav\" id=\"skip.navbar.top\">\n"
                 + "<!--   -->\n"
                 + "</span>",
                 // Bottom navbar <a href>
                 "<a href=\"#skip.navbar.bottom\" title=\"Skip navigation links\">Skip navigation links</a>",
                 // Bottom navbar <span id>
-                "<span class=\"skipNav\" id=\"skip.navbar.bottom\">\n"
+                "<span class=\"skip-nav\" id=\"skip.navbar.bottom\">\n"
                 + "<!--   -->\n"
                 + "</span>");
     }
 }
