diff a/src/hotspot/cpu/ppc/ppc.ad b/src/hotspot/cpu/ppc/ppc.ad
--- a/src/hotspot/cpu/ppc/ppc.ad
+++ b/src/hotspot/cpu/ppc/ppc.ad
@@ -15194,11 +15194,11 @@
 
   format %{ "ShouldNotReachHere" %}
   ins_encode %{
     if (is_reachable()) {
       // TODO: PPC port $archOpcode(ppc64Opcode_tdi);
-      __ should_not_reach_here();
+      __ stop(_halt_reason);
     }
   %}
   ins_pipe(pipe_class_default);
 %}
 
