diff a/test/nashorn/src/META-INF/services/jdk.dynalink.linker.GuardingDynamicLinkerExporter b/test/nashorn/src/META-INF/services/jdk.dynalink.linker.GuardingDynamicLinkerExporter
--- a/test/nashorn/src/META-INF/services/jdk.dynalink.linker.GuardingDynamicLinkerExporter
+++ /dev/null
@@ -1,3 +0,0 @@
-jdk.dynalink.test.UntrustedGuardingDynamicLinkerExporter
-jdk.dynalink.test.TrustedGuardingDynamicLinkerExporter
-jdk.dynalink.test.TrustedUnderscoreNameLinkerExporter
