diff a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java
@@ -23,12 +23,11 @@
  * questions.
  */
 
 package jdk.incubator.jextract.tool;
 
-import jdk.incubator.jextract.Declaration;
-import jdk.incubator.jextract.JextractTask;
+import jdk.incubator.jextract.*;
 import jdk.internal.joptsimple.OptionException;
 import jdk.internal.joptsimple.OptionParser;
 import jdk.internal.joptsimple.OptionSet;
 
 import javax.tools.JavaFileObject;
@@ -191,11 +190,11 @@
 
             Path output = Path.of(options.outputDir);
 
             JavaFileObject[] files = OutputFactory.generateWrapped(
                 toplevel,
-                Utils.javaSafeIdentifier(header.getFileName().toString().replace(".h", "_h"), true),
+                header.getFileName().toString(),
                 options.targetPackage,
                 options.libraryNames);
             jextractTask.write(output, files);
         } catch (RuntimeException re) {
             err.println(re);
