diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/MailingListUpdater.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/MailingListUpdater.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/MailingListUpdater.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/MailingListUpdater.java
@@ -330,15 +330,15 @@
         } else {
             printer.println("The new branch " + branch.name() + " is currently identical to the " + parent.name() + " branch.");
         }
 
         var subject = newBranchSubject(repository, commits, parent, branch);
-        var lastCommit = commits.get(commits.size() - 1);
+        var finalAuthor = commits.size() > 0 ? commitToAuthor(commits.get(commits.size() - 1)) : sender;
 
         var email = Email.create(subject, writer.toString())
                          .sender(sender)
-                         .author(commitToAuthor(lastCommit))
+                         .author(finalAuthor)
                          .recipient(recipient)
                          .headers(headers)
                          .build();
         list.post(email);
     }
