diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js b/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js
@@ -263,11 +263,11 @@
 {
     "use strict";
     
     const promiseCapability = @newPromiseCapability(@Promise);
     if (!@isObject(generator) || typeof @getByIdDirectPrivate(generator, "asyncGeneratorSuspendReason") !== 'number') {
-        promiseCapability.@reject.@call(@undefined, new @TypeError('|this| should be an async generator'));
+        promiseCapability.@reject.@call(@undefined, @makeTypeError('|this| should be an async generator'));
         return promiseCapability.@promise;
     }
 
     @asyncGeneratorQueueEnqueue(generator, {resumeMode, value, promiseCapability, @asyncGeneratorQueueItemNext: null, @asyncGeneratorQueueItemPrevious: null});
 
