diff a/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.cpp b/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.cpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.cpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.cpp
@@ -96,22 +96,10 @@
   StringDedupUnlinkOrOopsDoClosure sd_cl(&always_true, cl);
   StringDedupQueue::unlink_or_oops_do(&sd_cl);
   StringDedupTable::unlink_or_oops_do(&sd_cl, 0);
 }
 
-class ShenandoahIsMarkedNextClosure : public BoolObjectClosure {
-private:
-  ShenandoahMarkingContext* const _mark_context;
-
-public:
-  ShenandoahIsMarkedNextClosure() : _mark_context(ShenandoahHeap::heap()->marking_context()) { }
-
-  bool do_object_b(oop obj) {
-    return _mark_context->is_marked(obj);
-  }
-};
-
 //
 // Task for parallel unlink_or_oops_do() operation on the deduplication queue
 // and table.
 //
 class ShenandoahStringDedupUnlinkOrOopsDoTask : public AbstractGangTask {
