diff a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp
@@ -52,11 +52,11 @@
 bool AccessibilityTableHeaderContainer::computeAccessibilityIsIgnored() const
 {
     if (!m_parent)
         return true;
 
-#if PLATFORM(IOS_FAMILY) || PLATFORM(GTK)
+#if PLATFORM(IOS_FAMILY) || USE(ATK)
     return true;
 #endif
 
     return m_parent->accessibilityIsIgnored();
 }
