diff a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableColumn.cpp b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableColumn.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableColumn.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityTableColumn.cpp
@@ -174,11 +174,11 @@
 bool AccessibilityTableColumn::computeAccessibilityIsIgnored() const
 {
     if (!m_parent)
         return true;
 
-#if PLATFORM(IOS_FAMILY) || PLATFORM(GTK)
+#if PLATFORM(IOS_FAMILY) || USE(ATK)
     return true;
 #endif
 
     return m_parent->accessibilityIsIgnored();
 }
