diff a/src/java.sql.rowset/share/classes/javax/sql/rowset/spi/SyncProviderException.java b/src/java.sql.rowset/share/classes/javax/sql/rowset/spi/SyncProviderException.java
--- a/src/java.sql.rowset/share/classes/javax/sql/rowset/spi/SyncProviderException.java
+++ b/src/java.sql.rowset/share/classes/javax/sql/rowset/spi/SyncProviderException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2006, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -68,10 +68,11 @@
     /**
      * The instance of <code>javax.sql.rowset.spi.SyncResolver</code> that
      * this <code>SyncProviderException</code> object will return when its
      * <code>getSyncResolver</code> method is called.
      */
+     @SuppressWarnings("serial") // Not statically typed as Serializable
      private SyncResolver syncResolver = null;
 
     /**
      * Creates a new <code>SyncProviderException</code> object without a detail message.
      */
