diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmSignature.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmSignature.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmSignature.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmSignature.cpp
@@ -26,10 +26,11 @@
 #include "config.h"
 #include "WasmSignature.h"
 
 #if ENABLE(WEBASSEMBLY)
 
+#include "WasmSignatureInlines.h"
 #include <wtf/FastMalloc.h>
 #include <wtf/HashFunctions.h>
 #include <wtf/PrintStream.h>
 #include <wtf/text/WTFString.h>
 
