diff a/src/hotspot/share/adlc/adlparse.cpp b/src/hotspot/share/adlc/adlparse.cpp
--- a/src/hotspot/share/adlc/adlparse.cpp
+++ b/src/hotspot/share/adlc/adlparse.cpp
@@ -121,10 +121,11 @@
   RegisterForm *regBlock = _AD.get_registers();
   if (regBlock == NULL) {
     parse_err(SEMERR, "Did not declare 'register' definitions");
   }
   regBlock->addSpillRegClass();
+  regBlock->addDynamicRegClass();
 
   // Done with parsing, check consistency.
 
   if (_preproc_depth != 0) {
     parse_err(SYNERR, "End of file inside #ifdef");
