diff a/modules/javafx.graphics/src/main/native-glass/ios/GlassViewDelegate.m b/modules/javafx.graphics/src/main/native-glass/ios/GlassViewDelegate.m
--- a/modules/javafx.graphics/src/main/native-glass/ios/GlassViewDelegate.m
+++ b/modules/javafx.graphics/src/main/native-glass/ios/GlassViewDelegate.m
@@ -29,10 +29,11 @@
 #import "com_sun_glass_events_MouseEvent.h"
 #import "com_sun_glass_events_KeyEvent.h"
 #import "com_sun_glass_events_TouchEvent.h"
 
 #import "GlassStatics.h"
+#import "GlassHelper.h"
 #import "GlassMacros.h"
 #import "GlassWindow.h"
 
 //#define VERBOSE_DND
 #ifdef VERBOSE_DND
@@ -380,11 +381,11 @@
         self.uiView.delegate = self;
         self.jView = (*env)->NewGlobalRef(env, jview); // java side
 
         // Ensure JNI stuff related to gesture processing is ready
         if (NULL == jGestureSupportClass) {
-            (*env)->FindClass(env, "com/sun/glass/ui/ios/IosGestureSupport");
+            [GlassHelper ClassForName:"com.sun.glass.ui.ios.IosGestureSupport" withEnv:env];
         }
         self.touches = NULL;
         self.lastTouchId = 0;
 
         mouseTouch = nil;
