diff a/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java b/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java
--- a/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java
+++ b/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java
@@ -516,10 +516,15 @@
         return URIBuilder.base(getWebUrl()).appendPath("/diffs")
                          .setQuery(Map.of("start_sha", base.hex()))
                          .build();
     }
 
+    @Override
+    public boolean isDraft() {
+        return json.get("work_in_progress").asBoolean();
+    }
+
 
     @Override
     public void setState(State state) {
         request.put("")
                .body("state_event", state == State.CLOSED ? "close" : "reopen")
