diff a/test/src/main/java/org/openjdk/skara/test/TestHost.java b/test/src/main/java/org/openjdk/skara/test/TestHost.java
--- a/test/src/main/java/org/openjdk/skara/test/TestHost.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestHost.java
@@ -56,11 +56,11 @@
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
-    static TestHost createNew(List<HostUserDetails> users) {
+    public static TestHost createNew(List<HostUserDetails> users) {
         var data = new HostData();
         data.users.addAll(users);
         var host = new TestHost(data, 0);
         return host;
     }
