diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/filetools.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/filetools.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/filetools.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/filetools.cpp
@@ -132,9 +132,9 @@
 }
 
 /* Swap the file separater character given with the new one in the file path. */
 U_CAPI void U_EXPORT2
 swapFileSepChar(char *filePath, const char oldFileSepChar, const char newFileSepChar) {
-    for (int32_t i = 0, length = uprv_strlen(filePath); i < length; i++) {
+    for (int32_t i = 0, length = static_cast<int32_t>(uprv_strlen(filePath)); i < length; i++) {
         filePath[i] = (filePath[i] == oldFileSepChar ) ? newFileSepChar : filePath[i];
     }
 }
