diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/package.cpp b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/package.cpp
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/package.cpp
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/tools/toolutil/package.cpp
@@ -608,11 +608,11 @@
             memcpy(pkgPrefix, s, prefixLength);
             pkgPrefix[prefixLength]=0;
             memcpy(prefix, s, ++prefixLength);  // include the /
         } else {
             // Use the package basename as prefix.
-            int32_t inPkgNameLength=strlen(inPkgName);
+            int32_t inPkgNameLength= static_cast<int32_t>(strlen(inPkgName));
             memcpy(prefix, inPkgName, inPkgNameLength);
             prefixLength=inPkgNameLength;
 
             if( (int32_t)strlen(s)>=(inPkgNameLength+2) &&
                 0==memcmp(s, inPkgName, inPkgNameLength) &&
@@ -1041,11 +1041,11 @@
 
         // reset this Item entry
         memset(items+idx, 0, sizeof(Item));
 
         // copy the item's name
-        items[idx].name=allocString(TRUE, strlen(name));
+        items[idx].name=allocString(TRUE, static_cast<int32_t>(strlen(name)));
         strcpy(items[idx].name, name);
         pathToTree(items[idx].name);
     } else {
         // same-name item found, replace it
         if(items[idx].isDataOwned) {
