diff a/test/langtools/tools/javac/resolve/BitWiseOperators.java b/test/langtools/tools/javac/resolve/BitWiseOperators.java
--- a/test/langtools/tools/javac/resolve/BitWiseOperators.java
+++ b/test/langtools/tools/javac/resolve/BitWiseOperators.java
@@ -25,14 +25,11 @@
  * @test
  * @bug 8082311 8129962
  * @summary Verify that bitwise operators don't allow to mix numeric and boolean operands.
  * @library ../lib
  * @modules jdk.compiler/com.sun.tools.javac.api
- *          jdk.compiler/com.sun.tools.javac.code
- *          jdk.compiler/com.sun.tools.javac.comp
- *          jdk.compiler/com.sun.tools.javac.main
- *          jdk.compiler/com.sun.tools.javac.tree
+ *          jdk.compiler/com.sun.tools.javac.file
  *          jdk.compiler/com.sun.tools.javac.util
  * @build combo.ComboTestHelper
  * @run main BitWiseOperators
  */
 
