diff a/src/hotspot/cpu/zero/entry_zero.hpp b/src/hotspot/cpu/zero/entry_zero.hpp
--- a/src/hotspot/cpu/zero/entry_zero.hpp
+++ b/src/hotspot/cpu/zero/entry_zero.hpp
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2020, Oracle and/or its affiliates. All rights reserved.
  * Copyright 2008, 2009, 2010 Red Hat, Inc.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
@@ -24,11 +24,11 @@
  */
 
 #ifndef CPU_ZERO_ENTRY_ZERO_HPP
 #define CPU_ZERO_ENTRY_ZERO_HPP
 
-#include "interpreter/cppInterpreter.hpp"
+#include "interpreter/zero/zeroInterpreter.hpp"
 
 class ZeroEntry {
  public:
   ZeroEntry() {
     ShouldNotCallThis();
@@ -67,11 +67,11 @@
   }
 
  private:
   static void maybe_deoptimize(int deoptimized_frames, TRAPS) {
     if (deoptimized_frames)
-      CppInterpreter::main_loop(deoptimized_frames - 1, THREAD);
+      ZeroInterpreter::main_loop(deoptimized_frames - 1, THREAD);
   }
 
  public:
   static ByteSize entry_point_offset() {
     return byte_offset_of(ZeroEntry, _entry_point);
