diff a/bots/submit/src/main/java/org/openjdk/skara/bots/submit/SubmitBot.java b/bots/submit/src/main/java/org/openjdk/skara/bots/submit/SubmitBot.java
--- a/bots/submit/src/main/java/org/openjdk/skara/bots/submit/SubmitBot.java
+++ b/bots/submit/src/main/java/org/openjdk/skara/bots/submit/SubmitBot.java
@@ -39,11 +39,11 @@
         this.updateCache = new PullRequestUpdateCache();
     }
 
     @Override
     public List<WorkItem> getPeriodicItems() {
-        return repository.getPullRequests().stream()
+        return repository.pullRequests().stream()
                          .filter(updateCache::needsUpdate)
                          .flatMap(pr -> executors.stream()
                                                  .map(executor -> new SubmitBotWorkItem(this, executor, pr)))
                          .collect(Collectors.toList());
     }
