diff a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/package-info.java b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/package-info.java
--- a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/package-info.java
+++ b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/package-info.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -29,11 +29,11 @@
  *  <p>Doclets no longer have to implement their own version of standard tags
  *  such as &#64;param and &#64;throws.  Individual taglets provide
  *  common processing, independent of the output format.
  *  Each doclet must have a taglet writer that takes a taglet
  *  as input and writes doclet-dependent output. The taglet itself will
- *  do the tag processing. For example, suppose we are outputing
+ *  do the tag processing. For example, suppose we are outputting
  *  &#64;throws tags. The taglet would:
  *  <ul>
  *      <li> Retrieve the list of throws tags to be documented.
  *      <li> Replace {&#64;inheritDoc} with the appropriate documentation.
  *      <li> Add throws documentation for exceptions that are declared in
