diff a/storage/build.gradle b/storage/build.gradle
--- a/storage/build.gradle
+++ b/storage/build.gradle
@@ -24,20 +24,23 @@
 module {
     name = 'org.openjdk.skara.storage'
     test {
         requires 'org.junit.jupiter.api'
         requires 'org.junit.jupiter.params'
+        requires 'org.openjdk.skara.test'
         opens 'org.openjdk.skara.storage' to 'org.junit.platform.commons'
     }
 }
 
 dependencies {
     implementation project(':network')
     implementation project(':host')
     implementation project(':forge')
     implementation project(':issuetracker')
     implementation project(':vcs')
+
+    testImplementation project(':test')
 }
 
 publishing {
     publications {
         storage(MavenPublication) {
