diff a/modules/javafx.web/src/main/native/Source/WebCore/page/Crypto.cpp b/modules/javafx.web/src/main/native/Source/WebCore/page/Crypto.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/Crypto.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/Crypto.cpp
@@ -35,11 +35,12 @@
 #include "SubtleCrypto.h"
 #include <JavaScriptCore/ArrayBufferView.h>
 #include <wtf/CryptographicallyRandomNumber.h>
 
 #if OS(DARWIN)
-#include "CommonCryptoUtilities.h"
+#include <CommonCrypto/CommonCryptor.h>
+#include <CommonCrypto/CommonRandom.h>
 #endif
 
 namespace WebCore {
 
 Crypto::Crypto(ScriptExecutionContext* context)
