diff a/modules/javafx.web/src/main/native/Source/WebCore/page/PageOverlayController.h b/modules/javafx.web/src/main/native/Source/WebCore/page/PageOverlayController.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/PageOverlayController.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/PageOverlayController.h
@@ -91,11 +91,11 @@
     void updateSettingsForLayer(GraphicsLayer&);
     void updateForceSynchronousScrollLayerPositionUpdates();
 
     // GraphicsLayerClient
     void notifyFlushRequired(const GraphicsLayer*) override;
-    void paintContents(const GraphicsLayer*, GraphicsContext&, OptionSet<GraphicsLayerPaintingPhase>, const FloatRect& clipRect, GraphicsLayerPaintBehavior) override;
+    void paintContents(const GraphicsLayer*, GraphicsContext&, const FloatRect& clipRect, GraphicsLayerPaintBehavior) override;
     float deviceScaleFactor() const override;
     bool shouldSkipLayerInDump(const GraphicsLayer*, LayerTreeAsTextBehavior) const override;
     void tiledBackingUsageChanged(const GraphicsLayer*, bool) override;
 
     Page& m_page;
