diff a/modules/javafx.web/src/main/native/Source/WebCore/page/SecurityPolicy.h b/modules/javafx.web/src/main/native/Source/WebCore/page/SecurityPolicy.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/SecurityPolicy.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/SecurityPolicy.h
@@ -50,10 +50,12 @@
     // referrer header should be omitted.
     WEBCORE_EXPORT static String generateReferrerHeader(ReferrerPolicy, const URL&, const String& referrer);
 
     static bool shouldInheritSecurityOriginFromOwner(const URL&);
 
+    static bool isBaseURLSchemeAllowed(const URL&);
+
     enum LocalLoadPolicy {
         AllowLocalLoadsForAll, // No restriction on local loads.
         AllowLocalLoadsForLocalAndSubstituteData,
         AllowLocalLoadsForLocalOnly,
     };
