diff a/modules/javafx.web/src/main/native/Source/WebCore/page/VisitedLinkStore.h b/modules/javafx.web/src/main/native/Source/WebCore/page/VisitedLinkStore.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/VisitedLinkStore.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/VisitedLinkStore.h
@@ -29,11 +29,11 @@
 #include <wtf/HashSet.h>
 #include <wtf/RefCounted.h>
 
 namespace WebCore {
 
-typedef uint64_t SharedStringHash;
+using SharedStringHash = uint32_t;
 class Page;
 
 class VisitedLinkStore : public RefCounted<VisitedLinkStore> {
 public:
     WEBCORE_EXPORT VisitedLinkStore();
