diff a/modules/javafx.web/src/main/native/Source/WebCore/html/shadow/ProgressShadowElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/shadow/ProgressShadowElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/shadow/ProgressShadowElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/shadow/ProgressShadowElement.cpp
@@ -84,9 +84,9 @@
 {
 }
 
 void ProgressValueElement::setWidthPercentage(double width)
 {
-    setInlineStyleProperty(CSSPropertyWidth, width, CSSPrimitiveValue::CSS_PERCENTAGE);
+    setInlineStyleProperty(CSSPropertyWidth, width, CSSUnitType::CSS_PERCENTAGE);
 }
 
 }
