diff a/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java b/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java
--- a/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java
+++ b/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java
@@ -83,8 +83,13 @@
 
         var cwd = Path.of("").toAbsolutePath();
         var repo = Repository.get(cwd).or(die("error: no repository found at " + cwd.toString())).get();
         var remote = arguments.at(0).orString("origin");
 
-        System.exit(pushAndTrack(remote, repo.currentBranch()));
+        var currentBranch = repo.currentBranch();
+        if (currentBranch.isEmpty()) {
+            System.err.println("error: the repository is in a detached HEAD state");
+            System.exit(1);
+        }
+        System.exit(pushAndTrack(remote, repo.currentBranch().get()));
     }
 }
