diff a/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraHost.java b/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraHost.java
--- a/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraHost.java
+++ b/issuetracker/src/main/java/org/openjdk/skara/issuetracker/jira/JiraHost.java
@@ -91,18 +91,23 @@
     public IssueProject project(String name) {
         return new JiraProject(this, request, name);
     }
 
     @Override
-    public HostUser user(String username) {
+    public Optional<HostUser> user(String username) {
         var data = request.get("user")
                           .param("username", username)
+                          .onError(r -> JSON.of())
                           .execute();
+        if (data.isNull()) {
+            return Optional.empty();
+        }
+
         var user = new HostUser(data.get("name").asString(),
                                 data.get("name").asString(),
                                 data.get("displayName").asString());
-        return user;
+        return Optional.of(user);
     }
 
     @Override
     public HostUser currentUser() {
         var data = request.get("myself").execute();
