diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortController.idl b/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortController.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortController.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortController.idl
@@ -25,11 +25,12 @@
 
 [
     Constructor,
     ConstructorCallWith=ScriptExecutionContext,
     Exposed=(Window,Worker),
-    ImplementationLacksVTable
+    ImplementationLacksVTable,
+    JSCustomMarkFunction
 ] interface AbortController {
     [SameObject] readonly attribute AbortSignal signal;
 
     void abort();
 };
