diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gst_private.h b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gst_private.h
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gst_private.h
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gst_private.h
@@ -92,10 +92,18 @@
 struct _GstPluginPrivate {
   GList *deps;    /* list of GstPluginDep structures */
   GstStructure *cache_data;
 };
 
+/* Needed by GstMeta (to access meta seq) and GstBuffer (create/free/iterate) */
+typedef struct _GstMetaItem GstMetaItem;
+struct _GstMetaItem {
+  GstMetaItem *next;
+  guint64 seq_num;
+  GstMeta meta;
+};
+
 /* FIXME: could rename all priv_gst_* functions to __gst_* now */
 G_GNUC_INTERNAL  gboolean priv_gst_plugin_loading_have_whitelist (void);
 
 G_GNUC_INTERNAL  guint32  priv_gst_plugin_loading_get_whitelist_hash (void);
 
@@ -134,10 +142,11 @@
 
 /* cleanup functions called from gst_deinit(). */
 G_GNUC_INTERNAL  void  _priv_gst_allocator_cleanup (void);
 G_GNUC_INTERNAL  void  _priv_gst_caps_features_cleanup (void);
 G_GNUC_INTERNAL  void  _priv_gst_caps_cleanup (void);
+G_GNUC_INTERNAL  void  _priv_gst_debug_cleanup (void);
 
 /* called from gst_task_cleanup_all(). */
 G_GNUC_INTERNAL  void  _priv_gst_element_cleanup (void);
 
 /* Private registry functions */
@@ -193,14 +202,14 @@
 G_GNUC_INTERNAL
 gchar *priv_gst_string_take_and_wrap (gchar * s);
 
 /* registry cache backends */
 G_GNUC_INTERNAL
-gboolean        priv_gst_registry_binary_read_cache (GstRegistry * registry, const char *location);
+gboolean    priv_gst_registry_binary_read_cache (GstRegistry * registry, const char *location);
 
 G_GNUC_INTERNAL
-gboolean        priv_gst_registry_binary_write_cache    (GstRegistry * registry, GList * plugins, const char *location);
+gboolean    priv_gst_registry_binary_write_cache  (GstRegistry * registry, GList * plugins, const char *location);
 
 
 G_GNUC_INTERNAL
 void      __gst_element_factory_add_static_pad_template (GstElementFactory    * elementfactory,
                                                          GstStaticPadTemplate * templ);
@@ -359,14 +368,14 @@
   GstObject       object;
 
   /*< private >*/
   GstPluginDesc desc;
 
-  gchar *   filename;
-  gchar *   basename;       /* base name (non-dir part) of plugin path */
+  gchar * filename;
+  gchar * basename;       /* base name (non-dir part) of plugin path */
 
-  GModule * module;     /* contains the module if plugin is loaded */
+  GModule * module;   /* contains the module if plugin is loaded */
 
   off_t         file_size;
   time_t        file_mtime;
   gboolean      registered;     /* TRUE when the registry has seen a filename
                                  * that matches the plugin's basename */
