diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstallocator.h b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstallocator.h
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstallocator.h
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstallocator.h
@@ -104,13 +104,13 @@
  * @mem_unmap: the implementation of the GstMemoryUnmapFunction
  * @mem_copy: the implementation of the GstMemoryCopyFunction
  * @mem_share: the implementation of the GstMemoryShareFunction
  * @mem_is_span: the implementation of the GstMemoryIsSpanFunction
  * @mem_map_full: the implementation of the GstMemoryMapFullFunction.
- *      Will be used instead of @mem_map if present. (Since 1.6)
+ *      Will be used instead of @mem_map if present. (Since: 1.6)
  * @mem_unmap_full: the implementation of the GstMemoryUnmapFullFunction.
- *      Will be used instead of @mem_unmap if present. (Since 1.6)
+ *      Will be used instead of @mem_unmap if present. (Since: 1.6)
  *
  * The #GstAllocator is used to create new memory.
  */
 struct _GstAllocator
 {
