diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstpluginfeature.h b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstpluginfeature.h
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstpluginfeature.h
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gstpluginfeature.h
@@ -151,11 +151,11 @@
                                                          guint             min_major,
                                                          guint             min_minor,
                                                          guint             min_micro);
 GST_API
 gint            gst_plugin_feature_rank_compare_func    (gconstpointer p1,
-                             gconstpointer p2);
+               gconstpointer p2);
 
 #ifdef G_DEFINE_AUTOPTR_CLEANUP_FUNC
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(GstPluginFeature, gst_object_unref)
 #endif
 
