diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gsttracerutils.c b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gsttracerutils.c
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gsttracerutils.c
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gstreamer/gst/gsttracerutils.c
@@ -25,11 +25,11 @@
  * to attach to them.
  *
  * The user can activate tracers by setting the environment variable GST_TRACE
  * to a ';' separated list of tracers.
  *
- * Note that instanciating tracers at runtime is possible but is not thread safe
+ * Note that instantiating tracers at runtime is possible but is not thread safe
  * and needs to be done before any pipeline state is set to PAUSED.
  */
 
 #define GST_USE_UNSTABLE_API
 
