diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/Cone.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/Cone.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/Cone.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/Cone.h
@@ -33,11 +33,12 @@
 
 namespace WebCore {
 
 // Cone gain is defined according to the OpenAL specification
 
-class ConeEffect {
+class ConeEffect final {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     ConeEffect();
 
     // Returns scalar gain for the given source/listener positions/orientations
     double gain(FloatPoint3D sourcePosition, FloatPoint3D sourceOrientation, FloatPoint3D listenerPosition);
