diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3BasicBlockInlines.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3BasicBlockInlines.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3BasicBlockInlines.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3BasicBlockInlines.h
@@ -39,18 +39,10 @@
     ValueType* result = procedure.add<ValueType>(arguments...);
     append(result);
     return result;
 }
 
-template<typename ValueType, typename... Arguments>
-ValueType* BasicBlock::appendNewNonTerminal(Procedure& procedure, Arguments... arguments)
-{
-    ValueType* result = procedure.add<ValueType>(arguments...);
-    appendNonTerminal(result);
-    return result;
-}
-
 template<typename ValueType, typename... Arguments>
 ValueType* BasicBlock::replaceLastWithNew(Procedure& procedure, Arguments... arguments)
 {
     ValueType* result = procedure.add<ValueType>(arguments...);
     replaceLast(procedure, result);
