diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapSpecial.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapSpecial.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapSpecial.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3StackmapSpecial.h
@@ -71,11 +71,11 @@
 
     // Appends the reps for the Inst's args, starting with numIgnoredArgs, to the given vector.
     Vector<ValueRep> repsImpl(
         Air::GenerationContext&, unsigned numIgnoredB3Args, unsigned numIgnoredAirArgs, Air::Inst&);
 
-    static bool isArgValidForValue(const Air::Arg&, Value*);
+    static bool isArgValidForType(const Air::Arg&, Type);
     static bool isArgValidForRep(Air::Code&, const Air::Arg&, const ValueRep&);
     static ValueRep repForArg(Air::Code&, const Air::Arg&);
 };
 
 } } // namespace JSC::B3
