diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCPeerConnection.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCPeerConnection.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCPeerConnection.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCPeerConnection.h
@@ -75,10 +75,11 @@
     , public ActiveDOMObject
 #if !RELEASE_LOG_DISABLED
     , private LoggerHelper
 #endif
 {
+    WTF_MAKE_ISO_ALLOCATED(RTCPeerConnection);
 public:
     static Ref<RTCPeerConnection> create(ScriptExecutionContext&);
     virtual ~RTCPeerConnection();
 
     using DataChannelInit = RTCDataChannelInit;
