diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCRtpReceiver.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCRtpReceiver.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCRtpReceiver.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/RTCRtpReceiver.cpp
@@ -33,13 +33,16 @@
 
 #if ENABLE(WEB_RTC)
 
 #include "PeerConnectionBackend.h"
 #include "RTCRtpCapabilities.h"
+#include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
+WTF_MAKE_ISO_ALLOCATED_IMPL(RTCRtpReceiver);
+
 RTCRtpReceiver::RTCRtpReceiver(PeerConnectionBackend& connection, Ref<MediaStreamTrack>&& track, std::unique_ptr<RTCRtpReceiverBackend>&& backend)
     : m_track(WTFMove(track))
     , m_backend(WTFMove(backend))
     , m_connection(makeWeakPtr(&connection))
 {
