diff a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/ConstantHelper.java b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/ConstantHelper.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/ConstantHelper.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/ConstantHelper.java
@@ -134,13 +134,13 @@
     private static final Map<Class<?>, ClassDesc> CARRIERS = Map.ofEntries(
             Map.entry(Byte.TYPE,                desc(Byte.TYPE)),
             Map.entry(Short.TYPE,               desc(Short.TYPE)),
             Map.entry(Character.TYPE,           desc(Character.TYPE)),
             Map.entry(Integer.TYPE,             desc(Integer.TYPE)),
-            Map.entry(Long.TYPE,                desc(Short.TYPE)),
-            Map.entry(Float.TYPE,               desc(Short.TYPE)),
-            Map.entry(Double.TYPE,              desc(Short.TYPE)),
+            Map.entry(Long.TYPE,                desc(Long.TYPE)),
+            Map.entry(Float.TYPE,               desc(Float.TYPE)),
+            Map.entry(Double.TYPE,              desc(Double.TYPE)),
             Map.entry(MemoryAddress.class,      desc(Long.TYPE))
     );
 
     ConstantHelper(String parentClassName, ClassDesc runtimeHelper, ClassDesc cString, String[] libraryNames) {
         this.cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
