diff a/test/jdk/java/io/Serializable/typeStringBackRef/TypeStringBackRef.java b/test/jdk/java/io/Serializable/typeStringBackRef/TypeStringBackRef.java
--- a/test/jdk/java/io/Serializable/typeStringBackRef/TypeStringBackRef.java
+++ b/test/jdk/java/io/Serializable/typeStringBackRef/TypeStringBackRef.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2001, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2001, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -28,10 +28,11 @@
  */
 
 import java.io.*;
 
 public class TypeStringBackRef implements Serializable {
+    private static final long serialVersionUID = 1L;
 
     String a, b, c, d, e, f, g;
 
     public static void main(String[] args) throws Exception {
         ByteArrayOutputStream bout = new ByteArrayOutputStream();
