diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/Markable.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/Markable.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/Markable.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/Markable.h
@@ -81,10 +81,11 @@
 // similar to WTF::HashTable, which uses two values of T as an empty value and a deleted value.
 // This class is intended to be used as a member of a class to compact the size of the class.
 // Otherwise, you should use Optional.
 template<typename T, typename Traits>
 class Markable {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     constexpr Markable()
         : m_value(Traits::emptyValue())
     { }
 
