diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/OptionSet.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/OptionSet.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/OptionSet.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/OptionSet.h
@@ -35,15 +35,17 @@
 
 // OptionSet is a class that represents a set of enumerators in a space-efficient manner. The enumerators
 // must be powers of two greater than 0. This class is useful as a replacement for passing a bitmask of
 // enumerators around.
 template<typename T> class OptionSet {
+    WTF_MAKE_FAST_ALLOCATED;
     static_assert(std::is_enum<T>::value, "T is not an enum type");
     typedef typename std::make_unsigned<typename std::underlying_type<T>::type>::type StorageType;
 
 public:
     template<typename StorageType> class Iterator {
+        WTF_MAKE_FAST_ALLOCATED;
     public:
         // Isolate the rightmost set bit.
         T operator*() const { return static_cast<T>(m_value & -m_value); }
 
         // Iterates from smallest to largest enum value by turning off the rightmost set bit.
