diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/RedBlackTree.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/RedBlackTree.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/RedBlackTree.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/RedBlackTree.h
@@ -39,12 +39,13 @@
 //   it from the tree. The tree will subsequently no longer have a
 //   reference to this node.
 // - The key type must implement operator< and ==.
 
 template<class NodeType, typename KeyType>
-class RedBlackTree {
+class RedBlackTree final {
     WTF_MAKE_NONCOPYABLE(RedBlackTree);
+    WTF_MAKE_FAST_ALLOCATED;
 private:
     enum Color {
         Red = 1,
         Black
     };
