diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.cpp b/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.cpp
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.cpp
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.cpp
@@ -71,9 +71,9 @@
 
 void TimingScope::scopeDidEnd()
 {
     const auto& data = state().addToTotal(m_name, MonotonicTime::now() - m_startTime);
     if (!(data.callCount % m_logIterationInterval))
-        WTFLogAlways("%s: %u calls, mean duration: %.6fms", m_name, data.callCount, data.meanDuration().milliseconds());
+        WTFLogAlways("%s: %u calls, mean duration: %.6fms, total duration: %.6fms", m_name, data.callCount, data.meanDuration().milliseconds(), data.totalDuration.milliseconds());
 }
 
 } // namespace WebCore
