diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/TimingScope.h
@@ -35,10 +35,11 @@
 // Do not leave in shipping code.
 //
 // Mean scope durations are printed using WTFLogAlways every Nth call, where N is specified by logIterationInterval.
 
 class TimingScope {
+    WTF_MAKE_FAST_ALLOCATED;
     WTF_MAKE_NONCOPYABLE(TimingScope);
 public:
     TimingScope(const char* name, unsigned logIterationInterval = 1)
         : m_startTime(MonotonicTime::now())
         , m_name(name)
