diff a/src/hotspot/share/gc/z/zReferenceProcessor.cpp b/src/hotspot/share/gc/z/zReferenceProcessor.cpp
--- a/src/hotspot/share/gc/z/zReferenceProcessor.cpp
+++ b/src/hotspot/share/gc/z/zReferenceProcessor.cpp
@@ -68,11 +68,11 @@
 static volatile oop* reference_referent_addr(oop reference) {
   return (volatile oop*)java_lang_ref_Reference::referent_addr_raw(reference);
 }
 
 static oop reference_referent(oop reference) {
-  return *reference_referent_addr(reference);
+  return Atomic::load(reference_referent_addr(reference));
 }
 
 static void reference_set_referent(oop reference, oop referent) {
   java_lang_ref_Reference::set_referent_raw(reference, referent);
 }
