diff a/modules/javafx.web/src/main/native/Source/WebCore/storage/StorageArea.h b/modules/javafx.web/src/main/native/Source/WebCore/storage/StorageArea.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/storage/StorageArea.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/storage/StorageArea.h
@@ -56,10 +56,8 @@
     virtual size_t memoryBytesUsedByCache() = 0;
 
     virtual void incrementAccessCount() { }
     virtual void decrementAccessCount() { }
     virtual void closeDatabaseIfIdle() { }
-
-    virtual const SecurityOriginData& securityOrigin() const = 0;
 };
 
 } // namespace WebCore
