diff a/make/modules/jdk.management.agent/Gensrc.gmk b/make/modules/jdk.management.agent/Gensrc.gmk
--- a/make/modules/jdk.management.agent/Gensrc.gmk
+++ b/make/modules/jdk.management.agent/Gensrc.gmk
@@ -1,7 +1,7 @@
 #
-# Copyright (c) 2011, 2015, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2011, 2020, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.  Oracle designates this
@@ -33,11 +33,5 @@
     SRC_DIRS := $(TOPDIR)/src/jdk.management.agent/share/classes/jdk/internal/agent/resources, \
     CLASS := ListResourceBundle, \
 ))
 
 TARGETS += $(COMPILE_PROPERTIES)
-
-################################################################################
-
-all: $(TARGETS)
-
-.PHONY: all
