diff a/modules/javafx.media/src/main/java/com/sun/media/jfxmedia/locator/Locator.java b/modules/javafx.media/src/main/java/com/sun/media/jfxmedia/locator/Locator.java
--- a/modules/javafx.media/src/main/java/com/sun/media/jfxmedia/locator/Locator.java
+++ b/modules/javafx.media/src/main/java/com/sun/media/jfxmedia/locator/Locator.java
@@ -415,11 +415,11 @@
                                     isMediaUnAvailable = true;
                                 }
                             }
 
                             // FIXME: get cache settings from server, honor them
-                        } else if (scheme.equals("file") || scheme.equals("jar") || scheme.equals("jrt")) {
+                        } else if (scheme.equals("file") || scheme.equals("jar") || scheme.equals("jrt") || (scheme.equals("resource")) ) {
                             InputStream stream = getInputStream(uri);
                             stream.close();
                             isConnected = true;
                             contentType = MediaUtils.filenameToContentType(uriString); // We need to provide at least something
                         }
