diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h
@@ -33,10 +33,11 @@
 namespace WebCore {
 
 class PaymentContact;
 
 class ApplePayShippingContactSelectedEvent final : public Event {
+    WTF_MAKE_ISO_ALLOCATED(ApplePayShippingContactSelectedEvent);
 public:
     static Ref<ApplePayShippingContactSelectedEvent> create(const AtomString& type, unsigned version, const PaymentContact& shippingContact)
     {
         return adoptRef(*new ApplePayShippingContactSelectedEvent(type, version, shippingContact));
     }
