diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h
@@ -160,11 +160,11 @@
     Color m_debugBorderColor;
     float m_debugBorderWidth;
 
     TextureMapperPlatformLayer* m_contentsLayer;
     FloatRect m_needsDisplayRect;
-    TextureMapperAnimations m_animations;
+    Nicosia::Animations m_animations;
     MonotonicTime m_animationStartTime;
 };
 
 } // namespace WebCore
 
