diff a/test/langtools/jdk/javadoc/doclet/testParamTaglet/TestParamTaglet.java b/test/langtools/jdk/javadoc/doclet/testParamTaglet/TestParamTaglet.java
--- a/test/langtools/jdk/javadoc/doclet/testParamTaglet/TestParamTaglet.java
+++ b/test/langtools/jdk/javadoc/doclet/testParamTaglet/TestParamTaglet.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -21,11 +21,11 @@
  * questions.
  */
 
 /*
  * @test
- * @bug      4802275 4967243 8026567
+ * @bug      4802275 4967243 8026567 8239804
  * @summary  Make sure param tags are still printed even though they do not
  *           match up with a real parameters.
  *           Make sure inheritDoc cannot be used in an invalid param tag.
  * @library  ../../lib
  * @modules jdk.javadoc/jdk.javadoc.internal.tool
@@ -49,15 +49,15 @@
                 "pkg");
         checkExit(Exit.ERROR);
 
         checkOutput("pkg/C.html", true,
                 //Regular param tags.
-                "<span class=\"paramLabel\">Parameters:</span></dt>\n"
+                "<dt>Parameters:</dt>\n"
                 + "<dd><code>param1</code> - testing 1 2 3.</dd>\n"
                 + "<dd><code>param2</code> - testing 1 2 3.",
                 //Param tags that don't match with any real parameters.
-                "<span class=\"paramLabel\">Parameters:</span></dt>\n"
+                "<dt>Parameters:</dt>\n"
                 + "<dd><code>p1</code> - testing 1 2 3.</dd>\n"
                 + "<dd><code>p2</code> - testing 1 2 3.",
                 //{@inherit} doc misuse does not cause doclet to throw exception.
                 // Param is printed with nothing inherited.
                 //XXX: in the future when Configuration is available during doc inheritence,
