diff a/modules/javafx.web/src/main/native/Source/WebCore/inspector/InspectorFrontendHost.h b/modules/javafx.web/src/main/native/Source/WebCore/inspector/InspectorFrontendHost.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/inspector/InspectorFrontendHost.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/inspector/InspectorFrontendHost.h
@@ -53,13 +53,14 @@
     WEBCORE_EXPORT void disconnectClient();
 
     WEBCORE_EXPORT void addSelfToGlobalObjectInWorld(DOMWrapperWorld&);
 
     void loaded();
-    void requestSetDockSide(const String&);
     void closeWindow();
     void reopen();
+    void reset();
+
     void bringToFront();
     void inspectedURLChanged(const String&);
 
     bool supportsShowCertificate() const;
     bool showCertificate(const String& serializedCertificate);
@@ -67,13 +68,17 @@
     void setZoomFactor(float);
     float zoomFactor();
 
     String userInterfaceLayoutDirection();
 
+    void requestSetDockSide(const String&);
+
     void setAttachedWindowHeight(unsigned);
     void setAttachedWindowWidth(unsigned);
 
+    void setSheetRect(float x, float y, unsigned width, unsigned height);
+
     void startWindowDrag();
     void moveWindowBy(float x, float y) const;
 
     bool isRemote() const;
     String localizedStringsURL();
