diff a/modules/javafx.web/src/main/native/Source/WebCore/inspector/RecordingSwizzleTypes.h b/modules/javafx.web/src/main/native/Source/WebCore/inspector/RecordingSwizzleTypes.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/inspector/RecordingSwizzleTypes.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/inspector/RecordingSwizzleTypes.h
@@ -47,8 +47,13 @@
     WebGLTexture = 15,
     WebGLShader = 16,
     WebGLProgram = 17,
     WebGLUniformLocation = 18,
     ImageBitmap = 19,
+    WebGLQuery = 20,
+    WebGLSampler = 21,
+    WebGLSync = 22,
+    WebGLTransformFeedback = 23,
+    WebGLVertexArrayObject = 24,
 };
 
 } // namespace WebCore
