diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMarqueeElement.idl b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMarqueeElement.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMarqueeElement.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLMarqueeElement.idl
@@ -15,11 +15,13 @@
  * along with this library; see the file COPYING.LIB.  If not, write to
  * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
  * Boston, MA 02110-1301, USA.
  */
 
-interface HTMLMarqueeElement : HTMLElement {
+[
+    ActiveDOMObject,
+] interface HTMLMarqueeElement : HTMLElement {
     void start();
     void stop();
     
     [CEReactions=NotNeeded, Reflect] attribute DOMString behavior;
     [CEReactions=NotNeeded, Reflect] attribute DOMString bgColor;
