diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/AnimationEvent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/AnimationEvent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/AnimationEvent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/AnimationEvent.cpp
@@ -26,18 +26,18 @@
 #include "config.h"
 #include "AnimationEvent.h"
 
 namespace WebCore {
 
-AnimationEvent::AnimationEvent(const AtomicString& type, const Init& initializer, IsTrusted isTrusted)
+AnimationEvent::AnimationEvent(const AtomString& type, const Init& initializer, IsTrusted isTrusted)
     : Event(type, initializer, isTrusted)
     , m_animationName(initializer.animationName)
     , m_elapsedTime(initializer.elapsedTime)
 {
 }
 
-AnimationEvent::AnimationEvent(const AtomicString& type, const String& animationName, double elapsedTime)
+AnimationEvent::AnimationEvent(const AtomString& type, const String& animationName, double elapsedTime)
     : Event(type, CanBubble::Yes, IsCancelable::No)
     , m_animationName(animationName)
     , m_elapsedTime(elapsedTime)
 {
 }
