diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/DOMRect.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/DOMRect.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/DOMRect.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/DOMRect.h
@@ -32,11 +32,10 @@
 namespace WebCore {
 
 class FloatQuad;
 
 class DOMRect : public DOMRectReadOnly {
-    WTF_MAKE_FAST_ALLOCATED;
 public:
     static Ref<DOMRect> create() { return adoptRef(*new DOMRect()); }
     static Ref<DOMRect> create(double x, double y, double width, double height) { return adoptRef(*new DOMRect(x, y, width, height)); }
     static Ref<DOMRect> create(FloatRect rect) { return adoptRef(*new DOMRect(rect.x(), rect.y(), rect.width(), rect.height())); }
     static Ref<DOMRect> create(IntRect rect) { return adoptRef(*new DOMRect(rect.x(), rect.y(), rect.width(), rect.height())); }
@@ -54,7 +53,8 @@
     {
     }
 
     DOMRect() = default;
 };
+static_assert(sizeof(DOMRect) == sizeof(DOMRectReadOnly), "");
 
 }
