diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/IdTargetObserver.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/IdTargetObserver.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/IdTargetObserver.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/IdTargetObserver.h
@@ -23,11 +23,11 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #pragma once
 
-#include <wtf/text/AtomicString.h>
+#include <wtf/text/AtomString.h>
 
 namespace WebCore {
 
 class IdTargetObserverRegistry;
 
@@ -35,13 +35,13 @@
 public:
     virtual ~IdTargetObserver();
     virtual void idTargetChanged() = 0;
 
 protected:
-    IdTargetObserver(IdTargetObserverRegistry&, const AtomicString& id);
+    IdTargetObserver(IdTargetObserverRegistry&, const AtomString& id);
 
 private:
     IdTargetObserverRegistry* m_registry;
-    AtomicString m_id;
+    AtomString m_id;
 };
 
 } // namespace WebCore
