diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/InlineClassicScript.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/InlineClassicScript.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/InlineClassicScript.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/InlineClassicScript.h
@@ -37,12 +37,12 @@
 
     bool isClassicScript() const final { return true; }
     bool isModuleScript() const final { return false; }
 
 private:
-    InlineClassicScript(const String& nonce, const String& crossOriginMode, const String& charset, const AtomicString& initiatorName, bool isInUserAgentShadowTree)
-        : ScriptElementCachedScriptFetcher(nonce, crossOriginMode, charset, initiatorName, isInUserAgentShadowTree)
+    InlineClassicScript(const String& nonce, const String& crossOriginMode, const String& charset, const AtomString& initiatorName, bool isInUserAgentShadowTree)
+        : ScriptElementCachedScriptFetcher(nonce, ReferrerPolicy::EmptyString, crossOriginMode, charset, initiatorName, isInUserAgentShadowTree)
     {
     }
 };
 
 }
