diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/MessagePort.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/MessagePort.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/MessagePort.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/MessagePort.h
@@ -42,13 +42,13 @@
 
 namespace WebCore {
 
 class Frame;
 
-class MessagePort final : public ActiveDOMObject, public EventTargetWithInlineData, public CanMakeWeakPtr<MessagePort> {
+class MessagePort final : public ActiveDOMObject, public EventTargetWithInlineData, public CanMakeWeakPtr<MessagePort, WeakPtrFactoryInitialization::Eager> {
     WTF_MAKE_NONCOPYABLE(MessagePort);
-    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_ISO_ALLOCATED(MessagePort);
 public:
     static Ref<MessagePort> create(ScriptExecutionContext&, const MessagePortIdentifier& local, const MessagePortIdentifier& remote);
     virtual ~MessagePort();
 
     ExceptionOr<void> postMessage(JSC::ExecState&, JSC::JSValue message, Vector<JSC::Strong<JSC::JSObject>>&&);
@@ -97,12 +97,12 @@
     void derefEventTarget() final { deref(); }
 
 private:
     explicit MessagePort(ScriptExecutionContext&, const MessagePortIdentifier& local, const MessagePortIdentifier& remote);
 
-    bool addEventListener(const AtomicString& eventType, Ref<EventListener>&&, const AddEventListenerOptions&) final;
-    bool removeEventListener(const AtomicString& eventType, EventListener&, const ListenerOptions&) final;
+    bool addEventListener(const AtomString& eventType, Ref<EventListener>&&, const AddEventListenerOptions&) final;
+    bool removeEventListener(const AtomString& eventType, EventListener&, const ListenerOptions&) final;
 
     void disentangle();
 
     void registerLocalActivity();
 
