diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/StaticNodeList.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/StaticNodeList.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/StaticNodeList.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/StaticNodeList.cpp
@@ -27,12 +27,17 @@
  */
 
 #include "config.h"
 #include "StaticNodeList.h"
 
+#include <wtf/IsoMallocInlines.h>
+
 namespace WebCore {
 
+WTF_MAKE_ISO_ALLOCATED_IMPL(StaticNodeList);
+WTF_MAKE_ISO_ALLOCATED_IMPL(StaticElementList);
+
 unsigned StaticNodeList::length() const
 {
     return m_nodes.size();
 }
 
