diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/TransitionEvent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/TransitionEvent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/TransitionEvent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/TransitionEvent.cpp
@@ -27,19 +27,19 @@
 #include "config.h"
 #include "TransitionEvent.h"
 
 namespace WebCore {
 
-TransitionEvent::TransitionEvent(const AtomicString& type, const String& propertyName, double elapsedTime, const String& pseudoElement)
+TransitionEvent::TransitionEvent(const AtomString& type, const String& propertyName, double elapsedTime, const String& pseudoElement)
     : Event(type, CanBubble::Yes, IsCancelable::Yes)
     , m_propertyName(propertyName)
     , m_elapsedTime(elapsedTime)
     , m_pseudoElement(pseudoElement)
 {
 }
 
-TransitionEvent::TransitionEvent(const AtomicString& type, const Init& initializer, IsTrusted isTrusted)
+TransitionEvent::TransitionEvent(const AtomString& type, const Init& initializer, IsTrusted isTrusted)
     : Event(type, initializer, isTrusted)
     , m_propertyName(initializer.propertyName)
     , m_elapsedTime(initializer.elapsedTime)
     , m_pseudoElement(initializer.pseudoElement)
 {
