diff a/src/java.base/share/classes/java/lang/System.java b/src/java.base/share/classes/java/lang/System.java
--- a/src/java.base/share/classes/java/lang/System.java
+++ b/src/java.base/share/classes/java/lang/System.java
@@ -1617,11 +1617,11 @@
          * Returns the {@code LoggerFinder} instance. There is one
          * single system-wide {@code LoggerFinder} instance in
          * the Java Runtime.  See the class specification of how the
          * {@link LoggerFinder LoggerFinder} implementation is located and
          * loaded.
-
+         *
          * @return the {@link LoggerFinder LoggerFinder} instance.
          * @throws SecurityException if a security manager is present and its
          *         {@code checkPermission} method doesn't allow the
          *         {@code RuntimePermission("loggerFinder")}.
          */
@@ -2274,13 +2274,8 @@
                 return StringCoding.getBytesUTF8NoRepl(s);
             }
             public void setCause(Throwable t, Throwable cause) {
                 t.setCause(cause);
             }
-
-            public void loadLibrary(Class<?> caller, String library) {
-                assert library.indexOf(java.io.File.separatorChar) < 0;
-                ClassLoader.loadLibrary(caller, library, false);
-            }
         });
     }
 }
