diff a/bots/submit/src/test/java/org/openjdk/skara/bots/submit/ShellExecutorTests.java b/bots/submit/src/test/java/org/openjdk/skara/bots/submit/ShellExecutorTests.java
--- a/bots/submit/src/test/java/org/openjdk/skara/bots/submit/ShellExecutorTests.java
+++ b/bots/submit/src/test/java/org/openjdk/skara/bots/submit/ShellExecutorTests.java
@@ -20,11 +20,11 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
 package org.openjdk.skara.bots.submit;
 
-import org.openjdk.skara.host.*;
+import org.openjdk.skara.forge.*;
 import org.openjdk.skara.test.TemporaryDirectory;
 import org.openjdk.skara.vcs.Hash;
 
 import org.junit.jupiter.api.Test;
 
