diff a/bots/tester/src/test/java/org/openjdk/skara/bots/tester/InMemoryHost.java b/bots/tester/src/test/java/org/openjdk/skara/bots/tester/InMemoryHost.java
--- a/bots/tester/src/test/java/org/openjdk/skara/bots/tester/InMemoryHost.java
+++ b/bots/tester/src/test/java/org/openjdk/skara/bots/tester/InMemoryHost.java
@@ -35,12 +35,12 @@
     public boolean isValid() {
         return false;
     }
 
     @Override
-    public HostedRepository repository(String name) {
-        return null;
+    public Optional<HostedRepository> repository(String name) {
+        return Optional.empty();
     }
 
     @Override
     public HostUser user(String username) {
         return null;
