diff a/test/jdk/tools/jextract/Test8240657.java b/test/jdk/tools/jextract/Test8240657.java
--- a/test/jdk/tools/jextract/Test8240657.java
+++ b/test/jdk/tools/jextract/Test8240657.java
@@ -26,10 +26,11 @@
 import static org.testng.Assert.assertNotNull;
 
 /*
  * @test
  * @modules jdk.incubator.jextract
+ * @library /test/lib
  * @build JextractToolRunner
  * @bug 8240657
  * @summary when Java keywords are used as identifiers in C header, jextract generates non-compilable java code
  * @run testng/othervm -Djdk.incubator.foreign.Foreign=permit Test8240657
  */
