diff a/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingConstraints.cpp b/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingConstraints.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingConstraints.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/scrolling/ScrollingConstraints.cpp
@@ -98,10 +98,27 @@
 {
     FloatSize offset = computeStickyOffset(constrainingRect);
     return m_layerPositionAtLastLayout + offset - m_stickyOffsetAtLastLayout;
 }
 
+TextStream& operator<<(TextStream& ts, ScrollPositioningBehavior behavior)
+{
+    switch (behavior) {
+    case ScrollPositioningBehavior::None: ts << "none"; break;
+    case ScrollPositioningBehavior::Stationary: ts << "stationary"; break;
+    case ScrollPositioningBehavior::Moves: ts << "moves"; break;
+    }
+    return ts;
+}
+
+TextStream& operator<<(TextStream& ts, const AbsolutePositionConstraints& constraints)
+{
+    ts.dumpProperty("layer-position-at-last-layout", constraints.layerPositionAtLastLayout());
+
+    return ts;
+}
+
 TextStream& operator<<(TextStream& ts, const FixedPositionViewportConstraints& constraints)
 {
     ts.dumpProperty("viewport-rect-at-last-layout", constraints.viewportRectAtLastLayout());
     ts.dumpProperty("layer-position-at-last-layout", constraints.layerPositionAtLastLayout());
 
