diff a/test/langtools/tools/javac/patterns/FlowAnalysisTest.java b/test/langtools/tools/javac/patterns/FlowAnalysisTest.java
--- a/test/langtools/tools/javac/patterns/FlowAnalysisTest.java
+++ b/test/langtools/tools/javac/patterns/FlowAnalysisTest.java
@@ -4,11 +4,11 @@
  * @compile/fail/ref=FlowAnalysisTest.out -XDrawDiagnostics FlowAnalysisTest.java
  */
 
 class FlowAnalysisTest {
     void foo(Object object) {
-        
+
         switch (object) {
             case Object o: break;
             case String s: break;
         }
         switch (object) {
