diff a/test/langtools/tools/javac/patterns/PrimitiveSelectorTest.java b/test/langtools/tools/javac/patterns/PrimitiveSelectorTest.java
--- a/test/langtools/tools/javac/patterns/PrimitiveSelectorTest.java
+++ b/test/langtools/tools/javac/patterns/PrimitiveSelectorTest.java
@@ -37,10 +37,10 @@
                 worked = true;
                 break;
             default:
                 worked = false;
                 break;
-		}
+        }
         if (!worked)
             throw new AssertionError("Unexpected control flow");
     }
 }
