diff a/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java b/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java
--- a/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java
+++ b/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java
@@ -195,11 +195,11 @@
             mgfDigestId = AlgorithmId.get(mgfSpec.getDigestAlgorithm());
         } catch (NoSuchAlgorithmException nase) {
             throw new IOException("AlgorithmId " +
                     mgfSpec.getDigestAlgorithm() + " impl not found");
         }
-        mgfDigestId.encode(tmp2);
+        mgfDigestId.derEncode(tmp2);
         tmp3 = new DerOutputStream();
         tmp3.write(DerValue.tag_Sequence, tmp2);
         tmp.write(DerValue.createTag(DerValue.TAG_CONTEXT, true, (byte)1),
                   tmp3);
 
