diff a/src/jdk.incubator.jpackage/windows/native/libjpackage/IconSwap.cpp b/src/jdk.incubator.jpackage/windows/native/libjpackage/IconSwap.cpp
--- a/src/jdk.incubator.jpackage/windows/native/libjpackage/IconSwap.cpp
+++ b/src/jdk.incubator.jpackage/windows/native/libjpackage/IconSwap.cpp
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2012, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2012, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -94,11 +94,11 @@
 }
 
 // Note: We do not check here that iconTarget is valid icon.
 // Java code will already do this for us.
 
-bool ChangeIcon(wstring iconTarget, wstring launcher) {
+bool ChangeIcon(HANDLE update, const wstring& iconTarget) {
     WORD language = MAKELANGID(LANG_ENGLISH, SUBLANG_DEFAULT);
 
     HANDLE icon = CreateFile(iconTarget.c_str(), GENERIC_READ, 0, NULL,
             OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
     if (icon == INVALID_HANDLE_VALUE) {
@@ -152,19 +152,10 @@
         lpgid->idEntries[i].dwBytesInRes = lpid->idEntries[i].dwBytesInRes;
         lpgid->idEntries[i].nID = i + 1;
     }
 
     // Store images in .EXE
-    HANDLE update = BeginUpdateResource(launcher.c_str(), FALSE);
-    if (update == NULL) {
-        free(lpid);
-        free(lpgid);
-        CloseHandle(icon);
-        PrintError();
-        return false;
-    }
-
     for (int i = 0; i < lpid->idCount; i++) {
         LPBYTE lpBuffer = (LPBYTE) malloc(lpid->idEntries[i].dwBytesInRes);
         SetFilePointer(icon, lpid->idEntries[i].dwImageOffset,
                 NULL, FILE_BEGIN);
         ReadFile(icon, lpBuffer, lpid->idEntries[i].dwBytesInRes,
@@ -193,12 +184,7 @@
         return false;
     }
 
     free(lpgid);
 
-    if (EndUpdateResource(update, FALSE) == FALSE) {
-        PrintError();
-        return false;
-    }
-
     return true;
 }
