diff a/src/jdk.incubator.jpackage/windows/native/libjpackage/ResourceEditor.h b/src/jdk.incubator.jpackage/windows/native/libjpackage/ResourceEditor.h
--- a/src/jdk.incubator.jpackage/windows/native/libjpackage/ResourceEditor.h
+++ b/src/jdk.incubator.jpackage/windows/native/libjpackage/ResourceEditor.h
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -33,26 +33,33 @@
 
 class ResourceEditor {
 public:
     class FileLock {
     public:
-        FileLock(const std::wstring& binaryPath);
+        explicit FileLock(const std::wstring& binaryPath);
+        explicit FileLock(HANDLE h);
         ~FileLock();
 
         HANDLE get() const {
             return h;
         }
 
         void discard(bool v = true) {
             theDiscard = v;
         }
 
+        FileLock& ownHandle(bool v) {
+            theOwnHandle = v;
+            return *this;
+        }
+
     private:
         FileLock(const FileLock&);
         FileLock& operator=(const FileLock&);
     private:
         HANDLE h;
+        bool theOwnHandle;
         bool theDiscard;
     };
 
 public:
     ResourceEditor();
