diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedInlines.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedInlines.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedInlines.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedInlines.h
@@ -55,11 +55,11 @@
     return false;
 }
 
 template<typename T>
 template<typename FilterFunctionType>
-bool GCIncomingRefCounted<T>::filterIncomingReferences(FilterFunctionType& filterFunction)
+bool GCIncomingRefCounted<T>::filterIncomingReferences(FilterFunctionType&& filterFunction)
 {
     const bool verbose = false;
 
     if (verbose)
         dataLog("Filtering incoming references.\n");
