diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedSet.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedSet.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedSet.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/GCIncomingRefCountedSet.h
@@ -39,18 +39,15 @@
     void lastChanceToFinalize();
 
     // Returns true if the native object is new to this set.
     bool addReference(JSCell*, T*);
 
-    void sweep();
+    void sweep(VM&);
 
     size_t size() const { return m_bytes; };
 
 private:
-    static bool removeAll(JSCell*);
-    static bool removeDead(JSCell*);
-
     Vector<T*> m_vector;
     size_t m_bytes;
 };
 
 } // namespace JSC
