diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/LocalAllocator.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/LocalAllocator.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/LocalAllocator.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/LocalAllocator.h
@@ -23,10 +23,11 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #pragma once
 
+#include "AllocationFailureMode.h"
 #include "FreeList.h"
 #include "MarkedBlock.h"
 #include <wtf/Noncopyable.h>
 
 namespace JSC {
