diff a/test/langtools/jdk/javadoc/doclet/testJavaFX/pkg1/C.java b/test/langtools/jdk/javadoc/doclet/testJavaFX/pkg1/C.java
--- a/test/langtools/jdk/javadoc/doclet/testJavaFX/pkg1/C.java
+++ b/test/langtools/jdk/javadoc/doclet/testJavaFX/pkg1/C.java
@@ -24,17 +24,15 @@
 package pkg1;
 
 public class C {
 
     /**
-     * @propertySetter Property
      * @propertyDescription PropertyDescription
      */
     public void CC() {}
 
     /**
-     * @propertyGetter Property
      *
      */
     public void B() {}
 
     /**
