diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPU.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPU.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPU.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPU.h
@@ -26,11 +26,10 @@
 #pragma once
 
 #if ENABLE(WEBGPU)
 
 #include "JSDOMPromiseDeferred.h"
-#include <wtf/Optional.h>
 
 namespace WebCore {
 
 class WebGPUAdapter;
 
