diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUProgrammablePassEncoder.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUProgrammablePassEncoder.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUProgrammablePassEncoder.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUProgrammablePassEncoder.idl
@@ -20,18 +20,18 @@
  * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
  * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
-// https://github.com/gpuweb/gpuweb/blob/master/design/sketch.webidl
+// https://gpuweb.github.io/gpuweb
 
 typedef unsigned long u32;
 
 [
     Conditional=WEBGPU,
     EnabledAtRuntime=WebGPU,
+    InterfaceName=GPUProgrammablePassEncoder,
     SkipVTableValidation
 ] interface WebGPUProgrammablePassEncoder {
-    WebGPUCommandBuffer endPass();
+    void endPass();
     void setBindGroup(u32 index, WebGPUBindGroup bindGroup/*, optional sequence<u32> dynamicOffsets*/);
-    void setPipeline(WebGPURenderPipeline pipeline); // FIXME: Support WebGPUComputePipelines.
 };
