diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUTextureView.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUTextureView.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUTextureView.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WebGPUTextureView.cpp
@@ -28,19 +28,28 @@
 
 #if ENABLE(WEBGPU)
 
 namespace WebCore {
 
-Ref<WebGPUTextureView> WebGPUTextureView::create(Ref<GPUTexture>&& view)
+Ref<WebGPUTextureView> WebGPUTextureView::create(RefPtr<GPUTexture>&& view)
 {
     return adoptRef(*new WebGPUTextureView(WTFMove(view)));
 }
 
-WebGPUTextureView::WebGPUTextureView(Ref<GPUTexture>&& view)
+WebGPUTextureView::WebGPUTextureView(RefPtr<GPUTexture>&& view)
     : m_texture(WTFMove(view))
 {
 }
 
+void WebGPUTextureView::destroy()
+{
+    if (!m_texture)
+        return;
+
+    m_texture->destroy();
+    m_texture = nullptr;
+}
+
 } // namespace WebCore
 
 #endif // ENABLE(WEBGPU)
 
