diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/DisplayRefreshMonitor.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/DisplayRefreshMonitor.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/DisplayRefreshMonitor.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/DisplayRefreshMonitor.h
@@ -28,28 +28,31 @@
 #if USE(REQUEST_ANIMATION_FRAME_DISPLAY_MONITOR)
 
 #include "PlatformScreen.h"
 #include <wtf/HashSet.h>
 #include <wtf/Lock.h>
-#include <wtf/RefCounted.h>
+#include <wtf/ThreadSafeRefCounted.h>
 #include <wtf/RefPtr.h>
 
 namespace WebCore {
 
 class DisplayAnimationClient;
 class DisplayRefreshMonitorClient;
 
-class DisplayRefreshMonitor : public RefCounted<DisplayRefreshMonitor> {
+class DisplayRefreshMonitor : public ThreadSafeRefCounted<DisplayRefreshMonitor> {
 public:
     static RefPtr<DisplayRefreshMonitor> create(DisplayRefreshMonitorClient&);
     WEBCORE_EXPORT virtual ~DisplayRefreshMonitor();
 
     virtual void displayLinkFired() { }
 
     // Return true if callback request was scheduled, false if it couldn't be
     // (e.g., hardware refresh is not available)
     virtual bool requestRefreshCallback() = 0;
+
+    virtual void stop() { }
+
     void windowScreenDidChange(PlatformDisplayID);
 
     bool hasClients() const { return m_clients.size(); }
     void addClient(DisplayRefreshMonitorClient&);
     bool removeClient(DisplayRefreshMonitorClient&);
