diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FourCC.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FourCC.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FourCC.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FourCC.h
@@ -28,10 +28,11 @@
 #include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
 struct FourCC {
+    FourCC() = default;
     FourCC(uint32_t value) : value(value) { }
 
     template<std::size_t N>
     constexpr FourCC(const char (&data)[N])
     {
@@ -43,9 +44,9 @@
     WEBCORE_EXPORT static Optional<FourCC> fromString(const String&);
 
     bool operator==(const FourCC& other) const { return value == other.value; }
     bool operator!=(const FourCC& other) const { return value != other.value; }
 
-    uint32_t value;
+    uint32_t value { 0 };
 };
 
 }
