diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/RemoteVideoSample.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/RemoteVideoSample.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/RemoteVideoSample.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/RemoteVideoSample.h
@@ -39,10 +39,12 @@
 namespace WebCore {
 
 class RemoteVideoSample {
 public:
     RemoteVideoSample() = default;
+    RemoteVideoSample(RemoteVideoSample&&) = default;
+    RemoteVideoSample& operator=(RemoteVideoSample&&) = default;
     ~RemoteVideoSample() = default;
 
 #if HAVE(IOSURFACE)
     WEBCORE_EXPORT static std::unique_ptr<RemoteVideoSample> create(MediaSample&&);
     WEBCORE_EXPORT IOSurfaceRef surface();
