diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/text/PlatformLocale.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/text/PlatformLocale.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/text/PlatformLocale.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/text/PlatformLocale.h
@@ -38,11 +38,11 @@
 
 class Locale {
     WTF_MAKE_NONCOPYABLE(Locale); WTF_MAKE_FAST_ALLOCATED;
 
 public:
-    static std::unique_ptr<Locale> create(const AtomicString& localeIdentifier);
+    static std::unique_ptr<Locale> create(const AtomString& localeIdentifier);
     static std::unique_ptr<Locale> createDefault();
 
     // Converts the specified number string to another number string localized
     // for this Locale locale. The input string must conform to HTML
     // floating-point numbers, and is not empty.
