diff a/test/hotspot/jtreg/compiler/ciReplay/CiReplayBase.java b/test/hotspot/jtreg/compiler/ciReplay/CiReplayBase.java
--- a/test/hotspot/jtreg/compiler/ciReplay/CiReplayBase.java
+++ b/test/hotspot/jtreg/compiler/ciReplay/CiReplayBase.java
@@ -144,11 +144,11 @@
             if (needCoreDump) {
                 crashOut = ProcessTools.executeProcess(getTestJavaCommandlineWithPrefix(
                         RUN_SHELL_NO_LIMIT, options.toArray(new String[0])));
             } else {
                 crashOut = ProcessTools.executeProcess(ProcessTools.createJavaProcessBuilder(true,
-                        options.toArray(new String[0])));
+                        options));
             }
             crashOutputString = crashOut.getOutput();
             Asserts.assertNotEquals(crashOut.getExitValue(), 0, "Crash JVM exits gracefully");
             Files.write(Paths.get("crash.out"), crashOutputString.getBytes(),
                     StandardOpenOption.CREATE, StandardOpenOption.WRITE,
