diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanDoublePropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanDoublePropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanDoublePropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanDoublePropertyBuilder.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2011, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2011, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -58,10 +58,16 @@
  */
 public final class ReadOnlyJavaBeanDoublePropertyBuilder {
 
     private final ReadOnlyJavaBeanPropertyBuilderHelper helper = new ReadOnlyJavaBeanPropertyBuilderHelper();
 
+    /**
+     * @deprecated This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
+     */
+    @Deprecated(since="14", forRemoval=true)
+    public ReadOnlyJavaBeanDoublePropertyBuilder() {}
+
     /**
      * Create a new instance of {@code ReadOnlyJavaBeanDoublePropertyBuilder}
      *
      * @return the new {@code ReadOnlyJavaBeanDoublePropertyBuilder}
      */
