diff a/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java b/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
@@ -799,10 +799,26 @@
         try (var p = capture(cmd)) {
             await(p);
         }
     }
 
+    @Override
+    public void abortMerge() throws IOException {
+        try (var p = capture("hg", "merge", "--abort")) {
+            await(p);
+        }
+
+        try (var p = capture("hg", "status", "--unknown", "--no-status")) {
+            var res = await(p);
+            for (var path : res.stdout()) {
+                if (path.toString().endsWith(".orig")) {
+                    Files.delete(root().resolve(path));
+                }
+            }
+        }
+    }
+
     @Override
     public void addRemote(String name, String path) throws IOException {
         setPaths(name, path, path);
     }
 
