diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.h
@@ -64,13 +64,13 @@
             }
         } while (changed);
     }
 
     template<typename HasFunctor, typename AddFunctor>
-    void closeStartingWithLocal(VirtualRegister reg, const HasFunctor& has, const AddFunctor& add)
+    void closeStartingWithLocal(Operand op, const HasFunctor& has, const AddFunctor& add)
     {
-        Availability availability = m_locals.operand(reg);
+        Availability availability = m_locals.operand(op);
         if (!availability.hasNode())
             return;
 
         if (!add(availability.node()))
             return;
