diff a/test/jdk/java/util/concurrent/locks/ReentrantLock/LockOncePerThreadLoops.java b/test/jdk/java/util/concurrent/locks/ReentrantLock/LockOncePerThreadLoops.java
--- a/test/jdk/java/util/concurrent/locks/ReentrantLock/LockOncePerThreadLoops.java
+++ b/test/jdk/java/util/concurrent/locks/ReentrantLock/LockOncePerThreadLoops.java
@@ -92,11 +92,11 @@
                 pool.execute(this);
             barrier.await();
             barrier.await();
             if (print) {
                 long time = timer.getTime();
-                double secs = (double)(time) / 1000000000.0;
+                double secs = (double)time / 1000000000.0;
                 System.out.println("\t " + secs + "s run time");
             }
 
             int r = result;
             if (r == 0) // avoid overoptimization
