diff a/src/java.desktop/windows/native/libjsound/PLATFORM_API_WinOS_DirectSound.cpp b/src/java.desktop/windows/native/libjsound/PLATFORM_API_WinOS_DirectSound.cpp
--- a/src/java.desktop/windows/native/libjsound/PLATFORM_API_WinOS_DirectSound.cpp
+++ b/src/java.desktop/windows/native/libjsound/PLATFORM_API_WinOS_DirectSound.cpp
@@ -355,11 +355,11 @@
     BOOL underrun;
 
 } DS_Info;
 
 
-LPSTR TranslateDSError(HRESULT hr) {
+LPCSTR TranslateDSError(HRESULT hr) {
     switch(hr) {
         case DSERR_ALLOCATED:
             return "DSERR_ALLOCATED";
 
         case DSERR_CONTROLUNAVAIL:
