diff a/test/jdk/tools/jimage/JImageTest.java b/test/jdk/tools/jimage/JImageTest.java
--- a/test/jdk/tools/jimage/JImageTest.java
+++ b/test/jdk/tools/jimage/JImageTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2015, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -43,20 +43,21 @@
 
 /*
  * jimage testing.
  * @test
  * @summary Test jimage tool
+ * @bug 8222100
  * @library ../lib
  * @modules java.base/jdk.internal.jimage
  *          jdk.jdeps/com.sun.tools.classfile
  *          jdk.jlink/jdk.tools.jmod
  *          jdk.jlink/jdk.tools.jimage
  *          jdk.jlink/jdk.tools.jlink.internal
  *          jdk.compiler
  * @run build JImageTest
  * @run build tests.*
- * @run main/othervm -verbose:gc -Xmx1g JImageTest
+ * @run main/othervm/timeout=360 -verbose:gc -Xmx1g JImageTest
 */
 public class JImageTest {
 
     public static void main(String[] args) throws Exception {
         List<String> bootClasses = new ArrayList<>();
