diff a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/SolvesTracker.java b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/SolvesTracker.java
--- a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/SolvesTracker.java
+++ b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/SolvesTracker.java
@@ -44,14 +44,15 @@
         return String.format(solvesMarker, issue.id(), "");
     }
 
     static List<Issue> currentSolved(HostUser botUser, List<Comment> comments) {
         var solvesActions = comments.stream()
-                                     .filter(comment -> comment.author().equals(botUser))
-                                     .map(comment -> markerPattern.matcher(comment.body()))
-                                     .filter(Matcher::find)
-                                     .collect(Collectors.toList());
+                                    .filter(comment -> comment.author().equals(botUser))
+                                    .flatMap(comment -> comment.body().lines())
+                                    .map(markerPattern::matcher)
+                                    .filter(Matcher::find)
+                                    .collect(Collectors.toList());
         var current = new LinkedHashMap<String, Issue>();
         for (var action : solvesActions) {
             var key = action.group(1);
             if (action.group(2).equals("")) {
                 current.remove(key);
