diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediarecorder/MediaRecorder.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediarecorder/MediaRecorder.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediarecorder/MediaRecorder.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediarecorder/MediaRecorder.h
@@ -45,10 +45,11 @@
     , public RefCounted<MediaRecorder>
     , public EventTargetWithInlineData
     , public CanMakeWeakPtr<MediaRecorder>
     , private MediaStream::Observer
     , private MediaStreamTrackPrivate::Observer {
+    WTF_MAKE_ISO_ALLOCATED(MediaRecorder);
 public:
     enum class RecordingState { Inactive, Recording, Paused };
 
     struct Options {
         String mimeType;
