diff a/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h b/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h
--- a/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h
+++ b/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h
@@ -79,10 +79,35 @@
     int isSrcOpaque;
     int isDstOpaque;
     float extraAlpha;
 };
 
+struct TxtFrameOpRescaleUniforms {
+    vector_float4 color;
+    float extraAlpha;
+
+    int isSrcOpaque;
+    int isNonPremult;
+
+    vector_float4 normScaleFactors;
+    vector_float4 normOffsets;
+};
+
+struct TxtFrameOpConvolveUniforms {
+    float extraAlpha;
+    vector_float4 imgEdge;
+    int kernelSize;
+    int isEdgeZeroFill;
+};
+
+struct TxtFrameOpLookupUniforms {
+    float extraAlpha;
+    vector_float4 offset;
+    int isUseSrcAlpha;
+    int isNonPremult;
+};
+
 struct AnchorData
 {
     vector_float3 xParams;
     vector_float3 yParams;
 };
