diff a/modules/javafx.web/src/main/native/Source/WebCore/mapfile-vers b/modules/javafx.web/src/main/native/Source/WebCore/mapfile-vers
--- a/modules/javafx.web/src/main/native/Source/WebCore/mapfile-vers
+++ b/modules/javafx.web/src/main/native/Source/WebCore/mapfile-vers
@@ -1979,15 +1979,15 @@
                Java_com_sun_webkit_graphics_WCRenderQueue_twkRelease;
                Java_com_sun_webkit_network_SocketStreamHandle_twkDidClose;
                Java_com_sun_webkit_network_SocketStreamHandle_twkDidFail;
                Java_com_sun_webkit_network_SocketStreamHandle_twkDidOpen;
                Java_com_sun_webkit_network_SocketStreamHandle_twkDidReceiveData;
-               Java_com_sun_webkit_network_URLLoader_twkDidFail;
-               Java_com_sun_webkit_network_URLLoader_twkDidFinishLoading;
-               Java_com_sun_webkit_network_URLLoader_twkDidReceiveData;
-               Java_com_sun_webkit_network_URLLoader_twkDidReceiveResponse;
-               Java_com_sun_webkit_network_URLLoader_twkDidSendData;
-               Java_com_sun_webkit_network_URLLoader_twkWillSendRequest;
+               Java_com_sun_webkit_network_URLLoaderBase_twkDidFail;
+               Java_com_sun_webkit_network_URLLoaderBase_twkDidFinishLoading;
+               Java_com_sun_webkit_network_URLLoaderBase_twkDidReceiveData;
+               Java_com_sun_webkit_network_URLLoaderBase_twkDidReceiveResponse;
+               Java_com_sun_webkit_network_URLLoaderBase_twkDidSendData;
+               Java_com_sun_webkit_network_URLLoaderBase_twkWillSendRequest;
                kJSClassDefinitionEmpty;
         local:
                 *;
 };
