diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGDecoratedEnumeration.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGDecoratedEnumeration.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGDecoratedEnumeration.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/properties/SVGDecoratedEnumeration.h
@@ -37,11 +37,11 @@
 
 public:
     static auto create(const EnumType& value)
     {
         static_assert(std::is_integral<DecorationType>::value, "DecorationType form enum should be integral.");
-        return makeUnique<SVGDecoratedEnumeration>(value);
+        return adoptRef(*new SVGDecoratedEnumeration(value));
     }
 
 private:
     bool setValue(const DecorationType& value) override
     {
@@ -56,11 +56,11 @@
         if (Base::value() > SVGIDLEnumLimits<EnumType>::highestExposedEnumValue())
             return m_outOfRangeEnumValue;
         return Base::value();
     }
 
-    std::unique_ptr<SVGDecoratedProperty<DecorationType>> clone() override
+    Ref<SVGDecoratedProperty<DecorationType>> clone() override
     {
         return create(m_value);
     }
 
     static const DecorationType m_outOfRangeEnumValue = 0;
