diff a/test/langtools/jdk/javadoc/tool/sourceOnly/Test.java b/test/langtools/jdk/javadoc/tool/sourceOnly/Test.java
--- a/test/langtools/jdk/javadoc/tool/sourceOnly/Test.java
+++ b/test/langtools/jdk/javadoc/tool/sourceOnly/Test.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2002, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2002, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -23,11 +23,10 @@
 
 /*
  * @test
  * @bug 4548768
  * @summary Javadoc in JDK 1.4 uses classpath and not just source dir
- * @author gafter
  * @modules jdk.javadoc/jdk.javadoc.internal.tool
  * @compile p/SourceOnly.java p/NonSource.jasm
  * @run main Test
  */
 
