diff a/json/src/test/java/org/openjdk/skara/json/JSONParserTests.java b/json/src/test/java/org/openjdk/skara/json/JSONParserTests.java
--- a/json/src/test/java/org/openjdk/skara/json/JSONParserTests.java
+++ b/json/src/test/java/org/openjdk/skara/json/JSONParserTests.java
@@ -609,6 +609,14 @@
         assertFalse(json.get("foo").isBoolean());
         assertFalse(json.get("foo").isString());
         assertFalse(json.get("foo").isArray());
         assertFalse(json.get("foo").isNull());
     }
+
+    @Test
+    public void testJSONObjectWithNullField() {
+        var json = JSON.parse("{ \"foo\": null }");
+
+        assertNotNull(json.get("foo"));
+        assertTrue(json.get("foo").isNull());
+    }
 }
