diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionConstructor.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionConstructor.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionConstructor.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionConstructor.h
@@ -57,7 +57,8 @@
 
 private:
     GeneratorFunctionConstructor(VM&, Structure*);
     void finishCreation(VM&, GeneratorFunctionPrototype*);
 };
+STATIC_ASSERT_ISO_SUBSPACE_SHARABLE(GeneratorFunctionConstructor, InternalFunction);
 
 } // namespace JSC
