diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSCInlines.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSCInlines.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSCInlines.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/JSCInlines.h
@@ -48,8 +48,9 @@
 #include "JSObjectInlines.h"
 #include "JSProxy.h"
 #include "JSString.h"
 #include "Operations.h"
 #include "SlotVisitorInlines.h"
+#include "StrongInlines.h"
 #include "StructureInlines.h"
 #include "ThrowScope.h"
 #include "WeakGCMapInlines.h"
