diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/NullGetterFunction.h
@@ -35,11 +35,11 @@
 
     static NullGetterFunction* create(VM& vm, Structure* structure)
     {
         // Since NullGetterFunction is per JSGlobalObject, we use put-without-transition in InternalFunction::finishCreation.
         NullGetterFunction* function = new (NotNull, allocateCell< NullGetterFunction>(vm.heap))  NullGetterFunction(vm, structure);
-        function->finishCreation(vm, String(), NameVisibility::Visible, NameAdditionMode::WithoutStructureTransition);
+        function->finishCreation(vm, String(), NameAdditionMode::WithoutStructureTransition);
         return function;
     }
 
     DECLARE_EXPORT_INFO;
 
@@ -49,7 +49,8 @@
     }
 
 private:
     NullGetterFunction(VM&, Structure*);
 };
+STATIC_ASSERT_ISO_SUBSPACE_SHARABLE(NullGetterFunction, InternalFunction);
 
 } // namespace JSC
