diff a/test/jdk/sun/tools/jstack/DeadlockDetectionTest.java b/test/jdk/sun/tools/jstack/DeadlockDetectionTest.java
--- a/test/jdk/sun/tools/jstack/DeadlockDetectionTest.java
+++ b/test/jdk/sun/tools/jstack/DeadlockDetectionTest.java
@@ -48,10 +48,11 @@
     private static LingeredAppWithDeadlock theApp = null;
     private static ProcessBuilder processBuilder = new ProcessBuilder();
 
     private static OutputAnalyzer jstack(String... toolArgs) throws Exception {
         JDKToolLauncher launcher = JDKToolLauncher.createUsingTestJDK("jstack");
+        launcher.addVMArgs(Utils.getFilteredTestJavaOpts("-XX:+UsePerfData"));
         launcher.addVMArg("-XX:+UsePerfData");
         if (toolArgs != null) {
             for (String toolArg : toolArgs) {
                 launcher.addToolArg(toolArg);
             }
