diff a/test/jdk/javax/swing/JTextPane/TestJTextPaneHTMLRendering.java b/test/jdk/javax/swing/JTextPane/TestJTextPaneHTMLRendering.java
--- a/test/jdk/javax/swing/JTextPane/TestJTextPaneHTMLRendering.java
+++ b/test/jdk/javax/swing/JTextPane/TestJTextPaneHTMLRendering.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -21,11 +21,11 @@
  * questions.
  */
 
 /*
  * @test
- * @bug 8224475
+ * @bug 8224475 8238985
  * @summary Verify that JTextPane renders images properly for HTML text
  * @run main/manual TestJTextPaneHTMLRendering
  */
 
 import java.awt.BorderLayout;
@@ -58,11 +58,11 @@
     private static URL urlArrow;
 
     private static volatile boolean testResult = false;
     private static volatile CountDownLatch countDownLatch;
     private static final String INSTRUCTIONS = "INSTRUCTIONS:\n\n" +
-            "Verify that the JTextPane is filled with green arrow images.\n" +
+            "Verify that the JTextPane is filled with blue arrow images.\n" +
             "There should be 200 images (10 rows of 20 images each).\n" +
             "This test will run for 10 iterations and the current iteration\n" +
             "is being displayed at top of JTextPane. JTextpane will be\n" +
             "repainted each time  and should have same output\n"+
             "If yes, Press Pass, Otherwise, Press Fail.\n";
