diff a/application/l10n/org.openjdk.jmc.console.persistence.ja/META-INF/MANIFEST.MF b/application/l10n/org.openjdk.jmc.console.persistence.ja/META-INF/MANIFEST.MF
--- a/application/l10n/org.openjdk.jmc.console.persistence.ja/META-INF/MANIFEST.MF
+++ b/application/l10n/org.openjdk.jmc.console.persistence.ja/META-INF/MANIFEST.MF
@@ -1,10 +1,10 @@
 Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-RequiredExecutionEnvironment: JavaSE-1.8
 Bundle-Name: Japanese Language Fragment
 Bundle-SymbolicName: org.openjdk.jmc.console.persistence.ja
-Bundle-Version: 7.1.0.qualifier
+Bundle-Version: 8.0.0.qualifier
 Bundle-Vendor: Oracle Corporation
-Fragment-Host: org.openjdk.jmc.console.persistence;bundle-version="7.1.0"
+Fragment-Host: org.openjdk.jmc.console.persistence;bundle-version="8.0.0"
 Bundle-Localization: plugin
 Automatic-Module-Name: org.openjdk.jmc.console.persistence.ja
