diff a/test/jdk/tools/jextract/test8244938/Test8244938.java b/test/jdk/tools/jextract/test8244938/Test8244938.java
--- a/test/jdk/tools/jextract/test8244938/Test8244938.java
+++ b/test/jdk/tools/jextract/test8244938/Test8244938.java
@@ -37,12 +37,11 @@
 public class Test8244938 {
     @Test
     public void testNestedStructReturn() {
          var seg = func();
          assertEquals(seg.byteSize(), Point.sizeof());
-         var addr = seg.address();
-         assertEquals(Point.k$get(addr), 44);
-         var point2dAddr = Point.point2d$addr(addr);
+         assertEquals(Point.k$get(seg), 44);
+         var point2dAddr = Point.point2d$addr(seg);
          assertEquals(Point2D.i$get(point2dAddr), 567);
          assertEquals(Point2D.j$get(point2dAddr), 33);
     }
 }
