diff a/src/java.desktop/macosx/native/libawt_lwawt/awt/PrinterSurfaceData.m b/src/java.desktop/macosx/native/libawt_lwawt/awt/PrinterSurfaceData.m
--- a/src/java.desktop/macosx/native/libawt_lwawt/awt/PrinterSurfaceData.m
+++ b/src/java.desktop/macosx/native/libawt_lwawt/awt/PrinterSurfaceData.m
@@ -23,10 +23,11 @@
  * questions.
  */
 
 
 #import "PrinterSurfaceData.h"
+#import "jni_util.h"
 #import <JavaNativeFoundation/JavaNativeFoundation.h>
 
 
 //#define DEBUG 1
 #if defined DEBUG
@@ -85,10 +86,15 @@
 JNF_COCOA_ENTER(env);
 
 PRINT("Java_sun_lwawt_macosx_CPrinterSurfaceData_initOps")
 
     PrintSDOps *psdo = (PrintSDOps*)SurfaceData_InitOps(env, jthis, sizeof(PrintSDOps));
+    if (psdo == NULL) {
+        JNU_ThrowOutOfMemoryError(env, "Initialization of SurfaceData failed.");
+        return;
+    }
+
     psdo->nsRef            = (NSGraphicsContext*)jlong_to_ptr(nsRef);
     psdo->width            = width;
     psdo->height        = height;
 
     QuartzSDOps *qsdo = (QuartzSDOps*)psdo;
