diff a/test/jdk/jdk/jfr/tool/TestMetadata.java b/test/jdk/jdk/jfr/tool/TestMetadata.java
--- a/test/jdk/jdk/jfr/tool/TestMetadata.java
+++ b/test/jdk/jdk/jfr/tool/TestMetadata.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -24,10 +24,12 @@
  */
 
 package jdk.jfr.tool;
 
 import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Set;
 
 import jdk.jfr.EventType;
 import jdk.jfr.consumer.RecordingFile;
 import jdk.test.lib.process.OutputAnalyzer;
 
@@ -57,7 +59,20 @@
                 String name = t.getName();
                 name = name.substring(name.lastIndexOf(".") + 1);
                 output.shouldContain(name);
             }
         }
+        Set<String> annotations = new HashSet<>();
+        int lineNumber = 1;
+        for (String line : output.asLines()) {
+            if (line.startsWith("@")) {
+                if (annotations.contains(line)) {
+                    throw new Exception("Line " + lineNumber + ":" +  line + " repeats annotation");
+                }
+                annotations.add(line);
+            } else {
+                annotations.clear();
+            }
+            lineNumber++;
+        }
     }
 }
