diff a/test/jaxp/javax/xml/jaxp/unittest/transform/ProcessXSLT.java b/test/jaxp/javax/xml/jaxp/unittest/transform/ProcessXSLT.java
--- a/test/jaxp/javax/xml/jaxp/unittest/transform/ProcessXSLT.java
+++ b/test/jaxp/javax/xml/jaxp/unittest/transform/ProcessXSLT.java
@@ -143,19 +143,16 @@
    *
    * @param argv Input parameters from command line
    */
     public static void main(String argv[]) {
 
-        // Runtime.getRuntime().traceMethodCalls(false); // turns Java tracing off
         boolean doStackDumpOnError = false;
         boolean doDiag = false;
         boolean setQuietMode = false;
         String msg = null;
         boolean isSecureProcessing = false;
 
-        // Runtime.getRuntime().traceMethodCalls(false);
-        // Runtime.getRuntime().traceInstructions(false);
         /**
          * The default diagnostic writer...
          */
         java.io.PrintWriter diagnosticsWriter = new PrintWriter(System.err, true);
         java.io.PrintWriter dumpWriter = diagnosticsWriter;
