diff a/test/jdk/javax/swing/JRootPane/4670486/bug4670486.java b/test/jdk/javax/swing/JRootPane/4670486/bug4670486.java
--- a/test/jdk/javax/swing/JRootPane/4670486/bug4670486.java
+++ b/test/jdk/javax/swing/JRootPane/4670486/bug4670486.java
@@ -37,10 +37,11 @@
  */
 public class bug4670486 {
 
     public static volatile boolean actionExpected = false;
     public static volatile boolean actionRecieved = false;
+    public static JFrame frame;
 
     private static JMenuBar createMenuBar() {
         JMenuBar menubar = new JMenuBar();
 
         // Control with unique menu
@@ -99,46 +100,50 @@
             throw new RuntimeException("Action has not been received");
         }
     }
 
     public static void main(String[] args) throws Throwable {
-        Robot robot = new Robot();
-        robot.setAutoDelay(250);
-
-        UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
-
-        SwingUtilities.invokeAndWait(new Runnable() {
-
-            @Override
-            public void run() {
-                JFrame frame = new JFrame("Test");
-                frame.setContentPane(createPanel(frame));
-                frame.setJMenuBar(createMenuBar());
-                frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
-                frame.pack();
-                frame.setVisible(true);
-            }
-        });
-
-        robot.waitForIdle();
-
-        // Change the default button to
-        // force a call to BasicRootPaneUI.updateDefaultButtonBindings()
-        Util.hitKeys(robot, KeyEvent.VK_TAB);
-
-        // If the bug exists, then as soon as the menu appears,
-        // the VK_ENTER, VK_DOWN, VK_UP and VK_ESC will have no
-        // effect.
-        Util.hitMnemonics(robot, KeyEvent.VK_U);
-        Util.hitKeys(robot, KeyEvent.VK_ENTER);
-        robot.waitForIdle();
-
-        checkAction();
-
-        Util.hitMnemonics(robot, KeyEvent.VK_U);
-        Util.hitKeys(robot, KeyEvent.VK_DOWN);
-        Util.hitKeys(robot, KeyEvent.VK_ENTER);
-        robot.waitForIdle();
-
-        checkAction();
+        try {
+            Robot robot = new Robot();
+            robot.setAutoDelay(250);
+
+            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
+
+            SwingUtilities.invokeAndWait(new Runnable() {
+
+                @Override
+                public void run() {
+                    frame = new JFrame("Test");
+                    frame.setContentPane(createPanel(frame));
+                    frame.setJMenuBar(createMenuBar());
+                    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+                    frame.pack();
+                    frame.setVisible(true);
+                }
+            });
+
+            robot.waitForIdle();
+
+            // Change the default button to
+            // force a call to BasicRootPaneUI.updateDefaultButtonBindings()
+            Util.hitKeys(robot, KeyEvent.VK_TAB);
+
+            // If the bug exists, then as soon as the menu appears,
+            // the VK_ENTER, VK_DOWN, VK_UP and VK_ESC will have no
+            // effect.
+            Util.hitMnemonics(robot, KeyEvent.VK_U);
+            Util.hitKeys(robot, KeyEvent.VK_ENTER);
+            robot.waitForIdle();
+
+            checkAction();
+
+            Util.hitMnemonics(robot, KeyEvent.VK_U);
+            Util.hitKeys(robot, KeyEvent.VK_DOWN);
+            Util.hitKeys(robot, KeyEvent.VK_ENTER);
+            robot.waitForIdle();
+
+            checkAction();
+        } finally {
+            if (frame != null) SwingUtilities.invokeAndWait(() -> frame.dispose());
+        }
     }
 }
