diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/videooverlay.c b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/videooverlay.c
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/videooverlay.c
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/videooverlay.c
@@ -523,20 +523,20 @@
  * This helper shall be used by classes implementing the #GstVideoOverlay
  * interface that want the render rectangle to be controllable using
  * properties. This helper will install "render-rectangle" property into the
  * class.
  *
- * Since 1.14
+ * Since: 1.14
  */
 void
 gst_video_overlay_install_properties (GObjectClass * oclass, gint last_prop_id)
 {
   g_object_class_install_property (oclass, last_prop_id + PROP_RENDER_RECTANGLE,
       gst_param_spec_array ("render-rectangle", "Render Rectangle",
           "The render rectangle ('<x, y, width, height>')",
           g_param_spec_int ("rect-value", "Rectangle Value",
-              "One of x, y, width or height value.", -1, G_MAXINT, -1,
+              "One of x, y, width or height value.", G_MININT, G_MAXINT, -1,
               G_PARAM_WRITABLE | G_PARAM_STATIC_STRINGS),
           G_PARAM_WRITABLE | G_PARAM_STATIC_STRINGS));
 }
 
 /**
@@ -551,11 +551,11 @@
  * properties. This helper will parse and set the render rectangle calling
  * gst_video_overlay_set_render_rectangle().
  *
  * Returns: %TRUE if the @property_id matches the GstVideoOverlay property
  *
- * Since 1.14
+ * Since: 1.14
  */
 gboolean
 gst_video_overlay_set_property (GObject * object, gint last_prop_id,
     guint property_id, const GValue * value)
 {
