diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/video.h b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/video.h
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/video.h
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/gst-libs/gst/video/video.h
@@ -165,10 +165,13 @@
 
 G_END_DECLS
 
 #include <gst/video/colorbalancechannel.h>
 #include <gst/video/colorbalance.h>
+#ifndef GSTREAMER_LITE
+#include <gst/video/gstvideoaggregator.h>
+#endif // GSTREAMER_LITE
 #include <gst/video/gstvideodecoder.h>
 #include <gst/video/gstvideoencoder.h>
 #include <gst/video/gstvideofilter.h>
 #include <gst/video/gstvideometa.h>
 #include <gst/video/gstvideopool.h>
@@ -184,8 +187,9 @@
 #include <gst/video/video-overlay-composition.h>
 #include <gst/video/videooverlay.h>
 #include <gst/video/gstvideotimecode.h>
 #ifndef GSTREAMER_LITE
 #include <gst/video/gstvideoaffinetransformationmeta.h>
+#include <gst/video/video-anc.h>
 #endif // GSTREAMER_LITE
 
 #endif /* __GST_VIDEO_H__ */
