diff a/test/hotspot/jtreg/runtime/ErrorHandling/VeryEarlyAssertTest.java b/test/hotspot/jtreg/runtime/ErrorHandling/VeryEarlyAssertTest.java
--- a/test/hotspot/jtreg/runtime/ErrorHandling/VeryEarlyAssertTest.java
+++ b/test/hotspot/jtreg/runtime/ErrorHandling/VeryEarlyAssertTest.java
@@ -29,10 +29,11 @@
  * @summary No hs-err file if fatal error is raised during dynamic initialization.
  * @library /test/lib
  * @modules java.base/jdk.internal.misc
  * @requires (vm.debug == true)
  * @requires os.family == "linux"
+ * @run driver VeryEarlyAssertTest
  */
 
 import java.io.BufferedReader;
 import java.io.File;
 import java.io.FileInputStream;
