diff a/jcheck/src/test/java/org/openjdk/skara/jcheck/AuthorCheckTests.java b/jcheck/src/test/java/org/openjdk/skara/jcheck/AuthorCheckTests.java
--- a/jcheck/src/test/java/org/openjdk/skara/jcheck/AuthorCheckTests.java
+++ b/jcheck/src/test/java/org/openjdk/skara/jcheck/AuthorCheckTests.java
@@ -28,11 +28,11 @@
 
 import org.junit.jupiter.api.Test;
 import static org.junit.jupiter.api.Assertions.*;
 
 import java.util.*;
-import java.time.Instant;
+import java.time.ZonedDateTime;
 import java.io.IOException;
 
 class AuthorCheckTests {
     private static final JCheckConfiguration conf = JCheckConfiguration.parse(List.of(
         "[general]",
@@ -43,13 +43,13 @@
 
     private static Commit commit(Author author) {
         var committer = new Author("Foo", "foo@bar.org");
         var hash = new Hash("0123456789012345678901234567890123456789");
         var parents = List.of(new Hash("12345789012345789012345678901234567890"));
-        var timestamp = Instant.now();
+        var date = ZonedDateTime.now();
         var message = List.of("Initial commit");
-        var metadata = new CommitMetadata(hash, parents, author, committer, timestamp, message);
+        var metadata = new CommitMetadata(hash, parents, author, committer, date, message);
         return new Commit(metadata, List.of());
     }
 
     private static CommitMessage message(Commit c) {
         return CommitMessageParsers.v1.parse(c);
