diff a/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java b/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
@@ -1329,6 +1329,21 @@
         if (!Files.exists(hgrc)) {
             Files.createFile(hgrc);
         }
         Files.write(hgrc, lines, StandardOpenOption.WRITE, StandardOpenOption.APPEND);
     }
+
+    @Override
+    public String range(Hash h) {
+        return h.hex();
+    }
+
+    @Override
+    public String rangeInclusive(Hash from, Hash to) {
+        return from.hex() + ":" + to.hex();
+    }
+
+    @Override
+    public String rangeExclusive(Hash from, Hash to) {
+        return from.hex() + ":" + to.hex() + "-" + from.hex();
+    }
 }
