diff a/modules/javafx.web/src/main/native/Source/WebCore/workers/service/FetchEvent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/workers/service/FetchEvent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/workers/service/FetchEvent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/workers/service/FetchEvent.cpp
@@ -38,11 +38,11 @@
     FetchEvent::Init init;
     init.request = FetchRequest::create(context, { }, FetchHeaders::create(FetchHeaders::Guard::Immutable, { }), { }, { }, { });
     return FetchEvent::create("fetch", WTFMove(init), Event::IsTrusted::Yes);
 }
 
-FetchEvent::FetchEvent(const AtomicString& type, Init&& initializer, IsTrusted isTrusted)
+FetchEvent::FetchEvent(const AtomString& type, Init&& initializer, IsTrusted isTrusted)
     : ExtendableEvent(type, initializer, isTrusted)
     , m_request(initializer.request.releaseNonNull())
     , m_clientId(WTFMove(initializer.clientId))
     , m_reservedClientId(WTFMove(initializer.reservedClientId))
     , m_targetClientId(WTFMove(initializer.targetClientId))
