diff a/src/hotspot/share/gc/shared/genOopClosures.hpp b/src/hotspot/share/gc/shared/genOopClosures.hpp
--- a/src/hotspot/share/gc/shared/genOopClosures.hpp
+++ b/src/hotspot/share/gc/shared/genOopClosures.hpp
@@ -55,13 +55,10 @@
 
   // Derived classes that modify oops so that they might be old-to-young
   // pointers must call the method below.
   template <class T> void do_barrier(T* p);
 
-  // Version for use by closures that may be called in parallel code.
-  template <class T> void par_do_barrier(T* p);
-
  public:
   OopsInGenClosure() : OopIterateClosure(NULL),
     _orig_gen(NULL), _gen(NULL), _gen_boundary(NULL), _rs(NULL) {};
 
   OopsInGenClosure(Generation* gen);
