diff a/forge/src/main/java/org/openjdk/skara/forge/github/GitHubHost.java b/forge/src/main/java/org/openjdk/skara/forge/github/GitHubHost.java
--- a/forge/src/main/java/org/openjdk/skara/forge/github/GitHubHost.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/github/GitHubHost.java
@@ -194,11 +194,12 @@
 
         var name = details.get("name").asString();
         if (name == null) {
             name = login;
         }
-        return new HostUser(id, login, name);
+        var email = details.get("email").asString();
+        return new HostUser(id, login, name, email);
     }
 
     @Override
     public HostUser currentUser() {
         if (currentUser == null) {
