diff a/test/langtools/tools/javac/valhalla/lworld-values/GenericsAndValues3.java b/test/langtools/tools/javac/valhalla/lworld-values/GenericsAndValues3.java
--- a/test/langtools/tools/javac/valhalla/lworld-values/GenericsAndValues3.java
+++ b/test/langtools/tools/javac/valhalla/lworld-values/GenericsAndValues3.java
@@ -25,11 +25,11 @@
 
 /*
  * @test
  * @bug 8222784
  * @summary Check generics and values interplay
- * @compile -XDallowGenericsOverValues GenericsAndValues3.java
+ * @compile GenericsAndValues3.java
  */
 
 import java.util.stream.IntStream;
 
 @__inline__ class StreamBug {
