diff a/modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp b/modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp
@@ -649,10 +649,14 @@
     return is_data_set;
 }
 
 static gboolean dnd_destroy_drag_widget_callback(gpointer) {
     if (drag_widget) {
+        if (gtk_widget_has_grab(drag_widget)) {
+            gtk_grab_remove(drag_widget);
+        }
+
         gtk_widget_destroy(drag_widget);
         drag_widget = NULL;
     }
 
     return FALSE;
@@ -661,14 +665,10 @@
 static void dnd_end_callback(GtkWidget *widget,
                              GdkDragContext *context,
                              gpointer user_data)
 {
     if (drag_widget) {
-        if (gtk_widget_has_grab(drag_widget)) {
-            gtk_grab_remove(drag_widget);
-        }
-
         GdkDragAction action = gdk_drag_context_get_selected_action(context);
         dnd_set_performed_action(translate_gdk_action_to_glass(action));
     }
     gdk_threads_add_idle((GSourceFunc) dnd_destroy_drag_widget_callback, NULL);
 }
