diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotFactory.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotFactory.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotFactory.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotFactory.java
@@ -68,10 +68,11 @@
                                   .collect(Collectors.toSet());
         var readyComments = specific.get("ready").get("comments").stream()
                                     .map(JSONValue::asObject)
                                     .collect(Collectors.toMap(obj -> obj.get("user").asString(),
                                                               obj -> Pattern.compile(obj.get("pattern").asString())));
+        var integratorId = specific.get("integrator").asString();
 
         // Collect configuration applicable to all instances of a specific notifier
         var notifierFactories = NotifierFactory.getNotifierFactories();
         notifierFactories.forEach(notifierFactory -> log.info("Available notifier: " + notifierFactory.name()));
         var notifierConfiguration = new HashMap<String, JSONObject>();
@@ -133,10 +134,11 @@
                                .prStateStorageBuilder(prStateStorageBuilder)
                                .updaters(updaters)
                                .prUpdaters(prUpdaters)
                                .readyLabels(readyLabels)
                                .readyComments(readyComments)
+                               .integratorId(integratorId)
                                .build();
             ret.add(bot);
         }
 
         return ret;
