diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUError.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUError.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUError.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/gpu/GPUError.cpp
@@ -30,20 +30,20 @@
 
 #include "GPUErrorFilter.h"
 
 namespace WebCore {
 
-Optional<GPUError> createError(GPUErrorFilter filter, const String& message)
+GPUError createError(GPUErrorFilter filter, const String& message)
 {
     switch (filter) {
     case GPUErrorFilter::OutOfMemory:
         return GPUError(RefPtr<GPUOutOfMemoryError>(GPUOutOfMemoryError::create()));
     case GPUErrorFilter::Validation:
         return GPUError(RefPtr<GPUValidationError>(GPUValidationError::create(message)));
-    default:
+    case GPUErrorFilter::None:
         ASSERT_NOT_REACHED();
-        return WTF::nullopt;
+        return GPUError(RefPtr<GPUValidationError>(GPUValidationError::create("Bad error!")));
     }
 }
 
 } // namespace WebCore
 
