diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/FontPlatformDataJava.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/FontPlatformDataJava.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/FontPlatformDataJava.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/java/FontPlatformDataJava.cpp
@@ -63,11 +63,11 @@
     , m_size(size)
 {
 }
 
 std::unique_ptr<FontPlatformData> FontPlatformData::create(
-        const FontDescription& fontDescription, const AtomicString& family)
+        const FontDescription& fontDescription, const AtomString& family)
 {
     RefPtr<RQRef> wcFont = getJavaFont(
             family,
             fontDescription.computedSize(),
             isItalic(fontDescription.italic()),
