diff a/src/hotspot/share/ci/ciValueKlass.hpp b/src/hotspot/share/ci/ciValueKlass.hpp
--- a/src/hotspot/share/ci/ciValueKlass.hpp
+++ b/src/hotspot/share/ci/ciValueKlass.hpp
@@ -50,11 +50,11 @@
   ciValueKlass(Klass* h_k) : ciInstanceKlass(h_k), _declared_nonstatic_fields(NULL) {
     assert(is_final(), "ValueKlass must be final");
   };
 
   ciValueKlass(ciSymbol* name, jobject loader, jobject protection_domain) :
-    ciInstanceKlass(name, loader, protection_domain, T_VALUETYPE) {}
+    ciInstanceKlass(name, loader, protection_domain, T_INLINE_TYPE) {}
 
   int compute_nonstatic_fields();
   const char* type_string() { return "ciValueKlass"; }
 
 public:
