diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/DatabaseThread.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/DatabaseThread.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/DatabaseThread.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/DatabaseThread.cpp
@@ -37,11 +37,11 @@
 #include <wtf/AutodrainedPool.h>
 
 namespace WebCore {
 
 DatabaseThread::DatabaseThread()
-    : m_transactionCoordinator(std::make_unique<SQLTransactionCoordinator>())
+    : m_transactionCoordinator(makeUnique<SQLTransactionCoordinator>())
 {
     m_selfRef = this;
 }
 
 DatabaseThread::~DatabaseThread()
