diff a/test/jdk/TEST.groups b/test/jdk/TEST.groups
--- a/test/jdk/TEST.groups
+++ b/test/jdk/TEST.groups
@@ -74,11 +74,12 @@
     :build \
     :jdk_rmi \
     :jdk_beans \
     :jdk_imageio \
     :jdk_sound \
-    :jdk_client_sanity
+    :jdk_client_sanity \
+    :jdk_jfr_tier3
 
 ###############################################################################
 #
 # Other test definitions; generally smaller granularity than tiers
 #
@@ -284,10 +285,13 @@
     :svc_tools
 
 jdk_jfr = \
     jdk/jfr
 
+jdk_jfr_tier3 = \
+    jdk/jfr/event/metadata/TestLookForUntestedEvents.java
+
 #
 # Catch-all for other areas with a small number of tests
 #
 jdk_other = \
     java/sql \
