diff a/src/hotspot/share/interpreter/interpreterRuntime.hpp b/src/hotspot/share/interpreter/interpreterRuntime.hpp
--- a/src/hotspot/share/interpreter/interpreterRuntime.hpp
+++ b/src/hotspot/share/interpreter/interpreterRuntime.hpp
@@ -25,11 +25,10 @@
 #ifndef SHARE_INTERPRETER_INTERPRETERRUNTIME_HPP
 #define SHARE_INTERPRETER_INTERPRETERRUNTIME_HPP
 
 #include "interpreter/bytecode.hpp"
 #include "interpreter/linkResolver.hpp"
-#include "memory/universe.hpp"
 #include "oops/method.hpp"
 #include "runtime/frame.hpp"
 #include "runtime/signature.hpp"
 #include "runtime/thread.hpp"
 #include "utilities/macros.hpp"
