diff a/test/langtools/jdk/javadoc/doclet/testSerializedFormWithClassFile/TestSerializedFormWithClassFile.java b/test/langtools/jdk/javadoc/doclet/testSerializedFormWithClassFile/TestSerializedFormWithClassFile.java
--- a/test/langtools/jdk/javadoc/doclet/testSerializedFormWithClassFile/TestSerializedFormWithClassFile.java
+++ b/test/langtools/jdk/javadoc/doclet/testSerializedFormWithClassFile/TestSerializedFormWithClassFile.java
@@ -74,11 +74,11 @@
 
         checkOutput("serialized-form.html", true,
                 """
                     <div class="member-signature"><span class="modifiers">public</span>&nbsp;<span c\
                     lass="return-type">void</span>&nbsp;<span class="member-name">readObject</span>&\
-                    #8203;(<span class="parameters">java.io.ObjectInputStream&nbsp;arg0)</span>
+                    #8203;<span class="parameters">(java.io.ObjectInputStream&nbsp;arg0)</span>
                                     throws <span class="exceptions">java.lang.ClassNotFoundException,
                     java.io.IOException</span></div>
                     """);
     }
 
