diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java
@@ -110,11 +110,12 @@
         var parent = topCommentForHash(reviewComment.hash());
         if (reviewComment.parent().isPresent()) {
             var parentId = getStableMessageId(getMessageId(reviewComment.parent().get()));
             var last = Stream.concat(existing.stream(), generated.stream())
                              .filter(email -> (email.hasHeader("References") && email.headerValue("References").contains(parentId)) ||
-                                     (getStableMessageId(email.id()).equals(parentId)))
+                                     (getStableMessageId(email.id()).equals(parentId)) ||
+                                     (email.hasHeader("PR-Collapsed-IDs") && email.headerValue("PR-Collapsed-IDs").contains(parentId)))
                              .max(Comparator.comparingInt(email -> Integer.parseInt(email.headerValue("PR-Sequence"))));
 
             if (last.isEmpty()) {
                 throw new RuntimeException("Failed to find parent");
             } else {
diff a/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java b/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java
--- a/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java
+++ b/bots/mlbridge/src/test/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotTests.java
@@ -365,11 +365,11 @@
             TestBotRunner.runPeriodicItems(mlBot);
             listServer.processIncoming();
             listServer.processIncoming();
 
             // Make two file specific comments
-            pr.addReviewComment(masterHash, editHash, reviewFile.toString(), 2, "Review comment");
+            var first = pr.addReviewComment(masterHash, editHash, reviewFile.toString(), 2, "Review comment");
             pr.addReviewComment(masterHash, editHash, reviewFile.toString(), 2, "Another review comment");
             TestBotRunner.runPeriodicItems(mlBot);
             listServer.processIncoming();
 
             // The archive should contain a combined entry
@@ -393,10 +393,19 @@
             assertEquals(2, reviewReply.body().split("^On.*wrote:").length);
             assertEquals(2, reviewReply.body().split("> This is now ready").length, reviewReply.body());
             assertEquals("Re: RFR: This is a pull request", reviewReply.subject());
             assertTrue(reviewReply.body().contains("Review comment\n\n"), reviewReply.body());
             assertTrue(reviewReply.body().contains("Another review comment"), reviewReply.body());
+
+            // Now reply to the first (collapsed) comment
+            pr.addReviewCommentReply(first, "I agree");
+            TestBotRunner.runPeriodicItems(mlBot);
+            listServer.processIncoming();
+
+            // The archive should contain a new entry
+            Repository.materialize(archiveFolder.path(), archive.getUrl(), "master");
+            assertEquals(3, archiveContainsCount(archiveFolder.path(), "^On.*wrote:"));
         }
     }
 
     @Test
     void commentThreading(TestInfo testInfo) throws IOException {
