diff a/src/java.desktop/share/classes/sun/font/FileFont.java b/src/java.desktop/share/classes/sun/font/FileFont.java
--- a/src/java.desktop/share/classes/sun/font/FileFont.java
+++ b/src/java.desktop/share/classes/sun/font/FileFont.java
@@ -170,11 +170,11 @@
                     scaler.invalidateScalerContext(strike.pScalerContext);
                 }
             }
         }
         if (scaler != null) {
-            scaler.dispose();
+            scaler.disposeScaler();
         }
         scaler = FontScaler.getNullScaler();
     }
 
     StrikeMetrics getFontMetrics(long pScalerContext) {
