diff a/src/java.desktop/share/classes/sun/font/NullFontScaler.java b/src/java.desktop/share/classes/sun/font/NullFontScaler.java
--- a/src/java.desktop/share/classes/sun/font/NullFontScaler.java
+++ b/src/java.desktop/share/classes/sun/font/NullFontScaler.java
@@ -62,12 +62,10 @@
     GeneralPath getGlyphVectorOutline(long pScalerContext, int[] glyphs,
         int numGlyphs, float x, float y) {
         return new GeneralPath();
     }
 
-    long getLayoutTableCache() {return 0L;}
-
     long createScalerContext(double[] matrix, int aa,
         int fm, float boldness, float italic, boolean disableHinting) {
         return getNullScalerContext();
     }
 
