diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ArrayBufferView.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ArrayBufferView.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ArrayBufferView.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ArrayBufferView.h
@@ -34,11 +34,11 @@
 
 namespace JSC {
 
 class JSArrayBufferView;
 class JSGlobalObject;
-class ExecState;
+class CallFrame;
 
 class ArrayBufferView : public RefCounted<ArrayBufferView> {
 public:
     virtual TypedArrayType getType() const = 0;
 
@@ -108,11 +108,11 @@
         if (numElements > remainingElements)
             return false;
         return true;
     }
 
-    virtual JSArrayBufferView* wrap(ExecState*, JSGlobalObject*) = 0;
+    virtual JSArrayBufferView* wrap(JSGlobalObject*, JSGlobalObject*) = 0;
 
 protected:
     JS_EXPORT_PRIVATE ArrayBufferView(RefPtr<ArrayBuffer>&&, unsigned byteOffset, unsigned byteLength);
 
     inline bool setImpl(ArrayBufferView*, unsigned byteOffset);
