diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ECMAScriptSpecInternalFunctions.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ECMAScriptSpecInternalFunctions.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ECMAScriptSpecInternalFunctions.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ECMAScriptSpecInternalFunctions.cpp
@@ -31,12 +31,12 @@
 #include "JSCInlines.h"
 #include "RegExpObject.h"
 
 namespace JSC {
 
-EncodedJSValue JSC_HOST_CALL esSpecIsConstructor(ExecState* exec)
+EncodedJSValue JSC_HOST_CALL esSpecIsConstructor(JSGlobalObject* globalObject, CallFrame* callFrame)
 {
-    bool isConstructor = exec->uncheckedArgument(0).isConstructor(exec->vm());
+    bool isConstructor = callFrame->uncheckedArgument(0).isConstructor(globalObject->vm());
     return JSValue::encode(jsBoolean(isConstructor));
 }
 
 } // namespace JSC
