diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/PutPropertySlot.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/PutPropertySlot.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/PutPropertySlot.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/PutPropertySlot.h
@@ -36,11 +36,11 @@
 
 class PutPropertySlot {
 public:
     enum Type : uint8_t { Uncachable, ExistingProperty, NewProperty, SetterProperty, CustomValue, CustomAccessor };
     enum Context { UnknownContext, PutById, PutByIdEval };
-    typedef bool (*PutValueFunc)(ExecState*, EncodedJSValue thisObject, EncodedJSValue value);
+    typedef bool (*PutValueFunc)(JSGlobalObject*, EncodedJSValue thisObject, EncodedJSValue value);
 
     PutPropertySlot(JSValue thisValue, bool isStrictMode = false, Context context = UnknownContext, bool isInitialization = false)
         : m_base(0)
         , m_thisValue(thisValue)
         , m_offset(invalidOffset)
