diff a/src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp b/src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp
@@ -49,11 +49,11 @@
   shenandoah_padding(0);
   volatile jint         _current_index;
   shenandoah_padding(1);
 
 public:
-  ShenandoahCollectionSet(ShenandoahHeap* heap, char* heap_base, size_t size);
+  ShenandoahCollectionSet(ShenandoahHeap* heap, ReservedSpace space, char* heap_base);
 
   // Add region to collection set
   void add_region(ShenandoahHeapRegion* r);
 
   // MT version
