diff a/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp b/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp
--- a/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp
+++ b/src/hotspot/cpu/ppc/globalDefinitions_ppc.hpp
@@ -56,14 +56,10 @@
 #define INCLUDE_RTM_OPT 1
 #else
 #define INCLUDE_RTM_OPT 0
 #endif
 
-#if defined(AIX)
-#define PREFERRED_METASPACE_ALIGNMENT
-#endif
-
 #define SUPPORT_RESERVED_STACK_AREA
 
 // If UseSIGTRAP is active, we only use the poll bit and no polling page.
 // Otherwise, we fall back to usage of the polling page in nmethods.
 // Define the condition to use this -XX flag.
