diff a/make/CompileToolsJdk.gmk b/make/CompileToolsJdk.gmk
--- a/make/CompileToolsJdk.gmk
+++ b/make/CompileToolsJdk.gmk
@@ -26,11 +26,10 @@
 default: all
 
 include $(SPEC)
 include MakeBase.gmk
 include JavaCompilation.gmk
-include SetupJavaCompilers.gmk
 include TextFileProcessing.gmk
 
 ################################################################################
 
 $(eval $(call IncludeCustomExtension, CompileTools.gmk))
@@ -42,24 +41,24 @@
     $(TOPDIR)/make/jdk/src/classes \
     $(BUILDTOOLS_OUTPUTDIR)/interim_tzdb_classes \
     #
 
 $(eval $(call SetupJavaCompilation, BUILD_TOOLS_JDK, \
-    SETUP := GENERATE_OLDBYTECODE, \
+    TARGET_RELEASE := $(TARGET_RELEASE_BOOTJDK), \
     SRC := $(BUILD_TOOLS_SRC_DIRS), \
     EXCLUDES := \
         build/tools/classlist \
         build/tools/deps \
         build/tools/docs \
         build/tools/jigsaw \
         build/tools/depend \
         , \
     BIN := $(BUILDTOOLS_OUTPUTDIR)/jdk_tools_classes, \
-    ADD_JAVAC_FLAGS := \
+    DISABLED_WARNINGS := options, \
+    JAVAC_FLAGS := \
         --add-exports java.desktop/sun.awt=ALL-UNNAMED \
-        --add-exports java.base/sun.text=ALL-UNNAMED \
-        , \
+        --add-exports java.base/sun.text=ALL-UNNAMED, \
 ))
 
 TARGETS += $(BUILD_TOOLS_JDK)
 
 $(eval $(call SetupCopyFiles,COPY_NIMBUS_TEMPLATES, \
@@ -70,14 +69,15 @@
 TARGETS += $(COPY_NIMBUS_TEMPLATES)
 
 ################################################################################
 
 $(eval $(call SetupJavaCompilation, COMPILE_DEPEND, \
-    SETUP := GENERATE_OLDBYTECODE, \
+    TARGET_RELEASE := $(TARGET_RELEASE_BOOTJDK), \
     SRC := $(TOPDIR)/make/jdk/src/classes, \
     INCLUDES := build/tools/depend, \
     BIN := $(BUILDTOOLS_OUTPUTDIR)/depend, \
+    DISABLED_WARNINGS := options, \
 ))
 
 DEPEND_SERVICE_PROVIDER := $(BUILDTOOLS_OUTPUTDIR)/depend/META-INF/services/com.sun.source.util.Plugin
 
 $(DEPEND_SERVICE_PROVIDER):
