diff a/src/java.base/share/classes/sun/security/tools/KeyStoreUtil.java b/src/java.base/share/classes/sun/security/tools/KeyStoreUtil.java
--- a/src/java.base/share/classes/sun/security/tools/KeyStoreUtil.java
+++ b/src/java.base/share/classes/sun/security/tools/KeyStoreUtil.java
@@ -298,18 +298,15 @@
      * @throws ClassCastException if the class has not extended Provider
      */
     public static void loadProviderByClass(
             String provClass, String arg, ClassLoader cl) {
 
-        // For compatibility, SunPKCS11, OracleUcrypto, and SunMSCAPI
+        // For compatibility, SunPKCS11, and SunMSCAPI
         // can still be loadable with -providerClass.
         if (provClass.equals("sun.security.pkcs11.SunPKCS11")) {
             loadProviderByName("SunPKCS11", arg);
             return;
-        } else if (provClass.equals("com.oracle.security.crypto.UcryptoProvider")) {
-            loadProviderByName("OracleUcrypto", arg);
-            return;
         } else if (provClass.equals("sun.security.mscapi.SunMSCAPI")) {
             loadProviderByName("SunMSCAPI", arg);
             return;
         }
 
