diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/parser/ParserModes.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/parser/ParserModes.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/parser/ParserModes.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/parser/ParserModes.h
@@ -35,11 +35,15 @@
 enum class JSParserScriptMode { Classic, Module };
 
 enum class ConstructorKind { None, Base, Extends };
 enum class SuperBinding { Needed, NotNeeded };
 
-enum DebuggerMode { DebuggerOff, DebuggerOn };
+enum class CodeGenerationMode : uint8_t {
+    Debugger = 1 << 0,
+    TypeProfiler = 1 << 1,
+    ControlFlowProfiler = 1 << 2,
+};
 
 enum class FunctionMode { FunctionExpression, FunctionDeclaration, MethodDefinition };
 
 // Keep it less than 32, it means this should be within 5 bits.
 enum class SourceParseMode : uint8_t {
