diff a/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.asm.aarch64.test/src/org/graalvm/compiler/asm/aarch64/test/TestProtectedAssembler.java b/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.asm.aarch64.test/src/org/graalvm/compiler/asm/aarch64/test/TestProtectedAssembler.java
--- a/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.asm.aarch64.test/src/org/graalvm/compiler/asm/aarch64/test/TestProtectedAssembler.java
+++ b/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.asm.aarch64.test/src/org/graalvm/compiler/asm/aarch64/test/TestProtectedAssembler.java
@@ -321,12 +321,12 @@
     protected void lsr(int size, Register dst, Register src1, Register src2) {
         super.lsr(size, dst, src1, src2);
     }
 
     @Override
-    protected void ror(int size, Register dst, Register src1, Register src2) {
-        super.ror(size, dst, src1, src2);
+    protected void rorv(int size, Register dst, Register src1, Register src2) {
+        super.rorv(size, dst, src1, src2);
     }
 
     @Override
     protected void cls(int size, Register dst, Register src) {
         super.cls(size, dst, src);
