diff a/src/hotspot/share/gc/shared/barrierSet.hpp b/src/hotspot/share/gc/shared/barrierSet.hpp
--- a/src/hotspot/share/gc/shared/barrierSet.hpp
+++ b/src/hotspot/share/gc/shared/barrierSet.hpp
@@ -313,11 +313,11 @@
     // Clone barrier support
     static void clone_in_heap(oop src, oop dst, size_t size) {
       Raw::clone(src, dst, size);
     }
 
-    static void value_copy_in_heap(void* src, void* dst, ValueKlass* md) {
+    static void value_copy_in_heap(void* src, void* dst, InlineKlass* md) {
       Raw::value_copy(src, dst, md);
     }
 
     static oop resolve(oop obj) {
       return Raw::resolve(obj);
