diff a/test/langtools/jdk/javadoc/doclet/testLinkOption/TestOptionOrder.java b/test/langtools/jdk/javadoc/doclet/testLinkOption/TestOptionOrder.java
--- a/test/langtools/jdk/javadoc/doclet/testLinkOption/TestOptionOrder.java
+++ b/test/langtools/jdk/javadoc/doclet/testLinkOption/TestOptionOrder.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -114,11 +114,11 @@
 
          checkOrder("app/App.html",
                 // Instance in See Also
                 "<dd><a href=\"http://example.com/" + first + "/lib/LibClass.html",
                 // Instance in Field declaration
-                "<td class=\"colFirst\"><code><a href=\"http://example.com/" + first + "/lib/LibClass.html"
+                "<td class=\"col-first\"><code><a href=\"http://example.com/" + first + "/lib/LibClass.html"
                 );
     }
 
     private void createLib(Path base, String name, Kind kind) throws Exception {
         Path libBase = Files.createDirectories(base.resolve(name));
