diff a/modules/javafx.web/src/main/native/Source/WebCore/loader/icon/IconLoader.cpp b/modules/javafx.web/src/main/native/Source/WebCore/loader/icon/IconLoader.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/loader/icon/IconLoader.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/loader/icon/IconLoader.cpp
@@ -28,10 +28,11 @@
 
 #include "CachedRawResource.h"
 #include "CachedResourceLoader.h"
 #include "CachedResourceRequest.h"
 #include "CachedResourceRequestInitiators.h"
+#include "CustomHeaderFields.h"
 #include "Document.h"
 #include "DocumentLoader.h"
 #include "Frame.h"
 #include "FrameLoader.h"
 #include "FrameLoaderClient.h"
