diff a/modules/javafx.media/src/main/native/vs_project/plugins/plugins.vcxproj b/modules/javafx.media/src/main/native/vs_project/plugins/plugins.vcxproj
--- a/modules/javafx.media/src/main/native/vs_project/plugins/plugins.vcxproj
+++ b/modules/javafx.media/src/main/native/vs_project/plugins/plugins.vcxproj
@@ -82,11 +82,11 @@
     <ClCompile>
       <PrecompiledHeader>NotUsing</PrecompiledHeader>
       <WarningLevel>Level3</WarningLevel>
       <Optimization>Disabled</Optimization>
       <PreprocessorDefinitions>WIN32;_WINDOWS;_USRDLL;ENABLE_PULL_MODE=1;ENABLE_SOURCE_SEEKING=1;GSTREAMER_LITE;GST_REMOVE_DEPRECATED;GST_REMOVE_DISABLED;GST_DISABLE_GST_DEBUG;GST_DISABLE_LOADSAVE;G_DISABLE_DEPRECATED;G_DISABLE_ASSERT;G_DISABLE_CHECKS;_WINDLL;_MBCS;INITGUID;%(PreprocessorDefinitions)</PreprocessorDefinitions>
-      <AdditionalIncludeDirectories>$(SolutionDir)\..\gstreamer\3rd_party\glib\;$(SolutionDir)\..\gstreamer\3rd_party\glib\glib;$(SolutionDir)\..\gstreamer\3rd_party\glib\gmodule;$(SolutionDir)\..\gstreamer\3rd_party\glib\build\win32\vs100;$(SolutionDir)\..\gstreamer\gstreamer-lite\gstreamer;$(SolutionDir)\..\gstreamer\gstreamer-lite\gstreamer\libs;$(SolutionDir)\..\gstreamer\gstreamer-lite\gst-plugins-base;$(SolutionDir)\..\gstreamer\gstreamer-lite\gst-plugins-base\gst-libs;$(SolutionDir)\..\gstreamer\plugins;C:\Program Files\Microsoft SDKs\Windows\v7.1\Samples\multimedia\directshow\baseclasses;$(SolutionDir)\..\gstreamer\plugins\dshowwrapper;$(SolutionDir)\..\gstreamer\plugins\javasource;$(SolutionDir)\..\gstreamer\plugins\progressbuffer;$(SolutionDir)\..\gstreamer\plugins\progressbuffer\win32</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>$(SolutionDir)\..\gstreamer\3rd_party\glib\;$(SolutionDir)\..\gstreamer\3rd_party\glib\glib;$(SolutionDir)\..\gstreamer\3rd_party\glib\gmodule;$(SolutionDir)\..\gstreamer\3rd_party\glib\build\win32\vs100;$(SolutionDir)\..\gstreamer\gstreamer-lite\gstreamer;$(SolutionDir)\..\gstreamer\gstreamer-lite\gstreamer\libs;$(SolutionDir)\..\gstreamer\gstreamer-lite\gst-plugins-base;$(SolutionDir)\..\gstreamer\gstreamer-lite\gst-plugins-base\gst-libs;$(SolutionDir)\..\gstreamer\plugins;$(SolutionDir)\..\gstreamer\3rd_party\baseclasses;$(SolutionDir)\..\gstreamer\plugins\dshowwrapper;$(SolutionDir)\..\gstreamer\plugins\javasource;$(SolutionDir)\..\gstreamer\plugins\progressbuffer;$(SolutionDir)\..\gstreamer\plugins\progressbuffer\win32</AdditionalIncludeDirectories>
     </ClCompile>
     <Link>
       <SubSystem>Windows</SubSystem>
       <GenerateDebugInformation>true</GenerateDebugInformation>
     </Link>
