diff a/test/jdk/java/lang/management/GarbageCollectorMXBean/GcInfoCompositeType.java b/test/jdk/java/lang/management/GarbageCollectorMXBean/GcInfoCompositeType.java
--- a/test/jdk/java/lang/management/GarbageCollectorMXBean/GcInfoCompositeType.java
+++ b/test/jdk/java/lang/management/GarbageCollectorMXBean/GcInfoCompositeType.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2006, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2006, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -28,11 +28,11 @@
  * @author  Eamonn McManus
  * @modules jdk.management
  * @run     main/othervm -XX:-ExplicitGCInvokesConcurrent GcInfoCompositeType
  */
 // Passing "-XX:-ExplicitGCInvokesConcurrent" to force System.gc()
-// run on foreground when CMS is used and prevent situations when "GcInfo"
+// run on foreground when a concurrent collector is used and prevent situations when "GcInfo"
 // is missing even though System.gc() was successfuly processed.
 
 import java.util.*;
 import java.lang.management.*;
 import java.lang.reflect.*;
