diff a/src/hotspot/share/gc/shared/gcVMOperations.cpp b/src/hotspot/share/gc/shared/gcVMOperations.cpp
--- a/src/hotspot/share/gc/shared/gcVMOperations.cpp
+++ b/src/hotspot/share/gc/shared/gcVMOperations.cpp
@@ -283,11 +283,11 @@
   GCCauseSetter gccs(heap, _gc_cause);
 
   log_debug(gc)("Full GC for CodeCache");
 
   // Don't clear the soft refs yet.
-  heap->collect_as_vm_thread(GCCause::_metadata_GC_threshold);
+  heap->collect_as_vm_thread(GCCause::_codecache_GC_threshold);
 
   log_debug(gc)("After GC for CodeCache");
 
   if (GCLocker::is_active_and_needs_gc()) {
     set_gc_locked();
