diff a/test/langtools/tools/javac/switchexpr/ExpressionSwitchFallThrough1.java b/test/langtools/tools/javac/switchexpr/ExpressionSwitchFallThrough1.java
--- a/test/langtools/tools/javac/switchexpr/ExpressionSwitchFallThrough1.java
+++ b/test/langtools/tools/javac/switchexpr/ExpressionSwitchFallThrough1.java
@@ -23,12 +23,12 @@
 
 /*
  * @test
  * @bug 8206986
  * @summary Check fall through in switch expressions.
- * @compile --enable-preview -source ${jdk.version} ExpressionSwitchFallThrough1.java
- * @run main/othervm --enable-preview ExpressionSwitchFallThrough1
+ * @compile ExpressionSwitchFallThrough1.java
+ * @run main ExpressionSwitchFallThrough1
  */
 
 import java.util.Objects;
 
 public class ExpressionSwitchFallThrough1 {
@@ -48,21 +48,21 @@
     private String printExprFallThrough(Integer p) {
         String result = "";
         return switch (p) {
             case 0: result += "0";
             case 1: result += "1";
-                break result;
-            default: break "other";
+                yield result;
+            default: yield "other";
         };
     }
 
     private String printStatementFallThrough(Integer p) {
         String result = "";
         switch (p) {
             case 0: result += "0";
             case 1: result += "1";
-                break ;
+                break;
             default: result = "other";
                 break;
         }
         return result;
     }
