diff a/src/java.base/share/classes/java/lang/invoke/X-VarHandleByteArrayView.java.template b/src/java.base/share/classes/java/lang/invoke/X-VarHandleByteArrayView.java.template
--- a/src/java.base/share/classes/java/lang/invoke/X-VarHandleByteArrayView.java.template
+++ b/src/java.base/share/classes/java/lang/invoke/X-VarHandleByteArrayView.java.template
@@ -39,11 +39,11 @@
 
 #warn
 
 final class VarHandleByteArrayAs$Type$s extends VarHandleByteArrayBase {
 
-    static JavaNioAccess nioAccess = SharedSecrets.getJavaNioAccess();
+    static final JavaNioAccess NIO_ACCESS = SharedSecrets.getJavaNioAccess();
 
     static final int ALIGN = $BoxType$.BYTES - 1;
 
 #if[floatingPoint]
     @ForceInline
@@ -563,11 +563,11 @@
             return accessMode.at.accessModeType(ByteBuffer.class, $type$.class, int.class);
         }
 
         @ForceInline
         static int index(ByteBuffer bb, int index) {
-            MemorySegmentProxy segmentProxy = nioAccess.bufferSegment(bb);
+            MemorySegmentProxy segmentProxy = NIO_ACCESS.bufferSegment(bb);
             if (segmentProxy != null) {
                 segmentProxy.checkValidState();
             }
             return Preconditions.checkIndex(index, UNSAFE.getInt(bb, BUFFER_LIMIT) - ALIGN, null);
         }
