diff a/test/langtools/jdk/javadoc/doclet/testVisibleMembers/TestVisibleMembers.java b/test/langtools/jdk/javadoc/doclet/testVisibleMembers/TestVisibleMembers.java
--- a/test/langtools/jdk/javadoc/doclet/testVisibleMembers/TestVisibleMembers.java
+++ b/test/langtools/jdk/javadoc/doclet/testVisibleMembers/TestVisibleMembers.java
@@ -270,11 +270,11 @@
                 "p");
         checkExit(Exit.OK);
 
         checkOrder("p/C.html",
                 "METHOD DETAIL",
-                "public", "void", "method()",
+                "public", "void", "method",
                 "See Also:",
                 "sub()",
                 "sub1()");
 
         checkOrder("p/ILong.html",
@@ -305,11 +305,11 @@
                 "p");
         checkExit(Exit.OK);
 
         checkOrder("p/C.html",
                 "METHOD DETAIL",
-                "public", "void", "method()", "See Also:", "sub()", "I.sub1()",
+                "public", "void", "method", "See Also:", "sub()", "I.sub1()",
                 "public", "void", "m", "Method in C. See", "I.length()"
                 );
 
         checkOrder("p/ILong.html",
                 "METHOD DETAIL",
@@ -409,14 +409,14 @@
                 "p");
         checkExit(Exit.OK);
 
         checkOrder("p/C.html",
                 "METHOD DETAIL",
-                "public", "void", "m()", "Method m in p.B",
-                "public", "void", "n()", "Method n in p.A",
-                "public", "void", "o()", "Description copied from class:", ">A<", "Method o in p.A",
-                "public", "void", "p()", "Method p in p.B",
+                "public", "void", "m", "Method m in p.B",
+                "public", "void", "n", "Method n in p.A",
+                "public", "void", "o", "Description copied from class:", ">A<", "Method o in p.A",
+                "public", "void", "p", "Method p in p.B",
                 "END OF CLASS DATA");
 
         checkOutput("p/C.html", false,
                 "Overrides",
                 "Methods declared in class p");
@@ -469,14 +469,14 @@
                 "void", "p", "Method p in p.B",
                 "CONSTRUCTOR DETAIL");
 
         checkOrder("p/C.html",
                 "METHOD DETAIL",
-                "public", "void", "m()", "Method m in p.B",
-                "public", "void", "n()", "Method n in p.A",
-                "public", "void", "o()", "Description copied from class:", ">A<", "Method o in p.A",
-                "public", "void", "p()", "Method p in p.B",
+                "public", "void", "m", "Method m in p.B",
+                "public", "void", "n", "Method n in p.A",
+                "public", "void", "o", "Description copied from class:", ">A<", "Method o in p.A",
+                "public", "void", "p", "Method p in p.B",
                 "END OF CLASS DATA");
 
         checkOutput("p/C.html", false,
                 "Overrides",
                 "Methods declared in class p");
