diff a/modules/javafx.web/src/test/java/test/javafx/scene/web/LeakTest.java b/modules/javafx.web/src/test/java/test/javafx/scene/web/LeakTest.java
--- a/modules/javafx.web/src/test/java/test/javafx/scene/web/LeakTest.java
+++ b/modules/javafx.web/src/test/java/test/javafx/scene/web/LeakTest.java
@@ -54,11 +54,10 @@
 
 public class LeakTest extends TestBase {
 
     private static final int SLEEP_TIME = 1000;
 
-    @Ignore // RT-26710: javafx.scene.web.LeakTest hangs
     @Test public void testOleg() throws InterruptedException{
         final String URL = new File("src/test/resources/test/html/guimark2-vector.html").toURI().toASCIIString();
         final int CYCLE_COUNT = 16;
         final int CYCLE_LENGTH = 5;
         final CountDownLatch latch = new CountDownLatch(CYCLE_COUNT);
@@ -80,11 +79,10 @@
         }));
         time.play();
         latch.await();
     }
 
-    @Ignore // RT-26710: javafx.scene.web.LeakTest hangs
     @Test public void testGarbageCollectability() throws InterruptedException {
         final BlockingQueue<WeakReference<WebPage>> webPageRefQueue =
                 new LinkedBlockingQueue<WeakReference<WebPage>>();
         submit(() -> {
             WebView webView = new WebView();
