diff a/src/java.desktop/macosx/classes/com/apple/laf/AquaComboBoxRendererInternal.java b/src/java.desktop/macosx/classes/com/apple/laf/AquaComboBoxRendererInternal.java
--- a/src/java.desktop/macosx/classes/com/apple/laf/AquaComboBoxRendererInternal.java
+++ b/src/java.desktop/macosx/classes/com/apple/laf/AquaComboBoxRendererInternal.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2013, 2014, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -50,11 +50,11 @@
     public Dimension getPreferredSize() {
         // From BasicComboBoxRenderer - trick to avoid zero-height items
         final Dimension size;
 
         final String text = getText();
-        if ((text == null) || ("".equals(text))) {
+        if (text == null || text.isEmpty()) {
             setText(" ");
             size = super.getPreferredSize();
             setText("");
         } else {
             size = super.getPreferredSize();
