diff a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/linux/LinuxDebugger.java b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/linux/LinuxDebugger.java
--- a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/linux/LinuxDebugger.java
+++ b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/linux/LinuxDebugger.java
@@ -55,10 +55,11 @@
 
   // For LinuxCDebugger
   public List         getThreadList();
   public List         getLoadObjectList();
   public ClosestSymbol lookup(long address);
+  public String demangle(String sym);
 
   // NOTE: this interface implicitly contains the following methods:
   // From the Debugger interface via JVMDebugger
   //   public void attach(int processID) throws DebuggerException;
   //   public void attach(String executableName, String coreFileName) throws DebuggerException;
