diff a/test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java b/test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java
--- a/test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java
+++ b/test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java
@@ -476,11 +476,11 @@
         Set<ElementType> t = types.isEmpty() ? EnumSet.allOf(ElementType.class) : types;
         String anno = """
             <a href="Anno.html" title="annotation in p">@Anno</a>""";
         String rcAnno = t.contains(ElementType.RECORD_COMPONENT) ? anno + " " : "";
         String fAnno = t.contains(ElementType.FIELD) ? "<span class=\"annotations\">" + anno + "\n</span>" : "";
-        String pAnno = t.contains(ElementType.PARAMETER) ? anno + "\n" : "";
+        String pAnno = t.contains(ElementType.PARAMETER) ? anno + "\n " : "";
         String mAnno= t.contains(ElementType.METHOD) ? "<span class=\"annotations\">" + anno + "\n</span>" : "";
 
         checkOutput("p/R.html", true,
                 """
                     <pre>public record <span class="type-name-label">R</span>("""
@@ -493,11 +493,11 @@
                         + """
                             <span class="modifiers">private final</span>&nbsp;<span class="return-type">int<\
                             /span>&nbsp;<span class="member-name">i</span></div>""",
                 """
                     <div class="member-signature"><span class="modifiers">public</span>&nbsp;<span c\
-                    lass="member-name">R</span>&#8203;(<span class="parameters">"""
+                    lass="member-name">R</span>&#8203;<span class="parameters">("""
                         + pAnno
                         + "int&nbsp;i)</span></div>",
                 "<div class=\"member-signature\">"
                         + mAnno
                         + """
