diff a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java
--- a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java
+++ b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestCheckIssueVisitor.java
@@ -204,10 +204,17 @@
         messages.add(String.format("Executable files are not allowed (file: %s)", issue.path()));
         failedChecks.add(issue.check().getClass());
         readyForReview = false;
     }
 
+    @Override
+    public void visit(SymlinkIssue issue) {
+        messages.add(String.format("Symbolic linkes are not allowed (file: %s)", issue.path()));
+        failedChecks.add(issue.check().getClass());
+        readyForReview = false;
+    }
+
     @Override
     public void visit(BlacklistIssue issue) {
         log.fine("ignored: blacklisted commit");
     }
 
