diff a/src/hotspot/share/classfile/stackMapTable.cpp b/src/hotspot/share/classfile/stackMapTable.cpp
--- a/src/hotspot/share/classfile/stackMapTable.cpp
+++ b/src/hotspot/share/classfile/stackMapTable.cpp
@@ -192,11 +192,11 @@
       Symbol* fund_name = klass_name->fundamental_name(THREAD);
       if (fund_name == NULL) {
         _stream->stackmap_format_error("TBD something bad happened", THREAD);
         return VerificationType::bogus_type();
       }
-      return VerificationType::inlinetype_type(fund_name);
+      return VerificationType::inline_type(fund_name);
     } else {
       return VerificationType::reference_type(klass_name);
     }
   }
   if (tag == ITEM_UninitializedThis) {
