diff a/cli/src/main/java/org/openjdk/skara/cli/GitPr.java b/cli/src/main/java/org/openjdk/skara/cli/GitPr.java
--- a/cli/src/main/java/org/openjdk/skara/cli/GitPr.java
+++ b/cli/src/main/java/org/openjdk/skara/cli/GitPr.java
@@ -159,20 +159,20 @@
 
     private static String statusForPullRequest(PullRequest pr) {
         var labels = pr.labels();
         if (pr.isDraft()) {
             return "DRAFT";
-        } else if (labels.contains("rfr")) {
-            return "RFR";
+        } else if (labels.contains("integrated")) {
+            return "INTEGRATED";
         } else if (labels.contains("ready")) {
             return "READY";
+        } else if (labels.contains("rfr")) {
+            return "RFR";
         } else if (labels.contains("outdated")) {
             return "OUTDATED";
         } else if (labels.contains("oca")) {
             return "OCA";
-        } else if (labels.contains("integrated")) {
-            return "INTEGRATED";
         } else {
             var checks = pr.checks(pr.headHash());
             var jcheck = Optional.ofNullable(checks.get("jcheck"));
             if (jcheck.isPresent()) {
                 var checkStatus = jcheck.get().status();
