diff a/src/hotspot/share/gc/shenandoah/heuristics/shenandoahPassiveHeuristics.cpp b/src/hotspot/share/gc/shenandoah/heuristics/shenandoahPassiveHeuristics.cpp
--- a/src/hotspot/share/gc/shenandoah/heuristics/shenandoahPassiveHeuristics.cpp
+++ b/src/hotspot/share/gc/shenandoah/heuristics/shenandoahPassiveHeuristics.cpp
@@ -76,17 +76,5 @@
       live_cset = new_cset;
       cset->add_region(r);
     }
   }
 }
-
-const char* ShenandoahPassiveHeuristics::name() {
-  return "passive";
-}
-
-bool ShenandoahPassiveHeuristics::is_diagnostic() {
-  return true;
-}
-
-bool ShenandoahPassiveHeuristics::is_experimental() {
-  return false;
-}
