diff a/src/java.desktop/share/native/libsplashscreen/splashscreen_gfx_impl.c b/src/java.desktop/share/native/libsplashscreen/splashscreen_gfx_impl.c
--- a/src/java.desktop/share/native/libsplashscreen/splashscreen_gfx_impl.c
+++ b/src/java.desktop/share/native/libsplashscreen/splashscreen_gfx_impl.c
@@ -204,11 +204,11 @@
          int stride, void *pBits, ImageFormat * format)
 {
     int depthBytes = format->depthBytes;
 
     pRect->pBits = pBits;
-    INCPN(byte_t, pRect->pBits, y * stride + x * depthBytes);
+    INCPN(byte_t, pRect->pBits, (intptr_t) y * stride + x * depthBytes);
     pRect->numLines = height;
     pRect->numSamples = width;
     pRect->stride = stride * jump;
     pRect->depthBytes = depthBytes;
     pRect->format = format;
