diff a/host/src/main/java/org/openjdk/skara/host/github/GitHubRepository.java b/host/src/main/java/org/openjdk/skara/host/github/GitHubRepository.java
--- a/host/src/main/java/org/openjdk/skara/host/github/GitHubRepository.java
+++ b/host/src/main/java/org/openjdk/skara/host/github/GitHubRepository.java
@@ -132,17 +132,17 @@
                 .build();
     }
 
     @Override
     public URI getWebUrl() {
-        return URIBuilder.base(gitHubHost.getURI())
+        return URIBuilder.base(gitHubHost.getWebURI())
                          .setPath("/" + repository)
                          .build();    }
 
     @Override
     public URI getWebUrl(Hash hash) {
-        return URIBuilder.base(gitHubHost.getURI())
+        return URIBuilder.base(gitHubHost.getWebURI())
                 .setPath("/" + repository + "/commit/" + hash.abbreviate())
                 .build();
     }
 
     @Override
