diff a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/atk/WebKitAccessibleInterfaceAction.cpp b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/atk/WebKitAccessibleInterfaceAction.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/atk/WebKitAccessibleInterfaceAction.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/atk/WebKitAccessibleInterfaceAction.cpp
@@ -81,11 +81,11 @@
 {
     auto* accessible = WEBKIT_ACCESSIBLE(action);
     returnValIfWebKitAccessibleIsInvalid(accessible, nullptr);
 
     // FIXME: Construct a proper keybinding string.
-    return webkitAccessibleCacheAndReturnAtkProperty(accessible, AtkCachedActionKeyBinding, core(action)->accessKey().string().utf8());
+    return webkitAccessibleCacheAndReturnAtkProperty(accessible, AtkCachedActionKeyBinding, core(action)->accessKey().utf8());
 }
 
 static const gchar* webkitAccessibleActionGetName(AtkAction* action, gint)
 {
     auto* accessible = WEBKIT_ACCESSIBLE(action);
