diff a/src/java.desktop/share/classes/sun/awt/FontConfiguration.java b/src/java.desktop/share/classes/sun/awt/FontConfiguration.java
--- a/src/java.desktop/share/classes/sun/awt/FontConfiguration.java
+++ b/src/java.desktop/share/classes/sun/awt/FontConfiguration.java
@@ -81,12 +81,11 @@
     /* A default FontConfiguration must be created before an alternate
      * one to ensure proper static initialisation takes place.
      */
     public FontConfiguration(SunFontManager fm) {
         if (FontUtilities.debugFonts()) {
-            FontUtilities.getLogger()
-                .info("Creating standard Font Configuration");
+            FontUtilities.logInfo("Creating standard Font Configuration");
         }
         if (FontUtilities.debugFonts() && logger == null) {
             logger = PlatformLogger.getLogger("sun.awt.FontConfiguration");
         }
         fontManager = fm;
@@ -114,12 +113,11 @@
     public FontConfiguration(SunFontManager fm,
                              boolean preferLocaleFonts,
                              boolean preferPropFonts) {
         fontManager = fm;
         if (FontUtilities.debugFonts()) {
-            FontUtilities.getLogger()
-                .info("Creating alternate Font Configuration");
+            FontUtilities.logInfo("Creating alternate Font Configuration");
         }
         this.preferLocaleFonts = preferLocaleFonts;
         this.preferPropFonts = preferPropFonts;
         /* fontConfig should be initialised by default constructor, and
          * its data tables can be shared, since readFontConfigFile doesn't
