diff a/src/hotspot/share/ci/ciFlatArrayKlass.cpp b/src/hotspot/share/ci/ciFlatArrayKlass.cpp
--- a/src/hotspot/share/ci/ciFlatArrayKlass.cpp
+++ b/src/hotspot/share/ci/ciFlatArrayKlass.cpp
@@ -37,11 +37,11 @@
 // whose Klass part is a FlatArrayKlass.
 
 // ------------------------------------------------------------------
 // ciFlatArrayKlass::ciFlatArrayKlass
 //
-// Constructor for loaded inline array klasses.
+// Constructor for loaded inline type array klasses.
 ciFlatArrayKlass::ciFlatArrayKlass(Klass* h_k) : ciArrayKlass(h_k) {
   assert(get_Klass()->is_flatArray_klass(), "wrong type");
   InlineKlass* element_Klass = get_FlatArrayKlass()->element_klass();
   _base_element_klass = CURRENT_ENV->get_klass(element_Klass);
   assert(_base_element_klass->is_inlinetype(), "bad base klass");
