diff a/src/java.base/unix/native/libjava/jni_util_md.c b/src/java.base/unix/native/libjava/jni_util_md.c
--- a/src/java.base/unix/native/libjava/jni_util_md.c
+++ b/src/java.base/unix/native/libjava/jni_util_md.c
@@ -24,10 +24,11 @@
  */
 
 #include <errno.h>
 #include <string.h>
 
+#include "jvm.h"
 #include "jni.h"
 #include "jni_util.h"
 #include "dlfcn.h"
 
 #if defined(LINUX) && (defined(_GNU_SOURCE) || \
@@ -48,10 +49,14 @@
     procHandle = (void*)dlopen(NULL, RTLD_LAZY);
 #endif
     return procHandle;
 }
 
+void* findEntryInProcess(const char* name) {
+    return JVM_FindLibraryEntry(RTLD_DEFAULT, name);
+}
+
 void buildJniFunctionName(const char *sym, const char *cname,
                           char *jniEntryName) {
     strcpy(jniEntryName, sym);
     if (cname != NULL) {
         strcat(jniEntryName, "_");
