diff a/src/java.desktop/macosx/classes/sun/lwawt/macosx/CTaskbarPeer.java b/src/java.desktop/macosx/classes/sun/lwawt/macosx/CTaskbarPeer.java
--- a/src/java.desktop/macosx/classes/sun/lwawt/macosx/CTaskbarPeer.java
+++ b/src/java.desktop/macosx/classes/sun/lwawt/macosx/CTaskbarPeer.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2016, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -23,25 +23,28 @@
  * questions.
  */
 
 package sun.lwawt.macosx;
 
-import com.apple.eawt.Application;
 import java.awt.Image;
 import java.awt.PopupMenu;
 import java.awt.Taskbar.Feature;
 import java.awt.peer.TaskbarPeer;
 
+import apple.laf.JRSUIUtils;
+import com.apple.eawt.Application;
+
 final public class CTaskbarPeer implements TaskbarPeer {
 
     CTaskbarPeer() {}
 
     @Override
     public boolean isSupported(Feature feature) {
         switch(feature) {
             case ICON_BADGE_TEXT:
             case ICON_BADGE_NUMBER:
+                return JRSUIUtils.TaskBar.isIconBadgeSupported();
             case ICON_IMAGE:
             case MENU:
             case PROGRESS_VALUE:
             case USER_ATTENTION:
                 return true;
