diff a/test/jdk/java/net/URI/Test.java b/test/jdk/java/net/URI/Test.java
--- a/test/jdk/java/net/URI/Test.java
+++ b/test/jdk/java/net/URI/Test.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, 2014, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -1410,34 +1410,51 @@
 
     static void lt(String s, String t) throws URISyntaxException {
         lt(new URI(s), new URI(t));
     }
 
-   static void gt(URI u, URI v) throws URISyntaxException {
+    static void gt0(URI u, URI v) throws URISyntaxException {
+        ne0(u, v);
+        int c = u.compareTo(v);
+        if (c <= 0) {
+            show(u);
+            show(v);
+            throw new RuntimeException("Not greater than: " + u + " " + v
+                    + " " + c);
+        }
+        out.println(u + " < " + v);
+    }
+
+    static void gt(URI u, URI v) throws URISyntaxException {
         lt(v, u);
     }
 
     static void eqHashComp() throws URISyntaxException {
 
         header("Equality, hashing, and comparison");
 
         URI o = new URI("mailto:foo@bar.com");
         URI r = new URI("reg://some%20registry/b/c/d?q#f");
         URI s = new URI("http://jag:cafebabe@java.sun.com:94/b/c/d?q#f");
+        URI t = new URI("http://example.com/%5bsegment%5d");
         eq(o, o);
         lt(o, r);
         lt(s, o);
         lt(s, r);
+
         eq(o, new URI("MaILto:foo@bar.com"));
         gt(o, new URI("mailto:foo@bar.COM"));
         eq(r, new URI("rEg://some%20registry/b/c/d?q#f"));
         gt(r, new URI("reg://Some%20Registry/b/c/d?q#f"));
         gt(r, new URI("reg://some%20registry/b/c/D?q#f"));
         eq(s, new URI("hTtP://jag:cafebabe@Java.Sun.COM:94/b/c/d?q#f"));
         gt(s, new URI("http://jag:CafeBabe@java.sun.com:94/b/c/d?q#f"));
         lt(s, new URI("http://jag:cafebabe@java.sun.com:94/b/c/d?r#f"));
         lt(s, new URI("http://jag:cafebabe@java.sun.com:94/b/c/d?q#g"));
+        cmp0(t, new URI("http://example.com/%5Bsegment%5D"), true);
+        gt0(t, new URI("http://example.com/%5BSegment%5D"));
+        lt(new URI("http://example.com/%5Asegment%5D"), new URI("http://example.com/%5Bsegment%5D"));
         eq(new URI("http://host/a%00bcd"), new URI("http://host/a%00bcd"));
         ne(new URI("http://host/a%00bcd"), new URI("http://host/aZ00bcd"));
         eq0(new URI("http://host/abc%e2def%C3ghi"),
             new URI("http://host/abc%E2def%c3ghi"));
 
