diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/shapes/RasterShape.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/shapes/RasterShape.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/shapes/RasterShape.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/shapes/RasterShape.h
@@ -38,11 +38,11 @@
 namespace WebCore {
 
 class RasterShapeIntervals {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    RasterShapeIntervals(unsigned size, int offset = 0)
+    explicit RasterShapeIntervals(unsigned size, int offset = 0)
         : m_offset(offset)
     {
         m_intervals.resize(clampTo<int>(size));
     }
 
