diff a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java
--- a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java
+++ b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java
@@ -39,11 +39,11 @@
 class CheckWorkItem extends PullRequestWorkItem {
     private final HostedRepository censusRepo;
     private final String censusRef;
     private final Map<String, String> blockingLabels;
 
-    private final Pattern metadataComments = Pattern.compile("<!-- (add|remove) contributor");
+    private final Pattern metadataComments = Pattern.compile("<!-- (?:(add|remove) contributor)|(?:summary: ')");
     private final Logger log = Logger.getLogger("org.openjdk.skara.bots.pr");
 
     CheckWorkItem(PullRequest pr, HostedRepository censusRepo, String censusRef, Map<String, String> blockingLabels, Consumer<RuntimeException> errorHandler) {
         super(pr, errorHandler);
         this.censusRepo = censusRepo;
diff a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/SummaryTests.java b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/SummaryTests.java
--- a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/SummaryTests.java
+++ b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/SummaryTests.java
@@ -84,23 +84,24 @@
             TestBotRunner.runPeriodicItems(prBot);
 
             // The bot should reply with a success message
             assertLastCommentContains(pr,"Setting summary to");
 
+            // Approve it as another user
+            var approvalPr = integrator.getPullRequest(pr.getId());
+            approvalPr.addReview(Review.Verdict.APPROVED, "Approved");
+            TestBotRunner.runPeriodicItems(prBot);
+            TestBotRunner.runPeriodicItems(prBot);
+
             // Now update it
             pr.addComment("/summary Third time is surely the charm");
             TestBotRunner.runPeriodicItems(prBot);
+            TestBotRunner.runPeriodicItems(prBot);
 
             // The bot should reply with a success message
             assertLastCommentContains(pr,"Updating existing summary");
 
-            // Approve it as another user
-            var approvalPr = integrator.getPullRequest(pr.getId());
-            approvalPr.addReview(Review.Verdict.APPROVED, "Approved");
-            TestBotRunner.runPeriodicItems(prBot);
-            TestBotRunner.runPeriodicItems(prBot);
-
             // The commit message preview should contain the final summary
             var summaryLine = pr.getComments().stream()
                                .flatMap(comment -> comment.body().lines())
                                 .filter(line -> !line.contains("/summary"))
                                 .filter(line -> !line.contains("Updating existing"))
