diff a/modules/javafx.graphics/src/main/java/javafx/scene/Scene.java b/modules/javafx.graphics/src/main/java/javafx/scene/Scene.java
--- a/modules/javafx.graphics/src/main/java/javafx/scene/Scene.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/Scene.java
@@ -1216,15 +1216,15 @@
                         throw new IllegalArgumentException(_value +
                                 "is already set as root of another scene");
                     }
 
                     if (oldRoot != null) {
-                        oldRoot.setScenes(null, null);
+                        oldRoot.setScenes(null, null, /* reapplyCSS */ false);
                     }
                     oldRoot = _value;
                     _value.getStyleClass().add(0, "root");
-                    _value.setScenes(Scene.this, null);
+                    _value.setScenes(Scene.this, null, /* reapplyCSS */ true);
                     markDirty(DirtyBits.ROOT_DIRTY);
                     _value.resize(getWidth(), getHeight()); // maybe no-op if root is not resizable
                     _value.requestLayout();
                 }
 
