diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeys.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeys.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeys.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeys.idl
@@ -31,7 +31,7 @@
     EnabledAtRuntime=EncryptedMediaAPI,
     DisabledByQuirk=hasBrokenEncryptedMediaAPISupport,
     ImplementationLacksVTable,
 ] interface MediaKeys {
     [CallWith=ScriptExecutionContext, MayThrowException] MediaKeySession createSession(optional MediaKeySessionType sessionType = "temporary");
-    Promise<bool> setServerCertificate(BufferSource serverCertificate);
+    [CallWith=ScriptExecutionContext] Promise<bool> setServerCertificate(BufferSource serverCertificate);
 };
