diff a/src/java.base/share/classes/sun/util/calendar/Era.java b/src/java.base/share/classes/sun/util/calendar/Era.java
--- a/src/java.base/share/classes/sun/util/calendar/Era.java
+++ b/src/java.base/share/classes/sun/util/calendar/Era.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -46,11 +46,11 @@
  *   -----------------------------------------------------------------------
  *   Japanese calendar       Meiji            1868-01-01T00:00:00 local time
  *                           Taisho           1912-07-30T00:00:00 local time
  *                           Showa            1926-12-25T00:00:00 local time
  *                           Heisei           1989-01-08T00:00:00 local time
- *                           NewEra           2019-05-01T00:00:00 local time
+ *                           Reiwa            2019-05-01T00:00:00 local time
  *   -----------------------------------------------------------------------
  * }</pre>
  *
  * @author Masayoshi Okutsu
  * @since 1.5
