diff a/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java b/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java
--- a/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java
+++ b/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java
@@ -51,11 +51,11 @@
         throw new RuntimeException("Unknown operating system: " + System.getProperty("os.name"));
     }
 
     private static String getCPU() {
         var p = System.getProperty("os.arch").toLowerCase();
-        if (p.startsWith("amd64")) {
+        if (p.startsWith("amd64") || p.startsWith("x86_64") || p.startsWith("x64")) {
             return "x64";
         }
         if (p.startsWith("x86") || p.startsWith("i386")) {
             return "x86";
         }
