diff a/test/langtools/tools/javac/valhalla/lworld-values/InnerClassAttributeValuenessTest.java b/test/langtools/tools/javac/valhalla/lworld-values/InnerClassAttributeValuenessTest.java
--- a/test/langtools/tools/javac/valhalla/lworld-values/InnerClassAttributeValuenessTest.java
+++ b/test/langtools/tools/javac/valhalla/lworld-values/InnerClassAttributeValuenessTest.java
@@ -46,9 +46,9 @@
         public static final Inner.ref ZERO = Inner.create(0);
         public static final Inner.ref ZERO2 = Inner.create(0);
     }
 
     public static void main(String[] args) {
-        if ((Inner.class.getModifiers() & AccessFlags.ACC_VALUE) == 0)
+        if ((Inner.class.getModifiers() & AccessFlags.ACC_INLINE) == 0)
             throw new AssertionError("Value flag missing");
     }
 }
