diff a/test/langtools/tools/javac/valhalla/lworld-values/QTypeTest.java b/test/langtools/tools/javac/valhalla/lworld-values/QTypeTest.java
--- a/test/langtools/tools/javac/valhalla/lworld-values/QTypeTest.java
+++ b/test/langtools/tools/javac/valhalla/lworld-values/QTypeTest.java
@@ -45,11 +45,11 @@
         String [] params = new String [] { "-v",
                                             Paths.get(System.getProperty("test.classes"),
                                                 "QTypedValue.class").toString() };
         runCheck(params, new String [] {
               "final value class QTypedValue",
-              "  flags: (0x0130) ACC_FINAL, ACC_SUPER, ACC_VALUE",
+              "  flags: (0x0130) ACC_FINAL, ACC_SUPER, ACC_INLINE",
               "  this_class: #1                          // QTypedValue",
               "   #1 = Class              #2             // QTypedValue",
               "   #2 = Utf8               QTypedValue",
               "   #3 = Class              #4             // \"QQTypedValue;\"",
               "   #4 = Utf8               QQTypedValue;",
