diff a/test/jdk/sun/security/x509/X509CertImpl/V3Certificate.java b/test/jdk/sun/security/x509/X509CertImpl/V3Certificate.java
--- a/test/jdk/sun/security/x509/X509CertImpl/V3Certificate.java
+++ b/test/jdk/sun/security/x509/X509CertImpl/V3Certificate.java
@@ -21,11 +21,11 @@
  * questions.
  */
 
 /*
  * @test
- * @bug 8049237
+ * @bug 8049237 8242151
  * @modules java.base/sun.security.x509
  *          java.base/sun.security.util
  *          jdk.crypto.ec
  * @summary This test generates V3 certificate with all the supported
  * extensions. Writes back the generated certificate in to a file and checks for
@@ -153,11 +153,11 @@
 
         GeneralNameInterface ipInf = new IPAddressName(address);
         GeneralName ip = new GeneralName(ipInf);
 
         GeneralNameInterface oidInf =
-                new OIDName(new ObjectIdentifier("1.2.3.4"));
+                new OIDName(ObjectIdentifier.of("1.2.3.4"));
         GeneralName oid = new GeneralName(oidInf);
 
         SubjectAlternativeNameExtension subjectName
                 = new SubjectAlternativeNameExtension();
         IssuerAlternativeNameExtension issuerName
