diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentCoordinatorClient.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentCoordinatorClient.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentCoordinatorClient.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentCoordinatorClient.h
@@ -35,10 +35,11 @@
 #import <WebKitAdditions/PaymentCoordinatorClientAdditions.h>
 #endif
 
 namespace WebCore {
 
+class Document;
 class PaymentMerchantSession;
 struct PaymentAuthorizationResult;
 struct PaymentMethodUpdate;
 struct ShippingContactUpdate;
 struct ShippingMethodUpdate;
@@ -47,24 +48,28 @@
 public:
     bool supportsVersion(unsigned version);
 
     virtual Optional<String> validatedPaymentNetwork(const String&) = 0;
     virtual bool canMakePayments() = 0;
-    virtual void canMakePaymentsWithActiveCard(const String& merchantIdentifier, const String& domainName, WTF::Function<void (bool)>&& completionHandler) = 0;
-    virtual void openPaymentSetup(const String& merchantIdentifier, const String& domainName, WTF::Function<void (bool)>&& completionHandler) = 0;
+    virtual void canMakePaymentsWithActiveCard(const String& merchantIdentifier, const String& domainName, CompletionHandler<void(bool)>&&) = 0;
+    virtual void openPaymentSetup(const String& merchantIdentifier, const String& domainName, CompletionHandler<void(bool)>&&) = 0;
 
     virtual bool showPaymentUI(const URL& originatingURL, const Vector<URL>& linkIconURLs, const ApplePaySessionPaymentRequest&) = 0;
     virtual void completeMerchantValidation(const PaymentMerchantSession&) = 0;
     virtual void completeShippingMethodSelection(Optional<ShippingMethodUpdate>&&) = 0;
     virtual void completeShippingContactSelection(Optional<ShippingContactUpdate>&&) = 0;
     virtual void completePaymentMethodSelection(Optional<PaymentMethodUpdate>&&) = 0;
     virtual void completePaymentSession(Optional<PaymentAuthorizationResult>&&) = 0;
     virtual void abortPaymentSession() = 0;
     virtual void cancelPaymentSession() = 0;
     virtual void paymentCoordinatorDestroyed() = 0;
+    virtual bool supportsUnrestrictedApplePay() const = 0;
 
     virtual bool isMockPaymentCoordinator() const { return false; }
+    virtual bool isWebPaymentCoordinator() const { return false; }
+
+    virtual bool isAlwaysOnLoggingAllowed() const { return false; }
 
 protected:
     virtual ~PaymentCoordinatorClient() = default;
 
 #if defined(PAYMENTCOORDINATORCLIENT_ADDITIONS)
