diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/PaymentSession.h
@@ -39,10 +39,11 @@
 class PaymentMethod;
 
 class PaymentSession : public virtual PaymentSessionBase {
 public:
     static ExceptionOr<void> canCreateSession(Document&);
+    static bool enabledForContext(ScriptExecutionContext&);
 
     virtual unsigned version() const = 0;
     virtual void validateMerchant(URL&&) = 0;
     virtual void didAuthorizePayment(const Payment&) = 0;
     virtual void didSelectShippingMethod(const ApplePaySessionPaymentRequest::ShippingMethod&) = 0;
