diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/modern-media-controls/media/media-controller.js b/modules/javafx.web/src/main/native/Source/WebCore/Modules/modern-media-controls/media/media-controller.js
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/modern-media-controls/media/media-controller.js
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/modern-media-controls/media/media-controller.js
@@ -97,11 +97,11 @@
     get layoutTraits()
     {
         if (this.host && this.host.compactMode)
             return LayoutTraits.Compact;
 
-        let traits = window.navigator.platform === "MacIntel" ? LayoutTraits.macOS : LayoutTraits.iOS;
+        let traits = window.isIOSFamily ? LayoutTraits.iOS : LayoutTraits.macOS;
         if (this.isFullscreen)
             return traits | LayoutTraits.Fullscreen;
         return traits;
     }
 
@@ -282,11 +282,11 @@
     }
 
     _returnMediaLayerToInlineIfNeeded()
     {
         if (this.host)
-            window.requestAnimationFrame(() => this.host.setPreparedToReturnVideoLayerToInline(this.media.webkitPresentationMode !== PiPMode));
+            this.host.setPreparedToReturnVideoLayerToInline(this.media.webkitPresentationMode !== PiPMode);
     }
 
     _controlsClassForLayoutTraits(layoutTraits)
     {
         if (layoutTraits & LayoutTraits.Compact)
