diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/DeferrableRefCounted.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/DeferrableRefCounted.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/DeferrableRefCounted.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/DeferrableRefCounted.h
@@ -1,7 +1,7 @@
 /*
- * Copyright (C) 2013 Apple Inc. All rights reserved.
+ * Copyright (C) 2013-2019 Apple Inc. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
  * are met:
  * 1. Redistributions of source code must retain the above copyright
@@ -36,12 +36,12 @@
 // just being an additional "ref", except that you can detect if it has
 // specifically happened - this can be useful either for debugging, or
 // sometimes even for some additional functionality.
 
 class DeferrableRefCountedBase {
-    static const unsigned deferredFlag = 1;
-    static const unsigned normalIncrement = 2;
+    static constexpr unsigned deferredFlag = 1;
+    static constexpr unsigned normalIncrement = 2;
 
 public:
     void ref() const
     {
         m_refCount += normalIncrement;
