diff a/src/jdk.jdi/share/classes/com/sun/tools/jdi/JNITypeParser.java b/src/jdk.jdi/share/classes/com/sun/tools/jdi/JNITypeParser.java
--- a/src/jdk.jdi/share/classes/com/sun/tools/jdi/JNITypeParser.java
+++ b/src/jdk.jdi/share/classes/com/sun/tools/jdi/JNITypeParser.java
@@ -161,10 +161,11 @@
         switch(key) {
             case (JDWP.Tag.ARRAY):
                 return  key + nextSignature();
 
             case (JDWP.Tag.OBJECT):
+            case (JDWP.Tag.INLINE_OBJECT):
                 int endClass = signature.indexOf(SIGNATURE_ENDCLASS,
                                                  currentIndex);
                 String retVal = signature.substring(currentIndex - 1,
                                                     endClass + 1);
                 currentIndex = endClass + 1;
@@ -204,10 +205,11 @@
 
             case (JDWP.Tag.CHAR):
                 return "char";
 
             case (JDWP.Tag.OBJECT):
+            case (JDWP.Tag.INLINE_OBJECT):
                 int endClass = signature.indexOf(SIGNATURE_ENDCLASS,
                                                  currentIndex);
                 String retVal = signature.substring(currentIndex,
                                                     endClass);
                 int index = retVal.indexOf(".");
