diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTableCell.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTableCell.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTableCell.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTableCell.cpp
@@ -1316,16 +1316,16 @@
         return;
 
     LayoutRect paintRect = LayoutRect(paintOffset, frameRect().size());
     adjustBorderBoxRectForPainting(paintRect);
 
-    paintBoxShadow(paintInfo, paintRect, style(), Normal);
+    paintBoxShadow(paintInfo, paintRect, style(), ShadowStyle::Normal);
 
     // Paint our cell background.
     paintBackgroundsBehindCell(paintInfo, paintOffset, this);
 
-    paintBoxShadow(paintInfo, paintRect, style(), Inset);
+    paintBoxShadow(paintInfo, paintRect, style(), ShadowStyle::Inset);
 
     if (!style().hasBorder() || table->collapseBorders())
         return;
 
     paintBorder(paintInfo, paintRect, style());
