diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMultiColumnFlow.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMultiColumnFlow.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMultiColumnFlow.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMultiColumnFlow.cpp
@@ -72,10 +72,12 @@
     return nullptr;
 }
 
 RenderMultiColumnSet* RenderMultiColumnFlow::lastMultiColumnSet() const
 {
+    ASSERT(multiColumnBlockFlow());
+
     for (RenderObject* sibling = multiColumnBlockFlow()->lastChild(); sibling; sibling = sibling->previousSibling()) {
         if (is<RenderMultiColumnSet>(*sibling))
             return downcast<RenderMultiColumnSet>(sibling);
     }
     return nullptr;
