diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderScrollbar.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderScrollbar.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderScrollbar.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderScrollbar.h
@@ -54,11 +54,13 @@
 
     int minimumThumbLength();
 
     float opacity();
 
-    std::unique_ptr<RenderStyle> getScrollbarPseudoStyle(ScrollbarPart, PseudoId);
+    bool isHiddenByStyle() const final;
+
+    std::unique_ptr<RenderStyle> getScrollbarPseudoStyle(ScrollbarPart, PseudoId) const;
 
 private:
     RenderScrollbar(ScrollableArea&, ScrollbarOrientation, Element*, Frame*);
 
     bool isOverlayScrollbar() const override { return false; }
