diff a/issuetracker/src/main/java/module-info.java b/issuetracker/src/main/java/module-info.java
--- a/issuetracker/src/main/java/module-info.java
+++ b/issuetracker/src/main/java/module-info.java
@@ -31,6 +31,10 @@
     requires transitive org.openjdk.skara.host;
     requires java.net.http;
     requires java.logging;
 
     exports org.openjdk.skara.issuetracker;
+
+    uses org.openjdk.skara.issuetracker.IssueTrackerFactory;
+
+    provides org.openjdk.skara.issuetracker.IssueTrackerFactory with org.openjdk.skara.issuetracker.jira.JiraIssueTrackerFactory;
 }
