diff a/test/hotspot/jtreg/compiler/intrinsics/bmi/TestBlsmskL.java b/test/hotspot/jtreg/compiler/intrinsics/bmi/TestBlsmskL.java
--- a/test/hotspot/jtreg/compiler/intrinsics/bmi/TestBlsmskL.java
+++ b/test/hotspot/jtreg/compiler/intrinsics/bmi/TestBlsmskL.java
@@ -31,11 +31,10 @@
  * @modules java.base/jdk.internal.misc
  *          java.management
  *
  * @build sun.hotspot.WhiteBox
  * @run driver ClassFileInstaller sun.hotspot.WhiteBox
- *                                sun.hotspot.WhiteBox$WhiteBoxPermission
  * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions
  *                   -XX:+WhiteBoxAPI
  *                   compiler.intrinsics.bmi.TestBlsmskL
  */
 
