diff a/src/java.base/share/classes/javax/net/ssl/KeyManagerFactory.java b/src/java.base/share/classes/javax/net/ssl/KeyManagerFactory.java
--- a/src/java.base/share/classes/javax/net/ssl/KeyManagerFactory.java
+++ b/src/java.base/share/classes/javax/net/ssl/KeyManagerFactory.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1999, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1999, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -161,11 +161,11 @@
      * is returned.  The specified provider must be registered
      * in the security provider list.
      *
      * <p> Note that the list of registered providers may be retrieved via
      * the {@link Security#getProviders() Security.getProviders()} method.
-
+     *
      * @param algorithm the standard name of the requested algorithm.
      *          See the <a href=
      *          "{@docRoot}/../specs/security/standard-names.html#keymanagerfactory-algorithms">
      *          KeyManagerFactory section</a> in the Java Security Standard
      *          Algorithm Names Specification for information about standard
