diff a/modules/javafx.controls/src/main/java/javafx/scene/control/ToggleButton.java b/modules/javafx.controls/src/main/java/javafx/scene/control/ToggleButton.java
--- a/modules/javafx.controls/src/main/java/javafx/scene/control/ToggleButton.java
+++ b/modules/javafx.controls/src/main/java/javafx/scene/control/ToggleButton.java
@@ -191,21 +191,18 @@
      * {@code ToggleButton} can only be in one group at any one time. If the
      * group is changed, then the button is removed from the old group prior to
      * being added to the new group.
      */
     private ObjectProperty<ToggleGroup> toggleGroup;
-    @Override
     public final void setToggleGroup(ToggleGroup value) {
         toggleGroupProperty().set(value);
     }
 
-    @Override
     public final ToggleGroup getToggleGroup() {
         return toggleGroup == null ? null : toggleGroup.get();
     }
 
-    @Override
     public final ObjectProperty<ToggleGroup> toggleGroupProperty() {
         if (toggleGroup == null) {
             toggleGroup = new ObjectPropertyBase<ToggleGroup>() {
                 private ToggleGroup old;
                 private ChangeListener<Toggle> listener = (o, oV, nV) ->
