diff a/test/hotspot/jtreg/gc/arguments/GCTypes.java b/test/hotspot/jtreg/gc/arguments/GCTypes.java
--- a/test/hotspot/jtreg/gc/arguments/GCTypes.java
+++ b/test/hotspot/jtreg/gc/arguments/GCTypes.java
@@ -68,11 +68,10 @@
         String getGCName();
     }
 
     public static enum YoungGCType implements GCType {
         DefNew("Copy"),
-        ParNew("ParNew"),
         PSNew("PS Scavenge"),
         G1("G1 Young Generation");
 
         @Override
         public String getGCName() {
@@ -93,11 +92,10 @@
         }
     }
 
     public static enum OldGCType implements GCType {
         Serial("MarkSweepCompact"),
-        CMS("ConcurrentMarkSweep"),
         PSOld("PS MarkSweep"),
         G1("G1 Old Generation");
 
         private final String gcTypeName;
 
