diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBot.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBot.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBot.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBot.java
@@ -74,12 +74,11 @@
             log.warning("No previous history found for branch '" + branch + "' - resetting mark");
             history.setBranchHash(branch, curHead);
             return;
         }
 
-        var newCommits = localRepo.commits(lastRef.get() + ".." + curHead).stream()
-                                  .collect(Collectors.toList());
+        var newCommits = localRepo.commits(lastRef.get() + ".." + curHead).asList();
         if (newCommits.size() == 0) {
             return;
         }
 
         // Update the history first - if there is a problem here we don't want to send out multiple updates
@@ -116,12 +115,11 @@
             var previous = tag.previous();
             if (!previous.isPresent()) {
                 log.warning("No previous tag found for '" + tag.tag() + "' - ignoring");
                 continue;
             }
-            var commits = localRepo.commits(previous.get().tag() + ".." + tag.tag()).stream()
-                                   .collect(Collectors.toList());
+            var commits = localRepo.commits(previous.get().tag() + ".." + tag.tag()).asList();
             if (commits.size() == 0) {
                 continue;
             }
 
             // Update the history first - if there is a problem here we don't want to send out multiple updates
