diff a/test/jdk/sun/security/pkcs12/SameDN.java b/test/jdk/sun/security/pkcs12/SameDN.java
--- a/test/jdk/sun/security/pkcs12/SameDN.java
+++ b/test/jdk/sun/security/pkcs12/SameDN.java
@@ -53,11 +53,11 @@
             throw new Exception("user2 not signed by ca2");
         }
     }
 
     static void genkeypair(String alias, String dn) throws Exception {
-        keytool(COMMON + "-genkeypair -alias " + alias + " -dname " + dn)
+        keytool(COMMON + "-genkeypair -keyalg DSA -alias " + alias + " -dname " + dn)
                 .shouldHaveExitValue(0);
     }
 
     static void gencert(String issuer, String subject) throws Exception {
         keytool(COMMON + "-certreq -alias " + subject + " -file req")
