diff a/src/hotspot/os/windows/globals_windows.hpp b/src/hotspot/os/windows/globals_windows.hpp
--- a/src/hotspot/os/windows/globals_windows.hpp
+++ b/src/hotspot/os/windows/globals_windows.hpp
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2005, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -34,12 +34,11 @@
                          product_pd, \
                          diagnostic, \
                          diagnostic_pd, \
                          notproduct, \
                          range, \
-                         constraint, \
-                         writeable)
+                         constraint)
 
 
 //
 // Defines Windows-specific default values. The flags are available on all
 // platforms, but they may have different default values on other platforms.
