diff a/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/OpenJDKTag.java b/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/OpenJDKTag.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/OpenJDKTag.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/OpenJDKTag.java
@@ -104,11 +104,11 @@
      * Build number.
      *
      * @return
      */
     public int buildNum() {
-        return Integer.valueOf(buildNum);
+        return Integer.parseInt(buildNum);
     }
 
     /**
      * Tag of the previous build (if any).
      *
@@ -124,6 +124,27 @@
         var formattedBuildNum = String.format(buildPrefix.equals("+") ? "%d" : "%02d", previousBuildNum);
         var tagName = prefix + buildPrefix + formattedBuildNum;
         var tag = new Tag(tagName);
         return create(tag);
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        OpenJDKTag that = (OpenJDKTag) o;
+        return tag.equals(that.tag) &&
+                prefix.equals(that.prefix) &&
+                version.equals(that.version) &&
+                buildPrefix.equals(that.buildPrefix) &&
+                buildNum.equals(that.buildNum);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tag, prefix, version, buildPrefix, buildNum);
+    }
 }
