diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/SVGTextLayoutEngineBaseline.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/SVGTextLayoutEngineBaseline.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/SVGTextLayoutEngineBaseline.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/SVGTextLayoutEngineBaseline.cpp
@@ -35,11 +35,11 @@
 
 float SVGTextLayoutEngineBaseline::calculateBaselineShift(const SVGRenderStyle& style, SVGElement* context) const
 {
     if (style.baselineShift() == BaselineShift::Length) {
         auto baselineShiftValueLength = style.baselineShiftValue();
-        if (baselineShiftValueLength.unitType() == LengthTypePercentage)
+        if (baselineShiftValueLength.lengthType() == SVGLengthType::Percentage)
             return baselineShiftValueLength.valueAsPercentage() * m_font.pixelSize();
 
         SVGLengthContext lengthContext(context);
         return baselineShiftValueLength.value(lengthContext);
     }
