diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/CommentPosterWorkItem.java
@@ -63,12 +63,15 @@
     }
 
     private void postNewMessage(Email email) {
         var marker = String.format(bridgedMailMarker,
                                  Base64.getEncoder().encodeToString(email.id().address().getBytes(StandardCharsets.UTF_8)));
+
         var body = marker + "\n" +
-                "Mailing list message from " + email.author().toString() + "\n\n" +
+                "*Mailing list message from [" + email.author().fullName().orElse(email.author().localPart()) +
+                "](mailto:" + email.author().address() + ") on [" + email.sender().localPart() +
+                "](mailto:" + email.sender().address() + "):*\n\n" +
                 email.body();
         pr.addComment(body);
     }
 
     @Override
