diff a/make/autoconf/build-aux/config.sub b/make/autoconf/build-aux/config.sub
--- a/make/autoconf/build-aux/config.sub
+++ b/make/autoconf/build-aux/config.sub
@@ -27,10 +27,15 @@
 # autoconf system (which might easily get lost in a future update), we
 # wrap it and fix the broken property, if needed.
 
 DIR=`dirname $0`
 
+if [ "$1"x = "x86_64-unknown-linux-musl"x ]; then
+    echo $1
+    exit
+fi
+
 # Allow wsl
 if echo $* | grep x86_64-pc-wsl >/dev/null ; then
     echo $*
     exit
 fi
