diff a/test/hotspot/jtreg/serviceability/tmtools/jstat/utils/JstatGcCapacityResults.java b/test/hotspot/jtreg/serviceability/tmtools/jstat/utils/JstatGcCapacityResults.java
--- a/test/hotspot/jtreg/serviceability/tmtools/jstat/utils/JstatGcCapacityResults.java
+++ b/test/hotspot/jtreg/serviceability/tmtools/jstat/utils/JstatGcCapacityResults.java
@@ -68,10 +68,11 @@
         // Check exit code
         assertThat(getExitCode() == 0, "Unexpected exit code: " + getExitCode());
 
         // Check Young Gen consistency
         float NGCMN = getFloatValue("NGCMN");
+        assertThat(NGCMN >= 0, "NGCMN < 0 (min generation capacity is negative)");
         float NGCMX = getFloatValue("NGCMX");
         assertThat(NGCMX >= NGCMN, "NGCMN > NGCMX (min generation capacity > max generation capacity)");
 
         float NGC = getFloatValue("NGC");
         assertThat(NGC >= NGCMN, "NGC < NGCMN (generation capacity < min generation capacity)");
@@ -100,10 +101,11 @@
             assertThat(checkFloatIsSum(NGC, S0C, S1C, EC), errMsg);
         }
 
         // Check Old Gen consistency
         float OGCMN = getFloatValue("OGCMN");
+        assertThat(OGCMN >= 0, "OGCMN < 0 (min generation capacity is negative)");
         float OGCMX = getFloatValue("OGCMX");
         assertThat(OGCMX >= OGCMN, "OGCMN > OGCMX (min generation capacity > max generation capacity)");
 
         float OGC = getFloatValue("OGC");
         assertThat(OGC >= OGCMN, "OGC < OGCMN (generation capacity < min generation capacity)");
