diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/Logging.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/Logging.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/Logging.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/Logging.cpp
@@ -52,22 +52,22 @@
 bool isLogChannelEnabled(const String& name)
 {
     WTFLogChannel* channel = WTFLogChannelByName(logChannels, logChannelCount, name.utf8().data());
     if (!channel)
         return false;
-    return channel->state != WTFLogChannelOff;
+    return channel->state != WTFLogChannelState::Off;
 }
 
 static bool logChannelsNeedInitialization = true;
 
 void setLogChannelToAccumulate(const String& name)
 {
     WTFLogChannel* channel = WTFLogChannelByName(logChannels, logChannelCount, name.utf8().data());
     if (!channel)
         return;
 
-    channel->state = WTFLogChannelOnWithAccumulation;
+    channel->state = WTFLogChannelState::OnWithAccumulation;
     logChannelsNeedInitialization = true;
 }
 
 void initializeLogChannelsIfNecessary(Optional<String> logChannelString)
 {
