diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/ScrollAnimation.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/ScrollAnimation.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/ScrollAnimation.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/ScrollAnimation.h
@@ -32,10 +32,11 @@
 
 class FloatPoint;
 class ScrollableArea;
 
 class ScrollAnimation {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     virtual ~ScrollAnimation() { };
     virtual bool scroll(ScrollbarOrientation, ScrollGranularity, float /* step */, float /* multiplier */) { return true; };
     virtual void stop() = 0;
     virtual void updateVisibleLengths() { };
