diff a/test/langtools/jdk/javadoc/doclet/testDocRootInlineTag/TestDocRootInlineTag.java b/test/langtools/jdk/javadoc/doclet/testDocRootInlineTag/TestDocRootInlineTag.java
--- a/test/langtools/jdk/javadoc/doclet/testDocRootInlineTag/TestDocRootInlineTag.java
+++ b/test/langtools/jdk/javadoc/doclet/testDocRootInlineTag/TestDocRootInlineTag.java
@@ -54,14 +54,14 @@
                 testSrc("TestDocRootTag.java"), "pkg");
         checkExit(Exit.OK);
 
         checkOutput("TestDocRootTag.html", true,
                 "<a href=\"" + uri + "/java/io/File.html\" "
-                + "title=\"class or interface in java.io\" class=\"externalLink\"><code>File</code></a>",
+                + "title=\"class or interface in java.io\" class=\"external-link\"><code>File</code></a>",
                 "<a href=\"./index-all.html\">index</a>",
                 "<a href=\"" + uri + "/java/io/File.html\" "
-                + "title=\"class or interface in java.io\" class=\"externalLink\"><code>Second File Link</code></a>",
+                + "title=\"class or interface in java.io\" class=\"external-link\"><code>Second File Link</code></a>",
                 "The value of @docRoot is \"./\"");
 
         checkOutput("index-all.html", true,
                 "My package page is <a href=\"./pkg/package-summary.html\">here</a>");
     }
