diff a/test/src/main/java/org/openjdk/skara/test/TestHost.java b/test/src/main/java/org/openjdk/skara/test/TestHost.java
--- a/test/src/main/java/org/openjdk/skara/test/TestHost.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestHost.java
@@ -84,10 +84,13 @@
     public HostedRepository getRepository(String name) {
         Repository localRepository;
         if (data.repositories.containsKey(name)) {
             localRepository = data.repositories.get(name);
         } else {
+            if (data.repositories.size() > 0) {
+                throw new RuntimeException("A test host can only manage a single repository");
+            }
             localRepository = createLocalRepository();
             data.repositories.put(name, localRepository);
         }
         return new TestHostedRepository(this, name, localRepository);
     }
