diff a/bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java b/bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java
--- a/bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java
+++ b/bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java
@@ -79,11 +79,11 @@
                 log.info("Found existing scratch directory for " + hostedRepo.name());
                 repo = Repository.get(dir)
                         .orElseThrow(() -> new RuntimeException("Repository in " + dir + " has vanished"));
             }
 
-            repo.fetchAll();
+            repo.fetchAll(false);
             var depsFile = repo.root().resolve(depsFileName);
 
             var orderedBranches = orderedBranches(repo, depsFile);
             log.info("Merge order " + orderedBranches);
             for (var branch : orderedBranches) {
