diff a/test/hotspot/jtreg/vmTestbase/nsk/monitoring/share/thread/RunningThread.java b/test/hotspot/jtreg/vmTestbase/nsk/monitoring/share/thread/RunningThread.java
--- a/test/hotspot/jtreg/vmTestbase/nsk/monitoring/share/thread/RunningThread.java
+++ b/test/hotspot/jtreg/vmTestbase/nsk/monitoring/share/thread/RunningThread.java
@@ -34,11 +34,12 @@
         private volatile boolean running = false;
         private volatile boolean ready = false;
         private Object readyLock = new Object();
         private static final String[] expectedMethods = {
                 "nsk.monitoring.share.thread.RunningThread.runInside",
-                "java.lang.Thread.yield"
+                "java.lang.Thread.yield",
+                "java.lang.Thread.yield0"
         };
 
         public RunningThread(Log log, RunType recursionType, int maxDepth) {
                 super(log, recursionType, maxDepth);
         }
