diff a/src/hotspot/share/oops/metadata.hpp b/src/hotspot/share/oops/metadata.hpp
--- a/src/hotspot/share/oops/metadata.hpp
+++ b/src/hotspot/share/oops/metadata.hpp
@@ -48,12 +48,12 @@
   virtual int  size()                  const = 0;
   virtual MetaspaceObj::Type type()    const = 0;
   virtual const char* internal_name()  const = 0;
   virtual void metaspace_pointers_do(MetaspaceClosure* iter) {}
 
-  void print()       const { print_on(tty); }
-  void print_value() const { print_value_on(tty); }
+  void print()       const;
+  void print_value() const;
 
   static void print_value_on_maybe_null(outputStream* st, const Metadata* m) {
     if (NULL == m)
       st->print("NULL");
     else
