diff a/src/hotspot/share/oops/methodData.inline.hpp b/src/hotspot/share/oops/methodData.inline.hpp
--- a/src/hotspot/share/oops/methodData.inline.hpp
+++ b/src/hotspot/share/oops/methodData.inline.hpp
@@ -24,14 +24,14 @@
 
 #ifndef SHARE_OOPS_METHODDATA_INLINE_HPP
 #define SHARE_OOPS_METHODDATA_INLINE_HPP
 
 #include "oops/methodData.hpp"
-#include "runtime/orderAccess.hpp"
+#include "runtime/atomic.hpp"
 
 inline void DataLayout::release_set_cell_at(int index, intptr_t value) {
-  OrderAccess::release_store(&_cells[index], value);
+  Atomic::release_store(&_cells[index], value);
 }
 
 inline void ProfileData::release_set_intptr_at(int index, intptr_t value) {
   assert(0 <= index && index < cell_count(), "oob");
   data()->release_set_cell_at(index, value);
