diff a/modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java b/modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java
--- a/modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java
@@ -74,11 +74,11 @@
  * Constraints" for details.
  * <p>
  * HBox lays out each managed child regardless of the child's
  * visible property value; unmanaged children are ignored.</p>
  *
- * <h3>Resizable Range</h3>
+ * <h2>Resizable Range</h2>
  *
  * <p>
  * An hbox's parent will resize the hbox within the hbox's resizable range
  * during layout.   By default the hbox computes this range based on its content
  * as outlined in the table below.
@@ -111,11 +111,11 @@
  * <p>
  * HBox does not clip its content by default, so it is possible that children's
  * bounds may extend outside its own bounds if a child's min size prevents it from
  * being fit within the hbox.</p>
  *
- * <h3>Optional Layout Constraints</h3>
+ * <h2>Optional Layout Constraints</h2>
  *
  * <p>
  * An application may set constraints on individual children to customize HBox's layout.
  * For each constraint, HBox provides a static method for setting it on the child.
  * </p>
