diff a/modules/javafx.graphics/src/main/java/javafx/scene/layout/Pane.java b/modules/javafx.graphics/src/main/java/javafx/scene/layout/Pane.java
--- a/modules/javafx.graphics/src/main/java/javafx/scene/layout/Pane.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/layout/Pane.java
@@ -56,11 +56,11 @@
  *
  * <p>
  * Pane resizes each managed child regardless of the child's visible property value;
  * unmanaged children are ignored for all layout calculations.</p>
  *
- * <h3>Resizable Range</h3>
+ * <h2>Resizable Range</h2>
  *
  * <p>
  * A pane's parent will resize the pane within the pane's resizable range
  * during layout.   By default the pane computes this range based on its content
  * as outlined in the table below:
