diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaStreamAudioDestinationNode.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaStreamAudioDestinationNode.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaStreamAudioDestinationNode.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaStreamAudioDestinationNode.h
@@ -34,10 +34,11 @@
 
 class AudioContext;
 class MediaStreamAudioSource;
 
 class MediaStreamAudioDestinationNode final : public AudioBasicInspectorNode {
+    WTF_MAKE_ISO_ALLOCATED(MediaStreamAudioDestinationNode);
 public:
     static Ref<MediaStreamAudioDestinationNode> create(AudioContext&, size_t numberOfChannels);
 
     virtual ~MediaStreamAudioDestinationNode();
 
