diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/PannerNode.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/PannerNode.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/PannerNode.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/PannerNode.h
@@ -46,10 +46,11 @@
 // A distance effect will attenuate the gain as the position moves away from the listener.
 // A cone effect will attenuate the gain as the orientation moves away from the listener.
 // All of these effects follow the OpenAL specification very closely.
 
 class PannerNode final : public AudioNode {
+    WTF_MAKE_ISO_ALLOCATED(PannerNode);
 public:
     static Ref<PannerNode> create(AudioContext& context, float sampleRate)
     {
         return adoptRef(*new PannerNode(context, sampleRate));
     }
