diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NonRetriableException.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NonRetriableException.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NonRetriableException.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NonRetriableException.java
@@ -21,15 +21,15 @@
  * questions.
  */
 package org.openjdk.skara.bots.notify;
 
 public class NonRetriableException extends Exception {
-    private final RuntimeException cause;
+    private final Throwable cause;
 
-    public NonRetriableException(RuntimeException cause) {
+    public NonRetriableException(Throwable cause) {
         this.cause = cause;
     }
 
-    public RuntimeException cause() {
+    public Throwable cause() {
         return cause;
     }
 }
