diff a/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/DragClientJava.h b/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/DragClientJava.h
--- a/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/DragClientJava.h
+++ b/modules/javafx.web/src/main/native/Source/WebKitLegacy/java/WebCoreSupport/DragClientJava.h
@@ -35,22 +35,21 @@
 class Frame;
 class Image;
 class HTMLImageElement;
 
 class DragClientJava final : public DragClient {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     DragClientJava(const JLObject &webPage);
     ~DragClientJava() override;
 
     void willPerformDragDestinationAction(DragDestinationAction, const DragData&) override;
     void willPerformDragSourceAction(DragSourceAction, const IntPoint&, DataTransfer& clipboard) override;
     //We work in window rather than view coordinates here
     DragSourceAction dragSourceActionMaskForPoint(const IntPoint& windowPoint) override;
 
     void startDrag(DragItem, DataTransfer&, Frame&) override;
-
-    void dragControllerDestroyed() override;
 private:
     JGObject m_webPage;
 };
 
 } // namespace WebCore
