diff a/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java b/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
--- a/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
+++ b/modules/javafx.controls/src/main/java/javafx/scene/control/TableFocusModel.java
@@ -40,10 +40,16 @@
      *                                                                     *
      * Public API                                                          *
      *                                                                     *
      **********************************************************************/
 
+    /**
+     * Constructor for subclasses to call.
+     */
+    public TableFocusModel() {
+    }
+
     /**
      * Causes the item at the given index to receive the focus.
      *
      * @param row The row index of the item to give focus to.
      * @param column The column of the item to give focus to. Can be null.
