diff a/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthGraphicsUtils.java b/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthGraphicsUtils.java
--- a/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthGraphicsUtils.java
+++ b/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthGraphicsUtils.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2002, 2014, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2002, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -636,22 +636,22 @@
     }
 
     static void paintAccText(Graphics g, SynthMenuItemLayoutHelper lh,
                              MenuItemLayoutHelper.LayoutResult lr) {
         String accText = lh.getAccText();
-        if (accText != null && !accText.equals("")) {
+        if (accText != null && !accText.isEmpty()) {
             g.setColor(lh.getAccStyle().getColor(lh.getAccContext(),
                     ColorType.TEXT_FOREGROUND));
             g.setFont(lh.getAccStyle().getFont(lh.getAccContext()));
             lh.getAccGraphicsUtils().paintText(lh.getAccContext(), g, accText,
                     lr.getAccRect().x, lr.getAccRect().y, -1);
         }
     }
 
     static void paintText(Graphics g, SynthMenuItemLayoutHelper lh,
                           MenuItemLayoutHelper.LayoutResult lr) {
-        if (!lh.getText().equals("")) {
+        if (!lh.getText().isEmpty()) {
             if (lh.getHtmlView() != null) {
                 // Text is HTML
                 lh.getHtmlView().paint(g, lr.getTextRect());
             } else {
                 // Text isn't HTML
