diff a/src/java.base/share/classes/javax/security/auth/login/AccountExpiredException.java b/src/java.base/share/classes/javax/security/auth/login/AccountExpiredException.java
--- a/src/java.base/share/classes/javax/security/auth/login/AccountExpiredException.java
+++ b/src/java.base/share/classes/javax/security/auth/login/AccountExpiredException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -37,10 +37,11 @@
  *
  * @since 1.4
  */
 public class AccountExpiredException extends AccountException {
 
+    @java.io.Serial
     private static final long serialVersionUID = -6064064890162661560L;
 
     /**
      * Constructs a AccountExpiredException with no detail message. A detail
      * message is a String that describes this particular exception.
