diff a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSCanvasValue.h b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSCanvasValue.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSCanvasValue.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSCanvasValue.h
@@ -39,13 +39,13 @@
     static Ref<CSSCanvasValue> create(const String& name) { return adoptRef(*new CSSCanvasValue(name)); }
     ~CSSCanvasValue();
 
     String customCSSText() const;
 
-    RefPtr<Image> image(RenderElement*, const FloatSize&);
+    RefPtr<Image> image(RenderElement&, const FloatSize&);
     bool isFixedSize() const { return true; }
-    FloatSize fixedSize(const RenderElement*);
+    FloatSize fixedSize(const RenderElement&);
 
     HTMLCanvasElement* element() const { return m_element; }
 
     bool isPending() const { return false; }
     void loadSubimages(CachedResourceLoader&, const ResourceLoaderOptions&) { }
