diff a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontVariationValue.cpp b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontVariationValue.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontVariationValue.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/css/CSSFontVariationValue.cpp
@@ -26,12 +26,11 @@
 #include "config.h"
 
 #if ENABLE(VARIATION_FONTS)
 
 #include "CSSFontVariationValue.h"
-
-#include <wtf/text/StringBuilder.h>
+#include <wtf/text/StringConcatenateNumbers.h>
 
 namespace WebCore {
 
 CSSFontVariationValue::CSSFontVariationValue(FontTag tag, float value)
     : CSSValue(FontVariationClass)
@@ -40,17 +39,11 @@
 {
 }
 
 String CSSFontVariationValue::customCSSText() const
 {
-    StringBuilder builder;
-    builder.append('"');
-    for (char c : m_tag)
-        builder.append(c);
-    builder.appendLiteral("\" ");
-    builder.appendFixedPrecisionNumber(m_value);
-    return builder.toString();
+    return makeString('"', m_tag[0], m_tag[1], m_tag[2], m_tag[3], "\" ", m_value);
 }
 
 bool CSSFontVariationValue::equals(const CSSFontVariationValue& other) const
 {
     return m_tag == other.m_tag && m_value == other.m_value;
