diff a/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLRenderer.m b/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLRenderer.m
--- a/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLRenderer.m
+++ b/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLRenderer.m
@@ -100,12 +100,12 @@
 
     // Translate each vertex by a fraction so
     // that we hit pixel centers.
     float fx = (float)x + 0.2f;
     float fy = (float)y + 0.5f;
-    struct Vertex vert[1] = {{{fx, fy}}};
-    [mtlEncoder setVertexBytes:vert length:sizeof(vert) atIndex:MeshVertexBuffer];
+    struct Vertex vert = {{fx, fy}};
+    [mtlEncoder setVertexBytes:&vert length:sizeof(vert) atIndex:MeshVertexBuffer];
     [mtlEncoder drawPrimitives:MTLPrimitiveTypePoint vertexStart:0 vertexCount:1];
 }
 
 void MTLRenderer_DrawRect(MTLContext *mtlc, BMTLSDOps * dstOps, jint x, jint y, jint w, jint h) {
     if (mtlc == NULL || dstOps == NULL || dstOps->pTexture == NULL) {
