diff a/test/jdk/java/awt/KeyboardFocusmanager/TypeAhead/FreezeTest/FreezeTest.java b/test/jdk/java/awt/KeyboardFocusmanager/TypeAhead/FreezeTest/FreezeTest.java
--- a/test/jdk/java/awt/KeyboardFocusmanager/TypeAhead/FreezeTest/FreezeTest.java
+++ b/test/jdk/java/awt/KeyboardFocusmanager/TypeAhead/FreezeTest/FreezeTest.java
@@ -27,20 +27,37 @@
  * @summary Tests that type-ahead for dialog works and doesn't block program
  * @author Dmitry.Cherepanov@SUN.COM area=awt.focus
  * @run main FreezeTest
  */
 
-import java.awt.*;
-import java.lang.reflect.InvocationTargetException;
-import java.awt.event.*;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
-
 /*
  * Tests that type-ahead doesn't block program.
  */
 
+import java.awt.AWTEvent;
+import java.awt.Button;
+import java.awt.Component;
+import java.awt.Container;
+import java.awt.Dialog;
+import java.awt.Dimension;
+import java.awt.EventQueue;
+import java.awt.Frame;
+import java.awt.KeyboardFocusManager;
+import java.awt.Point;
+import java.awt.Robot;
+import java.awt.TextField;
+import java.awt.Toolkit;
+import java.awt.event.AWTEventListener;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.awt.event.FocusAdapter;
+import java.awt.event.FocusEvent;
+import java.awt.event.InputEvent;
+import java.awt.event.KeyEvent;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
 public class FreezeTest
 {
     static Frame f;
     static Button b;
     static Dialog d;
@@ -50,13 +67,23 @@
     static int click_count = 100;
     static int deliver_count = 0;
 
     public static void main(String args[]) throws Exception {
         FreezeTest test = new FreezeTest();
-        test.init();
-        test.start();
+        try {
+            test.init();
+            test.start();
+        } finally {
+            if (d != null) {
+                d.dispose();
+            }
+            if (f != null) {
+                f.dispose();
+            }
+        }
     }
+
     public void init()
     {
         Toolkit.getDefaultToolkit().addAWTEventListener(new AWTEventListener() {
                 public void eventDispatched(AWTEvent e) {
                     if (e instanceof KeyEvent){
