diff a/modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ContextMenuContent.java b/modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ContextMenuContent.java
--- a/modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ContextMenuContent.java
+++ b/modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ContextMenuContent.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2011, 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2011, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -649,22 +649,23 @@
     }
 
     private void showMenu(Menu menu) {
         menu.show();
 
-        // if there is a submenu
-        if (submenu != null) {
-            // request focus on the first item of the submenu after it is shown
-            ContextMenuContent cmContent = (ContextMenuContent)submenu.getSkin().getNode();
-            if (cmContent != null) {
-               if (cmContent.itemsContainer.getChildren().size() > 0) {
-                   cmContent.itemsContainer.getChildren().get(0).requestFocus();
-                   cmContent.currentFocusedIndex = 0;
-               } else {
-                   cmContent.requestFocus();
-               }
-            }
+        if (submenu == null) {
+            return;
+        }
+
+        // request focus on the first item of the submenu after it is shown
+        ContextMenuContent cmContent = (ContextMenuContent)submenu.getSkin().getNode();
+        if (cmContent != null) {
+           if (cmContent.itemsContainer.getChildren().size() > 0) {
+               cmContent.itemsContainer.getChildren().get(0).requestFocus();
+               cmContent.currentFocusedIndex = 0;
+           } else {
+               cmContent.requestFocus();
+           }
         }
     }
 
     private void selectMenuItem() {
         getFocusedNode().ifPresent(n -> {
diff a/modules/javafx.controls/src/test/java/test/javafx/scene/control/ContextMenuTest.java b/modules/javafx.controls/src/test/java/test/javafx/scene/control/ContextMenuTest.java
--- a/modules/javafx.controls/src/test/java/test/javafx/scene/control/ContextMenuTest.java
+++ b/modules/javafx.controls/src/test/java/test/javafx/scene/control/ContextMenuTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2010, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2010, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
