diff a/src/hotspot/share/runtime/abstract_vm_version.cpp b/src/hotspot/share/runtime/abstract_vm_version.cpp
--- a/src/hotspot/share/runtime/abstract_vm_version.cpp
+++ b/src/hotspot/share/runtime/abstract_vm_version.cpp
@@ -255,12 +255,12 @@
     #endif
   #else
     #define FLOAT_ARCH_STR XSTR(FLOAT_ARCH)
   #endif
 
-  #ifdef HOTSPOT_LIBC
-    #define LIBC_STR "-" HOTSPOT_LIBC
+  #ifdef MUSL_LIBC
+    #define LIBC_STR "-musl"
   #else
     #define LIBC_STR ""
   #endif
 
   #ifndef HOTSPOT_BUILD_TIME
