diff a/jcheck/src/main/java/org/openjdk/skara/jcheck/JCheckConfiguration.java b/jcheck/src/main/java/org/openjdk/skara/jcheck/JCheckConfiguration.java
--- a/jcheck/src/main/java/org/openjdk/skara/jcheck/JCheckConfiguration.java
+++ b/jcheck/src/main/java/org/openjdk/skara/jcheck/JCheckConfiguration.java
@@ -85,11 +85,11 @@
         if (checkMessage == null || !checkMessage.asString().equals("lax")) {
             error += ",message,hg-tag";
             shouldCheckMessage = true;
         }
         var checkDuplicateIssues = old.get("bugids");
-        if (checkDuplicateIssues == null || !checkDuplicateIssues.equals("dup")) {
+        if (checkDuplicateIssues == null || !checkDuplicateIssues.asString().equals("dup")) {
             error += ",duplicate-issues";
         }
         config.add(error);
 
         if (project.startsWith("jdk")) {
