diff a/src/java.base/unix/classes/sun/net/dns/ResolverConfigurationImpl.java b/src/java.base/unix/classes/sun/net/dns/ResolverConfigurationImpl.java
--- a/src/java.base/unix/classes/sun/net/dns/ResolverConfigurationImpl.java
+++ b/src/java.base/unix/classes/sun/net/dns/ResolverConfigurationImpl.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2002, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2002, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -175,20 +175,10 @@
             return sl;
         }
 
         // No search keyword so use local domain
 
-
-        // LOCALDOMAIN has absolute priority on Solaris
-
-        String localDomain = localDomain0();
-        if (localDomain != null && !localDomain.isEmpty()) {
-            sl = new LinkedList<>();
-            sl.add(localDomain);
-            return sl;
-        }
-
         // try domain keyword in /etc/resolv.conf
 
         sl = java.security.AccessController.doPrivileged(
                  new java.security.PrivilegedAction<>() {
                     public LinkedList<String> run() {
@@ -252,12 +242,10 @@
     }
 
 
     // --- Native methods --
 
-    static native String localDomain0();
-
     static native String fallbackDomain0();
 
     static {
         jdk.internal.loader.BootLoader.loadLibrary("net");
     }
