diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.h
@@ -34,21 +34,21 @@
 
 class PaymentContact;
 
 class ApplePayShippingContactSelectedEvent final : public Event {
 public:
-    static Ref<ApplePayShippingContactSelectedEvent> create(const AtomicString& type, unsigned version, const PaymentContact& shippingContact)
+    static Ref<ApplePayShippingContactSelectedEvent> create(const AtomString& type, unsigned version, const PaymentContact& shippingContact)
     {
         return adoptRef(*new ApplePayShippingContactSelectedEvent(type, version, shippingContact));
     }
 
     virtual ~ApplePayShippingContactSelectedEvent();
 
     const ApplePayPaymentContact& shippingContact() const { return m_shippingContact; }
 
 private:
-    ApplePayShippingContactSelectedEvent(const AtomicString& type, unsigned version, const PaymentContact&);
+    ApplePayShippingContactSelectedEvent(const AtomString& type, unsigned version, const PaymentContact&);
 
     // Event.
     EventInterface eventInterface() const override;
 
     const ApplePayPaymentContact m_shippingContact;
