diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h
@@ -1,7 +1,7 @@
 /*
- * Copyright (C) 2015-2017 Apple Inc. All rights reserved.
+ * Copyright (C) 2015-2019 Apple Inc. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
  * are met:
  * 1. Redistributions of source code must retain the above copyright
@@ -30,10 +30,14 @@
 #include "PaymentContact.h"
 #include <wtf/Optional.h>
 #include <wtf/Vector.h>
 #include <wtf/text/WTFString.h>
 
+#if USE(APPLE_INTERNAL_SDK)
+#include <WebKitAdditions/ApplePaySessionPaymentRequestAdditions.h>
+#endif
+
 namespace WebCore {
 
 enum class PaymentAuthorizationStatus;
 
 class ApplePaySessionPaymentRequest {
@@ -135,10 +139,15 @@
     };
 
     Requester requester() const { return m_requester; }
     void setRequester(Requester requester) { m_requester = requester; }
 
+#if defined(APPLEPAYSESSIONPAYMENTREQUEST_PUBLIC_ADDITIONS)
+APPLEPAYSESSIONPAYMENTREQUEST_PUBLIC_ADDITIONS
+#undef APPLEPAYSESSIONPAYMENTREQUEST_PUBLIC_ADDITIONS
+#endif
+
 private:
     unsigned m_version { 0 };
 
     String m_countryCode;
     String m_currencyCode;
@@ -160,10 +169,15 @@
 
     String m_applicationData;
     Vector<String> m_supportedCountries;
 
     Requester m_requester { Requester::ApplePayJS };
+
+#if defined(APPLEPAYSESSIONPAYMENTREQUEST_PRIVATE_ADDITIONS)
+APPLEPAYSESSIONPAYMENTREQUEST_PRIVATE_ADDITIONS
+#undef APPLEPAYSESSIONPAYMENTREQUEST_PRIVATE_ADDITIONS
+#endif
 };
 
 struct PaymentError {
     enum class Code {
         Unknown,
