diff a/application/org.openjdk.jmc.ide.ui/META-INF/MANIFEST.MF b/application/org.openjdk.jmc.ide.ui/META-INF/MANIFEST.MF
--- a/application/org.openjdk.jmc.ide.ui/META-INF/MANIFEST.MF
+++ b/application/org.openjdk.jmc.ide.ui/META-INF/MANIFEST.MF
@@ -1,11 +1,11 @@
 Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-RequiredExecutionEnvironment: JavaSE-1.8
 Bundle-Name: IDE UI Plug-in
 Bundle-SymbolicName: org.openjdk.jmc.ide.ui;singleton:=true
-Bundle-Version: 7.1.0.qualifier
+Bundle-Version: 8.0.0.qualifier
 Bundle-Vendor: Oracle Corporation
 Require-Bundle: org.openjdk.jmc.ui,
  org.eclipse.core.resources,
  org.eclipse.ui.ide
 Bundle-ActivationPolicy: lazy
