diff a/cli/src/main/java/org/openjdk/skara/cli/GitFork.java b/cli/src/main/java/org/openjdk/skara/cli/GitFork.java
--- a/cli/src/main/java/org/openjdk/skara/cli/GitFork.java
+++ b/cli/src/main/java/org/openjdk/skara/cli/GitFork.java
@@ -151,11 +151,10 @@
         }
         if (path.startsWith("/")) {
             path = path.substring(1);
         }
 
-        System.out.println("path: " + path);
         var fork = host.getRepository(path).fork();
 
         if (token == null) {
             GitCredentials.approve(credentials);
         }
diff a/skara.py b/skara.py
--- a/skara.py
+++ b/skara.py
@@ -23,10 +23,12 @@
 import os.path
 import subprocess
 import sys
 import shutil
 
+testedwith = '4.9.2'
+
 cmdtable = {}
 if hasattr(mercurial, 'registrar') and hasattr(mercurial.registrar, 'command'):
     command = mercurial.registrar.command(cmdtable)
 elif hasattr(mercurial.cmdutil, 'command'):
     command = mercurial.cmdutil.command(cmdtable)
@@ -150,6 +152,6 @@
     ('',  'columns', '', 'Comma separated list of columns to show'),
     ('', 'no-decoration', False, 'Do not prefix lines with any decoration')
 ]
 @command('pr', info_opts, 'hg pr <list|fetch|show|checkout|apply|integrate|approve|create|close|update>')
 def pr(ui, repo, action, **opts):
-    _skara(ui, ['pr', '--mercurial', action], **opts)
+    _skara(ui, ['pr', '--mercurial', action], **opts)
