diff a/test/jdk/sanity/client/SwingSet/src/InternalFrameDemoTest.java b/test/jdk/sanity/client/SwingSet/src/InternalFrameDemoTest.java
--- a/test/jdk/sanity/client/SwingSet/src/InternalFrameDemoTest.java
+++ b/test/jdk/sanity/client/SwingSet/src/InternalFrameDemoTest.java
@@ -62,10 +62,11 @@
 
 import com.sun.swingset3.demos.internalframe.InternalFrameDemo;
 
 /*
  * @test
+ * @bug 8211703
  * @key headful
  * @summary Verifies SwingSet3 InternalFrameDemo page by checking the different
  *  actions on the parent frame, internal frame and creating internal frame
  *  with different properties
  *
