diff a/test/jdk/javax/security/auth/Subject/doAs/Test.sh b/test/jdk/javax/security/auth/Subject/doAs/Test.sh
--- a/test/jdk/javax/security/auth/Subject/doAs/Test.sh
+++ b/test/jdk/javax/security/auth/Subject/doAs/Test.sh
@@ -1,9 +1,9 @@
 #! /bin/sh
 
 #
-# Copyright (c) 2001, 2012, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2001, 2020, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.
@@ -31,15 +31,10 @@
 # ${TESTJAVA} is pointing to the jre
 #
 # set platform-dependent variables
 OS=`uname -s`
 case "$OS" in
-  SunOS )
-    PS=":"
-    FS="/"
-    RM="/bin/rm -f"
-    ;;
   Linux )
     PS=":"
     FS="/"
     RM="/bin/rm -f"
     ;;
