diff a/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTreeBuilder.h b/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTreeBuilder.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTreeBuilder.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTreeBuilder.h
@@ -216,19 +216,19 @@
 
     bool m_shouldSkipLeadingNewline { false };
 
     bool m_framesetOk { true };
 
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
     bool m_destroyed { false };
     bool m_destructionProhibited { true };
 #endif
 };
 
 inline HTMLTreeBuilder::~HTMLTreeBuilder()
 {
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
     ASSERT(!m_destroyed);
     ASSERT(!m_destructionProhibited);
     m_destroyed = true;
 #endif
 }
