diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/entriesapi/FileSystemEntry.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/entriesapi/FileSystemEntry.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/entriesapi/FileSystemEntry.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/entriesapi/FileSystemEntry.idl
@@ -22,10 +22,11 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 [
+    ActiveDOMObject,
     CustomToJSObject,
     EnabledAtRuntime=DirectoryUpload,
 ] interface FileSystemEntry {
     readonly attribute boolean isFile;
     readonly attribute boolean isDirectory;
