diff a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/overview/ResultReportUi.java b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/overview/ResultReportUi.java
--- a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/overview/ResultReportUi.java
+++ b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/overview/ResultReportUi.java
@@ -223,11 +223,11 @@
 			if (image == null) {
 				return null;
 			}
 			ImageLoader loader = new ImageLoader();
 			ByteArrayOutputStream out = new ByteArrayOutputStream();
-			loader.data = new ImageData[] {image.getImageData()};
+			loader.data = new ImageData[] {image.getImageData(100)};
 			loader.save(out, SWT.IMAGE_PNG);
 			return Base64.getEncoder().encodeToString(out.toByteArray());
 		}
 
 		public DataPageDescriptor getDescriptor() {
