diff a/src/java.instrument/windows/native/libinstrument/FileSystemSupport_md.c b/src/java.instrument/windows/native/libinstrument/FileSystemSupport_md.c
--- a/src/java.instrument/windows/native/libinstrument/FileSystemSupport_md.c
+++ b/src/java.instrument/windows/native/libinstrument/FileSystemSupport_md.c
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2004, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2004, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -43,19 +43,10 @@
 
 static int isLetter(char c) {
     return ((c >= 'a') && (c <= 'z')) || ((c >= 'A') && (c <= 'Z'));
 }
 
-char pathSeparator() {
-    return ';';
-}
-
-/* filename are case insensitive on windows */
-int filenameStrcmp(const char* s1, const char* s2) {
-    return strcasecmp(s1, s2);
-}
-
 char* basePath(const char* path) {
     char* pos = strchr(path, slash);
     char* last = NULL;
     while (pos != NULL) {
         last = pos;
