diff a/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLOperatorDictionary.h b/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLOperatorDictionary.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLOperatorDictionary.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLOperatorDictionary.h
@@ -26,11 +26,11 @@
 #pragma once
 
 #if ENABLE(MATHML)
 
 #include <unicode/utypes.h>
-#include <wtf/Optional.h>
+#include <wtf/Forward.h>
 
 namespace WebCore {
 
 namespace MathMLOperatorDictionary {
 enum Form { Infix, Prefix, Postfix };
