diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MockRealtimeVideoSource.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MockRealtimeVideoSource.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MockRealtimeVideoSource.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/mock/MockRealtimeVideoSource.h
@@ -60,10 +60,11 @@
     ImageBuffer* imageBuffer() const;
 
     Seconds elapsedTime();
     void settingsDidChange(OptionSet<RealtimeMediaSourceSettings::Flag>) override;
     MediaSample::VideoRotation sampleRotation() const final { return m_deviceOrientation; }
+    void generatePresets() override;
 
 private:
     const RealtimeMediaSourceCapabilities& capabilities() final;
     const RealtimeMediaSourceSettings& settings() final;
 
@@ -74,11 +75,10 @@
     bool supportsSizeAndFrameRate(Optional<int> width, Optional<int> height, Optional<double>) final;
     void setSizeAndFrameRate(Optional<int> width, Optional<int> height, Optional<double>) final;
     void setFrameRateWithPreset(double, RefPtr<VideoPreset>) final;
     IntSize captureSize() const;
 
-    void generatePresets() final;
 
     bool isMockSource() const final { return true; }
 
     // OrientationNotifier::Observer
     void orientationChanged(int orientation) final;
