diff a/src/java.base/unix/classes/java/net/DefaultDatagramSocketImplFactory.java b/src/java.base/unix/classes/java/net/DefaultDatagramSocketImplFactory.java
--- a/src/java.base/unix/classes/java/net/DefaultDatagramSocketImplFactory.java
+++ b/src/java.base/unix/classes/java/net/DefaultDatagramSocketImplFactory.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2007, 2011, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2007, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -67,9 +67,9 @@
                 return result;
             } catch (Exception e) {
                 throw new SocketException("can't instantiate DatagramSocketImpl");
             }
         } else {
-            return new java.net.PlainDatagramSocketImpl();
+            return new java.net.PlainDatagramSocketImpl(isMulticast);
         }
     }
 }
