diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/MIMETypeRegistry.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/MIMETypeRegistry.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/MIMETypeRegistry.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/MIMETypeRegistry.cpp
@@ -637,10 +637,16 @@
     }
 
     return true;
 }
 
+bool MIMETypeRegistry::isXMLEntityMIMEType(StringView mimeType)
+{
+    return equalLettersIgnoringASCIICase(mimeType, "text/xml-external-parsed-entity")
+        || equalLettersIgnoringASCIICase(mimeType, "application/xml-external-parsed-entity");
+}
+
 bool MIMETypeRegistry::isJavaAppletMIMEType(const String& mimeType)
 {
     // Since this set is very limited and is likely to remain so we won't bother with the overhead
     // of using a hash set.
     // Any of the MIME types below may be followed by any number of specific versions of the JVM,
