diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/text/cf/TextBreakIteratorCF.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/text/cf/TextBreakIteratorCF.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/text/cf/TextBreakIteratorCF.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/text/cf/TextBreakIteratorCF.h
@@ -24,10 +24,11 @@
 #include <wtf/spi/cf/CFStringSPI.h>
 
 namespace WTF {
 
 class TextBreakIteratorCF {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     enum class Mode {
         Caret,
         Delete
     };
