diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
@@ -108,11 +108,11 @@
                               .collect(Collectors.toList());
         return nonScratch.size() == 1 && nonScratch.get(0).equals(fixVersion);
     }
 
     /**
-     * Return True if the issue's fixVersionList is a match for fixVersion, using "-pool" or "-open".
+     * Return true if the issue's fixVersionList is a match for fixVersion, using "-pool" or "-open".
      *
      * If fixVersion has a major release of <N>, it matches the fixVersionList has an
      * <N>-pool or <N>-open entry and all other entries are scratch values.
      */
     private boolean matchPoolVersion(Issue issue, String fixVersion) {
@@ -128,11 +128,11 @@
                               .collect(Collectors.toList());
         return nonScratch.size() == 1 && (nonScratch.get(0).equals(poolVersion) || nonScratch.get(0).equals(openVersion));
     }
 
     /**
-     * Return True if fixVersionList is empty or contains only scratch values.
+     * Return true if fixVersionList is empty or contains only scratch values.
      */
     private boolean matchScratchVersion(Issue issue) {
         var nonScratch = issue.fixVersions().stream()
                               .filter(this::isNonScratchVersion)
                               .collect(Collectors.toList());
@@ -152,16 +152,16 @@
         primary.addLink(backportLink);;
         return backport;
     }
 
     /**
-     * Return issue or one of its backports that applies to fixver.
+     * Return issue or one of its backports that applies to fixVersion.
      *
      * If the main issue       has the correct fixVersion, use it.
      * If an existing Backport has the correct fixVersion, use it.
-     * If the main issue       has a matching <N>-pool fixVersion, use it.
-     * If an existing Backport has a matching <N>-pool fixVersion, use it.
+     * If the main issue       has a matching <N>-pool/open fixVersion, use it.
+     * If an existing Backport has a matching <N>-pool/open fixVersion, use it.
      * If the main issue       has a "scratch" fixVersion, use it.
      * If an existing Backport has a "scratch" fixVersion, use it.
      *
      * Otherwise, create a new Backport.
      *
@@ -262,10 +262,11 @@
                     issue.addLink(linkBuilder.build());
                 }
 
                 if (setFixVersion) {
                     if (requestedVersion != null) {
+                        // Remove any previously set versions (can only be scratch versions)
                         for (var oldVersion : issue.fixVersions()) {
                             issue.removeFixVersion(oldVersion);
                         }
                         issue.addFixVersion(requestedVersion);
                     }
