diff a/bots/merge/src/main/java/org/openjdk/skara/bots/merge/MergeBot.java b/bots/merge/src/main/java/org/openjdk/skara/bots/merge/MergeBot.java
--- a/bots/merge/src/main/java/org/openjdk/skara/bots/merge/MergeBot.java
+++ b/bots/merge/src/main/java/org/openjdk/skara/bots/merge/MergeBot.java
@@ -466,11 +466,11 @@
                     message.add("");
                     message.add("```bash");
                     message.add("$ git push -u origin merge-" + fromBranch.name() + "-into-" + toBranch.name());
                     message.add("```");
                     message.add("");
-                    message.add("Now processed to create a pull request towards this repository.");
+                    message.add("Now proceed to create a pull request towards this repository.");
                     message.add("If you are using the [Skara](https://wiki.openjdk.java.net/display/skara#Skara-Skara)" +
                                 "CLI tooling then you can run the following command to create the pull request:");
                     message.add("");
                     message.add("```bash");
                     message.add("$ git pr create");
