diff a/test/src/main/java/org/openjdk/skara/test/CheckableRepository.java b/test/src/main/java/org/openjdk/skara/test/CheckableRepository.java
--- a/test/src/main/java/org/openjdk/skara/test/CheckableRepository.java
+++ b/test/src/main/java/org/openjdk/skara/test/CheckableRepository.java
@@ -76,11 +76,11 @@
             output.append("\n");
             output.append("[checks \"whitespace\"]\n");
             output.append("files=.*\\.txt\n");
             output.append("\n");
             output.append("[checks \"reviewers\"]\n");
-            output.append("minimum=1\n");
+            output.append("reviewers=1\n");
         }
         repo.add(checkConf);
 
         repo.commit("Initial commit", "testauthor", "ta@none.none");
 
