diff a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityAttachment.h b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityAttachment.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityAttachment.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/accessibility/AccessibilityAttachment.h
@@ -46,11 +46,11 @@
     bool isAttachmentElement() const override { return true; }
 
     String roleDescription() const override;
     float valueForRange() const override;
     bool computeAccessibilityIsIgnored() const override;
-    void accessibilityText(Vector<AccessibilityText>&) override;
+    void accessibilityText(Vector<AccessibilityText>&) const override;
     explicit AccessibilityAttachment(RenderAttachment*);
 };
 
 } // namespace WebCore
 
