diff a/test/jdk/javax/swing/text/DefaultEditorKit/4278839/bug4278839.java b/test/jdk/javax/swing/text/DefaultEditorKit/4278839/bug4278839.java
--- a/test/jdk/javax/swing/text/DefaultEditorKit/4278839/bug4278839.java
+++ b/test/jdk/javax/swing/text/DefaultEditorKit/4278839/bug4278839.java
@@ -22,11 +22,11 @@
  */
 
 /*
  * @test
  * @key headful
- * @bug 4278839
+ * @bug 4278839 8233634
  * @summary Incorrect cursor movement between words at the end of line
  * @author Anton Nashatyrev
  * @library ../../../regtesthelpers
  * @build Util
  * @run main bug4278839
@@ -39,16 +39,17 @@
 public class bug4278839 extends JFrame {
 
     private static boolean passed = true;
     private static JTextArea area;
     private static Robot robo;
+    private static JFrame frame;
 
-    public static void main(String[] args) {
+    public static void main(String[] args) throws Exception {
         try {
 
             robo = new Robot();
-            robo.setAutoDelay(100);
+            robo.setAutoDelay(200);
 
             SwingUtilities.invokeAndWait(new Runnable() {
                 @Override
                 public void run() {
                     createAndShowGUI();
@@ -58,10 +59,11 @@
             robo.waitForIdle();
 
             clickMouse();
             robo.waitForIdle();
 
+            area.setCaretPosition(0);
 
             if ("Aqua".equals(UIManager.getLookAndFeel().getID())) {
                 Util.hitKeys(robo, KeyEvent.VK_HOME);
             } else {
                 Util.hitKeys(robo, KeyEvent.VK_CONTROL, KeyEvent.VK_HOME);
@@ -84,10 +86,14 @@
             passed &= moveCaret(false) == 0;
 
         } catch (Exception e) {
             throw new RuntimeException("Test failed because of an exception:",
                     e);
+        } finally {
+            if (frame != null) {
+                SwingUtilities.invokeAndWait(() -> frame.dispose());
+            }
         }
 
         if (!passed) {
             throw new RuntimeException("Test failed.");
         }
@@ -141,11 +147,11 @@
 
         return KeyEvent.VK_CONTROL;
     }
 
     private static void createAndShowGUI() {
-        JFrame frame = new JFrame();
+        frame = new JFrame();
         frame.setTitle("Bug# 4278839");
         frame.setSize(200, 200);
         frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
         area = new JTextArea("\naaa bbb\nccc ddd\n");
         frame.getContentPane().add(new JScrollPane(area));
