diff a/src/java.xml/share/classes/javax/xml/datatype/DatatypeFactory.java b/src/java.xml/share/classes/javax/xml/datatype/DatatypeFactory.java
--- a/src/java.xml/share/classes/javax/xml/datatype/DatatypeFactory.java
+++ b/src/java.xml/share/classes/javax/xml/datatype/DatatypeFactory.java
@@ -181,11 +181,11 @@
      *
      * <p>Once an application has obtained a reference to a {@code DatatypeFactory}
      * it can use the factory to configure and obtain datatype instances.
      *
      *
-     * <h2>Tip for Trouble-shooting</h2>
+     * <h4>Tip for Trouble-shooting</h4>
      * <p>Setting the {@code jaxp.debug} system property will cause
      * this method to print a lot of debug messages
      * to {@code System.err} about what it is doing and where it is looking at.
      *
      * <p> If you have problems try:
