diff a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
@@ -382,11 +382,14 @@
 
                     save_cached_extents();
 
                     // set bounds again to set to correct window size that must
                     // be the total width and height accounting extents
-                    set_bounds(0, 0, false, false, -1, -1, -1, -1);
+                    // this is ignored if size is "content size" instead of "window size"
+                    if (geometry.window_size_set) {
+                        set_bounds(0, 0, false, false, geometry.current_w, geometry.current_h, -1, -1);
+                    }
                 }
             }
         }
     }
 }
@@ -969,10 +972,12 @@
     int newH = h > 0 ? h - geometry.adjust_h : ch;
 
     gboolean size_changed = FALSE;
     gboolean pos_changed = FALSE;
 
+    geometry.window_size_set = (w > 0 || h > 0);
+
     if (newW > 0 && newH > 0) {
         size_changed = TRUE;
 
         geometry.current_cw = newW;
         geometry.current_ch = newH;
diff a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
--- a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
@@ -93,11 +93,12 @@
                        enabled(true),
                        resizable(true),
                        minw(-1),
                        minh(-1),
                        maxw(-1),
-                       maxh(-1) {}
+                       maxh(-1),
+                       window_size_set(false) {}
 
     int current_x; // current position X
     int current_y; // current position Y
     int current_w; // current window width
     int current_h; // current window height
@@ -126,10 +127,14 @@
     int minw;
     int minh;
 
     int maxw;
     int maxh;
+
+    // if the window size was set (instead of content size) - this is used to
+    // "fix" the window size accouting extents.
+    bool window_size_set;
 };
 
 class WindowContext {
 private:
     jlong screen;
