diff a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/api/JavadocTool.java b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/api/JavadocTool.java
--- a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/api/JavadocTool.java
+++ b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/api/JavadocTool.java
@@ -60,11 +60,12 @@
  * If you write code that depends on this, you do so at your own
  * risk.  This code and its internal interfaces are subject to change
  * or deletion without notice.</b></p>
  */
 public class JavadocTool implements DocumentationTool {
-    // @Override // can't add @Override until bootstrap JDK provides Tool.name()
+
+    @Override
     public String name() {
         return "javadoc";
     }
 
     @Override
