diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGBlock.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGBlock.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGBlock.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGBlock.h
@@ -27,18 +27,18 @@
 class SVGElement;
 
 class RenderSVGBlock : public RenderBlockFlow {
     WTF_MAKE_ISO_ALLOCATED(RenderSVGBlock);
 public:
-    LayoutRect visualOverflowRect() const final;
-
     SVGGraphicsElement& graphicsElement() const { return downcast<SVGGraphicsElement>(nodeForNonAnonymous()); }
 
 protected:
     RenderSVGBlock(SVGGraphicsElement&, RenderStyle&&);
     void willBeDestroyed() override;
 
+    void computeOverflow(LayoutUnit oldClientAfterEdge, bool recomputeFloats = false) override;
+
 private:
     void element() const = delete;
 
     void updateFromStyle() final;
 
