diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/DeferredSourceDump.h
@@ -23,28 +23,28 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #pragma once
 
-#include "CodeOrigin.h"
 #include "JITCode.h"
 #include "Strong.h"
 
 namespace JSC {
 
 class CodeBlock;
 
 class DeferredSourceDump {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     DeferredSourceDump(CodeBlock*);
-    DeferredSourceDump(CodeBlock*, CodeBlock* rootCodeBlock, JITCode::JITType rootJITType, unsigned callerBytecodeIndex);
+    DeferredSourceDump(CodeBlock*, CodeBlock* rootCodeBlock, JITType rootJITType, unsigned callerBytecodeIndex);
 
     void dump();
 
 private:
     Strong<CodeBlock> m_codeBlock;
     Strong<CodeBlock> m_rootCodeBlock;
-    JITCode::JITType m_rootJITType;
+    JITType m_rootJITType;
     unsigned m_callerBytecodeIndex { UINT_MAX };
 };
 
 } // namespace JSC
