diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ObjectPropertyCondition.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ObjectPropertyCondition.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ObjectPropertyCondition.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ObjectPropertyCondition.cpp
@@ -140,19 +140,19 @@
         return false;
 
     return isWatchable(m_object->structure(), effort);
 }
 
-bool ObjectPropertyCondition::isStillLive() const
+bool ObjectPropertyCondition::isStillLive(VM& vm) const
 {
     if (!*this)
         return false;
 
-    if (!Heap::isMarked(m_object))
+    if (!vm.heap.isMarked(m_object))
         return false;
 
-    return m_condition.isStillLive();
+    return m_condition.isStillLive(vm);
 }
 
 void ObjectPropertyCondition::validateReferences(const TrackedReferences& tracked) const
 {
     if (!*this)
