diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/PutByIdStatus.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/PutByIdStatus.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/PutByIdStatus.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/PutByIdStatus.h
@@ -41,10 +41,11 @@
 class StructureStubInfo;
 
 typedef HashMap<CodeOrigin, StructureStubInfo*, CodeOriginApproximateHash> StubInfoMap;
 
 class PutByIdStatus {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     enum State {
         // It's uncached so we have no information.
         NoInformation,
         // It's cached as a simple store of some kind.
@@ -114,11 +115,11 @@
     const Vector<PutByIdVariant, 1>& variants() const { return m_variants; }
     const PutByIdVariant& at(size_t index) const { return m_variants[index]; }
     const PutByIdVariant& operator[](size_t index) const { return at(index); }
 
     void markIfCheap(SlotVisitor&);
-    bool finalize();
+    bool finalize(VM&);
 
     void merge(const PutByIdStatus&);
 
     void filter(const StructureSet&);
 
