diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasession/MediaRemoteControls.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasession/MediaRemoteControls.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasession/MediaRemoteControls.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediasession/MediaRemoteControls.h
@@ -32,11 +32,12 @@
 
 namespace WebCore {
 
 class MediaSession;
 
-class MediaRemoteControls : public RefCounted<MediaRemoteControls>, public EventTargetWithInlineData {
+class MediaRemoteControls final : public RefCounted<MediaRemoteControls>, public EventTargetWithInlineData {
+    WTF_MAKE_ISO_ALLOCATED(MediaRemoteControls);
 public:
     static Ref<MediaRemoteControls> create(ScriptExecutionContext& context, MediaSession* session = nullptr)
     {
         return adoptRef(*new MediaRemoteControls(context, session));
     }
