diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/client/TransactionOperation.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/client/TransactionOperation.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/client/TransactionOperation.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/client/TransactionOperation.cpp
@@ -40,11 +40,11 @@
     m_objectStoreIdentifier = request.sourceObjectStoreIdentifier();
     m_indexIdentifier = request.sourceIndexIdentifier();
     if (m_indexIdentifier)
         m_indexRecordType = request.requestedIndexRecordType();
     if (auto* cursor = request.pendingCursor())
-        m_cursorIdentifier = std::make_unique<IDBResourceIdentifier>(cursor->info().identifier());
+        m_cursorIdentifier = makeUnique<IDBResourceIdentifier>(cursor->info().identifier());
 
     m_idbRequest = &request;
 }
 
 } // namespace IDBClient
