diff a/application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js b/application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js
--- a/application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js
+++ b/application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js
@@ -131,11 +131,10 @@
 	var adjustTextToSearch = removeSpecialCharacters(textToSearch);
 	return (text !== undefined && removeSpecialCharacters(text).includes(adjustTextToSearch));
 };
 
 const removeSpecialCharacters = function(text) {
-	//return text.trim().replace(/[&\/\\#,+()$~%'":*?<>{}]/g, '_').toLowerCase();
 	const map = Array.prototype.map;
     const resultText = map.call(text.trim().toLowerCase(), element => {
         if(specialCharactersMap.has(element)){
             return specialCharactersMap.get(element);
         } else {
