diff a/apps/samples/Ensemble8/build.gradle b/apps/samples/Ensemble8/build.gradle
--- a/apps/samples/Ensemble8/build.gradle
+++ b/apps/samples/Ensemble8/build.gradle
@@ -2,13 +2,13 @@
 apply plugin:'application'
 
 def mainClassName = "ensemble.EnsembleApp"
 
 def FileCollection apachecp = files(
-   "./lib/lucene-core-7.7.1.jar",
-   "./lib/lucene-grouping-7.7.1.jar",
-   "./lib/lucene-queryparser-7.7.1.jar")
+   "./lib/lucene-core-7.7.2.jar",
+   "./lib/lucene-grouping-7.7.2.jar",
+   "./lib/lucene-queryparser-7.7.2.jar")
 
 sourceSets {
     main {
         java {
             srcDirs 'src/app/java',
