diff a/modules/javafx.controls/src/test/java/test/javafx/scene/control/TextInputControlTest.java b/modules/javafx.controls/src/test/java/test/javafx/scene/control/TextInputControlTest.java
--- a/modules/javafx.controls/src/test/java/test/javafx/scene/control/TextInputControlTest.java
+++ b/modules/javafx.controls/src/test/java/test/javafx/scene/control/TextInputControlTest.java
@@ -1903,12 +1903,12 @@
     @Test public void replaceSelectionAtEndWithListener() {
         StringBuilder selectedTextLog = new StringBuilder();
         StringBuilder selectionLog = new StringBuilder();
         textInput.setText("x xxx");
         textInput.selectRange(2, 5);
-        textInput.selectedTextProperty().addListener((__, ___, selection) -> selectedTextLog.append("|" + selection));
-        textInput.selectionProperty().addListener((__, ___, selection) -> selectionLog.append("|" + selection.getStart() + "," + selection.getEnd()));
+        textInput.selectedTextProperty().addListener((observable, oldValue, newValue) -> selectedTextLog.append("|" + newValue));
+        textInput.selectionProperty().addListener((observable, oldValue, newValue) -> selectionLog.append("|" + newValue.getStart() + "," + newValue.getEnd()));
         textInput.replaceSelection("a");
         assertEquals("|", selectedTextLog.toString());
         assertEquals("|3,3", selectionLog.toString());
         assertEquals("x a", textInput.getText());
     }
@@ -1917,13 +1917,13 @@
         textInput.setText("abcdefghij");
 
         StringBuilder selectedTextLog = new StringBuilder();
         StringBuilder selectionLog = new StringBuilder();
         StringBuilder textLog = new StringBuilder();
-        textInput.selectedTextProperty().addListener((__, ___, selection) -> selectedTextLog.append("|" + selection));
-        textInput.selectionProperty().addListener((__, ___, selection) -> selectionLog.append("|" + selection.getStart() + "," + selection.getEnd()));
-        textInput.textProperty().addListener((__, ___, text) -> textLog.append("|" + text));
+        textInput.selectedTextProperty().addListener((observable, oldValue, newValue) -> selectedTextLog.append("|" + newValue));
+        textInput.selectionProperty().addListener((observable, oldValue, newValue) -> selectionLog.append("|" + newValue.getStart() + "," + newValue.getEnd()));
+        textInput.textProperty().addListener((observable, oldValue, newValue) -> textLog.append("|" + newValue));
 
         textInput.selectRange(3, 6);
         assertEquals("|def", selectedTextLog.toString());
         assertEquals("|3,6", selectionLog.toString());
         assertEquals("", textLog.toString());
