diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBotFactory.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBotFactory.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBotFactory.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/JNotifyBotFactory.java
@@ -76,13 +76,10 @@
             }
             if (repo.value().contains("mailinglists")) {
                 var email = specific.get("email").asObject();
                 var smtp = email.get("smtp").asString();
                 var archive = URIBuilder.base(email.get("archive").asString()).build();
-                var senderName = email.get("name").asString();
-                var senderMail = email.get("address").asString();
-                var sender = EmailAddress.from(senderName, senderMail);
                 var listServer = MailingListServerFactory.createMailmanServer(archive, smtp);
 
                 for (var mailinglist : repo.value().get("mailinglists").asArray()) {
                     var recipient = mailinglist.get("recipient").asString();
                     var recipientAddress = EmailAddress.parse(recipient);
@@ -103,11 +100,11 @@
 
                     Map<String, String> headers = mailinglist.contains("headers") ?
                             mailinglist.get("headers").fields().stream()
                                        .collect(Collectors.toMap(JSONObject.Field::name, field -> field.value().asString())) :
                             Map.of();
-
+                    var sender = mailinglist.contains("sender") ? EmailAddress.parse(mailinglist.get("sender").asString()) : null;
                     updaters.add(new MailingListUpdater(listServer.getList(recipient), recipientAddress, sender,
                                                         includeBranchNames, mode, headers));
                 }
             }
 
