diff a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/swt/TreemapItem.java b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/swt/TreemapItem.java
--- a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/swt/TreemapItem.java
+++ b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/swt/TreemapItem.java
@@ -120,11 +120,11 @@
 			// adding a 0 weighted node to the end of decreasingly sorted list preserves the sorted structure
 			parentItem.children.add(this);
 		}
 	}
 
-	/* package-private */ static TreemapItem checkNull(TreemapItem item) {
+	static TreemapItem checkNull(TreemapItem item) {
 		if (item == null) {
 			SWT.error(SWT.ERROR_NULL_ARGUMENT);
 		}
 
 		return item;
@@ -169,11 +169,11 @@
 
 		sum += realWeight;
 		apparentWeight = sum;
 	}
 
-	/* package-private */ void paintItem(GC gc, Rectangle bounds, boolean all) {
+	void paintItem(GC gc, Rectangle bounds, boolean all) {
 		this.bounds = bounds;
 
 		Color bg = gc.getBackground();
 		Color fg = gc.getForeground();
 		Font font = gc.getFont();
