diff a/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.hpp b/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.hpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.hpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahStringDedup.hpp
@@ -24,10 +24,11 @@
 
 #ifndef SHARE_GC_SHENANDOAH_SHENANDOAHSTRINGDEDUP_HPP
 #define SHARE_GC_SHENANDOAH_SHENANDOAHSTRINGDEDUP_HPP
 
 #include "gc/shared/stringdedup/stringDedup.hpp"
+#include "gc/shenandoah/shenandoahPhaseTimings.hpp"
 #include "memory/iterator.hpp"
 
 class ShenandoahStringDedup : public StringDedup {
 public:
   // Initialize string deduplication.
@@ -37,11 +38,15 @@
   static void enqueue_candidate(oop java_string);
 
   // Deduplicate a string, the call is lock-free
   static void deduplicate(oop java_string);
 
-  static void parallel_oops_do(BoolObjectClosure* is_alive, OopClosure* cl, uint worker_id);
+  static void parallel_oops_do(ShenandoahPhaseTimings::Phase phase,
+                               BoolObjectClosure* is_alive,
+                               OopClosure* cl,
+                               uint worker_id);
+
   static void oops_do_slow(OopClosure* cl);
 
   static inline bool is_candidate(oop obj);
 
   static void unlink_or_oops_do(BoolObjectClosure* is_alive,
