diff a/src/hotspot/share/gc/shenandoah/shenandoahConcurrentRoots.cpp b/src/hotspot/share/gc/shenandoah/shenandoahConcurrentRoots.cpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahConcurrentRoots.cpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahConcurrentRoots.cpp
@@ -35,11 +35,11 @@
   return can_do_concurrent_roots() &&
          !ShenandoahHeap::heap()->is_stw_gc_in_progress();
 }
 
 bool ShenandoahConcurrentRoots::can_do_concurrent_class_unloading() {
-#if defined(X86) && !defined(SOLARIS)
+#if (defined(X86) || defined(AARCH64)) && !defined(SOLARIS)
   return ShenandoahCodeRootsStyle == 2 &&
          ClassUnloading;
 #else
   return false;
 #endif
