diff a/modules/javafx.web/src/main/native/Source/WebKitLegacy/Storage/StorageNamespaceImpl.h b/modules/javafx.web/src/main/native/Source/WebKitLegacy/Storage/StorageNamespaceImpl.h
--- a/modules/javafx.web/src/main/native/Source/WebKitLegacy/Storage/StorageNamespaceImpl.h
+++ b/modules/javafx.web/src/main/native/Source/WebKitLegacy/Storage/StorageNamespaceImpl.h
@@ -59,11 +59,11 @@
 
 private:
     StorageNamespaceImpl(WebCore::StorageType, const String& path, unsigned quota, PAL::SessionID);
 
     Ref<WebCore::StorageArea> storageArea(const WebCore::SecurityOriginData&) override;
-    Ref<StorageNamespace> copy(WebCore::Page* newPage) override;
+    Ref<StorageNamespace> copy(WebCore::Page& newPage) override;
 
     typedef HashMap<WebCore::SecurityOriginData, RefPtr<StorageAreaImpl>> StorageAreaMap;
     StorageAreaMap m_storageAreaMap;
 
     WebCore::StorageType m_storageType;
