diff a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/CheckTests.java b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/CheckTests.java
--- a/bots/pr/src/test/java/org/openjdk/skara/bots/pr/CheckTests.java
+++ b/bots/pr/src/test/java/org/openjdk/skara/bots/pr/CheckTests.java
@@ -30,10 +30,11 @@
 import java.io.IOException;
 import java.nio.file.Files;
 import java.util.Map;
 
 import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
 
 class CheckTests {
     @Test
     void simpleCommit(TestInfo testInfo) throws IOException {
         try (var credentials = new HostCredentials(testInfo);
@@ -374,10 +375,13 @@
         try (var credentials = new HostCredentials(testInfo);
              var tempFolder = new TemporaryDirectory()) {
             var author = credentials.getHostedRepository();
             var reviewer = credentials.getHostedRepository();
 
+            // This test is only relevant on hosts not supporting proper review comment bodies
+            assumeTrue(!author.host().supportsReviewBody());
+
             var censusBuilder = credentials.getCensusBuilder()
                                            .addAuthor(author.host().getCurrentUserDetails().id())
                                            .addReviewer(reviewer.host().getCurrentUserDetails().id());
             var checkBot = new PullRequestBot(author, censusBuilder.build(), "master");
 
