diff a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/ext/alsa/gstalsa.h b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/ext/alsa/gstalsa.h
--- a/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/ext/alsa/gstalsa.h
+++ b/modules/javafx.media/src/main/native/gstreamer/gstreamer-lite/gst-plugins-base/ext/alsa/gstalsa.h
@@ -71,11 +71,11 @@
                                             GstCaps   * caps);
 
 extern const GstAudioChannelPosition alsa_position[][8];
 #ifdef SND_CHMAP_API_VERSION
 gboolean alsa_chmap_to_channel_positions (const snd_pcm_chmap_t *chmap,
-                      GstAudioChannelPosition *pos);
+                                          GstAudioChannelPosition *pos);
 
 void alsa_detect_channels_mapping (GstObject * obj,
                                    snd_pcm_t * handle,
                                    GstAudioRingBufferSpec * spec,
                                    guint channels,
