diff a/test/hotspot/jtreg/runtime/sealedClasses/Pkg/NotPermitted.jcod b/test/hotspot/jtreg/runtime/sealedClasses/Pkg/NotPermitted.jcod
--- a/test/hotspot/jtreg/runtime/sealedClasses/Pkg/NotPermitted.jcod
+++ b/test/hotspot/jtreg/runtime/sealedClasses/Pkg/NotPermitted.jcod
@@ -29,11 +29,11 @@
 //    public class NotPermitted implements SealedInterface { }
 //
 class Pkg/NotPermitted {
   0xCAFEBABE;
   0; // minor version
-  59; // version
+  60; // version
   [15] { // Constant Pool
     ; // first element is empty
     Method #2 #3; // #1     at 0x0A
     class #4; // #2     at 0x0F
     NameAndType #5 #6; // #3     at 0x12
