diff a/forge/src/test/java/org/openjdk/skara/forge/CheckBuilderTests.java b/forge/src/test/java/org/openjdk/skara/forge/CheckBuilderTests.java
--- a/forge/src/test/java/org/openjdk/skara/forge/CheckBuilderTests.java
+++ b/forge/src/test/java/org/openjdk/skara/forge/CheckBuilderTests.java
@@ -32,22 +32,21 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 class CheckBuilderTests {
     @Test
     void testFrom() {
-        var hash = new Hash("0".repeat(40));
         var name = "test";
         var title = "title";
         var summary = "summary";
         var metadata = "metadata";
         var annotation = CheckAnnotationBuilder.create("README", 0, 1, CheckAnnotationLevel.NOTICE, "Message")
                                                .build();
         var startedAt = ZonedDateTime.now();
         var completedAt = ZonedDateTime.now();
         var success = true;
 
-        var existing = CheckBuilder.create(name, hash)
+        var existing = CheckBuilder.create(name, Hash.zero())
                                    .title(title)
                                    .summary(summary)
                                    .metadata(metadata)
                                    .annotation(annotation)
                                    .startedAt(startedAt)
