diff a/tests/system/src/testscriptapp1/java/mymod/pseudoScriptEngine/InvocationInfos.java b/tests/system/src/testscriptapp1/java/mymod/pseudoScriptEngine/InvocationInfos.java
--- a/tests/system/src/testscriptapp1/java/mymod/pseudoScriptEngine/InvocationInfos.java
+++ b/tests/system/src/testscriptapp1/java/mymod/pseudoScriptEngine/InvocationInfos.java
@@ -48,11 +48,11 @@
             Bindings binding = context.getBindings(scope);
             bindings.put(scope, binding == null ? new TreeMap<String,Object>() : new TreeMap<String,Object>(binding));
         }
     }
 
-    /** 
+    /**
      * Creates and returns a string having all information formatted to ease debugging.
      * @return string formatted to ease debugging
      */
     public String toDebugFormat(String indentation) {
         StringBuilder sb = new StringBuilder();
