diff a/src/java.desktop/unix/native/common/java2d/x11/X11FontScaler_md.c b/src/java.desktop/unix/native/common/java2d/x11/X11FontScaler_md.c
--- a/src/java.desktop/unix/native/common/java2d/x11/X11FontScaler_md.c
+++ b/src/java.desktop/unix/native/common/java2d/x11/X11FontScaler_md.c
@@ -271,19 +271,23 @@
     unsigned char *srcRow, *dstRow, *dstByte;
     int wholeByteCount, remainingBitsCount;
     unsigned int imageSize;
     JNIEnv *env;
 
+
     FONT_AWT_LOCK();
 /*     XTextExtents16(xFont, xChar, 1, &direction, &ascent, &descent, &xcs); */
     XQueryTextExtents16(awt_display,xFont->fid, xChar, 1,
                         &direction, &ascent, &descent, &xcs);
     width = xcs.rbearing - xcs.lbearing;
     height = xcs.ascent+xcs.descent;
     imageSize = width*height;
-
     glyphInfo = (GlyphInfo*)malloc(sizeof(GlyphInfo)+imageSize);
+    if (glyphInfo == NULL) {
+        AWT_UNLOCK();
+        return (jlong)(uintptr_t)NULL;
+    }
     glyphInfo->cellInfo = NULL;
     glyphInfo->width = width;
     glyphInfo->height = height;
     glyphInfo->topLeftX = xcs.lbearing;
     glyphInfo->topLeftY = -xcs.ascent;
