diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.h
@@ -31,11 +31,12 @@
 
 namespace WebCore {
 
 class MediaStreamTrack;
 
-class MediaStreamTrackEvent : public Event {
+class MediaStreamTrackEvent final : public Event {
+    WTF_MAKE_ISO_ALLOCATED(MediaStreamTrackEvent);
 public:
     virtual ~MediaStreamTrackEvent();
 
     static Ref<MediaStreamTrackEvent> create(const AtomString& type, CanBubble, IsCancelable, RefPtr<MediaStreamTrack>&&);
 
