diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/SimpleLineLayoutTextFragmentIterator.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/SimpleLineLayoutTextFragmentIterator.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/SimpleLineLayoutTextFragmentIterator.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/SimpleLineLayoutTextFragmentIterator.cpp
@@ -40,16 +40,18 @@
     , textAlign(style.textAlign())
     , hasKerningOrLigatures(font.enableKerning() || font.requiresShaping())
     , collapseWhitespace(style.collapseWhiteSpace())
     , preserveNewline(style.preserveNewline())
     , wrapLines(style.autoWrap())
+    , breakSpaces(style.whiteSpace() == WhiteSpace::BreakSpaces)
     , breakAnyWordOnOverflow(style.wordBreak() == WordBreak::BreakAll && wrapLines)
-    , breakFirstWordOnOverflow(breakAnyWordOnOverflow || (style.breakWords() && (wrapLines || preserveNewline)))
+    , breakWordOnOverflow(style.breakWords() && (wrapLines || preserveNewline))
+    , breakFirstWordOnOverflow(breakAnyWordOnOverflow || breakWordOnOverflow)
     , breakNBSP(wrapLines && style.nbspMode() == NBSPMode::Space)
     , keepAllWordsForCJK(style.wordBreak() == WordBreak::KeepAll)
     , wordSpacing(font.wordSpacing())
-    , tabWidth(collapseWhitespace ? 0 : style.tabSize())
+    , tabWidth(collapseWhitespace ? TabSize(0) : style.tabSize())
     , shouldHyphenate(style.hyphens() == Hyphens::Auto && canHyphenate(style.locale()))
     , hyphenStringWidth(shouldHyphenate ? font.width(TextRun(String(style.hyphenString()))) : 0)
     , hyphenLimitBefore(style.hyphenationLimitBefore() < 0 ? 2 : style.hyphenationLimitBefore())
     , hyphenLimitAfter(style.hyphenationLimitAfter() < 0 ? 2 : style.hyphenationLimitAfter())
     , locale(style.locale())
