diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/TextPainter.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/TextPainter.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/TextPainter.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/TextPainter.cpp
@@ -22,10 +22,11 @@
 
 #include "config.h"
 #include "TextPainter.h"
 
 #include "DisplayListReplayer.h"
+#include "FilterOperations.h"
 #include "GraphicsContext.h"
 #include "InlineTextBox.h"
 #include "RenderCombineText.h"
 #include "RenderLayer.h"
 #include "ShadowData.h"
@@ -96,11 +97,11 @@
 TextPainter::TextPainter(GraphicsContext& context)
     : m_context(context)
 {
 }
 
-void TextPainter::paintTextOrEmphasisMarks(const FontCascade& font, const TextRun& textRun, const AtomicString& emphasisMark,
+void TextPainter::paintTextOrEmphasisMarks(const FontCascade& font, const TextRun& textRun, const AtomString& emphasisMark,
     float emphasisMarkOffset, const FloatPoint& textOrigin, unsigned startOffset, unsigned endOffset)
 {
     ASSERT(startOffset < endOffset);
     if (!emphasisMark.isEmpty())
         m_context.drawEmphasisMarks(font, textRun, emphasisMark, textOrigin + FloatSize(0, emphasisMarkOffset), startOffset, endOffset);
@@ -115,11 +116,11 @@
     }
     m_glyphDisplayList = nullptr;
 }
 
 void TextPainter::paintTextWithShadows(const ShadowData* shadow, const FilterOperations* colorFilter, const FontCascade& font, const TextRun& textRun, const FloatRect& boxRect, const FloatPoint& textOrigin,
-    unsigned startOffset, unsigned endOffset, const AtomicString& emphasisMark, float emphasisMarkOffset, bool stroked)
+    unsigned startOffset, unsigned endOffset, const AtomString& emphasisMark, float emphasisMarkOffset, bool stroked)
 {
     if (!shadow) {
         paintTextOrEmphasisMarks(font, textRun, emphasisMark, emphasisMarkOffset, textOrigin, startOffset, endOffset);
         return;
     }
