diff a/src/java.base/unix/native/libnet/portconfig.c b/src/java.base/unix/native/libnet/portconfig.c
--- a/src/java.base/unix/native/libnet/portconfig.c
+++ b/src/java.base/unix/native/libnet/portconfig.c
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -58,17 +58,10 @@
             fclose(f);
             return ret == 2 ? 0 : -1;
         }
         return -1;
     }
-
-#elif defined(__solaris__)
-    {
-        range->higher = net_getParam("/dev/tcp", "tcp_largest_anon_port");
-        range->lower = net_getParam("/dev/tcp", "tcp_smallest_anon_port");
-        return 0;
-    }
 #elif defined(_ALLBSD_SOURCE)
     {
         int ret;
         size_t size = sizeof(range->lower);
         ret = sysctlbyname(
