diff a/src/hotspot/share/oops/klass.inline.hpp b/src/hotspot/share/oops/klass.inline.hpp
--- a/src/hotspot/share/oops/klass.inline.hpp
+++ b/src/hotspot/share/oops/klass.inline.hpp
@@ -30,11 +30,11 @@
 #include "oops/klass.hpp"
 #include "oops/markWord.hpp"
 #include "oops/oopHandle.inline.hpp"
 
 inline void Klass::set_prototype_header(markWord header) {
-  assert(!is_value() || header.is_always_locked(), "Unexpected prototype");
+  assert(!is_inline_klass() || header.is_always_locked(), "Unexpected prototype");
   assert(!header.has_bias_pattern() || is_instance_klass(), "biased locking currently only supported for Java instances");
   _prototype_header = header;
 }
 
 inline oop Klass::java_mirror() const {
