diff a/src/hotspot/share/oops/valueKlass.hpp b/src/hotspot/share/oops/valueKlass.hpp
--- a/src/hotspot/share/oops/valueKlass.hpp
+++ b/src/hotspot/share/oops/valueKlass.hpp
@@ -179,11 +179,11 @@
   // Specifically flat array klass
   Klass* value_array_klass(bool or_null, int rank, TRAPS);
 
  public:
   // Type testing
-  bool is_value_slow() const        { return true; }
+  bool is_inline_klass_slow() const        { return true; }
 
   // Casting from Klass*
   static ValueKlass* cast(Klass* k);
 
   // Use this to return the size of an instance in heap words
