diff a/make/data/tzdata/pacificnew b/make/data/tzdata/pacificnew
--- a/make/data/tzdata/pacificnew
+++ b/make/data/tzdata/pacificnew
@@ -42,11 +42,11 @@
 ## Rule	Twilite	XXXX	max	-	Apr	Sun>=1	2:00	1:00	D
 ## Rule	Twilite	XXXX	max	uspres	Oct	lastSun	2:00	1:00	PE
 ## Rule	Twilite	XXXX	max	uspres	Nov	Sun>=7	2:00	0	S
 ## Rule	Twilite	XXXX	max	nonpres	Oct	lastSun	2:00	0	S
 
-# Zone	NAME			GMTOFF	RULES/SAVE	FORMAT	[UNTIL]
+# Zone	NAME			STDOFF	RULES/SAVE	FORMAT	[UNTIL]
 ## Zone	America/Los_Angeles-PET	-8:00	US		P%sT	XXXX
 ##				-8:00	Twilite		P%sT
 
 # For now...
 Link	America/Los_Angeles	US/Pacific-New	##
