diff a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/MessageQueueBackend.java b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/MessageQueueBackend.java
--- a/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/MessageQueueBackend.java
+++ b/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/MessageQueueBackend.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -34,12 +34,12 @@
   // The two queues
   private MessageQueueImpl leftRightQueue;
   private MessageQueueImpl rightLeftQueue;
 
   public MessageQueueBackend() {
-    LinkedList leftRightPipe = new LinkedList();
-    LinkedList rightLeftPipe = new LinkedList();
+    LinkedList<Object> leftRightPipe = new LinkedList<>();
+    LinkedList<Object> rightLeftPipe = new LinkedList<>();
     leftRightQueue = new MessageQueueImpl(rightLeftPipe, leftRightPipe);
     rightLeftQueue = new MessageQueueImpl(leftRightPipe, rightLeftPipe);
   }
 
   /** Get one of the two symmetric sides of this message queue. */
@@ -51,14 +51,14 @@
   public MessageQueue getSecondQueue() {
     return rightLeftQueue;
   }
 
   private class MessageQueueImpl implements MessageQueue {
-    private LinkedList readList;
-    private LinkedList writeList;
+    private LinkedList<Object> readList;
+    private LinkedList<Object> writeList;
 
-    public MessageQueueImpl(LinkedList listToReadFrom, LinkedList listToWriteTo) {
+    public MessageQueueImpl(LinkedList<Object> listToReadFrom, LinkedList<Object> listToWriteTo) {
       readList = listToReadFrom;
       writeList = listToWriteTo;
     }
 
     public Object readMessage() {
