diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListBridgeBotFactory.java
@@ -119,11 +119,11 @@
             if (repoConfig.contains("branchname")) {
                 botBuilder.branchInSubject(Pattern.compile(repoConfig.get("branchname").asString()));
             }
             ret.add(botBuilder.build());
 
-            if (!repoConfig.contains("bidirectional") || !repoConfig.get("bidirectional").asBoolean()) {
+            if (!repoConfig.contains("bidirectional") || repoConfig.get("bidirectional").asBoolean()) {
                 listNamesForReading.add(list);
             }
             allRepositories.add(configuration.repository(repo));
         }
 
