diff a/test/make/TestCopyFiles.gmk b/test/make/TestCopyFiles.gmk
--- a/test/make/TestCopyFiles.gmk
+++ b/test/make/TestCopyFiles.gmk
@@ -1,7 +1,7 @@
 #
-# Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2017, 2019, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.  Oracle designates this
@@ -61,11 +61,11 @@
 	$(TOUCH) $@
 
 $(eval $(call SetupCopyFiles, COPY_1, \
     SRC := $(SRC_DIR), \
     DEST := $(DEST_DIR), \
-    FILES := $(call CacheFind, $(SRC_DIR)), \
+    FILES := $(call ShellFindFiles, $(SRC_DIR)), \
 ))
 
 # Optionally define a rule that deletes all the target files after the makefile
 # has been parsed. GNU make has specific problems with this in combination with
 # spaces in directory names.
