diff a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java
--- a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java
+++ b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java
@@ -316,25 +316,10 @@
                     "This PR has been reviewed by " +
                     formatReviewer(added.getKey().reviewer()) + " - " +
                     verdictToString(added.getKey().verdict()) + ".";
             pr.addComment(body);
         }
-
-        for (var updated : reviewTracker.updatedReviews().entrySet()) {
-            var body = updated.getValue() + "\n" +
-                    "The PR review by " + formatReviewer(updated.getKey().reviewer()) +
-                    " has been updated - " + verdictToString(updated.getKey().verdict()) + ".";
-            pr.addComment(body);
-        }
-
-        for (var removed : reviewTracker.removedReviews().entrySet()) {
-            var user = pr.repository().host().getUserDetails(removed.getKey());
-            var body = removed.getValue() + "\n" +
-                    "This PR is no longer reviewed by " +
-                    formatReviewer(user) + ".";
-            pr.addComment(body);
-        }
     }
 
     private Optional<Comment> findComment(List<Comment> comments, String marker) {
         var self = pr.repository().host().getCurrentUserDetails();
         return comments.stream()
