diff a/test/make/autoconf/test.m4 b/test/make/autoconf/test.m4
--- a/test/make/autoconf/test.m4
+++ b/test/make/autoconf/test.m4
@@ -44,17 +44,63 @@
   UTIL_GET_MATCHING_VALUES(ACTUAL, $FEW, $MANY)
   UTIL_ASSERT_STRING_EQUALS($ACTUAL, $EXPECTED_MATCHING, \
       [UTIL_GET_MATCHING_VALUES failed])
 ])
 
+AC_DEFUN([TEST_ARG_ENABLE],
+[
+  # fake '--enable-default-false=yes' on the command line
+  enable_default_false=yes
+  UTIL_ARG_ENABLE(NAME: default-false, DEFAULT: false, RESULT: TEST_RESULT)
+  UTIL_ASSERT_TRUE($TEST_RESULT)
+
+  # assume '--enable-default-true-but-unavailable=yes' not set
+  UTIL_ARG_ENABLE(NAME: default-true-but-unavailable, DEFAULT: auto, RESULT: TEST_RESULT, AVAILABLE: false)
+  UTIL_ASSERT_NOT_TRUE($TEST_RESULT)
+
+  enable_test_given_yes=yes
+  given=false
+  enabled=false
+  disabled=false
+  UTIL_ARG_ENABLE(NAME: test-given-yes, DEFAULT: false,
+      IF_GIVEN: [ given=true ], IF_ENABLED: [ enabled=true ],
+      IF_DISABLED: [ disabled=true ])
+  UTIL_ASSERT_TRUE($given)
+  UTIL_ASSERT_TRUE($enabled)
+  UTIL_ASSERT_NOT_TRUE($disabled)
+
+  enable_test_given_default=no
+  given=false
+  enabled=false
+  disabled=false
+  UTIL_ARG_ENABLE(NAME: test-given-default, DEFAULT: true,
+      IF_GIVEN: [ given=true ], IF_ENABLED: [ enabled=true ],
+      IF_DISABLED: [ disabled=true ])
+  UTIL_ASSERT_TRUE($given)
+  UTIL_ASSERT_NOT_TRUE($enabled)
+  UTIL_ASSERT_TRUE($disabled)
+
+  # assume '--enable-test-given-no' not set
+  given=false
+  enabled=false
+  disabled=false
+  UTIL_ARG_ENABLE(NAME: test-given-no, DEFAULT: true,
+      IF_GIVEN: [ given=true ], IF_ENABLED: [ enabled=true ],
+      IF_DISABLED: [ disabled=true ])
+  UTIL_ASSERT_NOT_TRUE($given)
+  UTIL_ASSERT_TRUE($enabled)
+  UTIL_ASSERT_NOT_TRUE($disabled)
+])
+
 # Use the CUSTOM_EARLY_HOOK to inject our test after basic init is done.
 AC_DEFUN_ONCE([CUSTOM_EARLY_HOOK],
 [
   $PRINTF "\nStarting configure tests\n"
   $PRINTF "==============================\n"
 
   TEST_STRING_OPS
+  TEST_ARG_ENABLE
 
   # If no assertions failed, report success
   $PRINTF "==============================\n"
   $PRINTF "Configure tests finished successfully\n\n"
   exit 0
