diff a/modules/javafx.web/src/main/native/Source/WebCore/editing/EditAction.cpp b/modules/javafx.web/src/main/native/Source/WebCore/editing/EditAction.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/editing/EditAction.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/editing/EditAction.cpp
@@ -25,10 +25,11 @@
 
 #include "config.h"
 #include "EditAction.h"
 
 #include "LocalizedStrings.h"
+#include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
 String undoRedoLabel(EditAction editAction)
 {
@@ -72,11 +73,12 @@
         return WEB_UI_STRING_KEY("Align Right", "Align Right (Undo action name)", "Undo action name");
     case EditAction::Center:
         return WEB_UI_STRING_KEY("Center", "Center (Undo action name)", "Undo action name");
     case EditAction::Justify:
         return WEB_UI_STRING_KEY("Justify", "Justify (Undo action name)", "Undo action name");
-    case EditAction::SetWritingDirection:
+    case EditAction::SetInlineWritingDirection:
+    case EditAction::SetBlockWritingDirection:
         return WEB_UI_STRING_KEY("Set Writing Direction", "Set Writing Direction (Undo action name)", "Undo action name");
     case EditAction::Subscript:
         return WEB_UI_STRING_KEY("Subscript", "Subscript (Undo action name)", "Undo action name");
     case EditAction::Superscript:
         return WEB_UI_STRING_KEY("Superscript", "Superscript (Undo action name)", "Undo action name");
