diff a/modules/javafx.web/src/main/native/Source/WebCore/editing/TextCheckingHelper.h b/modules/javafx.web/src/main/native/Source/WebCore/editing/TextCheckingHelper.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/editing/TextCheckingHelper.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/editing/TextCheckingHelper.h
@@ -109,7 +109,8 @@
 };
 
 void checkTextOfParagraph(TextCheckerClient&, StringView, OptionSet<TextCheckingType>, Vector<TextCheckingResult>&, const VisibleSelection& currentSelection);
 
 bool unifiedTextCheckerEnabled(const Frame*);
+bool platformDrivenTextCheckerEnabled();
 
 } // namespace WebCore
