diff a/modules/javafx.web/src/main/native/Source/WebCore/editing/TypingCommand.cpp b/modules/javafx.web/src/main/native/Source/WebCore/editing/TypingCommand.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/editing/TypingCommand.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/editing/TypingCommand.cpp
@@ -423,11 +423,11 @@
     if (m_currentTypingEditAction != EditAction::InsertReplacement || isEditingTextAreaOrTextInput())
         return nullptr;
 
     StringBuilder htmlText;
     MarkupAccumulator::appendCharactersReplacingEntities(htmlText, m_currentTextToInsert, 0, m_currentTextToInsert.length(), EntityMaskInHTMLPCDATA);
-    return DataTransfer::createForInputEvent(m_currentTextToInsert, htmlText.toString());
+    return DataTransfer::createForInputEvent(document(), m_currentTextToInsert, htmlText.toString());
 }
 
 void TypingCommand::didApplyCommand()
 {
     // TypingCommands handle applied editing separately (see TypingCommand::typingAddedToOpenCommand).
