diff a/src/java.base/share/classes/sun/net/spi/DefaultProxySelector.java b/src/java.base/share/classes/sun/net/spi/DefaultProxySelector.java
--- a/src/java.base/share/classes/sun/net/spi/DefaultProxySelector.java
+++ b/src/java.base/share/classes/sun/net/spi/DefaultProxySelector.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -383,11 +383,13 @@
      * @param disjunct non-null mask disjunct
      * @return java regex string corresponding to this mask
      */
     static String disjunctToRegex(String disjunct) {
         String regex;
-        if (disjunct.startsWith("*") && disjunct.endsWith("*")) {
+        if (disjunct.equals("*")) {
+            regex = ".*";
+        } else if (disjunct.startsWith("*") && disjunct.endsWith("*")) {
             regex = ".*" + quote(disjunct.substring(1, disjunct.length() - 1)) + ".*";
         } else if (disjunct.startsWith("*")) {
             regex = ".*" + quote(disjunct.substring(1));
         } else if (disjunct.endsWith("*")) {
             regex = quote(disjunct.substring(0, disjunct.length() - 1)) + ".*";
