diff a/make/jdk/src/classes/build/tools/jdwpgen/AbstractNamedNode.java b/make/jdk/src/classes/build/tools/jdwpgen/AbstractNamedNode.java
--- a/make/jdk/src/classes/build/tools/jdwpgen/AbstractNamedNode.java
+++ b/make/jdk/src/classes/build/tools/jdwpgen/AbstractNamedNode.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -60,12 +60,12 @@
         nameNode.constrain(ctx);
         super.constrain(ctx.subcontext(name));
     }
 
     void document(PrintWriter writer) {
-        writer.println("<h4 id=\"" + name + "\">" + name +
-                       " Command Set</h4>");
+        writer.println("<h2 id=\"" + name + "\">" + name +
+                       " Command Set</h2>");
         for (Node node : components) {
             node.document(writer);
         }
     }
 
