diff a/make/jdk/src/classes/build/tools/jdwpgen/ErrorNode.java b/make/jdk/src/classes/build/tools/jdwpgen/ErrorNode.java
--- a/make/jdk/src/classes/build/tools/jdwpgen/ErrorNode.java
+++ b/make/jdk/src/classes/build/tools/jdwpgen/ErrorNode.java
@@ -53,13 +53,14 @@
 
         String com = comment();
         if (com == null || com.length() == 0) {
             com = ConstantSetNode.getConstant("Error_" + name);
         }
-        writer.println("<tr><td>" + "<a href=\"#" + NAME_OF_ERROR_TABLE + "_" + name + "\">"
-                       + name + "</a></td>" +
-                       "<td>" + com + "&nbsp;</td></tr>");
+        writer.println("<tr>"
+                + "<th scope=\"row\">" + "<a href=\"#" + NAME_OF_ERROR_TABLE + "_" + name + "\">" + name + "</a>"
+                + "<td>" + com + "&nbsp;"
+                + "</tr>");
     }
 
     void genJavaComment(PrintWriter writer, int depth) {}
 
     void genJava(PrintWriter writer, int depth) {}
