diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/CallTracerTypes.h b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/CallTracerTypes.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/CallTracerTypes.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/CallTracerTypes.h
@@ -39,49 +39,71 @@
 #include "TypedOMCSSImageValue.h"
 #include <JavaScriptCore/ArrayBuffer.h>
 #include <JavaScriptCore/ArrayBufferView.h>
 #include <JavaScriptCore/Float32Array.h>
 #include <JavaScriptCore/Int32Array.h>
+#include <JavaScriptCore/Uint32Array.h>
+#include <wtf/Optional.h>
 #include <wtf/RefPtr.h>
 #include <wtf/Variant.h>
 #include <wtf/Vector.h>
 #include <wtf/text/WTFString.h>
 
 #if ENABLE(WEBGL)
 #include "WebGLBuffer.h"
 #include "WebGLFramebuffer.h"
 #include "WebGLProgram.h"
+#include "WebGLQuery.h"
 #include "WebGLRenderbuffer.h"
+#include "WebGLRenderingContextBase.h"
+#include "WebGLSampler.h"
 #include "WebGLShader.h"
+#include "WebGLSync.h"
 #include "WebGLTexture.h"
+#include "WebGLTransformFeedback.h"
 #include "WebGLUniformLocation.h"
+#include "WebGLVertexArrayObject.h"
+#endif
+
+#if ENABLE(WEBGL2)
+#include "WebGL2RenderingContext.h"
 #endif
 
 namespace WebCore {
 
-typedef Variant<
+using RecordCanvasActionVariant = Variant<
+    // enum
     CanvasDirection,
     CanvasFillRule,
     CanvasLineCap,
     CanvasLineJoin,
     CanvasTextAlign,
     CanvasTextBaseline,
+    ImageSmoothingQuality,
+
+    // struct
     DOMMatrix2DInit,
+
+    // pointer
     Element*,
     HTMLImageElement*,
     ImageBitmap*,
     ImageData*,
-    ImageSmoothingQuality,
     Path2D*,
 #if ENABLE(WEBGL)
     WebGLBuffer*,
     WebGLFramebuffer*,
     WebGLProgram*,
+    WebGLQuery*,
     WebGLRenderbuffer*,
+    WebGLSampler*,
     WebGLShader*,
+    WebGLSync*,
     WebGLTexture*,
+    WebGLTransformFeedback*,
     WebGLUniformLocation*,
+    WebGLVertexArrayObject*,
 #endif
     RefPtr<ArrayBuffer>,
     RefPtr<ArrayBufferView>,
     RefPtr<CanvasGradient>,
     RefPtr<CanvasPattern>,
@@ -90,23 +112,49 @@
     RefPtr<HTMLImageElement>,
 #if ENABLE(VIDEO)
     RefPtr<HTMLVideoElement>,
 #endif
     RefPtr<ImageBitmap>,
-#if ENABLE(CSS_PAINTING_API)
+#if ENABLE(CSS_TYPED_OM)
     RefPtr<TypedOMCSSImageValue>,
 #endif
     RefPtr<ImageData>,
     RefPtr<Int32Array>,
+    RefPtr<Uint32Array>,
+
+    // variant
+    CanvasImageSource,
+    CanvasRenderingContext2DBase::Style,
+#if ENABLE(WEBGL)
+    WebGLRenderingContextBase::BufferDataSource,
+    Optional<WebGLRenderingContextBase::BufferDataSource>,
+    WebGLRenderingContextBase::TexImageSource,
+    Optional<WebGLRenderingContextBase::TexImageSource>,
+#endif
+
+    // array
+    Vector<String>,
     Vector<float>,
-    Vector<int>,
+    Vector<uint32_t>,
+    Vector<int32_t>,
+#if ENABLE(WEBGL)
+    WebGLRenderingContextBase::Float32List::VariantType,
+    WebGLRenderingContextBase::Int32List::VariantType,
+#endif
+#if ENABLE(WEBGL2)
+    WebGL2RenderingContext::Uint32List::VariantType,
+#endif
+
+    // basic
     String,
     double,
     float,
+    Optional<float>,
+    uint64_t,
     int64_t,
     uint32_t,
     int32_t,
     uint8_t,
     bool
-> RecordCanvasActionVariant;
+>;
 
 } // namespace WebCore
