diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSCustomXPathNSResolver.cpp b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSCustomXPathNSResolver.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSCustomXPathNSResolver.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSCustomXPathNSResolver.cpp
@@ -1,7 +1,8 @@
 /*
  * Copyright (C) 2007, 2013 Alexey Proskuryakov (ap@nypop.com)
+ * Copyright (C) 2019 Apple Inc. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
  * are met:
  * 1. Redistributions of source code must retain the above copyright
@@ -70,11 +71,11 @@
     JSLockHolder lock(commonVM());
 
     ExecState* exec = m_globalObject->globalExec();
     VM& vm = exec->vm();
 
-    JSValue function = m_customResolver->get(exec, Identifier::fromString(exec, "lookupNamespaceURI"));
+    JSValue function = m_customResolver->get(exec, Identifier::fromString(vm, "lookupNamespaceURI"));
     CallData callData;
     CallType callType = getCallData(vm, function, callData);
     if (callType == CallType::None) {
         callType = m_customResolver->methodTable(vm)->getCallData(m_customResolver.get(), callData);
         if (callType == CallType::None) {
