diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSHistoryCustom.cpp b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSHistoryCustom.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSHistoryCustom.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSHistoryCustom.cpp
@@ -28,10 +28,11 @@
 
 #include "config.h"
 #include "JSHistory.h"
 
 #include "SerializedScriptValue.h"
+#include <JavaScriptCore/JSCInlines.h>
 
 namespace WebCore {
 
 using namespace JSC;
 
@@ -43,9 +44,9 @@
     });
 }
 
 void JSHistory::visitAdditionalChildren(SlotVisitor& visitor)
 {
-    wrapped().cachedState().visit(visitor);
+    wrapped().cachedStateForGC().visit(visitor);
 }
 
 } // namespace WebCore
