diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/WebCoreBuiltinNames.h b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/WebCoreBuiltinNames.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/WebCoreBuiltinNames.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/WebCoreBuiltinNames.h
@@ -1,8 +1,8 @@
 /*
  *  Copyright (c) 2015, Canon Inc. All rights reserved.
- *  Copyright (C) 2018 Apple Inc. All rights reserved.
+ *  Copyright (C) 2018-2019 Apple Inc. All rights reserved.
  *
  *  Redistribution and use in source and binary forms, with or without
  *  modification, are permitted provided that the following conditions
  *  are met:
  *  1.  Redistributions of source code must retain the above copyright
@@ -79,15 +79,43 @@
     macro(FileSystemEntry) \
     macro(FileSystemFileEntry) \
     macro(Gamepad) \
     macro(GamepadButton) \
     macro(GamepadEvent) \
+    macro(GPU) \
+    macro(GPUAdapter) \
+    macro(GPUBindGroup) \
+    macro(GPUBindGroupLayout) \
+    macro(GPUBuffer) \
+    macro(GPUBufferUsage) \
+    macro(GPUCanvasContext) \
+    macro(GPUColorWriteBits) \
+    macro(GPUCommandBuffer) \
+    macro(GPUCommandEncoder) \
+    macro(GPUComputePassEncoder) \
+    macro(GPUComputePipeline) \
+    macro(GPUDevice) \
+    macro(GPUOutOfMemoryError) \
+    macro(GPUPipelineLayout) \
+    macro(GPUProgrammablePassEncoder) \
+    macro(GPUQueue) \
+    macro(GPURenderPassEncoder) \
+    macro(GPURenderPipeline) \
+    macro(GPUSampler) \
+    macro(GPUShaderModule) \
+    macro(GPUShaderStageBit) \
+    macro(GPUSwapChain) \
+    macro(GPUTexture) \
     macro(GPUTextureUsage) \
+    macro(GPUTextureView) \
+    macro(GPUValidationError) \
     macro(HTMLAttachmentElement) \
     macro(HTMLAudioElement) \
+    macro(HTMLDialogElement) \
     macro(HTMLDataListElement) \
     macro(HTMLMenuItemElement) \
+    macro(HTMLKeygenElement) \
     macro(HTMLSlotElement) \
     macro(Headers) \
     macro(IDBCursor) \
     macro(IDBCursorWithValue) \
     macro(IDBDatabase) \
@@ -144,10 +172,12 @@
     macro(PerformanceObserverEntryList) \
     macro(PerformanceResourceTiming) \
     macro(PerformanceServerTiming) \
     macro(PointerEvent) \
     macro(PublicKeyCredential) \
+    macro(ResizeObserver) \
+    macro(ResizeObserverEntry) \
     macro(RTCCertificate) \
     macro(RTCDTMFSender) \
     macro(RTCDTMFToneChangeEvent) \
     macro(RTCDataChannel) \
     macro(RTCDataChannelEvent) \
@@ -189,54 +219,10 @@
     macro(VRFrameData) \
     macro(VRStageParameters) \
     macro(VisualViewport) \
     macro(WebGL2RenderingContext) \
     macro(WebGLVertexArrayObject) \
-    macro(WebGPU) \
-    macro(WebGPUAdapter) \
-    macro(WebGPUBindGroup) \
-    macro(WebGPUBindGroupLayout) \
-    macro(WebGPUBuffer) \
-    macro(WebGPUBufferUsage) \
-    macro(WebGPUCommandBuffer) \
-    macro(WebGPUDevice) \
-    macro(WebGPUIndexFormat) \
-    macro(WebGPUInputStepMode) \
-    macro(WebGPUQueue) \
-    macro(WebGPUPipelineLayout) \
-    macro(WebGPUProgrammablePassEncoder) \
-    macro(WebGPURenderingContext) \
-    macro(WebGPURenderPassEncoder) \
-    macro(WebGPURenderPipeline) \
-    macro(WebGPUShaderModule) \
-    macro(WebGPUShaderStageBit) \
-    macro(WebGPUSwapChain) \
-    macro(WebGPUTexture) \
-    macro(WebGPUTextureView) \
-    macro(WebGPUVertexFormat) \
-    macro(WebMetalBuffer) \
-    macro(WebMetalCommandBuffer) \
-    macro(WebMetalCommandQueue) \
-    macro(WebMetalComputeCommandEncoder) \
-    macro(WebMetalComputePipelineState) \
-    macro(WebMetalDepthStencilDescriptor) \
-    macro(WebMetalDepthStencilState) \
-    macro(WebMetalDrawable) \
-    macro(WebMetalFunction) \
-    macro(WebMetalLibrary) \
-    macro(WebMetalRenderCommandEncoder) \
-    macro(WebMetalRenderPassAttachmentDescriptor) \
-    macro(WebMetalRenderPassColorAttachmentDescriptor) \
-    macro(WebMetalRenderPassDepthAttachmentDescriptor) \
-    macro(WebMetalRenderPassDescriptor) \
-    macro(WebMetalRenderPipelineColorAttachmentDescriptor) \
-    macro(WebMetalRenderPipelineDescriptor) \
-    macro(WebMetalRenderPipelineState) \
-    macro(WebMetalRenderingContext) \
-    macro(WebMetalSize) \
-    macro(WebMetalTexture) \
-    macro(WebMetalTextureDescriptor) \
     macro(WebKitMediaKeyError) \
     macro(WebKitMediaKeyMessageEvent) \
     macro(WebKitMediaKeyNeededEvent) \
     macro(WebKitMediaKeySession) \
     macro(WebKitMediaKeys) \
@@ -277,10 +263,11 @@
     macro(finishConsumingStream) \
     macro(focus) \
     macro(frames) \
     macro(getTracks) \
     macro(getUserMedia) \
+    macro(gpu) \
     macro(header) \
     macro(href) \
     macro(indexedDB) \
     macro(initializeWith) \
     macro(isDisturbed) \
@@ -290,11 +277,15 @@
     macro(location) \
     macro(makeGetterTypeError) \
     macro(makeThisTypeError) \
     macro(matchingElementInFlatTree) \
     macro(mediaStreamTrackConstraints) \
-    macro(openDatabase) \
+    macro(ontouchcancel) \
+    macro(ontouchend) \
+    macro(ontouchmove) \
+    macro(ontouchstart) \
+    macro(ontouchforcechange) \
     macro(onvrdisplayactivate) \
     macro(onvrdisplayblur) \
     macro(onvrdisplayconnect) \
     macro(onvrdisplaydeactivate) \
     macro(onvrdisplaydisconnect) \
@@ -360,19 +351,18 @@
     macro(webkitIDBKeyRange) \
     macro(webkitIDBObjectStore) \
     macro(webkitIDBRequest) \
     macro(webkitIDBTransaction) \
     macro(webkitIndexedDB) \
-    macro(webgpu) \
     macro(window) \
     macro(writing) \
     WEBCORE_ADDITIONAL_PRIVATE_IDENTIFIERS(macro) \
 
 class WebCoreBuiltinNames {
 public:
-    explicit WebCoreBuiltinNames(JSC::VM* vm)
-        : m_vm(*vm)
+    explicit WebCoreBuiltinNames(JSC::VM& vm)
+        : m_vm(vm)
         WEBCORE_COMMON_PRIVATE_IDENTIFIERS_EACH_PROPERTY_NAME(INITIALIZE_BUILTIN_NAMES)
     {
 #define EXPORT_NAME(name) m_vm.propertyNames->appendExternalName(name##PublicName(), name##PrivateName());
         WEBCORE_COMMON_PRIVATE_IDENTIFIERS_EACH_PROPERTY_NAME(EXPORT_NAME)
 #undef EXPORT_NAME
