diff a/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/SaveError.java b/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/SaveError.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/SaveError.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/SaveError.java
@@ -35,14 +35,14 @@
 import static jdk.internal.clang.libclang.Index_h.CXSaveError_None;
 import static jdk.internal.clang.libclang.Index_h.CXSaveError_TranslationErrors;
 import static jdk.internal.clang.libclang.Index_h.CXSaveError_Unknown;
 
 public enum SaveError {
-    None(CXSaveError_None),
-    Unknown(CXSaveError_Unknown),
-    TranslationErrors(CXSaveError_TranslationErrors),
-    InvalidTU(CXSaveError_InvalidTU);
+    None(CXSaveError_None()),
+    Unknown(CXSaveError_Unknown()),
+    TranslationErrors(CXSaveError_TranslationErrors()),
+    InvalidTU(CXSaveError_InvalidTU());
 
     private final int code;
 
     SaveError(int code) {
         this.code = code;
