diff a/test/jdk/java/net/MulticastSocket/JoinLeave.java b/test/jdk/java/net/MulticastSocket/JoinLeave.java
--- a/test/jdk/java/net/MulticastSocket/JoinLeave.java
+++ b/test/jdk/java/net/MulticastSocket/JoinLeave.java
@@ -26,10 +26,11 @@
 import java.net.InetAddress;
 import java.net.MulticastSocket;
 import java.net.NetworkInterface;
 
 import jdk.test.lib.NetworkConfiguration;
+import jdk.test.lib.net.IPSupport;
 
 /**
  * @test
  * @bug 4091811 4148753 4102731
  * @summary Test java.net.MulticastSocket joinGroup and leaveGroup
@@ -40,10 +41,11 @@
  * @run main/othervm -Djava.net.preferIPv4Stack=true JoinLeave
  */
 public class JoinLeave {
 
     public static void main(String args[]) throws IOException {
+        IPSupport.throwSkippedExceptionIfNonOperational();
         InetAddress ip4Group = InetAddress.getByName("224.80.80.80");
         InetAddress ip6Group = InetAddress.getByName("ff02::a");
 
         NetworkConfiguration nc = NetworkConfiguration.probe();
         nc.ip4MulticastInterfaces().forEach(nic -> joinLeave(ip4Group, nic));
