diff a/test/jdk/java/net/MulticastSocket/Reuse.java b/test/jdk/java/net/MulticastSocket/Reuse.java
--- a/test/jdk/java/net/MulticastSocket/Reuse.java
+++ b/test/jdk/java/net/MulticastSocket/Reuse.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -22,10 +22,15 @@
  */
 
 import java.net.MulticastSocket;
 import java.net.BindException;
 
+/*
+ * @test
+ * @summary Check if MulticastSocket sets SO_REUSEADDR
+ */
+
 public class Reuse {
     public static void main(String[] args) throws Exception {
         MulticastSocket s1, s2;
 
         try {
