diff a/test/jdk/java/net/MulticastSocket/SetGetNetworkInterfaceTest.java b/test/jdk/java/net/MulticastSocket/SetGetNetworkInterfaceTest.java
--- a/test/jdk/java/net/MulticastSocket/SetGetNetworkInterfaceTest.java
+++ b/test/jdk/java/net/MulticastSocket/SetGetNetworkInterfaceTest.java
@@ -25,10 +25,11 @@
 import java.io.UncheckedIOException;
 import java.net.MulticastSocket;
 import java.net.NetworkInterface;
 
 import jdk.test.lib.NetworkConfiguration;
+import jdk.test.lib.net.IPSupport;
 
 /**
  * @test
  * @bug 6458027
  * @summary Disabling IPv6 on a specific network interface causes problems.
@@ -39,10 +40,11 @@
  * @run main/othervm -Djava.net.preferIPv4Stack=true SetGetNetworkInterfaceTest
 */
 public class SetGetNetworkInterfaceTest {
 
     public static void main(String[] args) throws Exception {
+        IPSupport.throwSkippedExceptionIfNonOperational();
         NetworkConfiguration nc = NetworkConfiguration.probe();
         try (MulticastSocket ms = new MulticastSocket()) {
             nc.multicastInterfaces(true).forEach(nif -> setGetNetworkInterface(ms, nif));
         } catch (IOException e) {
             e.printStackTrace();
