diff a/modules/javafx.web/src/main/native/Source/WebCore/workers/WorkerLoaderProxy.h b/modules/javafx.web/src/main/native/Source/WebCore/workers/WorkerLoaderProxy.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/workers/WorkerLoaderProxy.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/workers/WorkerLoaderProxy.h
@@ -43,10 +43,12 @@
 // context can be distinct.
 class WorkerLoaderProxy {
 public:
     virtual ~WorkerLoaderProxy() = default;
 
+    virtual bool isWorkerMessagingProxy() const { return false; }
+
     // Creates a cache storage connection to be used on the main thread. Method must be called on the main thread.
     virtual Ref<CacheStorageConnection> createCacheStorageConnection() = 0;
 
     // Posts a task to the thread which runs the loading code (normally, the main thread).
     virtual void postTaskToLoader(ScriptExecutionContext::Task&&) = 0;
