diff a/src/hotspot/share/logging/logDecorations.hpp b/src/hotspot/share/logging/logDecorations.hpp
--- a/src/hotspot/share/logging/logDecorations.hpp
+++ b/src/hotspot/share/logging/logDecorations.hpp
@@ -34,23 +34,20 @@
  private:
   char _decorations_buffer[DecorationsBufferSize];
   char* _decoration_offset[LogDecorators::Count];
   LogLevelType _level;
   const LogTagSet& _tagset;
-  static jlong _vm_start_time_millis;
   static const char* volatile _host_name;
 
   const char* host_name();
   void create_decorations(const LogDecorators& decorators);
 
 #define DECORATOR(name, abbr) char* create_##name##_decoration(char* pos);
   DECORATOR_LIST
 #undef DECORATOR
 
  public:
-  static void initialize(jlong vm_start_time);
-
   LogDecorations(LogLevelType level, const LogTagSet& tagset, const LogDecorators& decorators);
 
   void set_level(LogLevelType level) {
     _level = level;
   }
