diff a/test/hotspot/jtreg/gc/arguments/TestSurvivorAlignmentInBytesOption.java b/test/hotspot/jtreg/gc/arguments/TestSurvivorAlignmentInBytesOption.java
--- a/test/hotspot/jtreg/gc/arguments/TestSurvivorAlignmentInBytesOption.java
+++ b/test/hotspot/jtreg/gc/arguments/TestSurvivorAlignmentInBytesOption.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2014, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -37,11 +37,11 @@
  *           & vm.opt.UnlockExperimentalVMOptions == null
  *           & (vm.opt.IgnoreUnrecognizedVMOptions == null
  *              | vm.opt.IgnoreUnrecognizedVMOptions == "false")
  * @modules java.base/jdk.internal.misc
  *          java.management
- * @run main gc.arguments.TestSurvivorAlignmentInBytesOption
+ * @run driver gc.arguments.TestSurvivorAlignmentInBytesOption
  */
 public class TestSurvivorAlignmentInBytesOption {
     public static void main(String args[]) throws Throwable {
         String optionName = "SurvivorAlignmentInBytes";
         String unlockExperimentalVMOpts = "UnlockExperimentalVMOptions";
