diff a/test/hotspot/jtreg/gc/arguments/TestSelectDefaultGC.java b/test/hotspot/jtreg/gc/arguments/TestSelectDefaultGC.java
--- a/test/hotspot/jtreg/gc/arguments/TestSelectDefaultGC.java
+++ b/test/hotspot/jtreg/gc/arguments/TestSelectDefaultGC.java
@@ -43,19 +43,16 @@
     public static void assertVMOption(OutputAnalyzer output, String option, boolean value) {
         output.shouldMatch(" " + option + " .*=.* " + value + " ");
     }
 
     public static void testDefaultGC(boolean actAsServer) throws Exception {
-        String[] args = new String[] {
-          "-XX:" + (actAsServer ? "+" : "-") + "AlwaysActAsServerClassMachine",
-          "-XX:" + (actAsServer ? "-" : "+") + "NeverActAsServerClassMachine",
-          "-XX:+PrintFlagsFinal",
-          "-version"
-        };
-
         // Start VM without specifying GC
-        ProcessBuilder pb = GCArguments.createJavaProcessBuilder(args);
+        ProcessBuilder pb = GCArguments.createJavaProcessBuilder(
+            "-XX:" + (actAsServer ? "+" : "-") + "AlwaysActAsServerClassMachine",
+            "-XX:" + (actAsServer ? "-" : "+") + "NeverActAsServerClassMachine",
+            "-XX:+PrintFlagsFinal",
+            "-version");
         OutputAnalyzer output = new OutputAnalyzer(pb.start());
         output.shouldHaveExitValue(0);
 
         final boolean isServer = actAsServer;
 
