diff a/test/hotspot/jtreg/compiler/c1/Test6579789.java b/test/hotspot/jtreg/compiler/c1/Test6579789.java
--- a/test/hotspot/jtreg/compiler/c1/Test6579789.java
+++ b/test/hotspot/jtreg/compiler/c1/Test6579789.java
@@ -24,11 +24,11 @@
 /**
  * @test
  * @bug 6579789
  * @summary Internal error "c1_LinearScan.cpp:1429 Error: assert(false,"")" in debuggee with fastdebug VM
  *
- * @run main/othervm -Xcomp -XX:UseSSE=0
+ * @run main/othervm -Xcomp -XX:+IgnoreUnrecognizedVMOptions -XX:UseSSE=0
  *      -XX:CompileCommand=compileonly,compiler.c1.Test6579789::bug
  *      compiler.c1.Test6579789
  */
 
 package compiler.c1;
