diff a/src/jdk.jdi/share/classes/com/sun/tools/jdi/ValueImpl.java b/src/jdk.jdi/share/classes/com/sun/tools/jdi/ValueImpl.java
--- a/src/jdk.jdi/share/classes/com/sun/tools/jdi/ValueImpl.java
+++ b/src/jdk.jdi/share/classes/com/sun/tools/jdi/ValueImpl.java
@@ -42,10 +42,13 @@
         if (value == null) {
             JNITypeParser sig = new JNITypeParser(destination.signature());
             if (sig.isPrimitive()) {
                 throw new InvalidTypeException("Can't set a primitive type to null");
             }
+            if (destination.signature().charAt(0) == 'Q') {
+                throw new InvalidTypeException("Can't set an inline type to null");
+            }
             return null;    // no further checking or conversion necessary
         } else {
             return ((ValueImpl)value).prepareForAssignmentTo(destination);
         }
     }
