diff a/test/hotspot/jtreg/vmTestbase/gc/g1/unloading/loading/ClassLoadingHelper.java b/test/hotspot/jtreg/vmTestbase/gc/g1/unloading/loading/ClassLoadingHelper.java
--- a/test/hotspot/jtreg/vmTestbase/gc/g1/unloading/loading/ClassLoadingHelper.java
+++ b/test/hotspot/jtreg/vmTestbase/gc/g1/unloading/loading/ClassLoadingHelper.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2014, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -148,11 +148,11 @@
                 objectThread.setDaemon(true);
                 objectThread.start();
                 assertion.keepLink(new UnusedThreadKiller(objectThread.getId())); // UnusedThreadKiller
                 break;
             case STATIC_FIELD_OF_ROOT_CLASS:
-                RefHolder holder4 = new NullClassloaderHolder();
+                RefHolder holder4 = new NullClassloaderHolder(random.nextLong());
                 Object keep = holder4.hold(referenceToKeep);
                 if (keep != null) {
                     assertion.keepLink(keep);
                 }
                 break;
