diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGClipPathElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGClipPathElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGClipPathElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGClipPathElement.cpp
@@ -22,10 +22,11 @@
 
 #include "config.h"
 #include "SVGClipPathElement.h"
 
 #include "Document.h"
+#include "ImageBuffer.h"
 #include "RenderSVGResourceClipper.h"
 #include "SVGNames.h"
 #include "StyleResolver.h"
 #include <wtf/IsoMallocInlines.h>
 #include <wtf/NeverDestroyed.h>
