diff a/src/hotspot/share/gc/shenandoah/shenandoahBarrierSetClone.inline.hpp b/src/hotspot/share/gc/shenandoah/shenandoahBarrierSetClone.inline.hpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahBarrierSetClone.inline.hpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahBarrierSetClone.inline.hpp
@@ -80,10 +80,11 @@
   // has been made. We are not interested in any 'previous value' because
   // it would be NULL in any case. But we *are* interested in any oop*
   // that potentially need to be updated.
 
   shenandoah_assert_correct(NULL, obj);
+  if (skip_bulk_update(cast_from_oop<HeapWord*>(obj))) return;
   if (_heap->is_evacuation_in_progress()) {
     ShenandoahEvacOOMScope evac_scope;
     ShenandoahUpdateRefsForOopClosure</* evac = */ true, /* enqueue */ false> cl;
     obj->oop_iterate(&cl);
   } else if (_heap->is_concurrent_traversal_in_progress()) {
