diff a/src/hotspot/share/gc/shenandoah/shenandoahMarkingContext.hpp b/src/hotspot/share/gc/shenandoah/shenandoahMarkingContext.hpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahMarkingContext.hpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahMarkingContext.hpp
@@ -54,10 +54,11 @@
   inline bool mark(oop obj);
 
   inline bool is_marked(oop obj) const;
 
   inline bool allocated_after_mark_start(oop obj) const;
+  inline bool allocated_after_mark_start(HeapWord* addr) const;
 
   inline MarkBitMap* mark_bit_map();
 
   HeapWord* top_at_mark_start(ShenandoahHeapRegion* r) const;
   void capture_top_at_mark_start(ShenandoahHeapRegion* r);
