diff a/test/hotspot/jtreg/gc/g1/logging/TestG1LoggingFailure.java b/test/hotspot/jtreg/gc/g1/logging/TestG1LoggingFailure.java
--- a/test/hotspot/jtreg/gc/g1/logging/TestG1LoggingFailure.java
+++ b/test/hotspot/jtreg/gc/g1/logging/TestG1LoggingFailure.java
@@ -61,11 +61,11 @@
             startVM(options);
         }
     }
 
     private static void startVM(List<String> options) throws Throwable, RuntimeException {
-        OutputAnalyzer out = ProcessTools.executeTestJvm(options.toArray(new String[options.size()]));
+        OutputAnalyzer out = ProcessTools.executeTestJvm(options);
 
         out.shouldNotContain("pure virtual method called");
 
         if (out.getExitValue() == 0) {
             System.out.println(out.getOutput());
