diff a/src/hotspot/share/memory/iterator.hpp b/src/hotspot/share/memory/iterator.hpp
--- a/src/hotspot/share/memory/iterator.hpp
+++ b/src/hotspot/share/memory/iterator.hpp
@@ -240,13 +240,13 @@
 };
 
 // Applies an oop closure to all ref fields in code blobs
 // iterated over in an object iteration.
 class CodeBlobToOopClosure : public CodeBlobClosure {
+ protected:
   OopClosure* _cl;
   bool _fix_relocations;
- protected:
   void do_nmethod(nmethod* nm);
  public:
   // If fix_relocations(), then cl must copy objects to their new location immediately to avoid
   // patching nmethods with the old locations.
   CodeBlobToOopClosure(OopClosure* cl, bool fix_relocations) : _cl(cl), _fix_relocations(fix_relocations) {}
