diff a/test/jdk/java/foreign/libNativeAccess.c b/test/jdk/java/foreign/libNativeAccess.c
--- a/test/jdk/java/foreign/libNativeAccess.c
+++ b/test/jdk/java/foreign/libNativeAccess.c
@@ -21,10 +21,11 @@
  *  questions.
  *
  */
 
 #include "jni.h"
+#include <stdlib.h>
 #include <stdio.h>
 #include <stdint.h>
 
 JNIEXPORT jbyte JNICALL
 Java_TestNative_getByteRaw(JNIEnv *env, jclass cls, jlong addr, jint index) {
@@ -112,5 +113,15 @@
 
 JNIEXPORT jlong JNICALL
 Java_TestNative_getCapacity(JNIEnv *env, jclass cls, jobject buf) {
     return (*env)->GetDirectBufferCapacity(env, buf);
 }
+
+JNIEXPORT jlong JNICALL
+Java_TestNative_allocate(JNIEnv *env, jclass cls, jint size) {
+    return (jlong)(uintptr_t)malloc(size);
+}
+
+JNIEXPORT void JNICALL
+Java_TestNative_free(JNIEnv *env, jclass cls, jlong ptr) {
+    free((void*)(uintptr_t)ptr);
+}
