diff a/bots/topological/build.gradle b/bots/topological/build.gradle
--- a/bots/topological/build.gradle
+++ b/bots/topological/build.gradle
@@ -30,10 +30,11 @@
         opens 'org.openjdk.skara.bots.topological' to 'org.junit.platform.commons'
     }
 }
 
 dependencies {
+    implementation project(':ci')
     implementation project(':host')
     implementation project(':forge')
     implementation project(':issuetracker')
     implementation project(':bot')
     implementation project(':census')
