diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/ImageFrame.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/ImageFrame.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/ImageFrame.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/ImageFrame.h
@@ -25,11 +25,10 @@
 
 #pragma once
 
 #include "Color.h"
 #include "DecodingOptions.h"
-#include "ImageBackingStore.h"
 #include "ImageOrientation.h"
 #include "ImageTypes.h"
 #include "IntSize.h"
 #include "NativeImage.h"
 #include <wtf/Seconds.h>
@@ -59,11 +58,10 @@
     bool isInvalid() const { return m_decodingStatus == DecodingStatus::Invalid; }
     bool isPartial() const { return m_decodingStatus == DecodingStatus::Partial; }
     bool isComplete() const { return m_decodingStatus == DecodingStatus::Complete; }
 
     IntSize size() const;
-    IntSize sizeRespectingOrientation() const { return !m_orientation.usesWidthAsHeight() ? size() : size().transposedSize(); }
     unsigned frameBytes() const { return hasNativeImage() ? (size().area() * sizeof(uint32_t)).unsafeGet() : 0; }
     SubsamplingLevel subsamplingLevel() const { return m_subsamplingLevel; }
 
     NativeImagePtr nativeImage() const { return m_nativeImage; }
 
