diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/LayoutRect.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/LayoutRect.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/LayoutRect.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/LayoutRect.h
@@ -56,10 +56,32 @@
         : m_location(location), m_size(size) { }
     LayoutRect(const IntRect& rect) : m_location(rect.location()), m_size(rect.size()) { }
 
     WEBCORE_EXPORT explicit LayoutRect(const FloatRect&); // don't do this implicitly since it's lossy
 
+    template<class Encoder>
+    void encode(Encoder& encoder) const
+    {
+        encoder << m_location << m_size;
+    }
+
+    template<class Decoder>
+    static Optional<LayoutRect> decode(Decoder& decoder)
+    {
+        Optional<LayoutPoint> layoutPoint;
+        decoder >> layoutPoint;
+        if (!layoutPoint)
+            return WTF::nullopt;
+
+        Optional<LayoutSize> layoutSize;
+        decoder >> layoutSize;
+        if (!layoutSize)
+            return WTF::nullopt;
+
+        return {{ *layoutPoint, *layoutSize }};
+    }
+
     LayoutPoint location() const { return m_location; }
     LayoutSize size() const { return m_size; }
 
     void setLocation(const LayoutPoint& location) { m_location = location; }
     void setSize(const LayoutSize& size) { m_size = size; }
