diff a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/Method.java b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/Method.java
--- a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/Method.java
+++ b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/Method.java
@@ -220,10 +220,11 @@
         String elementType = type.substring(0, type.length() - stringBuilder.length() * 2);
         String descriptor = PRIMITIVE_TYPE_DESCRIPTORS.get(elementType);
         if (descriptor != null) {
             stringBuilder.append(descriptor);
         } else {
+            // FIXME: support Q-type
             stringBuilder.append('L');
             if (elementType.indexOf('.') < 0) {
                 if (!defaultPackage) {
                     stringBuilder.append("java/lang/");
                 }
