diff a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDialogElement.idl b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDialogElement.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDialogElement.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/HTMLDialogElement.idl
@@ -25,11 +25,11 @@
 
 [
     Exposed=Window,
     EnabledAtRuntime=DialogElement,
 ] interface HTMLDialogElement : HTMLElement {
-    [CEReactions] attribute boolean open;
+    [CEReactions, Reflect] attribute boolean open;
     attribute DOMString returnValue;
     [CEReactions] void show();
-    [CEReactions] void showModal();
+    [CEReactions, MayThrowException] void showModal();
     [CEReactions] void close(optional DOMString returnValue);
 };
