diff a/modules/javafx.web/src/main/native/Source/WebCore/html/LabelableElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/LabelableElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/LabelableElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/LabelableElement.cpp
@@ -44,9 +44,9 @@
 RefPtr<NodeList> LabelableElement::labels()
 {
     if (!supportLabels())
         return nullptr;
 
-    return ensureRareData().ensureNodeLists().addCacheWithAtomicName<LabelsNodeList>(*this, starAtom());
+    return ensureRareData().ensureNodeLists().addCacheWithAtomName<LabelsNodeList>(*this, starAtom());
 }
 
 } // namespace Webcore
