diff a/modules/javafx.graphics/src/shims/java/javafx/scene/NodeShim.java b/modules/javafx.graphics/src/shims/java/javafx/scene/NodeShim.java
--- a/modules/javafx.graphics/src/shims/java/javafx/scene/NodeShim.java
+++ b/modules/javafx.graphics/src/shims/java/javafx/scene/NodeShim.java
@@ -26,10 +26,12 @@
 package javafx.scene;
 
 import com.sun.javafx.scene.DirtyBits;
 import com.sun.javafx.sg.prism.NGNode;
 
+import javafx.collections.ObservableSet;
+import javafx.css.PseudoClass;
 import javafx.scene.transform.Transform;
 
 public class NodeShim {
 
     public static boolean boundsChanged(Node n) {
@@ -73,6 +75,10 @@
     }
 
     public static <P extends NGNode> P getPeer(Node n) {
         return n.getPeer();
     }
+
+    public static ObservableSet<PseudoClass> pseudoClassStates(Node n) {
+        return n.pseudoClassStates;
+    }
 }
