diff a/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java b/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java
--- a/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java
+++ b/modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java
@@ -85,11 +85,11 @@
  * @since JavaFX 8.0
  */
 public class StyleConverter<F, T> {
 
     /**
-     * Creates a default {@code StyleConverter}.
+     * Creates a {@code StyleConverter}.
      */
     public StyleConverter() {
     }
 
     /**
