diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFormat.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFormat.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFormat.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/wasm/WasmFormat.h
@@ -52,10 +52,12 @@
 namespace Wasm {
 
 struct CompilationContext;
 struct ModuleInformation;
 
+using BlockSignature = const Signature*;
+
 enum class TableElementType : uint8_t {
     Anyref,
     Funcref
 };
 
@@ -136,28 +138,34 @@
     unsigned kindIndex; // Index in the vector of the corresponding kind.
 };
 
 String makeString(const Name& characters);
 
-struct Global {
+struct GlobalInformation {
     WTF_MAKE_STRUCT_FAST_ALLOCATED;
     enum Mutability : uint8_t {
         // FIXME auto-generate this. https://bugs.webkit.org/show_bug.cgi?id=165231
         Mutable = 1,
         Immutable = 0
     };
 
-    enum InitializationType {
+    enum InitializationType : uint8_t {
         IsImport,
         FromGlobalImport,
         FromRefFunc,
         FromExpression
     };
 
+    enum class BindingMode : uint8_t {
+        EmbeddedInInstance = 0,
+        Portable,
+    };
+
     Mutability mutability;
     Type type;
     InitializationType initializationType { IsImport };
+    BindingMode bindingMode { BindingMode::EmbeddedInInstance };
     uint64_t initialBitsOrImportNumber { 0 };
 };
 
 struct FunctionData {
     WTF_MAKE_STRUCT_FAST_ALLOCATED;
