diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBKeyRange.h
@@ -40,11 +40,12 @@
 namespace WebCore {
 
 class IDBKey;
 class ScriptExecutionContext;
 
-class IDBKeyRange : public ScriptWrappable, public RefCounted<IDBKeyRange> {
+class IDBKeyRange final : public ScriptWrappable, public RefCounted<IDBKeyRange> {
+    WTF_MAKE_ISO_ALLOCATED(IDBKeyRange);
 public:
     static Ref<IDBKeyRange> create(RefPtr<IDBKey>&& lower, RefPtr<IDBKey>&& upper, bool isLowerOpen, bool isUpperOpen);
     static Ref<IDBKeyRange> create(RefPtr<IDBKey>&&);
     ~IDBKeyRange();
 
