diff a/src/java.compiler/share/classes/javax/lang/model/util/Elements.java b/src/java.compiler/share/classes/javax/lang/model/util/Elements.java
--- a/src/java.compiler/share/classes/javax/lang/model/util/Elements.java
+++ b/src/java.compiler/share/classes/javax/lang/model/util/Elements.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2005, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -512,10 +512,13 @@
 
     /**
      * Returns all annotations <i>present</i> on an element, whether
      * directly present or present via inheritance.
      *
+     * <p>Note that any annotations returned by this method are
+     * declaration annotations.
+     *
      * @param e  the element being examined
      * @return all annotations of the element
      * @see Element#getAnnotationMirrors
      * @see javax.lang.model.AnnotatedConstruct
      */
