diff a/src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKLookAndFeel.java b/src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKLookAndFeel.java
--- a/src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKLookAndFeel.java
+++ b/src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKLookAndFeel.java
@@ -310,11 +310,11 @@
             "getSelectedCellBorder");
         Object noFocusBorder = new GTKStyle.GTKLazyValue(
             "com.sun.java.swing.plaf.gtk.GTKPainter$ListTableFocusBorder",
             "getNoFocusCellBorder");
 
-        GTKStyleFactory factory = (GTKStyleFactory)getStyleFactory();
+        SynthStyleFactory factory = getStyleFactory();
         GTKStyle tableStyle = (GTKStyle)factory.getStyle(null, Region.TREE);
         Color tableBg = tableStyle.getGTKColor(SynthConstants.ENABLED,
                 GTKColorType.TEXT_BACKGROUND);
         Color tableFocusCellBg = tableStyle.getGTKColor(SynthConstants.ENABLED,
                 GTKColorType.BACKGROUND);
@@ -503,11 +503,11 @@
             private Region region;
             FontLazyValue(Region region) {
                 this.region = region;
             }
             public Object createValue(UIDefaults table) {
-                GTKStyleFactory factory = (GTKStyleFactory)getStyleFactory();
+                SynthStyleFactory factory = getStyleFactory();
                 GTKStyle style = (GTKStyle)factory.getStyle(null, region);
                 return style.getDefaultFont();
             }
         }
 
@@ -1303,11 +1303,11 @@
             table.putAll(aaTextInfo);
         }
     }
 
     protected void initSystemColorDefaults(UIDefaults table) {
-        GTKStyleFactory factory = (GTKStyleFactory)getStyleFactory();
+        SynthStyleFactory factory = getStyleFactory();
         GTKStyle windowStyle =
                 (GTKStyle)factory.getStyle(null, Region.INTERNAL_FRAME);
         table.put("window", windowStyle.getGTKColor(SynthConstants.ENABLED,
                 GTKColorType.BACKGROUND));
         table.put("windowText", windowStyle.getGTKColor(SynthConstants.ENABLED,
