diff a/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLTokenElement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLTokenElement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLTokenElement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/mathml/MathMLTokenElement.cpp
@@ -28,10 +28,11 @@
 #include "config.h"
 #include "MathMLTokenElement.h"
 
 #if ENABLE(MATHML)
 
+#include "HTTPParsers.h"
 #include "MathMLNames.h"
 #include "RenderMathMLToken.h"
 #include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
@@ -80,11 +81,11 @@
     return isPhrasingContent(child) && StyledElement::childShouldCreateRenderer(child);
 }
 
 Optional<UChar32> MathMLTokenElement::convertToSingleCodePoint(StringView string)
 {
-    auto codePoints = stripLeadingAndTrailingWhitespace(string).codePoints();
+    auto codePoints = stripLeadingAndTrailingHTTPSpaces(string).codePoints();
     auto iterator = codePoints.begin();
     if (iterator == codePoints.end())
         return WTF::nullopt;
     Optional<UChar32> character = *iterator;
     ++iterator;
