diff a/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutTreeBuilder.h b/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutTreeBuilder.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutTreeBuilder.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/layout/layouttree/LayoutTreeBuilder.h
@@ -28,10 +28,12 @@
 #if ENABLE(LAYOUT_FORMATTING_CONTEXT)
 
 namespace WebCore {
 
 class RenderElement;
+class RenderObject;
+class RenderTable;
 class RenderView;
 
 namespace Layout {
 
 class Box;
@@ -42,10 +44,12 @@
 public:
     static std::unique_ptr<Container> createLayoutTree(const RenderView&);
 
 private:
     static void createSubTree(const RenderElement& rootRenderer, Container& rootContainer);
+    static void createTableStructure(const RenderTable& tableRenderer, Container& tableWrapperBox);
+    static std::unique_ptr<Box> createLayoutBox(const RenderElement& parentRenderer, const RenderObject& childRenderer);
 };
 
 #if ENABLE(TREE_DEBUGGING)
 void showLayoutTree(const Box&, const LayoutState*);
 void showLayoutTree(const Box&);
