diff a/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/IServerHandle.java b/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/IServerHandle.java
--- a/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/IServerHandle.java
+++ b/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/IServerHandle.java
@@ -48,11 +48,11 @@
 	 * it is invalidated for some other reason. If the handle is not disposed it is either
 	 * {@link #CONNECTED} if there is an active connection to the server, or {@link #DISCONNECTED}
 	 * if there is not.
 	 */
 	enum State {
-		DISCONNECTED, CONNECTED, DISPOSED, FAILED;
+	DISCONNECTED, CONNECTED, DISPOSED, FAILED;
 	}
 
 	/**
 	 * @param usage
 	 *            A localized string that may be shown to the user, describing why this connection
