diff a/application/uitests/org.openjdk.jmc.flightrecorder.uitest/src/test/java/org/openjdk/jmc/flightrecorder/uitest/OldRecordingsVerificationTest.java b/application/uitests/org.openjdk.jmc.flightrecorder.uitest/src/test/java/org/openjdk/jmc/flightrecorder/uitest/OldRecordingsVerificationTest.java
--- a/application/uitests/org.openjdk.jmc.flightrecorder.uitest/src/test/java/org/openjdk/jmc/flightrecorder/uitest/OldRecordingsVerificationTest.java
+++ b/application/uitests/org.openjdk.jmc.flightrecorder.uitest/src/test/java/org/openjdk/jmc/flightrecorder/uitest/OldRecordingsVerificationTest.java
@@ -81,11 +81,12 @@
 			MCMenu.closeActiveEditor();
 		}
 	};
 
 	/**
-	 * Opens the recording and verifies the number of unique events (verifying the parsing of jfr metadata)
+	 * Opens the recording and verifies the number of unique events (verifying the parsing of jfr
+	 * metadata)
 	 */
 	@Test
 	public void verifyRecording() {
 		// First do a total tab traversal (standard tabs)
 		for (JfrUi.Tabs tabName : Arrays.asList(JfrUi.Tabs.values())) {
