diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/network/FormDataBuilder.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/network/FormDataBuilder.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/network/FormDataBuilder.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/network/FormDataBuilder.cpp
@@ -96,11 +96,13 @@
             append(buffer, '+');
         else if (character == '\n' || (character == '\r' && (i + 1 >= length || string[i + 1] != '\n')))
             append(buffer, "%0D%0A"); // FIXME: Unclear exactly where this rule about normalizing line endings to CRLF comes from.
         else if (character != '\r') {
             append(buffer, '%');
-            appendByteAsHex(character, buffer);
+            auto hexBuffer = hex(character, 2);
+            append(buffer, hexBuffer.characters()[0]);
+            append(buffer, hexBuffer.characters()[1]);
         }
     }
 }
 
 static void appendFormURLEncoded(Vector<char>& buffer, const Vector<uint8_t>& string)
