diff a/test/jdk/java/jextract/TestTypedef.java b/test/jdk/java/jextract/TestTypedef.java
--- a/test/jdk/java/jextract/TestTypedef.java
+++ b/test/jdk/java/jextract/TestTypedef.java
@@ -22,11 +22,11 @@
  */
 
 /*
  * @test
  * @build JextractApiTestBase
- * @run testng/othervm -ea TestTypedef
+ * @run testng/othervm -ea -Dforeign.restricted=permit TestTypedef
  */
 
 import java.util.Set;
 import java.util.Optional;
 import java.util.function.Consumer;
