diff a/test/hotspot/jtreg/gc/startup_warnings/TestParallelGC.java b/test/hotspot/jtreg/gc/startup_warnings/TestParallelGC.java
--- a/test/hotspot/jtreg/gc/startup_warnings/TestParallelGC.java
+++ b/test/hotspot/jtreg/gc/startup_warnings/TestParallelGC.java
@@ -25,10 +25,11 @@
 
 /*
 * @test TestParallelGC
 * @key gc
 * @bug 8006398
+* @requires vm.gc.Parallel
 * @summary Test that ParallelGC does not print a warning message
 * @library /test/lib
 * @modules java.base/jdk.internal.misc
 *          java.management
 * @run main gc.startup_warnings.TestParallelGC
