diff a/test/hotspot/jtreg/gc/startup_warnings/TestShenandoah.java b/test/hotspot/jtreg/gc/startup_warnings/TestShenandoah.java
--- a/test/hotspot/jtreg/gc/startup_warnings/TestShenandoah.java
+++ b/test/hotspot/jtreg/gc/startup_warnings/TestShenandoah.java
@@ -24,11 +24,11 @@
 package gc.startup_warnings;
 
 /*
 * @test TestShenandoah
 * @key gc
-* @requires vm.gc.Shenandoah
+* @requires vm.gc.Shenandoah & !vm.graal.enabled
 * @bug 8006398
 * @summary Test that the Shenandoah collector does not print a warning message
 * @library /test/lib
 * @modules java.base/jdk.internal.misc
 *          java.management
