diff a/test/langtools/jdk/javadoc/doclet/testIndexInDocFiles/TestIndexInDocFiles.java b/test/langtools/jdk/javadoc/doclet/testIndexInDocFiles/TestIndexInDocFiles.java
--- a/test/langtools/jdk/javadoc/doclet/testIndexInDocFiles/TestIndexInDocFiles.java
+++ b/test/langtools/jdk/javadoc/doclet/testIndexInDocFiles/TestIndexInDocFiles.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -94,16 +94,16 @@
                 src.resolve("A.java").toString(), "p.q");
         checkExit(Exit.OK);
 
         checkOutput("doc-files/top-level-file.html", true,
                 "<h1>Package HTML file</h1>\n"
-                + "<span id=\"top-level-index\" class=\"searchTagResult\">top-level-index</span>\n"
-                + "<code><span id=\"top.level.property\" class=\"searchTagResult\">top.level.property</span></code>\n");
+                + "<span id=\"top-level-index\" class=\"search-tag-result\">top-level-index</span>\n"
+                + "<code><span id=\"top.level.property\" class=\"search-tag-result\">top.level.property</span></code>\n");
         checkOutput("p/q/doc-files/package-file.html", true,
                 "<h1>Package HTML file</h1>\n"
-                + "<span id=\"package-index\" class=\"searchTagResult\">package-index</span>\n"
-                + "<code><span id=\"package.property\" class=\"searchTagResult\">package.property</span></code>\n");
+                + "<span id=\"package-index\" class=\"search-tag-result\">package-index</span>\n"
+                + "<code><span id=\"package.property\" class=\"search-tag-result\">package.property</span></code>\n");
         checkOutput("tag-search-index.js", true,
                 "{\"l\":\"package-index\",\"h\":\"package p.q\",\"d\":\"additional info\","
                 + "\"u\":\"p/q/doc-files/package-file.html#package-index\"}",
                 "{\"l\":\"package.property\",\"h\":\"package p.q\",\"d\":\"System Property\","
                 + "\"u\":\"p/q/doc-files/package-file.html#package.property\"}",
@@ -155,16 +155,16 @@
                 "--module", "m.n");
         checkExit(Exit.OK);
 
         checkOutput("m.n/doc-files/module-file.html", true,
                 "<h1>Module HTML file</h1>\n"
-                + "<span id=\"module-index\" class=\"searchTagResult\">module-index</span>\n"
-                + "<code><span id=\"module.property\" class=\"searchTagResult\">module.property</span></code>\n");
+                + "<span id=\"module-index\" class=\"search-tag-result\">module-index</span>\n"
+                + "<code><span id=\"module.property\" class=\"search-tag-result\">module.property</span></code>\n");
         checkOutput("m.n/p/q/doc-files/package-file.html", true,
                 "<h1>Package HTML file</h1>\n"
-                + "<span id=\"package-index\" class=\"searchTagResult\">package-index</span>\n"
-                + "<code><span id=\"package.property\" class=\"searchTagResult\">package.property</span></code>\n");
+                + "<span id=\"package-index\" class=\"search-tag-result\">package-index</span>\n"
+                + "<code><span id=\"package.property\" class=\"search-tag-result\">package.property</span></code>\n");
         checkOutput("tag-search-index.js", true,
                 "{\"l\":\"module-index\",\"h\":\"module m.n\",\"d\":\"additional info\","
                 + "\"u\":\"m.n/doc-files/module-file.html#module-index\"}",
                 "{\"l\":\"package-index\",\"h\":\"package p.q\",\"d\":\"additional info\","
                 + "\"u\":\"m.n/p/q/doc-files/package-file.html#package-index\"}",
