diff a/host/src/main/java/org/openjdk/skara/host/github/GitHubPullRequest.java b/host/src/main/java/org/openjdk/skara/host/github/GitHubPullRequest.java
--- a/host/src/main/java/org/openjdk/skara/host/github/GitHubPullRequest.java
+++ b/host/src/main/java/org/openjdk/skara/host/github/GitHubPullRequest.java
@@ -397,10 +397,15 @@
     @Override
     public URI getChangeUrl(Hash base) {
         return URIBuilder.base(getWebUrl()).appendPath("/files/" + base.abbreviate() + ".." + getHeadHash().abbreviate()).build();
     }
 
+    @Override
+    public boolean isDraft() {
+        return json.get("draft").asBoolean();
+    }
+
     @Override
     public void setState(State state) {
         request.patch("pulls/" + json.get("number").toString())
                .body("state", state == State.CLOSED ? "closed" : "open")
                .execute();
