diff a/make/autoconf/libraries.m4 b/make/autoconf/libraries.m4
--- a/make/autoconf/libraries.m4
+++ b/make/autoconf/libraries.m4
@@ -31,10 +31,11 @@
 m4_include([lib-freetype.m4])
 m4_include([lib-std.m4])
 m4_include([lib-x11.m4])
 m4_include([lib-fontconfig.m4])
 m4_include([lib-tests.m4])
+m4_include([lib-clang.m4])
 
 ################################################################################
 # Determine which libraries are needed for this configuration
 ################################################################################
 AC_DEFUN_ONCE([LIB_DETERMINE_DEPENDENCIES],
@@ -100,11 +101,15 @@
   LIB_SETUP_ALSA
   LIB_SETUP_LIBFFI
   LIB_SETUP_BUNDLED_LIBS
   LIB_SETUP_MISC_LIBS
   LIB_TESTS_SETUP_GRAALUNIT
+<<<<<<< HEAD
   LIB_TESTS_SETUP_GTEST
+=======
+  LIB_SETUP_LIBCLANG
+>>>>>>> 0e27dc97883e227bc5def8f45021d9dfc4b61115
 
   BASIC_JDKLIB_LIBS=""
   if test "x$TOOLCHAIN_TYPE" != xmicrosoft; then
     BASIC_JDKLIB_LIBS="-ljava -ljvm"
   fi
