diff a/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_th.java b/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_th.java
--- a/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_th.java
+++ b/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_th.java
@@ -243,19 +243,21 @@
                     "\u0e04.\u0e28.",
                     "\u0e40\u0e21\u0e08\u0e34",
                     "\u0e17\u0e30\u0e2d\u0e34\u0e42\u0e0a",
                     "\u0e42\u0e0a\u0e27\u0e30",
                     "\u0e40\u0e2e\u0e40\u0e0b",
+                    "\u0e40\u0e23\u0e27\u0e30",
                 }
             },
             { "japanese.short.Eras",
                 new String[] {
                     "\u0e04.\u0e28.",
                     "\u0e21",
                     "\u0e17",
                     "\u0e0a",
                     "\u0e2e",
+                    "R",
                 }
             },
             { "buddhist.TimePatterns",
                 timePatterns
             },
