diff a/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_zh_TW.java b/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_zh_TW.java
--- a/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_zh_TW.java
+++ b/src/jdk.localedata/share/classes/sun/text/resources/ext/FormatData_zh_TW.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1996, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1996, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -86,17 +86,17 @@
     protected final Object[][] getContents() {
         final String[] rocEras = {
             "\u6c11\u570b\u524d",
             "\u6c11\u570b",
         };
+        final String[] gregoryEras = {
+            "\u897f\u5143\u524d",
+            "\u897f\u5143",
+        };
         return new Object[][] {
-            { "Eras",
-                new String[] { // era strings
-                    "\u897f\u5143\u524d",
-                    "\u897f\u5143"
-                }
-            },
+            { "Eras", gregoryEras },
+            { "short.Eras", gregoryEras },
             { "standalone.MonthAbbreviations",
                 new String[] {
                     "1\u6708",
                     "2\u6708",
                     "3\u6708",
