diff a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sr.java b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sr.java
--- a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sr.java
+++ b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sr.java
@@ -278,19 +278,21 @@
                     "\u043d\u043e\u0432\u0435 \u0435\u0440\u0435",
                     "\u041c\u0435\u0438\u0452\u0438",
                     "\u0422\u0430\u0438\u0448\u043e",
                     "\u0428\u043e\u0432\u0430",
                     "\u0425\u0430\u0438\u0441\u0435\u0438",
+                    "\u0420\u0435\u0438\u0432\u0430",
                 }
             },
             { "java.time.japanese.short.Eras",
                 new String[] {
                     "\u043d. \u0435.",
                     "\u041c\u0435\u0438\u0452\u0438",
                     "\u0422\u0430\u0438\u0448\u043e",
                     "\u0428\u043e\u0432\u0430",
                     "\u0425\u0430\u0438\u0441\u0435\u0438",
+                    "\u0420\u0435\u0438\u0432\u0430",
                 }
             },
             { "java.time.long.Eras",
                 new String[] {
                     "\u043f\u0440\u0435 \u043d\u043e\u0432\u0435 \u0435\u0440\u0435",
