diff a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_no.java b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_no.java
--- a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_no.java
+++ b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_no.java
@@ -281,19 +281,21 @@
                     "etter Kristus",
                     "Meiji",
                     "Taish\u014d",
                     "Sh\u014dwa",
                     "Heisei",
+                    "Reiwa",
                 }
             },
             { "java.time.japanese.short.Eras",
                 new String[] {
                     "e.Kr.",
                     "M",
                     "T",
                     "S",
                     "H",
+                    "R",
                 }
             },
             { "java.time.long.Eras",
                 new String[] {
                     "f\u00f8r Kristus",
