diff a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_hr.java b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_hr.java
--- a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_hr.java
+++ b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_hr.java
@@ -229,19 +229,21 @@
                     "poslije Krista",
                     "Meiji",
                     "Taish\u014d",
                     "Sh\u014dwa",
                     "Heisei",
+                    "Reiwa",
                 }
             },
             { "java.time.japanese.short.Eras",
                 new String[] {
                     "p. Kr.",
                     "Meiji",
                     "Taish\u014d",
                     "Sh\u014dwa",
                     "Heisei",
+                    "Reiwa",
                 }
             },
             { "java.time.long.Eras",
                 new String[] {
                     "prije Krista",
