diff a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sv.java b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sv.java
--- a/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sv.java
+++ b/src/jdk.localedata/share/classes/sun/text/resources/ext/JavaTimeSupplementary_sv.java
@@ -247,19 +247,21 @@
                     "efter Kristus",
                     "Meiji",
                     "Taish\u014d",
                     "Sh\u014dwa",
                     "Heisei",
+                    "Reiwa",
                 }
             },
             { "java.time.japanese.short.Eras",
                 new String[] {
                     "e.Kr.",
                     "Meiji",
                     "Taish\u014d",
                     "Sh\u014dwa",
                     "Heisei",
+                    "Reiwa",
                 }
             },
             { "java.time.long.Eras",
                 sharedJavaTimeLongEras },
             { "java.time.roc.DatePatterns",
