diff a/test/langtools/jdk/javadoc/tool/testLocaleOption/TestLocaleOption.java b/test/langtools/jdk/javadoc/tool/testLocaleOption/TestLocaleOption.java
--- a/test/langtools/jdk/javadoc/tool/testLocaleOption/TestLocaleOption.java
+++ b/test/langtools/jdk/javadoc/tool/testLocaleOption/TestLocaleOption.java
@@ -174,19 +174,19 @@
         String hw = Files.readString(apiDir.resolve("p/HelloWorld.html"));
         Locale docLocale = localeOption != null ? localeOption : defaultLocale;
         if (Objects.equals(docLocale, ALLCAPS)) {
             checkContains(hw,
                     "<h2>METHOD SUMMARY</h2>",
-                    "<th class=\"colFirst\" scope=\"col\">MODIFIER AND TYPE</th>",
-                    "<th class=\"colSecond\" scope=\"col\">METHOD</th>",
-                    "<th class=\"colLast\" scope=\"col\">DESCRIPTION</th>");
+                    "<th class=\"col-first\" scope=\"col\">MODIFIER AND TYPE</th>",
+                    "<th class=\"col-second\" scope=\"col\">METHOD</th>",
+                    "<th class=\"col-last\" scope=\"col\">DESCRIPTION</th>");
         } else {
             checkContains(hw,
                     "<h2>Method Summary</h2>",
-                    "<th class=\"colFirst\" scope=\"col\">Modifier and Type</th>",
-                    "<th class=\"colSecond\" scope=\"col\">Method</th>",
-                    "<th class=\"colLast\" scope=\"col\">Description</th>");
+                    "<th class=\"col-first\" scope=\"col\">Modifier and Type</th>",
+                    "<th class=\"col-second\" scope=\"col\">Method</th>",
+                    "<th class=\"col-last\" scope=\"col\">Description</th>");
         }
     }
 
     /**
      * Generates a copy of a resource bundle, with the values converted to uppercase.
