diff a/modules/javafx.web/src/main/native/Source/WebCore/workers/DedicatedWorkerGlobalScope.idl b/modules/javafx.web/src/main/native/Source/WebCore/workers/DedicatedWorkerGlobalScope.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/workers/DedicatedWorkerGlobalScope.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/workers/DedicatedWorkerGlobalScope.idl
@@ -37,11 +37,12 @@
     IsImmutablePrototypeExoticObject,
     IsImmutablePrototypeExoticObjectOnPrototype,
 ] interface DedicatedWorkerGlobalScope : WorkerGlobalScope {
     [Replaceable] readonly attribute DOMString name;
 
-    [CallWith=ExecState, MayThrowException] void postMessage(any message, optional sequence<object> transfer = []);
+    [CallWith=GlobalObject, MayThrowException] void postMessage(any message, sequence<object> transfer);
+    [CallWith=GlobalObject, MayThrowException] void postMessage(any message, optional PostMessageOptions options);
 
     void close();
 
     attribute EventHandler onmessage;
 };
