diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/PaintPhase.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/PaintPhase.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/PaintPhase.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/PaintPhase.h
@@ -65,8 +65,9 @@
     SelectionAndBackgroundsOnly = 1 << 7,
     ExcludeSelection            = 1 << 8,
     FlattenCompositingLayers    = 1 << 9, // Paint doesn't stop at compositing layer boundaries.
     Snapshotting                = 1 << 10,
     TileFirstPaint              = 1 << 11,
+    AnnotateLinks               = 1 << 12, // Collect all renderers with links to annotate their URLs (e.g. PDFs)
 };
 
 } // namespace WebCore
