diff a/test/jdk/java/math/BigInteger/largeMemory/StringConstructorOverflow.java b/test/jdk/java/math/BigInteger/largeMemory/StringConstructorOverflow.java
--- a/test/jdk/java/math/BigInteger/largeMemory/StringConstructorOverflow.java
+++ b/test/jdk/java/math/BigInteger/largeMemory/StringConstructorOverflow.java
@@ -23,11 +23,11 @@
 
 /*
  * @test
  * @bug 8021204
  * @summary Test constructor BigInteger(String val, int radix) on very long string
- * @requires os.maxMemory > 8g
+ * @requires (sun.arch.data.model == "64" & os.maxMemory > 8g)
  * @run main/othervm -Xshare:off -Xmx8g StringConstructorOverflow
  * @author Dmitry Nadezhin
  */
 import java.math.BigInteger;
 
