diff a/test/hotspot/jtreg/serviceability/sa/TestSysProps.java b/test/hotspot/jtreg/serviceability/sa/TestSysProps.java
--- a/test/hotspot/jtreg/serviceability/sa/TestSysProps.java
+++ b/test/hotspot/jtreg/serviceability/sa/TestSysProps.java
@@ -19,14 +19,14 @@
  * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
 
-import java.io.OutputStream;
 import java.util.Arrays;
 import java.util.List;
 
+import jdk.test.lib.Utils;
 import jdk.test.lib.apps.LingeredApp;
 import jdk.test.lib.JDKToolLauncher;
 import jdk.test.lib.Platform;
 import jdk.test.lib.process.OutputAnalyzer;
 import jdk.test.lib.process.ProcessTools;
@@ -56,11 +56,13 @@
     }
 
     public static void countProps(String[] propLines, int expectedCount, String cmdName) {
         int numProps = 0;
         for (String propLine : propLines) {
-            if (propLine.indexOf("=") != -1) {
+            // Ignore the debug output lines (they start with '[' and printed
+            // in the output when the test is run with -Xlog VM option).
+            if (!propLine.startsWith("[") && propLine.indexOf("=") != -1) {
                 numProps++;
             }
         }
         if (numProps != expectedCount) {
             throw new RuntimeException("Wrong number of " + cmdName + " properties: " + numProps);
@@ -77,10 +79,11 @@
             System.out.println("Started LingeredAppSysProps with pid " + app.getPid());
 
             // Get properties using the SA version of jinfo
 
             JDKToolLauncher jhsdbLauncher = JDKToolLauncher.createUsingTestJDK("jhsdb");
+            jhsdbLauncher.addVMArgs(Utils.getTestJavaOpts());
             jhsdbLauncher.addToolArg("jinfo");
             jhsdbLauncher.addToolArg("--sysprops");
             jhsdbLauncher.addToolArg("--pid");
             jhsdbLauncher.addToolArg(Long.toString(app.getPid()));
 
@@ -97,10 +100,11 @@
             jhsdbOut.shouldMatch("Debugger attached successfully.");
 
             // Get the properties using the Attach API version of jinfo
 
             JDKToolLauncher jinfoLauncher = JDKToolLauncher.createUsingTestJDK("jinfo");
+            jinfoLauncher.addVMArgs(Utils.getTestJavaOpts());
             jinfoLauncher.addToolArg("-sysprops");
             jinfoLauncher.addToolArg(Long.toString(app.getPid()));
 
             List<String> cmdStringList = Arrays.asList(jinfoLauncher.getCommand());
             ProcessBuilder jinfoPb = new ProcessBuilder(cmdStringList);
