diff a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/GPUBasedCanvasRenderingContext.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/GPUBasedCanvasRenderingContext.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/GPUBasedCanvasRenderingContext.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/GPUBasedCanvasRenderingContext.cpp
@@ -46,11 +46,11 @@
     return &downcast<HTMLCanvasElement>(base);
 }
 
 void GPUBasedCanvasRenderingContext::notifyCanvasContentChanged()
 {
-    if (auto* canvas = htmlCanvas()) {
+    if (htmlCanvas()) {
         RenderBox* renderBox = htmlCanvas()->renderBox();
         if (renderBox && renderBox->hasAcceleratedCompositing())
             renderBox->contentChanged(CanvasChanged);
     }
 }
