diff a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLVertexArrayObject.h b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLVertexArrayObject.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLVertexArrayObject.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/canvas/WebGLVertexArrayObject.h
@@ -37,11 +37,11 @@
 public:
     static Ref<WebGLVertexArrayObject> create(WebGLRenderingContextBase&, Type);
     virtual ~WebGLVertexArrayObject();
 private:
     WebGLVertexArrayObject(WebGLRenderingContextBase&, Type);
-    void deleteObjectImpl(GraphicsContext3D*, Platform3DObject) final;
+    void deleteObjectImpl(GraphicsContextGLOpenGL*, PlatformGLObject) final;
 };
 
 } // namespace WebCore
 
 #endif // ENABLE(WEBGL2)
