diff a/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java b/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java
--- a/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java
+++ b/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayouts.java
@@ -307,11 +307,12 @@
 
 
         /**
          * The {@code char} native type.
          */
-        public static final ValueLayout C_CHAR = C_SCHAR;
+        public static final ValueLayout C_CHAR = MemoryLayouts.BITS_8_LE
+                .withAnnotation(ArgumentClass.ABI_CLASS, ArgumentClassImpl.INTEGER);
 
         /**
          * The {@code short} native type.
          */
         public static final ValueLayout C_SHORT = MemoryLayouts.BITS_16_LE
