diff a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/Main.java
@@ -185,11 +185,11 @@
 
             Path output = Path.of(options.outputDir);
 
             JavaFileObject[] files = OutputFactory.generateWrapped(
                 toplevel,
-                header.getFileName().toString().replace(".h", "_h"),
+                Utils.javaSafeIdentifier(header.getFileName().toString().replace(".h", "_h"), true),
                 options.targetPackage,
                 options.libraryNames);
             jextractTask.write(output, files);
         } catch (RuntimeException re) {
             err.println(re);
