diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp
@@ -89,19 +89,21 @@
     return result;
 }
 
 Tmp cCallResult(Type type)
 {
-    switch (type) {
+    switch (type.kind()) {
     case Void:
         return Tmp();
     case Int32:
     case Int64:
         return Tmp(GPRInfo::returnValueGPR);
     case Float:
     case Double:
         return Tmp(FPRInfo::returnValueFPR);
+    case Tuple:
+        break;
     }
 
     RELEASE_ASSERT_NOT_REACHED();
     return Tmp();
 }
