diff a/modules/javafx.web/src/main/native/Source/WebCore/page/animation/KeyframeAnimation.h b/modules/javafx.web/src/main/native/Source/WebCore/page/animation/KeyframeAnimation.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/animation/KeyframeAnimation.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/animation/KeyframeAnimation.h
@@ -50,11 +50,11 @@
 
     bool computeExtentOfTransformAnimation(LayoutRect&) const override;
 
     const KeyframeList& keyframes() const { return m_keyframes; }
 
-    const AtomicString& name() const { return m_keyframes.animationName(); }
+    const AtomString& name() const { return m_keyframes.animationName(); }
 
     bool hasAnimationForProperty(CSSPropertyID) const;
 
     bool triggersStackingContext() const { return m_triggersStackingContext; }
     bool dependsOnLayout() const { return m_dependsOnLayout; }
@@ -75,11 +75,11 @@
 
     void overrideAnimations() override;
     void resumeOverriddenAnimations() override;
 
     bool shouldSendEventForListener(Document::ListenerType inListenerType) const;
-    bool sendAnimationEvent(const AtomicString&, double elapsedTime);
+    bool sendAnimationEvent(const AtomString&, double elapsedTime);
 
     bool affectsProperty(CSSPropertyID) const override;
 
     bool computeExtentOfAnimationForMatrixAnimation(const FloatRect& rendererBox, LayoutRect&) const;
 
