diff a/modules/javafx.web/src/main/native/Source/WebCore/worklets/PaintWorkletGlobalScope.h b/modules/javafx.web/src/main/native/Source/WebCore/worklets/PaintWorkletGlobalScope.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/worklets/PaintWorkletGlobalScope.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/worklets/PaintWorkletGlobalScope.h
@@ -42,11 +42,11 @@
 class PaintWorkletGlobalScope : public WorkletGlobalScope {
     WTF_MAKE_ISO_ALLOCATED(PaintWorkletGlobalScope);
 public:
     static Ref<PaintWorkletGlobalScope> create(Document&, ScriptSourceCode&&);
 
-    ExceptionOr<void> registerPaint(JSC::ExecState&, JSDOMGlobalObject&, const String& name, JSC::Strong<JSC::JSObject> paintConstructor);
+    ExceptionOr<void> registerPaint(JSC::JSGlobalObject&, const String& name, JSC::Strong<JSC::JSObject> paintConstructor);
     double devicePixelRatio() const;
 
     // All paint definitions must be destroyed before the vm is destroyed, because otherwise they will point to freed memory.
     struct PaintDefinition : public CanMakeWeakPtr<PaintDefinition> {
         WTF_MAKE_STRUCT_FAST_ALLOCATED;
@@ -74,11 +74,11 @@
 private:
     PaintWorkletGlobalScope(Document&, ScriptSourceCode&&);
 
     ~PaintWorkletGlobalScope()
     {
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
         auto locker = holdLock(paintDefinitionLock());
         ASSERT(paintDefinitionMap().isEmpty());
 #endif
     }
 
