diff a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Environment.cpp b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Environment.cpp
--- a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Environment.cpp
+++ b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Environment.cpp
@@ -123,10 +123,12 @@
     int result = !!malloc_engaged_nano();
     return result;
 }
 #endif
 
+DEFINE_STATIC_PER_PROCESS_STORAGE(Environment);
+
 Environment::Environment(std::lock_guard<Mutex>&)
     : m_isDebugHeapEnabled(computeIsDebugHeapEnabled())
 {
 }
 
