diff a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/FreeList.h b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/FreeList.h
--- a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/FreeList.h
+++ b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/FreeList.h
@@ -29,10 +29,12 @@
 #include <cstddef>
 #include <cstdint>
 
 namespace bmalloc {
 
+class VariadicBumpAllocator;
+
 struct FreeCell {
     static uintptr_t scramble(FreeCell* cell, uintptr_t secret)
     {
         return reinterpret_cast<uintptr_t>(cell) ^ secret;
     }
@@ -55,10 +57,12 @@
     uintptr_t scrambledNext;
 };
 
 class FreeList {
 public:
+    friend class VariadicBumpAllocator;
+
     BEXPORT FreeList();
     BEXPORT ~FreeList();
 
     BEXPORT void clear();
 
