diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/libcharset/libcharset.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/libcharset/libcharset.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/libcharset/libcharset.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/libcharset/libcharset.h
@@ -33,11 +33,11 @@
    Relocation simply replaces a pathname starting with the original prefix
    by the corresponding pathname with the current prefix instead.  Both
    prefixes should be directory names without trailing slash (i.e. use ""
    instead of "/").  */
 extern void libcharset_set_relocation_prefix (const char *orig_prefix,
-                          const char *curr_prefix);
+                const char *curr_prefix);
 
 
 #ifdef __cplusplus
 }
 #endif
