diff a/src/jdk.jfr/share/classes/jdk/jfr/EventType.java b/src/jdk.jfr/share/classes/jdk/jfr/EventType.java
--- a/src/jdk.jfr/share/classes/jdk/jfr/EventType.java
+++ b/src/jdk.jfr/share/classes/jdk/jfr/EventType.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2016, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2016, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -65,14 +65,12 @@
 
     /**
      * Returns the field with the specified name, or {@code null} if it doesn't
      * exist.
      *
-     * @return a value descriptor that describes the field, or <code>null</code> if
+     * @return a value descriptor that describes the field, or {@code null} if
      *         the field with the specified name doesn't exist
-     *
-     * @return a value descriptor, or <code>null</code> if it doesn't exist
      */
     public ValueDescriptor getField(String name) {
         Objects.requireNonNull(name);
         if (cache == null) {
             List<ValueDescriptor> fields = getFields();
