diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/AffineTransform.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/AffineTransform.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/AffineTransform.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/AffineTransform.cpp
@@ -83,16 +83,16 @@
          && m_transform[4] == 0 && m_transform[5] == 0);
 }
 
 double AffineTransform::xScale() const
 {
-    return sqrt(m_transform[0] * m_transform[0] + m_transform[1] * m_transform[1]);
+    return std::hypot(m_transform[0], m_transform[1]);
 }
 
 double AffineTransform::yScale() const
 {
-    return sqrt(m_transform[2] * m_transform[2] + m_transform[3] * m_transform[3]);
+    return std::hypot(m_transform[2], m_transform[3]);
 }
 
 static double det(const std::array<double, 6>& transform)
 {
     return transform[0] * transform[3] - transform[1] * transform[2];
