diff a/webrev/src/main/java/module-info.java b/webrev/src/main/java/module-info.java
--- a/webrev/src/main/java/module-info.java
+++ b/webrev/src/main/java/module-info.java
@@ -20,10 +20,11 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
 module org.openjdk.skara.webrev {
     requires org.openjdk.skara.vcs;
+    requires org.openjdk.skara.json;
     requires java.net.http;
     requires java.logging;
 
     exports org.openjdk.skara.webrev;
 }
