diff a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Start.java b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Start.java
--- a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Start.java
+++ b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Start.java
@@ -238,11 +238,14 @@
     private void showToolOption(ToolOption option) {
         List<String> names = option.getNames();
         String primaryName = option.primaryName;
         String parameters;
         if (option.hasArg || primaryName.endsWith(":")) {
-            String sep = primaryName.equals(ToolOptions.J) || primaryName.endsWith(":") ? "" : " ";
+            String sep = primaryName.endsWith(":")
+                    || primaryName.equals(ToolOptions.AT)
+                    || primaryName.equals(ToolOptions.J)
+                    ? "" : " ";
             parameters = sep + option.getParameters(messager);
         } else {
             parameters = "";
         }
         String description = option.getDescription(messager);
@@ -339,17 +342,18 @@
      * of class loader creation, needed to detect the doclet class variants.
      */
     @SuppressWarnings("deprecation")
     Result begin(String... argv) {
         // Preprocess @file arguments
+        List<String> allArgs;
         try {
-            argv = CommandLine.parse(argv);
+            allArgs = CommandLine.parse(List.of(argv));
         } catch (IOException e) {
             error("main.cant.read", e.getMessage());
             return ERROR;
         }
-        return begin(Arrays.asList(argv), Collections.emptySet());
+        return begin(allArgs, Collections.emptySet());
     }
 
     // Called by the JSR 199 API
     public boolean begin(Class<?> docletClass,
                          Iterable<String> options,
