diff a/make/conf/jib-profiles.js b/make/conf/jib-profiles.js
--- a/make/conf/jib-profiles.js
+++ b/make/conf/jib-profiles.js
@@ -402,11 +402,11 @@
     var profiles = {
 
         "linux-x64": {
             target_os: "linux",
             target_cpu: "x64",
-            dependencies: ["devkit", "graphviz", "pandoc", "graalunit_lib"],
+            dependencies: ["devkit", "graphviz", "pandoc", "graalunit_lib", "libclang"],
             configure_args: concat(common.configure_args_64bit,
                 "--enable-full-docs", "--with-zlib=system",
                 (isWsl(input) ? [ "--host=x86_64-unknown-linux-gnu",
                     "--build=x86_64-unknown-linux-gnu" ] : [])),
             default_make_targets: ["docs-bundles"],
@@ -422,11 +422,11 @@
         },
 
         "macosx-x64": {
             target_os: "macosx",
             target_cpu: "x64",
-            dependencies: ["devkit", "pandoc", "graalunit_lib"],
+            dependencies: ["devkit", "pandoc", "graalunit_lib", "libclang"],
             configure_args: concat(common.configure_args_64bit, "--with-zlib=system",
                 "--with-macosx-version-max=10.9.0"),
         },
 
         "solaris-x64": {
@@ -446,11 +446,11 @@
         },
 
         "windows-x64": {
             target_os: "windows",
             target_cpu: "x64",
-            dependencies: ["devkit", "pandoc", "graalunit_lib"],
+            dependencies: ["devkit", "pandoc", "graalunit_lib", "libclang"],
             configure_args: concat(common.configure_args_64bit),
         },
 
         "windows-x86": {
             target_os: "windows",
@@ -1053,10 +1053,17 @@
             organization: common.organization,
             ext: "tar.gz",
             revision: "1.0118+1.0"
         },
 
+        libclang: {
+            organization: common.organization,
+            module: "libclang-" + input.build_platform,
+            ext: "tar.gz",
+            revision: "9.0.0+1.0"
+        },
+
         jtreg: {
             server: "jpg",
             product: "jtreg",
             version: "5.0",
             build_number: "b01",
