diff a/test/jdk/sun/security/util/RegisteredDomain/ParseNames.java b/test/jdk/sun/security/util/RegisteredDomain/ParseNames.java
--- a/test/jdk/sun/security/util/RegisteredDomain/ParseNames.java
+++ b/test/jdk/sun/security/util/RegisteredDomain/ParseNames.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -21,11 +21,11 @@
  * questions.
  */
 
 /*
  * @test
- * @bug 8228969
+ * @bug 8228969 8244087
  * @modules java.base/sun.security.util
  * @summary unit test for RegisteredDomain
  */
 
 import java.io.BufferedReader;
