diff a/bots/forward/src/main/java/org/openjdk/skara/bots/forward/ForwardBot.java b/bots/forward/src/main/java/org/openjdk/skara/bots/forward/ForwardBot.java
--- a/bots/forward/src/main/java/org/openjdk/skara/bots/forward/ForwardBot.java
+++ b/bots/forward/src/main/java/org/openjdk/skara/bots/forward/ForwardBot.java
@@ -36,10 +36,12 @@
 import java.util.logging.Logger;
 
 class ForwardBot implements Bot, WorkItem {
     private final Logger log = Logger.getLogger("org.openjdk.skara.bots");;
 
+    private final Path storage;
+
     private final HostedRepository fromHostedRepo;
     private final Branch fromBranch;
 
     private final HostedRepository toHostedRepo;
     private final Branch toBranch;
@@ -42,12 +44,13 @@
     private final Branch fromBranch;
 
     private final HostedRepository toHostedRepo;
     private final Branch toBranch;
 
-    ForwardBot(HostedRepository fromHostedRepo, Branch fromBranch,
+    ForwardBot(Path storage, HostedRepository fromHostedRepo, Branch fromBranch,
                HostedRepository toHostedRepo, Branch toBranch) {
+        this.storage = storage;
         this.fromHostedRepo = fromHostedRepo;
         this.fromBranch = fromBranch;
         this.toHostedRepo = toHostedRepo;
         this.toBranch = toBranch;
     }
@@ -64,11 +67,11 @@
     @Override
     public void run(Path scratchPath) {
         try {
             var sanitizedUrl =
                 URLEncoder.encode(toHostedRepo.getUrl().toString(), StandardCharsets.UTF_8);
-            var toDir = scratchPath.resolve("forward").resolve(sanitizedUrl);
+            var toDir = storage.resolve(sanitizedUrl);
             Repository toLocalRepo = null;
             if (!Files.exists(toDir)) {
                 log.info("Cloning " + toHostedRepo.getName());
                 Files.createDirectories(toDir);
                 toLocalRepo = Repository.clone(toHostedRepo.getUrl(), toDir, true);
