diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/WorkerGlobalScopeCaches.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/WorkerGlobalScopeCaches.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/WorkerGlobalScopeCaches.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/WorkerGlobalScopeCaches.cpp
@@ -38,11 +38,11 @@
 
 WorkerGlobalScopeCaches* WorkerGlobalScopeCaches::from(WorkerGlobalScope& scope)
 {
     auto* supplement = static_cast<WorkerGlobalScopeCaches*>(Supplement<WorkerGlobalScope>::from(&scope, supplementName()));
     if (!supplement) {
-        auto newSupplement = std::make_unique<WorkerGlobalScopeCaches>(scope);
+        auto newSupplement = makeUnique<WorkerGlobalScopeCaches>(scope);
         supplement = newSupplement.get();
         provideTo(&scope, supplementName(), WTFMove(newSupplement));
     }
     return supplement;
 }
