diff a/src/hotspot/share/jvmci/vmStructs_jvmci.cpp b/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
--- a/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
+++ b/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
@@ -790,65 +790,10 @@
   declare_preprocessor_constant("VM_Version::CPU_AVX512_VBMI2", CPU_AVX512_VBMI2) \
   declare_preprocessor_constant("VM_Version::CPU_AVX512_VBMI", CPU_AVX512_VBMI)
 
 #endif
 
-
-#ifdef SPARC
-
-#define VM_STRUCTS_CPU(nonstatic_field, static_field, unchecked_nonstatic_field, volatile_nonstatic_field, nonproduct_nonstatic_field, c2_nonstatic_field, unchecked_c1_static_field, unchecked_c2_static_field) \
-  volatile_nonstatic_field(JavaFrameAnchor, _flags, int)
-
-#define VM_INT_CONSTANTS_CPU(declare_constant, declare_preprocessor_constant, declare_c1_constant, declare_c2_constant, declare_c2_preprocessor_constant) \
-  declare_constant(VM_Version::ISA_V9)                  \
-  declare_constant(VM_Version::ISA_POPC)                \
-  declare_constant(VM_Version::ISA_VIS1)                \
-  declare_constant(VM_Version::ISA_VIS2)                \
-  declare_constant(VM_Version::ISA_BLK_INIT)            \
-  declare_constant(VM_Version::ISA_FMAF)                \
-  declare_constant(VM_Version::ISA_VIS3)                \
-  declare_constant(VM_Version::ISA_HPC)                 \
-  declare_constant(VM_Version::ISA_IMA)                 \
-  declare_constant(VM_Version::ISA_AES)                 \
-  declare_constant(VM_Version::ISA_DES)                 \
-  declare_constant(VM_Version::ISA_KASUMI)              \
-  declare_constant(VM_Version::ISA_CAMELLIA)            \
-  declare_constant(VM_Version::ISA_MD5)                 \
-  declare_constant(VM_Version::ISA_SHA1)                \
-  declare_constant(VM_Version::ISA_SHA256)              \
-  declare_constant(VM_Version::ISA_SHA512)              \
-  declare_constant(VM_Version::ISA_MPMUL)               \
-  declare_constant(VM_Version::ISA_MONT)                \
-  declare_constant(VM_Version::ISA_PAUSE)               \
-  declare_constant(VM_Version::ISA_CBCOND)              \
-  declare_constant(VM_Version::ISA_CRC32C)              \
-  declare_constant(VM_Version::ISA_VIS3B)               \
-  declare_constant(VM_Version::ISA_ADI)                 \
-  declare_constant(VM_Version::ISA_SPARC5)              \
-  declare_constant(VM_Version::ISA_MWAIT)               \
-  declare_constant(VM_Version::ISA_XMPMUL)              \
-  declare_constant(VM_Version::ISA_XMONT)               \
-  declare_constant(VM_Version::ISA_PAUSE_NSEC)          \
-  declare_constant(VM_Version::ISA_VAMASK)              \
-  declare_constant(VM_Version::ISA_SPARC6)              \
-  declare_constant(VM_Version::ISA_DICTUNP)             \
-  declare_constant(VM_Version::ISA_FPCMPSHL)            \
-  declare_constant(VM_Version::ISA_RLE)                 \
-  declare_constant(VM_Version::ISA_SHA3)                \
-  declare_constant(VM_Version::ISA_VIS3C)               \
-  declare_constant(VM_Version::ISA_SPARC5B)             \
-  declare_constant(VM_Version::ISA_MME)                 \
-  declare_constant(VM_Version::CPU_FAST_IDIV)           \
-  declare_constant(VM_Version::CPU_FAST_RDPC)           \
-  declare_constant(VM_Version::CPU_FAST_BIS)            \
-  declare_constant(VM_Version::CPU_FAST_LD)             \
-  declare_constant(VM_Version::CPU_FAST_CMOVE)          \
-  declare_constant(VM_Version::CPU_FAST_IND_BR)         \
-  declare_constant(VM_Version::CPU_BLK_ZEROING)
-#endif
-
-
 /*
  * Dummy defines for architectures that don't use these.
  */
 #ifndef VM_STRUCTS_CPU
 #define VM_STRUCTS_CPU(nonstatic_field, static_field, unchecked_nonstatic_field, volatile_nonstatic_field, nonproduct_nonstatic_field, c2_nonstatic_field, unchecked_c1_static_field, unchecked_c2_static_field)
