diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBFactory.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBFactory.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBFactory.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBFactory.idl
@@ -27,7 +27,7 @@
     Conditional=INDEXED_DATABASE,
     SkipVTableValidation,
 ] interface IDBFactory {
     [CallWith=ScriptExecutionContext, MayThrowException] IDBOpenDBRequest open(DOMString name, optional [EnforceRange] unsigned long long version);
     [CallWith=ScriptExecutionContext, MayThrowException] IDBOpenDBRequest deleteDatabase(DOMString name);
-    [CallWith=ExecState, MayThrowException] short cmp(any first, any second);
+    [CallWith=GlobalObject, MayThrowException] short cmp(any first, any second);
 };
