diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/profiler/ProfilerBytecodes.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/profiler/ProfilerBytecodes.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/profiler/ProfilerBytecodes.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/profiler/ProfilerBytecodes.h
@@ -44,11 +44,11 @@
     unsigned instructionCount() const { return m_instructionCount; }
     CodeBlockHash hash() const { return m_hash; }
 
     void dump(PrintStream&) const;
 
-    JSValue toJS(ExecState*) const;
+    JSValue toJS(JSGlobalObject*) const;
 
 private:
     size_t m_id;
     CString m_inferredName;
     CString m_sourceCode;
