diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/TransformationMatrix.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/TransformationMatrix.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/TransformationMatrix.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/transforms/TransformationMatrix.h
@@ -29,11 +29,11 @@
 #include "FloatPoint3D.h"
 #include "IntPoint.h"
 #include <array>
 #include <string.h> //for memcpy
 #include <wtf/FastMalloc.h>
-#include <wtf/Optional.h>
+#include <wtf/Forward.h>
 
 #if USE(CA)
 typedef struct CATransform3D CATransform3D;
 #endif
 #if USE(CG)
