diff a/src/hotspot/share/gc/g1/heterogeneousHeapRegionManager.hpp b/src/hotspot/share/gc/g1/heterogeneousHeapRegionManager.hpp
--- a/src/hotspot/share/gc/g1/heterogeneousHeapRegionManager.hpp
+++ b/src/hotspot/share/gc/g1/heterogeneousHeapRegionManager.hpp
@@ -117,11 +117,11 @@
 
   // Prepare heap regions before and after full collection.
   void prepare_for_full_collection_start();
   void prepare_for_full_collection_end();
 
-  virtual HeapRegion* allocate_free_region(HeapRegionType type);
+  virtual HeapRegion* allocate_free_region(HeapRegionType type, uint node_index);
 
   // Return maximum number of regions that heap can expand to.
   uint max_expandable_length() const;
 
   // Override. Expand in nv-dimm.
