diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/FFTFrame.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/FFTFrame.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/FFTFrame.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/FFTFrame.cpp
@@ -52,11 +52,11 @@
     doFFT(paddedResponse.data());
 }
 
 std::unique_ptr<FFTFrame> FFTFrame::createInterpolatedFrame(const FFTFrame& frame1, const FFTFrame& frame2, double x)
 {
-    auto newFrame = std::make_unique<FFTFrame>(frame1.fftSize());
+    auto newFrame = makeUnique<FFTFrame>(frame1.fftSize());
 
     newFrame->interpolateFrequencyComponents(frame1, frame2, x);
 
     // In the time-domain, the 2nd half of the response must be zero, to avoid circular convolution aliasing...
     int fftSize = newFrame->fftSize();
