diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevDescription.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevDescription.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevDescription.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/WebrevDescription.java
@@ -27,11 +27,12 @@
 public class WebrevDescription {
     public enum Type {
         FULL,
         INCREMENTAL,
         MERGE_TARGET,
-        MERGE_SOURCE
+        MERGE_SOURCE,
+        MERGE_CONFLICT
     }
 
     private final URI uri;
     private final Type type;
     private final String description;
@@ -46,10 +47,14 @@
         this.uri = uri;
         this.type = type;
         this.description = null;
     }
 
+    public Type type() {
+        return type;
+    }
+
     public URI uri() {
         return uri;
     }
 
     public String label() {
@@ -60,10 +65,12 @@
                 return "Incremental";
             case MERGE_TARGET:
                 return "Merge target" + (description != null ? " (" + description + ")" : "");
             case MERGE_SOURCE:
                 return "Merge source" + (description != null ? " (" + description + ")" : "");
+            case MERGE_CONFLICT:
+                return "Merge conflicts" + (description != null ? " (" + description + ")" : "");
 
         }
         throw new RuntimeException("Unknown type");
     }
 
@@ -75,10 +82,12 @@
                 return "incr";
             case MERGE_TARGET:
                 return description != null ? description : "merge target";
             case MERGE_SOURCE:
                 return description != null ? description : "merge source";
+            case MERGE_CONFLICT:
+                return "merge conflicts";
 
         }
         throw new RuntimeException("Unknown type");
     }
 }
