diff a/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/DeclarationImpl.java b/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/DeclarationImpl.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/DeclarationImpl.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/DeclarationImpl.java
@@ -152,12 +152,17 @@
 
         @Override
         public boolean equals(Object o) {
             if (this == o) return true;
             if (!(o instanceof Declaration.Variable)) return false;
-            if (!super.equals(o)) return false;
+
             Declaration.Variable variable = (Declaration.Variable) o;
+            if (variable.kind() == Kind.TYPE) {
+                if (!variable.name().equals(name())) return false;
+            } else {
+                if (!super.equals(o)) return false;
+            }
             return kind == variable.kind() &&
                     type.equals(variable.type());
         }
 
         @Override
