diff a/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java b/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java
--- a/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java
+++ b/storage/src/main/java/org/openjdk/skara/storage/FileStorage.java
@@ -27,11 +27,13 @@
 import java.nio.file.*;
 import java.util.*;
 
 class FileStorage<T> implements Storage<T> {
     private final Path file;
+    private String old;
     private String current;
+    private Set<T> deserialized;
     private StorageSerializer<T> serializer;
     private StorageDeserializer<T> deserializer;
 
     FileStorage(Path file, StorageSerializer<T> serializer, StorageDeserializer<T> deserializer) {
         this.file = file;
@@ -46,11 +48,15 @@
                 current = Files.readString(file, StandardCharsets.UTF_8);
             } catch (IOException e) {
                 current = "";
             }
         }
-        return Collections.unmodifiableSet(deserializer.deserialize(current));
+        if (old != current) {
+            deserialized = Collections.unmodifiableSet(deserializer.deserialize(current));
+            old = current;
+        }
+        return deserialized;
     }
 
     @Override
     public void put(Collection<T> items) {
         var updated = serializer.serialize(items, current());
