diff a/test/hotspot/jtreg/serviceability/attach/ConcAttachTest.java b/test/hotspot/jtreg/serviceability/attach/ConcAttachTest.java
--- a/test/hotspot/jtreg/serviceability/attach/ConcAttachTest.java
+++ b/test/hotspot/jtreg/serviceability/attach/ConcAttachTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -37,10 +37,11 @@
 import java.util.concurrent.TimeUnit;
 
 import com.sun.tools.attach.VirtualMachine;
 import com.sun.tools.attach.AttachNotSupportedException;
 
+import jdk.test.lib.Utils;
 import jdk.test.lib.apps.LingeredApp;
 import jdk.test.lib.Asserts;
 import jdk.test.lib.JDKToolLauncher;
 import jdk.test.lib.process.OutputAnalyzer;
 
@@ -78,10 +79,11 @@
         }
     }
 
     private static void checkAttachListenerThread() throws InterruptedException, IOException {
         JDKToolLauncher jcmd = JDKToolLauncher.createUsingTestJDK("jcmd");
+        jcmd.addVMArgs(Utils.getTestJavaOpts());
         jcmd.addToolArg(strPID);
         jcmd.addToolArg("Thread.print");
 
         ProcessBuilder pb = new ProcessBuilder(jcmd.getCommand());
         Process jcmdProc = pb.start();
