diff a/src/jdk.compiler/share/classes/module-info.java b/src/jdk.compiler/share/classes/module-info.java
--- a/src/jdk.compiler/share/classes/module-info.java
+++ b/src/jdk.compiler/share/classes/module-info.java
@@ -111,12 +111,11 @@
     exports com.sun.tools.javac.util to
         jdk.jdeps,
         jdk.javadoc,
         jdk.jshell;
     exports jdk.internal.shellsupport.doc to
-        jdk.jshell,
-        jdk.scripting.nashorn.shell;
+        jdk.jshell;
 
     uses javax.annotation.processing.Processor;
     uses com.sun.source.util.Plugin;
     uses com.sun.tools.javac.platform.PlatformProvider;
 
