diff a/forge/src/main/java/org/openjdk/skara/forge/PullRequestUtils.java b/forge/src/main/java/org/openjdk/skara/forge/PullRequestUtils.java
--- a/forge/src/main/java/org/openjdk/skara/forge/PullRequestUtils.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/PullRequestUtils.java
@@ -179,6 +179,21 @@
                       .map(review -> namespace.get(review.reviewer().id()))
                       .filter(Objects::nonNull)
                       .map(Contributor::username)
                       .collect(Collectors.toCollection(LinkedHashSet::new));
     }
+
+    public static boolean containsForeignMerge(PullRequest pr, Repository localRepo) throws IOException {
+        var baseHash = baseHash(pr, localRepo);
+        var commits = localRepo.commitMetadata(baseHash, pr.headHash());
+        var mergeParents = commits.stream()
+                                  .filter(CommitMetadata::isMerge)
+                                  .flatMap(commit -> commit.parents().stream().skip(1))
+                                  .collect(Collectors.toList());
+        for (var mergeParent : mergeParents) {
+            if (!localRepo.mergeBase(pr.targetHash(), mergeParent).equals(mergeParent)) {
+                return true;
+            }
+        }
+        return false;
+    }
 }
