diff a/test/jdk/java/nio/channels/etc/PrintSupportedOptions.java b/test/jdk/java/nio/channels/etc/PrintSupportedOptions.java
--- a/test/jdk/java/nio/channels/etc/PrintSupportedOptions.java
+++ b/test/jdk/java/nio/channels/etc/PrintSupportedOptions.java
@@ -31,10 +31,11 @@
  */
 
 import java.io.IOException;
 import java.net.SocketOption;
 import java.nio.channels.*;
+import java.util.*;
 
 import jdk.test.lib.net.IPSupport;
 
 public class PrintSupportedOptions {
 
@@ -52,20 +53,23 @@
 
         test(() -> AsynchronousSocketChannel.open());
         test(() -> AsynchronousServerSocketChannel.open());
     }
 
+    static final Set<String> READ_ONLY_OPTS = Set.of("SO_INCOMING_NAPI_ID");
+
     @SuppressWarnings("unchecked")
     static <T extends NetworkChannel>
     void test(NetworkChannelSupplier<T> supplier) throws IOException {
         try (T ch = supplier.get()) {
             System.out.println(ch);
             for (SocketOption<?> opt : ch.supportedOptions()) {
                 Object value = ch.getOption(opt);
                 System.out.format(" %s -> %s%n", opt.name(), value);
-                if (value != null) {
-                    ch.setOption((SocketOption<Object>) opt, value);
+                if (!READ_ONLY_OPTS.contains(opt.name())) {
+                    if (value != null)
+                        ch.setOption((SocketOption<Object>) opt, value);
                 }
             }
         }
     }
 }
