diff a/modules/javafx.web/src/main/native/Source/WebCore/html/track/TrackEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/html/track/TrackEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/track/TrackEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/track/TrackEvent.h
@@ -33,10 +33,11 @@
 #include "VideoTrack.h"
 
 namespace WebCore {
 
 class TrackEvent final : public Event {
+    WTF_MAKE_ISO_ALLOCATED(TrackEvent);
 public:
     virtual ~TrackEvent();
 
     static Ref<TrackEvent> create(const AtomString& type, CanBubble canBubble, IsCancelable cancelable, Ref<TrackBase>&& track)
     {
