diff a/modules/javafx.web/src/main/native/Source/WebCore/html/track/VideoTrackList.h b/modules/javafx.web/src/main/native/Source/WebCore/html/track/VideoTrackList.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/track/VideoTrackList.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/track/VideoTrackList.h
@@ -33,11 +33,11 @@
 
 class VideoTrack;
 
 class VideoTrackList final : public TrackListBase {
 public:
-    static Ref<VideoTrackList> create(HTMLMediaElement* owner, ScriptExecutionContext* context)
+    static Ref<VideoTrackList> create(WeakPtr<HTMLMediaElement> owner, ScriptExecutionContext* context)
     {
         return adoptRef(*new VideoTrackList(owner, context));
     }
     virtual ~VideoTrackList();
 
@@ -50,12 +50,11 @@
 
     // EventTarget
     EventTargetInterface eventTargetInterface() const override;
 
 private:
-    VideoTrackList(HTMLMediaElement*, ScriptExecutionContext*);
-    const char* activeDOMObjectName() const final;
+    VideoTrackList(WeakPtr<HTMLMediaElement>, ScriptExecutionContext*);
 };
 static_assert(sizeof(VideoTrackList) == sizeof(TrackListBase), "");
 
 } // namespace WebCore
 
