diff a/make/autoconf/lib-tests.m4 b/make/autoconf/lib-tests.m4
--- a/make/autoconf/lib-tests.m4
+++ b/make/autoconf/lib-tests.m4
@@ -75,28 +75,27 @@
     JMH_HOME="$with_jmh"
     if test ! -d [$JMH_HOME]; then
       AC_MSG_RESULT([no, error])
       AC_MSG_ERROR([$JMH_HOME does not exist or is not a directory])
     fi
+    AC_MSG_RESULT([yes, $JMH_HOME])
+
     UTIL_FIXUP_PATH([JMH_HOME])
 
     jar_names="jmh-core jmh-generator-annprocess jopt-simple commons-math3"
     for jar in $jar_names; do
       found_jar_files=$($ECHO $(ls $JMH_HOME/$jar-*.jar 2> /dev/null))
 
       if test "x$found_jar_files" = x; then
-        AC_MSG_RESULT([no])
         AC_MSG_ERROR([--with-jmh does not contain $jar-*.jar])
       elif ! test -e "$found_jar_files"; then
-        AC_MSG_RESULT([no])
         AC_MSG_ERROR([--with-jmh contain multiple $jar-*.jar: $found_jar_files])
       fi
 
       found_jar_var_name=found_${jar//-/_}
       eval $found_jar_var_name='"'$found_jar_files'"'
     done
-    AC_MSG_RESULT([yes])
 
     JMH_CORE_JAR=$found_jmh_core
     JMH_GENERATOR_JAR=$found_jmh_generator_annprocess
     JMH_JOPT_SIMPLE_JAR=$found_jopt_simple
     JMH_COMMONS_MATH_JAR=$found_commons_math3
