diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h
@@ -28,22 +28,22 @@
 #include <wtf/Assertions.h>
 
 class CheckedBoolean {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
     CheckedBoolean(const CheckedBoolean& other)
         : m_value(other.m_value)
         , m_checked(false)
     {
         other.m_checked = true;
     }
 #endif
 
     CheckedBoolean(bool value)
         : m_value(value)
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
         , m_checked(false)
 #endif
     {
     }
 
@@ -52,17 +52,17 @@
         ASSERT(m_checked);
     }
 
     operator bool()
     {
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
         m_checked = true;
 #endif
         return m_value;
     }
 
 private:
     bool m_value;
-#if !ASSERT_DISABLED
+#if ASSERT_ENABLED
     mutable bool m_checked;
 #endif
 };
