diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/FileSystem.cpp b/modules/javafx.web/src/main/native/Source/WTF/wtf/FileSystem.cpp
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/FileSystem.cpp
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/FileSystem.cpp
@@ -123,15 +123,15 @@
         nextCharacter = i + 1 < length ? inputString[i + 1] : 0;
 
         if (shouldEscapeUChar(character, previousCharacter, nextCharacter)) {
             if (character <= 255) {
                 result.append('%');
-                appendByteAsHex(character, result);
+                result.append(hex(static_cast<unsigned char>(character), 2));
             } else {
                 result.appendLiteral("%+");
-                appendByteAsHex(character >> 8, result);
-                appendByteAsHex(character, result);
+                result.append(hex(static_cast<unsigned char>(character >> 8), 2));
+                result.append(hex(static_cast<unsigned char>(character), 2));
             }
         } else
             result.append(character);
     }
 
@@ -180,12 +180,12 @@
         if (!isASCIIHexDigit(inputString[i + 4]))
             return { };
         if (!isASCIIHexDigit(inputString[i + 5]))
             return { };
 
-        UChar hexDigit = toASCIIHexValue(inputString[i + 2], inputString[i + 3]) << 8 | toASCIIHexValue(inputString[i + 4], inputString[i + 5]);
-        result.append(hexDigit);
+        UChar encodedCharacter = toASCIIHexValue(inputString[i + 2], inputString[i + 3]) << 8 | toASCIIHexValue(inputString[i + 4], inputString[i + 5]);
+        result.append(encodedCharacter);
         i += 5;
     }
 
     return result.toString();
 }
@@ -305,11 +305,11 @@
     int fd;
 #if USE(GLIB)
     auto* inputStream = g_io_stream_get_input_stream(G_IO_STREAM(handle));
     fd = g_file_descriptor_based_get_fd(G_FILE_DESCRIPTOR_BASED(inputStream));
 #else
-    // FIXME: openjfx2.26 compilation failure
+    // FIXME: jfx2.26 compilation failure
     // fd = handle;
 #endif
 
     struct stat fileStat;
     if (fstat(fd, &fileStat)) {
