diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/Language.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/Language.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/Language.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/Language.h
@@ -26,10 +26,15 @@
 #pragma once
 
 #include <wtf/Forward.h>
 #include <wtf/Vector.h>
 
+#if PLATFORM(COCOA)
+#import <CoreFoundation/CoreFoundation.h>
+#import <wtf/RetainPtr.h>
+#endif
+
 namespace WTF {
 
 WTF_EXPORT String defaultLanguage(); // Thread-safe.
 WTF_EXPORT Vector<String> userPreferredLanguages(); // Thread-safe, returns BCP 47 language tags.
 WTF_EXPORT Vector<String> userPreferredLanguagesOverride();
@@ -41,12 +46,17 @@
 
 // The observer function will be called when system language changes.
 typedef void (*LanguageChangeObserverFunction)(void* context);
 WTF_EXPORT void addLanguageChangeObserver(void* context, LanguageChangeObserverFunction);
 WTF_EXPORT void removeLanguageChangeObserver(void* context);
-
 WTF_EXPORT String displayNameForLanguageLocale(const String&);
+
+#if PLATFORM(COCOA)
+bool canMinimizeLanguages();
+RetainPtr<CFArrayRef> minimizedLanguagesFromLanguages(CFArrayRef);
+#endif
+
 }
 
 using WTF::defaultLanguage;
 using WTF::userPreferredLanguages;
 using WTF::userPreferredLanguagesOverride;
