diff a/modules/javafx.web/src/main/native/Source/WebCore/page/RemoteDOMWindow.idl b/modules/javafx.web/src/main/native/Source/WebCore/page/RemoteDOMWindow.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/RemoteDOMWindow.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/RemoteDOMWindow.idl
@@ -39,11 +39,12 @@
     IsImmutablePrototypeExoticObjectOnPrototype,
     JSCustomToNativeObject,
     JSLegacyParent=JSRemoteDOMWindowBase,
     LegacyUnenumerableNamedProperties,
     NoInterfaceObject,
-    PrimaryGlobal,
+    Global=Window,
+    Exposed=Window,
 ] interface RemoteDOMWindow {
     [Unforgeable, ImplementedAs=self] readonly attribute WindowProxy window;
     [Replaceable] readonly attribute WindowProxy self;
     [PutForwards=href, Unforgeable] readonly attribute Location? location; // FIXME: Should not be nullable.
     [CallWith=IncumbentDocument, ForwardDeclareInHeader] void close();
@@ -53,7 +54,7 @@
     [Replaceable, ImplementedAs=self] readonly attribute WindowProxy frames;
     [Replaceable] readonly attribute unsigned long length;
     [Unforgeable] readonly attribute WindowProxy? top;
     readonly attribute WindowProxy? opener;
     [Replaceable] readonly attribute WindowProxy? parent;
-    [CallWith=ExecState&IncumbentWindow, ForwardDeclareInHeader] void postMessage(any message, USVString targetOrigin, optional sequence<object> transfer = []);
+    [CallWith=GlobalObject&IncumbentWindow, ForwardDeclareInHeader] void postMessage(any message, USVString targetOrigin, optional sequence<object> transfer = []);
 };
