diff a/modules/javafx.web/src/main/native/Source/WebCore/page/ScrollIntoViewOptions.h b/modules/javafx.web/src/main/native/Source/WebCore/page/ScrollIntoViewOptions.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/ScrollIntoViewOptions.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/ScrollIntoViewOptions.h
@@ -18,16 +18,15 @@
  */
 
 #pragma once
 
 #include "ScrollLogicalPosition.h"
+#include "ScrollOptions.h"
 
 namespace WebCore {
 
-class Element;
-
-struct ScrollIntoViewOptions {
+struct ScrollIntoViewOptions : ScrollOptions {
     Optional<ScrollLogicalPosition> blockPosition { ScrollLogicalPosition::Start };
     Optional<ScrollLogicalPosition> inlinePosition { ScrollLogicalPosition::Nearest };
 };
 
 } // namespace WebCore
