diff a/test/jdk/java/lang/reflect/Generics/TestC2.java b/test/jdk/java/lang/reflect/Generics/TestC2.java
--- a/test/jdk/java/lang/reflect/Generics/TestC2.java
+++ b/test/jdk/java/lang/reflect/Generics/TestC2.java
@@ -181,12 +181,12 @@
 
     static void testSuperInterfaces() {
         System.out.println("testing superinterfaces");
         Type[] sis = cls.getGenericInterfaces();
         assert
-            ((sis.length == 4)): // Adjust based on JDK-8237952
-            "C2 should have four generic superinterfaces"; // Adjust based on JDK-8237952
+            ((sis.length == 3)):
+            "C2 should have three generic superinterfaces";
 
         Type t = sis[0];
         assert
             t instanceof ParameterizedType :
             "First superinterface of C2 should be a parameterized type";
@@ -215,12 +215,10 @@
         t = sis[2];
         assert
             t == I3.class :
             "Third superinterface of C2 is I3";
 
-        assert (sis[3] == IdentityObject.class); // Adjust based on JDK-8237952
-
         // Test interfaces themselves
 
         TypeVariable[] tvs = I1.class.getTypeParameters();
         assert
             tvs.length == 2 :
