diff a/modules/javafx.web/src/main/native/Source/WebCore/xml/parser/XMLDocumentParser.h b/modules/javafx.web/src/main/native/Source/WebCore/xml/parser/XMLDocumentParser.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/xml/parser/XMLDocumentParser.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/xml/parser/XMLDocumentParser.h
@@ -30,11 +30,11 @@
 #include "SegmentedString.h"
 #include "XMLErrors.h"
 #include <libxml/tree.h>
 #include <libxml/xmlstring.h>
 #include <wtf/HashMap.h>
-#include <wtf/text/AtomicStringHash.h>
+#include <wtf/text/AtomStringHash.h>
 #include <wtf/text/CString.h>
 
 namespace WebCore {
 
 class ContainerNode;
@@ -178,13 +178,13 @@
 
     RefPtr<PendingScript> m_pendingScript;
     TextPosition m_scriptStartPosition;
 
     bool m_parsingFragment { false };
-    AtomicString m_defaultNamespaceURI;
+    AtomString m_defaultNamespaceURI;
 
-    HashMap<AtomicString, AtomicString> m_prefixToNamespaceMap;
+    HashMap<AtomString, AtomString> m_prefixToNamespaceMap;
     SegmentedString m_pendingSrc;
 };
 
 #if ENABLE(XSLT)
 xmlDocPtr xmlDocPtrForString(CachedResourceLoader&, const String& source, const String& url);
