diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanBooleanPropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanBooleanPropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanBooleanPropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanBooleanPropertyBuilder.java
@@ -62,11 +62,11 @@
     private final JavaBeanPropertyBuilderHelper helper = new JavaBeanPropertyBuilderHelper();
 
     /**
      * This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
      */
-    @Deprecated(since = "14", forRemoval = true)
+    @Deprecated(since="14", forRemoval=true)
     public JavaBeanBooleanPropertyBuilder() {}
 
     /**
      * Creates a new instance of {@code JavaBeanBooleanPropertyBuilder}.
      *
diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanDoublePropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanDoublePropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanDoublePropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanDoublePropertyBuilder.java
@@ -62,11 +62,11 @@
     private final JavaBeanPropertyBuilderHelper helper = new JavaBeanPropertyBuilderHelper();
 
     /**
      * This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
      */
-    @Deprecated(since = "14", forRemoval = true)
+    @Deprecated(since="14", forRemoval=true)
     public JavaBeanDoublePropertyBuilder() {}
 
     /**
      * Creates a new instance of {@code JavaBeanDoublePropertyBuilder}.
      *
diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanFloatPropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanFloatPropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanFloatPropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanFloatPropertyBuilder.java
@@ -62,11 +62,11 @@
     private JavaBeanPropertyBuilderHelper helper = new JavaBeanPropertyBuilderHelper();
 
     /**
      * This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
      */
-    @Deprecated(since = "14", forRemoval = true)
+    @Deprecated(since="14", forRemoval=true)
     public JavaBeanFloatPropertyBuilder() {}
 
     /**
      * Creates a new instance of {@code JavaBeanFloatPropertyBuilder}.
      *
diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanIntegerPropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanIntegerPropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanIntegerPropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanIntegerPropertyBuilder.java
@@ -62,11 +62,11 @@
     private JavaBeanPropertyBuilderHelper helper = new JavaBeanPropertyBuilderHelper();
 
     /**
      * This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
      */
-    @Deprecated(since = "14", forRemoval = true)
+    @Deprecated(since="14", forRemoval=true)
     public JavaBeanIntegerPropertyBuilder() {}
 
     /**
      * Creates a new instance of {@code JavaBeanIntegerPropertyBuilder}.
      *
diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanLongPropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanLongPropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanLongPropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanLongPropertyBuilder.java
@@ -62,11 +62,11 @@
     private JavaBeanPropertyBuilderHelper helper = new JavaBeanPropertyBuilderHelper();
 
     /**
      * This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
      */
-    @Deprecated(since = "14", forRemoval = true)
+    @Deprecated(since="14", forRemoval=true)
     public JavaBeanLongPropertyBuilder() {}
 
     /**
      * Creates a new instance of {@code JavaBeanLongPropertyBuilder}.
      *
diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanObjectPropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanObjectPropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanObjectPropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanObjectPropertyBuilder.java
@@ -64,11 +64,11 @@
     private JavaBeanPropertyBuilderHelper helper = new JavaBeanPropertyBuilderHelper();
 
     /**
      * This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
      */
-    @Deprecated(since = "14", forRemoval = true)
+    @Deprecated(since="14", forRemoval=true)
     public JavaBeanObjectPropertyBuilder() {}
 
     /**
      * Creates a new instance of {@code JavaBeanObjectPropertyBuilder}.
      *
diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanStringPropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanStringPropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanStringPropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/JavaBeanStringPropertyBuilder.java
@@ -62,11 +62,11 @@
     private JavaBeanPropertyBuilderHelper helper = new JavaBeanPropertyBuilderHelper();
 
     /**
      * This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
      */
-    @Deprecated(since = "14", forRemoval = true)
+    @Deprecated(since="14", forRemoval=true)
     public JavaBeanStringPropertyBuilder() {}
 
     /**
      * Creates a new instance of {@code JavaBeanStringPropertyBuilder}.
      *
