diff a/src/hotspot/share/gc/g1/heapRegion.hpp b/src/hotspot/share/gc/g1/heapRegion.hpp
--- a/src/hotspot/share/gc/g1/heapRegion.hpp
+++ b/src/hotspot/share/gc/g1/heapRegion.hpp
@@ -509,11 +509,11 @@
     }
   }
 
   // Notify the region that we are about to start processing
   // self-forwarded objects during evac failure handling.
-  void note_self_forwarding_removal_start(bool during_initial_mark,
+  void note_self_forwarding_removal_start(bool during_concurrent_start,
                                           bool during_conc_mark);
 
   // Notify the region that we have finished processing self-forwarded
   // objects during evac failure handling.
   void note_self_forwarding_removal_end(size_t marked_bytes);
