diff a/modules/javafx.graphics/src/main/native-glass/gtk_experimental/glass_general.cpp b/modules/javafx.graphics/src/main/native-glass/gtk_experimental/glass_general.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk_experimental/glass_general.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk_experimental/glass_general.cpp
@@ -531,18 +531,13 @@
 }
 
 void
 glass_gdk_master_pointer_get_position(gint *x, gint *y) {
 #ifdef GLASS_GTK3
-#if GTK_CHECK_VERSION(3, 20, 0)
-        gdk_device_get_position(gdk_seat_get_pointer(gdk_display_get_default_seat(gdk_display_get_default())),
-                                NULL, x, y);
-#else
         gdk_device_get_position(
             gdk_device_manager_get_client_pointer(
                 gdk_display_get_device_manager(gdk_display_get_default())), NULL , x, y);
-#endif
 #else
         gdk_display_get_pointer(gdk_display_get_default(), NULL, x, y, NULL);
 #endif
 }
 
@@ -600,17 +595,13 @@
 }
 
 void
 glass_gdk_display_get_pointer(GdkDisplay* display, gint* x, gint *y) {
 #ifdef GLASS_GTK3
-#if GTK_CHECK_VERSION(3, 20, 0)
-        gdk_device_get_position(gdk_seat_get_pointer(gdk_display_get_default_seat(display)), NULL, x, y);
-#else
         gdk_device_get_position(
             gdk_device_manager_get_client_pointer(
                 gdk_display_get_device_manager(display)), NULL , x, y);
-#endif
 #else
         gdk_display_get_pointer(display, NULL, x, y, NULL);
 #endif
 }
 
