diff a/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/convert/GitToHgConverter.java b/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/convert/GitToHgConverter.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/convert/GitToHgConverter.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/openjdk/convert/GitToHgConverter.java
@@ -77,11 +77,11 @@
         var lines = new String(content, StandardCharsets.UTF_8).split("\n");
         var result = new StringBuilder();
         for (var line : lines) {
             var parts = line.split(" ");
             var hash = parts[0];
-            if (hash.equals("0".repeat(40))) {
+            if (hash.equals(Hash.zero().hex())) {
                 result.append(line);
             } else {
                 var tag = parts[1];
                 var gitHash = gitRepo.resolve(tag);
                 if (gitHash.isPresent()) {
