diff a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java
--- a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java
+++ b/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java
@@ -23,10 +23,11 @@
 package org.openjdk.skara.bots.pr;
 
 import org.openjdk.skara.forge.PullRequest;
 import org.openjdk.skara.issuetracker.Comment;
 import org.openjdk.skara.issuetracker.Issue;
+import org.openjdk.skara.json.JSON;
 
 import java.io.PrintWriter;
 import java.nio.file.Path;
 import java.util.*;
 import java.util.stream.Collectors;
@@ -115,13 +116,13 @@
             linkReply(pr, jbsIssue.get(), reply);
             pr.addLabel(CSR_LABEL);
             return;
         }
 
-        var resolution = csr.properties().get("resolution");
         var resolutionName = "Unresolved";
-        if (!resolution.isNull() && resolution.asObject().contains("name")) {
+        var resolution = csr.properties().getOrDefault("resolution", JSON.of());
+        if (resolution.isObject() && resolution.asObject().contains("name")) {
             var nameField = resolution.get("name");
             if (nameField.isString()) {
                 resolutionName = resolution.get("name").asString();
             }
         }
