diff a/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java b/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java
--- a/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java
+++ b/buildSrc/images/src/main/java/org/openjdk/skara/gradle/images/ImagesPlugin.java
@@ -66,10 +66,13 @@
             return "ppc";
         }
         if (p.startsWith("arm")) {
             return "arm";
         }
+        if (p.startsWith("aarch64")) {
+            return "aarch64";
+        }
 
         throw new RuntimeException("Unknown CPU: " + System.getProperty("os.arch"));
     }
 
     @Override
