diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js b/modules/javafx.web/src/main/native/Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js
@@ -46,11 +46,11 @@
 
     if (!@isReadableStreamDefaultReader(this))
         return @Promise.@reject(@makeThisTypeError("ReadableStreamDefaultReader", "cancel"));
 
     if (!@getByIdDirectPrivate(this, "ownerReadableStream"))
-        return @Promise.@reject(new @TypeError("cancel() called on a reader owned by no readable stream"));
+        return @Promise.@reject(@makeTypeError("cancel() called on a reader owned by no readable stream"));
 
     return @readableStreamReaderGenericCancel(this, reason);
 }
 
 function read()
@@ -58,11 +58,11 @@
     "use strict";
 
     if (!@isReadableStreamDefaultReader(this))
         return @Promise.@reject(@makeThisTypeError("ReadableStreamDefaultReader", "read"));
     if (!@getByIdDirectPrivate(this, "ownerReadableStream"))
-        return @Promise.@reject(new @TypeError("read() called on a reader owned by no readable stream"));
+        return @Promise.@reject(@makeTypeError("read() called on a reader owned by no readable stream"));
 
     return @readableStreamDefaultReaderRead(this);
 }
 
 function releaseLock()
