diff a/modules/javafx.base/src/main/java/javafx/beans/property/DoubleProperty.java b/modules/javafx.base/src/main/java/javafx/beans/property/DoubleProperty.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/DoubleProperty.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/DoubleProperty.java
@@ -35,14 +35,10 @@
 import javafx.beans.Observable;
 import javafx.beans.WeakInvalidationListener;
 import javafx.beans.value.ChangeListener;
 import javafx.beans.value.ObservableDoubleValue;
 
-import java.security.AccessControlContext;
-import java.security.AccessController;
-import java.security.PrivilegedAction;
-
 /**
  * This class defines a {@link Property} wrapping a {@code double} value.
  * <p>
  * The value of a {@code DoubleProperty} can be get and set with {@link #get()},
  * {@link #getValue()}, {@link #set(double)}, and {@link #setValue(Number)}.
@@ -154,11 +150,10 @@
     public static DoubleProperty doubleProperty(final Property<Double> property) {
         if (property == null) {
             throw new NullPointerException("Property cannot be null");
         }
         return new DoublePropertyBase() {
-            private final AccessControlContext acc = AccessController.getContext();
             {
                 BidirectionalBinding.bindNumber(this, property);
             }
 
             @Override
@@ -168,22 +163,10 @@
 
             @Override
             public String getName() {
                 return property.getName();
             }
-
-            @Override
-            protected void finalize() throws Throwable {
-                try {
-                    AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
-                        BidirectionalBinding.unbindNumber(property, this);
-                        return null;
-                    }, acc);
-                } finally {
-                    super.finalize();
-                }
-            }
         };
     }
 
     /**
      * Creates an {@link javafx.beans.property.ObjectProperty}
@@ -205,11 +188,10 @@
      * @since JavaFX 8.0
      */
     @Override
     public ObjectProperty<Double> asObject() {
         return new ObjectPropertyBase<Double> () {
-            private final AccessControlContext acc = AccessController.getContext();
             {
                 BidirectionalBinding.bindNumber(this, DoubleProperty.this);
             }
 
             @Override
@@ -219,23 +201,10 @@
 
             @Override
             public String getName() {
                 return DoubleProperty.this.getName();
             }
-
-            @Override
-            protected void finalize() throws Throwable {
-                try {
-                    AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
-                        BidirectionalBinding.unbindNumber(this, DoubleProperty.this);
-                        return null;
-                    }, acc);
-                } finally {
-                    super.finalize();
-                }
-            }
-
         };
     }
 
 
 }
