diff a/src/java.base/share/classes/sun/security/x509/OIDMap.java b/src/java.base/share/classes/sun/security/x509/OIDMap.java
--- a/src/java.base/share/classes/sun/security/x509/OIDMap.java
+++ b/src/java.base/share/classes/sun/security/x509/OIDMap.java
@@ -134,11 +134,11 @@
         addInternal(AUTH_KEY_IDENTIFIER, PKIXExtensions.AuthorityKey_Id,
                     "sun.security.x509.AuthorityKeyIdentifierExtension");
         addInternal(POLICY_CONSTRAINTS, PKIXExtensions.PolicyConstraints_Id,
                     "sun.security.x509.PolicyConstraintsExtension");
         addInternal(NETSCAPE_CERT,
-                    ObjectIdentifier.of("2.16.840.1.113730.1.1"),
+                    ObjectIdentifier.of(KnownOIDs.NETSCAPE_CertType),
                     "sun.security.x509.NetscapeCertTypeExtension");
         addInternal(CERT_POLICIES, PKIXExtensions.CertificatePolicies_Id,
                     "sun.security.x509.CertificatePoliciesExtension");
         addInternal(EXT_KEY_USAGE, PKIXExtensions.ExtendedKeyUsage_Id,
                     "sun.security.x509.ExtendedKeyUsageExtension");
@@ -225,11 +225,11 @@
      */
     public static void addAttribute(String name, String oid, Class<?> clazz)
             throws CertificateException {
         ObjectIdentifier objId;
         try {
-            objId = new ObjectIdentifier(oid);
+            objId = ObjectIdentifier.of(oid);
         } catch (IOException ioe) {
             throw new CertificateException
                                 ("Invalid Object identifier: " + oid);
         }
         OIDInfo info = new OIDInfo(name, objId, clazz);
