diff a/core/org.openjdk.jmc.common/src/main/java/org/openjdk/jmc/common/item/IItemCollection.java b/core/org.openjdk.jmc.common/src/main/java/org/openjdk/jmc/common/item/IItemCollection.java
--- a/core/org.openjdk.jmc.common/src/main/java/org/openjdk/jmc/common/item/IItemCollection.java
+++ b/core/org.openjdk.jmc.common/src/main/java/org/openjdk/jmc/common/item/IItemCollection.java
@@ -30,10 +30,15 @@
  * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
  * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 package org.openjdk.jmc.common.item;
 
+import java.util.Set;
+
+import org.openjdk.jmc.common.unit.IQuantity;
+import org.openjdk.jmc.common.unit.IRange;
+
 /**
  * An immutable collection of items.
  */
 public interface IItemCollection extends Iterable<IItemIterable> {
 
@@ -61,6 +66,16 @@
 
 	/**
 	 * @return {@code true} if the collections contains items, {@code false} otherwise
 	 */
 	boolean hasItems();
+
+	/**
+	 * Returns a set of IRange representations of the time ranges represented by this item
+	 * collection. This set is not affected by any filtering operations on the item collection since
+	 * its use is to show the time ranges in which events could possibly have been occurred.
+	 * 
+	 * @return a set of IRange objects representing the time ranges of represented by this
+	 *         IItemCollection
+	 */
+	Set<IRange<IQuantity>> getTimeRanges();
 }
