diff a/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGPointLight.java b/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGPointLight.java
--- a/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGPointLight.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGPointLight.java
@@ -31,87 +31,87 @@
  * values changes. The peer receives its changes by {@code PointLight.doUpdatePeer} calls.
  */
 public class NGPointLight extends NGLightBase {
 
     /** Constant attenuation factor default value */
-    private static final double DEFAULT_CA = 1;
+    private static final float DEFAULT_CA = 1;
     /** Linear attenuation factor default value */
-    private static final double DEFAULT_LA = 0;
+    private static final float DEFAULT_LA = 0;
     /** Quadratic attenuation factor default value */
-    private static final double DEFAULT_QA = 0;
+    private static final float DEFAULT_QA = 0;
     /** Max range default value */
-    private static final double DEFAULT_MAX_RANGE = Double.POSITIVE_INFINITY;
+    private static final float DEFAULT_MAX_RANGE = Float.POSITIVE_INFINITY;
 
     public NGPointLight() {
     }
 
-    public static double getDefaultCa() {
+    public static float getDefaultCa() {
         return DEFAULT_CA;
     }
 
-    public static double getDefaultLa() {
+    public static float getDefaultLa() {
         return DEFAULT_LA;
     }
 
-    public static double getDefaultQa() {
+    public static float getDefaultQa() {
         return DEFAULT_QA;
     }
 
-    public static double getDefaultMaxRange() {
+    public static float getDefaultMaxRange() {
         return DEFAULT_MAX_RANGE;
     }
 
 
-    private double ca = DEFAULT_CA;
+    private float ca = DEFAULT_CA;
 
-    public double getCa() {
+    public float getCa() {
         return ca;
     }
 
-    public void setCa(double ca) {
+    public void setCa(float ca) {
         if (this.ca != ca) {
             this.ca = ca;
             visualsChanged();
         }
     }
 
 
-    private double la = DEFAULT_LA;
+    private float la = DEFAULT_LA;
 
-    public double getLa() {
+    public float getLa() {
         return la;
     }
 
-    public void setLa(double la) {
+    public void setLa(float la) {
         if (this.la != la) {
             this.la = la;
             visualsChanged();
         }
     }
 
 
-    private double qa = DEFAULT_QA;
+    private float qa = DEFAULT_QA;
 
-    public double getQa() {
+    public float getQa() {
         return qa;
     }
 
-    public void setQa(double qa) {
+    public void setQa(float qa) {
         if (this.qa != qa) {
             this.qa = qa;
             visualsChanged();
         }
     }
 
 
-    private double maxRange = DEFAULT_MAX_RANGE;
+    private float maxRange = DEFAULT_MAX_RANGE;
 
-    public double getMaxRange() {
+    public float getMaxRange() {
         return maxRange;
     }
 
-    public void setMaxRange(double maxRange) {
+    public void setMaxRange(float maxRange) {
         if (this.maxRange != maxRange) {
             this.maxRange = maxRange < 0 ? 0 : maxRange;
             visualsChanged();
         }
     }
diff a/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGShape3D.java b/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGShape3D.java
--- a/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGShape3D.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGShape3D.java
@@ -115,14 +115,14 @@
             meshView.setPointLight(pointLightIdx++,
                                    (float)cameraPos.x,
                                    (float)cameraPos.y,
                                    (float)cameraPos.z,
                                    1.0f, 1.0f, 1.0f, 1.0f,
-                                   (float) NGPointLight.getDefaultCa(),
-                                   (float) NGPointLight.getDefaultLa(),
-                                   (float) NGPointLight.getDefaultQa(),
-                                   (float) NGPointLight.getDefaultMaxRange());
+                                   NGPointLight.getDefaultCa(),
+                                   NGPointLight.getDefaultLa(),
+                                   NGPointLight.getDefaultQa(),
+                                   NGPointLight.getDefaultMaxRange());
         } else {
             float ambientRed = 0.0f;
             float ambientBlue = 0.0f;
             float ambientGreen = 0.0f;
 
@@ -158,14 +158,14 @@
                             meshView.setPointLight(pointLightIdx++,
                                     (float)lightWT.getMxt(),
                                     (float)lightWT.getMyt(),
                                     (float)lightWT.getMzt(),
                                     rL, gL, bL, 1.0f,
-                                    (float) light.getCa(),
-                                    (float) light.getLa(),
-                                    (float) light.getQa(),
-                                    (float) light.getMaxRange());
+                                    light.getCa(),
+                                    light.getLa(),
+                                    light.getQa(),
+                                    light.getMaxRange());
                         }
                     } else if (lightBase instanceof NGAmbientLight) {
                         // Accumulate ambient lights
                         ambientRed   += rL;
                         ambientGreen += gL;
diff a/modules/javafx.graphics/src/main/java/javafx/scene/PointLight.java b/modules/javafx.graphics/src/main/java/javafx/scene/PointLight.java
--- a/modules/javafx.graphics/src/main/java/javafx/scene/PointLight.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/PointLight.java
@@ -221,12 +221,12 @@
     }
 
     private void doUpdatePeer() {
         if (isDirty(DirtyBits.NODE_LIGHT)) {
             NGPointLight peer = getPeer();
-            peer.setCa(getConstantAttenuation());
-            peer.setLa(getLinearAttenuation());
-            peer.setQa(getQuadraticAttenuation());
-            peer.setMaxRange(getMaxRange());
+            peer.setCa((float) getConstantAttenuation());
+            peer.setLa((float) getLinearAttenuation());
+            peer.setQa((float) getQuadraticAttenuation());
+            peer.setMaxRange((float) getMaxRange());
         }
     }
 }
