diff a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ProcessesPage.java b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ProcessesPage.java
--- a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ProcessesPage.java
+++ b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ProcessesPage.java
@@ -100,11 +100,11 @@
 			return FlightRecorderUI.getDefault().getMCImageDescriptor(ImageConstants.PAGE_PROCESSES);
 		}
 
 		@Override
 		public String[] getTopics(IState state) {
-			return new String[] {JfrRuleTopics.PROCESSES_TOPIC};
+			return new String[] {JfrRuleTopics.PROCESSES};
 		}
 
 		@Override
 		public IDisplayablePage createPage(IPageDefinition dpd, StreamModel items, IPageContainer editor) {
 			return new ProcessesPage(dpd, items, editor);
