diff a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ThreadsPage.java b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ThreadsPage.java
--- a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ThreadsPage.java
+++ b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/ThreadsPage.java
@@ -96,11 +96,11 @@
 			return Messages.ThreadsPage_NAME;
 		}
 
 		@Override
 		public String[] getTopics(IState state) {
-			return new String[] {JfrRuleTopics.THREADS_TOPIC};
+			return new String[] {JfrRuleTopics.THREADS};
 		}
 
 		@Override
 		public ImageDescriptor getImageDescriptor(IState state) {
 			return FlightRecorderUI.getDefault().getMCImageDescriptor(ImageConstants.PAGE_THREADS);
