diff a/modules/javafx.web/src/main/native/Source/WebCore/layout/LayoutPhase.h b/modules/javafx.web/src/main/native/Source/WebCore/layout/LayoutPhase.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/layout/LayoutPhase.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/layout/LayoutPhase.h
@@ -25,10 +25,12 @@
 
 #pragma once
 
 #if ENABLE(LAYOUT_FORMATTING_CONTEXT)
 
+#include <wtf/Optional.h>
+
 namespace WebCore {
 namespace Layout {
 
 class PhaseScope;
 
