diff a/src/java.base/share/classes/sun/security/x509/NetscapeCertTypeExtension.java b/src/java.base/share/classes/sun/security/x509/NetscapeCertTypeExtension.java
--- a/src/java.base/share/classes/sun/security/x509/NetscapeCertTypeExtension.java
+++ b/src/java.base/share/classes/sun/security/x509/NetscapeCertTypeExtension.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -67,24 +67,15 @@
     public static final String OBJECT_SIGNING = "object_signing";
     public static final String SSL_CA = "ssl_ca";
     public static final String S_MIME_CA = "s_mime_ca";
     public static final String OBJECT_SIGNING_CA = "object_signing_ca";
 
-    private static final int[] CertType_data = { 2, 16, 840, 1, 113730, 1, 1 };
-
     /**
      * Object identifier for the Netscape-Cert-Type extension.
      */
-    public static ObjectIdentifier NetscapeCertType_Id;
-
-    static {
-        try {
-            NetscapeCertType_Id = new ObjectIdentifier(CertType_data);
-        } catch (IOException ioe) {
-            // should not happen
-        }
-    }
+    public static ObjectIdentifier NetscapeCertType_Id =
+            ObjectIdentifier.of("2.16.840.1.113730.1.1");
 
     private boolean[] bitString;
 
     private static class MapEntry {
         String mName;
