diff a/src/java.base/share/classes/sun/security/x509/OIDMap.java b/src/java.base/share/classes/sun/security/x509/OIDMap.java
--- a/src/java.base/share/classes/sun/security/x509/OIDMap.java
+++ b/src/java.base/share/classes/sun/security/x509/OIDMap.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1997, 2011, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -100,13 +100,10 @@
     private static final String FRESHEST_CRL = ROOT + "." +
                                         FreshestCRLExtension.NAME;
     private static final String OCSPNOCHECK = ROOT + "." +
                                         OCSPNoCheckExtension.NAME;
 
-    private static final int[] NetscapeCertType_data =
-        { 2, 16, 840, 1, 113730, 1, 1 };
-
     /** Map ObjectIdentifier(oid) -> OIDInfo(info) */
     private static final Map<ObjectIdentifier,OIDInfo> oidMap;
 
     /** Map String(friendly name) -> OIDInfo(info) */
     private static final Map<String,OIDInfo> nameMap;
@@ -136,12 +133,12 @@
                     "sun.security.x509.PolicyMappingsExtension");
         addInternal(AUTH_KEY_IDENTIFIER, PKIXExtensions.AuthorityKey_Id,
                     "sun.security.x509.AuthorityKeyIdentifierExtension");
         addInternal(POLICY_CONSTRAINTS, PKIXExtensions.PolicyConstraints_Id,
                     "sun.security.x509.PolicyConstraintsExtension");
-        addInternal(NETSCAPE_CERT, ObjectIdentifier.newInternal
-                    (new int[] {2,16,840,1,113730,1,1}),
+        addInternal(NETSCAPE_CERT,
+                    ObjectIdentifier.of("2.16.840.1.113730.1.1"),
                     "sun.security.x509.NetscapeCertTypeExtension");
         addInternal(CERT_POLICIES, PKIXExtensions.CertificatePolicies_Id,
                     "sun.security.x509.CertificatePoliciesExtension");
         addInternal(EXT_KEY_USAGE, PKIXExtensions.ExtendedKeyUsage_Id,
                     "sun.security.x509.ExtendedKeyUsageExtension");
