diff a/jcheck/src/main/java/org/openjdk/skara/jcheck/IssueVisitor.java b/jcheck/src/main/java/org/openjdk/skara/jcheck/IssueVisitor.java
--- a/jcheck/src/main/java/org/openjdk/skara/jcheck/IssueVisitor.java
+++ b/jcheck/src/main/java/org/openjdk/skara/jcheck/IssueVisitor.java
@@ -36,10 +36,11 @@
     void visit(CommitterEmailIssue issue);
     void visit(AuthorNameIssue issue);
     void visit(AuthorEmailIssue issue);
     void visit(WhitespaceIssue issue);
     void visit(MessageIssue issue);
+    void visit(MessageWhitespaceIssue issue);
     void visit(IssuesIssue issue);
     void visit(ExecutableIssue issue);
     void visit(BlacklistIssue issue);
     void visit(BinaryIssue issue);
 }
