diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/IssueUpdater.java
@@ -60,10 +60,14 @@
         this.setFixVersion = setFixVersion;
         this.fixVersions = fixVersions;
         this.prOnly = prOnly;
     }
 
+    static IssueUpdaterBuilder newBuilder() {
+        return new IssueUpdaterBuilder();
+    }
+
     private final static Set<String> primaryTypes = Set.of("Bug", "New Feature", "Enhancement", "Task", "Sub-task");
 
     private boolean isPrimaryIssue(Issue issue) {
         var properties = issue.properties();
         if (!properties.containsKey("issuetype")) {
