diff a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/SysVVaList.java b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/SysVVaList.java
--- a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/SysVVaList.java
+++ b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/SysVVaList.java
@@ -137,13 +137,11 @@
         MemoryAddress base = ms.baseAddress();
         VH_gp_offset.set(base, MAX_GP_OFFSET);
         VH_fp_offset.set(base, MAX_FP_OFFSET);
         VH_overflow_arg_area.set(base, MemoryAddress.NULL);
         VH_reg_save_area.set(base, MemoryAddress.NULL);
-        MemorySegment unconfined = NativeMemorySegmentImpl.makeNativeSegmentUnchecked(
-                base, ms.byteSize(), null, null, null).withAccessModes(0);
-        return unconfined.baseAddress();
+        return ms.withAccessModes(0).baseAddress();
     }
 
     public static CSupport.VaList empty() {
         return EMPTY;
     }
