diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/uset.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/uset.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/uset.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/uset.h
@@ -31,14 +31,18 @@
 
 #include "unicode/utypes.h"
 #include "unicode/uchar.h"
 #include "unicode/localpointer.h"
 
-#ifndef UCNV_H
-struct USet;
+#ifndef USET_DEFINED
+
+#ifndef U_IN_DOXYGEN
+#define USET_DEFINED
+#endif
 /**
- * A UnicodeSet.  Use the uset_* API to manipulate.  Create with
+ * USet is the C API type corresponding to C++ class UnicodeSet.
+ * Use the uset_* API to manipulate.  Create with
  * uset_open*, and destroy with uset_close.
  * @stable ICU 2.4
  */
 typedef struct USet USet;
 #endif
