diff a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/usprep.h b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/usprep.h
--- a/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/usprep.h
+++ b/modules/javafx.web/src/main/native/Source/ThirdParty/icu/source/common/unicode/usprep.h
@@ -187,11 +187,11 @@
             UErrorCode* status);
 
 /**
  * Creates a StringPrep profile for the specified profile type.
  *
- * @param type      The profile type
+ * @param type        The profile type
  * @param status    ICU error code in/out parameter. Must not be NULL.
  *                  Must fulfill U_SUCCESS before the function call.
  * @return          Pointer to UStringPrepProfile that is opened. Should be closed by
  *                  calling usprep_close()
  * @see usprep_close()
