diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.cpp
@@ -24,11 +24,14 @@
  */
 
 #include "config.h"
 #include "DeviceOrientationOrMotionEvent.h"
 
+#include "DOMWindow.h"
+#include "DeviceOrientationAndMotionAccessController.h"
 #include "Document.h"
+#include "JSDOMPromiseDeferred.h"
 
 namespace WebCore {
 
 #if ENABLE(DEVICE_ORIENTATION)
 void DeviceOrientationOrMotionEvent::requestPermission(Document& document, PermissionPromise&& promise)
