diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/SimulatedClick.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/SimulatedClick.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/SimulatedClick.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/SimulatedClick.cpp
@@ -46,11 +46,11 @@
 
 private:
     SimulatedMouseEvent(const AtomString& eventType, RefPtr<WindowProxy>&& view, RefPtr<Event>&& underlyingEvent, Element& target, SimulatedClickSource source)
         : MouseEvent(eventType, CanBubble::Yes, IsCancelable::Yes, IsComposed::Yes,
             underlyingEvent ? underlyingEvent->timeStamp() : MonotonicTime::now(), WTFMove(view), /* detail */ 0,
-            { }, { }, { }, modifiersFromUnderlyingEvent(underlyingEvent), 0, 0, nullptr, 0, 0, nullptr, IsSimulated::Yes,
+            { }, { }, { }, modifiersFromUnderlyingEvent(underlyingEvent), 0, 0, nullptr, 0, 0, IsSimulated::Yes,
             source == SimulatedClickSource::UserAgent ? IsTrusted::Yes : IsTrusted::No)
     {
         setUnderlyingEvent(underlyingEvent.get());
 
         if (is<MouseEvent>(this->underlyingEvent())) {
