diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGArgumentsUtilities.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGArgumentsUtilities.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGArgumentsUtilities.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGArgumentsUtilities.cpp
@@ -52,11 +52,11 @@
     return reg >= argumentStart && reg < argumentStart + numArguments;
 }
 
 bool argumentsInvolveStackSlot(Node* candidate, VirtualRegister reg)
 {
-    return argumentsInvolveStackSlot(candidate->origin.semantic.inlineCallFrame, reg);
+    return argumentsInvolveStackSlot(candidate->origin.semantic.inlineCallFrame(), reg);
 }
 
 Node* emitCodeToGetArgumentsArrayLength(
     InsertionSet& insertionSet, Node* arguments, unsigned nodeIndex, NodeOrigin origin)
 {
@@ -104,11 +104,11 @@
     if (arguments->op() == NewArrayBuffer || arguments->op() == PhantomNewArrayBuffer) {
         return insertionSet.insertConstant(
             nodeIndex, origin, jsNumber(arguments->castOperand<JSImmutableButterfly*>()->length()));
     }
 
-    InlineCallFrame* inlineCallFrame = arguments->origin.semantic.inlineCallFrame;
+    InlineCallFrame* inlineCallFrame = arguments->origin.semantic.inlineCallFrame();
 
     unsigned numberOfArgumentsToSkip = 0;
     if (arguments->op() == CreateRest || arguments->op() == PhantomCreateRest)
         numberOfArgumentsToSkip = arguments->numberOfArgumentsToSkip();
 
