diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGBasicBlock.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGBasicBlock.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGBasicBlock.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGBasicBlock.h
@@ -255,11 +255,11 @@
 private:
     friend class InsertionSet;
     BlockNodeList m_nodes;
 };
 
-typedef Vector<BasicBlock*, 5> BlockList;
+typedef Vector<BasicBlock*> BlockList;
 
 static inline unsigned getBytecodeBeginForBlock(BasicBlock** basicBlock)
 {
     return (*basicBlock)->bytecodeBegin;
 }
