diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPhase.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPhase.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPhase.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPhase.h
@@ -51,10 +51,12 @@
 
     Graph& graph() { return m_graph; }
 
     // Each phase must have a run() method.
 
+    Prefix prefix;
+
 protected:
     // Things you need to have a DFG compiler phase.
     Graph& m_graph;
 
     VM& vm() { return m_graph.m_vm; }
@@ -80,11 +82,11 @@
     CompilerTimingScope timingScope("DFG", phase.name());
 
     bool result = phase.run();
 
     if (result && logCompilationChanges(phase.graph().m_plan.mode()))
-        dataLogF("Phase %s changed the IR.\n", phase.name());
+        dataLogLn(phase.graph().prefix(), "Phase ", phase.name(), " changed the IR.\n");
     return result;
 }
 
 template<typename PhaseType>
 bool runPhase(Graph& graph)
