diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGStoreBarrierClusteringPhase.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGStoreBarrierClusteringPhase.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGStoreBarrierClusteringPhase.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGStoreBarrierClusteringPhase.cpp
@@ -138,11 +138,11 @@
             for (auto iter = m_neededBarriers.begin(); iter != m_neededBarriers.end(); ++iter) {
                 Node* child = iter->child;
                 CodeOrigin semanticOrigin = iter->semanticOrigin;
 
                 NodeType type;
-                if (Options::useConcurrentBarriers() && iter == m_neededBarriers.begin())
+                if (iter == m_neededBarriers.begin())
                     type = FencedStoreBarrier;
                 else
                     type = StoreBarrier;
 
                 m_insertionSet.insertNode(
