diff a/src/hotspot/cpu/x86/macroAssembler_x86.cpp b/src/hotspot/cpu/x86/macroAssembler_x86.cpp
--- a/src/hotspot/cpu/x86/macroAssembler_x86.cpp
+++ b/src/hotspot/cpu/x86/macroAssembler_x86.cpp
@@ -7869,34 +7869,10 @@
   jcc(Assembler::notZero, copy_chars_loop);
 
   bind(done);
 }
 
-void MacroAssembler::spill_register(VMReg reg) {
-  assert(reg->is_reg(), "must be a register");
-  if (reg->is_Register()) {
-    push(reg->as_Register());
-  } else if (reg->is_XMMRegister()) {
-    subptr(rsp, 16); // 16 bytes
-    movdqu(Address(rsp, 0), reg->as_XMMRegister());
-  } else {
-    ShouldNotReachHere();
-  }
-}
-
-void MacroAssembler::fill_register(VMReg reg) {
-  assert(reg->is_reg(), "must be a register");
-  if (reg->is_Register()) {
-    pop(reg->as_Register());
-  } else if (reg->is_XMMRegister()) {
-    movdqu(reg->as_XMMRegister(), Address(rsp, 0));
-    addptr(rsp, 16); // 16 bytes
-  } else {
-    ShouldNotReachHere();
-  }
-}
-
 #ifdef _LP64
 void MacroAssembler::convert_f2i(Register dst, XMMRegister src) {
   Label done;
   cvttss2sil(dst, src);
   // Conversion instructions do not match JLS for overflow, underflow and NaN -> fixup in stub
