diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocument.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocument.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocument.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocument.h
@@ -28,29 +28,29 @@
 class SVGSVGElement;
 
 class SVGDocument final : public XMLDocument {
     WTF_MAKE_ISO_ALLOCATED(SVGDocument);
 public:
-    static Ref<SVGDocument> create(Frame*, const URL&);
+    static Ref<SVGDocument> create(PAL::SessionID, Frame*, const URL&);
 
     static RefPtr<SVGSVGElement> rootElement(const Document&);
 
     bool zoomAndPanEnabled() const;
     void startPan(const FloatPoint& start);
     void updatePan(const FloatPoint& position) const;
 
 private:
-    SVGDocument(Frame*, const URL&);
+    SVGDocument(PAL::SessionID, Frame*, const URL&);
 
     Ref<Document> cloneDocumentWithoutChildren() const override;
 
     FloatSize m_panningOffset;
 };
 
-inline Ref<SVGDocument> SVGDocument::create(Frame* frame, const URL& url)
+inline Ref<SVGDocument> SVGDocument::create(PAL::SessionID sessionID, Frame* frame, const URL& url)
 {
-    return adoptRef(*new SVGDocument(frame, url));
+    return adoptRef(*new SVGDocument(sessionID, frame, url));
 }
 
 } // namespace WebCore
 
 SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::SVGDocument)
