diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceElement.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceElement.h
@@ -55,11 +55,11 @@
 
 private:
     SVGFontFaceElement(const QualifiedName&, Document&);
     ~SVGFontFaceElement();
 
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
 
     void childrenChanged(const ChildChange&) final;
     InsertedIntoAncestorResult insertedIntoAncestor(InsertionType, ContainerNode&) final;
     void removedFromAncestor(RemovalType, ContainerNode&) final;
 
