diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceUriElement.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceUriElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceUriElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGFontFaceUriElement.h
@@ -39,11 +39,11 @@
     Ref<CSSFontFaceSrcValue> srcValue() const;
 
 private:
     SVGFontFaceUriElement(const QualifiedName&, Document&);
 
-    void parseAttribute(const QualifiedName&, const AtomicString&) final;
+    void parseAttribute(const QualifiedName&, const AtomString&) final;
     void childrenChanged(const ChildChange&) final;
     InsertedIntoAncestorResult insertedIntoAncestor(InsertionType, ContainerNode&) final;
     bool rendererIsNeeded(const RenderStyle&) final { return false; }
 
     void loadFont();
