diff a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/BVMTags.h b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/BVMTags.h
--- a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/BVMTags.h
+++ b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/BVMTags.h
@@ -44,13 +44,13 @@
 #else
 #define VM_TAG_FOR_EXECUTABLEALLOCATOR_MEMORY VM_MAKE_TAG(64)
 #endif // defined(VM_MEMORY_JAVASCRIPT_JIT_EXECUTABLE_ALLOCATOR)
 
 #if defined(VM_MEMORY_JAVASCRIPT_JIT_REGISTER_FILE)
-#define VM_TAG_FOR_REGISTERFILE_MEMORY VM_MAKE_TAG(VM_MEMORY_JAVASCRIPT_JIT_REGISTER_FILE)
+#define VM_TAG_FOR_ISOHEAP_MEMORY VM_MAKE_TAG(VM_MEMORY_JAVASCRIPT_JIT_REGISTER_FILE)
 #else
-#define VM_TAG_FOR_REGISTERFILE_MEMORY VM_MAKE_TAG(65)
+#define VM_TAG_FOR_ISOHEAP_MEMORY VM_MAKE_TAG(65)
 #endif // defined(VM_MEMORY_JAVASCRIPT_JIT_REGISTER_FILE)
 
 #if defined(VM_MEMORY_JAVASCRIPT_CORE)
 #define VM_TAG_FOR_GIGACAGE_MEMORY VM_MAKE_TAG(VM_MEMORY_JAVASCRIPT_CORE)
 #else
@@ -60,20 +60,20 @@
 #else // BOS(DARWIN)
 
 #define VM_TAG_FOR_TCMALLOC_MEMORY -1
 #define VM_TAG_FOR_GIGACAGE_MEMORY -1
 #define VM_TAG_FOR_EXECUTABLEALLOCATOR_MEMORY -1
-#define VM_TAG_FOR_REGISTERFILE_MEMORY -1
+#define VM_TAG_FOR_ISOHEAP_MEMORY -1
 
 #endif // BOS(DARWIN)
 
 namespace bmalloc {
 
 enum class VMTag {
     Unknown = -1,
     Malloc = VM_TAG_FOR_TCMALLOC_MEMORY,
-    JSVMStack = VM_TAG_FOR_REGISTERFILE_MEMORY,
+    IsoHeap = VM_TAG_FOR_ISOHEAP_MEMORY,
     JSJITCode = VM_TAG_FOR_EXECUTABLEALLOCATOR_MEMORY,
     JSGigacage = VM_TAG_FOR_GIGACAGE_MEMORY,
 };
 
 } // namespace bmalloc
