diff a/bots/notify/src/test/java/org/openjdk/skara/bots/notify/UpdaterTests.java b/bots/notify/src/test/java/org/openjdk/skara/bots/notify/UpdaterTests.java
--- a/bots/notify/src/test/java/org/openjdk/skara/bots/notify/UpdaterTests.java
+++ b/bots/notify/src/test/java/org/openjdk/skara/bots/notify/UpdaterTests.java
@@ -63,11 +63,11 @@
         try (var credentials = new HostCredentials(testInfo);
              var tempFolder = new TemporaryDirectory();
              var jsonFolder = new TemporaryDirectory()) {
             var repo = credentials.getHostedRepository();
             var localRepo = CheckableRepository.init(tempFolder.path(), repo.getRepositoryType());
-            localRepo.fetch(repo.getUrl(), "testlock:testlock");
+            credentials.commitLock(localRepo);
             localRepo.pushAll(repo.getUrl());
 
             var tagStorage = createTagStorage(repo);
             var branchStorage = createBranchStorage(repo);
 
@@ -97,11 +97,11 @@
         try (var credentials = new HostCredentials(testInfo);
              var tempFolder = new TemporaryDirectory();
              var jsonFolder = new TemporaryDirectory()) {
             var repo = credentials.getHostedRepository();
             var localRepo = CheckableRepository.init(tempFolder.path(), repo.getRepositoryType());
-            localRepo.fetch(repo.getUrl(), "testlock:testlock");
+            credentials.commitLock(localRepo);
             var masterHash = localRepo.resolve("master").orElseThrow();
             localRepo.tag(masterHash, "jdk-12+1", "Added tag 1", "Duke", "duke@openjdk.java.net");
             localRepo.pushAll(repo.getUrl());
 
             var tagStorage = createTagStorage(repo);
