diff a/tests/system/src/test/java/test/javafx/scene/web/SVGTest.java b/tests/system/src/test/java/test/javafx/scene/web/SVGTest.java
--- a/tests/system/src/test/java/test/javafx/scene/web/SVGTest.java
+++ b/tests/system/src/test/java/test/javafx/scene/web/SVGTest.java
@@ -35,10 +35,11 @@
 import javafx.scene.web.WebView;
 import javafx.stage.Stage;
 import org.junit.AfterClass;
 import org.junit.Before;
 import org.junit.BeforeClass;
+import org.junit.Ignore;
 import org.junit.Test;
 import test.util.Util;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
@@ -93,10 +94,11 @@
     /**
      * @test
      * @bug 8223298
      * summary Checks if svg pattern is displayed properly
      */
+    @Ignore("JDK-8243110")
     @Test public void testSVGRenderingWithPattern() {
         final CountDownLatch webViewStateLatch = new CountDownLatch(1);
         final String htmlSVGContent = "\n"
             + "<html>\n"
             + "<body style='margin: 0px 0px;'>\n"
