diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/WHLSLVisitor.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/WHLSLVisitor.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/WHLSLVisitor.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/WHLSLVisitor.h
@@ -61,11 +61,10 @@
 class SpecializationConstantSemantic;
 class StageInOutSemantic;
 class IntegerLiteral;
 class UnsignedIntegerLiteral;
 class FloatLiteral;
-class NullLiteral;
 class BooleanLiteral;
 class EnumerationMemberLiteral;
 class NumThreadsFunctionAttribute;
 class Block;
 class StatementList;
@@ -136,16 +135,14 @@
     virtual void visit(AST::SpecializationConstantSemantic&);
     virtual void visit(AST::StageInOutSemantic&);
     virtual void visit(AST::IntegerLiteral&);
     virtual void visit(AST::UnsignedIntegerLiteral&);
     virtual void visit(AST::FloatLiteral&);
-    virtual void visit(AST::NullLiteral&);
     virtual void visit(AST::BooleanLiteral&);
     virtual void visit(AST::IntegerLiteralType&);
     virtual void visit(AST::UnsignedIntegerLiteralType&);
     virtual void visit(AST::FloatLiteralType&);
-    virtual void visit(AST::NullLiteralType&);
     virtual void visit(AST::EnumerationMemberLiteral&);
     virtual void visit(AST::FunctionAttribute&);
     virtual void visit(AST::NumThreadsFunctionAttribute&);
     virtual void visit(AST::Block&);
     virtual void visit(AST::StatementList&);
