diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLStatement.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLStatement.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLStatement.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLStatement.cpp
@@ -27,10 +27,11 @@
  */
 #include "config.h"
 #include "SQLStatement.h"
 
 #include "Database.h"
+#include "Document.h"
 #include "Logging.h"
 #include "SQLError.h"
 #include "SQLResultSet.h"
 #include "SQLStatementCallback.h"
 #include "SQLStatementErrorCallback.h"
@@ -75,12 +76,12 @@
 namespace WebCore {
 
 SQLStatement::SQLStatement(Database& database, const String& statement, Vector<SQLValue>&& arguments, RefPtr<SQLStatementCallback>&& callback, RefPtr<SQLStatementErrorCallback>&& errorCallback, int permissions)
     : m_statement(statement.isolatedCopy())
     , m_arguments(WTFMove(arguments))
-    , m_statementCallbackWrapper(WTFMove(callback), &database.scriptExecutionContext())
-    , m_statementErrorCallbackWrapper(WTFMove(errorCallback), &database.scriptExecutionContext())
+    , m_statementCallbackWrapper(WTFMove(callback), &database.document())
+    , m_statementErrorCallbackWrapper(WTFMove(errorCallback), &database.document())
     , m_permissions(permissions)
 {
 }
 
 SQLStatement::~SQLStatement() = default;
