diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLTransaction.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLTransaction.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLTransaction.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webdatabase/SQLTransaction.h
@@ -33,11 +33,10 @@
 #include "SQLTransactionBackend.h"
 #include "SQLTransactionStateMachine.h"
 #include "SQLValue.h"
 #include <wtf/Deque.h>
 #include <wtf/Lock.h>
-#include <wtf/Optional.h>
 
 namespace WebCore {
 
 class Database;
 class SQLError;
