diff a/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp
@@ -59,11 +59,11 @@
     return { };
 }
 
 void WorkerNetworkAgent::setResourceCachingDisabled(bool disabled)
 {
-    m_workerGlobalScope.thread().workerDebuggerProxy().setResourceCachingDisabled(disabled);
+    m_workerGlobalScope.thread().workerDebuggerProxy().setResourceCachingDisabledByWebInspector(disabled);
 }
 
 ScriptExecutionContext* WorkerNetworkAgent::scriptExecutionContext(ErrorString&, const String&)
 {
     return &m_workerGlobalScope;
