diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/websockets/WebSocketChannelClient.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/websockets/WebSocketChannelClient.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/websockets/WebSocketChannelClient.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/websockets/WebSocketChannelClient.h
@@ -29,14 +29,15 @@
  */
 
 #pragma once
 
 #include <wtf/Forward.h>
+#include <wtf/WeakPtr.h>
 
 namespace WebCore {
 
-class WebSocketChannelClient {
+class WebSocketChannelClient : public CanMakeWeakPtr<WebSocketChannelClient> {
 public:
     virtual ~WebSocketChannelClient() = default;
     virtual void didConnect() = 0;
     virtual void didReceiveMessage(const String&) = 0;
     virtual void didReceiveBinaryData(Vector<uint8_t>&&) = 0;
