diff a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ImageConstants.java b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ImageConstants.java
--- a/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ImageConstants.java
+++ b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ImageConstants.java
@@ -108,11 +108,11 @@
 
 	public static final String ICON_IO_BY_HOST = "io_by_host.png"; //$NON-NLS-1$
 	public static final String ICON_IO_BY_PORT = "io_by_port.png"; //$NON-NLS-1$
 	public static final String ICON_IO_BY_HOST_AND_PORT = "io_by_host_and_port.png"; //$NON-NLS-1$
 
-	public static final String ICON_FLAMEVIEW_SEARCH = "magnifyer64.png"; //$NON-NLS-1$
+	public static final String ICON_MAGNIFIER = "magnifier.png"; //$NON-NLS-1$
 
 	public static final String PAGE_JAVA_APPLICATION = "pages/methodprofiling.png"; //$NON-NLS-1$
 	public static final String PAGE_METHOD = "pages/method.png"; //$NON-NLS-1$
 	public static final String PAGE_HEAP = "pages/objectstatistics.png"; //$NON-NLS-1$
 	public static final String PAGE_LOCKINSTANCES = "pages/locks.png"; //$NON-NLS-1$
