diff a/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/DFAMinimizer.cpp b/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/DFAMinimizer.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/DFAMinimizer.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/DFAMinimizer.cpp
@@ -407,21 +407,21 @@
     }
 
     bool isEmptyValue() const { return state == Empty; }
     bool isDeletedValue() const { return state == Deleted; }
 
-    unsigned hash;
+    unsigned hash { 0 };
 
-    const DFA* dfa;
-    uint32_t actionsStart;
-    uint16_t actionsLength;
+    const DFA* dfa { nullptr };
+    uint32_t actionsStart { 0 };
+    uint16_t actionsLength { 0 };
 
     enum {
-        Valid,
         Empty,
+        Valid,
         Deleted
-    } state;
+    } state { Empty };
 };
 
 struct ActionKeyHash {
     static unsigned hash(const ActionKey& actionKey)
     {
