diff a/src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.hpp b/src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.hpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.hpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.hpp
@@ -91,11 +91,20 @@
   SHENANDOAH_PAR_PHASE_DO(evac_,                    "    E: ", f)                      \
                                                                                        \
   f(conc_weak_roots,                                "Concurrent Weak Roots")           \
   SHENANDOAH_PAR_PHASE_DO(conc_weak_roots_,         "  CWR: ", f)                      \
   f(conc_cleanup_early,                             "Concurrent Cleanup")              \
-  f(conc_class_unloading,                           "Concurrent Class Unloading")      \
+  f(conc_class_unload,                              "Concurrent Class Unloading")      \
+  f(conc_class_unload_unlink,                       "  Unlink Stale")                  \
+  f(conc_class_unload_unlink_sd,                    "    System Dictionary")           \
+  f(conc_class_unload_unlink_weak_klass,            "    Weak Class Links")            \
+  f(conc_class_unload_unlink_code_roots,            "    Code Roots")                  \
+  f(conc_class_unload_rendezvous,                   "  Rendezvous")                    \
+  f(conc_class_unload_purge,                        "  Purge Unlinked")                \
+  f(conc_class_unload_purge_coderoots,              "    Code Roots")                  \
+  f(conc_class_unload_purge_cldg,                   "    CLDG")                        \
+  f(conc_class_unload_purge_ec,                     "    Exception Caches")            \
   f(conc_strong_roots,                              "Concurrent Strong Roots")         \
   SHENANDOAH_PAR_PHASE_DO(conc_strong_roots_,       "  CSR: ", f)                      \
   f(conc_evac,                                      "Concurrent Evacuation")           \
                                                                                        \
   f(init_update_refs_gross,                         "Pause Init  Update Refs (G)")     \
