diff a/src/java.compiler/share/classes/javax/lang/model/element/UnknownAnnotationValueException.java b/src/java.compiler/share/classes/javax/lang/model/element/UnknownAnnotationValueException.java
--- a/src/java.compiler/share/classes/javax/lang/model/element/UnknownAnnotationValueException.java
+++ b/src/java.compiler/share/classes/javax/lang/model/element/UnknownAnnotationValueException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2005, 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -58,11 +58,11 @@
      *
      * @param av the unknown annotation value, may be {@code null}
      * @param p an additional parameter, may be {@code null}
      */
     public UnknownAnnotationValueException(AnnotationValue av, Object p) {
-        super("Unknown annotation value: " + av);
+        super("Unknown annotation value: \"" + av + "\"");
         this.av = av;
         this.parameter = p;
     }
 
     /**
