diff a/src/java.base/share/classes/java/util/ConcurrentModificationException.java b/src/java.base/share/classes/java/util/ConcurrentModificationException.java
--- a/src/java.base/share/classes/java/util/ConcurrentModificationException.java
+++ b/src/java.base/share/classes/java/util/ConcurrentModificationException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1997, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -66,10 +66,11 @@
  * @see     TreeMap
  * @see     AbstractList
  * @since   1.2
  */
 public class ConcurrentModificationException extends RuntimeException {
+    @java.io.Serial
     private static final long serialVersionUID = -3666751008965953603L;
 
     /**
      * Constructs a ConcurrentModificationException with no
      * detail message.
@@ -104,11 +105,11 @@
 
     /**
      * Constructs a new exception with the specified detail message and
      * cause.
      *
-     * <p>Note that the detail message associated with <code>cause</code> is
+     * <p>Note that the detail message associated with {@code cause} is
      * <i>not</i> automatically incorporated in this exception's detail
      * message.
      *
      * @param  message the detail message (which is saved for later retrieval
      *         by the {@link Throwable#getMessage()} method).
