diff a/src/java.base/share/classes/java/util/InputMismatchException.java b/src/java.base/share/classes/java/util/InputMismatchException.java
--- a/src/java.base/share/classes/java/util/InputMismatchException.java
+++ b/src/java.base/share/classes/java/util/InputMismatchException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2008, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -32,12 +32,12 @@
  *
  * @author  unascribed
  * @see     java.util.Scanner
  * @since   1.5
  */
-public
-class InputMismatchException extends NoSuchElementException {
+public class InputMismatchException extends NoSuchElementException {
+    @java.io.Serial
     private static final long serialVersionUID = 8811230760997066428L;
 
     /**
      * Constructs an {@code InputMismatchException} with {@code null}
      * as its error message string.
