diff a/src/java.base/share/classes/java/util/Stack.java b/src/java.base/share/classes/java/util/Stack.java
--- a/src/java.base/share/classes/java/util/Stack.java
+++ b/src/java.base/share/classes/java/util/Stack.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1994, 2010, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1994, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -43,12 +43,11 @@
  *   Deque<Integer> stack = new ArrayDeque<Integer>();}</pre>
  *
  * @author  Jonathan Payne
  * @since   1.0
  */
-public
-class Stack<E> extends Vector<E> {
+public class Stack<E> extends Vector<E> {
     /**
      * Creates an empty Stack.
      */
     public Stack() {
     }
@@ -135,7 +134,8 @@
         }
         return -1;
     }
 
     /** use serialVersionUID from JDK 1.0.2 for interoperability */
+    @java.io.Serial
     private static final long serialVersionUID = 1224463164541339165L;
 }
