diff a/src/java.base/share/classes/java/util/TooManyListenersException.java b/src/java.base/share/classes/java/util/TooManyListenersException.java
--- a/src/java.base/share/classes/java/util/TooManyListenersException.java
+++ b/src/java.base/share/classes/java/util/TooManyListenersException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1996, 2008, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1996, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -25,11 +25,11 @@
 
 package java.util;
 
 /**
  * <p>
- * The <code> TooManyListenersException </code> Exception is used as part of
+ * The {@code  TooManyListenersException } Exception is used as part of
  * the Java Event model to annotate and implement a unicast special case of
  * a multicast Event Source.
  * </p>
  * <p>
  * The presence of a "throws TooManyListenersException" clause on any given
@@ -46,10 +46,11 @@
  * @author Laurence P. G. Cable
  * @since  1.1
  */
 
 public class TooManyListenersException extends Exception {
+    @java.io.Serial
     private static final long serialVersionUID = 5074640544770687831L;
 
     /**
      * Constructs a TooManyListenersException with no detail message.
      * A detail message is a String that describes this particular exception.
