diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/ContentData.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/ContentData.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/ContentData.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/ContentData.cpp
@@ -47,18 +47,18 @@
 RenderPtr<RenderObject> ImageContentData::createContentRenderer(Document& document, const RenderStyle& pseudoStyle) const
 {
     auto image = createRenderer<RenderImage>(document, RenderStyle::createStyleInheritingFromPseudoStyle(pseudoStyle), const_cast<StyleImage*>(m_image.ptr()));
     image->initializeStyle();
     image->setAltText(altText());
-    return WTFMove(image);
+    return image;
 }
 
 RenderPtr<RenderObject> TextContentData::createContentRenderer(Document& document, const RenderStyle&) const
 {
     auto fragment = createRenderer<RenderTextFragment>(document, m_text);
     fragment->setAltText(altText());
-    return WTFMove(fragment);
+    return fragment;
 }
 
 RenderPtr<RenderObject> CounterContentData::createContentRenderer(Document& document, const RenderStyle&) const
 {
     return createRenderer<RenderCounter>(document, *m_counter);
@@ -66,9 +66,9 @@
 
 RenderPtr<RenderObject> QuoteContentData::createContentRenderer(Document& document, const RenderStyle& pseudoStyle) const
 {
     auto quote = createRenderer<RenderQuote>(document, RenderStyle::createStyleInheritingFromPseudoStyle(pseudoStyle), m_quote);
     quote->initializeStyle();
-    return WTFMove(quote);
+    return quote;
 }
 
 } // namespace WebCore
