diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/dtoa/fixed-dtoa.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/dtoa/fixed-dtoa.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/dtoa/fixed-dtoa.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/dtoa/fixed-dtoa.h
@@ -48,11 +48,11 @@
 // shortest representation of the input.
 //
 // This method only works for some parameters. If it can't handle the input it
 // returns false. The output is null-terminated when the function succeeds.
 bool FastFixedDtoa(double v, int fractional_count,
-                       BufferReference<char> buffer, int* length, int* decimal_point);
+                   BufferReference<char> buffer, int* length, int* decimal_point);
 
 }  // namespace double_conversion
-} // namespace WTF
+}  // namespace WTF
 
 #endif  // DOUBLE_CONVERSION_FIXED_DTOA_H_
