diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleInheritedData.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleInheritedData.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleInheritedData.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/StyleInheritedData.cpp
@@ -24,10 +24,12 @@
 
 #include "RenderStyle.h"
 
 namespace WebCore {
 
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StyleInheritedData);
+
 StyleInheritedData::StyleInheritedData()
     : horizontalBorderSpacing(RenderStyle::initialHorizontalBorderSpacing())
     , verticalBorderSpacing(RenderStyle::initialVerticalBorderSpacing())
     , lineHeight(RenderStyle::initialLineHeight())
 #if ENABLE(TEXT_AUTOSIZING)
