diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/SVGRenderStyleDefs.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/SVGRenderStyleDefs.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/SVGRenderStyleDefs.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/style/SVGRenderStyleDefs.h
@@ -32,10 +32,14 @@
 #include "SVGLengthValue.h"
 #include "ShadowData.h"
 #include <wtf/RefCounted.h>
 #include <wtf/RefPtr.h>
 
+namespace WTF {
+class TextStream;
+}
+
 namespace WebCore {
 
 class CSSValue;
 class CSSValueList;
 class SVGPaint;
@@ -135,11 +139,13 @@
     Luminance,
     Alpha
 };
 
 // Inherited/Non-Inherited Style Datastructures
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StyleFillData);
 class StyleFillData : public RefCounted<StyleFillData> {
+    WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER(StyleFillData);
 public:
     static Ref<StyleFillData> create() { return adoptRef(*new StyleFillData); }
     Ref<StyleFillData> copy() const;
 
     bool operator==(const StyleFillData&) const;
@@ -280,11 +286,10 @@
     bool operator!=(const StyleResourceData& other) const
     {
         return !(*this == other);
     }
 
-    String clipper;
     String masker;
 
 private:
     StyleResourceData();
     StyleResourceData(const StyleResourceData&);
@@ -334,6 +339,30 @@
 private:
     StyleLayoutData();
     StyleLayoutData(const StyleLayoutData&);
 };
 
+
+WTF::TextStream& operator<<(WTF::TextStream&, AlignmentBaseline);
+WTF::TextStream& operator<<(WTF::TextStream&, BaselineShift);
+WTF::TextStream& operator<<(WTF::TextStream&, BufferedRendering);
+WTF::TextStream& operator<<(WTF::TextStream&, ColorInterpolation);
+WTF::TextStream& operator<<(WTF::TextStream&, ColorRendering);
+WTF::TextStream& operator<<(WTF::TextStream&, DominantBaseline);
+WTF::TextStream& operator<<(WTF::TextStream&, GlyphOrientation);
+WTF::TextStream& operator<<(WTF::TextStream&, MaskType);
+WTF::TextStream& operator<<(WTF::TextStream&, SVGPaintType);
+WTF::TextStream& operator<<(WTF::TextStream&, ShapeRendering);
+WTF::TextStream& operator<<(WTF::TextStream&, TextAnchor);
+WTF::TextStream& operator<<(WTF::TextStream&, VectorEffect);
+
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleFillData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleStrokeData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleStopData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleTextData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleMiscData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleShadowSVGData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleResourceData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleInheritedResourceData&);
+WTF::TextStream& operator<<(WTF::TextStream&, const StyleLayoutData&);
+
 } // namespace WebCore
