diff a/test/src/main/java/org/openjdk/skara/test/TestMailmanServer.java b/test/src/main/java/org/openjdk/skara/test/TestMailmanServer.java
--- a/test/src/main/java/org/openjdk/skara/test/TestMailmanServer.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestMailmanServer.java
@@ -103,21 +103,25 @@
         var listPath = Files.createTempFile("list-" + name, ".txt");
         lists.put(name, listPath);
         return listName;
     }
 
-    public void processIncoming() throws IOException {
-        var email = smtpServer.receive(Duration.ofSeconds(10));
+    public void processIncoming(Duration timeout) throws IOException {
+        var email = smtpServer.receive(timeout);
         var mboxEntry = Mbox.fromMail(email);
 
         var listPath = email.recipients().stream()
                             .filter(recipient -> lists.containsKey(recipient.localPart()))
                             .map(recipient -> lists.get(recipient.localPart()))
+    public void processIncoming() throws IOException {
+        processIncoming(Duration.ofSeconds(10));
+    }
+
     @Override
     public void close() throws IOException {
         httpServer.stop(0);
         smtpServer.close();
     }
