diff a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
--- a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
@@ -29,10 +29,11 @@
 import java.io.*;
 import java.net.*;
 import java.nio.file.Path;
 import java.util.*;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 public class TestHostedRepository implements HostedRepository {
     private final TestHost host;
     private final String projectName;
     private final Repository localRepository;
@@ -68,10 +69,21 @@
     @Override
     public List<PullRequest> getPullRequests() {
         return new ArrayList<>(host.getPullRequests(this));
     }
 
+    @Override
+    public List<PullRequest> findPullRequestsWithComment(String author, String body) {
+        return getPullRequests().stream()
+                                .filter(pr -> pr.getComments().stream()
+                                        .filter(comment -> author == null || comment.author().userName().equals(author))
+                                        .filter(comment -> comment == null ||comment.body().contains(body))
+                                        .count() > 0
+                                )
+                                .collect(Collectors.toList());
+    }
+
     @Override
     public Optional<PullRequest> parsePullRequestUrl(String url) {
         var matcher = pullRequestPattern.matcher(url);
         if (matcher.find()) {
             return Optional.of(getPullRequest(matcher.group(1)));
