diff a/test/jdk/java/util/Locale/Bug8040211.java b/test/jdk/java/util/Locale/Bug8040211.java
--- a/test/jdk/java/util/Locale/Bug8040211.java
+++ b/test/jdk/java/util/Locale/Bug8040211.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2016, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2016, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -21,13 +21,13 @@
  * questions.
  */
 
 /*
  * @test
- * @bug 8040211 8191404 8203872 8222980 8225435
+ * @bug 8040211 8191404 8203872 8222980 8225435 8241082 8242010
  * @summary Checks the IANA language subtag registry data update
- *          (LSR Revision: 2019-09-16) with Locale and Locale.LanguageRange
+ *          (LSR Revision: 2020-04-01) with Locale and Locale.LanguageRange
  *          class methods.
  * @run main Bug8040211
  */
 
 import java.util.ArrayList;
@@ -41,13 +41,13 @@
 public class Bug8040211 {
 
     static boolean err = false;
 
     private static final String ACCEPT_LANGUAGE =
-        "Accept-Language: aam, adp, aog, aue, bcg, cey, cqu, dif, ema,"
-        + " en-gb-oed, gti, kdz, koj, kwq, kxe, lii, lmm, lsn, lsv, lvi, mtm,"
-        + " ngv, nns, oyb, phr, pnd, pub, snz, suj, szy,taj, tjj, tjp, tvx,"
+        "Accept-Language: aam, adp, aog, aue, bcg, bpp, cey, cnp, cqu, csp, dif, dmw, ema,"
+        + " en-gb-oed, gti, kdz, koj, kru, kwq, kxe, kzk, lii, lmm, lsn, lsv, lvi, mtm,"
+        + " ngv, nns, oyb, phr, pnd, pub, scv, snz, suj, szy, taj, tjj, tjp, tvx,"
         + " uss, uth, wkr;q=0.9, ar-hyw;q=0.8, yug;q=0.5, gfx;q=0.4";
     private static final List<LanguageRange> EXPECTED_RANGE_LIST = List.of(
             new LanguageRange("aam", 1.0),
             new LanguageRange("aas", 1.0),
             new LanguageRange("adp", 1.0),
@@ -56,35 +56,50 @@
             new LanguageRange("myd", 1.0),
             new LanguageRange("aue", 1.0),
             new LanguageRange("ktz", 1.0),
             new LanguageRange("bcg", 1.0),
             new LanguageRange("bgm", 1.0),
+            new LanguageRange("bpp", 1.0),
+            new LanguageRange("nxu", 1.0),
             new LanguageRange("cey", 1.0),
+            new LanguageRange("cnp", 1.0),
+            new LanguageRange("zh-cnp", 1.0),
             new LanguageRange("cqu", 1.0),
             new LanguageRange("quh", 1.0),
+            new LanguageRange("csp", 1.0),
+            new LanguageRange("zh-csp", 1.0),
             new LanguageRange("dif", 1.0),
             new LanguageRange("dit", 1.0),
+            new LanguageRange("dmw", 1.0),
+            new LanguageRange("xrq", 1.0),
             new LanguageRange("ema", 1.0),
             new LanguageRange("uok", 1.0),
             new LanguageRange("en-gb-oed", 1.0),
             new LanguageRange("en-gb-oxendict", 1.0),
             new LanguageRange("gti", 1.0),
             new LanguageRange("nyc", 1.0),
             new LanguageRange("kdz", 1.0),
             new LanguageRange("ncp", 1.0),
             new LanguageRange("koj", 1.0),
             new LanguageRange("kwv", 1.0),
+            new LanguageRange("kru", 1.0),
+            new LanguageRange("kxl", 1.0),
             new LanguageRange("kwq", 1.0),
             new LanguageRange("yam", 1.0),
             new LanguageRange("kxe", 1.0),
             new LanguageRange("tvd", 1.0),
+            new LanguageRange("kzk", 1.0),
+            new LanguageRange("gli", 1.0),
+            new LanguageRange("drr", 1.0),
             new LanguageRange("lii", 1.0),
             new LanguageRange("raq", 1.0),
             new LanguageRange("lmm", 1.0),
             new LanguageRange("rmx", 1.0),
             new LanguageRange("lsn", 1.0),
+            new LanguageRange("sgn-lsn", 1.0),
             new LanguageRange("lsv", 1.0),
+            new LanguageRange("sgn-lsv", 1.0),
             new LanguageRange("lvi", 1.0),
             new LanguageRange("mtm", 1.0),
             new LanguageRange("ymt", 1.0),
             new LanguageRange("ngv", 1.0),
             new LanguageRange("nnx", 1.0),
@@ -97,10 +112,12 @@
             new LanguageRange("phr", 1.0),
             new LanguageRange("pmu", 1.0),
             new LanguageRange("pnd", 1.0),
             new LanguageRange("pub", 1.0),
             new LanguageRange("puz", 1.0),
+            new LanguageRange("scv", 1.0),
+            new LanguageRange("zir", 1.0),
             new LanguageRange("snz", 1.0),
             new LanguageRange("asd", 1.0),
             new LanguageRange("suj", 1.0),
             new LanguageRange("szy", 1.0),
             new LanguageRange("taj", 1.0),
