diff a/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java b/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java
--- a/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java
+++ b/test/langtools/jdk/javadoc/doclet/testHtmlTag/TestHtmlTag.java
@@ -23,11 +23,10 @@
 
 /*
  * @test
  * @bug 6786682 4649116 8182765
  * @summary This test verifies the use of lang attribute by <HTML>.
- * @author Bhavesh Patel
  * @library ../../lib
  * @modules jdk.javadoc/jdk.javadoc.internal.tool
  * @build javadoc.tester.*
  * @run main TestHtmlTag
  */
@@ -107,13 +106,11 @@
                 "pkg3");
         checkExit(Exit.OK);
 
         checkOutput("pkg3/package-summary.html", true,
                 "<div class=\"contentContainer\">\n"
-                + "<section role=\"region\"><a id=\"package.description\">\n"
-                + "<!--   -->\n"
-                + "</a>\n"
+                + "<section class=\"packageDescription\" id=\"package.description\">\n"
                 + "<div class=\"block\"><p>This is the first line. Note the newlines before the &lt;p&gt; is relevant.</div>\n"
                 + "</section>");
 
         checkOutput("pkg3/A.DatatypeFactory.html", true,
                 "<div class=\"block\"><p>\n"
