diff a/test/hotspot/jtreg/runtime/jni/atExit/libatExit.c b/test/hotspot/jtreg/runtime/jni/atExit/libatExit.c
--- a/test/hotspot/jtreg/runtime/jni/atExit/libatExit.c
+++ b/test/hotspot/jtreg/runtime/jni/atExit/libatExit.c
@@ -102,11 +102,12 @@
     report("JNI_CreateJavaVM", res, using_system_exit ? JNI_EEXIST : JNI_ERR);
   }
   // else test has already failed
 }
 
-jint JNI_OnLoad(JavaVM *vm, void *reserved) {
+JNIEXPORT jint JNICALL
+JNI_OnLoad(JavaVM *vm, void *reserved) {
   printf("JNI_OnLoad: registering atexit handler\n");
   jvm = vm;
   atexit(at_exit_handler);
 
   return JNI_VERSION_1_1;
