diff a/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java b/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java
--- a/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java
+++ b/src/jdk.incubator.jextract/share/classes/jdk/internal/clang/Type.java
@@ -107,11 +107,11 @@
     }
 
     // Struct/RecordType
     private long getOffsetOf0(String fieldName) {
         try (MemorySegment cfname = Utils.toNativeString(fieldName)) {
-            return Index_h.clang_Type_getOffsetOf(type, cfname.baseAddress());
+            return Index_h.clang_Type_getOffsetOf(type, cfname.address());
         }
     }
 
     public long getOffsetOf(String fieldName) {
         long res = getOffsetOf0(fieldName);
