diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/LazyOperandValueProfile.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/LazyOperandValueProfile.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/LazyOperandValueProfile.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/LazyOperandValueProfile.h
@@ -127,21 +127,22 @@
         , m_operand(VirtualRegister())
     {
     }
 
     explicit LazyOperandValueProfile(const LazyOperandValueProfileKey& key)
-        : MinimalValueProfile(key.bytecodeOffset())
-        , m_operand(key.operand())
+        : MinimalValueProfile()
+        , m_key(key)
     {
     }
 
     LazyOperandValueProfileKey key() const
     {
-        return LazyOperandValueProfileKey(m_bytecodeOffset, m_operand);
+        return m_key;
     }
 
     VirtualRegister m_operand;
+    LazyOperandValueProfileKey m_key;
 
     typedef SegmentedVector<LazyOperandValueProfile, 8> List;
 };
 
 class LazyOperandValueProfileParser;
