diff a/src/hotspot/share/runtime/signature.cpp b/src/hotspot/share/runtime/signature.cpp
--- a/src/hotspot/share/runtime/signature.cpp
+++ b/src/hotspot/share/runtime/signature.cpp
@@ -420,17 +420,11 @@
   }
   Klass* klass = as_klass(class_loader, protection_domain, failure_mode, CHECK_NULL);
   if (klass == NULL) {
     return NULL;
   }
-  if (klass->is_value()) {
-    ValueKlass* vk = ValueKlass::cast(InstanceKlass::cast(klass));
-    return _type == T_VALUETYPE ? vk->value_mirror() : vk->indirect_mirror();
-  } else {
-    assert(_type != T_VALUETYPE, "must not be value type");
-    return klass->java_mirror();
-  }
+  return klass->java_mirror();
 }
 
 void SignatureStream::skip_to_return_type() {
   while (!at_return_type()) {
     next();
