diff a/test/langtools/jdk/javadoc/doclet/testMethodTypes/TestMethodTypes.java b/test/langtools/jdk/javadoc/doclet/testMethodTypes/TestMethodTypes.java
--- a/test/langtools/jdk/javadoc/doclet/testMethodTypes/TestMethodTypes.java
+++ b/test/langtools/jdk/javadoc/doclet/testMethodTypes/TestMethodTypes.java
@@ -47,70 +47,67 @@
                 "pkg1");
         checkExit(Exit.OK);
 
         checkOutput("pkg1/A.html", true,
                 "var data = {",
-                "<div role=\"tablist\" aria-orientation=\"horizontal\"><button role=\"tab\""
-                + " aria-selected=\"true\" aria-controls=\"member-summary_tabpanel\" tabindex=\"0\""
+                "<div class=\"table-tabs\" role=\"tablist\" aria-orientation=\"horizontal\"><button role=\"tab\""
+                + " aria-selected=\"true\" aria-controls=\"method-summary-table.tabpanel\" tabindex=\"0\""
                 + " onkeydown=\"switchTab(event)\" id=\"t0\" class=\"active-table-tab\">All Methods</button>"
-                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"member-summary_tabpanel\""
+                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"method-summary-table.tabpanel\""
                 + " tabindex=\"-1\" onkeydown=\"switchTab(event)\" id=\"t1\" class=\"table-tab\""
                 + " onclick=\"show(1);\">Static Methods</button><button role=\"tab\" aria-selected=\"false\""
-                + " aria-controls=\"member-summary_tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
+                + " aria-controls=\"method-summary-table.tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
                 + " id=\"t2\" class=\"table-tab\" onclick=\"show(2);\">Instance Methods</button>"
-                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"member-summary_tabpanel\""
+                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"method-summary-table.tabpanel\""
                 + " tabindex=\"-1\" onkeydown=\"switchTab(event)\" id=\"t4\" class=\"table-tab\""
                 + " onclick=\"show(8);\">Concrete Methods</button><button role=\"tab\""
-                + " aria-selected=\"false\" aria-controls=\"member-summary_tabpanel\" tabindex=\"-1\""
+                + " aria-selected=\"false\" aria-controls=\"method-summary-table.tabpanel\" tabindex=\"-1\""
                 + " onkeydown=\"switchTab(event)\" id=\"t6\" class=\"table-tab\" onclick=\"show(32);\">"
                 + "Deprecated Methods</button></div>",
                 "<tr class=\"alt-color\" id=\"i0\">");
 
         checkOutput("pkg1/B.html", true,
                 "var data = {\"i0\":6,\"i1\":18,\"i2\":18,\"i3\":1,\"i4\":1,"
                 + "\"i5\":6,\"i6\":6,\"i7\":6,\"i8\":6};\n",
-                "<div role=\"tablist\" aria-orientation=\"horizontal\"><button role=\"tab\""
-                + " aria-selected=\"true\" aria-controls=\"member-summary_tabpanel\" tabindex=\"0\""
+                "<div class=\"table-tabs\" role=\"tablist\" aria-orientation=\"horizontal\"><button role=\"tab\""
+                + " aria-selected=\"true\" aria-controls=\"method-summary-table.tabpanel\" tabindex=\"0\""
                 + " onkeydown=\"switchTab(event)\" id=\"t0\" class=\"active-table-tab\">All Methods"
                 + "</button><button role=\"tab\" aria-selected=\"false\""
-                + " aria-controls=\"member-summary_tabpanel\" tabindex=\"-1\""
+                + " aria-controls=\"method-summary-table.tabpanel\" tabindex=\"-1\""
                 + " onkeydown=\"switchTab(event)\" id=\"t1\" class=\"table-tab\" onclick=\"show(1);\">"
                 + "Static Methods</button><button role=\"tab\" aria-selected=\"false\""
-                + " aria-controls=\"member-summary_tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
+                + " aria-controls=\"method-summary-table.tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
                 + " id=\"t2\" class=\"table-tab\" onclick=\"show(2);\">Instance Methods</button>"
-                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"member-summary_tabpanel\""
+                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"method-summary-table.tabpanel\""
                 + " tabindex=\"-1\" onkeydown=\"switchTab(event)\" id=\"t3\" class=\"table-tab\""
                 + " onclick=\"show(4);\">Abstract Methods</button><button role=\"tab\""
-                + " aria-selected=\"false\" aria-controls=\"member-summary_tabpanel\" tabindex=\"-1\""
+                + " aria-selected=\"false\" aria-controls=\"method-summary-table.tabpanel\" tabindex=\"-1\""
                 + " onkeydown=\"switchTab(event)\" id=\"t5\" class=\"table-tab\" onclick=\"show(16);\">"
                 + "Default Methods</button></div>");
 
         checkOutput("pkg1/D.html", true,
                 "var data = {",
-                "<div role=\"tablist\" aria-orientation=\"horizontal\"><button role=\"tab\""
-                + " aria-selected=\"true\" aria-controls=\"member-summary_tabpanel\" tabindex=\"0\""
+                "<div class=\"table-tabs\" role=\"tablist\" aria-orientation=\"horizontal\"><button role=\"tab\""
+                + " aria-selected=\"true\" aria-controls=\"method-summary-table.tabpanel\" tabindex=\"0\""
                 + " onkeydown=\"switchTab(event)\" id=\"t0\" class=\"active-table-tab\">All Methods</button>"
-                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"member-summary_tabpanel\""
+                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"method-summary-table.tabpanel\""
                 + " tabindex=\"-1\" onkeydown=\"switchTab(event)\" id=\"t2\" class=\"table-tab\""
                 + " onclick=\"show(2);\">Instance Methods</button><button role=\"tab\" aria-selected=\"false\""
-                + " aria-controls=\"member-summary_tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
+                + " aria-controls=\"method-summary-table.tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
                 + " id=\"t3\" class=\"table-tab\" onclick=\"show(4);\">Abstract Methods</button>"
-                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"member-summary_tabpanel\""
+                + "<button role=\"tab\" aria-selected=\"false\" aria-controls=\"method-summary-table.tabpanel\""
                 + " tabindex=\"-1\" onkeydown=\"switchTab(event)\" id=\"t4\" class=\"table-tab\""
                 + " onclick=\"show(8);\">Concrete Methods</button><button role=\"tab\" aria-selected=\"false\""
-                + " aria-controls=\"member-summary_tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
+                + " aria-controls=\"method-summary-table.tabpanel\" tabindex=\"-1\" onkeydown=\"switchTab(event)\""
                 + " id=\"t6\" class=\"table-tab\" onclick=\"show(32);\">Deprecated Methods</button></div>",
                 "<tr class=\"alt-color\" id=\"i0\">");
 
         checkOutput("pkg1/A.html", false,
-                "<caption><span>Methods</span><span class=\"tab-end\">&nbsp;</span>"
-                + "</caption>");
+                "<caption><span>Methods</span></caption>");
 
         checkOutput("pkg1/B.html", false,
-                "<caption><span>Methods</span><span class=\"tab-end\">&nbsp;</span>"
-                + "</caption>");
+                "<caption><span>Methods</span></caption>");
 
         checkOutput("pkg1/D.html", false,
-                "<caption><span>Methods</span><span class=\"tab-end\">&nbsp;</span>"
-                + "</caption>");
+                "<caption><span>Methods</span></caption>");
     }
 }
