diff a/test/jdk/java/net/SocketPermission/SocketPermissionCollection.java b/test/jdk/java/net/SocketPermission/SocketPermissionCollection.java
--- a/test/jdk/java/net/SocketPermission/SocketPermissionCollection.java
+++ b/test/jdk/java/net/SocketPermission/SocketPermissionCollection.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -21,11 +21,11 @@
  * questions.
  */
 
 /*
  * @test
- * @bug 8056179
+ * @bug 8056179 8231785
  * @summary Unit test for PermissionCollection subclasses
  */
 
 import java.net.SocketPermission;
 import java.security.Permission;
@@ -124,12 +124,14 @@
         if (!perms.implies(new SocketPermission("204.160.241.0:1025", "connect"))) {
             System.err.println("Expected true, returned false");
             testFail++;
         }
 
-        // test 13
-        System.out.println("test 13: elements returns correct number of perms");
+
+        // test 10
+        System.out.println("test 10: elements returns correct number of perms");
+        perms.add(new SocketPermission("www.example.us", "resolve"));
         int numPerms = 0;
         Enumeration<Permission> e = perms.elements();
         while (e.hasMoreElements()) {
             numPerms++;
             System.out.println(e.nextElement());
