diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.idl b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.idl
@@ -23,15 +23,17 @@
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
  */
 
 interface SVGNumberList {
+    readonly attribute unsigned long length;
     readonly attribute unsigned long numberOfItems;
 
     [MayThrowException] void clear();
-    [MayThrowException] SVGNumber initialize(SVGNumber item);
-    [MayThrowException] SVGNumber getItem(unsigned long index);
-    [MayThrowException] SVGNumber insertItemBefore(SVGNumber item, unsigned long index);
-    [MayThrowException] SVGNumber replaceItem(SVGNumber item, unsigned long index);
+    [MayThrowException] SVGNumber initialize(SVGNumber newItem);
+    [MayThrowException] getter SVGNumber getItem(unsigned long index);
+    [MayThrowException] SVGNumber insertItemBefore(SVGNumber newItem, unsigned long index);
+    [MayThrowException] SVGNumber replaceItem(SVGNumber newItem, unsigned long index);
     [MayThrowException] SVGNumber removeItem(unsigned long index);
-    [MayThrowException] SVGNumber appendItem(SVGNumber item);
+    [MayThrowException] SVGNumber appendItem(SVGNumber newItem);
+    [MayThrowException] setter void (unsigned long index, SVGNumber newItem);
 };
