diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/API/JSHeapFinalizerPrivate.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/API/JSHeapFinalizerPrivate.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/API/JSHeapFinalizerPrivate.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/API/JSHeapFinalizerPrivate.cpp
@@ -27,20 +27,18 @@
 #include "JSHeapFinalizerPrivate.h"
 
 #include "APICast.h"
 #include "JSCInlines.h"
 
-using namespace JSC;
-
 void JSContextGroupAddHeapFinalizer(JSContextGroupRef group, JSHeapFinalizer finalizer, void *userData)
 {
-    VM* vm = toJS(group);
-    JSLockHolder locker(vm);
-    vm->heap.addHeapFinalizerCallback(HeapFinalizerCallback(finalizer, userData));
+    JSC::VM* vm = toJS(group);
+    JSC::JSLockHolder locker(vm);
+    vm->heap.addHeapFinalizerCallback(JSC::HeapFinalizerCallback(finalizer, userData));
 }
 
 void JSContextGroupRemoveHeapFinalizer(JSContextGroupRef group, JSHeapFinalizer finalizer, void *userData)
 {
-    VM* vm = toJS(group);
-    JSLockHolder locker(vm);
-    vm->heap.removeHeapFinalizerCallback(HeapFinalizerCallback(finalizer, userData));
+    JSC::VM* vm = toJS(group);
+    JSC::JSLockHolder locker(vm);
+    vm->heap.removeHeapFinalizerCallback(JSC::HeapFinalizerCallback(finalizer, userData));
 }
