diff a/application/org.openjdk.jmc.joverflow.ui/META-INF/MANIFEST.MF b/application/org.openjdk.jmc.joverflow.ui/META-INF/MANIFEST.MF
--- a/application/org.openjdk.jmc.joverflow.ui/META-INF/MANIFEST.MF
+++ b/application/org.openjdk.jmc.joverflow.ui/META-INF/MANIFEST.MF
@@ -11,7 +11,5 @@
 Eclipse-BuddyPolicy: app
 Bundle-ActivationPolicy: lazy
 Bundle-Activator: org.openjdk.jmc.joverflow.ui.JOverflowPlugin
 Automatic-Module-Name: org.openjdk.jmc.joverflow.ui
 Import-Package: org.openjdk.jmc.flightrecorder.ui
-Export-Package: org.openjdk.jmc.joverflow.ui,
- org.openjdk.jmc.joverflow.ui.model
diff a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowEditor.java b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowEditor.java
--- a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowEditor.java
+++ b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowEditor.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
  *
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * The contents of this file are subject to the terms of either the Universal Permissive License
  * v 1.0 as shown at http://oss.oracle.com/licenses/upl
@@ -265,28 +265,28 @@
 	@Override
 	public boolean isSaveAsAllowed() {
 		return false;
 	}
 
-	public JOverflowUi getJOverflowUi() {
+	JOverflowUi getJOverflowUi() {
 		return mJOverflowUi;
 	}
 
-	public Snapshot getSnapshot() {
+	Snapshot getSnapshot() {
 		return mSnapshot;
 	}
 
-	public void addUiLoadedListener(UiLoadedListener listener) {
+	void addUiLoadedListener(UiLoadedListener listener) {
 		mUiLoadedListeners.add(listener);
 		if (mJOverflowUi != null) {
 			listener.uiLoaded(mJOverflowUi);
 		}
 	}
 
-	public void removeUiLoadedListener(UiLoadedListener listener) {
+	void removeUiLoadedListener(UiLoadedListener listener) {
 		mUiLoadedListeners.remove(listener);
 	}
 
-	public interface UiLoadedListener {
+	interface UiLoadedListener {
 		void uiLoaded(JOverflowUi ui);
 	}
 }
diff a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowPlugin.java b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowPlugin.java
--- a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowPlugin.java
+++ b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowPlugin.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
  *
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * The contents of this file are subject to the terms of either the Universal Permissive License
  * v 1.0 as shown at http://oss.oracle.com/licenses/upl
diff a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowUi.java b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowUi.java
--- a/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowUi.java
+++ b/application/org.openjdk.jmc.joverflow.ui/src/main/java/org/openjdk/jmc/joverflow/ui/JOverflowUi.java
@@ -1,8 +1,8 @@
 /*
- * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2019, 2020, Red Hat Inc. All rights reserved.
+ * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, Red Hat Inc. All rights reserved.
  *
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * The contents of this file are subject to the terms of either the Universal Permissive License
  * v 1.0 as shown at http://oss.oracle.com/licenses/upl
@@ -181,17 +181,17 @@
 		mAncestorViewer.reset();
 
 		updateModel();
 	}
 
-	public void addModelListener(final ModelListener listener) {
+	void addModelListener(final ModelListener listener) {
 		mModelListeners.add(listener);
 
 		if (mModel != null) {
 			updateModel();
 		}
 	}
 
-	public void removeModelListener(final ModelListener listener) {
+	void removeModelListener(final ModelListener listener) {
 		mModelListeners.remove(listener);
 	}
 }
diff a/application/pom.xml b/application/pom.xml
--- a/application/pom.xml
+++ b/application/pom.xml
@@ -75,11 +75,10 @@
 		<module>org.openjdk.jmc.feature.flightrecorder.metadata</module>
 		<module>org.openjdk.jmc.feature.ide</module>
 		<module>org.openjdk.jmc.feature.ide.launch</module>
 		<module>org.openjdk.jmc.feature.jconsole</module>
 		<module>org.openjdk.jmc.feature.joverflow</module>
-		<module>org.openjdk.jmc.feature.joverflow.ext.treemap</module>
 		<module>org.openjdk.jmc.feature.license</module>
 		<module>org.openjdk.jmc.feature.pde</module>
 		<module>org.openjdk.jmc.feature.rcp</module>
 		<module>org.openjdk.jmc.feature.rcp.update</module>
 		<module>org.openjdk.jmc.feature.twitter</module>
@@ -98,11 +97,10 @@
 		<module>org.openjdk.jmc.ide.jdt</module>
 		<module>org.openjdk.jmc.ide.launch</module>
 		<module>org.openjdk.jmc.ide.ui</module>
 		<module>org.openjdk.jmc.joverflow</module>
 		<module>org.openjdk.jmc.joverflow.ui</module>
-		<module>org.openjdk.jmc.joverflow.ext.treemap</module>
 		<module>org.openjdk.jmc.osgi.extension</module>
 		<module>org.openjdk.jmc.pde</module>
 		<module>org.openjdk.jmc.rcp.application</module>
 		<module>org.openjdk.jmc.rcp.intro</module>
 		<module>org.openjdk.jmc.rcp.product</module>
