diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/text/LocaleNone.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/text/LocaleNone.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/text/LocaleNone.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/text/LocaleNone.cpp
@@ -53,13 +53,13 @@
     Vector<String> m_shortMonthLabels;
     Vector<String> m_monthLabels;
 #endif
 };
 
-std::unique_ptr<Locale> Locale::create(const AtomicString&)
+std::unique_ptr<Locale> Locale::create(const AtomString&)
 {
-    return std::make_unique<LocaleNone>();
+    return makeUnique<LocaleNone>();
 }
 
 LocaleNone::~LocaleNone() = default;
 
 void LocaleNone::initializeLocaleData()
