diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FloatSize.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FloatSize.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FloatSize.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/FloatSize.cpp
@@ -48,15 +48,10 @@
         std::max(min.width(), std::min(max.width(), m_width)),
         std::max(min.height(), std::min(max.height(), m_height))
     };
 }
 
-float FloatSize::diagonalLength() const
-{
-    return sqrtf(diagonalLengthSquared());
-}
-
 bool FloatSize::isZero() const
 {
     return fabs(m_width) < std::numeric_limits<float>::epsilon() && fabs(m_height) < std::numeric_limits<float>::epsilon();
 }
 
