diff a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/AvailableMemory.cpp b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/AvailableMemory.cpp
--- a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/AvailableMemory.cpp
+++ b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/AvailableMemory.cpp
@@ -51,11 +51,11 @@
 #include <unistd.h>
 #endif
 
 namespace bmalloc {
 
-static const size_t availableMemoryGuess = 512 * bmalloc::MB;
+static constexpr size_t availableMemoryGuess = 512 * bmalloc::MB;
 
 #if BOS(DARWIN)
 static size_t memorySizeAccordingToKernel()
 {
 #if BPLATFORM(IOS_FAMILY_SIMULATOR)
