diff a/modules/javafx.graphics/src/main/java/javafx/animation/Transition.java b/modules/javafx.graphics/src/main/java/javafx/animation/Transition.java
--- a/modules/javafx.graphics/src/main/java/javafx/animation/Transition.java
+++ b/modules/javafx.graphics/src/main/java/javafx/animation/Transition.java
@@ -38,11 +38,11 @@
  * This class offers a simple framework to define animation. It provides all the
  * basic functionality defined in {@link Animation}. {@code Transition} requires
  * the implementation of a method {@link #interpolate(double)} which is the
  * called in each frame, while the {@code Transition} is running.
  * <p>
- * In addition an extending class needs to set the duration of a single cycle
+ * In addition, an extending class needs to set the duration of a single cycle
  * with {@link Animation#setCycleDuration(javafx.util.Duration)}. This duration
  * is usually set by the user via a duration property (as in
  * {@link FadeTransition#durationProperty() duration}) for example. But it can also be calculated
  * by the extending class as is done in {@link ParallelTransition} and
  * {@link FadeTransition}.
