diff a/src/java.net.http/share/classes/jdk/internal/net/http/Stream.java b/src/java.net.http/share/classes/jdk/internal/net/http/Stream.java
--- a/src/java.net.http/share/classes/jdk/internal/net/http/Stream.java
+++ b/src/java.net.http/share/classes/jdk/internal/net/http/Stream.java
@@ -387,14 +387,14 @@
         if ((frame instanceof HeaderFrame)) {
             HeaderFrame hframe = (HeaderFrame)frame;
             if (hframe.endHeaders()) {
                 Log.logTrace("handling response (streamid={0})", streamid);
                 handleResponse();
-                if (hframe.getFlag(HeaderFrame.END_STREAM)) {
-                    if (debug.on()) debug.log("handling END_STREAM: %d", streamid);
-                    receiveDataFrame(new DataFrame(streamid, DataFrame.END_STREAM, List.of()));
-                }
+            }
+            if (hframe.getFlag(HeaderFrame.END_STREAM)) {
+                if (debug.on()) debug.log("handling END_STREAM: %d", streamid);
+                receiveDataFrame(new DataFrame(streamid, DataFrame.END_STREAM, List.of()));
             }
         } else if (frame instanceof DataFrame) {
             receiveDataFrame((DataFrame)frame);
         } else {
             otherFrame(frame);
