diff a/src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java b/src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java
--- a/src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java
+++ b/src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java
@@ -423,23 +423,11 @@
             super(evt, evt.requestID);
             this.classSignature = evt.signature;
         }
 
         public String className() {
-            assert classSignature.startsWith("L") && classSignature.endsWith(";");
-
-            // trim leading "L" and trailing ";"
-            String name = classSignature.substring(1, classSignature.length() - 1);
-            int index = name.indexOf(".");  // check if it is a hidden class
-            if (index < 0) {
-                return name.replace('/', '.');
-            }  else {
-                // map the type descriptor from: "L" + N + "." + <suffix> + ";"
-                // to class name: N.replace('/', '.') + "/" + <suffix>
-                return name.substring(0, index).replace('/', '.')
-                        + "/" + name.substring(index + 1);
-            }
+            return JNITypeParser.convertSignatureToClassname(classSignature);
         }
 
         public String classSignature() {
             return classSignature;
         }
