diff a/modules/javafx.graphics/src/main/java/com/sun/prism/impl/PrismSettings.java b/modules/javafx.graphics/src/main/java/com/sun/prism/impl/PrismSettings.java
--- a/modules/javafx.graphics/src/main/java/com/sun/prism/impl/PrismSettings.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/prism/impl/PrismSettings.java
@@ -25,12 +25,10 @@
 
 package com.sun.prism.impl;
 
 import java.security.AccessController;
 import java.security.PrivilegedAction;
-import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
 import java.util.Properties;
 import java.util.StringTokenizer;
 import com.sun.javafx.PlatformUtil;
 import com.sun.javafx.util.Utils;
@@ -224,11 +222,11 @@
             } else {
                 tryOrderArr = new String[] { "sw" };
             }
         }
 
-        tryOrder = Collections.unmodifiableList(Arrays.asList(tryOrderArr));
+        tryOrder = List.of(tryOrderArr);
 
         RasterizerType rSpec = null;
         String rOrder = systemProperties.getProperty("prism.rasterizerorder");
         if (rOrder != null) {
             for (String s : split(rOrder.toLowerCase(), ",")) {
