diff a/src/hotspot/share/utilities/macros.hpp b/src/hotspot/share/utilities/macros.hpp
--- a/src/hotspot/share/utilities/macros.hpp
+++ b/src/hotspot/share/utilities/macros.hpp
@@ -259,13 +259,15 @@
 #define INCLUDE_JFR 1
 #endif
 
 #if INCLUDE_JFR
 #define JFR_ONLY(code) code
+#define NOT_JFR_RETURN()      /* next token must be ; */
 #define NOT_JFR_RETURN_(code) /* next token must be ; */
 #else
 #define JFR_ONLY(code)
+#define NOT_JFR_RETURN()      {}
 #define NOT_JFR_RETURN_(code) { return code; }
 #endif
 
 #ifndef INCLUDE_JVMCI
 #define INCLUDE_JVMCI 1
