diff a/application/org.openjdk.jmc.flightrecorder.controlpanel.ui.configuration/META-INF/MANIFEST.MF b/application/org.openjdk.jmc.flightrecorder.controlpanel.ui.configuration/META-INF/MANIFEST.MF
--- a/application/org.openjdk.jmc.flightrecorder.controlpanel.ui.configuration/META-INF/MANIFEST.MF
+++ b/application/org.openjdk.jmc.flightrecorder.controlpanel.ui.configuration/META-INF/MANIFEST.MF
@@ -1,10 +1,10 @@
 Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-Name: Flight Recorder Configuration UI
 Bundle-SymbolicName: org.openjdk.jmc.flightrecorder.controlpanel.ui.configuration;singleton:=true
-Bundle-Version: 7.1.0.qualifier
+Bundle-Version: 8.0.0.qualifier
 Require-Bundle: org.openjdk.jmc.ui,
  org.openjdk.jmc.flightrecorder.configuration,
  org.openjdk.jmc.ui.common
 Bundle-ActivationPolicy: lazy
 Bundle-RequiredExecutionEnvironment: JavaSE-1.8
