diff a/src/jdk.internal.le/share/classes/jdk/internal/org/jline/utils/windows-vtp.caps b/src/jdk.internal.le/share/classes/jdk/internal/org/jline/utils/windows-vtp.caps
--- a/src/jdk.internal.le/share/classes/jdk/internal/org/jline/utils/windows-vtp.caps
+++ b/src/jdk.internal.le/share/classes/jdk/internal/org/jline/utils/windows-vtp.caps
@@ -6,11 +6,11 @@
 	cuf=\E[%p1%dC, cuf1=\E[C, cup=\E[%i%p1%d;%p2%dH,
 	cuu=\E[%p1%dA, cuu1=\E[A,
 	il=\E[%p1%dL, il1=\E[L,
 	dl=\E[%p1%dM, dl1=\E[M,
 	ech=\E[%p1%dX,
-	el=\E[K, ed=\E[2K,
+	el=\E[K, ed=\E[J,
 	el1=\E[1K, home=\E[H, hpa=\E[%i%p1%dG,
 	ind=^J,
 	invis=\E[8m, kbs=^H, kcbt=\E[Z,
 	kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA,
 	khome=\E[H,
