diff a/modules/javafx.base/src/main/java/module-info.java b/modules/javafx.base/src/main/java/module-info.java
--- a/modules/javafx.base/src/main/java/module-info.java
+++ b/modules/javafx.base/src/main/java/module-info.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2015, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -47,10 +47,11 @@
 
     exports com.sun.javafx to
         javafx.controls,
         javafx.graphics,
         javafx.fxml,
+        javafx.media,
         javafx.swing;
     exports com.sun.javafx.beans to
         javafx.controls,
         javafx.fxml,
         javafx.graphics;
