diff a/src/hotspot/cpu/aarch64/globalDefinitions_aarch64.hpp b/src/hotspot/cpu/aarch64/globalDefinitions_aarch64.hpp
--- a/src/hotspot/cpu/aarch64/globalDefinitions_aarch64.hpp
+++ b/src/hotspot/cpu/aarch64/globalDefinitions_aarch64.hpp
@@ -54,10 +54,8 @@
 // evidence that it's worth doing.
 #define DEOPTIMIZE_WHEN_PATCHING
 
 #define SUPPORT_RESERVED_STACK_AREA
 
-#define PREFERRED_METASPACE_ALIGNMENT
-
 #define COMPRESSED_CLASS_POINTERS_DEPENDS_ON_COMPRESSED_OOPS true
 
 #endif // CPU_AARCH64_GLOBALDEFINITIONS_AARCH64_HPP
