diff a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
--- a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java
@@ -64,11 +64,11 @@
         return json.get("iid").toString();
     }
 
     @Override
     public HostUser author() {
-        return repository.forge().user(json.get("author").get("username").asString());
+        return repository.forge().user(json.get("author").get("username").asString()).get();
     }
 
     @Override
     public List<Review> reviews() {
 
@@ -120,11 +120,11 @@
                               if (createdAt.isAfter(cd.date)) {
                                   hash = cd.hash;
                               }
                           }
                           var id = obj.get("id").asInt();
-                          return new Review(createdAt, reviewer, verdict, hash, id, null);
+                          return new Review(createdAt, reviewer.get(), verdict, hash, id, null);
                       })
                       .collect(Collectors.toList());
     }
 
     @Override
@@ -622,11 +622,11 @@
     @Override
     public List<HostUser> assignees() {
         var assignee = json.get("assignee").asObject();
         if (assignee != null) {
             var user = repository.forge().user(assignee.get("username").asString());
-            return List.of(user);
+            return List.of(user.get());
         }
         return Collections.emptyList();
     }
 
     @Override
