diff a/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLResourcePreloader.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLResourcePreloader.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLResourcePreloader.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLResourcePreloader.cpp
@@ -59,11 +59,11 @@
         if (crossOriginMode.isNull())
             crossOriginMode = "omit"_s;
     }
     if (m_resourceType == CachedResource::Type::Script)
         options.referrerPolicy = m_referrerPolicy;
-    auto request = createPotentialAccessControlRequest(completeURL(document), document, crossOriginMode, WTFMove(options));
+    auto request = createPotentialAccessControlRequest(completeURL(document), WTFMove(options), document, crossOriginMode);
     request.setInitiator(m_initiator);
     return request;
 }
 
 void HTMLResourcePreloader::preload(PreloadRequestStream requests)
