diff a/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTokenizer.cpp b/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTokenizer.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTokenizer.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/html/parser/HTMLTokenizer.cpp
@@ -1416,11 +1416,11 @@
     else if (tagName == styleTag
         || tagName == iframeTag
         || tagName == xmpTag
         || (tagName == noembedTag)
         || tagName == noframesTag
-        || (tagName == noscriptTag && m_options.scriptEnabled))
+        || (tagName == noscriptTag && m_options.scriptingFlag))
         m_state = RAWTEXTState;
 }
 
 inline void HTMLTokenizer::appendToTemporaryBuffer(UChar character)
 {
