diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBVersionChangeEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBVersionChangeEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBVersionChangeEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/IDBVersionChangeEvent.h
@@ -32,10 +32,11 @@
 #include <wtf/Optional.h>
 
 namespace WebCore {
 
 class IDBVersionChangeEvent final : public Event {
+    WTF_MAKE_ISO_ALLOCATED(IDBVersionChangeEvent);
 public:
     static Ref<IDBVersionChangeEvent> create(uint64_t oldVersion, uint64_t newVersion, const AtomString& eventType)
     {
         return adoptRef(*new IDBVersionChangeEvent(IDBResourceIdentifier::emptyValue(), oldVersion, newVersion, eventType));
     }
