diff a/application/org.openjdk.jmc.jdp/src/main/java/org/openjdk/jmc/jdp/common/ConfigurationFactory.java b/application/org.openjdk.jmc.jdp/src/main/java/org/openjdk/jmc/jdp/common/ConfigurationFactory.java
--- a/application/org.openjdk.jmc.jdp/src/main/java/org/openjdk/jmc/jdp/common/ConfigurationFactory.java
+++ b/application/org.openjdk.jmc.jdp/src/main/java/org/openjdk/jmc/jdp/common/ConfigurationFactory.java
@@ -47,11 +47,12 @@
 		InetAddress tmp = null;
 		try {
 			tmp = InetAddress.getByName(Configuration.DEFAULT_MULTICAST_ADDRESS);
 		} catch (UnknownHostException e) {
 			// Multicast address by IP, should never happen!
-			Logger.getLogger("org.openjdk.jmc.jdp.common").log(Level.SEVERE, "Could not create default mulitcast address!", //$NON-NLS-1$ //$NON-NLS-2$
+			Logger.getLogger("org.openjdk.jmc.jdp.common").log(Level.SEVERE, //$NON-NLS-1$
+					"Could not create default mulitcast address!", //$NON-NLS-1$
 					e);
 		}
 		DEFAULT_MULTICAST_ADDRESS = tmp;
 	}
 
