diff a/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/jdk/JdkAggregators.java b/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/jdk/JdkAggregators.java
--- a/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/jdk/JdkAggregators.java
+++ b/core/org.openjdk.jmc.flightrecorder/src/main/java/org/openjdk/jmc/flightrecorder/jdk/JdkAggregators.java
@@ -387,11 +387,12 @@
 	public static final IAggregator<IQuantity, ?> VM_OPERATION_COUNT = Aggregators.count(
 			Messages.getString(Messages.AGGR_VM_OPERATION_COUNT),
 			Messages.getString(Messages.AGGR_VM_OPERATION_COUNT_DESC), JdkFilters.VM_OPERATIONS);
 	public static final IAggregator<IQuantity, ?> VM_OPERATION_DURATION = Aggregators.sum(
 			Messages.getString(Messages.AGGR_VM_OPERATION_DURATION),
-			Messages.getString(Messages.AGGR_VM_OPERATION_DURATION_DESC), JfrAttributes.DURATION);
+			Messages.getString(Messages.AGGR_VM_OPERATION_DURATION_DESC), JdkTypeIDs.VM_OPERATIONS,
+			JfrAttributes.DURATION);
 
 	public static final IAggregator<IQuantity, ?> COMPILATIONS_COUNT = Aggregators.count(
 			Messages.getString(Messages.AGGR_COMPILATIONS_COUNT),
 			Messages.getString(Messages.AGGR_COMPILATIONS_COUNT_DESC), JdkFilters.COMPILATION);
 	public static final IAggregator<IQuantity, ?> LONGEST_COMPILATION = Aggregators.filter(Aggregators.max(DURATION),
