diff a/src/jdk.jfr/share/classes/jdk/jfr/internal/consumer/Parser.java b/src/jdk.jfr/share/classes/jdk/jfr/internal/consumer/Parser.java
--- a/src/jdk.jfr/share/classes/jdk/jfr/internal/consumer/Parser.java
+++ b/src/jdk.jfr/share/classes/jdk/jfr/internal/consumer/Parser.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2016, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2016, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -40,11 +40,11 @@
      *         problems
      */
     public abstract Object parse(RecordingInput input) throws IOException;
 
     /**
-     * Skips data that would usually be by parsed the {@code #parse(RecordingInput)} method.
+     * Skips data that would usually be by parsed the {@link #parse(RecordingInput)} method.
      *
      * @param input input to read from
      * @throws IOException if operation couldn't be completed due to I/O
      *         problems
      */
