diff a/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java b/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java
--- a/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java
+++ b/src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/AbstractLayout.java
@@ -166,10 +166,15 @@
 
     boolean hasNaturalAlignment() {
         return size.isPresent() && size.getAsLong() == alignment;
     }
 
+    @Override
+    public boolean isPadding() {
+        return this instanceof PaddingLayout;
+    }
+
     @Override
     public int hashCode() {
         return attributes.hashCode() << Long.hashCode(alignment);
     }
 
