diff a/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack016.java b/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack016.java
--- a/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack016.java
+++ b/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack016.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -55,11 +55,11 @@
  */
 
 package nsk.stress.stack;
 
 
-import nsk.share.Harakiri;
+import nsk.share.Terminator;
 
 import java.io.PrintStream;
 
 public class stack016 extends Thread {
     private final static int THREADS = 10;
@@ -80,11 +80,11 @@
             if (args[i].toLowerCase().equals("-verbose"))
                 verbose = true;
             else if (args[i].toLowerCase().equals("-eager"))
                 eager = true;
         if (!eager)
-            Harakiri.appoint(Harakiri.parseAppointment(args));
+            Terminator.appoint(Terminator.parseAppointment(args));
         stack016.out = out;
         stack016 test = new stack016();
         return test.doRun();
     }
 
