diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/gobject/gsourceclosure.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/gobject/gsourceclosure.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/gobject/gsourceclosure.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/gobject/gsourceclosure.h
@@ -26,11 +26,11 @@
 
 G_BEGIN_DECLS
 
 GLIB_AVAILABLE_IN_ALL
 void g_source_set_closure        (GSource  *source,
-                  GClosure *closure);
+          GClosure *closure);
 
 GLIB_AVAILABLE_IN_ALL
 void g_source_set_dummy_callback (GSource  *source);
 
 G_END_DECLS
