diff a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Printer.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Printer.java
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Printer.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Printer.java
@@ -324,11 +324,11 @@
              s =  sym.getQualifiedName().toString();
         } else {
             s =  sym.name.toString();
         }
 
-        return sym.isProjectedNullable() ? s + '?' : s;
+        return s;
     }
 
     /**
      * Converts a set of method argument types into their corresponding
      * localized string representation.
