diff a/test/hotspot/jtreg/gc/shenandoah/options/TestHeuristicsUnlock.java b/test/hotspot/jtreg/gc/shenandoah/options/TestHeuristicsUnlock.java
--- a/test/hotspot/jtreg/gc/shenandoah/options/TestHeuristicsUnlock.java
+++ b/test/hotspot/jtreg/gc/shenandoah/options/TestHeuristicsUnlock.java
@@ -43,18 +43,14 @@
         DIAGNOSTIC,
         EXPERIMENTAL,
     }
 
     public static void main(String[] args) throws Exception {
-        testWith("-XX:ShenandoahGCHeuristics=adaptive", Mode.PRODUCT);
-        testWith("-XX:ShenandoahGCHeuristics=static", Mode.PRODUCT);
-        testWith("-XX:ShenandoahGCHeuristics=compact", Mode.PRODUCT);
-
-        testWith("-XX:ShenandoahGCMode=iu", Mode.EXPERIMENTAL);
-
+        testWith("-XX:ShenandoahGCHeuristics=adaptive",   Mode.PRODUCT);
+        testWith("-XX:ShenandoahGCHeuristics=static",     Mode.PRODUCT);
+        testWith("-XX:ShenandoahGCHeuristics=compact",    Mode.PRODUCT);
         testWith("-XX:ShenandoahGCHeuristics=aggressive", Mode.DIAGNOSTIC);
-        testWith("-XX:ShenandoahGCHeuristics=passive", Mode.DIAGNOSTIC);
     }
 
     private static void testWith(String h, Mode mode) throws Exception {
         if (false) { // When ShenandoahGC is experimental flag, this makes no sense to test
             ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(
