diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Kind.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Kind.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Kind.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/B3Kind.h
@@ -1,7 +1,7 @@
 /*
- * Copyright (C) 2016 Apple Inc. All rights reserved.
+ * Copyright (C) 2016-2019 Apple Inc. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
  * are met:
  * 1. Redistributions of source code must retain the above copyright
@@ -209,11 +209,11 @@
 }
 
 struct KindHash {
     static unsigned hash(const Kind& key) { return key.hash(); }
     static bool equal(const Kind& a, const Kind& b) { return a == b; }
-    static const bool safeToCompareToEmptyOrDeleted = true;
+    static constexpr bool safeToCompareToEmptyOrDeleted = true;
 };
 
 } } // namespace JSC::B3
 
 namespace WTF {
@@ -223,11 +223,11 @@
     typedef JSC::B3::KindHash Hash;
 };
 
 template<typename T> struct HashTraits;
 template<> struct HashTraits<JSC::B3::Kind> : public SimpleClassHashTraits<JSC::B3::Kind> {
-    static const bool emptyValueIsZero = false;
+    static constexpr bool emptyValueIsZero = false;
 };
 
 } // namespace WTF
 
 #endif // ENABLE(B3_JIT)
