diff a/agent/src/main/java/org/openjdk/jmc/agent/Transformer.java b/agent/src/main/java/org/openjdk/jmc/agent/Transformer.java
--- a/agent/src/main/java/org/openjdk/jmc/agent/Transformer.java
+++ b/agent/src/main/java/org/openjdk/jmc/agent/Transformer.java
@@ -106,11 +106,11 @@
 					new JFRNextClassVisitor(classWriter, td, definingClassLoader, classBeingRedefined, protectionDomain,
 							inspectionClassLoader) :
 					new JFRClassVisitor(classWriter, td, definingClassLoader, classBeingRedefined, protectionDomain,
 							inspectionClassLoader); 
 			ClassReader reader = new ClassReader(classfileBuffer);
-			reader.accept(visitor, 0);
+			reader.accept(visitor, ClassReader.EXPAND_FRAMES);
 			return classWriter.toByteArray();
 		} catch (Throwable t) {
 			Logger.getLogger(getClass().getName()).log(Level.SEVERE,
 					"Failed to instrument " + td.getMethod().toString(), t); //$NON-NLS-1$
 			return classfileBuffer;
