diff a/test/langtools/tools/doclint/html/EntitiesTest.java b/test/langtools/tools/doclint/html/EntitiesTest.java
--- a/test/langtools/tools/doclint/html/EntitiesTest.java
+++ b/test/langtools/tools/doclint/html/EntitiesTest.java
@@ -292,27 +292,39 @@
     * &Dagger; &#8225;
     * &permil; &#8240;
     * &lsaquo; &#8249;
     * &rsaquo; &#8250;
     * &euro; &#8364;
+    * &sol; &#47;
+    * &bigotimes; &#10754;
+    * &mho; &#2127;
+    * &horbar; &#8213;
+    * &cylcty; &#9005;
     */
     void symbolic_entities() { }
 
+    /**
+     * &#9003;
+     * &#32;
+     */
+    void other_numeric() {}
+
     /**
      * &bad;
      */
     void bad_name() { }
 
     /**
+     * &#0;
+     * &#13;
      * &#31;
-     * &#2127;
+     * &#xd800;
+     * &#xdfff;
      */
     void out_of_range() { }
 
     /**
-     * &#8213;
-     * &#9003;
      * &#xffff;
      */
     void sparse_negative() { }
 }
 
