diff a/src/hotspot/share/classfile/classFileParser.cpp b/src/hotspot/share/classfile/classFileParser.cpp
--- a/src/hotspot/share/classfile/classFileParser.cpp
+++ b/src/hotspot/share/classfile/classFileParser.cpp
@@ -2477,13 +2477,13 @@
       SignatureStream ss((Symbol*) signature, true);
       while (!ss.at_return_type())  ss.next();
       if (ss.is_reference()) {
         Symbol* ret = ss.as_symbol();
         const Symbol* required = class_name();
-        if (is_unsafe_anonymous()) {
-          // The original class name in the UAC byte stream gets changed.  So
-          // using the original name in the return type is no longer valid.
+        if (is_hidden() || is_unsafe_anonymous()) {
+          // The original class name in hidden classes and in the UAC byte stream gets
+          // changed.  So using the original name in the return type is no longer valid.
           required = vmSymbols::java_lang_Object();
         }
         ok = (ret == required);
       }
       if (!ok) {
