diff a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
--- a/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
+++ b/test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java
@@ -183,9 +183,22 @@
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
     }
 
+    @Override
+    public List<HostedBranch> branches() {
+        try {
+            var result = new ArrayList<HostedBranch>();
+            for (var b : localRepository.branches()) {
+                result.add(new TestHostedBranch(b.name(), localRepository.resolve(b).orElseThrow()));
+            }
+            return result;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
     Repository localRepository() {
         return localRepository;
     }
 }
