diff a/modules/javafx.web/src/main/native/Source/WebCore/xml/XPathValue.cpp b/modules/javafx.web/src/main/native/Source/WebCore/xml/XPathValue.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/xml/XPathValue.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/xml/XPathValue.cpp
@@ -124,11 +124,11 @@
                 return "NaN"_s;
             if (m_number == 0)
                 return "0"_s;
             if (std::isinf(m_number))
                 return std::signbit(m_number) ? "-Infinity"_s : "Infinity"_s;
-            return String::number(m_number);
+            return String::numberToStringFixedPrecision(m_number);
         case BooleanValue:
             return m_bool ? "true"_s : "false"_s;
     }
 
     ASSERT_NOT_REACHED();
