diff a/modules/javafx.web/src/main/native/Source/WebCore/xml/XSLTProcessor.cpp b/modules/javafx.web/src/main/native/Source/WebCore/xml/XSLTProcessor.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/xml/XSLTProcessor.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/xml/XSLTProcessor.cpp
@@ -70,14 +70,14 @@
     bool sourceIsDocument = (sourceNode == &ownerDocument.get());
     String documentSource = sourceString;
 
     RefPtr<Document> result;
     if (sourceMIMEType == "text/plain") {
-        result = XMLDocument::createXHTML(frame, sourceIsDocument ? ownerDocument->url() : URL());
+        result = XMLDocument::createXHTML(ownerDocument->sessionID(), frame, sourceIsDocument ? ownerDocument->url() : URL());
         transformTextStringToXHTMLDocumentString(documentSource);
     } else
-        result = DOMImplementation::createDocument(sourceMIMEType, frame, sourceIsDocument ? ownerDocument->url() : URL());
+        result = DOMImplementation::createDocument(ownerDocument->sessionID(), sourceMIMEType, frame, sourceIsDocument ? ownerDocument->url() : URL());
 
     // Before parsing, we need to save & detach the old document and get the new document
     // in place. We have to do this only if we're rendering the result document.
     if (frame) {
         if (FrameView* view = frame->view())
