diff a/test/langtools/tools/javac/diags/examples/MalformedSupported/MalformedSupported.java b/test/langtools/tools/javac/diags/examples/MalformedSupported/MalformedSupported.java
--- a/test/langtools/tools/javac/diags/examples/MalformedSupported/MalformedSupported.java
+++ b/test/langtools/tools/javac/diags/examples/MalformedSupported/MalformedSupported.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2010, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2010, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -20,8 +20,8 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
 
 // key: compiler.warn.proc.malformed.supported.string
-// options: -processor AnnoProc
+// options: -processor AnnoProc -Xlint:processing
 
 class MalformedSupported { }
