diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGAnimateMotionElement.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGAnimateMotionElement.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGAnimateMotionElement.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGAnimateMotionElement.h
@@ -39,16 +39,16 @@
     bool hasValidAttributeType() const override;
     bool hasValidAttributeName() const override;
 
     void parseAttribute(const QualifiedName&, const AtomString&) override;
 
-    void resetAnimatedType() override;
-    void clearAnimatedType(SVGElement* targetElement) override;
+    void startAnimation() override;
+    void stopAnimation(SVGElement* targetElement) override;
     bool calculateToAtEndOfDurationValue(const String& toAtEndOfDurationString) override;
     bool calculateFromAndToValues(const String& fromString, const String& toString) override;
     bool calculateFromAndByValues(const String& fromString, const String& byString) override;
-    void calculateAnimatedValue(float percentage, unsigned repeatCount, SVGSMILElement* resultElement) override;
+    void calculateAnimatedValue(float percentage, unsigned repeatCount) override;
     void applyResultsToTarget() override;
     Optional<float> calculateDistance(const String& fromString, const String& toString) override;
 
     enum RotateMode {
         RotateAngle,
