diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocumentExtensions.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocumentExtensions.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocumentExtensions.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGDocumentExtensions.h
@@ -51,11 +51,11 @@
     RenderSVGResourceContainer* resourceById(const AtomString& id) const;
 
     void startAnimations();
     void pauseAnimations();
     void unpauseAnimations();
-    void dispatchSVGLoadEventToOutermostSVGElements();
+    void dispatchLoadEventToOutermostSVGElements();
     bool areAnimationsPaused() const { return m_areAnimationsPaused; }
 
     void reportWarning(const String&);
     void reportError(const String&);
 
