diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGMarkerTypes.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGMarkerTypes.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGMarkerTypes.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGMarkerTypes.h
@@ -38,14 +38,14 @@
 
 enum SVGMarkerOrientType {
     SVGMarkerOrientUnknown = 0,
     SVGMarkerOrientAuto,
     SVGMarkerOrientAngle,
-    SVGMarkerOrientAutoStartReverse,
 
-    // Add new elements before here.
-    SVGMarkerOrientMax
+    // The DOM can't set the property 'orientType' to this value. It is used only
+    // internally when setting the 'orient' attribute to "auto-start-reverse".
+    SVGMarkerOrientAutoStartReverse = SVGMarkerOrientUnknown
 };
 
 template<>
 struct SVGPropertyTraits<SVGMarkerUnitsType> {
     static unsigned highestEnumValue() { return SVGMarkerUnitsStrokeWidth; }
@@ -83,11 +83,11 @@
     static const String autoStartReverseString()
     {
         static const NeverDestroyed<String> autoStartReverseString = MAKE_STATIC_STRING_IMPL("auto-start-reverse");
         return autoStartReverseString;
     }
-    static unsigned highestEnumValue() { return SVGMarkerOrientAutoStartReverse; }
+    static unsigned highestEnumValue() { return SVGMarkerOrientAngle; }
     static SVGMarkerOrientType fromString(const String& string)
     {
         if (string == autoString())
             return SVGMarkerOrientAuto;
         if (string == autoStartReverseString())
@@ -102,13 +102,10 @@
             return autoStartReverseString();
         return emptyString();
     }
 };
 
-template<>
-inline unsigned SVGIDLEnumLimits<SVGMarkerOrientType>::highestExposedEnumValue() { return SVGMarkerOrientAngle; }
-
 template<>
 struct SVGPropertyTraits<std::pair<SVGAngleValue, SVGMarkerOrientType>> {
     static std::pair<SVGAngleValue, SVGMarkerOrientType> fromString(const String& string)
     {
         SVGAngleValue angle;
