diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGNumberList.h
@@ -76,11 +76,11 @@
 
         for (const auto& number : m_items) {
             if (builder.length())
                 builder.append(' ');
 
-            builder.appendFixedPrecisionNumber(number->value());
+            builder.append(number->value());
         }
 
         return builder.toString();
     }
 };
