diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGTransformList.idl b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGTransformList.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGTransformList.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGTransformList.idl
@@ -35,8 +35,8 @@
     [MayThrowException] SVGTransform replaceItem(SVGTransform newItem, unsigned long index);
     [MayThrowException] SVGTransform removeItem(unsigned long index);
     [MayThrowException] SVGTransform appendItem(SVGTransform newItem);
     [MayThrowException] setter void (unsigned long index, SVGTransform newItem);
 
-    [MayThrowException, NewObject] SVGTransform createSVGTransformFromMatrix(SVGMatrix matrix);
+    [MayThrowException, NewObject] SVGTransform createSVGTransformFromMatrix(optional DOMMatrix2DInit matrix);
     [MayThrowException] SVGTransform consolidate();
 };
