diff a/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLMaskFill.m b/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLMaskFill.m
--- a/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLMaskFill.m
+++ b/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLMaskFill.m
@@ -116,11 +116,11 @@
                     MTLVC_MASK_CACHE_TILE_WIDTH, pMask, dstOps,
                     MTLVC_MASK_CACHE_TILE_WIDTH);
         }
     }
 
-    MTLVertexCache_FlushVertexCache(mtlc);
+    MTLVertexCache_DisableMaskCache(mtlc);
 }
 
 JNIEXPORT void JNICALL
 Java_sun_java2d_metal_MTLMaskFill_maskFill
     (JNIEnv *env, jobject self,
