diff a/test/jdk/jdk/internal/platform/cgroup/TestCgroupMetrics.java b/test/jdk/jdk/internal/platform/cgroup/TestCgroupMetrics.java
--- a/test/jdk/jdk/internal/platform/cgroup/TestCgroupMetrics.java
+++ b/test/jdk/jdk/internal/platform/cgroup/TestCgroupMetrics.java
@@ -41,17 +41,10 @@
             System.out.println("TEST PASSED!!!");
             return;
         }
 
         MetricsTester metricsTester = new MetricsTester();
-        metricsTester.setup();
-        metricsTester.testCpuAccounting();
-        metricsTester.testCpuSchedulingMetrics();
-        metricsTester.testCpuSets();
-        metricsTester.testMemorySubsystem();
-        metricsTester.testBlkIO();
-        metricsTester.testCpuConsumption();
-        metricsTester.testMemoryUsage();
+        metricsTester.testAll(metrics);
         System.out.println("TEST PASSED!!!");
     }
 
 }
