diff a/modules/javafx.web/src/main/native/Source/WebCore/editing/RenderedPosition.h b/modules/javafx.web/src/main/native/Source/WebCore/editing/RenderedPosition.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/editing/RenderedPosition.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/editing/RenderedPosition.h
@@ -70,12 +70,12 @@
 
 private:
     bool operator==(const RenderedPosition&) const { return false; }
     explicit RenderedPosition(RenderObject*, InlineBox*, int offset);
 
-    InlineBox* prevLeafChild() const;
-    InlineBox* nextLeafChild() const;
+    InlineBox* previousLeafOnLine() const;
+    InlineBox* nextLeafOnLine() const;
     bool atLeftmostOffsetInBox() const { return m_inlineBox && m_offset == m_inlineBox->caretLeftmostOffset(); }
     bool atRightmostOffsetInBox() const { return m_inlineBox && m_offset == m_inlineBox->caretRightmostOffset(); }
     bool atLeftBoundaryOfBidiRun(ShouldMatchBidiLevel, unsigned char bidiLevelOfRun) const;
     bool atRightBoundaryOfBidiRun(ShouldMatchBidiLevel, unsigned char bidiLevelOfRun) const;
 
@@ -84,27 +84,27 @@
     int m_offset;
 
     static InlineBox* uncachedInlineBox() { return reinterpret_cast<InlineBox*>(1); }
     // Needs to be different form 0 so pick 1 because it's also on the null page.
 
-    mutable InlineBox* m_prevLeafChild;
-    mutable InlineBox* m_nextLeafChild;
+    mutable InlineBox* m_previousLeafOnLine;
+    mutable InlineBox* m_nextLeafOnLine;
 };
 
 inline RenderedPosition::RenderedPosition()
     : m_offset(0)
-    , m_prevLeafChild(uncachedInlineBox())
-    , m_nextLeafChild(uncachedInlineBox())
+    , m_previousLeafOnLine(uncachedInlineBox())
+    , m_nextLeafOnLine(uncachedInlineBox())
 {
 }
 
 inline RenderedPosition::RenderedPosition(RenderObject* renderer, InlineBox* box, int offset)
     : m_renderer(renderer)
     , m_inlineBox(box)
     , m_offset(offset)
-    , m_prevLeafChild(uncachedInlineBox())
-    , m_nextLeafChild(uncachedInlineBox())
+    , m_previousLeafOnLine(uncachedInlineBox())
+    , m_nextLeafOnLine(uncachedInlineBox())
 {
 }
 
 bool renderObjectContainsPosition(RenderObject*, const Position&);
 
