diff a/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/ConnectionToolkit.java b/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/ConnectionToolkit.java
--- a/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/ConnectionToolkit.java
+++ b/application/org.openjdk.jmc.rjmx/src/main/java/org/openjdk/jmc/rjmx/ConnectionToolkit.java
@@ -54,10 +54,11 @@
 import javax.management.remote.JMXServiceURL;
 
 import org.openjdk.jmc.common.version.JavaVMVersionToolkit;
 import org.openjdk.jmc.common.version.JavaVersion;
 import org.openjdk.jmc.rjmx.internal.RJMXConnection;
+import org.openjdk.jmc.ui.common.jvm.JVMDescriptor;
 
 /**
  * Toolkit providing utility methods to retrieve MBean proxy objects, invoke JMX operations and
  * query a connection about its properties.
  */
@@ -368,11 +369,29 @@
 	 * Returns {@code true} if the connection handle is associated with an Oracle built JVM,
 	 * {@code false} otherwise. This method <b>does not</b> require the connection handle to be
 	 * connected.
 	 */
 	public static boolean isOracle(IConnectionHandle handle) {
-		String vendor = handle.getServerDescriptor().getJvmInfo().getJvmVendor();
+		JVMDescriptor descriptor = handle.getServerDescriptor().getJvmInfo();
+		// This should normally not happen for discovered JVMs, but users can create custom connections
+		String vendor = null;
+		if (descriptor != null) {
+			vendor = descriptor.getJvmVendor();
+		} else {
+			// We try checking if connected
+			if (handle.isConnected()) {
+				MBeanServerConnection connection = handle.getServiceOrNull(MBeanServerConnection.class);
+				if (connection != null) {
+					try {
+						vendor = getRuntimeBean(connection).getVmVendor();
+					} catch (IOException e) {
+						// Worst case we classify JVM vendor wrong
+						RJMXPlugin.getDefault().getLogger().log(Level.WARNING, "Could not check if Oracle JVM", e);
+					}
+				}
+			}
+		}
 		return vendor != null && vendor.contains("Oracle");
 	}
 
 	/**
 	 * This will return true if the java version is above or equal the supplied value. (For example
