diff a/src/jdk.incubator.jpackage/macosx/classes/jdk/incubator/jpackage/internal/MacDmgBundler.java b/src/jdk.incubator.jpackage/macosx/classes/jdk/incubator/jpackage/internal/MacDmgBundler.java
--- a/src/jdk.incubator.jpackage/macosx/classes/jdk/incubator/jpackage/internal/MacDmgBundler.java
+++ b/src/jdk.incubator.jpackage/macosx/classes/jdk/incubator/jpackage/internal/MacDmgBundler.java
@@ -357,11 +357,11 @@
                 hdiutil,
                 "attach",
                 protoDMG.toAbsolutePath().toString(),
                 hdiUtilVerbosityFlag,
                 "-mountroot", imagesRoot.toAbsolutePath().toString());
-        IOUtils.exec(pb, false, null, true);
+        IOUtils.exec(pb, false, null, true, Executor.INFINITE_TIMEOUT);
 
         Path mountedRoot = imagesRoot.resolve(APP_NAME.fetchFrom(params));
 
         // Copy app image, since we did not create DMG with it, but instead we created
         // empty one.
@@ -390,11 +390,11 @@
             // to install-dir in DMG as critical error, since it can fail in
             // headless enviroment.
             try {
                 pb = new ProcessBuilder("osascript",
                         getConfig_VolumeScript(params).toAbsolutePath().toString());
-                IOUtils.exec(pb);
+                IOUtils.exec(pb, 180); // Wait 3 minutes. See JDK-8248248.
             } catch (IOException ex) {
                 Log.verbose(ex);
             }
 
             // volume icon
