diff a/src/jdk.hotspot.agent/linux/native/libsaproc/DwarfParser.cpp b/src/jdk.hotspot.agent/linux/native/libsaproc/DwarfParser.cpp
--- a/src/jdk.hotspot.agent/linux/native/libsaproc/DwarfParser.cpp
+++ b/src/jdk.hotspot.agent/linux/native/libsaproc/DwarfParser.cpp
@@ -102,11 +102,13 @@
   jlong result = 0L;
 
   DwarfParser *parser = new DwarfParser(reinterpret_cast<lib_info *>(lib));
   if (!parser->is_parseable()) {
     jclass ex_cls = env->FindClass("sun/jvm/hotspot/debugger/DebuggerException");
-    env->ThrowNew(ex_cls, "DWARF not found");
+    if (!env->ExceptionOccurred()) {
+        env->ThrowNew(ex_cls, "DWARF not found");
+    }
     return 0L;
   }
 
   return reinterpret_cast<jlong>(parser);
 }
@@ -144,11 +146,13 @@
 JNIEXPORT void JNICALL Java_sun_jvm_hotspot_debugger_linux_amd64_DwarfParser_processDwarf0
   (JNIEnv *env, jobject this_obj, jlong pc) {
   DwarfParser *parser = reinterpret_cast<DwarfParser *>(get_dwarf_context(env, this_obj));
   if (!parser->process_dwarf(pc)) {
     jclass ex_cls = env->FindClass("sun/jvm/hotspot/debugger/DebuggerException");
-    env->ThrowNew(ex_cls, "Could not find PC in DWARF");
+    if (!env->ExceptionOccurred()) {
+        env->ThrowNew(ex_cls, "Could not find PC in DWARF");
+    }
     return;
   }
 }
 
 /*
