diff a/webrev/src/test/java/org/openjdk/skara/webrev/WebrevTests.java b/webrev/src/test/java/org/openjdk/skara/webrev/WebrevTests.java
--- a/webrev/src/test/java/org/openjdk/skara/webrev/WebrevTests.java
+++ b/webrev/src/test/java/org/openjdk/skara/webrev/WebrevTests.java
@@ -34,11 +34,11 @@
 
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 class WebrevTests {
     void assertContains(Path file, String text) throws IOException {
-        var contents = Files.readString(file);
+        var contents = Files.readString(file).replaceAll("\\R", "\n");
         assertTrue(contents.contains(text));
     }
 
     @ParameterizedTest
     @EnumSource(VCS.class)
