diff a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/C-X.java.template b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/C-X.java.template
--- a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/C-X.java.template
+++ b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/C-X.java.template
@@ -21,15 +21,10 @@
     public static final MemoryLayout LAYOUT = ${LAYOUT};
     public static final Class<?> CARRIER = ${CARRIER}.class;
     private static final VarHandle handle = LAYOUT.varHandle(CARRIER);
     private static final VarHandle arrayHandle = arrayHandle(LAYOUT, CARRIER);
 
-    private static void copy(MemoryAddress addr, ${CARRIER}[] arr) {
-        var heapSegment = MemorySegment.ofArray(arr);
-        MemoryAddress.copy(heapSegment.baseAddress(), addr, arr.length * sizeof());
-    }
-
     public static ${CARRIER} get(MemoryAddress addr) {
         return (${CARRIER}) handle.get(addr);
     }
 
     public static void set(MemoryAddress addr, ${CARRIER} value) {
@@ -67,18 +62,18 @@
     }
 
     public static MemorySegment allocateArray(${CARRIER}[] arr) {
         var arrLayout = MemoryLayout.ofSequence(arr.length, LAYOUT);
         var seg = MemorySegment.allocateNative(arrLayout);
-        copy(seg.baseAddress(), arr);
+        seg.copyFrom(MemorySegment.ofArray(arr));
         return seg;
     }
 
     public static MemoryAddress allocateArray(${CARRIER}[] arr, NativeAllocationScope scope) {
         var arrLayout = MemoryLayout.ofSequence(arr.length, LAYOUT);
         var addr = scope.allocate(arrLayout);
-        copy(addr, arr);
+        addr.segment().copyFrom(MemorySegment.ofArray(arr));
         return addr;
     }
 
     public static long sizeof() {
         return LAYOUT.byteSize();
@@ -89,9 +84,9 @@
         var elemSize = sizeof();
         if (segSize % elemSize != 0) {
             throw new UnsupportedOperationException("segment cannot contain integral number of elements");
         }
         ${CARRIER}[] array = new ${CARRIER}[(int) (segSize / elemSize)];
-        MemoryAddress.copy(seg.baseAddress(), MemorySegment.ofArray(array).baseAddress(), array.length * elemSize);
+        MemorySegment.ofArray(array).copyFrom(seg);
         return array;
     }
 }
diff a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/Cstring.java.template b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/Cstring.java.template
--- a/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/Cstring.java.template
+++ b/src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/resources/Cstring.java.template
@@ -17,16 +17,10 @@
         return MemoryLayout.ofSequence(elemLayout)
                 .varHandle(elemCarrier, MemoryLayout.PathElement.sequenceElement());
     }
     private final static VarHandle byteArrHandle = arrayHandle(C_CHAR, byte.class);
 
-    private static void copy(MemoryAddress addr, byte[] bytes) {
-        var heapSegment = MemorySegment.ofArray(bytes);
-        MemoryAddress.copy(heapSegment.baseAddress(), addr, bytes.length);
-        byteArrHandle.set(addr, (long)bytes.length, (byte)0);
-    }
-
     private static MemorySegment toCString(byte[] bytes) {
         MemoryLayout strLayout = MemoryLayout.ofSequence(bytes.length + 1, C_CHAR);
         MemorySegment segment = MemorySegment.allocateNative(strLayout);
         MemoryAddress addr = segment.baseAddress();
         copy(addr, bytes);
@@ -34,11 +28,11 @@
     }
 
     private static MemoryAddress toCString(byte[] bytes, NativeAllocationScope scope) {
         MemoryLayout strLayout = MemoryLayout.ofSequence(bytes.length + 1, C_CHAR);
         MemoryAddress addr = scope.allocate(strLayout);
-        copy(addr, bytes);
+        addr.segment().copyFrom(MemorySegment.ofArray(bytes));
         return addr;
     }
 
     public static void copy(MemoryAddress addr, String str) {
         copy(addr, str.getBytes());
@@ -46,10 +40,19 @@
 
     public static void copy(MemoryAddress addr, String str, Charset charset) {
         copy(addr, str.getBytes(charset));
     }
 
+    //where
+    private static void copy(MemoryAddress addr, byte[] bytes) {
+            var heapSegment = MemorySegment.ofArray(bytes);
+            addr.segment()
+                    .asSlice(addr.segmentOffset(), bytes.length)
+                    .copyFrom(heapSegment);
+            byteArrHandle.set(addr, (long)bytes.length, (byte)0);
+        }
+
     public static MemorySegment toCString(String str) {
          return toCString(str.getBytes());
     }
 
     public static MemorySegment toCString(String str, Charset charset) {
