diff a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Frame.java b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Frame.java
--- a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Frame.java
+++ b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/Frame.java
@@ -358,10 +358,11 @@
             case 'J':
                 return LONG;
             case 'D':
                 return DOUBLE;
             case 'L':
+            case 'Q':
                 internalName = buffer.substring(offset + 1, buffer.length() - 1);
                 return REFERENCE_KIND | symbolTable.addType(internalName);
             case '[':
                 int elementDescriptorOffset = offset + 1;
                 while (buffer.charAt(elementDescriptorOffset) == '[') {
@@ -392,10 +393,11 @@
                         break;
                     case 'D':
                         typeValue = DOUBLE;
                         break;
                     case 'L':
+                    case 'Q':
                         internalName = buffer.substring(elementDescriptorOffset + 1, buffer.length() - 1);
                         typeValue = REFERENCE_KIND | symbolTable.addType(internalName);
                         break;
                     default:
                         throw new IllegalArgumentException();
