diff a/src/java.base/share/classes/sun/text/CollatorUtilities.java b/src/java.base/share/classes/sun/text/CollatorUtilities.java
--- a/src/java.base/share/classes/sun/text/CollatorUtilities.java
+++ b/src/java.base/share/classes/sun/text/CollatorUtilities.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2000, 2005, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -23,11 +23,11 @@
  * questions.
  */
 
 package sun.text;
 
-import sun.text.normalizer.NormalizerBase;
+import jdk.internal.icu.text.NormalizerBase;
 
 public class CollatorUtilities {
 
     public static int toLegacyMode(NormalizerBase.Mode mode) {
         // find the index of the legacy mode in the table;
