diff a/src/hotspot/share/runtime/reflection.cpp b/src/hotspot/share/runtime/reflection.cpp
--- a/src/hotspot/share/runtime/reflection.cpp
+++ b/src/hotspot/share/runtime/reflection.cpp
@@ -391,12 +391,11 @@
         THROW_0(vmSymbols::java_lang_IllegalArgumentException());
       }
       dim += k_dim;
     }
   }
-  ArrayStorageProperties storage_props = ArrayStorageProperties::for_signature(klass->name());
-  klass = klass->array_klass(storage_props, dim, CHECK_NULL);
+  klass = klass->array_klass(dim, CHECK_NULL);
   oop obj = ArrayKlass::cast(klass)->multi_allocate(len, dimensions, CHECK_NULL);
   assert(obj->is_array(), "just checking");
   return arrayOop(obj);
 }
 
