diff a/src/java.management/share/classes/sun/management/MemoryNotifInfoCompositeData.java b/src/java.management/share/classes/sun/management/MemoryNotifInfoCompositeData.java
--- a/src/java.management/share/classes/sun/management/MemoryNotifInfoCompositeData.java
+++ b/src/java.management/share/classes/sun/management/MemoryNotifInfoCompositeData.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2004, 2008, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2004, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -36,10 +36,11 @@
  * A CompositeData for MemoryNotificationInfo for the local management support.
  * This class avoids the performance penalty paid to the
  * construction of a CompositeData use in the local case.
  */
 public class MemoryNotifInfoCompositeData extends LazyCompositeData {
+    @SuppressWarnings("serial") // Not statically typed as Serializable
     private final MemoryNotificationInfo memoryNotifInfo;
 
     private MemoryNotifInfoCompositeData(MemoryNotificationInfo info) {
         this.memoryNotifInfo = info;
     }
