diff a/modules/javafx.graphics/src/main/native-glass/win/GlassApplication.cpp b/modules/javafx.graphics/src/main/native-glass/win/GlassApplication.cpp
--- a/modules/javafx.graphics/src/main/native-glass/win/GlassApplication.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/win/GlassApplication.cpp
@@ -353,26 +353,33 @@
  * JNI section
  *******************************************************/
 
 extern "C" {
 
+#ifndef STATIC_BUILD
 BOOL WINAPI DllMain(HANDLE hinstDLL, DWORD dwReason, LPVOID lpvReserved)
 {
     if (dwReason == DLL_PROCESS_ATTACH) {
         GlassApplication::SetHInstance((HINSTANCE)hinstDLL);
     }
     return TRUE;
 }
+#endif
 
 /*
  * Class:     com_sun_glass_ui_win_WinApplication
  * Method:    initIDs
  * Signature: ()V
  */
 JNIEXPORT void JNICALL Java_com_sun_glass_ui_win_WinApplication_initIDs
   (JNIEnv *env, jclass cls, jfloat overrideUIScale)
 {
+#ifdef STATIC_BUILD
+    HINSTANCE hInstExe = ::GetModuleHandle(NULL);
+    GlassApplication::SetHInstance((HINSTANCE)hInstExe);
+#endif
+
     GlassApplication::overrideUIScale = overrideUIScale;
 
     javaIDs.Application.reportExceptionMID =
         env->GetStaticMethodID(cls, "reportException", "(Ljava/lang/Throwable;)V");
     ASSERT(javaIDs.Application.reportExceptionMID);
