diff a/modules/javafx.graphics/src/main/native-glass/win/Utils.cpp b/modules/javafx.graphics/src/main/native-glass/win/Utils.cpp
--- a/modules/javafx.graphics/src/main/native-glass/win/Utils.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/win/Utils.cpp
@@ -107,11 +107,15 @@
     return modifiers;
 }
 
 extern "C" {
 
+#ifdef STATIC_BUILD
+JNIEXPORT jint JNICALL JNI_OnLoad_glass(JavaVM *vm, void *reserved)
+#else
 JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, void *reserved)
+#endif
 {
     memset(&javaIDs, 0, sizeof(javaIDs));
     jvm = vm;
     return JNI_VERSION_1_2;
 }
