diff a/test/jdk/java/net/MulticastSocket/MulticastAddresses.java b/test/jdk/java/net/MulticastSocket/MulticastAddresses.java
--- a/test/jdk/java/net/MulticastSocket/MulticastAddresses.java
+++ b/test/jdk/java/net/MulticastSocket/MulticastAddresses.java
@@ -25,10 +25,12 @@
  * @test
  * @bug 4488458
  * @library /test/lib
  * @summary Test that MutlicastSocket.joinGroup is working for
  *          various multicast and non-multicast addresses.
+ * @run main MulticastAddresses
+ * @run main/othervm -Djdk.net.usePlainDatagramSocketImpl MulticastAddresses
  */
 
 import jdk.test.lib.NetworkConfiguration;
 
 import java.net.*;
