diff a/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h b/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h
--- a/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h
+++ b/src/java.desktop/macosx/native/libawt_lwawt/awt/common.h
@@ -68,10 +68,11 @@
 struct TxtFrameUniforms {
     vector_float4 color;
     int mode;
     int isSrcOpaque;
     int isDstOpaque;
+    float extraAlpha;
 };
 
 struct AnchorData
 {
     vector_float3 xParams;
