diff a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
--- a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
@@ -169,11 +169,11 @@
     GdkWindow *gdk_window;
     BgColor bg_color;
     GdkDevice *pointer_device;
 
     static WindowContext* sm_mouse_drag_window;
-
+    static WindowContext* sm_grab_window;
 public:
     WindowContext(jobject, WindowContext *, long, WindowFrameType, WindowType, GdkWMFunction);
 
     bool hasIME();
     bool filterIME(GdkEvent *);
@@ -259,11 +259,11 @@
     void size_position_notify(bool, bool);
     void update_ontop_tree(bool);
     bool on_top_inherited();
     bool effective_on_top();
     GdkDevice * get_pointer_device();
-    void grab_mouse_drag_focus();
+    bool grab_mouse_drag_focus(GdkCursor *cursor, bool owner_events);
     void ungrab_mouse_drag_focus();
 };
 
 void destroy_and_delete_ctx(WindowContext *ctx);
 
