diff a/modules/javafx.graphics/src/test/java/test/com/sun/javafx/scene/layout/region/BackgroundRepeatConverterTest.java b/modules/javafx.graphics/src/test/java/test/com/sun/javafx/scene/layout/region/BackgroundRepeatConverterTest.java
--- a/modules/javafx.graphics/src/test/java/test/com/sun/javafx/scene/layout/region/BackgroundRepeatConverterTest.java
+++ b/modules/javafx.graphics/src/test/java/test/com/sun/javafx/scene/layout/region/BackgroundRepeatConverterTest.java
@@ -30,11 +30,11 @@
 import javafx.css.ParsedValue;
 import com.sun.javafx.css.ParsedValueImpl;
 import com.sun.javafx.scene.layout.region.RepeatStruct;
 
 import static org.junit.Assert.assertEquals;
-import com.sun.javafx.scene.layout.region.RepeatStructConverter;
+import static org.junit.Assert.fail;
 import com.sun.javafx.scene.layout.region.RepeatStructConverter;
 
 /**
  */
 public class BackgroundRepeatConverterTest {
@@ -51,21 +51,26 @@
     }
 
     /*
         -fx-background-repeat: null
      */
-    @Test(expected=NullPointerException.class)
+    @Test
     public void scenario2() {
         ParsedValue<String,BackgroundRepeat>[][] values = new ParsedValueImpl[][] {
                 {null}
         };
 
         ParsedValueImpl<ParsedValue<String,BackgroundRepeat>[][], RepeatStruct[]> value =
                 new ParsedValueImpl<ParsedValue<String,BackgroundRepeat>[][], RepeatStruct[]>(
                         values, null
                 );
-        RepeatStruct[] results = RepeatStructConverter.getInstance().convert(value, null);
+        try {
+            RepeatStruct[] results = RepeatStructConverter.getInstance().convert(value, null);
+            fail("Expected NullPointerException");
+        } catch (NullPointerException expected) {
+            // Test Pass, NPE is expected
+        }
     }
 
     /*
         -fx-background-repeat: repeat round
      */
