diff a/test/jdk/com/sun/nio/sctp/SctpMultiChannel/Branch.java b/test/jdk/com/sun/nio/sctp/SctpMultiChannel/Branch.java
--- a/test/jdk/com/sun/nio/sctp/SctpMultiChannel/Branch.java
+++ b/test/jdk/com/sun/nio/sctp/SctpMultiChannel/Branch.java
@@ -208,11 +208,12 @@
                 check(assocs.size() == 1, "there should be only one association");
 
                 /* echo the message */
                 debug("Server: echoing first message");
                 buffer.flip();
-                int bytes = serverChannel.send(buffer, info);
+                MessageInfo sendInfo = MessageInfo.createOutgoing(info.association(), null, 0);
+                int bytes = serverChannel.send(buffer, sendInfo);
                 debug("Server: sent " + bytes + "bytes");
 
                 clientFinishedLatch.await(10L, TimeUnit.SECONDS);
                 serverFinishedLatch.countDown();
             } catch (IOException ioe) {
