diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/CheckedBoolean.h
@@ -26,10 +26,11 @@
 #pragma once
 
 #include <wtf/Assertions.h>
 
 class CheckedBoolean {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
 #if !ASSERT_DISABLED
     CheckedBoolean(const CheckedBoolean& other)
         : m_value(other.m_value)
         , m_checked(false)
