diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/KeyValuePair.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/KeyValuePair.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/KeyValuePair.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/KeyValuePair.h
@@ -29,10 +29,12 @@
 
 namespace WTF {
 
 template<typename KeyTypeArg, typename ValueTypeArg>
 struct KeyValuePair {
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
+
     using KeyType = KeyTypeArg;
     using ValueType = ValueTypeArg;
 
     KeyValuePair()
     {
