diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/NeverDestroyed.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/NeverDestroyed.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/NeverDestroyed.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/NeverDestroyed.h
@@ -1,7 +1,7 @@
 /*
- * Copyright (C) 2013-2017 Apple Inc. All rights reserved.
+ * Copyright (C) 2013-2019 Apple Inc. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
  * are met:
  * 1. Redistributions of source code must retain the above copyright
@@ -58,13 +58,17 @@
     }
 
     operator T&() { return *storagePointer(); }
     T& get() { return *storagePointer(); }
 
+    T* operator->() { return storagePointer(); }
+
     operator const T&() const { return *storagePointer(); }
     const T& get() const { return *storagePointer(); }
 
+    const T* operator->() const { return storagePointer(); }
+
 private:
     using PointerType = typename std::remove_const<T>::type*;
 
     PointerType storagePointer() const { return const_cast<PointerType>(reinterpret_cast<const T*>(&m_storage)); }
 
