diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/ParallelJobsOpenMP.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/ParallelJobsOpenMP.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/ParallelJobsOpenMP.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/ParallelJobsOpenMP.h
@@ -34,10 +34,11 @@
 
 namespace WTF {
 
 class ParallelEnvironment {
     WTF_MAKE_NONCOPYABLE(ParallelEnvironment);
+    WTF_MAKE_FAST_ALLOCATED;
 public:
     typedef void (*ThreadFunction)(void*);
 
     ParallelEnvironment(ThreadFunction threadFunction, size_t sizeOfParameter, int requestedJobNumber) :
         m_threadFunction(threadFunction),
