diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/StreamBuffer.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/StreamBuffer.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/StreamBuffer.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/StreamBuffer.h
@@ -32,11 +32,13 @@
 
 #include <wtf/Deque.h>
 
 namespace WTF {
 
-template <typename T, size_t BlockSize> class StreamBuffer {
+template <typename T, size_t BlockSize>
+class StreamBuffer {
+    WTF_MAKE_FAST_ALLOCATED;
 private:
     typedef Vector<T> Block;
 public:
     StreamBuffer()
         : m_size(0)
@@ -56,11 +58,11 @@
             return;
 
         m_size += size;
         while (size) {
             if (!m_buffer.size() || m_buffer.last()->size() == BlockSize)
-                m_buffer.append(std::make_unique<Block>());
+                m_buffer.append(makeUnique<Block>());
             size_t appendSize = std::min(BlockSize - m_buffer.last()->size(), size);
             m_buffer.last()->append(data, appendSize);
             data += appendSize;
             size -= appendSize;
         }
