diff a/modules/javafx.web/src/main/native/Source/WTF/wtf/URLHelpers.h b/modules/javafx.web/src/main/native/Source/WTF/wtf/URLHelpers.h
--- a/modules/javafx.web/src/main/native/Source/WTF/wtf/URLHelpers.h
+++ b/modules/javafx.web/src/main/native/Source/WTF/wtf/URLHelpers.h
@@ -28,12 +28,10 @@
  */
 
 #pragma once
 
 #include <wtf/Forward.h>
-#include <wtf/Optional.h>
-#include <wtf/text/WTFString.h>
 
 namespace WTF {
 namespace URLHelpers {
 
 using URLDecodeFunction = String(&)(const String&);
