diff a/test/jdk/java/nio/file/attribute/BasicFileAttributeView/UnixSocketFile.java b/test/jdk/java/nio/file/attribute/BasicFileAttributeView/UnixSocketFile.java
--- a/test/jdk/java/nio/file/attribute/BasicFileAttributeView/UnixSocketFile.java
+++ b/test/jdk/java/nio/file/attribute/BasicFileAttributeView/UnixSocketFile.java
@@ -59,10 +59,18 @@
         if (b == -1) {
             System.err.println("Netcat command unavailable; skipping test.");
             return;
         }
 
+        // Verify that 'nc' accepts '-U' for Unix domain sockets.
+        // Skip the test if it is not.
+        Process procHelp = Runtime.getRuntime().exec(CMD_BASE + " -h");
+        if (procHelp.waitFor() != 0) {
+            System.err.println("Netcat does not accept required options; skipping test.");
+            return;
+        }
+
         // Create a new sub-directory of the nominal test directory in which
         // 'nc' will create the socket file.
         String testSubDir = System.getProperty("test.dir", ".")
             + File.separator + TEST_SUB_DIR;
         Path socketTestDir = Paths.get(testSubDir);
