diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/glx/GLContextGLX.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/glx/GLContextGLX.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/glx/GLContextGLX.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/glx/GLContextGLX.cpp
@@ -18,11 +18,10 @@
 
 #include "config.h"
 #include "GLContextGLX.h"
 
 #if USE(GLX)
-#include "GraphicsContext3D.h"
 #include "OpenGLShims.h"
 #include "PlatformDisplayX11.h"
 #include "XErrorTrapper.h"
 #include <GL/glx.h>
 #include <cairo.h>
@@ -412,12 +411,12 @@
 #endif
 
     return m_cairoDevice;
 }
 
-#if ENABLE(GRAPHICS_CONTEXT_3D)
-PlatformGraphicsContext3D GLContextGLX::platformContext()
+#if ENABLE(GRAPHICS_CONTEXT_GL)
+PlatformGraphicsContextGL GLContextGLX::platformContext()
 {
     return m_context.get();
 }
 #endif
 
