diff a/src/hotspot/cpu/x86/x86_64.ad b/src/hotspot/cpu/x86/x86_64.ad
--- a/src/hotspot/cpu/x86/x86_64.ad
+++ b/src/hotspot/cpu/x86/x86_64.ad
@@ -459,10 +459,11 @@
   return offset;
 }
 
 int MachCallNativeNode::ret_addr_offset() {
   int offset = 13; // movq r10,#addr; callq (r10)
+  offset += clear_avx_size();
   return offset;
 }
 //
 // Compute padding required for nodes which need alignment
 //
@@ -12567,13 +12568,11 @@
   match(CallNative);
   effect(USE meth);
 
   ins_cost(300);
   format %{ "call_native " %}
-  ins_encode %{
-    __ call(RuntimeAddress((address) $meth$$method));
-  %}
+  ins_encode(clear_avx, Java_To_Runtime(meth));
   ins_pipe(pipe_slow);
 %}
 
 // Call runtime without safepoint
 instruct CallLeafNoFPDirect(method meth)
