diff a/application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/formpage/commands/internal/ContextLookup.java b/application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/formpage/commands/internal/ContextLookup.java
--- a/application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/formpage/commands/internal/ContextLookup.java
+++ b/application/org.openjdk.jmc.ui/src/main/java/org/openjdk/jmc/ui/formpage/commands/internal/ContextLookup.java
@@ -36,26 +36,25 @@
 
 import org.eclipse.ui.IEditorPart;
 import org.eclipse.ui.IWorkbenchPage;
 import org.eclipse.ui.IWorkbenchPart;
 import org.eclipse.ui.IWorkbenchWindow;
+import org.eclipse.ui.PlatformUI;
 import org.eclipse.ui.forms.editor.FormEditor;
 import org.eclipse.ui.forms.editor.IFormPage;
 
-import org.openjdk.jmc.ui.UIPlugin;
-
 public class ContextLookup {
 	public static Object getContext(PrintStream out) {
 		IWorkbenchPart p = getActiveEditor();
 		if (p instanceof IFormPage) {
 			return ((IFormPage) p).getPartControl();
 		}
 		return null;
 	}
 
 	final static IEditorPart getActiveMainEditor() {
-		IWorkbenchWindow ww = UIPlugin.getDefault().getWorkbench().getActiveWorkbenchWindow();
+		IWorkbenchWindow ww = PlatformUI.getWorkbench().getActiveWorkbenchWindow();
 		if (ww != null) {
 			IWorkbenchPage wp = ww.getActivePage();
 			if (wp == null) {
 				return null;
 			}
