diff a/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java b/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java
--- a/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java
+++ b/src/java.base/share/classes/com/sun/crypto/provider/OAEPParameters.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2003, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -53,28 +53,14 @@
 public final class OAEPParameters extends AlgorithmParametersSpi {
 
     private String mdName;
     private MGF1ParameterSpec mgfSpec;
     private byte[] p;
-    private static ObjectIdentifier OID_MGF1;
-    private static ObjectIdentifier OID_PSpecified;
-
-    static {
-        try {
-            OID_MGF1 = new ObjectIdentifier(new int[] {1,2,840,113549,1,1,8});
-        } catch (IOException ioe) {
-            // should not happen
-            OID_MGF1 = null;
-        }
-        try {
-            OID_PSpecified =
-                new ObjectIdentifier(new int[] {1,2,840,113549,1,1,9});
-        } catch (IOException ioe) {
-            // should not happen
-            OID_PSpecified = null;
-        }
-    }
+    private static ObjectIdentifier OID_MGF1 =
+            ObjectIdentifier.of("1.2.840.113549.1.1.8");
+    private static ObjectIdentifier OID_PSpecified =
+            ObjectIdentifier.of("1.2.840.113549.1.1.9");
 
     public OAEPParameters() {
     }
 
     protected void engineInit(AlgorithmParameterSpec paramSpec)
