diff a/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java b/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java
--- a/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java
+++ b/cli/src/main/java/org/openjdk/skara/cli/GitPublish.java
@@ -229,8 +229,11 @@
             }
             err = pushAndFollow(remote, branch, isQuiet, browser);
         } else {
             err = push(remote, branch, isQuiet);
         }
-        System.exit(err);
+
+        if (err != 0) {
+            System.exit(err);
+        }
     }
 }
