diff a/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java b/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java
--- a/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java
+++ b/host/src/main/java/org/openjdk/skara/host/gitlab/GitLabMergeRequest.java
@@ -206,10 +206,14 @@
         for (var note : discussion.get("notes").asArray()) {
             // Ignore system generated comments
             if (note.get("system").asBoolean()) {
                 continue;
             }
+            // Ignore plain comments
+            if (!note.contains("position")) {
+                continue;
+            }
 
             var comment = parseReviewComment(discussion.get("id").asString(), parent, note.asObject());
             parent = comment;
             ret.add(comment);
         }
@@ -277,13 +281,12 @@
         return ret;
     }
 
     @Override
     public List<Comment> getComments() {
-        // FIXME: sort order doesn't seem to affect anything
         return request.get("notes").param("sort", "asc").execute().stream()
-                      .filter(entry -> !entry.get("resolvable").asBoolean()) // Ignore discussions - they are review comments
+                      .filter(entry -> !entry.contains("position")) // Ignore comments with a position - they are review comments
                       .filter(entry -> !entry.get("system").asBoolean()) // Ignore system generated comments
                 .map(this::parseComment)
                 .collect(Collectors.toList());
     }
 
