diff a/test/hotspot/jtreg/runtime/lambda-features/TestStaticandInstance.java b/test/hotspot/jtreg/runtime/lambda-features/TestStaticandInstance.java
--- a/test/hotspot/jtreg/runtime/lambda-features/TestStaticandInstance.java
+++ b/test/hotspot/jtreg/runtime/lambda-features/TestStaticandInstance.java
@@ -24,11 +24,11 @@
 /*
  * @test
  * @bug 8087342
  * @summary Test linkresolver search static, instance and overpass duplicates
  * @modules java.base/jdk.internal.org.objectweb.asm
- * @run main/othervm -XX:-BytecodeVerificationRemote -XX:-BytecodeVerificationLocal TestStaticandInstance
+ * @run main/othervm -XX:+UnlockDiagnosticVMOptions -XX:-BytecodeVerificationRemote -XX:-BytecodeVerificationLocal TestStaticandInstance
  */
 
 
 import java.util.*;
 import jdk.internal.org.objectweb.asm.*;
