diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/yarr/YarrUnicodeProperties.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/yarr/YarrUnicodeProperties.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/yarr/YarrUnicodeProperties.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/yarr/YarrUnicodeProperties.h
@@ -24,12 +24,11 @@
  */
 
 #pragma once
 
 #include "Yarr.h"
-#include <wtf/Optional.h>
-#include <wtf/text/WTFString.h>
+#include <wtf/Forward.h>
 
 namespace JSC { namespace Yarr {
 
 struct CharacterClass;
 
