diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/NavigatorMediaDevices.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/NavigatorMediaDevices.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/NavigatorMediaDevices.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/NavigatorMediaDevices.idl
@@ -31,6 +31,7 @@
 [
     Conditional=MEDIA_STREAM,
     EnabledAtRuntime=MediaDevices,
 ] partial interface Navigator {
     [SameObject, SecureContext, ContextAllowsMediaDevices] readonly attribute MediaDevices mediaDevices;
+    [Custom, EnabledBySetting=LegacyGetUserMedia] void getUserMedia(object constraints, object? successCallback, object? errorCallback);
 };
