diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/OverconstrainedError.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/OverconstrainedError.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/OverconstrainedError.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/mediastream/OverconstrainedError.h
@@ -42,10 +42,11 @@
         return adoptRef(*new OverconstrainedError(constraint, message));
     }
 
     String constraint() const { return m_constraint; }
     String message() const { return m_message; }
+    String name() const { return "OverconstrainedError"_s; }
 
 protected:
     explicit OverconstrainedError(const String& constraint, const String& message)
         : m_constraint(constraint)
         , m_message(message)
