diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/TempRegisterSet.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/TempRegisterSet.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/TempRegisterSet.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/jit/TempRegisterSet.h
@@ -192,14 +192,14 @@
     {
         ASSERT(i < totalNumberOfRegisters);
         return !!(m_set[i >> 3] & (1 << (i & 7)));
     }
 
-    static const unsigned totalNumberOfRegisters =
+    static constexpr unsigned totalNumberOfRegisters =
         GPRInfo::numberOfRegisters + FPRInfo::numberOfRegisters;
 
-    static const unsigned numberOfBytesInTempRegisterSet =
+    static constexpr unsigned numberOfBytesInTempRegisterSet =
         (totalNumberOfRegisters + 7) >> 3;
 
     uint8_t m_set[numberOfBytesInTempRegisterSet];
 };
 
