diff a/src/hotspot/share/gc/z/zNMethod.hpp b/src/hotspot/share/gc/z/zNMethod.hpp
--- a/src/hotspot/share/gc/z/zNMethod.hpp
+++ b/src/hotspot/share/gc/z/zNMethod.hpp
@@ -45,16 +45,17 @@
 
   static bool supports_entry_barrier(nmethod* nm);
 
   static bool is_armed(nmethod* nm);
   static void disarm(nmethod* nm);
+  static void arm(nmethod* nm, int arm_value);
 
-  static void nmethod_oops_do(nmethod* nm, OopClosure* cl);
+  static void nmethod_oops_do(nmethod* nm, OopClosure* cl, bool keepalive_is_strong);
 
   static void oops_do_begin();
   static void oops_do_end();
-  static void oops_do(OopClosure* cl);
+  static void oops_do(OopClosure* cl, bool keepalive_is_strong);
 
   static ZReentrantLock* lock_for_nmethod(nmethod* nm);
 
   static void unlink(ZWorkers* workers, bool unloading_occurred);
   static void purge(ZWorkers* workers);
