diff a/src/java.desktop/share/classes/sun/print/PathGraphics.java b/src/java.desktop/share/classes/sun/print/PathGraphics.java
--- a/src/java.desktop/share/classes/sun/print/PathGraphics.java
+++ b/src/java.desktop/share/classes/sun/print/PathGraphics.java
@@ -33,10 +33,12 @@
 import sun.font.Font2DHandle;
 import sun.font.FontManager;
 import sun.font.FontManagerFactory;
 import sun.font.FontUtilities;
 
+import java.awt.AlphaComposite;
+import java.awt.Composite;
 import java.awt.Color;
 import java.awt.Font;
 import java.awt.Graphics2D;
 import java.awt.Image;
 import java.awt.Paint;
@@ -1888,6 +1890,27 @@
         drawImageToPlatform(bufferedImage, xform, null,
                             0, 0, srcWidth, srcHeight, false);
 
     }
 
+    protected boolean isCompositing(Composite composite) {
+
+        boolean isCompositing = false;
+
+        if (composite instanceof AlphaComposite) {
+            AlphaComposite alphaComposite = (AlphaComposite) composite;
+            float alpha = alphaComposite.getAlpha();
+            int rule = alphaComposite.getRule();
+
+            if (alpha != 1.0
+                    || (rule != AlphaComposite.SRC
+                        && rule != AlphaComposite.SRC_OVER))
+            {
+                isCompositing = true;
+            }
+
+        } else {
+            isCompositing = true;
+        }
+        return isCompositing;
+    }
 }
