diff a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
--- a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp
@@ -382,11 +382,14 @@
 
                     save_cached_extents();
 
                     // set bounds again to set to correct window size that must
                     // be the total width and height accounting extents
-                    set_bounds(0, 0, false, false, -1, -1, -1, -1);
+                    // this is ignored if size is "content size" instead of "window size"
+                    if (geometry.window_size_set) {
+                        set_bounds(0, 0, false, false, geometry.current_w, geometry.current_h, -1, -1);
+                    }
                 }
             }
         }
     }
 }
@@ -969,10 +972,12 @@
     int newH = h > 0 ? h - geometry.adjust_h : ch;
 
     gboolean size_changed = FALSE;
     gboolean pos_changed = FALSE;
 
+    geometry.window_size_set = (w > 0 || h > 0);
+
     if (newW > 0 && newH > 0) {
         size_changed = TRUE;
 
         geometry.current_cw = newW;
         geometry.current_ch = newH;
