diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java
@@ -69,11 +69,11 @@
                               .map(Hash::new)
                               .findAny();
     }
 
     private boolean hasLegacyIntegrationNotice(Hash hash) {
-        var legacyIntegrationPattern = Pattern.compile("Changeset: " + hash.abbreviate());
+        var legacyIntegrationPattern = Pattern.compile("Changeset\\\\?: " + hash.abbreviate());
         return ignoredComments.stream()
                               .map(Comment::body)
                               .map(legacyIntegrationPattern::matcher)
                               .anyMatch(Matcher::find);
     }
