diff a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/IsoSharedHeap.h b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/IsoSharedHeap.h
--- a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/IsoSharedHeap.h
+++ b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/IsoSharedHeap.h
@@ -51,20 +51,20 @@
     unsigned m_remaining { 0 };
 };
 
 class IsoSharedHeap : public StaticPerProcess<IsoSharedHeap> {
 public:
-    IsoSharedHeap(std::lock_guard<Mutex>&)
+    IsoSharedHeap(const LockHolder&)
     {
     }
 
     template<unsigned>
     void* allocateNew(bool abortOnFailure);
 
 private:
     template<unsigned>
-    void* allocateSlow(bool abortOnFailure);
+    void* allocateSlow(const LockHolder&, bool abortOnFailure);
 
     IsoSharedPage* m_currentPage { nullptr };
     VariadicBumpAllocator m_allocator;
 };
 DECLARE_STATIC_PER_PROCESS_STORAGE(IsoSharedHeap);
