diff a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Vector.h b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Vector.h
--- a/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Vector.h
+++ b/modules/javafx.web/src/main/native/Source/bmalloc/bmalloc/Vector.h
@@ -70,12 +70,12 @@
     void resize(size_t);
 
     void shrinkToFit();
 
 private:
-    static const size_t growFactor = 2;
-    static const size_t shrinkFactor = 4;
+    static constexpr size_t growFactor = 2;
+    static constexpr size_t shrinkFactor = 4;
     static size_t initialCapacity() { return vmPageSize() / sizeof(T); }
 
     void growCapacity();
     void shrinkCapacity();
     void reallocateBuffer(size_t);
