diff a/test/jdk/java/net/DatagramSocket/ReuseAddressTest.java b/test/jdk/java/net/DatagramSocket/ReuseAddressTest.java
--- a/test/jdk/java/net/DatagramSocket/ReuseAddressTest.java
+++ b/test/jdk/java/net/DatagramSocket/ReuseAddressTest.java
@@ -34,10 +34,11 @@
  * @summary This test might fail intermittently as it needs a UDP socket that
  *          binds to the wildcard address.
  * @summary Expected SocketException not thrown when calling bind() with
  *   setReuseAddress(false)
  * @run main/othervm ReuseAddressTest
+ * @run main/othervm -Djdk.net.usePlainDatagramSocketImpl ReuseAddressTest
  */
 
 public class ReuseAddressTest {
 
     String getInfo(DatagramSocket soc) {
