diff a/test/jdk/java/net/DatagramSocket/SetGetReceiveBufferSize.java b/test/jdk/java/net/DatagramSocket/SetGetReceiveBufferSize.java
--- a/test/jdk/java/net/DatagramSocket/SetGetReceiveBufferSize.java
+++ b/test/jdk/java/net/DatagramSocket/SetGetReceiveBufferSize.java
@@ -25,10 +25,11 @@
  * @test
  * @bug 4173717 8243488
  * @summary Check that setReceiveBufferSize and getReceiveBufferSize work as expected
  * @run testng SetGetReceiveBufferSize
  * @run testng/othervm -Djava.net.preferIPv4Stack=true SetGetReceiveBufferSize
+ * @run testng/othervm -Djdk.net.usePlainDatagramSocketImpl SetGetReceiveBufferSize
  */
 
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
