diff a/modules/javafx.web/src/main/native/Source/WTF/CMakeLists.txt b/modules/javafx.web/src/main/native/Source/WTF/CMakeLists.txt
--- a/modules/javafx.web/src/main/native/Source/WTF/CMakeLists.txt
+++ b/modules/javafx.web/src/main/native/Source/WTF/CMakeLists.txt
@@ -1,5 +1,5 @@
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.10)
 include(WebKitCommon)
 set_property(DIRECTORY . PROPERTY FOLDER "WTF")
 
 add_subdirectory(wtf)
