diff a/tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java b/tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java
--- a/tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java
+++ b/tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java
@@ -189,11 +189,11 @@
         }
     }
 
     // Test for JDK-8237602
     @Test
-    public void testPermutGetTabsWithMoreTabs1() {
+    public void testAddingNewTabsWithExistingTabsAtSameIndex() {
         // Step #1
         Util.runAndWait(() -> {
             tabPane.getTabs().setAll(tab[0], tab[1]);
             tabPane.getSelectionModel().select(tab[1]);
         });
