diff a/test/hotspot/jtreg/compiler/whitebox/OSRFailureLevel4Test.java b/test/hotspot/jtreg/compiler/whitebox/OSRFailureLevel4Test.java
--- a/test/hotspot/jtreg/compiler/whitebox/OSRFailureLevel4Test.java
+++ b/test/hotspot/jtreg/compiler/whitebox/OSRFailureLevel4Test.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -41,17 +41,21 @@
 package compiler.whitebox;
 
 import sun.hotspot.WhiteBox;
 import java.lang.reflect.Executable;
 import java.lang.reflect.Method;
+import jtreg.SkippedException;
 
 public class OSRFailureLevel4Test extends Thread {
     private static final WhiteBox WHITE_BOX = WhiteBox.getWhiteBox();
     private static final long BACKEDGE_THRESHOLD = 150000;
     private Method method;
 
     public static void main(String[] args) throws Exception {
+        if (CompilerWhiteBoxTest.skipOnTieredCompilation(false)) {
+            throw new SkippedException("Test isn't applicable for non-tiered mode");
+        }
         OSRFailureLevel4Test test = new OSRFailureLevel4Test();
         test.test();
     }
 
     /**
