diff a/src/java.base/share/classes/sun/security/util/SecurityProperties.java b/src/java.base/share/classes/sun/security/util/SecurityProperties.java
--- a/src/java.base/share/classes/sun/security/util/SecurityProperties.java
+++ b/src/java.base/share/classes/sun/security/util/SecurityProperties.java
@@ -41,19 +41,24 @@
      *
      * @param  propName the name of the system or security property
      * @return the value of the system or security property
      */
     public static String privilegedGetOverridable(String propName) {
-        return AccessController.doPrivileged((PrivilegedAction<String>)
-            () -> {
-                String val = System.getProperty(propName);
-                if (val == null) {
-                    return Security.getProperty(propName);
-                } else {
-                    return val;
-                }
-            });
+        if (System.getSecurityManager() == null) {
+            return getOverridableProperty(propName);
+        } else {
+            return AccessController.doPrivileged((PrivilegedAction<String>) () -> getOverridableProperty(propName));
+        }
+    }
+
+    private static String getOverridableProperty(String propName) {
+        String val = System.getProperty(propName);
+        if (val == null) {
+            return Security.getProperty(propName);
+        } else {
+            return val;
+        }
     }
 
     /**
      * Returns true in case the system or security property "jdk.includeInExceptions"
      * contains the category refName
