diff a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListArchiveReaderBot.java b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListArchiveReaderBot.java
--- a/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListArchiveReaderBot.java
+++ b/bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/MailingListArchiveReaderBot.java
@@ -47,10 +47,14 @@
         this.archivePoster = archivePoster;
         this.lists = lists;
         this.repositories = repositories;
     }
 
+    private synchronized void invalidate(List<Email> messages) {
+        messages.forEach(m -> parsedEmailIds.remove(m.id()));
+    }
+
     synchronized void inspect(Conversation conversation) {
         // Is this a new conversation?
         if (!parsedConversations.containsKey(conversation.first().id())) {
             var first = conversation.first();
 
@@ -109,11 +113,11 @@
                 .collect(Collectors.toList());
         if (bridgeCandidates.isEmpty()) {
             return;
         }
 
-        var workItem = new CommentPosterWorkItem(pr, bridgeCandidates);
+        var workItem = new CommentPosterWorkItem(pr, bridgeCandidates, e -> invalidate(bridgeCandidates));
         commentQueue.add(workItem);
     }
 
     @Override
     public List<WorkItem> getPeriodicItems() {
