diff a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java
--- a/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java
@@ -177,10 +177,17 @@
         return URIBuilder.base(gitLabHost.getUri())
                          .setPath("/" + projectName + "/commit/" + hash.abbreviate())
                          .build();
     }
 
+    @Override
+    public URI webUrl(String baseRef, String headRef) {
+        return URIBuilder.base(gitLabHost.getUri())
+                         .setPath("/" + projectName + "/compare/" + baseRef + "..." + headRef)
+                         .build();
+    }
+
     @Override
     public VCS repositoryType() {
         return VCS.GIT;
     }
 
