diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/coordinated/TiledBackingStore.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/coordinated/TiledBackingStore.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/coordinated/TiledBackingStore.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/graphics/texmap/coordinated/TiledBackingStore.cpp
@@ -218,11 +218,11 @@
 
     // Now construct the tile(s) within the shortest distance.
     unsigned tilesToCreateCount = tilesToCreate.size();
     for (unsigned n = 0; n < tilesToCreateCount; ++n) {
         Tile::Coordinate coordinate = tilesToCreate[n];
-        m_tiles.add(coordinate, std::make_unique<Tile>(*this, coordinate));
+        m_tiles.add(coordinate, makeUnique<Tile>(*this, coordinate));
     }
     requiredTileCount -= tilesToCreateCount;
 
     // Re-call createTiles on a timer to cover the visible area with the newest shortest distance.
     m_pendingTileCreation = requiredTileCount;
