diff a/modules/javafx.graphics/src/main/java/com/sun/javafx/scene/PointLightHelper.java b/modules/javafx.graphics/src/main/java/com/sun/javafx/scene/PointLightHelper.java
--- a/modules/javafx.graphics/src/main/java/com/sun/javafx/scene/PointLightHelper.java
+++ b/modules/javafx.graphics/src/main/java/com/sun/javafx/scene/PointLightHelper.java
@@ -72,6 +72,6 @@
 
     public interface PointLightAccessor {
         NGNode doCreatePeer(Node node);
         void doUpdatePeer(Node node);
     }
-}
+}
