diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/IsoAlignedMemoryAllocator.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/IsoAlignedMemoryAllocator.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/IsoAlignedMemoryAllocator.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/IsoAlignedMemoryAllocator.cpp
@@ -23,10 +23,11 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #include "config.h"
 #include "IsoAlignedMemoryAllocator.h"
+#include "MarkedBlock.h"
 
 namespace JSC {
 
 IsoAlignedMemoryAllocator::IsoAlignedMemoryAllocator()
 {
@@ -86,7 +87,22 @@
 void IsoAlignedMemoryAllocator::dump(PrintStream& out) const
 {
     out.print("Iso(", RawPointer(this), ")");
 }
 
+void* IsoAlignedMemoryAllocator::tryAllocateMemory(size_t)
+{
+    RELEASE_ASSERT_NOT_REACHED();
+}
+
+void IsoAlignedMemoryAllocator::freeMemory(void*)
+{
+    RELEASE_ASSERT_NOT_REACHED();
+}
+
+void* IsoAlignedMemoryAllocator::tryReallocateMemory(void*, size_t)
+{
+    RELEASE_ASSERT_NOT_REACHED();
+}
+
 } // namespace JSC
 
