diff a/test/jdk/sun/security/pkcs11/tls/TestPRF.java b/test/jdk/sun/security/pkcs11/tls/TestPRF.java
--- a/test/jdk/sun/security/pkcs11/tls/TestPRF.java
+++ b/test/jdk/sun/security/pkcs11/tls/TestPRF.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2005, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -109,18 +109,10 @@
                     SecretKey key;
                     try {
                         kg.init(spec);
                         key = kg.generateKey();
                     } catch (Exception e) {
-                        if (secret == null) {
-                            // This fails on Solaris, but since we never call this
-                            // API for this case in JSSE, ignore the failure.
-                            // (SunJSSE uses the CKM_TLS_KEY_AND_MAC_DERIVE
-                            // mechanism)
-                            System.out.print("X");
-                            continue;
-                        }
                         System.out.println();
                         throw new Exception("Error on line: " + lineNumber, e);
                     }
                     byte[] enc = key.getEncoded();
                     if (Arrays.equals(output, enc) == false) {
