diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/D2L.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/D2L.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/D2L.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/D2L.java
@@ -23,11 +23,10 @@
 
 /**
  * D2L - Convert double to long
  * <PRE>Stack: ..., value.word1, value.word2 -&gt; ..., result.word1, result.word2</PRE>
  *
- * @version $Id: D2L.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class D2L extends ConversionInstruction {
 
     /** Convert double to long
      */
