diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP2.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP2.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP2.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP2.java
@@ -23,11 +23,10 @@
 
 /**
  * DUP2 - Duplicate two top operand stack words
  * <PRE>Stack: ..., word2, word1 -&gt; ..., word2, word1, word2, word1</PRE>
  *
- * @version $Id: DUP2.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class DUP2 extends StackInstruction implements PushInstruction {
 
     public DUP2() {
         super(com.sun.org.apache.bcel.internal.Const.DUP2);
