diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP_X2.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP_X2.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP_X2.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/DUP_X2.java
@@ -23,11 +23,10 @@
 
 /**
  * DUP_X2 - Duplicate top operand stack word and put three down
  * <PRE>Stack: ..., word3, word2, word1 -&gt; ..., word1, word3, word2, word1</PRE>
  *
- * @version $Id: DUP_X2.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class DUP_X2 extends StackInstruction {
 
     public DUP_X2() {
         super(com.sun.org.apache.bcel.internal.Const.DUP_X2);
