diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/F2I.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/F2I.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/F2I.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/F2I.java
@@ -23,11 +23,10 @@
 
 /**
  * F2I - Convert float to int
  * <PRE>Stack: ..., value -&gt; ..., result</PRE>
  *
- * @version $Id: F2I.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class F2I extends ConversionInstruction {
 
     /** Convert float to int
      */
