diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/FDIV.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/FDIV.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/FDIV.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/FDIV.java
@@ -23,11 +23,10 @@
 
 /**
  * FDIV - Divide floats
  * <PRE>Stack: ..., value1, value2 -&gt; result</PRE>
  *
- * @version $Id: FDIV.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class FDIV extends ArithmeticInstruction {
 
     /** Divide floats
      */
