diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/I2L.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/I2L.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/I2L.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/I2L.java
@@ -23,11 +23,10 @@
 
 /**
  * I2L - Convert int to long
  * <PRE>Stack: ..., value -&gt; ..., result.word1, result.word2</PRE>
  *
- * @version $Id: I2L.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class I2L extends ConversionInstruction {
 
     /** Convert int to long
      */
