diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/IUSHR.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/IUSHR.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/IUSHR.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/IUSHR.java
@@ -23,11 +23,10 @@
 
 /**
  * IUSHR - Logical shift right int
  * <PRE>Stack: ..., value1, value2 -&gt; ..., result</PRE>
  *
- * @version $Id: IUSHR.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class IUSHR extends ArithmeticInstruction {
 
     public IUSHR() {
         super(com.sun.org.apache.bcel.internal.Const.IUSHR);
