diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/LSHR.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/LSHR.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/LSHR.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/LSHR.java
@@ -23,11 +23,10 @@
 
 /**
  * LSHR - Arithmetic shift right long
  * <PRE>Stack: ..., value1.word1, value1.word2, value2 -&gt; ..., result.word1, result.word2</PRE>
  *
- * @version $Id: LSHR.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class LSHR extends ArithmeticInstruction {
 
     public LSHR() {
         super(com.sun.org.apache.bcel.internal.Const.LSHR);
