diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/POP2.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/POP2.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/POP2.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/POP2.java
@@ -24,11 +24,10 @@
 /**
  * POP2 - Pop two top operand stack words
  *
  * <PRE>Stack: ..., word2, word1 -&gt; ...</PRE>
  *
- * @version $Id: POP2.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public class POP2 extends StackInstruction implements PopInstruction {
 
     public POP2() {
         super(com.sun.org.apache.bcel.internal.Const.POP2);
