diff a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/Visitor.java b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/Visitor.java
--- a/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/Visitor.java
+++ b/src/java.xml/share/classes/com/sun/org/apache/bcel/internal/generic/Visitor.java
@@ -25,11 +25,10 @@
  * Interface implementing the Visitor pattern programming style.
  * I.e., a class that implements this interface can handle all types of
  * instructions with the properly typed methods just by calling the accept()
  * method.
  *
- * @version $Id: Visitor.java 1747278 2016-06-07 17:28:43Z britter $
  */
 public interface Visitor {
 
     void visitStackInstruction( StackInstruction obj );
 
