diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotBuilder.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotBuilder.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotBuilder.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/NotifyBotBuilder.java
@@ -33,11 +33,11 @@
     private HostedRepository repository;
     private Path storagePath;
     private Pattern branches;
     private StorageBuilder<UpdatedTag> tagStorageBuilder;
     private StorageBuilder<UpdatedBranch> branchStorageBuilder;
-    private StorageBuilder<PullRequestIssues> prIssuesStorageBuilder;
+    private StorageBuilder<PullRequestState> prStateStorageBuilder;
     private List<RepositoryUpdateConsumer> updaters = List.of();
     private List<PullRequestUpdateConsumer> prUpdaters = List.of();
     private Set<String> readyLabels = Set.of();
     private Map<String, Pattern> readyComments = Map.of();
 
@@ -64,12 +64,12 @@
     public NotifyBotBuilder branchStorageBuilder(StorageBuilder<UpdatedBranch> branchStorageBuilder) {
         this.branchStorageBuilder = branchStorageBuilder;
         return this;
     }
 
-    public NotifyBotBuilder prIssuesStorageBuilder(StorageBuilder<PullRequestIssues> prIssuesStorageBuilder) {
-        this.prIssuesStorageBuilder = prIssuesStorageBuilder;
+    public NotifyBotBuilder prStateStorageBuilder(StorageBuilder<PullRequestState> prStateStorageBuilder) {
+        this.prStateStorageBuilder = prStateStorageBuilder;
         return this;
     }
 
     public NotifyBotBuilder updaters(List<RepositoryUpdateConsumer> updaters) {
         this.updaters = updaters;
@@ -90,8 +90,8 @@
         this.readyComments = readyComments;
         return this;
     }
 
     public NotifyBot build() {
-        return new NotifyBot(repository, storagePath, branches, tagStorageBuilder, branchStorageBuilder, prIssuesStorageBuilder, updaters, prUpdaters, readyLabels, readyComments);
+        return new NotifyBot(repository, storagePath, branches, tagStorageBuilder, branchStorageBuilder, prStateStorageBuilder, updaters, prUpdaters, readyLabels, readyComments);
     }
-}
+}
