diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSNodeListCustom.cpp b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSNodeListCustom.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSNodeListCustom.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSNodeListCustom.cpp
@@ -72,11 +72,11 @@
     // https://bugs.webkit.org/show_bug.cgi?id=142595
     globalObject.vm().heap.deprecatedReportExtraMemory(nodeList->memoryCost());
     return createWrapper<NodeList>(&globalObject, WTFMove(nodeList));
 }
 
-JSC::JSValue toJSNewlyCreated(ExecState*, JSDOMGlobalObject* globalObject, Ref<NodeList>&& nodeList)
+JSC::JSValue toJSNewlyCreated(JSGlobalObject*, JSDOMGlobalObject* globalObject, Ref<NodeList>&& nodeList)
 {
     return createWrapper(*globalObject, WTFMove(nodeList));
 }
 
 } // namespace WebCore
