diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSPluginElementFunctions.h b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSPluginElementFunctions.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSPluginElementFunctions.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSPluginElementFunctions.h
@@ -33,12 +33,12 @@
     class JSHTMLElement;
 
     // JavaScript access to plug-in-exported properties for JSHTMLAppletElement, JSHTMLEmbedElement and JSHTMLObjectElement.
 
     JSC::Bindings::Instance* pluginInstance(HTMLElement&);
-    WEBCORE_EXPORT JSC::JSObject* pluginScriptObject(JSC::ExecState*, JSHTMLElement*);
+    WEBCORE_EXPORT JSC::JSObject* pluginScriptObject(JSC::JSGlobalObject*, JSHTMLElement*);
 
-    bool pluginElementCustomGetOwnPropertySlot(JSHTMLElement*, JSC::ExecState*, JSC::PropertyName, JSC::PropertySlot&);
-    bool pluginElementCustomPut(JSHTMLElement*, JSC::ExecState*, JSC::PropertyName, JSC::JSValue, JSC::PutPropertySlot&, bool& putResult);
+    bool pluginElementCustomGetOwnPropertySlot(JSHTMLElement*, JSC::JSGlobalObject*, JSC::PropertyName, JSC::PropertySlot&);
+    bool pluginElementCustomPut(JSHTMLElement*, JSC::JSGlobalObject*, JSC::PropertyName, JSC::JSValue, JSC::PutPropertySlot&, bool& putResult);
     JSC::CallType pluginElementCustomGetCallData(JSHTMLElement*, JSC::CallData&);
 
 } // namespace WebCore
