diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/BigIntObject.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/BigIntObject.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/BigIntObject.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/BigIntObject.cpp
@@ -54,16 +54,16 @@
     Base::finishCreation(vm);
     ASSERT(inherits(vm, info()));
     setInternalValue(vm, bigInt);
 }
 
-String BigIntObject::toStringName(const JSObject*, ExecState*)
+String BigIntObject::toStringName(const JSObject*, JSGlobalObject*)
 {
     return "Object"_s;
 }
 
-JSValue BigIntObject::defaultValue(const JSObject* object, ExecState*, PreferredPrimitiveType)
+JSValue BigIntObject::defaultValue(const JSObject* object, JSGlobalObject*, PreferredPrimitiveType)
 {
     const BigIntObject* bigIntObject = jsCast<const BigIntObject*>(object);
     return bigIntObject->internalValue();
 }
 
