diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CachePayload.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CachePayload.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CachePayload.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/CachePayload.cpp
@@ -35,11 +35,11 @@
 CachePayload CachePayload::makeMappedPayload(FileSystem::MappedFileData&& data)
 {
     return CachePayload(true, data.leakHandle(), data.size());
 }
 
-CachePayload CachePayload::makeMallocPayload(MallocPtr<uint8_t>&& data, size_t size)
+CachePayload CachePayload::makeMallocPayload(MallocPtr<uint8_t, VMMalloc>&& data, size_t size)
 {
     return CachePayload(false, data.leakPtr(), size);
 }
 
 CachePayload CachePayload::makeEmptyPayload()
