diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConfigFile.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConfigFile.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConfigFile.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/ConfigFile.cpp
@@ -463,11 +463,11 @@
     if (parseResult == ParseOK) {
         if (strlen(logPathname))
             WTF::setDataFile(logPathname);
 
         if (!jscOptionsBuilder.isEmpty()) {
-            Options::enableRestrictedOptions(true);
+            JSC::Config::enableRestrictedOptions();
             Options::setOptions(jscOptionsBuilder.toString().utf8().data());
         }
     } else
         WTF::dataLogF("Error in JSC Config file on or near line %u, parsing '%s'\n", scanner.lineNumber(), scanner.currentBuffer());
 }
