diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionPrototype.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionPrototype.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionPrototype.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/GeneratorFunctionPrototype.cpp
@@ -50,9 +50,9 @@
 void GeneratorFunctionPrototype::finishCreation(VM& vm)
 {
     Base::finishCreation(vm);
     ASSERT(inherits(vm, info()));
     putDirectWithoutTransition(vm, vm.propertyNames->length, jsNumber(0), PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum);
-    putDirectWithoutTransition(vm, vm.propertyNames->toStringTagSymbol, jsString(vm, "GeneratorFunction"), PropertyAttribute::DontEnum | PropertyAttribute::ReadOnly);
+    putDirectWithoutTransition(vm, vm.propertyNames->toStringTagSymbol, jsNontrivialString(vm, "GeneratorFunction"_s), PropertyAttribute::DontEnum | PropertyAttribute::ReadOnly);
 }
 
 } // namespace JSC
