diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Intrinsic.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Intrinsic.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Intrinsic.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/runtime/Intrinsic.cpp
@@ -77,14 +77,58 @@
         return "ArrayPushIntrinsic";
     case ArrayPopIntrinsic:
         return "ArrayPopIntrinsic";
     case ArraySliceIntrinsic:
         return "ArraySliceIntrinsic";
+    case ArrayValuesIntrinsic:
+        return "ArrayValuesIntrinsic";
+    case ArrayKeysIntrinsic:
+        return "ArrayKeysIntrinsic";
+    case ArrayEntriesIntrinsic:
+        return "ArrayEntriesIntrinsic";
     case CharCodeAtIntrinsic:
         return "CharCodeAtIntrinsic";
     case CharAtIntrinsic:
         return "CharAtIntrinsic";
+    case DatePrototypeGetTimeIntrinsic:
+        return "DatePrototypeGetTimeIntrinsic";
+    case DatePrototypeGetFullYearIntrinsic:
+        return "DatePrototypeGetFullYearIntrinsic";
+    case DatePrototypeGetUTCFullYearIntrinsic:
+        return "DatePrototypeGetUTCFullYearIntrinsic";
+    case DatePrototypeGetMonthIntrinsic:
+        return "DatePrototypeGetMonthIntrinsic";
+    case DatePrototypeGetUTCMonthIntrinsic:
+        return "DatePrototypeGetUTCMonthIntrinsic";
+    case DatePrototypeGetDateIntrinsic:
+        return "DatePrototypeGetDateIntrinsic";
+    case DatePrototypeGetUTCDateIntrinsic:
+        return "DatePrototypeGetUTCDateIntrinsic";
+    case DatePrototypeGetDayIntrinsic:
+        return "DatePrototypeGetDayIntrinsic";
+    case DatePrototypeGetUTCDayIntrinsic:
+        return "DatePrototypeGetUTCDayIntrinsic";
+    case DatePrototypeGetHoursIntrinsic:
+        return "DatePrototypeGetHoursIntrinsic";
+    case DatePrototypeGetUTCHoursIntrinsic:
+        return "DatePrototypeGetUTCHoursIntrinsic";
+    case DatePrototypeGetMinutesIntrinsic:
+        return "DatePrototypeGetMinutesIntrinsic";
+    case DatePrototypeGetUTCMinutesIntrinsic:
+        return "DatePrototypeGetUTCMinutesIntrinsic";
+    case DatePrototypeGetSecondsIntrinsic:
+        return "DatePrototypeGetSecondsIntrinsic";
+    case DatePrototypeGetUTCSecondsIntrinsic:
+        return "DatePrototypeGetUTCSecondsIntrinsic";
+    case DatePrototypeGetMillisecondsIntrinsic:
+        return "DatePrototypeGetMillisecondsIntrinsic";
+    case DatePrototypeGetUTCMillisecondsIntrinsic:
+        return "DatePrototypeGetUTCMillisecondsIntrinsic";
+    case DatePrototypeGetTimezoneOffsetIntrinsic:
+        return "DatePrototypeGetTimezoneOffsetIntrinsic";
+    case DatePrototypeGetYearIntrinsic:
+        return "DatePrototypeGetYearIntrinsic";
     case FromCharCodeIntrinsic:
         return "FromCharCodeIntrinsic";
     case PowIntrinsic:
         return "PowIntrinsic";
     case FloorIntrinsic:
@@ -121,10 +165,12 @@
         return "ObjectIsIntrinsic";
     case ObjectKeysIntrinsic:
         return "ObjectKeysIntrinsic";
     case ReflectGetPrototypeOfIntrinsic:
         return "ReflectGetPrototypeOfIntrinsic";
+    case StringPrototypeCodePointAtIntrinsic:
+        return "StringPrototypeCodePointAtIntrinsic";
     case StringPrototypeValueOfIntrinsic:
         return "StringPrototypeValueOfIntrinsic";
     case StringPrototypeReplaceIntrinsic:
         return "StringPrototypeReplaceIntrinsic";
     case StringPrototypeReplaceRegExpIntrinsic:
@@ -143,14 +189,20 @@
         return "RandomIntrinsic";
     case FRoundIntrinsic:
         return "FRoundIntrinsic";
     case TruncIntrinsic:
         return "TruncIntrinsic";
+    case TypedArrayValuesIntrinsic:
+        return "TypedArrayValuesIntrinsic";
+    case TypedArrayKeysIntrinsic:
+        return "TypedArrayKeysIntrinsic";
+    case TypedArrayEntriesIntrinsic:
+        return "TypedArrayEntriesIntrinsic";
     case IsTypedArrayViewIntrinsic:
         return "IsTypedArrayViewIntrinsic";
-    case BoundThisNoArgsFunctionCallIntrinsic:
-        return "BoundThisNoArgsFunctionCallIntrinsic";
+    case BoundFunctionCallIntrinsic:
+        return "BoundFunctionCallIntrinsic";
     case JSMapGetIntrinsic:
         return "JSMapGetIntrinsic";
     case JSMapHasIntrinsic:
         return "JSMapHasIntrinsic";
     case JSMapSetIntrinsic:
