diff a/src/hotspot/share/gc/z/zMarkCache.inline.hpp b/src/hotspot/share/gc/z/zMarkCache.inline.hpp
--- a/src/hotspot/share/gc/z/zMarkCache.inline.hpp
+++ b/src/hotspot/share/gc/z/zMarkCache.inline.hpp
@@ -42,11 +42,11 @@
 }
 
 inline void ZMarkCacheEntry::evict() {
   if (_page != NULL) {
     // Write cached data out to page
-    _page->inc_live_atomic(_objects, _bytes);
+    _page->inc_live(_objects, _bytes);
     _page = NULL;
   }
 }
 
 inline void ZMarkCache::inc_live(ZPage* page, size_t bytes) {
