diff a/src/java.base/share/classes/sun/security/rsa/PSSParameters.java b/src/java.base/share/classes/sun/security/rsa/PSSParameters.java
--- a/src/java.base/share/classes/sun/security/rsa/PSSParameters.java
+++ b/src/java.base/share/classes/sun/security/rsa/PSSParameters.java
@@ -241,11 +241,11 @@
         }
 
         if (!mgfDigestId.getOID().equals(AlgorithmId.SHA_oid)) {
             tmp2 = new DerOutputStream();
             tmp2.putOID(AlgorithmId.MGF1_oid);
-            mgfDigestId.encode(tmp2);
+            mgfDigestId.derEncode(tmp2);
             tmp3 = new DerOutputStream();
             tmp3.write(DerValue.tag_Sequence, tmp2);
             tmp.write(DerValue.createTag(DerValue.TAG_CONTEXT, true, (byte) 1),
                     tmp3);
         }
