diff a/src/jdk.incubator.jpackage/share/native/applauncher/AppLauncher.h b/src/jdk.incubator.jpackage/share/native/applauncher/AppLauncher.h
--- a/src/jdk.incubator.jpackage/share/native/applauncher/AppLauncher.h
+++ b/src/jdk.incubator.jpackage/share/native/applauncher/AppLauncher.h
@@ -49,10 +49,15 @@
     AppLauncher& setAppDir(const tstring& v) {
         appDirPath = v;
         return *this;
     }
 
+    AppLauncher& setInitJvmFromCmdlineOnly(bool v) {
+        initJvmFromCmdlineOnly = v;
+        return *this;
+    }
+
     AppLauncher& addJvmLibName(const tstring& v) {
         jvmLibNames.push_back(v);
         return *this;
     }
 
@@ -76,8 +81,9 @@
     tstring launcherPath;
     tstring defaultRuntimePath;
     tstring appDirPath;
     tstring imageRoot;
     tstring_array jvmLibNames;
+    bool initJvmFromCmdlineOnly;
 };
 
 #endif // AppLauncher_h
