diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/Metal/WHLSLFunctionWriter.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/Metal/WHLSLFunctionWriter.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/Metal/WHLSLFunctionWriter.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/Metal/WHLSLFunctionWriter.h
@@ -25,21 +25,33 @@
 
 #pragma once
 
 #if ENABLE(WEBGPU)
 
+#include "WHLSLMangledNames.h"
+#include "WHLSLSemanticMatcher.h"
+
 namespace WebCore {
 
 namespace WHLSL {
 
 class Program;
 
 namespace Metal {
 
 class TypeNamer;
 
-String metalFunctions(Program&, TypeNamer&);
+struct RenderMetalFunctionEntryPoints {
+    MangledFunctionName mangledVertexEntryPointName;
+    MangledFunctionName mangledFragmentEntryPointName;
+};
+RenderMetalFunctionEntryPoints emitMetalFunctions(StringBuilder&, Program&, TypeNamer&, MatchedRenderSemantics&&, Layout&);
+
+struct ComputeMetalFunctionEntryPoints {
+    MangledFunctionName mangledEntryPointName;
+};
+ComputeMetalFunctionEntryPoints emitMetalFunctions(StringBuilder&, Program&, TypeNamer&, MatchedComputeSemantics&&, Layout&);
 
 }
 
 }
 
