diff a/src/hotspot/share/gc/shenandoah/shenandoahHeapRegionCounters.cpp b/src/hotspot/share/gc/shenandoah/shenandoahHeapRegionCounters.cpp
--- a/src/hotspot/share/gc/shenandoah/shenandoahHeapRegionCounters.cpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahHeapRegionCounters.cpp
@@ -22,11 +22,11 @@
  *
  */
 
 #include "precompiled.hpp"
 #include "gc/shenandoah/shenandoahHeap.inline.hpp"
-#include "gc/shenandoah/shenandoahHeapRegion.hpp"
+#include "gc/shenandoah/shenandoahHeapRegion.inline.hpp"
 #include "gc/shenandoah/shenandoahHeapRegionSet.hpp"
 #include "gc/shenandoah/shenandoahHeapRegionCounters.hpp"
 #include "memory/resourceArea.hpp"
 #include "runtime/atomic.hpp"
 #include "runtime/perfData.inline.hpp"
@@ -83,11 +83,10 @@
       ShenandoahHeap* heap = ShenandoahHeap::heap();
       jlong status = 0;
       if (heap->is_concurrent_mark_in_progress())      status |= 1 << 0;
       if (heap->is_evacuation_in_progress())           status |= 1 << 1;
       if (heap->is_update_refs_in_progress())          status |= 1 << 2;
-      if (heap->is_concurrent_traversal_in_progress()) status |= 1 << 3;
       _status->set_value(status);
 
       _timestamp->set_value(os::elapsed_counter());
 
       size_t num_regions = heap->num_regions();
