diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/RemoteCommandListener.cpp b/modules/javafx.web/src/main/native/Source/WebCore/platform/RemoteCommandListener.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/RemoteCommandListener.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/RemoteCommandListener.cpp
@@ -29,10 +29,10 @@
 namespace WebCore {
 
 #if (!PLATFORM(IOS_FAMILY) && !PLATFORM(MAC)) || !HAVE(MEDIA_PLAYER)
 std::unique_ptr<RemoteCommandListener> RemoteCommandListener::create(RemoteCommandListenerClient& client)
 {
-    return std::make_unique<RemoteCommandListener>(client);
+    return makeUnique<RemoteCommandListener>(client);
 }
 #endif
 
 }
