diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/RuntimeApplicationChecks.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/RuntimeApplicationChecks.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/RuntimeApplicationChecks.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/RuntimeApplicationChecks.h
@@ -41,11 +41,11 @@
 #if PLATFORM(COCOA)
 
 bool isInWebProcess();
 
 WEBCORE_EXPORT void setApplicationSDKVersion(uint32_t);
-uint32_t applicationSDKVersion();
+WEBCORE_EXPORT uint32_t applicationSDKVersion();
 
 WEBCORE_EXPORT void setApplicationBundleIdentifier(const String&);
 String applicationBundleIdentifier();
 
 #if PLATFORM(MAC)
@@ -75,11 +75,13 @@
 #if PLATFORM(IOS_FAMILY)
 
 namespace IOSApplication {
 
 WEBCORE_EXPORT bool isMobileMail();
+WEBCORE_EXPORT bool isMailCompositionService();
 WEBCORE_EXPORT bool isMobileSafari();
+WEBCORE_EXPORT bool isIMDb();
 WEBCORE_EXPORT bool isWebBookmarksD();
 WEBCORE_EXPORT bool isDumpRenderTree();
 bool isMobileStore();
 bool isSpringBoard();
 WEBCORE_EXPORT bool isWebApp();
@@ -89,10 +91,12 @@
 WEBCORE_EXPORT bool isTheSecretSocietyHiddenMystery();
 WEBCORE_EXPORT bool isCardiogram();
 WEBCORE_EXPORT bool isNike();
 bool isMoviStarPlus();
 WEBCORE_EXPORT bool isFirefox();
+WEBCORE_EXPORT bool isAppleApplication();
+WEBCORE_EXPORT bool isEvernote();
 
 } // IOSApplication
 
 #endif // PLATFORM(IOS_FAMILY)
 
