diff a/test/jdk/java/awt/SplashScreen/MultiResolutionSplash/MultiResolutionSplashTest.java b/test/jdk/java/awt/SplashScreen/MultiResolutionSplash/MultiResolutionSplashTest.java
--- a/test/jdk/java/awt/SplashScreen/MultiResolutionSplash/MultiResolutionSplashTest.java
+++ b/test/jdk/java/awt/SplashScreen/MultiResolutionSplash/MultiResolutionSplashTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2014, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -138,11 +138,12 @@
         robot.waitForIdle();
 
         frame.dispose();
 
         if (!textField.getText().equals("ab")) {
-            throw new RuntimeException("Focus is lost!");
+            throw new RuntimeException("Focus is lost! " +
+                "Expected 'ab' got " + "'" + textField.getText() + "'.");
         }
     }
 
     static boolean compare(Color c1, Color c2) {
         return compare(c1.getRed(), c2.getRed())
