diff a/test/jdk/javax/swing/JMenu/4515762/bug4515762.java b/test/jdk/javax/swing/JMenu/4515762/bug4515762.java
--- a/test/jdk/javax/swing/JMenu/4515762/bug4515762.java
+++ b/test/jdk/javax/swing/JMenu/4515762/bug4515762.java
@@ -119,10 +119,11 @@
                     frame = new JFrame("Test");
                     frame.setJMenuBar(createMenuBar());
                     frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
                     frame.pack();
                     frame.setVisible(true);
+                    frame.toFront();
                 }
             });
 
             robot.waitForIdle();
 
@@ -138,35 +139,42 @@
 
             // Unique menu items.
             actionExpected = true;
             Util.hitMnemonics(robot, KeyEvent.VK_U);
 
+            robot.waitForIdle();
             robot.keyPress(KeyEvent.VK_S);
             robot.keyRelease(KeyEvent.VK_S);
             robot.waitForIdle();
 
             checkAction();
 
             Util.hitMnemonics(robot, KeyEvent.VK_U);
+            robot.waitForIdle();
+
             robot.keyPress(KeyEvent.VK_M);
             robot.keyRelease(KeyEvent.VK_M);
             robot.waitForIdle();
 
             checkAction();
 
             Util.hitMnemonics(robot, KeyEvent.VK_U);
+            robot.waitForIdle();
             Util.hitKeys(robot, KeyEvent.VK_T);
             robot.waitForIdle();
 
             checkAction();
+
             Util.hitMnemonics(robot, KeyEvent.VK_U);
+            robot.waitForIdle();
             Util.hitKeys(robot, KeyEvent.VK_W);
             robot.waitForIdle();
 
             checkAction();
 
             Util.hitMnemonics(robot, KeyEvent.VK_U);
+            robot.waitForIdle();
             Util.hitKeys(robot, KeyEvent.VK_U);
             robot.waitForIdle();
 
             checkAction();
         } finally {
