diff a/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/BackportsTests.java b/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/BackportsTests.java
--- a/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/BackportsTests.java
+++ b/bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/BackportsTests.java
@@ -173,11 +173,11 @@
                 resolvedInBuild = version.split("/", 2)[1];
                 version = version.split("/", 2)[0];
             }
             issue.setProperty("fixVersions", JSON.array().add(version));
             if (!resolvedInBuild.isEmpty()) {
-                issue.setProperty("customfield_10006", JSON.of(resolvedInBuild));
+                issue.setProperty("customfield_10006", JSON.object().put("value", resolvedInBuild));
             }
         }
 
         BackportManager(HostCredentials credentials, String initialVersion) {
             this.credentials = credentials;
