diff a/test/langtools/jdk/javadoc/doclet/testHelpFile/TestHelpFile.java b/test/langtools/jdk/javadoc/doclet/testHelpFile/TestHelpFile.java
--- a/test/langtools/jdk/javadoc/doclet/testHelpFile/TestHelpFile.java
+++ b/test/langtools/jdk/javadoc/doclet/testHelpFile/TestHelpFile.java
@@ -21,11 +21,11 @@
  * questions.
  */
 
 /*
  * @test
- * @bug      7132631
+ * @bug      7132631 8241693
  * @summary  Make sure that the help file is generated correctly.
  * @library  ../../lib
  * @modules jdk.javadoc/jdk.javadoc.internal.tool
  * @build    javadoc.tester.*
  * @run main TestHelpFile
@@ -47,7 +47,23 @@
                 testSrc("TestHelpFile.java"));
         checkExit(Exit.OK);
 
         checkOutput("help-doc.html", true,
             "<a href=\"constant-values.html\">Constant Field Values</a>");
+
+        // check a representative sample of the contents
+        checkOrder("help-doc.html",
+                "</div>\n"
+                + "<section class=\"help-section\">\n"
+                + "<h2>Package</h2>",
+                "<ul class=\"help-section-list\">\n"
+                + "<li>Interfaces</li>\n"
+                + "<li>Classes</li>\n"
+                + "<li>Enums</li>",
+                "</section>\n"
+                + "<section class=\"help-section\">\n"
+                + "<h2>Class or Interface</h2>",
+                "<ul class=\"help-section-list\">\n"
+                + "<li>Class Inheritance Diagram</li>\n"
+                + "<li>Direct Subclasses</li>\n");
     }
 }
