diff a/cli/src/main/java/org/openjdk/skara/cli/GitFork.java b/cli/src/main/java/org/openjdk/skara/cli/GitFork.java
--- a/cli/src/main/java/org/openjdk/skara/cli/GitFork.java
+++ b/cli/src/main/java/org/openjdk/skara/cli/GitFork.java
@@ -333,10 +333,11 @@
             System.out.println("Cloning " + cloneURI + "...");
 
             var cloneArgs = new ArrayList<String>();
             if (reference != null) {
                 cloneArgs.add("--reference-if-able=" + reference);
+                cloneArgs.add("--dissociate");
             }
             if (depth != null) {
                 cloneArgs.add("--depth=" + depth);
             }
             if (shallowSince != null) {
