diff a/modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuButtonSkinBase.java b/modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuButtonSkinBase.java
--- a/modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuButtonSkinBase.java
+++ b/modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuButtonSkinBase.java
@@ -216,12 +216,10 @@
      *                                                                         *
      **************************************************************************/
 
     /** {@inheritDoc} */
     @Override public void dispose() {
-        // FIXME : JDK-8244112 - backout if we are already disposed
-        // should check for getSkinnable to be null and return
         if (getSkinnable() == null) return;
 
         // Cleanup accelerators
         if (getSkinnable().getScene() != null) {
             ControlAcceleratorSupport.removeAcceleratorsFromScene(getSkinnable().getItems(), getSkinnable().getScene());
