diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.h
@@ -53,11 +53,11 @@
 
     void detachSession();
 
     unsigned long size();
     bool has(const BufferSource&);
-    JSC::JSValue get(JSC::ExecState&, const BufferSource&);
+    JSC::JSValue get(JSC::JSGlobalObject&, const BufferSource&);
 
     class Iterator {
     public:
         explicit Iterator(MediaKeyStatusMap&);
         Optional<WTF::KeyValuePair<BufferSource::VariantType, MediaKeyStatus>> next();
