diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeySystemAccess.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeySystemAccess.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeySystemAccess.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeySystemAccess.cpp
@@ -31,10 +31,11 @@
 
 #if ENABLE(ENCRYPTED_MEDIA)
 
 #include "CDM.h"
 #include "CDMInstance.h"
+#include "JSDOMPromiseDeferred.h"
 #include "JSMediaKeys.h"
 #include "MediaKeys.h"
 #include "MediaKeysRequirement.h"
 
 namespace WebCore {
