diff a/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIModule.cpp b/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIModule.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIModule.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/inspector/CommandLineAPIModule.cpp
@@ -51,16 +51,16 @@
 String CommandLineAPIModule::source() const
 {
     return StringImpl::createWithoutCopying(CommandLineAPIModuleSource_js, sizeof(CommandLineAPIModuleSource_js));
 }
 
-JSValue CommandLineAPIModule::host(InjectedScriptManager* injectedScriptManager, ExecState* exec) const
+JSValue CommandLineAPIModule::host(InjectedScriptManager* injectedScriptManager, JSGlobalObject* lexicalGlobalObject) const
 {
     // CommandLineAPIModule should only ever be used by a WebInjectedScriptManager.
     WebInjectedScriptManager* pageInjectedScriptManager = static_cast<WebInjectedScriptManager*>(injectedScriptManager);
     ASSERT(pageInjectedScriptManager->commandLineAPIHost());
 
-    JSDOMGlobalObject* globalObject = jsCast<JSDOMGlobalObject*>(exec->lexicalGlobalObject());
-    return pageInjectedScriptManager->commandLineAPIHost()->wrapper(exec, globalObject);
+    JSDOMGlobalObject* globalObject = jsCast<JSDOMGlobalObject*>(lexicalGlobalObject);
+    return pageInjectedScriptManager->commandLineAPIHost()->wrapper(lexicalGlobalObject, globalObject);
 }
 
 } // namespace WebCore
