diff a/src/java.base/share/classes/java/lang/module/ResolvedModule.java b/src/java.base/share/classes/java/lang/module/ResolvedModule.java
--- a/src/java.base/share/classes/java/lang/module/ResolvedModule.java
+++ b/src/java.base/share/classes/java/lang/module/ResolvedModule.java
@@ -77,11 +77,11 @@
      * }</pre>
      *
      * @return The module descriptor
      */
     ModuleDescriptor descriptor() {
-        return reference().descriptor();
+        return mref.descriptor();
     }
 
     /**
      * Returns the module name.
      *
@@ -91,11 +91,11 @@
      * }</pre>
      *
      * @return The module name
      */
     public String name() {
-        return reference().descriptor().name();
+        return mref.descriptor().name();
     }
 
     /**
      * Returns the set of resolved modules that this resolved module reads.
      *
