diff a/test/jdk/sun/security/x509/AVA/AVAEqualsHashCode.java b/test/jdk/sun/security/x509/AVA/AVAEqualsHashCode.java
--- a/test/jdk/sun/security/x509/AVA/AVAEqualsHashCode.java
+++ b/test/jdk/sun/security/x509/AVA/AVAEqualsHashCode.java
@@ -44,11 +44,11 @@
         String name = "CN=eve s. dropper";
         X500Name dn = new X500Name(name);
         DerOutputStream deros = new DerOutputStream();
         ObjectIdentifier oid = ObjectIdentifier.of("1.2.840.113549.2.5");
 
-        dn.encode(deros);
+        dn.derEncode(deros);
         byte[] ba = deros.toByteArray();
         DerValue dv = new DerValue(ba);
 
         GetAVAConstructor a = new GetAVAConstructor();
         java.security.AccessController.doPrivileged(a);
