diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/DOMCache.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/DOMCache.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/DOMCache.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/cache/DOMCache.idl
@@ -24,10 +24,11 @@
  */
 
 typedef (FetchRequest or USVString) RequestInfo;
 
 [
+    ActiveDOMObject,
     SecureContext,
     Exposed=(Window,Worker),
     EnabledAtRuntime=CacheAPI,
     InterfaceName=Cache,
 ] interface DOMCache {
