diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.cpp
@@ -37,14 +37,12 @@
 
 namespace AST {
 
 int64_t IntegerLiteral::valueForSelectedType() const
 {
-    ASSERT(m_type.resolvedType());
-    auto& typeReference = downcast<TypeReference>(*m_type.resolvedType());
-    ASSERT(typeReference.resolvedType());
-    auto& nativeTypeDeclaration = downcast<NativeTypeDeclaration>(*typeReference.resolvedType());
+    auto& typeReference = downcast<TypeReference>(m_type.resolvedType());
+    auto& nativeTypeDeclaration = downcast<NativeTypeDeclaration>(typeReference.resolvedType());
     return nativeTypeDeclaration.formatValueFromInteger()(m_value);
 }
 
 } // namespace AST
 
