diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLContinue.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLContinue.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLContinue.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLContinue.h
@@ -25,40 +25,42 @@
 
 #pragma once
 
 #if ENABLE(WEBGPU)
 
-#include "WHLSLLexer.h"
+#include "WHLSLCodeLocation.h"
 #include "WHLSLStatement.h"
+#include <wtf/FastMalloc.h>
 
 namespace WebCore {
 
 namespace WHLSL {
 
 namespace AST {
 
-class Continue : public Statement {
+class Continue final : public Statement {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
-    Continue(Lexer::Token&& origin)
-        : Statement(WTFMove(origin))
+    Continue(CodeLocation location)
+        : Statement(location, Kind::Continue)
     {
     }
 
-    virtual ~Continue() = default;
+    ~Continue() = default;
 
     Continue(const Continue&) = delete;
     Continue(Continue&&) = default;
 
-    bool isContinue() const override { return true; }
-
 private:
 };
 
 } // namespace AST
 
 }
 
 }
 
+DEFINE_DEFAULT_DELETE(Continue)
+
 SPECIALIZE_TYPE_TRAITS_WHLSL_STATEMENT(Continue, isContinue())
 
 #endif
