diff a/src/java.base/share/classes/java/lang/NegativeArraySizeException.java b/src/java.base/share/classes/java/lang/NegativeArraySizeException.java
--- a/src/java.base/share/classes/java/lang/NegativeArraySizeException.java
+++ b/src/java.base/share/classes/java/lang/NegativeArraySizeException.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1994, 2008, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1994, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -29,24 +29,24 @@
  * Thrown if an application tries to create an array with negative size.
  *
  * @author  unascribed
  * @since   1.0
  */
-public
-class NegativeArraySizeException extends RuntimeException {
+public class NegativeArraySizeException extends RuntimeException {
+    @java.io.Serial
     private static final long serialVersionUID = -8960118058596991861L;
 
     /**
-     * Constructs a <code>NegativeArraySizeException</code> with no
+     * Constructs a {@code NegativeArraySizeException} with no
      * detail message.
      */
     public NegativeArraySizeException() {
         super();
     }
 
     /**
-     * Constructs a <code>NegativeArraySizeException</code> with the
+     * Constructs a {@code NegativeArraySizeException} with the
      * specified detail message.
      *
      * @param   s   the detail message.
      */
     public NegativeArraySizeException(String s) {
