diff a/test/hotspot/jtreg/vmTestbase/vm/mlvm/share/MlvmTestExecutor.java b/test/hotspot/jtreg/vmTestbase/vm/mlvm/share/MlvmTestExecutor.java
--- a/test/hotspot/jtreg/vmTestbase/vm/mlvm/share/MlvmTestExecutor.java
+++ b/test/hotspot/jtreg/vmTestbase/vm/mlvm/share/MlvmTestExecutor.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2014, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -354,11 +354,10 @@
     }
 
     private static void setupMlvmTest(MlvmTest instance) {
         MlvmTest.setInstance(instance);
         OptionSupport.setup(instance, Env.getArgParser().getRawArguments(), new IgnoreUnknownArgumentsHandler());
-        instance.initRNG();
     }
 
     private static boolean runMlvmTestInstance(MlvmTest instance) throws Throwable {
         List<Class<? extends Throwable>> expectedExceptions = instance.getRequiredExceptions();
         int runsCount = instance.getRunsNumber() * instance.getStressOptions().getRunsFactor();
