diff a/modules/javafx.web/src/main/native/Source/WebCore/replay/UserInputBridge.h b/modules/javafx.web/src/main/native/Source/WebCore/replay/UserInputBridge.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/replay/UserInputBridge.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/replay/UserInputBridge.h
@@ -53,11 +53,11 @@
     WTF_MAKE_NONCOPYABLE(UserInputBridge);
 public:
     UserInputBridge(Page&);
 
     // User input APIs.
-#if ENABLE(CONTEXT_MENUS)
+#if ENABLE(CONTEXT_MENU_EVENT)
     WEBCORE_EXPORT bool handleContextMenuEvent(const PlatformMouseEvent&, Frame&, InputSource = InputSource::User);
 #endif
     WEBCORE_EXPORT bool handleMousePressEvent(const PlatformMouseEvent&, InputSource source = InputSource::User);
     WEBCORE_EXPORT bool handleMouseReleaseEvent(const PlatformMouseEvent&, InputSource source = InputSource::User);
     WEBCORE_EXPORT bool handleMouseMoveEvent(const PlatformMouseEvent&, InputSource source = InputSource::User);
