diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/BuiltinUtils.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/BuiltinUtils.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/BuiltinUtils.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/builtins/BuiltinUtils.h
@@ -25,10 +25,11 @@
  */
 
 #pragma once
 
 #include "ConstructAbility.h"
+#include "ConstructorKind.h"
 
 namespace JSC {
 
 #define INITIALIZE_BUILTIN_NAMES(name) , m_##name(JSC::Identifier::fromString(vm, #name)), m_##name##PrivateName(JSC::Identifier::fromUid(JSC::PrivateName(JSC::PrivateName::PrivateSymbol, #name ""_s)))
 #define DECLARE_BUILTIN_NAMES(name) const JSC::Identifier m_##name; const JSC::Identifier m_##name##PrivateName;
@@ -39,8 +40,8 @@
 class Identifier;
 class SourceCode;
 class UnlinkedFunctionExecutable;
 class VM;
 
-JS_EXPORT_PRIVATE UnlinkedFunctionExecutable* createBuiltinExecutable(VM&, const SourceCode&, const Identifier&, ConstructAbility);
+JS_EXPORT_PRIVATE UnlinkedFunctionExecutable* createBuiltinExecutable(VM&, const SourceCode&, const Identifier&, ConstructorKind, ConstructAbility);
 
 } // namespace JSC
