diff a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java
@@ -432,11 +432,11 @@
     public Symbol valueProjection() {
         return null;
     }
 
     /**
-     * Return the reference projection IFF 'this' happens to be value projection, null
+     * Return the reference projection IFF 'this' happens to be inline class, null
      * otherwise.
      */
     public Symbol referenceProjection() {
         return null;
     }
@@ -1669,11 +1669,14 @@
             return isReferenceProjection() ? projection : null;
         }
 
         @Override
         public ClassSymbol referenceProjection() {
-            if (!isValue() || projection != null)
+            if (!isValue())
+                return null;
+
+            if (projection != null)
                 return projection;
 
             ClassType ct = (ClassType) this.type;
             ClassType projectedType = new ClassType(ct.getEnclosingType(), ct.typarams_field, null);
             projectedType.allparams_field = ct.allparams_field;
diff a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java
@@ -237,18 +237,30 @@
 
     public boolean isValue() {
         return false;
     }
 
+    /**
+     * @return true IFF the receiver is a reference projection of an inline type and false
+     * for primitives or plain references
+     */
     public boolean isReferenceProjection() {
         return false;
     }
 
+    /**
+     * @return the value projection type IFF the receiver is a reference projection of an inline type
+     * and null otherwise
+     */
     public Type valueProjection() {
         return null;
     }
 
+    /**
+     * @return the reference projection type IFF the receiver is an inline type
+     * and null otherwise
+     */
     public Type referenceProjection() {
         return null;
     }
 
     /**
@@ -1192,11 +1204,14 @@
             return tsym != null && tsym.isReferenceProjection();
         }
 
         @Override
         public Type valueProjection() {
-            if (!isReferenceProjection() || projection !=  null)
+            if (!isReferenceProjection())
+                return null;
+
+            if (projection !=  null)
                 return projection;
 
             // Make a best case effort to cache the other projection.
             ClassSymbol valueClass = (ClassSymbol) tsym.valueProjection();
 
