diff a/issuetracker/src/test/java/org/openjdk/skara/issuetracker/IssueTrackerTests.java b/issuetracker/src/test/java/org/openjdk/skara/issuetracker/IssueTrackerTests.java
--- a/issuetracker/src/test/java/org/openjdk/skara/issuetracker/IssueTrackerTests.java
+++ b/issuetracker/src/test/java/org/openjdk/skara/issuetracker/IssueTrackerTests.java
@@ -20,10 +20,11 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
 package org.openjdk.skara.issuetracker;
 
+import org.openjdk.skara.json.JSON;
 import org.openjdk.skara.test.HostCredentials;
 
 import org.junit.jupiter.api.*;
 
 import java.io.IOException;
@@ -58,16 +59,16 @@
             issue.updateComment(comment.id(), "Now it is updated");
             issue.addLabel("label");
             issue.addLabel("another");
             issue.removeLabel("label");
             issue.setAssignees(List.of(project.issueTracker().currentUser()));
-            issue.addFixVersion("1.0");
-            issue.addFixVersion("2.0");
-            issue.removeFixVersion("1.0");
+            issue.setProperty("fixVersions", JSON.array().add("1.0"));
+            issue.setProperty("fixVersions", JSON.array().add("1.0").add("2.0"));
+            issue.setProperty("fixVersions", JSON.array().add("3.0"));
             var updated = project.issue(issue.id()).orElseThrow();
             assertEquals(List.of("another"), updated.labels());
-            assertEquals(List.of("2.0"), updated.fixVersions());
+            assertEquals(List.of(JSON.array().add("3.0")), updated.properties().get("fixVersions"));
             assertEquals(List.of(project.issueTracker().currentUser()), updated.assignees());
             assertEquals(1, updated.comments().size());
             assertEquals("Updated title", updated.title());
             assertEquals("Now it is updated", updated.comments().get(0).body());
 
