diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/CellContainer.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/CellContainer.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/CellContainer.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/CellContainer.cpp
@@ -30,12 +30,12 @@
 
 namespace JSC {
 
 bool CellContainer::isNewlyAllocated(HeapCell* cell) const
 {
-    if (isLargeAllocation())
-        return largeAllocation().isNewlyAllocated();
+    if (isPreciseAllocation())
+        return preciseAllocation().isNewlyAllocated();
     MarkedBlock& block = markedBlock();
     return !block.isNewlyAllocatedStale()
         && block.isNewlyAllocated(cell);
 }
 
