diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/Synchronousness.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/Synchronousness.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/Synchronousness.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/heap/Synchronousness.cpp
@@ -26,14 +26,14 @@
 #include "config.h"
 #include "Synchronousness.h"
 
 #include <wtf/PrintStream.h>
 
-using namespace JSC;
-
 namespace WTF {
 
+using namespace JSC;
+
 void printInternal(PrintStream& out, Synchronousness synchronousness)
 {
     switch (synchronousness) {
     case Async:
         out.print("Async");
