diff a/make/launcher/Launcher-jdk.incubator.jextract.gmk b/make/launcher/Launcher-jdk.incubator.jextract.gmk
--- a/make/launcher/Launcher-jdk.incubator.jextract.gmk
+++ b/make/launcher/Launcher-jdk.incubator.jextract.gmk
@@ -26,8 +26,8 @@
 include LauncherCommon.gmk
 
 $(eval $(call SetupBuildLauncher, jextract,\
     CFLAGS := -DENABLE_ARG_FILES, \
     MAIN_CLASS := jdk.incubator.jextract.tool.Main, \
-    JAVA_ARGS := -Djdk.incubator.foreign.permitUncheckedSegments=true, \
+    JAVA_ARGS := -Djdk.incubator.foreign.restrictedMethods=permit, \
 ))
 
diff a/test/jdk/java/jextract/SmokeTest.java b/test/jdk/java/jextract/SmokeTest.java
--- a/test/jdk/java/jextract/SmokeTest.java
+++ b/test/jdk/java/jextract/SmokeTest.java
@@ -25,11 +25,11 @@
  */
 
 /*
  * @test
  * @build JextractApiTestBase
- * @run testng/othervm -Djdk.incubator.foreign.permitUncheckedSegments=true SmokeTest
+ * @run testng/othervm -Djdk.incubator.foreign.restrictedMethods=permit SmokeTest
  */
 
 import jdk.incubator.jextract.Declaration;
 import jdk.incubator.jextract.Type;
 import org.testng.annotations.Test;
diff a/test/jdk/java/jextract/TestMacros.java b/test/jdk/java/jextract/TestMacros.java
--- a/test/jdk/java/jextract/TestMacros.java
+++ b/test/jdk/java/jextract/TestMacros.java
@@ -26,11 +26,11 @@
 
 /*
  * @test
  * @bug 8239128
  * @build JextractApiTestBase
- * @run testng/othervm -Djdk.incubator.foreign.permitUncheckedSegments=true TestMacros
+ * @run testng/othervm -Djdk.incubator.foreign.restrictedMethods=permit TestMacros
  */
 
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import jdk.incubator.foreign.MemoryLayouts;
diff a/test/jdk/tools/jextract/JextractToolProviderTest.java b/test/jdk/tools/jextract/JextractToolProviderTest.java
--- a/test/jdk/tools/jextract/JextractToolProviderTest.java
+++ b/test/jdk/tools/jextract/JextractToolProviderTest.java
@@ -36,11 +36,11 @@
 /*
  * @test
  * @modules jdk.incubator.jextract
  * @build JextractToolRunner
  * @run testng/othervm
- *          -Djdk.incubator.foreign.permitUncheckedSegments=true
+ *          -Djdk.incubator.foreign.restrictedMethods=permit
  *          -Duser.language=en
  *          --add-modules jdk.incubator.jextract
  *          JextractToolProviderTest
  */
 public class JextractToolProviderTest extends JextractToolRunner {
