diff a/test/hotspot/jtreg/runtime/valhalla/valuetypes/identityObject/InterfaceExtendingIdentityObject.java b/test/hotspot/jtreg/runtime/valhalla/valuetypes/identityObject/InterfaceExtendingIdentityObject.java
--- /dev/null
+++ b/test/hotspot/jtreg/runtime/valhalla/valuetypes/identityObject/InterfaceExtendingIdentityObject.java
@@ -0,0 +1,3 @@
+public interface InterfaceExtendingIdentityObject extends IdentityObject {
+
+}
