diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AnalyserNode.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AnalyserNode.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AnalyserNode.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AnalyserNode.h
@@ -28,10 +28,11 @@
 #include "RealtimeAnalyser.h"
 
 namespace WebCore {
 
 class AnalyserNode final : public AudioBasicInspectorNode {
+    WTF_MAKE_ISO_ALLOCATED(AnalyserNode);
 public:
     static Ref<AnalyserNode> create(AudioContext& context, float sampleRate)
     {
         return adoptRef(*new AnalyserNode(context, sampleRate));
     }
