diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AudioBasicInspectorNode.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AudioBasicInspectorNode.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AudioBasicInspectorNode.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/AudioBasicInspectorNode.cpp
@@ -28,18 +28,22 @@
 
 #include "AudioBasicInspectorNode.h"
 
 #include "AudioNodeInput.h"
 #include "AudioNodeOutput.h"
+#include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
+WTF_MAKE_ISO_ALLOCATED_IMPL(AudioBasicInspectorNode);
+
 AudioBasicInspectorNode::AudioBasicInspectorNode(AudioContext& context, float sampleRate, unsigned outputChannelCount)
     : AudioNode(context, sampleRate)
 {
-    addInput(std::make_unique<AudioNodeInput>(this));
-    addOutput(std::make_unique<AudioNodeOutput>(this, outputChannelCount));
+    setNodeType(NodeTypeBasicInspector);
+    addInput(makeUnique<AudioNodeInput>(this));
+    addOutput(makeUnique<AudioNodeOutput>(this, outputChannelCount));
 }
 
 // We override pullInputs() as an optimization allowing this node to take advantage of in-place processing,
 // where the input is simply passed through unprocessed to the output.
 // Note: this only applies if the input and output channel counts match.
