diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h
@@ -29,10 +29,11 @@
 namespace WebCore {
 
 class AudioDestination;
 
 class DefaultAudioDestinationNode final : public AudioDestinationNode {
+    WTF_MAKE_ISO_ALLOCATED(DefaultAudioDestinationNode);
 public:
     static Ref<DefaultAudioDestinationNode> create(AudioContext& context)
     {
         return adoptRef(*new DefaultAudioDestinationNode(context));
     }
