diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.h
@@ -36,10 +36,11 @@
 namespace WebCore {
 
 class AudioContext;
 
 class MediaElementAudioSourceNode final : public AudioNode, public AudioSourceProviderClient {
+    WTF_MAKE_ISO_ALLOCATED(MediaElementAudioSourceNode);
 public:
     static Ref<MediaElementAudioSourceNode> create(AudioContext&, HTMLMediaElement&);
 
     virtual ~MediaElementAudioSourceNode();
 
