diff a/src/hotspot/os/solaris/os_solaris.inline.hpp b/src/hotspot/os/solaris/os_solaris.inline.hpp
--- a/src/hotspot/os/solaris/os_solaris.inline.hpp
+++ b/src/hotspot/os/solaris/os_solaris.inline.hpp
@@ -47,16 +47,10 @@
   guarantee (r == 0 || r == 1, "CR6501650 or CR6493689") ;
   return r;
 }
 
 
-// On Solaris, reservations are made on a page by page basis, nothing to do.
-inline void os::pd_split_reserved_memory(char *base, size_t size,
-                                      size_t split, bool realloc) {
-}
-
-
 // Bang the shadow pages if they need to be touched to be mapped.
 inline void os::map_stack_shadow_pages(address sp) {
 }
 
 inline void os::dll_unload(void *lib) { ::dlclose(lib); }
