diff a/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java b/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
--- a/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
+++ b/vcs/src/main/java/org/openjdk/skara/vcs/hg/HgRepository.java
@@ -587,10 +587,20 @@
         }
 
         return new Branch(name);
     }
 
+    @Override
+    public void prune(Branch branch, String remote) throws IOException {
+        try (var p = capture("hg", "bookmark", "--delete", branch.name())) {
+            await(p);
+        }
+        try (var p = capture("hg", "push", "--bookmark", branch.name(), remote)) {
+            await(p);
+        }
+    }
+
     @Override
     public Hash mergeBase(Hash first, Hash second) throws IOException {
         var revset = "ancestor(" + first.hex() + ", " + second.hex() + ")";
         try (var p = capture("hg", "log", "--rev=" + revset, "--template={node}\n")) {
             var res = await(p);
