diff a/email/src/test/java/org/openjdk/skara/email/SMTPTests.java b/email/src/test/java/org/openjdk/skara/email/SMTPTests.java
--- a/email/src/test/java/org/openjdk/skara/email/SMTPTests.java
+++ b/email/src/test/java/org/openjdk/skara/email/SMTPTests.java
@@ -65,6 +65,23 @@
             SMTP.send(server.address(), recipient, sentMail);
             var email = server.receive(Duration.ofSeconds(10));
             assertEquals(sentMail, email);
         }
     }
+
+    @Test
+    void encoded() throws IOException {
+        log.info("Hello");
+        try (var server = new SMTPServer()) {
+            var sender = EmailAddress.from("Señor Dévèlöper", "test@test.email");
+            var recipient = EmailAddress.from("Dêst", "dest@dest.email");
+            var sentMail = Email.create(sender, "Sübject", "Bödÿ")
+                                .recipient(recipient)
+                                .header("Something", "Öthè®")
+                                .build();
+
+            SMTP.send(server.address(), recipient, sentMail);
+            var email = server.receive(Duration.ofSeconds(10));
+            assertEquals(sentMail, email);
+        }
+    }
 }
