diff a/agent/src/main/java/org/openjdk/jmc/agent/jmx/AgentManagementFactory.java b/agent/src/main/java/org/openjdk/jmc/agent/jmx/AgentManagementFactory.java
--- a/agent/src/main/java/org/openjdk/jmc/agent/jmx/AgentManagementFactory.java
+++ b/agent/src/main/java/org/openjdk/jmc/agent/jmx/AgentManagementFactory.java
@@ -48,11 +48,11 @@
 public final class AgentManagementFactory {
 	private static final String AGENT_OBJECT_NAME = "org.openjdk.jmc.jfr.agent:type=AgentController"; //$NON-NLS-1$
 
 	private static AgentController agentControllerMBean;
 
-	public static AgentControllerMBean getAgentControllerBean() {
+	public static AgentControllerMXBean getAgentControllerBean() {
 		return agentControllerMBean;
 	}
 
 	public static void createAndRegisterAgentControllerMBean(
 		Instrumentation instrumentation, TransformRegistry registry) {
