diff a/test/jdk/sun/security/pkcs/pkcs9/UnknownAttribute.java b/test/jdk/sun/security/pkcs/pkcs9/UnknownAttribute.java
--- a/test/jdk/sun/security/pkcs/pkcs9/UnknownAttribute.java
+++ b/test/jdk/sun/security/pkcs/pkcs9/UnknownAttribute.java
@@ -32,10 +32,11 @@
 
 import java.io.*;
 import java.util.Arrays;
 
 import sun.security.pkcs.PKCS9Attribute;
+import sun.security.util.DerOutputStream;
 import sun.security.util.DerValue;
 import sun.security.util.ObjectIdentifier;
 import jdk.test.lib.hexdump.HexPrinter;
 
 public class UnknownAttribute {
@@ -55,11 +56,11 @@
         };
         PKCS9Attribute p2 = new PKCS9Attribute(new DerValue(data));
         if (p2.isKnown()) {
             throw new Exception();
         }
-        ByteArrayOutputStream bout = new ByteArrayOutputStream();
+        DerOutputStream bout = new DerOutputStream();
         p2.derEncode(bout);
         HexPrinter.simple().dest(System.err).format(bout.toByteArray());
         if (!Arrays.equals(data, bout.toByteArray())) {
             throw new Exception();
         }
@@ -73,11 +74,11 @@
         PKCS9Attribute p3 = new PKCS9Attribute(
                 ObjectIdentifier.of("1.2.3"), new byte[]{0x31,0x02,0x05,0x00});
         if (p3.isKnown()) {
             throw new Exception();
         }
-        bout = new ByteArrayOutputStream();
+        bout = new DerOutputStream();
         p3.derEncode(bout);
         if (!Arrays.equals(data, bout.toByteArray())) {
             throw new Exception();
         }
     }
