diff a/agent/src/main/java/org/openjdk/jmc/agent/Parameter.java b/agent/src/main/java/org/openjdk/jmc/agent/Parameter.java
--- a/agent/src/main/java/org/openjdk/jmc/agent/Parameter.java
+++ b/agent/src/main/java/org/openjdk/jmc/agent/Parameter.java
@@ -30,10 +30,12 @@
  * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
  * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 package org.openjdk.jmc.agent;
 
+import javax.management.openmbean.CompositeData;
+
 import org.openjdk.jmc.agent.util.TypeUtils;
 
 /**
  * Metadata for a parameter to be logged by the agent.
  */
@@ -57,10 +59,15 @@
 		this.relationKey = relationKey;
 		this.converterClassName = converterClassName;
 		this.fieldName = "field" + TypeUtils.deriveIdentifierPart(name); //$NON-NLS-1$
 	}
 
+	public static Parameter from(CompositeData cd) {
+		return new Parameter((int) cd.get("index"), (String) cd.get("name"), (String) cd.get("description"),
+				(String) cd.get("contentType"), (String) cd.get("relationKey"), (String) cd.get("converterClassName"));
+	}
+
 	public int getIndex() {
 		return index;
 	}
 
 	public String getName() {
