diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSMediaStreamTrackCustom.cpp b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSMediaStreamTrackCustom.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSMediaStreamTrackCustom.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSMediaStreamTrackCustom.cpp
@@ -31,20 +31,20 @@
 #include "JSCanvasCaptureMediaStreamTrack.h"
 #include "JSDOMBinding.h"
 
 namespace WebCore {
 
-JSC::JSValue toJSNewlyCreated(JSC::ExecState*, JSDOMGlobalObject* globalObject, Ref<MediaStreamTrack>&& object)
+JSC::JSValue toJSNewlyCreated(JSC::JSGlobalObject*, JSDOMGlobalObject* globalObject, Ref<MediaStreamTrack>&& object)
 {
     if (is<CanvasCaptureMediaStreamTrack>(object))
         return createWrapper<CanvasCaptureMediaStreamTrack>(globalObject, WTFMove(object));
     return createWrapper<MediaStreamTrack>(globalObject, WTFMove(object));
 }
 
-JSC::JSValue toJS(JSC::ExecState* state, JSDOMGlobalObject* globalObject, MediaStreamTrack& track)
+JSC::JSValue toJS(JSC::JSGlobalObject* lexicalGlobalObject, JSDOMGlobalObject* globalObject, MediaStreamTrack& track)
 {
-    return wrap(state, globalObject, track);
+    return wrap(lexicalGlobalObject, globalObject, track);
 }
 
 }
 
 #endif
