diff a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSRemoteDOMWindowBase.cpp b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSRemoteDOMWindowBase.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSRemoteDOMWindowBase.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/bindings/js/JSRemoteDOMWindowBase.cpp
@@ -37,11 +37,11 @@
 
 const GlobalObjectMethodTable JSRemoteDOMWindowBase::s_globalObjectMethodTable = {
     nullptr, // shellSupportsRichSourceInfo
     nullptr, // shouldInterruptScript
     &javaScriptRuntimeFlags,
-    nullptr, // queueTaskToEventLoop
+    nullptr, // queueMicrotaskToEventLoop
     nullptr, // shouldInterruptScriptBeforeTimeout
     nullptr, // moduleLoaderImportModule
     nullptr, // moduleLoaderResolve
     nullptr, // moduleLoaderFetch
     nullptr, // moduleLoaderCreateImportMetaProperties
