diff a/src/java.base/share/classes/java/text/DateFormat.java b/src/java.base/share/classes/java/text/DateFormat.java
--- a/src/java.base/share/classes/java/text/DateFormat.java
+++ b/src/java.base/share/classes/java/text/DateFormat.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1996, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1996, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -363,14 +363,14 @@
      */
     public abstract StringBuffer format(Date date, StringBuffer toAppendTo,
                                         FieldPosition fieldPosition);
 
     /**
-      * Formats a {@link Date} into a date-time string.
-      *
-      * @param date the time value to be formatted into a date-time string.
-      * @return the formatted date-time string.
+     * Formats a {@link Date} into a date-time string.
+     *
+     * @param date the time value to be formatted into a date-time string.
+     * @return the formatted date-time string.
      */
     public final String format(Date date)
     {
         return format(date, new StringBuffer(),
                       DontCareFieldPosition.INSTANCE).toString();
