diff a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/HRTFPanner.h b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/HRTFPanner.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/HRTFPanner.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/platform/audio/HRTFPanner.h
@@ -25,15 +25,16 @@
 #ifndef HRTFPanner_h
 #define HRTFPanner_h
 
 #include "DelayDSPKernel.h"
 #include "FFTConvolver.h"
-#include "HRTFDatabaseLoader.h"
 #include "Panner.h"
 
 namespace WebCore {
 
+class HRTFDatabaseLoader;
+
 class HRTFPanner final : public Panner {
 public:
     explicit HRTFPanner(float sampleRate, HRTFDatabaseLoader*);
     virtual ~HRTFPanner();
 
