diff a/test/jdk/java/util/concurrent/tck/ConcurrentHashMapTest.java b/test/jdk/java/util/concurrent/tck/ConcurrentHashMapTest.java
--- a/test/jdk/java/util/concurrent/tck/ConcurrentHashMapTest.java
+++ b/test/jdk/java/util/concurrent/tck/ConcurrentHashMapTest.java
@@ -53,12 +53,10 @@
     }
     public static Test suite() {
         class Implementation implements MapImplementation {
             public Class<?> klazz() { return ConcurrentHashMap.class; }
             public Map emptyMap() { return new ConcurrentHashMap(); }
-            public Object makeKey(int i) { return i; }
-            public Object makeValue(int i) { return i; }
             public boolean isConcurrent() { return true; }
             public boolean permitsNullKeys() { return false; }
             public boolean permitsNullValues() { return false; }
             public boolean supportsSetValue() { return true; }
         }
