diff a/test/jdk/java/lang/reflect/AccessibleObject/CanAccessTest.java b/test/jdk/java/lang/reflect/AccessibleObject/CanAccessTest.java
--- a/test/jdk/java/lang/reflect/AccessibleObject/CanAccessTest.java
+++ b/test/jdk/java/lang/reflect/AccessibleObject/CanAccessTest.java
@@ -92,11 +92,11 @@
     /**
      * the specified object must be an instance of the declaring class
      * for instance members
      */
     public void testInstanceMethod() throws Exception {
-        Method m = Unsafe.class.getDeclaredMethod("addressSize0");
+        Method m = Unsafe.class.getDeclaredMethod("allocateMemory0", long.class);
         assertFalse(m.canAccess(INSTANCE));
 
         try {
             m.canAccess(null);
             assertTrue(false);
