diff a/src/jdk.compiler/share/classes/com/sun/source/tree/AnnotationTree.java b/src/jdk.compiler/share/classes/com/sun/source/tree/AnnotationTree.java
--- a/src/jdk.compiler/share/classes/com/sun/source/tree/AnnotationTree.java
+++ b/src/jdk.compiler/share/classes/com/sun/source/tree/AnnotationTree.java
@@ -34,11 +34,11 @@
  * <pre>
  *    {@code @}<em>annotationType</em>
  *    {@code @}<em>annotationType</em> ( <em>arguments</em> )
  * </pre>
  *
- * @jls section 9.7
+ * @jls 9.7 Annotations
  *
  * @author Peter von der Ah&eacute;
  * @author Jonathan Gibbons
  * @since 1.6
  */
