diff a/src/jdk.compiler/share/classes/com/sun/source/tree/ArrayTypeTree.java b/src/jdk.compiler/share/classes/com/sun/source/tree/ArrayTypeTree.java
--- a/src/jdk.compiler/share/classes/com/sun/source/tree/ArrayTypeTree.java
+++ b/src/jdk.compiler/share/classes/com/sun/source/tree/ArrayTypeTree.java
@@ -31,11 +31,11 @@
  * For example:
  * <pre>
  *   <em>type</em> []
  * </pre>
  *
- * @jls section 10.1
+ * @jls 10.1 Array Types
  *
  * @author Peter von der Ah&eacute;
  * @author Jonathan Gibbons
  * @since 1.6
  */
