diff a/src/jdk.compiler/share/classes/com/sun/source/tree/ForLoopTree.java b/src/jdk.compiler/share/classes/com/sun/source/tree/ForLoopTree.java
--- a/src/jdk.compiler/share/classes/com/sun/source/tree/ForLoopTree.java
+++ b/src/jdk.compiler/share/classes/com/sun/source/tree/ForLoopTree.java
@@ -34,11 +34,11 @@
  * <pre>
  *   for ( <em>initializer</em> ; <em>condition</em> ; <em>update</em> )
  *       <em>statement</em>
  * </pre>
  *
- * @jls section 14.14.1
+ * @jls 14.14.1 The basic for Statement
  *
  * @author Peter von der Ah&eacute;
  * @author Jonathan Gibbons
  * @since 1.6
  */
