diff a/src/jdk.compiler/share/classes/com/sun/source/tree/ParameterizedTypeTree.java b/src/jdk.compiler/share/classes/com/sun/source/tree/ParameterizedTypeTree.java
--- a/src/jdk.compiler/share/classes/com/sun/source/tree/ParameterizedTypeTree.java
+++ b/src/jdk.compiler/share/classes/com/sun/source/tree/ParameterizedTypeTree.java
@@ -33,11 +33,11 @@
  * For example:
  * <pre>
  *   <em>type</em> &lt; <em>typeArguments</em> &gt;
  * </pre>
  *
- * @jls section 4.5.1
+ * @jls 4.5.1 Type Arguments of Parameterized Types
  *
  * @author Peter von der Ah&eacute;
  * @author Jonathan Gibbons
  * @since 1.6
  */
