diff a/src/jdk.compiler/share/classes/com/sun/source/tree/SwitchTree.java b/src/jdk.compiler/share/classes/com/sun/source/tree/SwitchTree.java
--- a/src/jdk.compiler/share/classes/com/sun/source/tree/SwitchTree.java
+++ b/src/jdk.compiler/share/classes/com/sun/source/tree/SwitchTree.java
@@ -35,11 +35,11 @@
  *   switch ( <em>expression</em> ) {
  *     <em>cases</em>
  *   }
  * </pre>
  *
- * @jls section 14.11
+ * @jls 14.11 The switch Statement
  *
  * @author Peter von der Ah&eacute;
  * @author Jonathan Gibbons
  * @since 1.6
  */
