diff a/src/jdk.compiler/share/classes/com/sun/source/tree/ThrowTree.java b/src/jdk.compiler/share/classes/com/sun/source/tree/ThrowTree.java
--- a/src/jdk.compiler/share/classes/com/sun/source/tree/ThrowTree.java
+++ b/src/jdk.compiler/share/classes/com/sun/source/tree/ThrowTree.java
@@ -31,11 +31,11 @@
  * For example:
  * <pre>
  *   throw <em>expression</em>;
  * </pre>
  *
- * @jls section 14.18
+ * @jls 14.18 The throw Statement
  *
  * @author Peter von der Ah&eacute;
  * @author Jonathan Gibbons
  * @since 1.6
  */
