diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGTextPath.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGTextPath.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGTextPath.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGTextPath.h
@@ -22,19 +22,22 @@
 
 #include "RenderSVGInline.h"
 
 namespace WebCore {
 
+class SVGGeometryElement;
+
 class RenderSVGTextPath final : public RenderSVGInline {
     WTF_MAKE_ISO_ALLOCATED(RenderSVGTextPath);
 public:
     RenderSVGTextPath(SVGTextPathElement&, RenderStyle&&);
 
     SVGTextPathElement& textPathElement() const;
+    SVGGeometryElement* targetElement() const;
 
     Path layoutPath() const;
-    float startOffset() const;
+    const SVGLengthValue& startOffset() const;
     bool exactAlignment() const;
     bool stretchMethod() const;
 
 private:
     void graphicsElement() const = delete;
