diff a/application/uitests/org.openjdk.jmc.console.uitest/src/test/java/org/openjdk/jmc/console/uitest/mbeanhelpers/Testable.java b/application/uitests/org.openjdk.jmc.console.uitest/src/test/java/org/openjdk/jmc/console/uitest/mbeanhelpers/Testable.java
--- a/application/uitests/org.openjdk.jmc.console.uitest/src/test/java/org/openjdk/jmc/console/uitest/mbeanhelpers/Testable.java
+++ b/application/uitests/org.openjdk.jmc.console.uitest/src/test/java/org/openjdk/jmc/console/uitest/mbeanhelpers/Testable.java
@@ -825,13 +825,13 @@
 		while (!stop) {
 			synchronized (this) {
 				long oldLastTime = lastTime;
 				lastTime = System.currentTimeMillis();
 //				System.out.println("Current time: " + getAliveTime());
-				sendNotification(new AttributeChangeNotification(this, sequenceNumber++, lastTime, "Update",
-						"AliveTime", "long", Long.valueOf(oldLastTime - startTime),
-						Long.valueOf(lastTime - startTime)));
+				sendNotification(
+						new AttributeChangeNotification(this, sequenceNumber++, lastTime, "Update", "AliveTime", "long",
+								Long.valueOf(oldLastTime - startTime), Long.valueOf(lastTime - startTime)));
 				try {
 					wait(Math.max(1, updateTime));
 				} catch (InterruptedException e) {
 				}
 			}
