diff a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGLengthList.h b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGLengthList.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGLengthList.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/svg/SVGLengthList.h
@@ -25,19 +25,20 @@
 
 #pragma once
 
 #include "SVGLength.h"
 #include "SVGValuePropertyList.h"
+#include <wtf/text/StringBuilder.h>
 
 namespace WebCore {
 
 class SVGLengthList : public SVGValuePropertyList<SVGLength> {
     using Base = SVGValuePropertyList<SVGLength>;
     using Base::Base;
 
 public:
-    static Ref<SVGLengthList> create(SVGLengthMode lengthMode = LengthModeOther)
+    static Ref<SVGLengthList> create(SVGLengthMode lengthMode = SVGLengthMode::Other)
     {
         return adoptRef(*new SVGLengthList(lengthMode));
     }
 
     static Ref<SVGLengthList> create(SVGPropertyOwner* owner, SVGPropertyAccess access, SVGLengthMode lengthMode)
@@ -108,9 +109,9 @@
         : Base(other, access)
         , m_lengthMode(other.lengthMode())
     {
     }
 
-    SVGLengthMode m_lengthMode { LengthModeOther };
+    SVGLengthMode m_lengthMode { SVGLengthMode::Other };
 };
 
 } // namespace WebCore
