diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGText.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGText.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGText.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/svg/RenderSVGText.h
@@ -85,11 +85,10 @@
     const RenderObject* pushMappingToContainer(const RenderLayerModelObject* ancestorToStopAt, RenderGeometryMap&) const override;
     void willBeDestroyed() override;
 
     const AffineTransform& localToParentTransform() const override { return m_localTransform; }
     AffineTransform localTransform() const override { return m_localTransform; }
-    std::unique_ptr<RootInlineBox> createRootInlineBox() override;
 
     RenderBlock* firstLineBlock() const override;
 
     bool shouldHandleSubtreeMutations() const;
 
