diff a/make/modules/java.base/gensrc/GensrcLocaleData.gmk b/make/modules/java.base/gensrc/GensrcLocaleData.gmk
--- a/make/modules/java.base/gensrc/GensrcLocaleData.gmk
+++ b/make/modules/java.base/gensrc/GensrcLocaleData.gmk
@@ -1,7 +1,7 @@
 #
-# Copyright (c) 2011, 2019, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2011, 2020, Oracle and/or its affiliates. All rights reserved.
 # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 #
 # This code is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License version 2 only, as
 # published by the Free Software Foundation.  Oracle designates this
@@ -145,9 +145,9 @@
 	$(SED) $(SED_NONBASEARGS) $< > $@
 
 GENSRC_BASELOCALEDATA := $(SUPPORT_OUTPUTDIR)/gensrc/java.base/sun/util/locale/provider/BaseLocaleDataMetaInfo.java
 GENSRC_LOCALEDATA := $(SUPPORT_OUTPUTDIR)/gensrc/jdk.localedata/sun/util/resources/provider/NonBaseLocaleDataMetaInfo.java
 
-GENSRC_JAVA_BASE += $(GENSRC_BASELOCALEDATA)
-GENSRC_JDK_LOCALEDATA += $(GENSRC_LOCALEDATA)
-
 ################################################################################
+
+# This file is included twice, by java.base and jdk.localedata. The includer must
+# add either GENSRC_BASELOCALEDATA or GENSRC_LOCALEDATA to TARGETS.
