diff a/jcheck/src/test/java/org/openjdk/skara/jcheck/TestRepository.java b/jcheck/src/test/java/org/openjdk/skara/jcheck/TestRepository.java
--- a/jcheck/src/test/java/org/openjdk/skara/jcheck/TestRepository.java
+++ b/jcheck/src/test/java/org/openjdk/skara/jcheck/TestRepository.java
@@ -299,10 +299,15 @@
 
     public List<Submodule> submodules() throws IOException {
         return null;
     }
 
+    @Override
+    public Tree tree(Hash h) throws IOException {
+        return null;
+    }
+
     public Optional<Tag.Annotated> annotate(Tag tag) throws IOException {
         return null;
     }
 
     public String range(Hash h) {
