diff a/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java b/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java
--- a/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java
+++ b/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java
@@ -56,12 +56,12 @@
                                                  "-Xmx32m",
                                                  "-Xms32m",
                                                  "-version"});
 
     runTest("-XX:+UseG1GC");
-    runTest("-XX:+UseParallelOldGC -XX:-UseAdaptiveGCBoundary");
-    runTest("-XX:+UseParallelOldGC -XX:+UseAdaptiveGCBoundary");
+    runTest("-XX:+UseParallelGC -XX:-UseAdaptiveGCBoundary");
+    runTest("-XX:+UseParallelGC -XX:+UseAdaptiveGCBoundary");
   }
 
   private static void runTest(String... extraFlags) throws Exception {
     ArrayList<String> testOpts = new ArrayList<>();
     Collections.addAll(testOpts, commonOpts.toArray(new String[commonOpts.size()]));
