diff a/src/hotspot/cpu/x86/x86.ad b/src/hotspot/cpu/x86/x86.ad
--- a/src/hotspot/cpu/x86/x86.ad
+++ b/src/hotspot/cpu/x86/x86.ad
@@ -2339,13 +2339,15 @@
 
 // ============================================================================
 
 instruct ShouldNotReachHere() %{
   match(Halt);
-  format %{ "ud2\t# ShouldNotReachHere" %}
+  format %{ "stop\t# ShouldNotReachHere" %}
   ins_encode %{
-    __ stop(_halt_reason);
+    if (is_reachable()) {
+      __ stop(_halt_reason);
+    }
   %}
   ins_pipe(pipe_slow);
 %}
 
 // =================================EVEX special===============================
