diff a/jcheck/src/main/java/org/openjdk/skara/jcheck/ChecksConfiguration.java b/jcheck/src/main/java/org/openjdk/skara/jcheck/ChecksConfiguration.java
--- a/jcheck/src/main/java/org/openjdk/skara/jcheck/ChecksConfiguration.java
+++ b/jcheck/src/main/java/org/openjdk/skara/jcheck/ChecksConfiguration.java
@@ -33,31 +33,35 @@
         new ChecksConfiguration(List.of(),
                                 List.of(),
                                 WhitespaceConfiguration.DEFAULT,
                                 ReviewersConfiguration.DEFAULT,
                                 MergeConfiguration.DEFAULT,
-                                CommitterConfiguration.DEFAULT);
+                                CommitterConfiguration.DEFAULT,
+                                IssuesConfiguration.DEFAULT);
 
     private final List<String> error;
     private final List<String> warning;
     private final WhitespaceConfiguration whitespace;
     private final ReviewersConfiguration reviewers;
     private final MergeConfiguration merge;
     private final CommitterConfiguration committer;
+    private final IssuesConfiguration issues;
 
     ChecksConfiguration(List<String> error,
                         List<String> warning,
                         WhitespaceConfiguration whitespace,
                         ReviewersConfiguration reviewers,
                         MergeConfiguration merge,
-                        CommitterConfiguration committer) {
+                        CommitterConfiguration committer,
+                        IssuesConfiguration issues) {
         this.error = error;
         this.warning = warning;
         this.whitespace = whitespace;
         this.reviewers = reviewers;
         this.merge = merge;
+        this.issues = issues;
     }
 
     public List<String> error() {
         return error;
     }
@@ -98,10 +102,14 @@
 
     public CommitterConfiguration committer() {
         return committer;
     }
 
+    public IssuesConfiguration issues() {
+        return issues;
+    }
+
     static String name() {
         return "checks";
     }
 
     static ChecksConfiguration parse(Section s) {
@@ -114,9 +122,10 @@
 
         var whitespace = WhitespaceConfiguration.parse(s.subsection(WhitespaceConfiguration.name()));
         var reviewers = ReviewersConfiguration.parse(s.subsection(ReviewersConfiguration.name()));
         var merge = MergeConfiguration.parse(s.subsection(MergeConfiguration.name()));
         var committer = CommitterConfiguration.parse(s.subsection(CommitterConfiguration.name()));
+        var issues = IssuesConfiguration.parse(s.subsection(IssuesConfiguration.name()));
 
-        return new ChecksConfiguration(error, warning, whitespace, reviewers, merge, committer);
+        return new ChecksConfiguration(error, warning, whitespace, reviewers, merge, committer, issues);
     }
 }
