diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGAvailabilityMap.cpp
@@ -63,14 +63,14 @@
     m_heap = WTFMove(newHeap);
 }
 
 void AvailabilityMap::pruneByLiveness(Graph& graph, CodeOrigin where)
 {
-    Operands<Availability> localsCopy(m_locals.numberOfArguments(), m_locals.numberOfLocals(), Availability::unavailable());
+    Operands<Availability> localsCopy(OperandsLike, m_locals, Availability::unavailable());
     graph.forAllLiveInBytecode(
         where,
-        [&] (VirtualRegister reg) {
+        [&] (Operand reg) {
             localsCopy.operand(reg) = m_locals.operand(reg);
         });
     m_locals = WTFMove(localsCopy);
     pruneHeap();
 }
