diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPredictionInjectionPhase.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPredictionInjectionPhase.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPredictionInjectionPhase.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGPredictionInjectionPhase.cpp
@@ -70,11 +70,11 @@
                 continue;
             if (block->bytecodeBegin != m_graph.m_plan.osrEntryBytecodeIndex())
                 continue;
             const Operands<Optional<JSValue>>& mustHandleValues = m_graph.m_plan.mustHandleValues();
             for (size_t i = 0; i < mustHandleValues.size(); ++i) {
-                int operand = mustHandleValues.operandForIndex(i);
+                Operand operand = mustHandleValues.operandForIndex(i);
                 Optional<JSValue> value = mustHandleValues[i];
                 if (!value)
                     continue;
                 Node* node = block->variablesAtHead.operand(operand);
                 if (!node)
