diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/scripts/codegen/objc_generator_templates.py b/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/scripts/codegen/objc_generator_templates.py
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/scripts/codegen/objc_generator_templates.py
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/inspector/scripts/codegen/objc_generator_templates.py
@@ -111,10 +111,16 @@
 };""")
 
     BackendDispatcherHeaderDomainHandlerImplementation = (
     """void ObjCInspector${domainName}BackendDispatcher::${commandName}(${parameters})
 {
+    if (!${respondsToSelector}) {
+        backendDispatcher()->reportProtocolError(requestId, BackendDispatcher::MethodNotFound, "'${domainName}.${commandName}' was not found"_s);
+        backendDispatcher()->sendPendingErrors();
+        return;
+    }
+
     id errorCallback = ^(NSString *error) {
         backendDispatcher()->reportProtocolError(requestId, BackendDispatcher::ServerError, error);
         backendDispatcher()->sendPendingErrors();
     };
 
