diff a/src/hotspot/share/oops/inlineKlass.cpp b/src/hotspot/share/oops/inlineKlass.cpp
--- a/src/hotspot/share/oops/inlineKlass.cpp
+++ b/src/hotspot/share/oops/inlineKlass.cpp
@@ -176,11 +176,11 @@
 }
 
 // Arrays of...
 
 bool InlineKlass::flatten_array() {
-  if (!FlatArrayFlatten) {
+  if (!UseFlatArray) {
     return false;
   }
   // Too big
   int elem_bytes = raw_value_byte_size();
   if ((InlineArrayElemMaxFlatSize >= 0) && (elem_bytes > InlineArrayElemMaxFlatSize)) {
