diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/opcode_generator.rb b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/opcode_generator.rb
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/opcode_generator.rb
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/b3/air/opcode_generator.rb
@@ -802,10 +802,11 @@
 }
 
 writeH("OpcodeGenerated") {
     | outp |
     outp.puts "#include \"AirInstInlines.h\""
+    outp.puts "#include \"CCallHelpers.h\""
     outp.puts "#include \"wtf/PrintStream.h\""
     outp.puts "namespace WTF {"
     outp.puts "using namespace JSC::B3::Air;"
     outp.puts "void printInternal(PrintStream& out, Opcode opcode)"
     outp.puts "{"
