diff a/test/jdk/java/foreign/CallGeneratorHelper.java b/test/jdk/java/foreign/CallGeneratorHelper.java
--- a/test/jdk/java/foreign/CallGeneratorHelper.java
+++ b/test/jdk/java/foreign/CallGeneratorHelper.java
@@ -388,11 +388,11 @@
             if (Utils.isPadding(l)) continue;
             VarHandle accessor = g.varHandle(structFieldCarrier(l), MemoryLayout.PathElement.groupElement(l.name().get()));
             List<Consumer<Object>> fieldsCheck = new ArrayList<>();
             Object value = makeArg(l, fieldsCheck, check);
             if (isPointer(l)) {
-                value = MemoryAddress.getUnsafeOffset((MemoryAddress)value);
+                value = MemoryAddress.asLong((MemoryAddress)value);
             }
             //set value
             accessor.set(str.baseAddress(), value);
             //add check
             if (check) {
