diff a/modules/javafx.web/src/main/native/Source/CMakeLists.txt b/modules/javafx.web/src/main/native/Source/CMakeLists.txt
--- a/modules/javafx.web/src/main/native/Source/CMakeLists.txt
+++ b/modules/javafx.web/src/main/native/Source/CMakeLists.txt
@@ -11,11 +11,11 @@
     add_subdirectory(ThirdParty/capstone)
 endif ()
 
 add_subdirectory(JavaScriptCore)
 
-if (WIN32 AND ENABLE_GRAPHICS_CONTEXT_3D)
+if (ENABLE_GRAPHICS_CONTEXT_3D)
     add_subdirectory(ThirdParty/ANGLE)
 endif ()
 
 if (ENABLE_API_TESTS)
     add_subdirectory(ThirdParty/gtest)
