diff a/test/hotspot/jtreg/vmTestbase/gc/huge/quicklook/largeheap/MemOptions/MemOptionsTest.java b/test/hotspot/jtreg/vmTestbase/gc/huge/quicklook/largeheap/MemOptions/MemOptionsTest.java
--- a/test/hotspot/jtreg/vmTestbase/gc/huge/quicklook/largeheap/MemOptions/MemOptionsTest.java
+++ b/test/hotspot/jtreg/vmTestbase/gc/huge/quicklook/largeheap/MemOptions/MemOptionsTest.java
@@ -92,12 +92,11 @@
     private void positive(String name, String... opts) throws IOException {
         System.out.println(name);
         var cmd = new ArrayList<String>();
         Collections.addAll(cmd, opts);
         cmd.add(MemStat.class.getName());
-        var pb = ProcessTools.createJavaProcessBuilder(true,
-                cmd.toArray(new String[cmd.size()]));
+        var pb = ProcessTools.createJavaProcessBuilder(true, cmd);
         var output = new OutputAnalyzer(pb.start());
         if (output.getExitValue() != 0) {
             output.reportDiagnosticSummary();
             failed.add(name);
         }
@@ -106,12 +105,11 @@
     private void negative(String name, String... opts) throws IOException {
         System.out.println(name);
         var cmd = new ArrayList<String>();
         Collections.addAll(cmd, opts);
         cmd.add(MemStat.class.getName());
-        var pb = ProcessTools.createJavaProcessBuilder(true,
-                cmd.toArray(new String[cmd.size()]));
+        var pb = ProcessTools.createJavaProcessBuilder(true, cmd);
         var output = new OutputAnalyzer(pb.start());
         if (output.getExitValue() == 0) {
             output.reportDiagnosticSummary();
             failed.add(name);
         }
