diff a/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack019.java b/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack019.java
--- a/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack019.java
+++ b/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack019.java
@@ -40,11 +40,11 @@
  *     See the bug:
  *     4366625 (P4/S4) multiple stack overflow causes HS crash
  *
  * @ignore 8139875
  * @requires vm.opt.DeoptimizeALot != true
- * @run main/othervm nsk.stress.stack.stack019 -eager
+ * @run main/othervm/timeout=900 nsk.stress.stack.stack019 -eager
  */
 
 package nsk.stress.stack;
 
 
