diff a/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack014.java b/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack014.java
--- a/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack014.java
+++ b/test/hotspot/jtreg/vmTestbase/nsk/stress/stack/stack014.java
@@ -48,11 +48,11 @@
  *     However, it passes against all these HS versions on Win32.
  *     See the bug:
  *     4366625 (P4/S4) multiple stack overflow causes HS crash
  *
  * @requires vm.opt.DeoptimizeALot != true
- * @run main/othervm nsk.stress.stack.stack014
+ * @run main/othervm/timeout=900 nsk.stress.stack.stack014
  */
 
 package nsk.stress.stack;
 
 
