diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/galloca.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/galloca.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/galloca.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/galloca.h
@@ -86,11 +86,11 @@
  *   will be freed together with the variable sized array upon exit of that scope, and
  *   not upon exit of the enclosing function scope.
  *
  * Returns: space for @size bytes, allocated on the stack
  */
-#define g_alloca(size)       alloca (size)
+#define g_alloca(size)     alloca (size)
 /**
  * g_newa:
  * @struct_type: Type of memory chunks to be allocated
  * @n_structs: Number of chunks to be allocated
  *
