diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gcharset.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gcharset.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gcharset.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gcharset.h
@@ -29,10 +29,12 @@
 
 GLIB_AVAILABLE_IN_ALL
 gboolean              g_get_charset         (const char **charset);
 GLIB_AVAILABLE_IN_ALL
 gchar *               g_get_codeset         (void);
+GLIB_AVAILABLE_IN_2_62
+gboolean              g_get_console_charset (const char **charset);
 
 GLIB_AVAILABLE_IN_ALL
 const gchar * const * g_get_language_names  (void);
 GLIB_AVAILABLE_IN_2_58
 const gchar * const * g_get_language_names_with_category
