diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconstructor.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconstructor.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconstructor.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconstructor.h
@@ -26,10 +26,12 @@
 #define G_DEFINE_DESTRUCTOR(_func) static void __attribute__((destructor)) _func (void);
 
 #elif defined (_MSC_VER) && (_MSC_VER >= 1500)
 /* Visual studio 2008 and later has _Pragma */
 
+#include <stdlib.h>
+
 #define G_HAS_CONSTRUCTORS 1
 
 /* We do some weird things to avoid the constructors being optimized
  * away on VS2015 if WholeProgramOptimization is enabled. First we
  * make a reference to the array from the wrapper to make sure its
