diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconvert.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconvert.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconvert.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gconvert.h
@@ -82,87 +82,87 @@
  */
 typedef struct _GIConv *GIConv;
 
 GLIB_AVAILABLE_IN_ALL
 GIConv g_iconv_open   (const gchar  *to_codeset,
-               const gchar  *from_codeset);
+           const gchar  *from_codeset);
 GLIB_AVAILABLE_IN_ALL
 gsize  g_iconv        (GIConv        converter,
-               gchar       **inbuf,
-               gsize        *inbytes_left,
-               gchar       **outbuf,
-               gsize        *outbytes_left);
+           gchar       **inbuf,
+           gsize        *inbytes_left,
+           gchar       **outbuf,
+           gsize        *outbytes_left);
 GLIB_AVAILABLE_IN_ALL
 gint   g_iconv_close  (GIConv        converter);
 
 
 GLIB_AVAILABLE_IN_ALL
 gchar* g_convert               (const gchar  *str,
-                gssize        len,
-                const gchar  *to_codeset,
-                const gchar  *from_codeset,
-                gsize        *bytes_read,
-                gsize        *bytes_written,
-                GError      **error) G_GNUC_MALLOC;
+        gssize        len,
+        const gchar  *to_codeset,
+        const gchar  *from_codeset,
+        gsize        *bytes_read,
+        gsize        *bytes_written,
+        GError      **error) G_GNUC_MALLOC;
 GLIB_AVAILABLE_IN_ALL
 gchar* g_convert_with_iconv    (const gchar  *str,
-                gssize        len,
-                GIConv        converter,
-                gsize        *bytes_read,
-                gsize        *bytes_written,
-                GError      **error) G_GNUC_MALLOC;
+        gssize        len,
+        GIConv        converter,
+        gsize        *bytes_read,
+        gsize        *bytes_written,
+        GError      **error) G_GNUC_MALLOC;
 GLIB_AVAILABLE_IN_ALL
 gchar* g_convert_with_fallback (const gchar  *str,
-                gssize        len,
-                const gchar  *to_codeset,
-                const gchar  *from_codeset,
-                const gchar  *fallback,
-                gsize        *bytes_read,
-                gsize        *bytes_written,
-                GError      **error) G_GNUC_MALLOC;
+        gssize        len,
+        const gchar  *to_codeset,
+        const gchar  *from_codeset,
+        const gchar  *fallback,
+        gsize        *bytes_read,
+        gsize        *bytes_written,
+        GError      **error) G_GNUC_MALLOC;
 
 
 /* Convert between libc's idea of strings and UTF-8.
  */
 GLIB_AVAILABLE_IN_ALL
 gchar* g_locale_to_utf8   (const gchar  *opsysstring,
-               gssize        len,
-               gsize        *bytes_read,
-               gsize        *bytes_written,
-               GError      **error) G_GNUC_MALLOC;
+         gssize        len,
+         gsize        *bytes_read,
+         gsize        *bytes_written,
+         GError      **error) G_GNUC_MALLOC;
 GLIB_AVAILABLE_IN_ALL
 gchar* g_locale_from_utf8 (const gchar  *utf8string,
-               gssize        len,
-               gsize        *bytes_read,
-               gsize        *bytes_written,
-               GError      **error) G_GNUC_MALLOC;
+         gssize        len,
+         gsize        *bytes_read,
+         gsize        *bytes_written,
+         GError      **error) G_GNUC_MALLOC;
 
 /* Convert between the operating system (or C runtime)
  * representation of file names and UTF-8.
  */
 GLIB_AVAILABLE_IN_ALL
 gchar* g_filename_to_utf8   (const gchar  *opsysstring,
-                 gssize        len,
-                 gsize        *bytes_read,
-                 gsize        *bytes_written,
-                 GError      **error) G_GNUC_MALLOC;
+           gssize        len,
+           gsize        *bytes_read,
+           gsize        *bytes_written,
+           GError      **error) G_GNUC_MALLOC;
 GLIB_AVAILABLE_IN_ALL
 gchar* g_filename_from_utf8 (const gchar  *utf8string,
-                 gssize        len,
-                 gsize        *bytes_read,
-                 gsize        *bytes_written,
-                 GError      **error) G_GNUC_MALLOC;
+           gssize        len,
+           gsize        *bytes_read,
+           gsize        *bytes_written,
+           GError      **error) G_GNUC_MALLOC;
 
 GLIB_AVAILABLE_IN_ALL
 gchar *g_filename_from_uri (const gchar *uri,
-                gchar      **hostname,
-                GError     **error) G_GNUC_MALLOC;
+          gchar      **hostname,
+          GError     **error) G_GNUC_MALLOC;
 
 GLIB_AVAILABLE_IN_ALL
 gchar *g_filename_to_uri   (const gchar *filename,
-                const gchar *hostname,
-                GError     **error) G_GNUC_MALLOC;
+          const gchar *hostname,
+          GError     **error) G_GNUC_MALLOC;
 GLIB_AVAILABLE_IN_ALL
 gchar *g_filename_display_name (const gchar *filename) G_GNUC_MALLOC;
 GLIB_AVAILABLE_IN_ALL
 gboolean g_get_filename_charsets (const gchar ***filename_charsets);
 
