diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gdatasetprivate.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gdatasetprivate.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gdatasetprivate.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gdatasetprivate.h
@@ -31,11 +31,11 @@
 G_BEGIN_DECLS
 
 /* GET_FLAGS is implemented via atomic pointer access, to allow memory
  * barriers to take effect without acquiring the global dataset mutex.
  */
-#define G_DATALIST_GET_FLAGS(datalist)              \
+#define G_DATALIST_GET_FLAGS(datalist)        \
   ((gsize) g_atomic_pointer_get (datalist) & G_DATALIST_FLAGS_MASK)
 
 
 G_END_DECLS
 
