diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/ghmac.c b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/ghmac.c
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/ghmac.c
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/ghmac.c
@@ -263,11 +263,11 @@
 
 /**
  * g_hmac_get_string:
  * @hmac: a #GHmac
  *
- * Gets the HMAC as an hexadecimal string.
+ * Gets the HMAC as a hexadecimal string.
  *
  * Once this function has been called the #GHmac can no longer be
  * updated with g_hmac_update().
  *
  * The hexadecimal characters will be lower case.
