diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-unix.c b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-unix.c
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-unix.c
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-unix.c
@@ -308,11 +308,11 @@
 
   return (* func) (fd_source->fd, g_source_query_unix_fd (source, fd_source->tag), user_data);
 }
 
 GSourceFuncs g_unix_fd_source_funcs = {
-  NULL, NULL, g_unix_fd_source_dispatch, NULL
+  NULL, NULL, g_unix_fd_source_dispatch, NULL, NULL, NULL
 };
 
 /**
  * g_unix_fd_source_new:
  * @fd: a file descriptor
