diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-object.h b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-object.h
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-object.h
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/glib-object.h
@@ -32,10 +32,13 @@
 #include <gobject/gtypemodule.h>
 #include <gobject/gtypeplugin.h>
 #include <gobject/gvalue.h>
 #include <gobject/gvaluearray.h>
 #include <gobject/gvaluetypes.h>
+#ifndef GSTREAMER_LITE
+#include <gobject/glib-enumtypes.h>
+#endif // GSTREAMER_LITE
 
 #include <gobject/gobject-autocleanups.h>
 
 #undef __GLIB_GOBJECT_H_INSIDE__
 
