diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gsequence.c b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gsequence.c
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gsequence.c
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gsequence.c
@@ -825,11 +825,11 @@
 }
 
 /**
  * g_sequence_lookup:
  * @seq: a #GSequence
- * @data: data to lookup
+ * @data: data to look up
  * @cmp_func: the function used to compare items in the sequence
  * @cmp_data: user data passed to @cmp_func
  *
  * Returns an iterator pointing to the position of the first item found
  * equal to @data according to @cmp_func and @cmp_data. If more than one
@@ -1113,11 +1113,11 @@
 }
 
 /**
  * g_sequence_lookup_iter:
  * @seq: a #GSequence
- * @data: data to lookup
+ * @data: data to look up
  * @iter_cmp: the function used to compare iterators in the sequence
  * @cmp_data: user data passed to @iter_cmp
  *
  * Like g_sequence_lookup(), but uses a #GSequenceIterCompareFunc
  * instead of a #GCompareDataFunc as the compare function.
