diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtrashstack.c b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtrashstack.c
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtrashstack.c
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtrashstack.c
@@ -22,10 +22,13 @@
  * GLib at ftp://ftp.gtk.org/pub/gtk/.
  */
 
 #include "config.h"
 
+/* we know we are deprecated here, no need for warnings */
+#define GLIB_DISABLE_DEPRECATION_WARNINGS
+
 #include "gtrashstack.h"
 
 /**
  * SECTION:trash_stack
  * @title: Trash Stacks
