diff a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtree.c b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtree.c
--- a/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtree.c
+++ b/modules/javafx.media/src/main/native/gstreamer/3rd_party/glib/glib/gtree.c
@@ -46,11 +46,11 @@
  *
  * To create a new #GTree use g_tree_new().
  *
  * To insert a key/value pair into a #GTree use g_tree_insert().
  *
- * To lookup the value corresponding to a given key, use
+ * To look up the value corresponding to a given key, use
  * g_tree_lookup() and g_tree_lookup_extended().
  *
  * To find out the number of nodes in a #GTree, use g_tree_nnodes(). To
  * get the height of a #GTree, use g_tree_height().
  *
@@ -871,12 +871,12 @@
 
 /**
  * g_tree_lookup_extended:
  * @tree: a #GTree
  * @lookup_key: the key to look up
- * @orig_key: (optional) (nullable): returns the original key
- * @value: (optional) (nullable): returns the value associated with the key
+ * @orig_key: (out) (optional) (nullable): returns the original key
+ * @value: (out) (optional) (nullable): returns the value associated with the key
  *
  * Looks up a key in the #GTree, returning the original key and the
  * associated value. This is useful if you need to free the memory
  * allocated for the original key, for example before calling
  * g_tree_remove().
