diff a/network/src/test/java/org/openjdk/skara/network/RestRequestTests.java b/network/src/test/java/org/openjdk/skara/network/RestRequestTests.java
--- a/network/src/test/java/org/openjdk/skara/network/RestRequestTests.java
+++ b/network/src/test/java/org/openjdk/skara/network/RestRequestTests.java
@@ -41,11 +41,11 @@
     private final List<JSONObject> requests;
     private final String response;
     private int responseCode;
 
     private int truncatedResponseCount = 0;
-    private boolean usedCache = false;
+    private volatile boolean usedCache = false;
 
     class Handler implements HttpHandler {
         private String checksum(String body) {
             try {
                 var digest = MessageDigest.getInstance("SHA-256");
