diff a/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java b/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java
--- a/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java
+++ b/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java
@@ -350,11 +350,11 @@
             return null;
         }, snapshotParams, img);
     }
 
     // Test tiled snapshots
-    
+
     private void doTestTiledSnapshotImm(int w, int h) {
         setupImageScene(w, h);
         Image original = ((ImageView) tmpNode).getImage();
         assertNotNull(original);
         WritableImage buffer = useImage ? new WritableImage(w, h) : null;
@@ -434,11 +434,11 @@
 
     @Test
     public void testSnapshot2x2TilesDifferentSizeImm() {
         doTestTiledSnapshotImm(4101, 4101);
     }
-    
+
     @Test
     public void testSnapshot2x2TilesSameHeightImm() {
         doTestTiledSnapshotImm(4101, 4100);
     }
 
@@ -505,11 +505,11 @@
 
             assertEquals(WIDTH, (int)wimg.getWidth());
             assertEquals(HEIGHT, (int)wimg.getHeight());
         });
     }
-    
+
     private void doTestSnapshotScaleNodeDefer(int xScale, int yScale) {
         setupSimpleScene();
         final SnapshotParameters snapshotParams = new SnapshotParameters();
         snapshotParams.setTransform(Transform.scale(xScale, yScale));
         final int WIDTH = NODE_W * xScale;
@@ -557,11 +557,11 @@
 
     @Test
     public void testSnapshotBigYScaleNodeDefer() {
         doTestSnapshotScaleNodeDefer(1, 200);
     }
-    
+
     // Test node snapshot with a 90 degree rotate transform
 
     @Test
     public void testSnapshotRotateNodeImm() {
         setupSimpleScene();
