diff a/test/hotspot/jtreg/runtime/cds/appcds/SharedBaseAddress.java b/test/hotspot/jtreg/runtime/cds/appcds/SharedBaseAddress.java
--- a/test/hotspot/jtreg/runtime/cds/appcds/SharedBaseAddress.java
+++ b/test/hotspot/jtreg/runtime/cds/appcds/SharedBaseAddress.java
@@ -48,10 +48,13 @@
         String appJar = JarBuilder.getOrCreateHelloJar();
 
         for (String testEntry : testTable) {
             System.out.println("sharedBaseAddress = " + testEntry);
 
+            // Note: some platforms may restrict valid values for SharedBaseAddress; the VM should print
+            // a warning and use the default value instead. Similar, ASLR may prevent the given address
+            // from being used; this too should handled gracefully by using the default base address.
             OutputAnalyzer dumpOutput = TestCommon.dump(
                 appJar, new String[] {"Hello"}, "-XX:SharedBaseAddress=" + testEntry);
             TestCommon.checkDump(dumpOutput, "Loading classes to share");
 
             OutputAnalyzer execOutput = TestCommon.exec(appJar, "Hello");
