diff a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifierFactory.java b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifierFactory.java
--- a/bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifierFactory.java
+++ b/bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifierFactory.java
@@ -61,8 +61,12 @@
 
         if (notifierConfiguration.contains("security")) {
             builder.securityLevel(notifierConfiguration.get("security").asString());
         }
 
+        if (notifierConfiguration.contains("pronly")) {
+            builder.prOnly(notifierConfiguration.get("pronly").asBoolean());
+        }
+
         return builder.build();
     }
 }
