diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLStageInOutSemantic.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLStageInOutSemantic.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLStageInOutSemantic.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLStageInOutSemantic.cpp
@@ -45,17 +45,16 @@
     if (is<ArrayType>(unnamedType))
         return true;
     if (!is<TypeReference>(unnamedType))
         return false;
     auto& typeReference = downcast<TypeReference>(unnamedType);
-    ASSERT(typeReference.resolvedType());
-    auto& resolvedType = *typeReference.resolvedType();
+    auto& resolvedType = typeReference.resolvedType();
     if (is<EnumerationDefinition>(resolvedType))
         return true;
     if (!is<NativeTypeDeclaration>(resolvedType))
         return false;
-    auto& nativeTypeDeclaration = downcast<NativeTypeDeclaration>(*typeReference.resolvedType());
+    auto& nativeTypeDeclaration = downcast<NativeTypeDeclaration>(typeReference.resolvedType());
     return nativeTypeDeclaration.isNumber()
         || nativeTypeDeclaration.isVector()
         || nativeTypeDeclaration.isMatrix();
 }
 
