diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLIntegerLiteral.h
@@ -27,45 +27,46 @@
 
 #if ENABLE(WEBGPU)
 
 #include "WHLSLExpression.h"
 #include "WHLSLIntegerLiteralType.h"
-#include "WHLSLLexer.h"
+#include <wtf/FastMalloc.h>
 
 namespace WebCore {
 
 namespace WHLSL {
 
 namespace AST {
 
-class IntegerLiteral : public Expression {
+class IntegerLiteral final : public Expression {
+    WTF_MAKE_FAST_ALLOCATED;
 public:
-    IntegerLiteral(Lexer::Token&& origin, int value)
-        : Expression(Lexer::Token(origin))
-        , m_type(WTFMove(origin), value)
+    IntegerLiteral(CodeLocation location, int value)
+        : Expression(location, Kind::IntegerLiteral)
+        , m_type(location, value)
         , m_value(value)
     {
     }
 
-    virtual ~IntegerLiteral() = default;
+    ~IntegerLiteral() = default;
 
     IntegerLiteral(const IntegerLiteral&) = delete;
     IntegerLiteral(IntegerLiteral&&) = default;
 
     IntegerLiteral& operator=(const IntegerLiteral&) = delete;
     IntegerLiteral& operator=(IntegerLiteral&&) = default;
 
     IntegerLiteralType& type() { return m_type; }
     int value() const { return m_value; }
 
-    bool isIntegerLiteral() const override { return true; }
-
     IntegerLiteral clone() const
     {
-        IntegerLiteral result(Lexer::Token(origin()), m_value);
-        if (result.m_type.resolvedType())
-            result.m_type.resolve(result.m_type.resolvedType()->clone());
+        IntegerLiteral result(codeLocation(), m_value);
+        result.m_type = m_type.clone();
+        if (auto* resolvedType = m_type.maybeResolvedType())
+            result.m_type.resolve(const_cast<AST::UnnamedType&>(*resolvedType));
+        copyTypeTo(result);
         return result;
     }
 
     int64_t valueForSelectedType() const;
 
@@ -78,8 +79,10 @@
 
 }
 
 }
 
+DEFINE_DEFAULT_DELETE(IntegerLiteral)
+
 SPECIALIZE_TYPE_TRAITS_WHLSL_EXPRESSION(IntegerLiteral, isIntegerLiteral())
 
 #endif
