diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLSpecializationConstantSemantic.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLSpecializationConstantSemantic.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLSpecializationConstantSemantic.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/webgpu/WHLSL/AST/WHLSLSpecializationConstantSemantic.cpp
@@ -40,14 +40,13 @@
 bool SpecializationConstantSemantic::isAcceptableType(const UnnamedType& unnamedType, const Intrinsics&) const
 {
     if (!is<TypeReference>(unnamedType))
         return false;
     auto& typeReference = downcast<TypeReference>(unnamedType);
-    ASSERT(typeReference.resolvedType());
-    if (!is<NativeTypeDeclaration>(*typeReference.resolvedType()))
+    if (!is<NativeTypeDeclaration>(typeReference.resolvedType()))
         return false;
-    return downcast<NativeTypeDeclaration>(*typeReference.resolvedType()).isNumber();
+    return downcast<NativeTypeDeclaration>(typeReference.resolvedType()).isNumber();
 }
 
 bool SpecializationConstantSemantic::isAcceptableForShaderItemDirection(ShaderItemDirection direction, const Optional<EntryPointType>&) const
 {
     return direction == ShaderItemDirection::Input;
