diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchBodyOwner.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchBodyOwner.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchBodyOwner.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchBodyOwner.h
@@ -55,11 +55,11 @@
 
     void loadBlob(const Blob&, FetchBodyConsumer*);
 
     bool isActive() const { return !!m_blobLoader; }
 
-    RefPtr<ReadableStream> readableStream(JSC::ExecState&);
+    RefPtr<ReadableStream> readableStream(JSC::JSGlobalObject&);
     bool hasReadableStreamBody() const { return m_body && m_body->hasReadableStream(); }
 
 #if ENABLE(STREAMS_API)
     virtual void consumeBodyAsStream();
     virtual void feedStream() { }
@@ -80,11 +80,11 @@
     ExceptionOr<void> extractBody(FetchBody::Init&&);
     void updateContentType();
     void consumeOnceLoadingFinished(FetchBodyConsumer::Type, Ref<DeferredPromise>&&);
 
     void setBody(FetchBody&& body) { m_body = WTFMove(body); }
-    void createReadableStream(JSC::ExecState&);
+    void createReadableStream(JSC::JSGlobalObject&);
 
     // ActiveDOMObject API
     void stop() override;
 
     void setDisturbed() { m_isDisturbed = true; }
