diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchResponse.idl b/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchResponse.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchResponse.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/fetch/FetchResponse.idl
@@ -31,11 +31,11 @@
 
 enum FetchResponseType { "basic", "cors", "default", "error", "opaque", "opaqueredirect" };
 
 dictionary FetchResponseInit {
     unsigned short status = 200;
-    ByteString statusText = "OK";
+    ByteString statusText = "";
     HeadersInit headers;
 };
 
 [
     ActiveDOMObject,
