diff a/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java b/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java
--- a/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java
+++ b/forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java
@@ -198,10 +198,16 @@
     public URI webUrl(Hash hash) {
         var endpoint = "/" + repository + "/commit/" + hash.abbreviate();
         return gitHubHost.getWebURI(endpoint);
     }
 
+    @Override
+    public URI webUrl(String baseRef, String headRef) {
+        var endpoint = "/" + repository + "/compare/" + baseRef + "..." + headRef;
+        return gitHubHost.getWebURI(endpoint);
+    }
+
     @Override
     public VCS repositoryType() {
         return VCS.GIT;
     }
 
