diff a/application/org.openjdk.jmc.browser/META-INF/MANIFEST.MF b/application/org.openjdk.jmc.browser/META-INF/MANIFEST.MF
--- a/application/org.openjdk.jmc.browser/META-INF/MANIFEST.MF
+++ b/application/org.openjdk.jmc.browser/META-INF/MANIFEST.MF
@@ -1,9 +1,9 @@
 Manifest-Version: 1.0
 Bundle-Name: JVMBrowser Plug-in
 Bundle-SymbolicName: org.openjdk.jmc.browser;singleton:=true
-Bundle-Version: 7.1.0.qualifier
+Bundle-Version: 8.0.0.qualifier
 Bundle-Activator: org.openjdk.jmc.browser.JVMBrowserPlugin
 Bundle-Localization: plugin
 Require-Bundle: org.openjdk.jmc.ui,
  org.eclipse.ui.forms,
  org.openjdk.jmc.rjmx,
