diff a/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java b/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java
--- a/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java
+++ b/tests/system/src/test/java/test/javafx/scene/Snapshot2Test.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2012, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2012, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -39,11 +39,10 @@
 import javafx.util.Callback;
 import org.junit.After;
 import org.junit.AfterClass;
 import org.junit.Before;
 import org.junit.BeforeClass;
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 import org.junit.runners.Parameterized.Parameters;
 import test.util.Util;
@@ -371,30 +370,26 @@
     @Test
     public void testSnapshotScaleNodeDefer() {
         doTestSnapshotScaleNodeDefer(3, 3);
     }
 
-    // TODO: Re-enable this test when RT-22073 is fixed
-    @Ignore @Test
+    @Test
     public void testSnapshotBigXScaleNodeImm() {
         doTestSnapshotScaleNodeImm(100, 1);
     }
 
-    // TODO: Re-enable this test when RT-22073 is fixed
-    @Ignore @Test
+    @Test
     public void testSnapshotBigXScaleNodeDefer() {
         doTestSnapshotScaleNodeDefer(100, 1);
     }
 
-    // TODO: Re-enable this test when RT-22073 is fixed
-    @Ignore @Test
+    @Test
     public void testSnapshotBigYScaleNodeImm() {
         doTestSnapshotScaleNodeImm(1, 200);
     }
 
-    // TODO: Re-enable this test when RT-22073 is fixed
-    @Ignore @Test
+    @Test
     public void testSnapshotBigYScaleNodeDefer() {
         doTestSnapshotScaleNodeDefer(1, 200);
     }
 
     // Test node snapshot with a 90 degree rotate transform
