diff a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java
@@ -432,11 +432,11 @@
     public Symbol valueProjection() {
         return null;
     }
 
     /**
-     * Return the reference projection IFF 'this' happens to be value projection, null
+     * Return the reference projection IFF 'this' happens to be inline class, null
      * otherwise.
      */
     public Symbol referenceProjection() {
         return null;
     }
@@ -1669,11 +1669,14 @@
             return isReferenceProjection() ? projection : null;
         }
 
         @Override
         public ClassSymbol referenceProjection() {
-            if (!isValue() || projection != null)
+            if (!isValue())
+                return null;
+
+            if (projection != null)
                 return projection;
 
             ClassType ct = (ClassType) this.type;
             ClassType projectedType = new ClassType(ct.getEnclosingType(), ct.typarams_field, null);
             projectedType.allparams_field = ct.allparams_field;
