diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/Scripts/generate-js-builtins.py b/modules/javafx.web/src/main/native/Source/JavaScriptCore/Scripts/generate-js-builtins.py
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/Scripts/generate-js-builtins.py
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/Scripts/generate-js-builtins.py
@@ -137,11 +137,11 @@
     cli_parser.add_option("--wrappers-only", action="store_true", help="Produce .h/.cpp wrapper files to ease integration of the builtins.")
     cli_parser.add_option("-v", "--debug", action="store_true", help="Log extra output for debugging the generator itself.")
     cli_parser.add_option("-t", "--test", action="store_true", help="Enable test mode.")
 
     arg_options, arg_values = cli_parser.parse_args()
-    if len(arg_values) is 0 and not arg_options.input_directory:
+    if len(arg_values) == 0 and not arg_options.input_directory:
         raise ParseException("At least one input file or directory expected.")
 
     if not arg_options.output_directory:
         raise ParseException("Missing output directory.")
 
