diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCapabilities.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCapabilities.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCapabilities.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCapabilities.h
@@ -164,9 +164,9 @@
     return inlineFunctionForConstructCapabilityLevel(codeBlock);
 }
 
 inline bool isSmallEnoughToInlineCodeInto(CodeBlock* codeBlock)
 {
-    return codeBlock->instructionCount() <= Options::maximumInliningCallerSize();
+    return codeBlock->bytecodeCost() <= Options::maximumInliningCallerBytecodeCost();
 }
 
 } } // namespace JSC::DFG
