diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCommon.h b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCommon.h
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCommon.h
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/dfg/DFGCommon.h
@@ -171,11 +171,11 @@
     // locals is extended; for example you can replace a GetLocal with a Phantom
     // and so long as the Phantom retains the GetLocal's children (i.e. the Phi
     // most likely) then it implies that the local is still live but that it need
     // not be stored to the stack necessarily. This implies that Phantom can
     // reference nodes that have no result, as long as those nodes are valid
-    // GetLocal children (i.e. Phi, SetLocal, SetArgument).
+    // GetLocal children (i.e. Phi, SetLocal, SetArgumentDefinitely, SetArgumentMaybe).
     //
     // LoadStore form also implies that Phis need not have children. By default,
     // they end up having no children if you enter LoadStore using the canonical
     // way (call Graph::dethread).
     //
