diff a/modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceWithSecurityManagerTest.java b/modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceWithSecurityManagerTest.java
--- a/modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceWithSecurityManagerTest.java
+++ b/modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceWithSecurityManagerTest.java
@@ -41,11 +41,11 @@
  * It is ad-hoc, a more rigorous analysis on what the permissions should be
  * would be great, and then we could add the ability to do security manager
  * related tests on lots of different unit tests.
  */
 @RunWith(ServiceWithSecurityManagerTest.ServiceTestRunner.class)
-@Ignore("This class doesn't appear to run correctly, often s.evaluate isn't called. Likely bogus test at present.")
+@Ignore("JDK-8234175") // This class doesn't appear to run correctly, often s.evaluate isn't called. Likely bogus test at present.
 public class ServiceWithSecurityManagerTest extends ServiceLifecycleTest {
 
     public static final class ServiceTestRunner extends BlockJUnit4ClassRunner {
         private ThreadGroup mainThreadGroup;
 
