diff a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
--- a/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
+++ b/modules/javafx.graphics/src/main/native-glass/gtk/glass_window.h
@@ -160,11 +160,10 @@
     } xim = { NULL, NULL, FALSE};
 
     size_t events_processing_cnt;
 
     WindowGeometry geometry;
-    GtkWindowGroup *win_group; // this is used for window grabs
     std::set<WindowContext *> children;
     GdkWMFunction gdk_windowManagerFunctions;
     GtkWidget *gtk_widget;
     GdkWindow *gdk_window;
     BgColor bg_color;
