diff a/src/hotspot/os_cpu/aix_ppc/os_aix_ppc.cpp b/src/hotspot/os_cpu/aix_ppc/os_aix_ppc.cpp
--- a/src/hotspot/os_cpu/aix_ppc/os_aix_ppc.cpp
+++ b/src/hotspot/os_cpu/aix_ppc/os_aix_ppc.cpp
@@ -373,16 +373,16 @@
         }
         stub = SharedRuntime::get_handle_wrong_method_stub();
         goto run_stub;
       }
 
-      else if ((SafepointMechanism::uses_thread_local_poll() && USE_POLL_BIT_ONLY)
+      else if (USE_POLL_BIT_ONLY
                ? (sig == SIGTRAP && ((NativeInstruction*)pc)->is_safepoint_poll())
-               : (sig == SIGSEGV && os::is_poll_address(addr))) {
+               : (sig == SIGSEGV && SafepointMechanism::is_poll_address(addr))) {
         if (TraceTraps) {
           tty->print_cr("trap: safepoint_poll at " INTPTR_FORMAT " (%s)", p2i(pc),
-                        (SafepointMechanism::uses_thread_local_poll() && USE_POLL_BIT_ONLY) ? "SIGTRAP" : "SIGSEGV");
+                        USE_POLL_BIT_ONLY ? "SIGTRAP" : "SIGSEGV");
         }
         stub = SharedRuntime::get_poll_stub(pc);
         goto run_stub;
       }
 
