diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp
@@ -73,11 +73,11 @@
     auto& statuses = m_session->statuses();
     return std::any_of(statuses.begin(), statuses.end(),
         [&keyId] (auto& it) { return keyIdsMatch(it.first, keyId); });
 }
 
-JSC::JSValue MediaKeyStatusMap::get(JSC::ExecState& state, const BufferSource& keyId)
+JSC::JSValue MediaKeyStatusMap::get(JSC::JSGlobalObject& state, const BufferSource& keyId)
 {
     if (!m_session)
         return JSC::jsUndefined();
 
     auto& statuses = m_session->statuses();
