diff a/test/hotspot/jtreg/testlibrary_tests/ctw/ClassesDirTest.java b/test/hotspot/jtreg/testlibrary_tests/ctw/ClassesDirTest.java
--- a/test/hotspot/jtreg/testlibrary_tests/ctw/ClassesDirTest.java
+++ b/test/hotspot/jtreg/testlibrary_tests/ctw/ClassesDirTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2013, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -30,11 +30,10 @@
  *          java.base/jdk.internal.misc
  *          java.base/jdk.internal.reflect
  *          java.management
  * @build sun.hotspot.WhiteBox Foo Bar
  * @run driver ClassFileInstaller sun.hotspot.WhiteBox Foo Bar
- *                                sun.hotspot.WhiteBox$WhiteBoxPermission
  * @run driver ClassesDirTest prepare
  * @run driver ClassesDirTest compile classes
  * @run driver ClassesDirTest check
  * @summary testing of CompileTheWorld :: classes in directory
  * @author igor.ignatyev@oracle.com
