diff a/test/hotspot/jtreg/testlibrary_tests/ctw/JarsTest.java b/test/hotspot/jtreg/testlibrary_tests/ctw/JarsTest.java
--- a/test/hotspot/jtreg/testlibrary_tests/ctw/JarsTest.java
+++ b/test/hotspot/jtreg/testlibrary_tests/ctw/JarsTest.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2013, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2013, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -32,11 +32,10 @@
  *          java.compiler
  *          java.management
  *          jdk.internal.jvmstat/sun.jvmstat.monitor
  * @build sun.hotspot.WhiteBox Foo Bar
  * @run driver ClassFileInstaller sun.hotspot.WhiteBox Foo Bar
- *                                sun.hotspot.WhiteBox$WhiteBoxPermission
  * @run driver JarsTest prepare
  * @run driver JarsTest compile foo.jar bar.jar
  * @run driver JarsTest check
  * @summary testing of CompileTheWorld :: jars
  * @author igor.ignatyev@oracle.com
