diff a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/MemberSummaryWriter.java b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/MemberSummaryWriter.java
--- a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/MemberSummaryWriter.java
+++ b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/MemberSummaryWriter.java
@@ -97,16 +97,16 @@
      * @return a content tree containing the inherited summary links
      */
     Content getInheritedSummaryLinksTree();
 
     /**
-     * Add the member tree to the member summary tree.
+     * Adds the given summary to the list of summaries.
      *
-     * @param memberSummaryTree the content tree representing the member summary
-     * @param memberTree the content tree representing the member
+     * @param summariesList the list of summaries
+     * @param content       the summary
      */
-    void addMemberTree(Content memberSummaryTree, Content memberTree);
+    void addSummary(Content summariesList, Content content);
 
     /**
      * Get the member tree.
      *
      * @param memberTree the content tree representing the member
