diff a/test/jdk/TEST.groups b/test/jdk/TEST.groups
--- a/test/jdk/TEST.groups
+++ b/test/jdk/TEST.groups
@@ -39,10 +39,11 @@
 
 tier1_part3 = \
     :jdk_math \
     :jdk_svc_sanity \
     :jdk_foreign \
+    :jdk_jextract \
     java/nio/Buffer \
     com/sun/crypto/provider/Cipher \
     sun/nio/cs/ISO8859x.java
 
 # When adding tests to tier2, make sure they end up in one of the tier2_partX groups
@@ -334,10 +335,14 @@
     :svc_tools
 
 jdk_foreign = \
     java/foreign
 
+jdk_jextract = \
+    java/jextract \
+    tools/jextract
+
 #############################
 
 #
 # Client area groups
 #
