diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IDBConnectionToClient.cpp b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IDBConnectionToClient.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IDBConnectionToClient.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IDBConnectionToClient.cpp
@@ -37,15 +37,15 @@
 {
     return adoptRef(*new IDBConnectionToClient(delegate));
 }
 
 IDBConnectionToClient::IDBConnectionToClient(IDBConnectionToClientDelegate& delegate)
-    : m_delegate(makeWeakPtr(delegate))
+    : m_delegate(&delegate)
 {
 }
 
-uint64_t IDBConnectionToClient::identifier() const
+IDBConnectionIdentifier IDBConnectionToClient::identifier() const
 {
     ASSERT(m_delegate);
     return m_delegate->identifier();
 }
 
