diff a/modules/javafx.base/src/main/java/javafx/beans/property/ReadOnlyBooleanPropertyBase.java b/modules/javafx.base/src/main/java/javafx/beans/property/ReadOnlyBooleanPropertyBase.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/ReadOnlyBooleanPropertyBase.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/ReadOnlyBooleanPropertyBase.java
@@ -39,10 +39,16 @@
  */
 public abstract class ReadOnlyBooleanPropertyBase extends ReadOnlyBooleanProperty {
 
     ExpressionHelper<Boolean> helper;
 
+    /**
+     * Creates a default {@code ReadOnlyBooleanPropertyBase}.
+     */
+    public ReadOnlyBooleanPropertyBase() {
+    }
+
     @Override
     public void addListener(InvalidationListener listener) {
         helper = ExpressionHelper.addListener(helper, this, listener);
     }
 
