diff a/src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Poly1305Parameters.java b/src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Poly1305Parameters.java
--- a/src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Poly1305Parameters.java
+++ b/src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Poly1305Parameters.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -38,13 +38,13 @@
  * <a href="https://tools.ietf.org/html/rfc8103"><i>RFC 8103</i></a>
  * and is defined according to the following ASN.1:
  *
  * <pre>
  * id-alg-AEADChaCha20Poly1305 OBJECT IDENTIFIER ::=
-          { iso(1) member-body(2) us(840) rsadsi(113549) pkcs(1)
-            pkcs9(9) smime(16) alg(3) 18 }
-
+ *        { iso(1) member-body(2) us(840) rsadsi(113549) pkcs(1)
+ *          pkcs9(9) smime(16) alg(3) 18 }
+ *
  * AEADChaCha20Poly1305Nonce ::= OCTET STRING (SIZE(12))
  * </pre>
  *
  * The AlgorithmParameters may be instantiated either by its name
  * ("ChaCha20-Poly1305") or via its OID (1.2.840.113549.1.9.16.3.18)
@@ -84,11 +84,11 @@
     }
 
     /**
      * Initialize the ChaCha20Poly1305Parameters from a DER encoded
      * parameter block.
-
+     *
      * @param encoded the DER encoding of the nonce as an OCTET STRING.
      *
      * @throws IOException if the encoded nonce is not 12 bytes long or a DER
      *      decoding error occurs.
      */
