diff a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/ClassRemapper.java b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/ClassRemapper.java
--- a/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/ClassRemapper.java
+++ b/src/java.base/share/classes/jdk/internal/org/objectweb/asm/commons/ClassRemapper.java
@@ -250,17 +250,17 @@
     }
 
     /**
       * <b>Experimental, use at your own risk.</b>.
       *
-      * @param permittedSubtype the internal name of a permitted subtype.
+      * @param permittedSubclass the internal name of a permitted subclass.
       * @deprecated this API is experimental.
       */
     @Override
     @Deprecated
-    public void visitPermittedSubtypeExperimental(final String permittedSubtype) {
-        super.visitPermittedSubtypeExperimental(remapper.mapType(permittedSubtype));
+    public void visitPermittedSubclassExperimental(final String permittedSubclass) {
+        super.visitPermittedSubclassExperimental(remapper.mapType(permittedSubclass));
     }
 
     /**
       * Constructs a new remapper for fields. The default implementation of this method returns a new
       * {@link FieldRemapper}.
