diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/ClassCollection.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/ClassCollection.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/ClassCollection.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/ClassCollection.cpp
@@ -30,14 +30,17 @@
 #include "config.h"
 #include "ClassCollection.h"
 
 #include "NodeRareData.h"
 #include "StyledElement.h"
+#include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
-Ref<ClassCollection> ClassCollection::create(ContainerNode& rootNode, CollectionType type, const AtomicString& classNames)
+WTF_MAKE_ISO_ALLOCATED_IMPL(ClassCollection);
+
+Ref<ClassCollection> ClassCollection::create(ContainerNode& rootNode, CollectionType type, const AtomString& classNames)
 {
     ASSERT(type == ByClass);
     return adoptRef(*new ClassCollection(rootNode, type, classNames));
 }
 
