diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/DocumentFragment.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/DocumentFragment.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/DocumentFragment.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/DocumentFragment.cpp
@@ -77,11 +77,11 @@
         break;
     case CloningOperation::Everything:
         cloneChildNodes(clone);
         break;
     }
-    return WTFMove(clone);
+    return clone;
 }
 
 void DocumentFragment::parseHTML(const String& source, Element* contextElement, ParserContentPolicy parserContentPolicy)
 {
     ASSERT(contextElement);
@@ -91,11 +91,11 @@
 bool DocumentFragment::parseXML(const String& source, Element* contextElement, ParserContentPolicy parserContentPolicy)
 {
     return XMLDocumentParser::parseDocumentFragment(source, *this, contextElement, parserContentPolicy);
 }
 
-Element* DocumentFragment::getElementById(const AtomicString& id) const
+Element* DocumentFragment::getElementById(const AtomString& id) const
 {
     if (id.isNull())
         return nullptr;
 
     // Fast path for ShadowRoot, where we are both a DocumentFragment and a TreeScope.
