diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/ExtensionStyleSheets.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/ExtensionStyleSheets.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/ExtensionStyleSheets.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/ExtensionStyleSheets.cpp
@@ -45,11 +45,13 @@
 #include "UserContentURLPattern.h"
 #include "UserStyleSheet.h"
 
 namespace WebCore {
 
+#if ENABLE(CONTENT_EXTENSIONS)
 using namespace ContentExtensions;
+#endif
 using namespace HTMLNames;
 
 ExtensionStyleSheets::ExtensionStyleSheets(Document& document)
     : m_document(document)
 {
