diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationObserver.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationObserver.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationObserver.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationObserver.cpp
@@ -88,11 +88,11 @@
     if (init.attributeOldValue.valueOr(false))
         options |= AttributeOldValue;
     if (init.characterDataOldValue.valueOr(false))
         options |= CharacterDataOldValue;
 
-    HashSet<AtomicString> attributeFilter;
+    HashSet<AtomString> attributeFilter;
     if (init.attributeFilter) {
         for (auto& value : init.attributeFilter.value())
             attributeFilter.add(value);
         options |= AttributeFilter;
     }
@@ -173,11 +173,11 @@
 static void queueMutationObserverCompoundMicrotask()
 {
     if (mutationObserverCompoundMicrotaskQueuedFlag)
         return;
     mutationObserverCompoundMicrotaskQueuedFlag = true;
-    MicrotaskQueue::mainThreadQueue().append(std::make_unique<MutationObserverMicrotask>());
+    MicrotaskQueue::mainThreadQueue().append(makeUnique<MutationObserverMicrotask>());
 }
 
 void MutationObserver::enqueueMutationRecord(Ref<MutationRecord>&& mutation)
 {
     ASSERT(isMainThread());
