diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationEvent.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationEvent.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationEvent.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/MutationEvent.cpp
@@ -23,19 +23,19 @@
 #include "config.h"
 #include "MutationEvent.h"
 
 namespace WebCore {
 
-MutationEvent::MutationEvent(const AtomicString& type, CanBubble canBubble, IsCancelable cancelable, Node* relatedNode, const String& prevValue, const String& newValue)
+MutationEvent::MutationEvent(const AtomString& type, CanBubble canBubble, IsCancelable cancelable, Node* relatedNode, const String& prevValue, const String& newValue)
     : Event(type, canBubble, cancelable)
     , m_relatedNode(relatedNode)
     , m_prevValue(prevValue)
     , m_newValue(newValue)
 {
 }
 
-void MutationEvent::initMutationEvent(const AtomicString& type, bool canBubble, bool cancelable, Node* relatedNode, const String& prevValue, const String& newValue, const String& attrName, unsigned short attrChange)
+void MutationEvent::initMutationEvent(const AtomString& type, bool canBubble, bool cancelable, Node* relatedNode, const String& prevValue, const String& newValue, const String& attrName, unsigned short attrChange)
 {
     if (isBeingDispatched())
         return;
 
     initEvent(type, canBubble, cancelable);
