diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/NodeIterator.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/NodeIterator.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/NodeIterator.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/NodeIterator.cpp
@@ -25,13 +25,16 @@
 #include "config.h"
 #include "NodeIterator.h"
 
 #include "Document.h"
 #include "NodeTraversal.h"
+#include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
+WTF_MAKE_ISO_ALLOCATED_IMPL(NodeIterator);
+
 inline NodeIterator::NodePointer::NodePointer(Node& node, bool isPointerBeforeNode)
     : node(&node)
     , isPointerBeforeNode(isPointerBeforeNode)
 {
 }
@@ -110,11 +113,11 @@
             break;
         }
     }
 
     m_candidateNode.clear();
-    return WTFMove(result);
+    return result;
 }
 
 ExceptionOr<RefPtr<Node>> NodeIterator::previousNode()
 {
     RefPtr<Node> result;
@@ -139,11 +142,11 @@
             break;
         }
     }
 
     m_candidateNode.clear();
-    return WTFMove(result);
+    return result;
 }
 
 void NodeIterator::nodeWillBeRemoved(Node& removedNode)
 {
     updateForNodeRemoval(removedNode, m_candidateNode);
