diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/PointerEvent.idl b/modules/javafx.web/src/main/native/Source/WebCore/dom/PointerEvent.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/PointerEvent.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/PointerEvent.idl
@@ -37,10 +37,11 @@
 };
 
 [
     Conditional=POINTER_EVENTS,
     EnabledAtRuntime=PointerEvents,
+    DisabledByQuirk=shouldDisablePointerEvents,
     Constructor(DOMString type, optional PointerEventInit eventInitDict),
     Exposed=Window
 ] interface PointerEvent : MouseEvent {
     readonly attribute long pointerId;
     readonly attribute double width;
