diff a/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java b/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java
--- a/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java
+++ b/modules/javafx.base/src/main/java/com/sun/javafx/PlatformUtil.java
@@ -236,12 +236,11 @@
             String s = classUrlString.substring(4,
                     classUrlString.lastIndexOf('!'));
             // Strip everything after the last "/" or "\" to get rid of the jar filename
             int lastIndexOfSlash = Math.max(
                     s.lastIndexOf('/'), s.lastIndexOf('\\'));
-            return new File(new URL(s.substring(0, lastIndexOfSlash + 1)).getPath())
-                    .getParentFile();
+            return new File(new URL(s.substring(0, lastIndexOfSlash + 1)).getPath());
         } catch (MalformedURLException e) {
             return null;
         }
     }
 
