diff a/src/jdk.jdwp.agent/share/native/libjdwp/MethodImpl.c b/src/jdk.jdwp.agent/share/native/libjdwp/MethodImpl.c
--- a/src/jdk.jdwp.agent/share/native/libjdwp/MethodImpl.c
+++ b/src/jdk.jdwp.agent/share/native/libjdwp/MethodImpl.c
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 1998, 2005, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -229,12 +229,14 @@
     (void)outStream_writeBoolean(out, isObsolete);
 
     return JNI_TRUE;
 }
 
-void *Method_Cmds[] = { (void *)0x5
-    ,(void *)lineTable
-    ,(void *)variableTable
-    ,(void *)bytecodes
-    ,(void *)isObsolete
-    ,(void *)variableTableWithGenerics
+Command Method_Commands[] = {
+    {lineTable, "LineTable"},
+    {variableTable, "VariableTable"},
+    {bytecodes, "Bytecodes"},
+    {isObsolete, "IsObsolete"},
+    {variableTableWithGenerics, "VariableTableWithGenerics"}
 };
+
+DEBUG_DISPATCH_DEFINE_CMDSET(Method)
