diff a/test/langtools/tools/javac/doctree/positions/TestPositionSource.java b/test/langtools/tools/javac/doctree/positions/TestPositionSource.java
--- a/test/langtools/tools/javac/doctree/positions/TestPositionSource.java
+++ b/test/langtools/tools/javac/doctree/positions/TestPositionSource.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2010, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2010, 2019, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.
@@ -25,10 +25,12 @@
 
     /**First sentence.
      *
      * <p>Description with {@link java.io.InputStream link}
      *
+     * <em>text &lt; &#60; <!--some comment--> suffix</em>
+     *
      * @param first description
      * @param second description
      * @return whatever
      * @throws IllegalStateException why?
      * @since 1.15
@@ -52,11 +54,11 @@
      */
     public boolean erroneous(int first, int second) throws IllegalStateException {
         return true;
     }
 
-    /**First sentence.
+    /**First sentence. <em  >text <!--  some comment  --> suffix</em  >
      *
      * <p>Description with {@link    }, {@link java.util.List#add(   int   )},
      * {@link java.util.List#add(   int   ) some   text   with   whitespaces}, {@link
      *
      * @param     first
@@ -67,6 +69,24 @@
      */
     public boolean withWhiteSpaces(int first, int second) throws IllegalStateException {
         return true;
     }
 
+    /**First sentence.
+     *
+     * <p>Description with {@unknownInlineTag }, {@unknownInlineTag text}, {@unknownInlineTag
+     *
+     * @param p1 p {@unknownInlineTag text
+     * @param p2 p <
+     * @param p3 p <em
+     * @param p4 p <!--
+     * @param p5 p <!--  --
+     * @param p6 p <!--  --
+     * @param p7 p &
+     * @param p8 p &lt
+     * @param p9 p <em> </
+     * @param pa p <em> </em
+     */
+    public void erroneous2(int p1, int p2, int p3, int p4, int p5,
+                           int p6, int p7, int p8, int p9, int pa) {
+    }
 }
