diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMenuList.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMenuList.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMenuList.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderMenuList.cpp
@@ -610,10 +610,11 @@
 }
 
 void RenderMenuList::popupDidHide()
 {
 #if !PLATFORM(IOS_FAMILY)
+    // PopupMenuMac::show in WebKitLegacy can call this callback even when popup had already been dismissed.
     m_popupIsVisible = false;
 #endif
 }
 
 bool RenderMenuList::itemIsSeparator(unsigned listIndex) const
