diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTheme.cpp b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTheme.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTheme.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderTheme.cpp
@@ -1210,27 +1210,28 @@
 {
 }
 
 void RenderTheme::purgeCaches()
 {
-    m_colorCache = ColorCache();
-    m_darkColorCache = ColorCache();
+    m_colorCacheMap.clear();
 }
 
 void RenderTheme::platformColorsDidChange()
 {
-    m_colorCache = ColorCache();
-    m_darkColorCache = ColorCache();
+    m_colorCacheMap.clear();
 
     Page::updateStyleForAllPagesAfterGlobalChangeInEnvironment();
 }
 
 auto RenderTheme::colorCache(OptionSet<StyleColor::Options> options) const -> ColorCache&
 {
-    if (options.contains(StyleColor::Options::UseDarkAppearance))
-        return m_darkColorCache;
-    return m_colorCache;
+    auto optionsIgnoringVisitedLink = options;
+    optionsIgnoringVisitedLink.remove(StyleColor::Options::ForVisitedLink);
+
+    return m_colorCacheMap.ensure(optionsIgnoringVisitedLink.toRaw(), [] {
+        return ColorCache();
+    }).iterator->value;
 }
 
 FontCascadeDescription& RenderTheme::cachedSystemFontDescription(CSSValueID systemFontID) const
 {
     static NeverDestroyed<FontCascadeDescription> caption;
