diff a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderLayerModelObject.h b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderLayerModelObject.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderLayerModelObject.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/rendering/RenderLayerModelObject.h
@@ -24,10 +24,11 @@
 
 #include "RenderElement.h"
 
 namespace WebCore {
 
+class KeyframeList;
 class RenderLayer;
 
 struct RepaintLayoutRects {
     LayoutRect m_repaintRect; // This rect is clipped by enclosing objects (e.g., overflow:hidden).
     LayoutRect m_outlineBox; // This rect is unclipped.
@@ -66,10 +67,21 @@
 
     bool hasRepaintLayoutRects() const;
     void setRepaintLayoutRects(const RepaintLayoutRects&);
     void clearRepaintLayoutRects();
 
+    bool startTransition(double timeOffset, CSSPropertyID, const RenderStyle* fromStyle, const RenderStyle* toStyle) override;
+    void transitionPaused(double timeOffset, CSSPropertyID) override;
+    void transitionFinished(CSSPropertyID) override;
+
+    bool startAnimation(double timeOffset, const Animation&, const KeyframeList&) override;
+    void animationPaused(double timeOffset, const String& name) override;
+    void animationSeeked(double timeOffset, const String& name) override;
+    void animationFinished(const String& name) override;
+
+    void suspendAnimations(MonotonicTime = MonotonicTime()) override;
+
 protected:
     RenderLayerModelObject(Element&, RenderStyle&&, BaseTypeFlags);
     RenderLayerModelObject(Document&, RenderStyle&&, BaseTypeFlags);
 
     void createLayer();
