diff a/src/java.base/share/classes/sun/security/util/DerEncoder.java b/src/java.base/share/classes/sun/security/util/DerEncoder.java
--- a/src/java.base/share/classes/sun/security/util/DerEncoder.java
+++ b/src/java.base/share/classes/sun/security/util/DerEncoder.java
@@ -24,11 +24,10 @@
  */
 
 package sun.security.util;
 
 import java.io.IOException;
-import java.io.OutputStream;
 
 /**
  * Interface to an object that knows how to write its own DER
  * encoding to an output stream.
  *
@@ -39,9 +38,8 @@
     /**
      * DER encode this object and write the results to a stream.
      *
      * @param out  the stream on which the DER encoding is written.
      */
-    public void derEncode(OutputStream out)
-        throws IOException;
+    void derEncode(DerOutputStream out);
 
 }
