diff a/src/java.base/share/classes/sun/security/util/NamedCurve.java b/src/java.base/share/classes/sun/security/util/NamedCurve.java
--- a/src/java.base/share/classes/sun/security/util/NamedCurve.java
+++ b/src/java.base/share/classes/sun/security/util/NamedCurve.java
@@ -57,15 +57,11 @@
         System.arraycopy(aliases, 0, nameAndAliases, 1, aliases.length);
 
         this.oid = ko.value();
 
         DerOutputStream out = new DerOutputStream();
-        try {
-            out.putOID(ObjectIdentifier.of(ko));
-        } catch (IOException e) {
-            throw new RuntimeException("Internal error", e);
-        }
+        out.putOID(ObjectIdentifier.of(ko));
         encoded = out.toByteArray();
     }
 
     // returns the curve's standard name followed by its aliases
     public String[] getNameAndAliases() {
