diff a/modules/javafx.controls/src/shims/java/javafx/scene/control/skin/MenuBarSkinShim.java b/modules/javafx.controls/src/shims/java/javafx/scene/control/skin/MenuBarSkinShim.java
--- a/modules/javafx.controls/src/shims/java/javafx/scene/control/skin/MenuBarSkinShim.java
+++ b/modules/javafx.controls/src/shims/java/javafx/scene/control/skin/MenuBarSkinShim.java
@@ -54,6 +54,11 @@
     }
 
     public static int getFocusedIndex(MenuBarSkin skin) {
         return skin.getFocusedMenuIndex();
     }
+
+    public static void setFocusedIndex(MenuBarSkin skin, int index) {
+        skin.setFocusedIndex(index);
+    }
+
 }
