diff a/src/hotspot/share/runtime/rtmLocking.hpp b/src/hotspot/share/runtime/rtmLocking.hpp
--- a/src/hotspot/share/runtime/rtmLocking.hpp
+++ b/src/hotspot/share/runtime/rtmLocking.hpp
@@ -80,10 +80,11 @@
   //     3     Set if an internal buffer overflowed.
   //     4     Set if a debug breakpoint was hit.
   //     5     Set if an abort occurred during execution of a nested transaction.
  private:
   uintx _abortX_count[ABORT_STATUS_LIMIT];
+  static const char* _abortX_desc[ABORT_STATUS_LIMIT];
 
  public:
   static uintx _calculation_flag;
   static uintx* rtm_calculation_flag_addr() { return &_calculation_flag; }
 
@@ -104,10 +105,10 @@
   static int abortX_count_offset()        { return (int)offset_of(RTMLockingCounters, _abortX_count[0]); }
 
 
   bool nonzero() {  return (_abort_count + _total_count) > 0; }
 
-  void print_on(outputStream* st);
-  void print() { print_on(tty); }
+  void print_on(outputStream* st) const;
+  void print() const;
 };
 
 #endif // SHARE_RUNTIME_RTMLOCKING_HPP
