diff a/src/hotspot/share/interpreter/interpreterRuntime.cpp b/src/hotspot/share/interpreter/interpreterRuntime.cpp
--- a/src/hotspot/share/interpreter/interpreterRuntime.cpp
+++ b/src/hotspot/share/interpreter/interpreterRuntime.cpp
@@ -1509,11 +1509,11 @@
     default:  ShouldNotReachHere(); return;
   }
 
   // Both Q-signatures and L-signatures are mapped to atos
   if (cp_entry->flag_state() == atos && ik->field_signature(index)->is_Q_signature()) {
-    sig_type = JVM_SIGNATURE_VALUETYPE;
+    sig_type = JVM_SIGNATURE_INLINETYPE;
   }
 
   bool is_static = (obj == NULL);
   bool is_inlined = cp_entry->is_inlined();
 
