diff a/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java b/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
--- a/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
+++ b/webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java
@@ -27,10 +27,11 @@
 import java.io.*;
 import java.net.URISyntaxException;
 import java.nio.channels.FileChannel;
 import java.nio.file.*;
 import java.util.*;
+import java.util.function.Function;
 
 import static java.nio.file.StandardOpenOption.*;
 
 public class Webrev {
     private static final String ANCNAV_HTML = "navigation.html";
@@ -58,10 +59,12 @@
         private String username;
         private String upstream;
         private String pullRequest;
         private String branch;
         private String issue;
+        private Function<String, String> issueLinker;
+        private Function<String, String> commitLinker;
         private String version;
         private List<Path> files = List.of();
 
         Builder(ReadOnlyRepository repository, Path output) {
             this.repository = repository;
@@ -96,10 +99,20 @@
         public Builder issue(String issue) {
             this.issue = issue;
             return this;
         }
 
+        public Builder issueLinker(Function<String, String> issueLinker) {
+            this.issueLinker = issueLinker;
+            return this;
+        }
+
+        public Builder commitLinker(Function<String, String> commitLinker) {
+            this.commitLinker = commitLinker;
+            return this;
+        }
+
         public Builder version(String version) {
             this.version = version;
             return this;
         }
 
@@ -172,11 +185,11 @@
                 navigations.addLast(new Navigation(prev, next));
             }
 
             var headHash = head == null ? repository.head() : head;
             var fileViews = new ArrayList<FileView>();
-            var formatter = new MetadataFormatter();
+            var formatter = new MetadataFormatter(issueLinker);
             for (var patch : patches) {
                 var status = patch.status();
                 var path = status.isDeleted() ?
                     patch.source().path().get() :
                     patch.target().path().get();
@@ -192,20 +205,23 @@
             }
 
             var total = fileViews.stream().map(FileView::stats).mapToInt(WebrevStats::total).sum();
             var stats = new WebrevStats(diff.added(), diff.removed(), diff.modified(), total);
 
+            var issueForWebrev = issue == null ? null : issueLinker.apply(issue);
+            var tailEndURL = commitLinker == null ? null : commitLinker.apply(tailEnd.hex());
             try (var w = Files.newBufferedWriter(output.resolve("index.html"))) {
                 var index = new IndexView(fileViews,
                                           title,
                                           username,
                                           upstream,
                                           branch,
                                           pullRequest,
-                                          issue,
+                                          issueForWebrev,
                                           version,
                                           tailEnd,
+                                          tailEndURL,
                                           output.relativize(patchFile),
                                           stats);
                 index.render(w);
 
             }
