diff a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IndexValueEntry.h b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IndexValueEntry.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IndexValueEntry.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/Modules/indexeddb/server/IndexValueEntry.h
@@ -38,11 +38,11 @@
 namespace IDBServer {
 
 class IndexValueEntry {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    IndexValueEntry(bool unique);
+    explicit IndexValueEntry(bool unique);
     ~IndexValueEntry();
 
     void addKey(const IDBKeyData&);
 
     // Returns true if a key was actually removed.
