diff a/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java b/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java
--- a/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java
+++ b/test/hotspot/jtreg/gc/nvdimm/TestAllocateOldGenAt.java
@@ -28,54 +28,40 @@
  * @summary Test to check allocation of Java Heap with AllocateOldGenAt option
  * @requires vm.gc=="null" & os.family != "aix"
  * @requires test.vm.gc.nvdimm
  * @library /test/lib
  * @modules java.base/jdk.internal.misc
- * @run main gc.nvdimm.TestAllocateOldGenAt
+ * @run driver gc.nvdimm.TestAllocateOldGenAt
  */
 
 import jdk.test.lib.JDKToolFinder;
 import jdk.test.lib.process.ProcessTools;
 import jdk.test.lib.process.OutputAnalyzer;
 import java.util.ArrayList;
 import java.util.Collections;
 
 public class TestAllocateOldGenAt {
-  private static ArrayList<String> commonOpts;
+  private static String[] commonFlags;
 
   public static void main(String args[]) throws Exception {
-    commonOpts = new ArrayList<>();
-
-    String testVmOptsStr = System.getProperty("test.java.opts");
-    if (!testVmOptsStr.isEmpty()) {
-      String[] testVmOpts = testVmOptsStr.split(" ");
-      Collections.addAll(commonOpts, testVmOpts);
-    }
     String test_dir = System.getProperty("test.dir", ".");
-    Collections.addAll(commonOpts, new String[] {"-XX:+UnlockExperimentalVMOptions",
-                                                 "-XX:AllocateOldGenAt=" + test_dir,
-                                                 "-Xmx32m",
-                                                 "-Xms32m",
-                                                 "-version"});
+    commonFlags = new String[] {
+        "-XX:+UnlockExperimentalVMOptions",
+        "-XX:AllocateOldGenAt=" + test_dir,
+        "-Xmx32m",
+        "-Xms32m",
+        "-version"};
 
     runTest("-XX:+UseG1GC");
     runTest("-XX:+UseParallelGC");
   }
 
   private static void runTest(String... extraFlags) throws Exception {
-    ArrayList<String> testOpts = new ArrayList<>();
-    Collections.addAll(testOpts, commonOpts.toArray(new String[commonOpts.size()]));
-    Collections.addAll(testOpts, extraFlags);
-
-    System.out.print("Testing:\n" + JDKToolFinder.getJDKTool("java"));
-    for (int i = 0; i < testOpts.size(); i += 1) {
-      System.out.print(" " + testOpts.get(i));
-    }
-    System.out.println();
-
-    ProcessBuilder pb =
-      ProcessTools.createJavaProcessBuilder(testOpts.toArray(new String[testOpts.size()]));
+    ArrayList<String> flags = new ArrayList<>();
+    Collections.addAll(flags, commonFlags);
+    Collections.addAll(flags, extraFlags);
+    ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(true, flags);
     OutputAnalyzer output = new OutputAnalyzer(pb.start());
 
     output.shouldHaveExitValue(0);
 
   }
