diff a/modules/javafx.web/src/main/native/Source/WebCore/page/PerformanceNavigation.cpp b/modules/javafx.web/src/main/native/Source/WebCore/page/PerformanceNavigation.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/PerformanceNavigation.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/PerformanceNavigation.cpp
@@ -29,10 +29,11 @@
  */
 
 #include "config.h"
 #include "PerformanceNavigation.h"
 
+#include "CustomHeaderFields.h"
 #include "DocumentLoader.h"
 #include "Frame.h"
 #include "FrameLoader.h"
 #include "FrameLoaderTypes.h"
 
