diff a/modules/javafx.web/src/main/native/Source/WebCore/page/RuntimeEnabledFeatures.cpp b/modules/javafx.web/src/main/native/Source/WebCore/page/RuntimeEnabledFeatures.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/RuntimeEnabledFeatures.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/RuntimeEnabledFeatures.cpp
@@ -31,10 +31,11 @@
 
 #include "config.h"
 #include "RuntimeEnabledFeatures.h"
 
 #include "MediaPlayer.h"
+#include "PlatformScreen.h"
 #include <JavaScriptCore/Options.h>
 #include <wtf/NeverDestroyed.h>
 
 namespace WebCore {
 
@@ -61,6 +62,13 @@
 bool RuntimeEnabledFeatures::spectreGadgetsEnabled() const
 {
     return JSC::Options::enableSpectreGadgets();
 }
 
+#if ENABLE(TOUCH_EVENTS)
+bool RuntimeEnabledFeatures::touchEventsEnabled() const
+{
+    return m_touchEventsEnabled.valueOr(screenHasTouchDevice());
+}
+#endif
+
 } // namespace WebCore
