diff a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanBooleanPropertyBuilder.java b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanBooleanPropertyBuilder.java
--- a/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanBooleanPropertyBuilder.java
+++ b/modules/javafx.base/src/main/java/javafx/beans/property/adapter/ReadOnlyJavaBeanBooleanPropertyBuilder.java
@@ -58,15 +58,11 @@
  */
 public final class ReadOnlyJavaBeanBooleanPropertyBuilder {
 
     private final ReadOnlyJavaBeanPropertyBuilderHelper helper = new ReadOnlyJavaBeanPropertyBuilderHelper();
 
-    /**
-     * @deprecated This constructor was exposed erroneously and will be removed in the next version. Use {@link #create()} instead.
-     */
-    @Deprecated(since="14", forRemoval=true)
-    public ReadOnlyJavaBeanBooleanPropertyBuilder() {}
+    private ReadOnlyJavaBeanBooleanPropertyBuilder() {}
 
     /**
      * Create a new instance of {@code ReadOnlyJavaBeanBooleanPropertyBuilder}
      *
      * @return the new {@code ReadOnlyJavaBeanBooleanPropertyBuilder}
