diff a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ExecutableToCodeBlockEdge.cpp b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ExecutableToCodeBlockEdge.cpp
--- a/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ExecutableToCodeBlockEdge.cpp
+++ b/modules/javafx.web/src/main/native/Source/JavaScriptCore/bytecode/ExecutableToCodeBlockEdge.cpp
@@ -26,10 +26,12 @@
 #include "config.h"
 #include "ExecutableToCodeBlockEdge.h"
 
 #include "CodeBlock.h"
 #include "IsoCellSetInlines.h"
+#include "JSObjectInlines.h"
+#include "StructureInlines.h"
 
 namespace JSC {
 
 const ClassInfo ExecutableToCodeBlockEdge::s_info = { "ExecutableToCodeBlockEdge", nullptr, nullptr, nullptr, CREATE_METHOD_TABLE(ExecutableToCodeBlockEdge) };
 
