diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortSignal.idl b/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortSignal.idl
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortSignal.idl
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/AbortSignal.idl
@@ -22,10 +22,11 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 [
+    CustomIsReachable,
     Exposed=(Window,Worker)
 ] interface AbortSignal : EventTarget {
     readonly attribute boolean aborted;
 
     attribute EventHandler onabort;
