diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/DeviceOrientationOrMotionEvent.h
@@ -24,16 +24,18 @@
  */
 
 #pragma once
 
 #include "DeviceOrientationOrMotionPermissionState.h"
-#include "JSDOMPromiseDeferred.h"
+#include "IDLTypes.h"
 
 namespace WebCore {
 
 class Document;
 
+template<typename IDLType> class DOMPromiseDeferred;
+
 class DeviceOrientationOrMotionEvent {
 public:
 
 #if ENABLE(DEVICE_ORIENTATION)
     using PermissionState = DeviceOrientationOrMotionPermissionState;
