diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/Event.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/Event.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/Event.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/Event.h
@@ -37,10 +37,11 @@
 class EventPath;
 class EventTarget;
 class ScriptExecutionContext;
 
 class Event : public ScriptWrappable, public RefCounted<Event> {
+    WTF_MAKE_ISO_ALLOCATED(Event);
 public:
     enum class IsTrusted : uint8_t { No, Yes };
     enum class CanBubble : uint8_t { No, Yes };
     enum class IsCancelable : uint8_t { No, Yes };
     enum class IsComposed : uint8_t { No, Yes };
