diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/QualifiedName.cpp b/modules/javafx.web/src/main/native/Source/WebCore/dom/QualifiedName.cpp
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/QualifiedName.cpp
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/QualifiedName.cpp
@@ -24,10 +24,13 @@
 #include "ThreadGlobalData.h"
 #include <wtf/Assertions.h>
 
 namespace WebCore {
 
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(QualifiedName);
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(QualifiedNameQualifiedNameImpl);
+
 QualifiedName::QualifiedName(const AtomString& p, const AtomString& l, const AtomString& n)
     : m_impl(threadGlobalData().qualifiedNameCache().getOrCreate(QualifiedNameComponents { p.impl(), l.impl(), n.isEmpty() ? nullptr : n.impl() }))
 {
 }
 
