diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEventWithKeyState.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEventWithKeyState.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEventWithKeyState.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEventWithKeyState.h
@@ -28,10 +28,11 @@
 #include "UIEvent.h"
 
 namespace WebCore {
 
 class UIEventWithKeyState : public UIEvent {
+    WTF_MAKE_ISO_ALLOCATED(UIEventWithKeyState);
 public:
     using Modifier = PlatformEvent::Modifier;
 
     bool ctrlKey() const { return m_modifiers.contains(Modifier::ControlKey); }
     bool shiftKey() const { return m_modifiers.contains(Modifier::ShiftKey); }
