diff a/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEvent.h b/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEvent.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEvent.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/dom/UIEvent.h
@@ -31,10 +31,11 @@
 
 // FIXME: Remove this when no one is depending on it anymore.
 typedef WindowProxy AbstractView;
 
 class UIEvent : public Event {
+    WTF_MAKE_ISO_ALLOCATED(UIEvent);
 public:
     static Ref<UIEvent> create(const AtomString& type, CanBubble canBubble, IsCancelable isCancelable, IsComposed isComposed, RefPtr<WindowProxy>&& view, int detail)
     {
         return adoptRef(*new UIEvent(type, canBubble, isCancelable, isComposed, WTFMove(view), detail));
     }
