diff a/src/hotspot/share/opto/phaseX.cpp b/src/hotspot/share/opto/phaseX.cpp
--- a/src/hotspot/share/opto/phaseX.cpp
+++ b/src/hotspot/share/opto/phaseX.cpp
@@ -1604,10 +1604,19 @@
           }
         }
       }
     }
 
+    // Inline type nodes can have other inline types as users. If an input gets
+    // updated, make sure that inline type users get a chance for optimization.
+    if (use->is_InlineTypeBase()) {
+      for (DUIterator_Fast i2max, i2 = use->fast_outs(i2max); i2 < i2max; i2++) {
+        Node* u = use->fast_out(i2);
+        if (u->is_InlineTypeBase())
+          _worklist.push(u);
+      }
+    }
     // If changed Cast input, check Phi users for simple cycles
     if (use->is_ConstraintCast()) {
       for (DUIterator_Fast i2max, i2 = use->fast_outs(i2max); i2 < i2max; i2++) {
         Node* u = use->fast_out(i2);
         if (u->is_Phi())
