diff a/modules/javafx.web/src/main/native/Source/WebCore/page/animation/CSSPropertyAnimation.h b/modules/javafx.web/src/main/native/Source/WebCore/page/animation/CSSPropertyAnimation.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/page/animation/CSSPropertyAnimation.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/page/animation/CSSPropertyAnimation.h
@@ -46,9 +46,9 @@
     static int getNumProperties();
 
     static HashSet<CSSPropertyID> animatableShorthandsAffectingProperty(CSSPropertyID);
 
     // Return true if we need to start software animation timers
-    static bool blendProperties(const CSSPropertyBlendingClient*, CSSPropertyID, RenderStyle* dst, const RenderStyle* a, const RenderStyle* b, double progress);
+    static void blendProperties(const CSSPropertyBlendingClient*, CSSPropertyID, RenderStyle* dst, const RenderStyle* a, const RenderStyle* b, double progress);
 };
 
 } // namespace WebCore
