diff a/test/jdk/javax/net/ssl/compatibility/UseCase.java b/test/jdk/javax/net/ssl/compatibility/UseCase.java
--- a/test/jdk/javax/net/ssl/compatibility/UseCase.java
+++ b/test/jdk/javax/net/ssl/compatibility/UseCase.java
@@ -28,14 +28,14 @@
  * The TLS communication use case.
  */
 public class UseCase {
 
     private static final boolean FULL_CASES
-            = Utils.getBoolProperty("fullCases");
+            = Boolean.getBoolean("fullCases");
 
     public static final boolean FULL_CIPHER_SUITES
-            = Utils.getBoolProperty("fullCipherSuites");
+            = Boolean.getBoolean("fullCipherSuites");
 
     public static final Protocol[] PROTOCOLS = new Protocol[] {
             Protocol.TLSV1,
             Protocol.TLSV1_1,
             Protocol.TLSV1_2,
@@ -127,12 +127,12 @@
             this.negoAppProtocol = negoAppProtocol;
         }
     }
 
     private static final Object[][] PARAMS = new Object[][] {
-            FULL_CASES ? PROTOCOLS : PROTOCOLS,
-            FULL_CASES ? CIPHER_SUITES : MANDATORY_CIPHER_SUITES,
+            PROTOCOLS,
+            FULL_CASES || FULL_CIPHER_SUITES ? CIPHER_SUITES : MANDATORY_CIPHER_SUITES,
             FULL_CASES ? new Boolean[] { false, true } : new Boolean[] { true },
             FULL_CASES
                     ? new ServerName[] { ServerName.NONE, ServerName.EXAMPLE }
                     : new ServerName[] { ServerName.EXAMPLE },
             FULL_CASES
