diff a/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/ContentExtensionActions.h b/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/ContentExtensionActions.h
--- a/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/ContentExtensionActions.h
+++ b/modules/javafx.web/src/main/native/Source/WebCore/contentextensions/ContentExtensionActions.h
@@ -30,16 +30,18 @@
 #include <wtf/HashSet.h>
 #include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
+struct ContentRuleListResults;
 class Page;
 class ResourceRequest;
 
 namespace ContentExtensions {
 
-typedef uint8_t SerializedActionByte;
+struct Action;
+using SerializedActionByte = uint8_t;
 
 enum class ActionType : uint8_t {
     BlockLoad,
     BlockCookies,
     CSSDisplayNoneSelector,
@@ -62,18 +64,17 @@
     }
     ASSERT_NOT_REACHED();
     return false;
 }
 
-struct BlockedStatus {
-    bool blockedLoad { false };
-    bool blockedCookies { false };
-    bool madeHTTPS { false };
-    HashSet<std::pair<String, String>> notifications;
+struct ActionsFromContentRuleList {
+    String contentRuleListIdentifier;
+    bool sawIgnorePreviousRules { false };
+    Vector<Action> actions;
 };
 
-WEBCORE_EXPORT void applyBlockedStatusToRequest(const BlockedStatus&, Page*, ResourceRequest&);
+WEBCORE_EXPORT void applyResultsToRequest(ContentRuleListResults&&, Page*, ResourceRequest&);
 
 } // namespace ContentExtensions
 
 } // namespace WebCore
 
