diff a/modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCBufferedContext.java b/modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCBufferedContext.java
--- a/modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCBufferedContext.java
+++ b/modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCBufferedContext.java
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2011, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2011, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License version 2 only, as
  * published by the Free Software Foundation.  Oracle designates this
@@ -202,10 +202,15 @@
     @Override public void saveState() {
         init();
         super.saveState();
     }
 
+    @Override public void scale(float sx, float sy) {
+        init();
+        super.scale(sx, sy);
+    }
+
     @Override public void setTransform(WCTransform tm) {
         init();
         super.setTransform(tm);
     }
 
