diff a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java
--- a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java
+++ b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java
@@ -207,12 +207,12 @@
     }
 
     public void addClassesSummary(SortedSet<TypeElement> classes, String label,
             TableHeader tableHeader, Content summaryContentTree) {
         if(!classes.isEmpty()) {
-            Table table = new Table(HtmlStyle.typeSummary)
-                    .setCaption(getTableCaption(new StringContent(label)))
+            Table table = new Table(HtmlStyle.typeSummary, HtmlStyle.summaryTable)
+                    .setCaption(new StringContent(label))
                     .setHeader(tableHeader)
                     .setColumnStyles(HtmlStyle.colFirst, HtmlStyle.colLast);
 
             for (TypeElement klass : classes) {
                 if (!utils.isCoreClass(klass) || !configuration.isGeneratedDoc(klass)) {
